// vscode/src/vs/workbench/parts/extensions/electron-browser/extensions.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var instantiation_1 = require('vs/platform/instantiation/common/instantiation');
exports.VIEWLET_ID = 'workbench.view.extensions';
(function (ExtensionState) {
    ExtensionState[ExtensionState["Installing"] = 0] = "Installing";
    ExtensionState[ExtensionState["Installed"] = 1] = "Installed";
    ExtensionState[ExtensionState["NeedsRestart"] = 2] = "NeedsRestart";
    ExtensionState[ExtensionState["Uninstalled"] = 3] = "Uninstalled";
})(exports.ExtensionState || (exports.ExtensionState = {}));
var ExtensionState = exports.ExtensionState;
exports.SERVICE_ID = 'extensionsWorkbenchService';
exports.IExtensionsWorkbenchService = instantiation_1.createDecorator(exports.SERVICE_ID);
exports.ConfigurationKey = 'extensions';
