// vscode/src/vs/workbench/api/node/mainThreadEditors.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var lifecycle_1 = require('vs/base/common/lifecycle');
var winjs_base_1 = require('vs/base/common/winjs.base');
var threadService_1 = require('vs/workbench/services/thread/common/threadService');
var editorCommon_1 = require('vs/editor/common/editorCommon');
var codeEditorService_1 = require('vs/editor/common/services/codeEditorService');
var editorService_1 = require('vs/workbench/services/editor/common/editorService');
var groupService_1 = require('vs/workbench/services/group/common/groupService');
var modelService_1 = require('vs/editor/common/services/modelService');
var mainThreadEditorsTracker_1 = require('vs/workbench/api/node/mainThreadEditorsTracker');
var telemetry_1 = require('vs/platform/telemetry/common/telemetry');
var event_1 = require('vs/platform/event/common/event');
var arrays_1 = require('vs/base/common/arrays');
var objects_1 = require('vs/base/common/objects');
var extHost_protocol_1 = require('./extHost.protocol');
var MainThreadEditors = (function (_super) {
    __extends(MainThreadEditors, _super);
    function MainThreadEditors(threadService, workbenchEditorService, editorGroupService, telemetryService, editorService, eventService, modelService) {
        var _this = this;
        _super.call(this);
        this._proxy = threadService.get(extHost_protocol_1.ExtHostContext.ExtHostEditors);
        this._workbenchEditorService = workbenchEditorService;
        this._telemetryService = telemetryService;
        this._toDispose = [];
        this._textEditorsListenersMap = Object.create(null);
        this._textEditorsMap = Object.create(null);
        this._activeTextEditor = null;
        this._visibleEditors = [];
        this._editorPositionData = null;
        this._editorTracker = new mainThreadEditorsTracker_1.MainThreadEditorsTracker(editorService, modelService);
        this._toDispose.push(this._editorTracker);
        this._toDispose.push(this._editorTracker.onTextEditorAdd(function (textEditor) { return _this._onTextEditorAdd(textEditor); }));
        this._toDispose.push(this._editorTracker.onTextEditorRemove(function (textEditor) { return _this._onTextEditorRemove(textEditor); }));
        this._toDispose.push(this._editorTracker.onDidUpdateTextEditors(function () { return _this._updateActiveAndVisibleTextEditors(); }));
        this._toDispose.push(this._editorTracker.onChangedFocusedTextEditor(function (focusedTextEditorId) { return _this._updateActiveAndVisibleTextEditors(); }));
        this._toDispose.push(editorGroupService.onEditorsChanged(function () { return _this._updateActiveAndVisibleTextEditors(); }));
        this._toDispose.push(editorGroupService.onEditorsMoved(function () { return _this._updateActiveAndVisibleTextEditors(); }));
    }
    MainThreadEditors.prototype.dispose = function () {
        var _this = this;
        Object.keys(this._textEditorsListenersMap).forEach(function (editorId) {
            lifecycle_1.dispose(_this._textEditorsListenersMap[editorId]);
        });
        this._textEditorsListenersMap = Object.create(null);
        this._toDispose = lifecycle_1.dispose(this._toDispose);
    };
    MainThreadEditors.prototype._onTextEditorAdd = function (textEditor) {
        var _this = this;
        var id = textEditor.getId();
        var toDispose = [];
        toDispose.push(textEditor.onConfigurationChanged(function (opts) {
            _this._proxy.$acceptOptionsChanged(id, opts);
        }));
        toDispose.push(textEditor.onSelectionChanged(function (event) {
            _this._proxy.$acceptSelectionsChanged(id, event);
        }));
        this._proxy.$acceptTextEditorAdd({
            id: id,
            document: textEditor.getModel().uri,
            options: textEditor.getConfiguration(),
            selections: textEditor.getSelections(),
            editorPosition: this._findEditorPosition(textEditor)
        });
        this._textEditorsListenersMap[id] = toDispose;
        this._textEditorsMap[id] = textEditor;
    };
    MainThreadEditors.prototype._onTextEditorRemove = function (textEditor) {
        var id = textEditor.getId();
        lifecycle_1.dispose(this._textEditorsListenersMap[id]);
        delete this._textEditorsListenersMap[id];
        delete this._textEditorsMap[id];
        this._proxy.$acceptTextEditorRemove(id);
    };
    MainThreadEditors.prototype._updateActiveAndVisibleTextEditors = function () {
        // active and visible editors
        var visibleEditors = this._editorTracker.getVisibleTextEditorIds();
        var activeEditor = this._findActiveTextEditorId();
        if (activeEditor !== this._activeTextEditor || !arrays_1.equals(this._visibleEditors, visibleEditors, function (a, b) { return a === b; })) {
            this._activeTextEditor = activeEditor;
            this._visibleEditors = visibleEditors;
            this._proxy.$acceptActiveEditorAndVisibleEditors(this._activeTextEditor, this._visibleEditors);
        }
        // editor columns
        var editorPositionData = this._getTextEditorPositionData();
        if (!objects_1.equals(this._editorPositionData, editorPositionData)) {
            this._editorPositionData = editorPositionData;
            this._proxy.$acceptEditorPositionData(this._editorPositionData);
        }
    };
    MainThreadEditors.prototype._findActiveTextEditorId = function () {
        var focusedTextEditorId = this._editorTracker.getFocusedTextEditorId();
        if (focusedTextEditorId) {
            return focusedTextEditorId;
        }
        var activeEditor = this._workbenchEditorService.getActiveEditor();
        if (!activeEditor) {
            return null;
        }
        var editor = activeEditor.getControl();
        // Substitute for (editor instanceof ICodeEditor)
        if (!editor || typeof editor.getEditorType !== 'function') {
            // Not a text editor...
            return null;
        }
        if (editor.getEditorType() === editorCommon_1.EditorType.ICodeEditor) {
            return this._editorTracker.findTextEditorIdFor(editor);
        }
        // Must be a diff editor => use the modified side
        return this._editorTracker.findTextEditorIdFor(editor.getModifiedEditor());
    };
    MainThreadEditors.prototype._findEditorPosition = function (editor) {
        for (var _i = 0, _a = this._workbenchEditorService.getVisibleEditors(); _i < _a.length; _i++) {
            var workbenchEditor = _a[_i];
            if (editor.matches(workbenchEditor)) {
                return workbenchEditor.position;
            }
        }
    };
    MainThreadEditors.prototype._getTextEditorPositionData = function () {
        var result = Object.create(null);
        for (var _i = 0, _a = this._workbenchEditorService.getVisibleEditors(); _i < _a.length; _i++) {
            var workbenchEditor = _a[_i];
            var editor = workbenchEditor.getControl();
            // Substitute for (editor instanceof ICodeEditor)
            if (!editor || typeof editor.getEditorType !== 'function') {
                // Not a text editor...
                continue;
            }
            if (editor.getEditorType() === editorCommon_1.EditorType.ICodeEditor) {
                var id = this._editorTracker.findTextEditorIdFor(editor);
                if (id) {
                    result[id] = workbenchEditor.position;
                }
            }
        }
        return result;
    };
    // --- from extension host process
    MainThreadEditors.prototype.$tryShowTextDocument = function (resource, position, preserveFocus) {
        var _this = this;
        var input = {
            resource: resource,
            options: { preserveFocus: preserveFocus, pinned: true }
        };
        return this._workbenchEditorService.openEditor(input, position).then(function (editor) {
            if (!editor) {
                return;
            }
            return new winjs_base_1.TPromise(function (c) {
                // not very nice but the way it is: changes to the editor state aren't
                // send to the ext host as they happen but stuff is delayed a little. in
                // order to provide the real editor on #openTextEditor we need to sync on
                // that update
                var subscription;
                var handle;
                function contd() {
                    subscription.dispose();
                    clearTimeout(handle);
                    c(undefined);
                }
                subscription = _this._editorTracker.onDidUpdateTextEditors(function () {
                    contd();
                });
                handle = setTimeout(function () {
                    contd();
                }, 1000);
            }).then(function () {
                // find the editor we have just opened and return the
                // id we have assigned to it.
                for (var id in _this._textEditorsMap) {
                    if (_this._textEditorsMap[id].matches(editor)) {
                        return id;
                    }
                }
            });
        });
    };
    MainThreadEditors.prototype.$tryShowEditor = function (id, position) {
        // check how often this is used
        this._telemetryService.publicLog('api.deprecated', { function: 'TextEditor.show' });
        var mainThreadEditor = this._textEditorsMap[id];
        if (mainThreadEditor) {
            var model = mainThreadEditor.getModel();
            return this._workbenchEditorService.openEditor({
                resource: model.uri,
                options: { preserveFocus: false }
            }, position).then(function () { return; });
        }
    };
    MainThreadEditors.prototype.$tryHideEditor = function (id) {
        // check how often this is used
        this._telemetryService.publicLog('api.deprecated', { function: 'TextEditor.hide' });
        var mainThreadEditor = this._textEditorsMap[id];
        if (mainThreadEditor) {
            var editors = this._workbenchEditorService.getVisibleEditors();
            for (var _i = 0, editors_1 = editors; _i < editors_1.length; _i++) {
                var editor = editors_1[_i];
                if (mainThreadEditor.matches(editor)) {
                    return this._workbenchEditorService.closeEditor(editor.position, editor.input).then(function () { return; });
                }
            }
        }
    };
    MainThreadEditors.prototype.$trySetSelections = function (id, selections) {
        if (!this._textEditorsMap[id]) {
            return winjs_base_1.TPromise.wrapError('TextEditor disposed');
        }
        this._textEditorsMap[id].setSelections(selections);
        return winjs_base_1.TPromise.as(null);
    };
    MainThreadEditors.prototype.$trySetDecorations = function (id, key, ranges) {
        if (!this._textEditorsMap[id]) {
            return winjs_base_1.TPromise.wrapError('TextEditor disposed');
        }
        this._textEditorsMap[id].setDecorations(key, ranges);
        return winjs_base_1.TPromise.as(null);
    };
    MainThreadEditors.prototype.$tryRevealRange = function (id, range, revealType) {
        if (!this._textEditorsMap[id]) {
            return winjs_base_1.TPromise.wrapError('TextEditor disposed');
        }
        this._textEditorsMap[id].revealRange(range, revealType);
    };
    MainThreadEditors.prototype.$trySetOptions = function (id, options) {
        if (!this._textEditorsMap[id]) {
            return winjs_base_1.TPromise.wrapError('TextEditor disposed');
        }
        this._textEditorsMap[id].setConfiguration(options);
        return winjs_base_1.TPromise.as(null);
    };
    MainThreadEditors.prototype.$tryApplyEdits = function (id, modelVersionId, edits, opts) {
        if (!this._textEditorsMap[id]) {
            return winjs_base_1.TPromise.wrapError('TextEditor disposed');
        }
        return winjs_base_1.TPromise.as(this._textEditorsMap[id].applyEdits(modelVersionId, edits, opts));
    };
    MainThreadEditors.prototype.$registerTextEditorDecorationType = function (key, options) {
        this._editorTracker.registerTextEditorDecorationType(key, options);
    };
    MainThreadEditors.prototype.$removeTextEditorDecorationType = function (key) {
        this._editorTracker.removeTextEditorDecorationType(key);
    };
    MainThreadEditors = __decorate([
        __param(0, threadService_1.IThreadService),
        __param(1, editorService_1.IWorkbenchEditorService),
        __param(2, groupService_1.IEditorGroupService),
        __param(3, telemetry_1.ITelemetryService),
        __param(4, codeEditorService_1.ICodeEditorService),
        __param(5, event_1.IEventService),
        __param(6, modelService_1.IModelService)
    ], MainThreadEditors);
    return MainThreadEditors;
}(extHost_protocol_1.MainThreadEditorsShape));
exports.MainThreadEditors = MainThreadEditors;
