// vscode/src/vs/editor/common/model/mirrorModel2.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var prefixSumComputer_1 = require("vs/editor/common/viewModel/prefixSumComputer");
var MirrorModel2 = (function () {
    function MirrorModel2(uri, lines, eol, versionId) {
        this._uri = uri;
        this._lines = lines;
        this._eol = eol;
        this._versionId = versionId;
    }
    MirrorModel2.prototype.dispose = function () {
        this._lines.length = 0;
    };
    Object.defineProperty(MirrorModel2.prototype, "version", {
        get: function () {
            return this._versionId;
        },
        enumerable: true,
        configurable: true
    });
    MirrorModel2.prototype.getText = function () {
        return this._lines.join(this._eol);
    };
    MirrorModel2.prototype.onEvents = function (events) {
        var newEOL = null;
        for (var i = 0, len = events.length; i < len; i++) {
            var e = events[i];
            if (e.eol) {
                newEOL = e.eol;
            }
        }
        if (newEOL && newEOL !== this._eol) {
            this._eol = newEOL;
            this._lineStarts = null;
        }
        // Update my lines
        var lastVersionId = -1;
        for (var i = 0, len = events.length; i < len; i++) {
            var e = events[i];
            this._acceptDeleteRange(e.range);
            this._acceptInsertText({
                lineNumber: e.range.startLineNumber,
                column: e.range.startColumn
            }, e.text);
            lastVersionId = Math.max(lastVersionId, e.versionId);
        }
        if (lastVersionId !== -1) {
            this._versionId = lastVersionId;
        }
    };
    MirrorModel2.prototype._ensureLineStarts = function () {
        if (!this._lineStarts) {
            var lineStartValues = [];
            var eolLength = this._eol.length;
            for (var i = 0, len = this._lines.length; i < len; i++) {
                lineStartValues.push(this._lines[i].length + eolLength);
            }
            this._lineStarts = new prefixSumComputer_1.PrefixSumComputer(lineStartValues);
        }
    };
    /**
     * All changes to a line's text go through this method
     */
    MirrorModel2.prototype._setLineText = function (lineIndex, newValue) {
        this._lines[lineIndex] = newValue;
        if (this._lineStarts) {
            // update prefix sum
            this._lineStarts.changeValue(lineIndex, this._lines[lineIndex].length + this._eol.length);
        }
    };
    MirrorModel2.prototype._acceptDeleteRange = function (range) {
        if (range.startLineNumber === range.endLineNumber) {
            if (range.startColumn === range.endColumn) {
                // Nothing to delete
                return;
            }
            // Delete text on the affected line
            this._setLineText(range.startLineNumber - 1, this._lines[range.startLineNumber - 1].substring(0, range.startColumn - 1)
                + this._lines[range.startLineNumber - 1].substring(range.endColumn - 1));
            return;
        }
        // Take remaining text on last line and append it to remaining text on first line
        this._setLineText(range.startLineNumber - 1, this._lines[range.startLineNumber - 1].substring(0, range.startColumn - 1)
            + this._lines[range.endLineNumber - 1].substring(range.endColumn - 1));
        // Delete middle lines
        this._lines.splice(range.startLineNumber, range.endLineNumber - range.startLineNumber);
        if (this._lineStarts) {
            // update prefix sum
            this._lineStarts.removeValues(range.startLineNumber, range.endLineNumber - range.startLineNumber);
        }
    };
    MirrorModel2.prototype._acceptInsertText = function (position, insertText) {
        if (insertText.length === 0) {
            // Nothing to insert
            return;
        }
        var insertLines = insertText.split(/\r\n|\r|\n/);
        if (insertLines.length === 1) {
            // Inserting text on one line
            this._setLineText(position.lineNumber - 1, this._lines[position.lineNumber - 1].substring(0, position.column - 1)
                + insertLines[0]
                + this._lines[position.lineNumber - 1].substring(position.column - 1));
            return;
        }
        // Append overflowing text from first line to the end of text to insert
        insertLines[insertLines.length - 1] += this._lines[position.lineNumber - 1].substring(position.column - 1);
        // Delete overflowing text from first line and insert text on first line
        this._setLineText(position.lineNumber - 1, this._lines[position.lineNumber - 1].substring(0, position.column - 1)
            + insertLines[0]);
        // Insert new lines & store lengths
        var newLengths = new Array(insertLines.length - 1);
        for (var i = 1; i < insertLines.length; i++) {
            this._lines.splice(position.lineNumber + i - 1, 0, insertLines[i]);
            newLengths[i - 1] = insertLines[i].length + this._eol.length;
        }
        if (this._lineStarts) {
            // update prefix sum
            this._lineStarts.insertValues(position.lineNumber, newLengths);
        }
    };
    return MirrorModel2;
}());
exports.MirrorModel2 = MirrorModel2;
