// vscode/src/vs/base/browser/ui/actionbar/actionbar.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
require('./actionbar.css');
var nls = require('vs/nls');
var lifecycle = require('vs/base/common/lifecycle');
var builder_1 = require('vs/base/browser/builder');
var platform = require('vs/base/common/platform');
var actions_1 = require('vs/base/common/actions');
var DOM = require('vs/base/browser/dom');
var events_1 = require('vs/base/common/events');
var types = require('vs/base/common/types');
var eventEmitter_1 = require('vs/base/common/eventEmitter');
var touch_1 = require('vs/base/browser/touch');
var keyboardEvent_1 = require('vs/base/browser/keyboardEvent');
var BaseActionItem = (function (_super) {
    __extends(BaseActionItem, _super);
    function BaseActionItem(context, action) {
        var _this = this;
        _super.call(this);
        this._callOnDispose = [];
        this._context = context || this;
        this._action = action;
        if (action instanceof actions_1.Action) {
            this._callOnDispose.push(action.onDidChange(function (event) {
                if (!_this.builder) {
                    // we have not been rendered yet, so there
                    // is no point in updating the UI
                    return;
                }
                _this._handleActionChangeEvent(event);
            }));
        }
    }
    BaseActionItem.prototype._handleActionChangeEvent = function (event) {
        if (event.enabled !== void 0) {
            this._updateEnabled();
        }
        if (event.checked !== void 0) {
            this._updateChecked();
        }
        if (event.class !== void 0) {
            this._updateClass();
        }
        if (event.label !== void 0) {
            this._updateLabel();
            this._updateTooltip();
        }
        if (event.tooltip !== void 0) {
            this._updateTooltip();
        }
    };
    Object.defineProperty(BaseActionItem.prototype, "callOnDispose", {
        get: function () {
            return this._callOnDispose;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BaseActionItem.prototype, "actionRunner", {
        get: function () {
            return this._actionRunner;
        },
        set: function (actionRunner) {
            this._actionRunner = actionRunner;
        },
        enumerable: true,
        configurable: true
    });
    BaseActionItem.prototype.getAction = function () {
        return this._action;
    };
    BaseActionItem.prototype.isEnabled = function () {
        return this._action.enabled;
    };
    BaseActionItem.prototype.setActionContext = function (newContext) {
        this._context = newContext;
    };
    BaseActionItem.prototype.render = function (container) {
        var _this = this;
        this.builder = builder_1.$(container);
        this.gesture = new touch_1.Gesture(container);
        this.builder.on(touch_1.EventType.Tap, function (e) { return _this.onClick(e); });
        if (platform.isMacintosh) {
            this.builder.on(DOM.EventType.CONTEXT_MENU, function (event) { return _this.onClick(event); }); // https://github.com/Microsoft/vscode/issues/1011
        }
        this.builder.on(DOM.EventType.MOUSE_DOWN, function (e) {
            DOM.EventHelper.stop(e);
            if (_this._action.enabled) {
                _this.builder.addClass('active');
            }
        });
        this.builder.on(DOM.EventType.CLICK, function (e) {
            DOM.EventHelper.stop(e, true);
            setTimeout(function () { return _this.onClick(e); }, 50);
        });
        this.builder.on([DOM.EventType.MOUSE_UP, DOM.EventType.MOUSE_OUT], function (e) {
            DOM.EventHelper.stop(e);
            _this.builder.removeClass('active');
        });
    };
    BaseActionItem.prototype.onClick = function (event) {
        DOM.EventHelper.stop(event, true);
        var context;
        if (types.isUndefinedOrNull(this._context)) {
            context = event;
        }
        else {
            context = this._context;
            context.event = event;
        }
        this._actionRunner.run(this._action, context);
    };
    BaseActionItem.prototype.focus = function () {
        if (this.builder) {
            this.builder.domFocus();
        }
    };
    BaseActionItem.prototype.blur = function () {
        if (this.builder) {
            this.builder.domBlur();
        }
    };
    BaseActionItem.prototype._updateEnabled = function () {
        // implement in subclass
    };
    BaseActionItem.prototype._updateLabel = function () {
        // implement in subclass
    };
    BaseActionItem.prototype._updateTooltip = function () {
        // implement in subclass
    };
    BaseActionItem.prototype._updateClass = function () {
        // implement in subclass
    };
    BaseActionItem.prototype._updateChecked = function () {
        // implement in subclass
    };
    BaseActionItem.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        if (this.builder) {
            this.builder.destroy();
            this.builder = null;
        }
        if (this.gesture) {
            this.gesture.dispose();
            this.gesture = null;
        }
        this._callOnDispose = lifecycle.dispose(this._callOnDispose);
    };
    return BaseActionItem;
}(eventEmitter_1.EventEmitter));
exports.BaseActionItem = BaseActionItem;
var Separator = (function (_super) {
    __extends(Separator, _super);
    function Separator(label, order) {
        _super.call(this, Separator.ID, label, label ? 'separator text' : 'separator');
        this.checked = false;
        this.enabled = false;
        this.order = order;
    }
    Separator.ID = 'vs.actions.separator';
    return Separator;
}(actions_1.Action));
exports.Separator = Separator;
var ActionItem = (function (_super) {
    __extends(ActionItem, _super);
    function ActionItem(context, action, options) {
        if (options === void 0) { options = {}; }
        _super.call(this, context, action);
        this.options = options;
        this.options.icon = options.icon !== undefined ? options.icon : false;
        this.options.label = options.label !== undefined ? options.label : true;
        this.cssClass = '';
    }
    ActionItem.prototype.render = function (container) {
        _super.prototype.render.call(this, container);
        this.$e = builder_1.$('a.action-label').appendTo(this.builder);
        this.$e.attr({ role: 'button' });
        if (this.options.label && this.options.keybinding) {
            builder_1.$('span.keybinding').text(this.options.keybinding).appendTo(this.builder);
        }
        this._updateClass();
        this._updateLabel();
        this._updateTooltip();
        this._updateEnabled();
        this._updateChecked();
    };
    ActionItem.prototype.focus = function () {
        _super.prototype.focus.call(this);
        this.$e.domFocus();
    };
    ActionItem.prototype._updateLabel = function () {
        if (this.options.label) {
            this.$e.text(this.getAction().label);
        }
    };
    ActionItem.prototype._updateTooltip = function () {
        var title = null;
        if (this.getAction().tooltip) {
            title = this.getAction().tooltip;
        }
        else if (!this.options.label && this.getAction().label && this.options.icon) {
            title = this.getAction().label;
            if (this.options.keybinding) {
                title = nls.localize({ key: 'titleLabel', comment: ['action title', 'action keybinding'] }, "{0} ({1})", title, this.options.keybinding);
            }
        }
        if (title) {
            this.$e.attr({ title: title });
        }
    };
    ActionItem.prototype._updateClass = function () {
        if (this.cssClass) {
            this.$e.removeClass(this.cssClass);
        }
        if (this.options.icon) {
            this.cssClass = this.getAction().class;
            this.$e.addClass('icon');
            if (this.cssClass) {
                this.$e.addClass(this.cssClass);
            }
            this._updateEnabled();
        }
        else {
            this.$e.removeClass('icon');
        }
    };
    ActionItem.prototype._updateEnabled = function () {
        if (this.getAction().enabled) {
            this.builder.removeClass('disabled');
            this.$e.removeClass('disabled');
            this.$e.attr({ tabindex: 0 });
        }
        else {
            this.builder.addClass('disabled');
            this.$e.addClass('disabled');
            DOM.removeTabIndexAndUpdateFocus(this.$e.getHTMLElement());
        }
    };
    ActionItem.prototype._updateChecked = function () {
        if (this.getAction().checked) {
            this.$e.addClass('checked');
        }
        else {
            this.$e.removeClass('checked');
        }
    };
    return ActionItem;
}(BaseActionItem));
exports.ActionItem = ActionItem;
(function (ActionsOrientation) {
    ActionsOrientation[ActionsOrientation["HORIZONTAL"] = 1] = "HORIZONTAL";
    ActionsOrientation[ActionsOrientation["VERTICAL"] = 2] = "VERTICAL";
})(exports.ActionsOrientation || (exports.ActionsOrientation = {}));
var ActionsOrientation = exports.ActionsOrientation;
var defaultOptions = {
    orientation: ActionsOrientation.HORIZONTAL,
    context: null
};
var ActionBar = (function (_super) {
    __extends(ActionBar, _super);
    function ActionBar(container, options) {
        var _this = this;
        if (options === void 0) { options = defaultOptions; }
        _super.call(this);
        this.options = options;
        this._context = options.context;
        this.toDispose = [];
        this._actionRunner = this.options.actionRunner;
        if (!this._actionRunner) {
            this._actionRunner = new actions_1.ActionRunner();
            this.toDispose.push(this._actionRunner);
        }
        this.toDispose.push(this.addEmitter2(this._actionRunner));
        this.items = [];
        this.focusedItem = undefined;
        this.domNode = document.createElement('div');
        this.domNode.className = 'monaco-action-bar';
        if (options.animated !== false) {
            DOM.addClass(this.domNode, 'animated');
        }
        var isVertical = this.options.orientation === ActionsOrientation.VERTICAL;
        if (isVertical) {
            this.domNode.className += ' vertical';
        }
        builder_1.$(this.domNode).on(DOM.EventType.KEY_DOWN, function (e) {
            var event = new keyboardEvent_1.StandardKeyboardEvent(e);
            var eventHandled = true;
            if (event.equals(isVertical ? 16 /* UpArrow */ : 15 /* LeftArrow */)) {
                _this.focusPrevious();
            }
            else if (event.equals(isVertical ? 18 /* DownArrow */ : 17 /* RightArrow */)) {
                _this.focusNext();
            }
            else if (event.equals(9 /* Escape */)) {
                _this.cancel();
            }
            else if (event.equals(3 /* Enter */) || event.equals(10 /* Space */)) {
            }
            else {
                eventHandled = false;
            }
            if (eventHandled) {
                event.preventDefault();
                event.stopPropagation();
            }
        });
        // Prevent native context menu on actions
        builder_1.$(this.domNode).on(DOM.EventType.CONTEXT_MENU, function (e) {
            e.preventDefault();
            e.stopPropagation();
        });
        builder_1.$(this.domNode).on(DOM.EventType.KEY_UP, function (e) {
            var event = new keyboardEvent_1.StandardKeyboardEvent(e);
            // Run action on Enter/Space
            if (event.equals(3 /* Enter */) || event.equals(10 /* Space */)) {
                _this.doTrigger(event);
                event.preventDefault();
                event.stopPropagation();
            }
            else if (event.equals(2 /* Tab */) || event.equals(16384 /* Shift */ | 2 /* Tab */)) {
                _this.updateFocusedItem();
            }
        });
        this.focusTracker = DOM.trackFocus(this.domNode);
        this.focusTracker.addBlurListener(function () {
            if (document.activeElement === _this.domNode || !DOM.isAncestor(document.activeElement, _this.domNode)) {
                _this.emit(DOM.EventType.BLUR, {});
                _this.focusedItem = undefined;
            }
        });
        this.focusTracker.addFocusListener(function () { return _this.updateFocusedItem(); });
        this.actionsList = document.createElement('ul');
        this.actionsList.className = 'actions-container';
        this.actionsList.setAttribute('role', 'toolbar');
        if (this.options.ariaLabel) {
            this.actionsList.setAttribute('aria-label', this.options.ariaLabel);
        }
        this.domNode.appendChild(this.actionsList);
        ((container instanceof builder_1.Builder) ? container.getHTMLElement() : container).appendChild(this.domNode);
    }
    ActionBar.prototype.setAriaLabel = function (label) {
        if (label) {
            this.actionsList.setAttribute('aria-label', label);
        }
        else {
            this.actionsList.removeAttribute('aria-label');
        }
    };
    ActionBar.prototype.updateFocusedItem = function () {
        for (var i = 0; i < this.actionsList.children.length; i++) {
            var elem = this.actionsList.children[i];
            if (DOM.isAncestor(document.activeElement, elem)) {
                this.focusedItem = i;
                break;
            }
        }
    };
    Object.defineProperty(ActionBar.prototype, "context", {
        get: function () {
            return this._context;
        },
        set: function (context) {
            this._context = context;
            this.items.forEach(function (i) { return i.setActionContext(context); });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ActionBar.prototype, "actionRunner", {
        get: function () {
            return this._actionRunner;
        },
        set: function (actionRunner) {
            if (actionRunner) {
                this._actionRunner = actionRunner;
                this.items.forEach(function (item) { return item.actionRunner = actionRunner; });
            }
        },
        enumerable: true,
        configurable: true
    });
    ActionBar.prototype.getContainer = function () {
        return builder_1.$(this.domNode);
    };
    ActionBar.prototype.push = function (arg, options) {
        var _this = this;
        if (options === void 0) { options = {}; }
        var actions = !Array.isArray(arg) ? [arg] : arg;
        var index = types.isNumber(options.index) ? options.index : null;
        actions.forEach(function (action) {
            var actionItemElement = document.createElement('li');
            actionItemElement.className = 'action-item';
            actionItemElement.setAttribute('role', 'presentation');
            var item = null;
            if (_this.options.actionItemProvider) {
                item = _this.options.actionItemProvider(action);
            }
            if (!item) {
                item = new ActionItem(_this.context, action, options);
            }
            item.actionRunner = _this._actionRunner;
            item.setActionContext(_this.context);
            _this.addEmitter2(item);
            item.render(actionItemElement);
            if (index === null || index < 0 || index >= _this.actionsList.children.length) {
                _this.actionsList.appendChild(actionItemElement);
            }
            else {
                _this.actionsList.insertBefore(actionItemElement, _this.actionsList.children[index++]);
            }
            _this.items.push(item);
        });
    };
    ActionBar.prototype.clear = function () {
        // Do not dispose action items if they were provided from outside
        this.items = this.options.actionItemProvider ? [] : lifecycle.dispose(this.items);
        builder_1.$(this.actionsList).empty();
    };
    ActionBar.prototype.length = function () {
        return this.items.length;
    };
    ActionBar.prototype.isEmpty = function () {
        return this.items.length === 0;
    };
    ActionBar.prototype.focus = function (selectFirst) {
        if (selectFirst && typeof this.focusedItem === 'undefined') {
            this.focusedItem = 0;
        }
        this.updateFocus();
    };
    ActionBar.prototype.focusNext = function () {
        if (typeof this.focusedItem === 'undefined') {
            this.focusedItem = this.items.length - 1;
        }
        var startIndex = this.focusedItem;
        var item;
        do {
            this.focusedItem = (this.focusedItem + 1) % this.items.length;
            item = this.items[this.focusedItem];
        } while (this.focusedItem !== startIndex && !item.isEnabled());
        if (this.focusedItem === startIndex && !item.isEnabled()) {
            this.focusedItem = undefined;
        }
        this.updateFocus();
    };
    ActionBar.prototype.focusPrevious = function () {
        if (typeof this.focusedItem === 'undefined') {
            this.focusedItem = 0;
        }
        var startIndex = this.focusedItem;
        var item;
        do {
            this.focusedItem = this.focusedItem - 1;
            if (this.focusedItem < 0) {
                this.focusedItem = this.items.length - 1;
            }
            item = this.items[this.focusedItem];
        } while (this.focusedItem !== startIndex && !item.isEnabled());
        if (this.focusedItem === startIndex && !item.isEnabled()) {
            this.focusedItem = undefined;
        }
        this.updateFocus();
    };
    ActionBar.prototype.updateFocus = function () {
        if (typeof this.focusedItem === 'undefined') {
            this.domNode.focus();
            return;
        }
        for (var i = 0; i < this.items.length; i++) {
            var item = this.items[i];
            var actionItem = item;
            if (i === this.focusedItem) {
                if (types.isFunction(actionItem.focus)) {
                    actionItem.focus();
                }
            }
            else {
                if (types.isFunction(actionItem.blur)) {
                    actionItem.blur();
                }
            }
        }
    };
    ActionBar.prototype.doTrigger = function (event) {
        if (typeof this.focusedItem === 'undefined') {
            return; //nothing to focus
        }
        // trigger action
        var actionItem = this.items[this.focusedItem];
        var context = (actionItem._context === null || actionItem._context === undefined) ? event : actionItem._context;
        this.run(actionItem._action, context).done();
    };
    ActionBar.prototype.cancel = function () {
        if (document.activeElement instanceof HTMLElement) {
            document.activeElement.blur(); // remove focus from focussed action
        }
        this.emit(events_1.EventType.CANCEL);
    };
    ActionBar.prototype.run = function (action, context) {
        return this._actionRunner.run(action, context);
    };
    ActionBar.prototype.dispose = function () {
        if (this.items !== null) {
            lifecycle.dispose(this.items);
        }
        this.items = null;
        if (this.focusTracker) {
            this.focusTracker.dispose();
            this.focusTracker = null;
        }
        this.toDispose = lifecycle.dispose(this.toDispose);
        this.getContainer().destroy();
        _super.prototype.dispose.call(this);
    };
    return ActionBar;
}(eventEmitter_1.EventEmitter));
exports.ActionBar = ActionBar;
var SelectActionItem = (function (_super) {
    __extends(SelectActionItem, _super);
    function SelectActionItem(ctx, action, options, selected) {
        _super.call(this, ctx, action);
        this.select = document.createElement('select');
        this.select.className = 'action-bar-select';
        this.options = options;
        this.selected = selected;
        this.toDispose = [];
        this.registerListeners();
    }
    SelectActionItem.prototype.setOptions = function (options, selected) {
        this.options = options;
        if (selected >= 0) {
            this.selected = selected;
        }
        else if (this.selected < 0 || this.selected > this.options.length) {
            this.selected = 0;
        }
        this.doSetOptions();
    };
    SelectActionItem.prototype.registerListeners = function () {
        var _this = this;
        this.toDispose.push(DOM.addStandardDisposableListener(this.select, 'change', function (e) {
            _this.actionRunner.run(_this._action, _this.getActionContext(e.target.value)).done();
        }));
    };
    SelectActionItem.prototype.getActionContext = function (option) {
        return option;
    };
    SelectActionItem.prototype.focus = function () {
        if (this.select) {
            this.select.focus();
        }
    };
    Object.defineProperty(SelectActionItem.prototype, "enabled", {
        set: function (value) {
            this.select.disabled = !value;
        },
        enumerable: true,
        configurable: true
    });
    SelectActionItem.prototype.blur = function () {
        if (this.select) {
            this.select.blur();
        }
    };
    SelectActionItem.prototype.render = function (container) {
        DOM.addClass(container, 'select-container');
        container.appendChild(this.select);
        this.doSetOptions();
    };
    SelectActionItem.prototype.getSelected = function () {
        return this.options && this.selected >= 0 && this.selected < this.options.length ? this.options[this.selected] : null;
    };
    SelectActionItem.prototype.doSetOptions = function () {
        var _this = this;
        this.select.options.length = 0;
        this.options.forEach(function (option) {
            _this.select.add(_this.createOption(option));
        });
        if (this.selected >= 0) {
            this.select.selectedIndex = this.selected;
            this.select.title = this.options[this.selected];
        }
    };
    SelectActionItem.prototype.createOption = function (value) {
        var option = document.createElement('option');
        option.value = value;
        option.text = value;
        return option;
    };
    SelectActionItem.prototype.dispose = function () {
        this.toDispose = lifecycle.dispose(this.toDispose);
        _super.prototype.dispose.call(this);
    };
    return SelectActionItem;
}(BaseActionItem));
exports.SelectActionItem = SelectActionItem;
