diff --git a/src/lockfile.rs b/src/lockfile.rs
index 8f9ebd83..b4842c70 100644
--- a/src/lockfile.rs
+++ b/src/lockfile.rs
@@ -131,7 +131,7 @@ impl Digest {
     }

     pub fn bin_version(binary: &Path) -> Result<String> {
-        let safe_vars = [OsStr::new("HOMEDRIVE"), OsStr::new("PATHEXT")];
+        let safe_vars = [OsStr::new("HOMEDRIVE"), OsStr::new("PATHEXT"), OsStr::new("NIX_LD_LIBRARY_PATH"), OsStr::new("NIX_LD"), OsStr::new("LD_LIBRARY_PATH")];
         let env = std::env::vars_os().filter(|(var, _)| safe_vars.contains(&var.as_os_str()));

         let output = Command::new(binary)
diff --git a/src/metadata.rs b/src/metadata.rs
index c84bfbb8..9ef2d13d 100644
--- a/src/metadata.rs
+++ b/src/metadata.rs
@@ -151,6 +150,9 @@ impl Cargo {
                 "sparse".into(),
             );
         }
+        map.insert("NIX_LD".into(), env::var("NIX_LD").unwrap_or_default());
+        map.insert("NIX_LD_LIBRARY_PATH".into(), env::var("NIX_LD_LIBRARY_PATH").unwrap_or_default());
+        map.insert("LD_LIBRARY_PATH".into(), env::var("LD_LIBRARY_PATH").unwrap_or_default());
         Ok(map)
     }
 }
