// vscode/src/vs/workbench/parts/markers/common/markersModel.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var paths = require("vs/base/common/paths");
var types = require("vs/base/common/types");
var Map = require("vs/base/common/map");
var severity_1 = require("vs/base/common/severity");
var uri_1 = require("vs/base/common/uri");
var range_1 = require("vs/editor/common/core/range");
var filters_1 = require("vs/base/common/filters");
var messages_1 = require("vs/workbench/parts/markers/common/messages");
var Resource = (function () {
    function Resource(uri, markers, statistics, matches) {
        if (matches === void 0) { matches = []; }
        this.uri = uri;
        this.markers = markers;
        this.statistics = statistics;
        this.matches = matches;
        this._name = null;
        this._path = null;
    }
    Object.defineProperty(Resource.prototype, "path", {
        get: function () {
            if (this._path === null) {
                this._path = this.uri.fsPath;
            }
            return this._path;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Resource.prototype, "name", {
        get: function () {
            if (this._name === null) {
                this._name = paths.basename(this.uri.fsPath);
            }
            return this._name;
        },
        enumerable: true,
        configurable: true
    });
    return Resource;
}());
exports.Resource = Resource;
var Marker = (function () {
    function Marker(id, marker, labelMatches, sourceMatches) {
        if (labelMatches === void 0) { labelMatches = []; }
        if (sourceMatches === void 0) { sourceMatches = []; }
        this.id = id;
        this.marker = marker;
        this.labelMatches = labelMatches;
        this.sourceMatches = sourceMatches;
    }
    Object.defineProperty(Marker.prototype, "resource", {
        get: function () {
            return this.marker.resource;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(Marker.prototype, "range", {
        get: function () {
            return this.marker;
        },
        enumerable: true,
        configurable: true
    });
    Marker.prototype.toString = function () {
        return ["file: '" + this.marker.resource + "'",
            "severity: '" + severity_1.default.toString(this.marker.severity) + "'",
            "message: '" + this.marker.message + "'",
            "at: '" + this.marker.startLineNumber + "," + this.marker.startColumn + "'",
            "source: '" + (this.marker.source ? this.marker.source : '') + "'"].join('\n');
    };
    return Marker;
}());
exports.Marker = Marker;
var FilterOptions = (function () {
    function FilterOptions(filter) {
        if (filter === void 0) { filter = ''; }
        this._filterErrors = false;
        this._filterWarnings = false;
        this._filterInfos = false;
        this._filter = '';
        this._completeFilter = '';
        if (filter) {
            this.parse(filter);
        }
    }
    Object.defineProperty(FilterOptions.prototype, "filterErrors", {
        get: function () {
            return this._filterErrors;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FilterOptions.prototype, "filterWarnings", {
        get: function () {
            return this._filterWarnings;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FilterOptions.prototype, "filterInfos", {
        get: function () {
            return this._filterInfos;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FilterOptions.prototype, "filter", {
        get: function () {
            return this._filter;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FilterOptions.prototype, "completeFilter", {
        get: function () {
            return this._completeFilter;
        },
        enumerable: true,
        configurable: true
    });
    FilterOptions.prototype.hasFilters = function () {
        return !!this._filter;
    };
    FilterOptions.prototype.parse = function (filter) {
        this._completeFilter = filter;
        this._filter = filter.trim();
        this._filterErrors = this.matches(this._filter, messages_1.default.MARKERS_PANEL_FILTER_ERRORS);
        this._filterWarnings = this.matches(this._filter, messages_1.default.MARKERS_PANEL_FILTER_WARNINGS);
        this._filterInfos = this.matches(this._filter, messages_1.default.MARKERS_PANEL_FILTER_INFOS);
    };
    FilterOptions.prototype.matches = function (prefix, word) {
        var result = filters_1.matchesPrefix(prefix, word);
        return result && result.length > 0;
    };
    return FilterOptions;
}());
FilterOptions._filter = filters_1.or(filters_1.matchesPrefix, filters_1.matchesContiguousSubString);
FilterOptions._fuzzyFilter = filters_1.or(filters_1.matchesPrefix, filters_1.matchesContiguousSubString, filters_1.matchesFuzzy);
exports.FilterOptions = FilterOptions;
var MarkersModel = (function () {
    function MarkersModel(markers) {
        if (markers === void 0) { markers = []; }
        this.markersByResource = new Map.LinkedMap();
        this._filterOptions = new FilterOptions();
        this.update(markers);
    }
    Object.defineProperty(MarkersModel.prototype, "filterOptions", {
        get: function () {
            return this._filterOptions;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MarkersModel.prototype, "filteredResources", {
        get: function () {
            return this._filteredResources;
        },
        enumerable: true,
        configurable: true
    });
    MarkersModel.prototype.hasFilteredResources = function () {
        return this.filteredResources.length > 0;
    };
    MarkersModel.prototype.hasResources = function () {
        return this.markersByResource.size > 0;
    };
    MarkersModel.prototype.hasResource = function (resource) {
        return this.markersByResource.has(resource);
    };
    Object.defineProperty(MarkersModel.prototype, "nonFilteredResources", {
        get: function () {
            return this._nonFilteredResources;
        },
        enumerable: true,
        configurable: true
    });
    MarkersModel.prototype.getBulkUpdater = function () {
        var _this = this;
        return {
            add: function (resourceUri, markers) {
                _this.updateResource(resourceUri, markers);
            },
            done: function () {
                _this.refresh();
            }
        };
    };
    MarkersModel.prototype.update = function (arg1, arg2) {
        if (arg1 instanceof FilterOptions) {
            this._filterOptions = arg1;
        }
        if (arg1 instanceof uri_1.default) {
            this.updateResource(arg1, arg2);
        }
        if (types.isArray(arg1)) {
            this.updateMarkers(arg1);
        }
        this.refresh();
    };
    MarkersModel.prototype.refresh = function () {
        this.refreshResources();
    };
    MarkersModel.prototype.refreshResources = function () {
        this._nonFilteredResources = [];
        this._filteredResources = [];
        for (var _i = 0, _a = this.markersByResource.entries(); _i < _a.length; _i++) {
            var entry = _a[_i];
            var filteredResource = this.toFilteredResource(entry);
            if (filteredResource.markers.length) {
                this._filteredResources.push(filteredResource);
            }
            else {
                this._nonFilteredResources.push(filteredResource);
            }
        }
    };
    MarkersModel.prototype.updateResource = function (resourceUri, markers) {
        if (this.markersByResource.has(resourceUri)) {
            this.markersByResource.delete(resourceUri);
        }
        if (markers.length > 0) {
            this.markersByResource.set(resourceUri, markers);
        }
    };
    MarkersModel.prototype.updateMarkers = function (markers) {
        var _this = this;
        markers.forEach(function (marker) {
            var uri = marker.resource;
            var markers = _this.markersByResource.get(uri);
            if (!markers) {
                markers = [];
                _this.markersByResource.set(uri, markers);
            }
            markers.push(marker);
        });
    };
    MarkersModel.prototype.toFilteredResource = function (entry) {
        var markers = [];
        for (var i = 0; i < entry.value.length; i++) {
            var m = entry.value[i];
            var uri = entry.key.toString();
            if (!this._filterOptions.hasFilters() || this.filterMarker(m)) {
                markers.push(this.toMarker(m, i, uri));
            }
        }
        var matches = this._filterOptions.hasFilters() ? FilterOptions._filter(this._filterOptions.filter, paths.basename(entry.key.fsPath)) : [];
        return new Resource(entry.key, markers, this.getStatistics(entry.value), matches || []);
    };
    MarkersModel.prototype.toMarker = function (marker, index, uri) {
        var labelMatches = this._filterOptions.hasFilters() ? FilterOptions._fuzzyFilter(this._filterOptions.filter, marker.message) : [];
        var sourceMatches = marker.source && this._filterOptions.hasFilters() ? FilterOptions._filter(this._filterOptions.filter, marker.source) : [];
        return new Marker(uri + index, marker, labelMatches || [], sourceMatches || []);
    };
    MarkersModel.prototype.filterMarker = function (marker) {
        if (this._filterOptions.filterErrors && severity_1.default.Error === marker.severity) {
            return true;
        }
        if (this._filterOptions.filterWarnings && severity_1.default.Warning === marker.severity) {
            return true;
        }
        if (this._filterOptions.filterInfos && severity_1.default.Info === marker.severity) {
            return true;
        }
        if (!!FilterOptions._fuzzyFilter(this._filterOptions.filter, marker.message)) {
            return true;
        }
        if (!!FilterOptions._filter(this._filterOptions.filter, paths.basename(marker.resource.fsPath))) {
            return true;
        }
        if (!!marker.source && !!FilterOptions._filter(this._filterOptions.filter, marker.source)) {
            return true;
        }
        return false;
    };
    MarkersModel.prototype.getStatistics = function (markers) {
        var errors = 0, warnings = 0, infos = 0, unknowns = 0;
        for (var _i = 0, markers_1 = markers; _i < markers_1.length; _i++) {
            var marker = markers_1[_i];
            switch (marker.severity) {
                case severity_1.default.Error:
                    errors++;
                    break;
                case severity_1.default.Warning:
                    warnings++;
                    break;
                case severity_1.default.Info:
                    infos++;
                    break;
                default:
                    unknowns++;
                    break;
            }
        }
        return { errors: errors, warnings: warnings, infos: infos, unknowns: unknowns };
    };
    MarkersModel.prototype.dispose = function () {
        this.markersByResource.clear();
        this._filteredResources = [];
        this._nonFilteredResources = [];
    };
    MarkersModel.prototype.getTitle = function (markerStatistics) {
        var title = MarkersModel.getStatisticsLabel(markerStatistics);
        return title ? title : messages_1.default.MARKERS_PANEL_TITLE_PROBLEMS;
    };
    MarkersModel.prototype.getMessage = function () {
        if (this.hasFilteredResources()) {
            return '';
        }
        if (this.hasResources()) {
            if (this._filterOptions.hasFilters()) {
                return messages_1.default.MARKERS_PANEL_NO_PROBLEMS_FILTERS;
            }
        }
        return messages_1.default.MARKERS_PANEL_NO_PROBLEMS_BUILT;
    };
    MarkersModel.getStatisticsLabel = function (markerStatistics, onlyErrors) {
        if (onlyErrors === void 0) { onlyErrors = false; }
        var label = this.getLabel('', markerStatistics.errors, messages_1.default.MARKERS_PANEL_SINGLE_ERROR_LABEL, messages_1.default.MARKERS_PANEL_MULTIPLE_ERRORS_LABEL);
        if (!onlyErrors) {
            label = this.getLabel(label, markerStatistics.warnings, messages_1.default.MARKERS_PANEL_SINGLE_WARNING_LABEL, messages_1.default.MARKERS_PANEL_MULTIPLE_WARNINGS_LABEL);
            label = this.getLabel(label, markerStatistics.infos, messages_1.default.MARKERS_PANEL_SINGLE_INFO_LABEL, messages_1.default.MARKERS_PANEL_MULTIPLE_INFOS_LABEL);
            label = this.getLabel(label, markerStatistics.unknowns, messages_1.default.MARKERS_PANEL_SINGLE_UNKNOWN_LABEL, messages_1.default.MARKERS_PANEL_MULTIPLE_UNKNOWNS_LABEL);
        }
        return label;
    };
    MarkersModel.getLabel = function (title, markersCount, singleMarkerString, multipleMarkersFunction) {
        if (markersCount <= 0) {
            return title;
        }
        title = title ? title + ', ' : '';
        title += markersCount === 1 ? singleMarkerString : multipleMarkersFunction(markersCount);
        return title;
    };
    MarkersModel.compare = function (a, b) {
        if (a instanceof Resource && b instanceof Resource) {
            return MarkersModel.compareResources(a, b);
        }
        if (a instanceof Marker && b instanceof Marker) {
            return MarkersModel.compareMarkers(a, b);
        }
        return 0;
    };
    MarkersModel.compareResources = function (a, b) {
        if (a.statistics.errors === 0 && b.statistics.errors > 0) {
            return 1;
        }
        if (b.statistics.errors === 0 && a.statistics.errors > 0) {
            return -1;
        }
        return a.path.localeCompare(b.path) || a.name.localeCompare(b.name);
    };
    MarkersModel.compareMarkers = function (a, b) {
        if (a.marker.severity === b.marker.severity) {
            return range_1.Range.compareRangesUsingStarts(a.marker, b.marker);
        }
        return a.marker.severity > b.marker.severity ? -1 : 1;
    };
    return MarkersModel;
}());
exports.MarkersModel = MarkersModel;
