// vscode/src/vs/workbench/common/editor/untitledEditorModel.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var stringEditorModel_1 = require("vs/workbench/common/editor/stringEditorModel");
var modesRegistry_1 = require("vs/editor/common/modes/modesRegistry");
var editorCommon_1 = require("vs/editor/common/editorCommon");
var files_1 = require("vs/platform/files/common/files");
var configuration_1 = require("vs/platform/configuration/common/configuration");
var modeService_1 = require("vs/editor/common/services/modeService");
var modelService_1 = require("vs/editor/common/services/modelService");
var event_1 = require("vs/base/common/event");
var async_1 = require("vs/base/common/async");
var backup_1 = require("vs/workbench/services/backup/common/backup");
var textfiles_1 = require("vs/workbench/services/textfile/common/textfiles");
var UntitledEditorModel = (function (_super) {
    __extends(UntitledEditorModel, _super);
    function UntitledEditorModel(modeId, resource, hasAssociatedFilePath, modeService, modelService, backupFileService, textFileService, configurationService) {
        var _this = _super.call(this, '', modeId, resource, modeService, modelService) || this;
        _this.backupFileService = backupFileService;
        _this.textFileService = textFileService;
        _this.configurationService = configurationService;
        _this.hasAssociatedFilePath = hasAssociatedFilePath;
        _this.dirty = false;
        _this.versionId = 0;
        _this._onDidChangeContent = new event_1.Emitter();
        _this._onDidChangeDirty = new event_1.Emitter();
        _this._onDidChangeEncoding = new event_1.Emitter();
        _this.contentChangeEventScheduler = new async_1.RunOnceScheduler(function () { return _this._onDidChangeContent.fire(); }, UntitledEditorModel.DEFAULT_CONTENT_CHANGE_BUFFER_DELAY);
        _this.registerListeners();
        return _this;
    }
    Object.defineProperty(UntitledEditorModel.prototype, "onDidChangeContent", {
        get: function () {
            return this._onDidChangeContent.event;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UntitledEditorModel.prototype, "onDidChangeDirty", {
        get: function () {
            return this._onDidChangeDirty.event;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UntitledEditorModel.prototype, "onDidChangeEncoding", {
        get: function () {
            return this._onDidChangeEncoding.event;
        },
        enumerable: true,
        configurable: true
    });
    UntitledEditorModel.prototype.getOrCreateMode = function (modeService, modeId, firstLineText) {
        if (!modeId || modeId === modesRegistry_1.PLAINTEXT_MODE_ID) {
            return modeService.getOrCreateModeByFilenameOrFirstLine(this.resource.fsPath, firstLineText); // lookup mode via resource path if the provided modeId is unspecific
        }
        return _super.prototype.getOrCreateMode.call(this, modeService, modeId, firstLineText);
    };
    UntitledEditorModel.prototype.registerListeners = function () {
        var _this = this;
        // Config Changes
        this.configurationChangeListener = this.configurationService.onDidUpdateConfiguration(function (e) { return _this.onConfigurationChange(e.config); });
    };
    UntitledEditorModel.prototype.onConfigurationChange = function (configuration) {
        this.configuredEncoding = configuration && configuration.files && configuration.files.encoding;
    };
    UntitledEditorModel.prototype.getVersionId = function () {
        return this.versionId;
    };
    UntitledEditorModel.prototype.getValue = function () {
        if (this.textEditorModel) {
            return this.textEditorModel.getValue(editorCommon_1.EndOfLinePreference.TextDefined, true /* Preserve BOM */);
        }
        return null;
    };
    UntitledEditorModel.prototype.getModeId = function () {
        if (this.textEditorModel) {
            return this.textEditorModel.getModeId();
        }
        return null;
    };
    UntitledEditorModel.prototype.getEncoding = function () {
        return this.preferredEncoding || this.configuredEncoding;
    };
    UntitledEditorModel.prototype.setEncoding = function (encoding) {
        var oldEncoding = this.getEncoding();
        this.preferredEncoding = encoding;
        // Emit if it changed
        if (oldEncoding !== this.preferredEncoding) {
            this._onDidChangeEncoding.fire();
        }
    };
    UntitledEditorModel.prototype.isDirty = function () {
        return this.dirty;
    };
    UntitledEditorModel.prototype.setDirty = function (dirty) {
        if (this.dirty === dirty) {
            return;
        }
        this.dirty = dirty;
        this._onDidChangeDirty.fire();
    };
    UntitledEditorModel.prototype.getResource = function () {
        return this.resource;
    };
    UntitledEditorModel.prototype.revert = function () {
        this.setDirty(false);
        // Handle content change event buffered
        this.contentChangeEventScheduler.schedule();
    };
    UntitledEditorModel.prototype.load = function () {
        var _this = this;
        // Check for backups first
        return this.backupFileService.loadBackupResource(this.resource).then(function (backupResource) {
            if (backupResource) {
                return _this.textFileService.resolveTextContent(backupResource, backup_1.BACKUP_FILE_RESOLVE_OPTIONS).then(function (rawTextContent) {
                    return _this.backupFileService.parseBackupContent(rawTextContent);
                });
            }
            return null;
        }).then(function (backupContent) {
            if (backupContent) {
                _this.setValue(backupContent);
            }
            _this.setDirty(_this.hasAssociatedFilePath || !!backupContent); // untitled associated to file path are dirty right away as well as untitled with content
            return _super.prototype.load.call(_this).then(function (model) {
                var configuration = _this.configurationService.getConfiguration();
                // Encoding
                _this.configuredEncoding = configuration && configuration.files && configuration.files.encoding;
                // Listen to content changes
                _this.textModelChangeListener = _this.textEditorModel.onDidChangeContent(function (e) { return _this.onModelContentChanged(); });
                return model;
            });
        });
    };
    UntitledEditorModel.prototype.onModelContentChanged = function () {
        this.versionId++;
        // mark the untitled editor as non-dirty once its content becomes empty and we do
        // not have an associated path set. we never want dirty indicator in that case.
        if (!this.hasAssociatedFilePath && this.textEditorModel.getLineCount() === 1 && this.textEditorModel.getLineContent(1) === '') {
            this.setDirty(false);
        }
        else {
            this.setDirty(true);
        }
        // Handle content change event buffered
        this.contentChangeEventScheduler.schedule();
    };
    UntitledEditorModel.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        if (this.textModelChangeListener) {
            this.textModelChangeListener.dispose();
            this.textModelChangeListener = null;
        }
        if (this.configurationChangeListener) {
            this.configurationChangeListener.dispose();
            this.configurationChangeListener = null;
        }
        this.contentChangeEventScheduler.dispose();
        this._onDidChangeContent.dispose();
        this._onDidChangeDirty.dispose();
        this._onDidChangeEncoding.dispose();
    };
    return UntitledEditorModel;
}(stringEditorModel_1.StringEditorModel));
UntitledEditorModel.DEFAULT_CONTENT_CHANGE_BUFFER_DELAY = files_1.CONTENT_CHANGE_EVENT_BUFFER_DELAY;
UntitledEditorModel = __decorate([
    __param(3, modeService_1.IModeService),
    __param(4, modelService_1.IModelService),
    __param(5, backup_1.IBackupFileService),
    __param(6, textfiles_1.ITextFileService),
    __param(7, configuration_1.IConfigurationService)
], UntitledEditorModel);
exports.UntitledEditorModel = UntitledEditorModel;
