// vscode/src/vs/platform/extensions/common/extensionsRegistry.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var nls = require('vs/nls');
var errors_1 = require('vs/base/common/errors');
var paths = require('vs/base/common/paths');
var severity_1 = require('vs/base/common/severity');
var jsonContributionRegistry_1 = require('vs/platform/jsonschemas/common/jsonContributionRegistry');
var platform_1 = require('vs/platform/platform');
var ExtensionMessageCollector = (function () {
    function ExtensionMessageCollector(messageHandler, source) {
        this._messageHandler = messageHandler;
        this._source = source;
    }
    ExtensionMessageCollector.prototype._msg = function (type, message) {
        this._messageHandler({
            type: type,
            message: message,
            source: this._source
        });
    };
    ExtensionMessageCollector.prototype.error = function (message) {
        this._msg(severity_1.default.Error, message);
    };
    ExtensionMessageCollector.prototype.warn = function (message) {
        this._msg(severity_1.default.Warning, message);
    };
    ExtensionMessageCollector.prototype.info = function (message) {
        this._msg(severity_1.default.Info, message);
    };
    return ExtensionMessageCollector;
}());
function isValidExtensionDescription(extensionFolderPath, extensionDescription, notices) {
    if (!extensionDescription) {
        notices.push(nls.localize('extensionDescription.empty', "Got empty extension description"));
        return false;
    }
    if (typeof extensionDescription.publisher !== 'string') {
        notices.push(nls.localize('extensionDescription.publisher', "property `{0}` is mandatory and must be of type `string`", 'publisher'));
        return false;
    }
    if (typeof extensionDescription.name !== 'string') {
        notices.push(nls.localize('extensionDescription.name', "property `{0}` is mandatory and must be of type `string`", 'name'));
        return false;
    }
    if (typeof extensionDescription.version !== 'string') {
        notices.push(nls.localize('extensionDescription.version', "property `{0}` is mandatory and must be of type `string`", 'version'));
        return false;
    }
    if (!extensionDescription.engines) {
        notices.push(nls.localize('extensionDescription.engines', "property `{0}` is mandatory and must be of type `object`", 'engines'));
        return false;
    }
    if (typeof extensionDescription.engines.vscode !== 'string') {
        notices.push(nls.localize('extensionDescription.engines.vscode', "property `{0}` is mandatory and must be of type `string`", 'engines.vscode'));
        return false;
    }
    if (typeof extensionDescription.extensionDependencies !== 'undefined') {
        if (!_isStringArray(extensionDescription.extensionDependencies)) {
            notices.push(nls.localize('extensionDescription.extensionDependencies', "property `{0}` can be omitted or must be of type `string[]`", 'extensionDependencies'));
            return false;
        }
    }
    if (typeof extensionDescription.activationEvents !== 'undefined') {
        if (!_isStringArray(extensionDescription.activationEvents)) {
            notices.push(nls.localize('extensionDescription.activationEvents1', "property `{0}` can be omitted or must be of type `string[]`", 'activationEvents'));
            return false;
        }
        if (typeof extensionDescription.main === 'undefined') {
            notices.push(nls.localize('extensionDescription.activationEvents2', "properties `{0}` and `{1}` must both be specified or must both be omitted", 'activationEvents', 'main'));
            return false;
        }
    }
    if (typeof extensionDescription.main !== 'undefined') {
        if (typeof extensionDescription.main !== 'string') {
            notices.push(nls.localize('extensionDescription.main1', "property `{0}` can be omitted or must be of type `string`", 'main'));
            return false;
        }
        else {
            var normalizedAbsolutePath = paths.normalize(paths.join(extensionFolderPath, extensionDescription.main));
            if (normalizedAbsolutePath.indexOf(extensionFolderPath)) {
                notices.push(nls.localize('extensionDescription.main2', "Expected `main` ({0}) to be included inside extension's folder ({1}). This might make the extension non-portable.", normalizedAbsolutePath, extensionFolderPath));
            }
        }
        if (typeof extensionDescription.activationEvents === 'undefined') {
            notices.push(nls.localize('extensionDescription.main3', "properties `{0}` and `{1}` must both be specified or must both be omitted", 'activationEvents', 'main'));
            return false;
        }
    }
    return true;
}
exports.isValidExtensionDescription = isValidExtensionDescription;
var hasOwnProperty = Object.hasOwnProperty;
var schemaRegistry = platform_1.Registry.as(jsonContributionRegistry_1.Extensions.JSONContribution);
var ExtensionPoint = (function () {
    function ExtensionPoint(name, registry) {
        this.name = name;
        this._registry = registry;
        this._handler = null;
        this._messageHandler = null;
    }
    ExtensionPoint.prototype.setHandler = function (handler) {
        if (this._handler) {
            throw new Error('Handler already set!');
        }
        this._handler = handler;
        this._handle();
    };
    ExtensionPoint.prototype.handle = function (messageHandler) {
        this._messageHandler = messageHandler;
        this._handle();
    };
    ExtensionPoint.prototype._handle = function () {
        var _this = this;
        if (!this._handler || !this._messageHandler) {
            return;
        }
        this._registry.registerPointListener(this.name, function (descriptions) {
            var users = descriptions.map(function (desc) {
                return {
                    description: desc,
                    value: desc.contributes[_this.name],
                    collector: new ExtensionMessageCollector(_this._messageHandler, desc.extensionFolderPath)
                };
            });
            _this._handler(users);
        });
    };
    return ExtensionPoint;
}());
var schemaId = 'vscode://schemas/vscode-extensions';
var schema = {
    default: {
        'name': '{{name}}',
        'description': '{{description}}',
        'author': '{{author}}',
        'version': '{{1.0.0}}',
        'main': '{{pathToMain}}',
        'dependencies': {}
    },
    properties: {
        // engines: {
        // 	required: [ 'vscode' ],
        // 	properties: {
        // 		'vscode': {
        // 			type: 'string',
        // 			description: nls.localize('vscode.extension.engines.vscode', 'Specifies that this package only runs inside VSCode of the given version.'),
        // 		}
        // 	}
        // },
        publisher: {
            description: nls.localize('vscode.extension.publisher', 'The publisher of the VS Code extension.'),
            type: 'string'
        },
        displayName: {
            description: nls.localize('vscode.extension.displayName', 'The display name for the extension used in the VS Code gallery.'),
            type: 'string'
        },
        categories: {
            description: nls.localize('vscode.extension.categories', 'The categories used by the VS Code gallery to categorize the extension.'),
            type: 'array',
            uniqueItems: true,
            items: {
                type: 'string',
                enum: ['Languages', 'Snippets', 'Linters', 'Themes', 'Debuggers', 'Productivity', 'Other']
            }
        },
        galleryBanner: {
            type: 'object',
            description: nls.localize('vscode.extension.galleryBanner', 'Banner used in the VS Code marketplace.'),
            properties: {
                color: {
                    description: nls.localize('vscode.extension.galleryBanner.color', 'The banner color on the VS Code marketplace page header.'),
                    type: 'string'
                },
                theme: {
                    description: nls.localize('vscode.extension.galleryBanner.theme', 'The color theme for the font used in the banner.'),
                    type: 'string',
                    enum: ['dark', 'light']
                }
            }
        },
        contributes: {
            description: nls.localize('vscode.extension.contributes', 'All contributions of the VS Code extension represented by this package.'),
            type: 'object',
            properties: {},
            default: {}
        },
        preview: {
            type: 'boolean',
            description: nls.localize('vscode.extension.preview', 'Sets the extension to be flagged as a Preview in the Marketplace.'),
        },
        activationEvents: {
            description: nls.localize('vscode.extension.activationEvents', 'Activation events for the VS Code extension.'),
            type: 'array',
            items: {
                type: 'string',
                defaultSnippets: [{ label: 'onLanguage', body: 'onLanguage:{{languageId}}' }, { label: 'onCommand', body: 'onCommand:{{commandId}}' }, { label: 'onDebug', body: 'onDebug:{{type}}' }, { label: 'workspaceContains', body: 'workspaceContains:{{fileName}}' }],
            }
        },
        badges: {
            type: 'array',
            description: nls.localize('vscode.extension.badges', 'Array of badges to display in the sidebar of the Marketplace\'s extension page.'),
            items: {
                type: 'object',
                required: ['url', 'href', 'description'],
                properties: {
                    url: {
                        type: 'string',
                        description: nls.localize('vscode.extension.badges.url', 'Badge image URL.')
                    },
                    href: {
                        type: 'string',
                        description: nls.localize('vscode.extension.badges.href', 'Badge link.')
                    },
                    description: {
                        type: 'string',
                        description: nls.localize('vscode.extension.badges.description', 'Badge description.')
                    }
                }
            }
        },
        extensionDependencies: {
            description: nls.localize('vscode.extension.extensionDependencies', 'Dependencies to other extensions. The identifier of an extension is always ${publisher}.${name}. For example: vscode.csharp.'),
            type: 'array',
            uniqueItems: true,
            items: {
                type: 'string'
            }
        },
        scripts: {
            type: 'object',
            properties: {
                'vscode:prepublish': {
                    description: nls.localize('vscode.extension.scripts.prepublish', 'Script executed before the package is published as a VS Code extension.'),
                    type: 'string'
                }
            }
        },
        icon: {
            type: 'string',
            description: nls.localize('vscode.extension.icon', 'The path to a 128x128 pixel icon.')
        }
    }
};
var ExtensionsRegistryImpl = (function () {
    function ExtensionsRegistryImpl() {
        this._extensionsMap = {};
        this._extensionsArr = [];
        this._activationMap = {};
        this._pointListeners = [];
        this._extensionPoints = {};
        this._oneTimeActivationEventListeners = {};
    }
    ExtensionsRegistryImpl.prototype.registerPointListener = function (point, handler) {
        var entry = {
            extensionPoint: point,
            listener: handler
        };
        this._pointListeners.push(entry);
        this._triggerPointListener(entry, ExtensionsRegistryImpl._filterWithExtPoint(this.getAllExtensionDescriptions(), point));
    };
    ExtensionsRegistryImpl.prototype.registerExtensionPoint = function (extensionPoint, jsonSchema) {
        if (hasOwnProperty.call(this._extensionPoints, extensionPoint)) {
            throw new Error('Duplicate extension point: ' + extensionPoint);
        }
        var result = new ExtensionPoint(extensionPoint, this);
        this._extensionPoints[extensionPoint] = result;
        schema.properties['contributes'].properties[extensionPoint] = jsonSchema;
        schemaRegistry.registerSchema(schemaId, schema);
        return result;
    };
    ExtensionsRegistryImpl.prototype.handleExtensionPoints = function (messageHandler) {
        var _this = this;
        Object.keys(this._extensionPoints).forEach(function (extensionPointName) {
            _this._extensionPoints[extensionPointName].handle(messageHandler);
        });
    };
    ExtensionsRegistryImpl.prototype._triggerPointListener = function (handler, desc) {
        // console.log('_triggerPointListeners: ' + desc.length + ' OF ' + handler.extensionPoint);
        if (!desc || desc.length === 0) {
            return;
        }
        try {
            handler.listener(desc);
        }
        catch (e) {
            errors_1.onUnexpectedError(e);
        }
    };
    ExtensionsRegistryImpl.prototype.registerExtensions = function (extensionDescriptions) {
        for (var i = 0, len = extensionDescriptions.length; i < len; i++) {
            var extensionDescription = extensionDescriptions[i];
            if (hasOwnProperty.call(this._extensionsMap, extensionDescription.id)) {
                // No overwriting allowed!
                console.error('Extension `' + extensionDescription.id + '` is already registered');
                continue;
            }
            this._extensionsMap[extensionDescription.id] = extensionDescription;
            this._extensionsArr.push(extensionDescription);
            if (Array.isArray(extensionDescription.activationEvents)) {
                for (var j = 0, lenJ = extensionDescription.activationEvents.length; j < lenJ; j++) {
                    var activationEvent = extensionDescription.activationEvents[j];
                    this._activationMap[activationEvent] = this._activationMap[activationEvent] || [];
                    this._activationMap[activationEvent].push(extensionDescription);
                }
            }
        }
        for (var i = 0, len = this._pointListeners.length; i < len; i++) {
            var listenerEntry = this._pointListeners[i];
            var descriptions = ExtensionsRegistryImpl._filterWithExtPoint(extensionDescriptions, listenerEntry.extensionPoint);
            this._triggerPointListener(listenerEntry, descriptions);
        }
    };
    ExtensionsRegistryImpl._filterWithExtPoint = function (input, point) {
        return input.filter(function (desc) {
            return (desc.contributes && hasOwnProperty.call(desc.contributes, point));
        });
    };
    ExtensionsRegistryImpl.prototype.getExtensionDescriptionsForActivationEvent = function (activationEvent) {
        if (!hasOwnProperty.call(this._activationMap, activationEvent)) {
            return [];
        }
        return this._activationMap[activationEvent].slice(0);
    };
    ExtensionsRegistryImpl.prototype.getAllExtensionDescriptions = function () {
        return this._extensionsArr.slice(0);
    };
    ExtensionsRegistryImpl.prototype.getExtensionDescription = function (extensionId) {
        if (!hasOwnProperty.call(this._extensionsMap, extensionId)) {
            return null;
        }
        return this._extensionsMap[extensionId];
    };
    ExtensionsRegistryImpl.prototype.registerOneTimeActivationEventListener = function (activationEvent, listener) {
        if (!hasOwnProperty.call(this._oneTimeActivationEventListeners, activationEvent)) {
            this._oneTimeActivationEventListeners[activationEvent] = [];
        }
        this._oneTimeActivationEventListeners[activationEvent].push(listener);
    };
    ExtensionsRegistryImpl.prototype.triggerActivationEventListeners = function (activationEvent) {
        if (hasOwnProperty.call(this._oneTimeActivationEventListeners, activationEvent)) {
            var listeners = this._oneTimeActivationEventListeners[activationEvent];
            delete this._oneTimeActivationEventListeners[activationEvent];
            for (var i = 0, len = listeners.length; i < len; i++) {
                var listener = listeners[i];
                try {
                    listener();
                }
                catch (e) {
                    errors_1.onUnexpectedError(e);
                }
            }
        }
    };
    return ExtensionsRegistryImpl;
}());
function _isStringArray(arr) {
    if (!Array.isArray(arr)) {
        return false;
    }
    for (var i = 0, len = arr.length; i < len; i++) {
        if (typeof arr[i] !== 'string') {
            return false;
        }
    }
    return true;
}
var PRExtensions = {
    ExtensionsRegistry: 'ExtensionsRegistry'
};
platform_1.Registry.add(PRExtensions.ExtensionsRegistry, new ExtensionsRegistryImpl());
exports.ExtensionsRegistry = platform_1.Registry.as(PRExtensions.ExtensionsRegistry);
schemaRegistry.registerSchema(schemaId, schema);
