// vscode/src/vs/platform/message/common/message.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var nls = require("vs/nls");
var winjs_base_1 = require("vs/base/common/winjs.base");
var severity_1 = require("vs/base/common/severity");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var actions_1 = require("vs/base/common/actions");
exports.CloseAction = new actions_1.Action('close.message', nls.localize('close', "Close"), null, true, function () { return winjs_base_1.TPromise.as(true); });
exports.LaterAction = new actions_1.Action('later.message', nls.localize('later', "Later"), null, true, function () { return winjs_base_1.TPromise.as(true); });
exports.CancelAction = new actions_1.Action('cancel.message', nls.localize('cancel', "Cancel"), null, true, function () { return winjs_base_1.TPromise.as(true); });
exports.IMessageService = instantiation_1.createDecorator('messageService');
exports.IChoiceService = instantiation_1.createDecorator('choiceService');
exports.Severity = severity_1.default;
