// vscode/src/vs/workbench/parts/debug/common/debug.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var instantiation_1 = require('vs/platform/instantiation/common/instantiation');
var contextkey_1 = require('vs/platform/contextkey/common/contextkey');
exports.VIEWLET_ID = 'workbench.view.debug';
exports.REPL_ID = 'workbench.panel.repl';
exports.DEBUG_SERVICE_ID = 'debugService';
exports.CONTEXT_IN_DEBUG_MODE = new contextkey_1.RawContextKey('inDebugMode', false);
exports.CONTEXT_NOT_IN_DEBUG_MODE = exports.CONTEXT_IN_DEBUG_MODE.toNegated();
exports.CONTEXT_IN_DEBUG_REPL = new contextkey_1.RawContextKey('inDebugRepl', false);
exports.CONTEXT_NOT_IN_DEBUG_REPL = exports.CONTEXT_IN_DEBUG_REPL.toNegated();
exports.CONTEXT_ON_FIRST_DEBUG_REPL_LINE = new contextkey_1.RawContextKey('onFirsteDebugReplLine', false);
exports.CONTEXT_ON_LAST_DEBUG_REPL_LINE = new contextkey_1.RawContextKey('onLastDebugReplLine', false);
exports.EDITOR_CONTRIBUTION_ID = 'editor.contrib.debug';
exports.DEBUG_SCHEME = 'debug';
;
// service enums
(function (State) {
    State[State["Disabled"] = 0] = "Disabled";
    State[State["Inactive"] = 1] = "Inactive";
    State[State["Initializing"] = 2] = "Initializing";
    State[State["Stopped"] = 3] = "Stopped";
    State[State["Running"] = 4] = "Running";
    State[State["RunningNoDebug"] = 5] = "RunningNoDebug";
})(exports.State || (exports.State = {}));
var State = exports.State;
exports.IDebugService = instantiation_1.createDecorator(exports.DEBUG_SERVICE_ID);
// utils
var _formatPIIRegexp = /{([^}]+)}/g;
function formatPII(value, excludePII, args) {
    return value.replace(_formatPIIRegexp, function (match, group) {
        if (excludePII && group.length > 0 && group[0] !== '_') {
            return match;
        }
        return args && args.hasOwnProperty(group) ?
            args[group] :
            match;
    });
}
exports.formatPII = formatPII;
