// vscode/src/vs/base/browser/ui/countBadge/countBadge.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
require("./countBadge.css");
var dom_1 = require("vs/base/browser/dom");
var strings_1 = require("vs/base/common/strings");
var CountBadge = (function () {
    function CountBadge(container, count, titleFormat) {
        this.element = dom_1.append(container, dom_1.$('.monaco-count-badge'));
        this.titleFormat = titleFormat || '';
        this.setCount(count || 0);
    }
    CountBadge.prototype.setCount = function (count) {
        this.count = count;
        this.render();
    };
    CountBadge.prototype.setTitleFormat = function (titleFormat) {
        this.titleFormat = titleFormat;
        this.render();
    };
    CountBadge.prototype.render = function () {
        this.element.textContent = '' + this.count;
        this.element.title = strings_1.format(this.titleFormat, this.count);
    };
    return CountBadge;
}());
exports.CountBadge = CountBadge;
