diff --git a/ui/node_modules/vscode/src/vs/base/browser/ui/scrollbar/media/scrollbars.css b/ui/node_modules/vscode/src/vs/base/browser/ui/scrollbar/media/scrollbars.css
index 0931517a3..2b15e20a4 100644
--- a/ui/node_modules/vscode/src/vs/base/browser/ui/scrollbar/media/scrollbars.css
+++ b/ui/node_modules/vscode/src/vs/base/browser/ui/scrollbar/media/scrollbars.css
@@ -72,7 +72,7 @@
 	left: 3px;
 	height: 3px;
 	width: 100%;
-	box-shadow: #DDD 0 6px 6px -6px inset;
+	box-shadow: #000 0 6px 6px -6px inset;
 }
 .monaco-scrollable-element > .shadow.left {
 	display: block;
diff --git a/ui/node_modules/vscode/src/vs/base/common/worker/simpleWorker.ts b/ui/node_modules/vscode/src/vs/base/common/worker/simpleWorker.ts
index 648c36a37..e1fd59037 100644
--- a/ui/node_modules/vscode/src/vs/base/common/worker/simpleWorker.ts
+++ b/ui/node_modules/vscode/src/vs/base/common/worker/simpleWorker.ts
@@ -194,7 +194,12 @@ export class SimpleWorkerClient<T> extends Disposable {
 		super();
 
 		let lazyProxyFulfill: (v: T) => void = null;
-		let lazyProxyReject: (err: any) => void = null;
+
+		// HACK: lazyProxyReject doesn't SEEM to do anything, but the
+		// call to it fails with 'Uncaught TypeError: lazyProxyReject
+		// is not a function' because the assignment to it doesn't get
+		// called for some reason. So, just make it a no-op.
+		let lazyProxyReject: (err: any) => void = () => (void 0);
 
 		this._worker = this._register(workerFactory.create(
 			'vs/base/common/worker/simpleWorker',
diff --git a/ui/node_modules/vscode/src/vs/base/worker/defaultWorkerFactory.ts b/ui/node_modules/vscode/src/vs/base/worker/defaultWorkerFactory.ts
index 52bca4b7a..df6a7de55 100644
--- a/ui/node_modules/vscode/src/vs/base/worker/defaultWorkerFactory.ts
+++ b/ui/node_modules/vscode/src/vs/base/worker/defaultWorkerFactory.ts
@@ -19,7 +19,13 @@ function environment(name: string, fallback: any = false): any {
 }
 
 function defaultGetWorkerUrl(workerId: string, label: string): string {
-	return require.toUrl('./' + workerId) + '#' + label;
+	// HACK: Disable the web worker, and remove the require call that would
+	// otherwise cause Webpack to process it. If we require using worker-loader,
+	// it tries to fetch the script from a URL "https://sourcegraph.com/myrepo/-/myfile.go function() { ..."
+	// for some reason. This is a bug in Webpack and/or worker-loader. But we
+	// can safely disable the web worker without loss of functionality, so
+	// we choose that approach instead of debugging it for now.
+	throw new Error('web worker disabled; this should be unreachable');
 }
 var getWorkerUrl = getCrossOriginWorkerScriptUrl || defaultGetWorkerUrl;
 
diff --git a/ui/node_modules/vscode/src/vs/base/worker/workerMain.ts b/ui/node_modules/vscode/src/vs/base/worker/workerMain.ts
index f80f9d44e..62cab872e 100644
--- a/ui/node_modules/vscode/src/vs/base/worker/workerMain.ts
+++ b/ui/node_modules/vscode/src/vs/base/worker/workerMain.ts
@@ -9,17 +9,12 @@
 	let MonacoEnvironment = (<any>self).MonacoEnvironment;
 	let monacoBaseUrl = MonacoEnvironment && MonacoEnvironment.baseUrl ? MonacoEnvironment.baseUrl : '../../../';
 
-	if (typeof (<any>self).define !== 'function' || !(<any>self).define.amd) {
-		importScripts(monacoBaseUrl + 'vs/loader.js');
-	}
-
-	require.config({
-		baseUrl: monacoBaseUrl,
-		catchError: true
-	});
-
 	let loadCode = function (moduleId) {
-		require([moduleId], function (ws) {
+		// HACK: webpack doesn't support completely dynamic requires, so constrain this and make it be known if anything else is ever passed to it.
+		if (moduleId !== 'vs/base/common/worker/simpleWorker') {
+			throw new Error('can\'t dynamically load module ' + moduleId);
+		}
+		require(['vs/base/common/worker/simpleWorker'], function (ws) {
 			setTimeout(function () {
 				let messageHandler = ws.create((msg: any) => {
 					(<any>self).postMessage(msg);
diff --git a/ui/node_modules/vscode/src/vs/editor/common/editorCommon.ts b/ui/node_modules/vscode/src/vs/editor/common/editorCommon.ts
index 09b38d10a..da61944b8 100644
--- a/ui/node_modules/vscode/src/vs/editor/common/editorCommon.ts
+++ b/ui/node_modules/vscode/src/vs/editor/common/editorCommon.ts
@@ -3668,7 +3668,7 @@ export interface IEditor {
 	 * will not be destroyed.
 	 * It is safe to call setModel(null) to simply detach the current model from the editor.
 	 */
-	setModel(model: IEditorModel): void;
+	setModel(model: IEditorModel | undefined): void;
 
 	/**
 	 * Change the decorations. All decorations added through this changeAccessor
diff --git a/ui/node_modules/vscode/src/vs/editor/common/services/editorSimpleWorker.ts b/ui/node_modules/vscode/src/vs/editor/common/services/editorSimpleWorker.ts
index bbef60f8f..e424678d2 100644
--- a/ui/node_modules/vscode/src/vs/editor/common/services/editorSimpleWorker.ts
+++ b/ui/node_modules/vscode/src/vs/editor/common/services/editorSimpleWorker.ts
@@ -21,7 +21,6 @@ import { computeLinks } from 'vs/editor/common/modes/linkComputer';
 import { BasicInplaceReplace } from 'vs/editor/common/modes/supports/inplaceReplaceSupport';
 import { IRawModelData } from 'vs/editor/common/services/editorSimpleWorkerCommon';
 import { getWordAtText, ensureValidWordDefinition } from 'vs/editor/common/model/wordHelper';
-import { createMonacoBaseAPI } from 'vs/editor/common/standalone/standaloneBase';
 
 export interface IMirrorModel {
 	readonly uri: URI;
@@ -518,9 +517,3 @@ export class EditorSimpleWorkerImpl extends BaseEditorSimpleWorker implements IR
 export function create(): IRequestHandler {
 	return new EditorSimpleWorkerImpl();
 }
-
-var global: any = self;
-let isWebWorker = (typeof global.importScripts === 'function');
-if (isWebWorker) {
-	global.monaco = createMonacoBaseAPI();
-}
diff --git a/ui/node_modules/vscode/src/vs/platform/contextkey/common/contextkey.ts b/ui/node_modules/vscode/src/vs/platform/contextkey/common/contextkey.ts
index 71ba74893..f05d17a87 100644
--- a/ui/node_modules/vscode/src/vs/platform/contextkey/common/contextkey.ts
+++ b/ui/node_modules/vscode/src/vs/platform/contextkey/common/contextkey.ts
@@ -451,7 +451,7 @@ export interface IContextKeyServiceTarget {
 	setAttribute(attr: string, value: string): void;
 	removeAttribute(attr: string): void;
 	hasAttribute(attr: string): boolean;
-	getAttribute(attr: string): string;
+	getAttribute(attr: string): string | null;
 }
 
 export let IContextKeyService = createDecorator<IContextKeyService>('contextKeyService');
diff --git a/ui/node_modules/vscode/src/vs/base/common/platform.ts b/ui/node_modules/vscode/src/vs/base/common/platform.ts
index 3075beb0a..de6365bbe 100644
--- a/ui/node_modules/vscode/src/vs/base/common/platform.ts
+++ b/ui/node_modules/vscode/src/vs/base/common/platform.ts
@@ -43,7 +43,9 @@ declare let self: any;
 export const LANGUAGE_DEFAULT = 'en';
 
 // OS detection
-if (typeof process === 'object') {
+// We add the check for process.title because webpack injects a process
+// variable, causing this check to mistakenly succeed in the web browser.
+if (typeof process === 'object' && (process as any).title !== "browser") {
 	_isWindows = (process.platform === 'win32');
 	_isMacintosh = (process.platform === 'darwin');
 	_isLinux = (process.platform === 'linux');
diff --git a/ui/node_modules/vscode/src/vs/editor/common/model/textModel.ts b/ui/node_modules/vscode/src/vs/editor/common/model/textModel.ts
index 8475280eb..fc508dcaf 100644
--- a/ui/node_modules/vscode/src/vs/editor/common/model/textModel.ts
+++ b/ui/node_modules/vscode/src/vs/editor/common/model/textModel.ts
@@ -893,12 +893,7 @@ export class TextModel extends OrderGuaranteeEventEmitter implements editorCommo
 			return [];
 		}
 
-		let searchRange: Range;
-		if (Range.isIRange(rawSearchScope)) {
-			searchRange = this.validateRange(rawSearchScope);
-		} else {
-			searchRange = this.getFullModelRange();
-		}
+		const searchRange = this.getFullModelRange();
 
 		if (regex.multiline) {
 			return this._doFindMatchesMultiline(searchRange, regex, limitResultCount);
diff --git a/ui/node_modules/vscode/src/vs/editor/contrib/goToDeclaration/browser/goToDeclaration.ts b/ui/node_modules/vscode/src/vs/editor/contrib/goToDeclaration/browser/goToDeclaration.ts
index 2c1980787..726da1b83 100644
--- a/ui/node_modules/vscode/src/vs/editor/contrib/goToDeclaration/browser/goToDeclaration.ts
+++ b/ui/node_modules/vscode/src/vs/editor/contrib/goToDeclaration/browser/goToDeclaration.ts
@@ -368,7 +368,7 @@ class GotoDefinitionWithMouseEditorContribution implements editorCommon.IEditorC
 							source += '\n\u2026';
 						}
 
-						const language = this.modeService.getModeIdByFilenameOrFirstLine(editorModel.uri.fsPath);
+						const language = this.modeService.getModeIdByFilenameOrFirstLine(editorModel.uri.fragment);
 						hoverMessage = {
 							language,
 							value: source
diff --git a/ui/node_modules/vscode/src/vs/editor/editor.main.ts b/ui/node_modules/vscode/src/vs/editor/editor.main.ts
index 34afc07f3..3ce47776f 100644
--- a/ui/node_modules/vscode/src/vs/editor/editor.main.ts
+++ b/ui/node_modules/vscode/src/vs/editor/editor.main.ts
@@ -10,17 +10,9 @@ import 'vs/editor/contrib/quickOpen/browser/quickOutline';
 import 'vs/editor/contrib/quickOpen/browser/gotoLine';
 import 'vs/editor/contrib/quickOpen/browser/quickCommand';
 
-import { createMonacoBaseAPI } from 'vs/editor/common/standalone/standaloneBase';
-import { createMonacoEditorAPI } from 'vs/editor/browser/standalone/standaloneEditor';
-import { createMonacoLanguagesAPI } from 'vs/editor/browser/standalone/standaloneLanguages';
 import { DefaultConfig } from 'vs/editor/common/config/defaultConfig';
 
 // Set defaults for standalone editor
 DefaultConfig.editor.wrappingIndent = 'none';
 DefaultConfig.editor.folding = false;
-DefaultConfig.editor.glyphMargin = false;
-
-var global: any = self;
-global.monaco = createMonacoBaseAPI();
-global.monaco.editor = createMonacoEditorAPI();
-global.monaco.languages = createMonacoLanguagesAPI();
\ No newline at end of file
+DefaultConfig.editor.glyphMargin = false;
\ No newline at end of file
diff --git a/ui/node_modules/vscode/src/vs/editor/browser/standalone/standaloneEditor.ts b/ui/node_modules/vscode/src/vs/editor/browser/standalone/standaloneEditor.ts
index 37ca08de6..11fadd38a 100644
--- a/ui/node_modules/vscode/src/vs/editor/browser/standalone/standaloneEditor.ts
+++ b/ui/node_modules/vscode/src/vs/editor/browser/standalone/standaloneEditor.ts
@@ -254,66 +254,3 @@ export function colorize(text: string, languageId: string, options: IColorizerOp
 export function colorizeModelLine(model: IModel, lineNumber: number, tabSize: number = 4): string {
 	return Colorizer.colorizeModelLine(model, lineNumber, tabSize);
 }
-
-/**
- * @internal
- */
-export function createMonacoEditorAPI(): typeof monaco.editor {
-	return {
-		// methods
-		create: <any>create,
-		createDiffEditor: createDiffEditor,
-		createDiffNavigator: createDiffNavigator,
-
-		createModel: createModel,
-		setModelLanguage: setModelLanguage,
-		setModelMarkers: setModelMarkers,
-		getModels: getModels,
-		getModel: getModel,
-		onDidCreateModel: onDidCreateModel,
-		onWillDisposeModel: onWillDisposeModel,
-		onDidChangeModelLanguage: onDidChangeModelLanguage,
-
-
-		createWebWorker: createWebWorker,
-		colorizeElement: colorizeElement,
-		colorize: colorize,
-		colorizeModelLine: colorizeModelLine,
-
-		// enums
-		ScrollbarVisibility: ScrollbarVisibility,
-		WrappingIndent: editorCommon.WrappingIndent,
-		OverviewRulerLane: editorCommon.OverviewRulerLane,
-		EndOfLinePreference: editorCommon.EndOfLinePreference,
-		DefaultEndOfLine: editorCommon.DefaultEndOfLine,
-		EndOfLineSequence: editorCommon.EndOfLineSequence,
-		TrackedRangeStickiness: editorCommon.TrackedRangeStickiness,
-		CursorChangeReason: editorCommon.CursorChangeReason,
-		MouseTargetType: editorCommon.MouseTargetType,
-		TextEditorCursorStyle: editorCommon.TextEditorCursorStyle,
-		TextEditorCursorBlinkingStyle: editorCommon.TextEditorCursorBlinkingStyle,
-		ContentWidgetPositionPreference: ContentWidgetPositionPreference,
-		OverlayWidgetPositionPreference: OverlayWidgetPositionPreference,
-
-		// classes
-		InternalEditorScrollbarOptions: <any>editorCommon.InternalEditorScrollbarOptions,
-		EditorWrappingInfo: <any>editorCommon.EditorWrappingInfo,
-		InternalEditorViewOptions: <any>editorCommon.InternalEditorViewOptions,
-		EditorContribOptions: <any>editorCommon.EditorContribOptions,
-		InternalEditorOptions: <any>editorCommon.InternalEditorOptions,
-		OverviewRulerPosition: <any>editorCommon.OverviewRulerPosition,
-		EditorLayoutInfo: <any>editorCommon.EditorLayoutInfo,
-		BareFontInfo: <any>editorCommon.BareFontInfo,
-		FontInfo: <any>editorCommon.FontInfo,
-		TextModelResolvedOptions: <any>editorCommon.TextModelResolvedOptions,
-
-		// vars
-		EditorType: editorCommon.EditorType,
-		CursorMoveByUnit: editorCommon.CursorMoveByUnit,
-		CursorMovePosition: editorCommon.CursorMovePosition,
-		EditorScrollDirection: editorCommon.EditorScrollDirection,
-		EditorScrollByUnit: editorCommon.EditorScrollByUnit,
-		RevealLineAtArgument: editorCommon.RevealLineAtArgument,
-		Handler: editorCommon.Handler,
-	};
-}
diff --git a/ui/node_modules/vscode/src/vs/editor/browser/standalone/standaloneLanguages.ts b/ui/node_modules/vscode/src/vs/editor/browser/standalone/standaloneLanguages.ts
index 9bd1213b9..350eaa79a 100644
--- a/ui/node_modules/vscode/src/vs/editor/browser/standalone/standaloneLanguages.ts
+++ b/ui/node_modules/vscode/src/vs/editor/browser/standalone/standaloneLanguages.ts
@@ -480,39 +480,3 @@ class SuggestAdapter {
 		});
 	}
 }
-
-/**
- * @internal
- */
-export function createMonacoLanguagesAPI(): typeof monaco.languages {
-	return {
-		register: register,
-		getLanguages: getLanguages,
-		onLanguage: onLanguage,
-
-		// provider methods
-		setLanguageConfiguration: setLanguageConfiguration,
-		setTokensProvider: setTokensProvider,
-		setMonarchTokensProvider: setMonarchTokensProvider,
-		registerReferenceProvider: registerReferenceProvider,
-		registerRenameProvider: registerRenameProvider,
-		registerCompletionItemProvider: registerCompletionItemProvider,
-		registerSignatureHelpProvider: registerSignatureHelpProvider,
-		registerHoverProvider: registerHoverProvider,
-		registerDocumentSymbolProvider: registerDocumentSymbolProvider,
-		registerDocumentHighlightProvider: registerDocumentHighlightProvider,
-		registerDefinitionProvider: registerDefinitionProvider,
-		registerCodeLensProvider: registerCodeLensProvider,
-		registerCodeActionProvider: registerCodeActionProvider,
-		registerDocumentFormattingEditProvider: registerDocumentFormattingEditProvider,
-		registerDocumentRangeFormattingEditProvider: registerDocumentRangeFormattingEditProvider,
-		registerOnTypeFormattingEditProvider: registerOnTypeFormattingEditProvider,
-		registerLinkProvider: registerLinkProvider,
-
-		// enums
-		DocumentHighlightKind: modes.DocumentHighlightKind,
-		CompletionItemKind: CompletionItemKind,
-		SymbolKind: modes.SymbolKind,
-		IndentAction: IndentAction
-	};
-}
diff --git a/ui/node_modules/vscode/src/vs/platform/instantiation/common/instantiationService.ts b/ui/node_modules/vscode/src/vs/platform/instantiation/common/instantiationService.ts
index 5a004276c..95d84c875 100644
--- a/ui/node_modules/vscode/src/vs/platform/instantiation/common/instantiationService.ts
+++ b/ui/node_modules/vscode/src/vs/platform/instantiation/common/instantiationService.ts
@@ -98,7 +98,7 @@ export class InstantiationService implements IInstantiationService {
 		let serviceArgs = serviceDependencies.map(dependency => {
 			let service = this._getOrCreateServiceInstance(dependency.id);
 			if (!service && this._strict && !dependency.optional) {
-				throw new Error(`[createInstance] ${desc.ctor.name} depends on UNKNOWN service ${dependency.id}.`);
+				console.warn(`[createInstance] ${desc.ctor.name} depends on UNKNOWN service ${dependency.id}.`);
 			}
 			return service;
 		});
diff --git a/ui/node_modules/vscode/src/vs/editor/contrib/referenceSearch/browser/referencesWidget.ts b/ui/node_modules/vscode/src/vs/editor/contrib/referenceSearch/browser/referencesWidget.ts
index cd5d7b479..75c607245 100644
--- a/ui/node_modules/vscode/src/vs/editor/contrib/referenceSearch/browser/referencesWidget.ts
+++ b/ui/node_modules/vscode/src/vs/editor/contrib/referenceSearch/browser/referencesWidget.ts
@@ -8,7 +8,6 @@ import 'vs/css!./referencesWidget';
 import * as nls from 'vs/nls';
 import * as collections from 'vs/base/common/collections';
 import { onUnexpectedError } from 'vs/base/common/errors';
-import { getPathLabel } from 'vs/base/common/labels';
 import Event, { Emitter } from 'vs/base/common/event';
 import { IDisposable, dispose, Disposables, empty as EmptyDisposable } from 'vs/base/common/lifecycle';
 import { Schemas } from 'vs/base/common/network';
@@ -710,7 +710,12 @@ export class ReferenceWidget extends PeekViewWidget {
 
 		// Update widget header
 		if (reference.uri.scheme !== Schemas.inMemory) {
-			this.setTitle(reference.name, getPathLabel(reference.directory, this._contextService));
+			const file = reference.uri;
+			const path = file.path + "/" + file.fragment;
+			const dirs = path.split("/");
+			dirs.splice(0, 1);
+			const base = dirs.pop();
+			this.setTitle(base, dirs.join("/"));
 		} else {
 			this.setTitle(nls.localize('peekView.alternateTitle', "References"));
 		}
