// tslint-react/rules/jsxNoLambdaRule.js
/**
 * @license
 * Copyright 2016 Palantir Technologies, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Lint = require("tslint/lib/lint");
var ts = require("typescript");
var Rule = (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        _super.apply(this, arguments);
    }
    Rule.prototype.apply = function (sourceFile) {
        var jsxNoLambdaWalker = new JsxNoLambdaWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(jsxNoLambdaWalker);
    };
    Rule.FAILURE_STRING = "Lambdas are forbidden in JSX attributes due to their rendering performance impact";
    return Rule;
}(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var JsxNoLambdaWalker = (function (_super) {
    __extends(JsxNoLambdaWalker, _super);
    function JsxNoLambdaWalker() {
        _super.apply(this, arguments);
        this.isInJsxAttribute = false;
    }
    JsxNoLambdaWalker.prototype.visitNode = function (node) {
        if (node.kind === ts.SyntaxKind.JsxAttribute) {
            this.isInJsxAttribute = true;
            _super.prototype.visitNode.call(this, node);
            this.isInJsxAttribute = false;
        }
        else {
            _super.prototype.visitNode.call(this, node);
        }
    };
    JsxNoLambdaWalker.prototype.visitFunctionExpression = function (node) {
        if (this.isInJsxAttribute) {
            this.reportFailure(node);
        }
        _super.prototype.visitFunctionExpression.call(this, node);
    };
    JsxNoLambdaWalker.prototype.visitArrowFunction = function (node) {
        if (this.isInJsxAttribute) {
            this.reportFailure(node);
        }
        _super.prototype.visitArrowFunction.call(this, node);
    };
    JsxNoLambdaWalker.prototype.reportFailure = function (node) {
        this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING));
    };
    return JsxNoLambdaWalker;
}(Lint.RuleWalker));
