// vscode/src/vs/base/browser/browser.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var types = require("vs/base/common/types");
var Platform = require("vs/base/common/platform");
var event_1 = require("vs/base/common/event");
var WindowManager = (function () {
    function WindowManager() {
        this._zoomLevel = 0;
        this._zoomFactor = 0;
        this._pixelRatioCache = 0;
        this._pixelRatioComputed = false;
        this._onDidChangeZoomLevel = new event_1.Emitter();
        this.onDidChangeZoomLevel = this._onDidChangeZoomLevel.event;
        this._onDidChangeFullscreen = new event_1.Emitter();
        this.onDidChangeFullscreen = this._onDidChangeFullscreen.event;
    }
    WindowManager.prototype.getZoomLevel = function () {
        return this._zoomLevel;
    };
    WindowManager.prototype.setZoomLevel = function (zoomLevel) {
        if (this._zoomLevel === zoomLevel) {
            return;
        }
        this._zoomLevel = zoomLevel;
        this._pixelRatioComputed = false;
        this._onDidChangeZoomLevel.fire(this._zoomLevel);
    };
    WindowManager.prototype.getZoomFactor = function () {
        return this._zoomFactor;
    };
    WindowManager.prototype.setZoomFactor = function (zoomFactor) {
        this._zoomFactor = zoomFactor;
    };
    WindowManager.prototype.getPixelRatio = function () {
        if (!this._pixelRatioComputed) {
            this._pixelRatioCache = this._computePixelRatio();
            this._pixelRatioComputed = true;
        }
        return this._pixelRatioCache;
    };
    WindowManager.prototype._computePixelRatio = function () {
        var ctx = document.createElement('canvas').getContext('2d');
        var dpr = window.devicePixelRatio || 1;
        var bsr = ctx.webkitBackingStorePixelRatio ||
            ctx.mozBackingStorePixelRatio ||
            ctx.msBackingStorePixelRatio ||
            ctx.oBackingStorePixelRatio ||
            ctx.backingStorePixelRatio || 1;
        return dpr / bsr;
    };
    WindowManager.prototype.setFullscreen = function (fullscreen) {
        if (this._fullscreen === fullscreen) {
            return;
        }
        this._fullscreen = fullscreen;
        this._onDidChangeFullscreen.fire();
    };
    WindowManager.prototype.isFullscreen = function () {
        return this._fullscreen;
    };
    return WindowManager;
}());
WindowManager.INSTANCE = new WindowManager();
/** A zoom index, e.g. 1, 2, 3 */
function getZoomLevel() {
    return WindowManager.INSTANCE.getZoomLevel();
}
exports.getZoomLevel = getZoomLevel;
/** The zoom scale for an index, e.g. 1, 1.2, 1.4 */
function getZoomFactor() {
    return WindowManager.INSTANCE.getZoomFactor();
}
exports.getZoomFactor = getZoomFactor;
function getPixelRatio() {
    return WindowManager.INSTANCE.getPixelRatio();
}
exports.getPixelRatio = getPixelRatio;
function setZoomLevel(zoomLevel) {
    WindowManager.INSTANCE.setZoomLevel(zoomLevel);
}
exports.setZoomLevel = setZoomLevel;
function setZoomFactor(zoomFactor) {
    WindowManager.INSTANCE.setZoomFactor(zoomFactor);
}
exports.setZoomFactor = setZoomFactor;
function onDidChangeZoomLevel(callback) {
    return WindowManager.INSTANCE.onDidChangeZoomLevel(callback);
}
exports.onDidChangeZoomLevel = onDidChangeZoomLevel;
function setFullscreen(fullscreen) {
    WindowManager.INSTANCE.setFullscreen(fullscreen);
}
exports.setFullscreen = setFullscreen;
function isFullscreen() {
    return WindowManager.INSTANCE.isFullscreen();
}
exports.isFullscreen = isFullscreen;
function onDidChangeFullscreen(callback) {
    return WindowManager.INSTANCE.onDidChangeFullscreen(callback);
}
exports.onDidChangeFullscreen = onDidChangeFullscreen;
var userAgent = navigator.userAgent;
// DOCUMENTED FOR FUTURE REFERENCE:
// When running IE11 in IE10 document mode, the code below will identify the browser as being IE10,
// which is correct because IE11 in IE10 document mode will reimplement all the bugs of IE10
exports.isIE11 = (userAgent.indexOf('Trident') >= 0 && userAgent.indexOf('MSIE') < 0);
exports.isIE10 = (userAgent.indexOf('MSIE 10') >= 0);
exports.isIE9 = (userAgent.indexOf('MSIE 9') >= 0);
exports.isIE11orEarlier = exports.isIE11 || exports.isIE10 || exports.isIE9;
exports.isIE10orEarlier = exports.isIE10 || exports.isIE9;
exports.isIE10orLater = exports.isIE11 || exports.isIE10;
exports.isEdge = (userAgent.indexOf('Edge/') >= 0);
exports.isEdgeOrIE = exports.isEdge || exports.isIE11 || exports.isIE10 || exports.isIE9;
exports.isOpera = (userAgent.indexOf('Opera') >= 0);
exports.isFirefox = (userAgent.indexOf('Firefox') >= 0);
exports.isWebKit = (userAgent.indexOf('AppleWebKit') >= 0);
exports.isChrome = (userAgent.indexOf('Chrome') >= 0);
exports.isSafari = (userAgent.indexOf('Chrome') === -1) && (userAgent.indexOf('Safari') >= 0);
exports.isIPad = (userAgent.indexOf('iPad') >= 0);
exports.canUseTranslate3d = !exports.isIE9 && !exports.isFirefox;
exports.enableEmptySelectionClipboard = exports.isWebKit;
/**
 * Returns if the browser supports CSS 3 animations.
 */
function hasCSSAnimationSupport() {
    if (this._hasCSSAnimationSupport === true || this._hasCSSAnimationSupport === false) {
        return this._hasCSSAnimationSupport;
    }
    var supported = false;
    var element = document.createElement('div');
    var properties = ['animationName', 'webkitAnimationName', 'msAnimationName', 'MozAnimationName', 'OAnimationName'];
    for (var i = 0; i < properties.length; i++) {
        var property = properties[i];
        if (!types.isUndefinedOrNull(element.style[property]) || element.style.hasOwnProperty(property)) {
            supported = true;
            break;
        }
    }
    if (supported) {
        this._hasCSSAnimationSupport = true;
    }
    else {
        this._hasCSSAnimationSupport = false;
    }
    return this._hasCSSAnimationSupport;
}
exports.hasCSSAnimationSupport = hasCSSAnimationSupport;
function supportsExecCommand(command) {
    return ((exports.isIE11orEarlier || Platform.isNative)
        && document.queryCommandSupported(command));
}
exports.supportsExecCommand = supportsExecCommand;
