diff --git a/repositories.bzl b/repositories.bzl
index 8a1d0b3..e921d50 100644
--- a/repositories.bzl
+++ b/repositories.bzl
@@ -22,6 +22,7 @@ def go_repositories():

     go_repository(
         name = "com_github_antihax_optional",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/antihax/optional",
         sum = "h1:xK2lYat7ZLaVVcIuj82J8kIro4V6kDe0AUDFboUCwcg=",
         version = "v1.0.0",
@@ -90,6 +91,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_census_instrumentation_opencensus_proto",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/census-instrumentation/opencensus-proto",
         sum = "h1:glEXhBS5PSLLv4IXzLA5yPRVX4bilULVyxxbrfOtDAk=",
         version = "v0.2.1",
@@ -102,6 +104,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_cespare_xxhash_v2",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/cespare/xxhash/v2",
         sum = "h1:6MnRN8NT7+YBpUIWxHtefFZOKTAPgGjpQSxqLNn0+qY=",
         version = "v2.1.1",
@@ -134,12 +137,14 @@ def go_repositories():

     go_repository(
         name = "com_github_cncf_udpa_go",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/cncf/udpa/go",
         sum = "h1:hzAQntlaYRkVSFEfj9OTWlVV1H155FMD8BTKktLv0QI=",
         version = "v0.0.0-20210930031921-04548b0d99d4",
     )
     go_repository(
         name = "com_github_cncf_xds_go",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/cncf/xds/go",
         sum = "h1:zH8ljVhhq7yC0MIeUL/IviMtY8hx2mK8cN9wEYb8ggw=",
         version = "v0.0.0-20211011173535-cb28da3451f1",
@@ -203,12 +208,14 @@ def go_repositories():

     go_repository(
         name = "com_github_envoyproxy_go_control_plane",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/envoyproxy/go-control-plane",
         sum = "h1:xvqufLtNVwAhN8NMyWklVgxnWohi+wtMGQMhtxexlm0=",
         version = "v0.10.2-0.20220325020618-49ff273808a1",
     )
     go_repository(
         name = "com_github_envoyproxy_protoc_gen_validate",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/envoyproxy/protoc-gen-validate",
         sum = "h1:EQciDnbrYxy13PgWoY8AqoxGiPrpgBZ1R8UNe3ddc+A=",
         version = "v0.1.0",
@@ -284,6 +291,7 @@ def go_repositories():

     go_repository(
         name = "com_github_golang_glog",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/golang/glog",
         sum = "h1:nfP3RFugxnNRyKgeWd4oI1nYvXpxrx8ck8ZrcizshdQ=",
         version = "v1.0.0",
@@ -303,6 +311,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_golang_protobuf",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/golang/protobuf",
         sum = "h1:ROPKBNFfQgOUMifHyP+KYbvpjbdoFNs+aK7DXlji0Tw=",
         version = "v1.5.2",
@@ -316,6 +325,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_google_go_cmp",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/google/go-cmp",
         sum = "h1:O2Tfq5qg4qc4AmwVlvv0oLiVAGB7enBSJ2x2DqQFi38=",
         version = "v0.5.9",
@@ -347,6 +357,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_google_uuid",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/google/uuid",
         sum = "h1:EVhdT+1Kseyi1/pUmXKaFxYsDNy9RQYkMWRH68J/W7Y=",
         version = "v1.1.2",
@@ -593,18 +604,21 @@ def go_repositories():

     go_repository(
         name = "com_github_kr_pretty",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/kr/pretty",
         sum = "h1:L/CwN0zerZDmRFUapSPitk6f+Q3+0za1rQkzVuMiMFI=",
         version = "v0.1.0",
     )
     go_repository(
         name = "com_github_kr_pty",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/kr/pty",
         sum = "h1:VkoXIwSboBpnk99O/KFauAEILuNHv5DVFKZMBN/gUgw=",
         version = "v1.1.1",
     )
     go_repository(
         name = "com_github_kr_text",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/kr/text",
         sum = "h1:45sCR5RtlFHMR4UwH9sdQ5TC8v0qDQCHnXt+kaKSTVE=",
         version = "v0.1.0",
@@ -782,6 +796,7 @@ def go_repositories():

     go_repository(
         name = "com_github_rogpeppe_fastuuid",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/rogpeppe/fastuuid",
         sum = "h1:Ppwyp6VYCF1nvBTXL3trRso7mXMlRrw9ooo375wvi2s=",
         version = "v1.2.0",
@@ -929,18 +944,21 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go",
         sum = "h1:DNtEKRBAAzeS4KyIory52wWHuClNaXJ5x1F7xa4q+5Y=",
         version = "v0.105.0",
     )
     go_repository(
         name = "com_google_cloud_go_accessapproval",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/accessapproval",
         sum = "h1:/nTivgnV/n1CaAeo+ekGexTYUsKEU9jUVkoY5359+3Q=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_accesscontextmanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/accesscontextmanager",
         sum = "h1:CFhNhU7pcD11cuDkQdrE6PQJgv0EXNKNv06jIzbLlCU=",
         version = "v1.4.0",
@@ -948,6 +966,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_aiplatform",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/aiplatform",
         sum = "h1:QqHZT1IMldf/daXoSnkJWBIqGBsw50X+xP6HSVzLRPo=",
         version = "v1.24.0",
@@ -955,24 +974,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_analytics",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/analytics",
         sum = "h1:NKw6PpQi6V1O+KsjuTd+bhip9d0REYu4NevC45vtGp8=",
         version = "v0.12.0",
     )
     go_repository(
         name = "com_google_cloud_go_apigateway",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/apigateway",
         sum = "h1:IIoXKR7FKrEAQhMTz5hK2wiDz2WNFHS7eVr/L1lE/rM=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_apigeeconnect",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/apigeeconnect",
         sum = "h1:AONoTYJviyv1vS4IkvWzq69gEVdvHx35wKXc+e6wjZQ=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_appengine",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/appengine",
         sum = "h1:lmG+O5oaR9xNwaRBwE2XoMhwQHsHql5IoiGr1ptdDwU=",
         version = "v1.5.0",
@@ -980,12 +1003,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_area120",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/area120",
         sum = "h1:TCMhwWEWhCn8d44/Zs7UCICTWje9j3HuV6nVGMjdpYw=",
         version = "v0.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_artifactregistry",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/artifactregistry",
         sum = "h1:3d0LRAU1K6vfqCahhl9fx2oGHcq+s5gftdix4v8Ibrc=",
         version = "v1.9.0",
@@ -993,36 +1018,42 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_asset",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/asset",
         sum = "h1:aCrlaLGJWTODJX4G56ZYzJefITKEWNfbjjtHSzWpxW0=",
         version = "v1.10.0",
     )
     go_repository(
         name = "com_google_cloud_go_assuredworkloads",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/assuredworkloads",
         sum = "h1:hhIdCOowsT1GG5eMCIA0OwK6USRuYTou/1ZeNxCSRtA=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_automl",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/automl",
         sum = "h1:BMioyXSbg7d7xLibn47cs0elW6RT780IUWr42W8rp2Q=",
         version = "v1.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_baremetalsolution",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/baremetalsolution",
         sum = "h1:g9KO6SkakcYPcc/XjAzeuUrEOXlYPnMpuiaywYaGrmQ=",
         version = "v0.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_batch",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/batch",
         sum = "h1:1jvEBY55OH4Sd2FxEXQfxGExFWov1A/IaRe+Z5Z71Fw=",
         version = "v0.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_beyondcorp",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/beyondcorp",
         sum = "h1:w+4kThysgl0JiKshi2MKDCg2NZgOyqOI0wq2eBZyrzA=",
         version = "v0.3.0",
@@ -1030,42 +1061,49 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_bigquery",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/bigquery",
         sum = "h1:u0fvz5ysJBe1jwUPI4LuPwAX+o+6fCUwf3ECeg6eDUQ=",
         version = "v1.43.0",
     )
     go_repository(
         name = "com_google_cloud_go_billing",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/billing",
         sum = "h1:Xkii76HWELHwBtkQVZvqmSo9GTr0O+tIbRNnMcGdlg4=",
         version = "v1.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_binaryauthorization",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/binaryauthorization",
         sum = "h1:pL70vXWn9TitQYXBWTK2abHl2JHLwkFRjYw6VflRqEA=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_certificatemanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/certificatemanager",
         sum = "h1:tzbR4UHBbgsewMWUD93JHi8EBi/gHBoSAcY1/sThFGk=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_channel",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/channel",
         sum = "h1:pNuUlZx0Jb0Ts9P312bmNMuH5IiFWIR4RUtLb70Ke5s=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_cloudbuild",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/cloudbuild",
         sum = "h1:TAAmCmAlOJ4uNBu6zwAjwhyl/7fLHHxIEazVhr3QBbQ=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_clouddms",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/clouddms",
         sum = "h1:UhzHIlgFfMr6luVYVNydw/pl9/U5kgtjCMJHnSvoVws=",
         version = "v1.4.0",
@@ -1073,6 +1111,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_cloudtasks",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/cloudtasks",
         sum = "h1:faUiUgXjW8yVZ7XMnKHKm1WE4OldPBUWWfIRN/3z1dc=",
         version = "v1.8.0",
@@ -1086,6 +1125,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go_compute_metadata",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/compute/metadata",
         sum = "h1:nBbNSZyDpkNlo3DepaaLKVuO7ClyifSAmNloSCZrHnQ=",
         version = "v0.2.0",
@@ -1093,12 +1133,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_contactcenterinsights",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/contactcenterinsights",
         sum = "h1:tTQLI/ZvguUf9Hv+36BkG2+/PeC8Ol1q4pBW+tgCx0A=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_container",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/container",
         sum = "h1:nbEK/59GyDRKKlo1SqpohY1TK8LmJ2XNcvS9Gyom2A0=",
         version = "v1.7.0",
@@ -1106,6 +1148,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_containeranalysis",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/containeranalysis",
         sum = "h1:2824iym832ljKdVpCBnpqm5K94YT/uHTVhNF+dRTXPI=",
         version = "v0.6.0",
@@ -1113,24 +1156,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_datacatalog",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datacatalog",
         sum = "h1:6kZ4RIOW/uT7QWC5SfPfq/G8sYzr/v+UOmOAxy4Z1TE=",
         version = "v1.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_dataflow",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataflow",
         sum = "h1:CW3541Fm7KPTyZjJdnX6NtaGXYFn5XbFC5UcjgALKvU=",
         version = "v0.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_dataform",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataform",
         sum = "h1:vLwowLF2ZB5J5gqiZCzv076lDI/Rd7zYQQFu5XO1PSg=",
         version = "v0.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_datafusion",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datafusion",
         sum = "h1:j5m2hjWovTZDTQak4MJeXAR9yN7O+zMfULnjGw/OOLg=",
         version = "v1.5.0",
@@ -1138,18 +1185,21 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_datalabeling",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datalabeling",
         sum = "h1:dp8jOF21n/7jwgo/uuA0RN8hvLcKO4q6s/yvwevs2ZM=",
         version = "v0.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_dataplex",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataplex",
         sum = "h1:cNxeA2DiWliQGi21kPRqnVeQ5xFhNoEjPRt1400Pm8Y=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_dataproc",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataproc",
         sum = "h1:gVOqNmElfa6n/ccG/QDlfurMWwrK3ezvy2b2eDoCmS0=",
         version = "v1.8.0",
@@ -1157,6 +1207,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_dataqna",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataqna",
         sum = "h1:gx9jr41ytcA3dXkbbd409euEaWtofCVXYBvJz3iYm18=",
         version = "v0.6.0",
@@ -1170,12 +1221,14 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go_datastream",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datastream",
         sum = "h1:PgIgbhedBtYBU6POGXFMn2uSl9vpqubc3ewTNdcU8Mk=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_deploy",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/deploy",
         sum = "h1:kI6dxt8Ml0is/x7YZjLveTvR7YPzXAUD/8wQZ2nH5zA=",
         version = "v1.5.0",
@@ -1183,12 +1236,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_dialogflow",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dialogflow",
         sum = "h1:HYHVOkoxQ9bSfNIelSZYNAtUi4CeSrCnROyOsbOqPq8=",
         version = "v1.19.0",
     )
     go_repository(
         name = "com_google_cloud_go_dlp",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dlp",
         sum = "h1:9I4BYeJSVKoSKgjr70fLdRDumqcUeVmHV4fd5f9LR6Y=",
         version = "v1.7.0",
@@ -1196,12 +1251,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_documentai",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/documentai",
         sum = "h1:jfq09Fdjtnpnmt/MLyf6A3DM3ynb8B2na0K+vSXvpFM=",
         version = "v1.10.0",
     )
     go_repository(
         name = "com_google_cloud_go_domains",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/domains",
         sum = "h1:pu3JIgC1rswIqi5romW0JgNO6CTUydLYX8zyjiAvO1c=",
         version = "v0.7.0",
@@ -1209,24 +1266,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_edgecontainer",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/edgecontainer",
         sum = "h1:hd6J2n5dBBRuAqnNUEsKWrp6XNPKsaxwwIyzOPZTokk=",
         version = "v0.2.0",
     )
     go_repository(
         name = "com_google_cloud_go_essentialcontacts",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/essentialcontacts",
         sum = "h1:b6csrQXCHKQmfo9h3dG/pHyoEh+fQG1Yg78a53LAviY=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_eventarc",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/eventarc",
         sum = "h1:AgCqrmMMIcel5WWKkzz5EkCUKC3Rl5LNMMYsS+LvsI0=",
         version = "v1.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_filestore",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/filestore",
         sum = "h1:yjKOpzvqtDmL5AXbKttLc8j0hL20kuC1qPdy5HPcxp0=",
         version = "v1.4.0",
@@ -1240,18 +1301,21 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go_functions",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/functions",
         sum = "h1:35tgv1fQOtvKqH/uxJMzX3w6usneJ0zXpsFr9KAVhNE=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_gaming",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gaming",
         sum = "h1:97OAEQtDazAJD7yh/kvQdSCQuTKdR0O+qWAJBZJ4xiA=",
         version = "v1.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_gkebackup",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkebackup",
         sum = "h1:4K+jiv4ocqt1niN8q5Imd8imRoXBHTrdnJVt/uFFxF4=",
         version = "v0.3.0",
@@ -1259,54 +1323,63 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_gkeconnect",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkeconnect",
         sum = "h1:zAcvDa04tTnGdu6TEZewaLN2tdMtUOJJ7fEceULjguA=",
         version = "v0.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_gkehub",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkehub",
         sum = "h1:JTcTaYQRGsVm+qkah7WzHb6e9sf1C0laYdRPn9aN+vg=",
         version = "v0.10.0",
     )
     go_repository(
         name = "com_google_cloud_go_gkemulticloud",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkemulticloud",
         sum = "h1:8F1NhJj8ucNj7lK51UZMtAjSWTgP1zO18XF6vkfiPPU=",
         version = "v0.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_gsuiteaddons",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gsuiteaddons",
         sum = "h1:TGT2oGmO5q3VH6SjcrlgPUWI0njhYv4kywLm6jag0to=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_iam",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/iam",
         sum = "h1:k4MuwOsS7zGJJ+QfZ5vBK8SgHBAvYN/23BWsiihJ1vs=",
         version = "v0.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_iap",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/iap",
         sum = "h1:BGEXovwejOCt1zDk8hXq0bOhhRu9haXKWXXXp2B4wBM=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_ids",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/ids",
         sum = "h1:LncHK4HHucb5Du310X8XH9/ICtMwZ2PCfK0ScjWiJoY=",
         version = "v1.2.0",
     )
     go_repository(
         name = "com_google_cloud_go_iot",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/iot",
         sum = "h1:Y9+oZT9jD4GUZzORXTU45XsnQrhxmDT+TFbPil6pRVQ=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_kms",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/kms",
         sum = "h1:OWRZzrPmOZUzurjI2FBGtgY2mB1WaJkqhw6oIwSj0Yg=",
         version = "v1.6.0",
@@ -1314,24 +1387,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_language",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/language",
         sum = "h1:3Wa+IUMamL4JH3Zd3cDZUHpwyqplTACt6UZKRD2eCL4=",
         version = "v1.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_lifesciences",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/lifesciences",
         sum = "h1:tIqhivE2LMVYkX0BLgG7xL64oNpDaFFI7teunglt1tI=",
         version = "v0.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_longrunning",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/longrunning",
         sum = "h1:NjljC+FYPV3uh5/OwWT6pVU+doBqMg2x/rZlE+CamDs=",
         version = "v0.3.0",
     )
     go_repository(
         name = "com_google_cloud_go_managedidentities",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/managedidentities",
         sum = "h1:3Kdajn6X25yWQFhFCErmKSYTSvkEd3chJROny//F1A0=",
         version = "v1.4.0",
@@ -1339,24 +1416,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_mediatranslation",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/mediatranslation",
         sum = "h1:qAJzpxmEX+SeND10Y/4868L5wfZpo4Y3BIEnIieP4dk=",
         version = "v0.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_memcache",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/memcache",
         sum = "h1:yLxUzJkZVSH2kPaHut7k+7sbIBFpvSh1LW9qjM2JDjA=",
         version = "v1.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_metastore",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/metastore",
         sum = "h1:3KcShzqWdqxrDEXIBWpYJpOOrgpDj+HlBi07Grot49Y=",
         version = "v1.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_monitoring",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/monitoring",
         sum = "h1:c9riaGSPQ4dUKWB+M1Fl0N+iLxstMbCktdEwYSPGDvA=",
         version = "v1.8.0",
@@ -1364,12 +1445,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_networkconnectivity",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/networkconnectivity",
         sum = "h1:BVdIKaI68bihnXGdCVL89Jsg9kq2kg+II30fjVqo62E=",
         version = "v1.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_networkmanagement",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/networkmanagement",
         sum = "h1:mDHA3CDW00imTvC5RW6aMGsD1bH+FtKwZm/52BxaiMg=",
         version = "v1.5.0",
@@ -1377,30 +1460,35 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_networksecurity",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/networksecurity",
         sum = "h1:qDEX/3sipg9dS5JYsAY+YvgTjPR63cozzAWop8oZS94=",
         version = "v0.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_notebooks",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/notebooks",
         sum = "h1:AC8RPjNvel3ExgXjO1YOAz+teg9+j+89TNxa7pIZfww=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_optimization",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/optimization",
         sum = "h1:7PxOq9VTT7TMib/6dMoWpMvWS2E4dJEvtYzjvBreaec=",
         version = "v1.2.0",
     )
     go_repository(
         name = "com_google_cloud_go_orchestration",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/orchestration",
         sum = "h1:39d6tqvNjd/wsSub1Bn4cEmrYcet5Ur6xpaN+SxOxtY=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_orgpolicy",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/orgpolicy",
         sum = "h1:erF5PHqDZb6FeFrUHiYj2JK2BMhsk8CyAg4V4amJ3rE=",
         version = "v1.5.0",
@@ -1408,24 +1496,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_osconfig",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/osconfig",
         sum = "h1:NO0RouqCOM7M2S85Eal6urMSSipWwHU8evzwS+siqUI=",
         version = "v1.10.0",
     )
     go_repository(
         name = "com_google_cloud_go_oslogin",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/oslogin",
         sum = "h1:pKGDPfeZHDybtw48WsnVLjoIPMi9Kw62kUE5TXCLCN4=",
         version = "v1.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_phishingprotection",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/phishingprotection",
         sum = "h1:OrwHLSRSZyaiOt3tnY33dsKSedxbMzsXvqB21okItNQ=",
         version = "v0.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_policytroubleshooter",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/policytroubleshooter",
         sum = "h1:NQklJuOUoz1BPP+Epjw81COx7IISWslkZubz/1i0UN8=",
         version = "v1.4.0",
@@ -1433,6 +1525,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_privatecatalog",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/privatecatalog",
         sum = "h1:Vz86uiHCtNGm1DeC32HeG2VXmOq5JRYA3VRPf8ZEcSg=",
         version = "v0.6.0",
@@ -1446,36 +1539,42 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go_recaptchaenterprise_v2",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/recaptchaenterprise/v2",
         sum = "h1:UqzFfb/WvhwXGDF1eQtdHLrmni+iByZXY4h3w9Kdyv8=",
         version = "v2.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_recommendationengine",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/recommendationengine",
         sum = "h1:6w+WxPf2LmUEqX0YyvfCoYb8aBYOcbIV25Vg6R0FLGw=",
         version = "v0.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_recommender",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/recommender",
         sum = "h1:9kMZQGeYfcOD/RtZfcNKGKtoex3DdoB4zRgYU/WaIwE=",
         version = "v1.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_redis",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/redis",
         sum = "h1:/zTwwBKIAD2DEWTrXZp8WD9yD/gntReF/HkPssVYd0U=",
         version = "v1.10.0",
     )
     go_repository(
         name = "com_google_cloud_go_resourcemanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/resourcemanager",
         sum = "h1:NDao6CHMwEZIaNsdWy+tuvHaavNeGP06o1tgrR0kLvU=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_resourcesettings",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/resourcesettings",
         sum = "h1:eTzOwB13WrfF0kuzG2ZXCfB3TLunSHBur4s+HFU6uSM=",
         version = "v1.4.0",
@@ -1483,12 +1582,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_retail",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/retail",
         sum = "h1:N9fa//ecFUOEPsW/6mJHfcapPV0wBSwIUwpVZB7MQ3o=",
         version = "v1.11.0",
     )
     go_repository(
         name = "com_google_cloud_go_run",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/run",
         sum = "h1:AWPuzU7Xtaj3Jf+QarDWIs6AJ5hM1VFQ+F6Q+VZ6OT4=",
         version = "v0.3.0",
@@ -1496,12 +1597,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_scheduler",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/scheduler",
         sum = "h1:K/mxOewgHGeKuATUJNGylT75Mhtjmx1TOkKukATqMT8=",
         version = "v1.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_secretmanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/secretmanager",
         sum = "h1:xE6uXljAC1kCR8iadt9+/blg1fvSbmenlsDN4fT9gqw=",
         version = "v1.9.0",
@@ -1509,18 +1612,21 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_security",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/security",
         sum = "h1:KSKzzJMyUoMRQzcz7azIgqAUqxo7rmQ5rYvimMhikqg=",
         version = "v1.10.0",
     )
     go_repository(
         name = "com_google_cloud_go_securitycenter",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/securitycenter",
         sum = "h1:QTVtk/Reqnx2bVIZtJKm1+mpfmwRwymmNvlaFez7fQY=",
         version = "v1.16.0",
     )
     go_repository(
         name = "com_google_cloud_go_servicecontrol",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/servicecontrol",
         sum = "h1:ImIzbOu6y4jL6ob65I++QzvqgFaoAKgHOG+RU9/c4y8=",
         version = "v1.5.0",
@@ -1528,24 +1634,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_servicedirectory",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/servicedirectory",
         sum = "h1:f7M8IMcVzO3T425AqlZbP3yLzeipsBHtRza8vVFYMhQ=",
         version = "v1.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_servicemanagement",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/servicemanagement",
         sum = "h1:TpkCO5M7dhKSy1bKUD9o/sSEW/U1Gtx7opA1fsiMx0c=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_serviceusage",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/serviceusage",
         sum = "h1:b0EwJxPJLpavSljMQh0RcdHsUrr5DQ+Nelt/3BAs5ro=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_shell",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/shell",
         sum = "h1:b1LFhFBgKsG252inyhtmsUUZwchqSz3WTvAIf3JFo4g=",
         version = "v1.4.0",
@@ -1553,6 +1663,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_speech",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/speech",
         sum = "h1:yK0ocnFH4Wsf0cMdUyndJQ/hPv02oTJOxzi6AgpBy4s=",
         version = "v1.9.0",
@@ -1566,6 +1677,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go_storagetransfer",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/storagetransfer",
         sum = "h1:fUe3OydbbvHcAYp07xY+2UpH4AermGbmnm7qdEj3tGE=",
         version = "v1.6.0",
@@ -1573,36 +1685,42 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_talent",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/talent",
         sum = "h1:MrekAGxLqAeAol4Sc0allOVqUGO8j+Iim8NMvpiD7tM=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_texttospeech",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/texttospeech",
         sum = "h1:ccPiHgTewxgyAeCWgQWvZvrLmbfQSFABTMAfrSPLPyY=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_tpu",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/tpu",
         sum = "h1:ztIdKoma1Xob2qm6QwNh4Xi9/e7N3IfvtwG5AcNsj1g=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_trace",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/trace",
         sum = "h1:qO9eLn2esajC9sxpqp1YKX37nXC3L4BfGnPS0Cx9dYo=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_translate",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/translate",
         sum = "h1:AOYOH3MspzJ/bH1YXzB+xTE8fMpn3mwhLjugwGXvMPI=",
         version = "v1.4.0",
     )
     go_repository(
         name = "com_google_cloud_go_video",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/video",
         sum = "h1:ttlvO4J5c1VGq6FkHqWPD/aH6PfdxujHt+muTJlW1Zk=",
         version = "v1.9.0",
@@ -1610,24 +1728,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_videointelligence",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/videointelligence",
         sum = "h1:RPFgVVXbI2b5vnrciZjtsUgpNKVtHO/WIyXUhEfuMhA=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_vision_v2",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/vision/v2",
         sum = "h1:TQHxRqvLMi19azwm3qYuDbEzZWmiKJNTpGbkNsfRCik=",
         version = "v2.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_vmmigration",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/vmmigration",
         sum = "h1:A2Tl2ZmwMRpvEmhV2ibISY85fmQR+Y5w9a0PlRz5P3s=",
         version = "v1.3.0",
     )
     go_repository(
         name = "com_google_cloud_go_vpcaccess",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/vpcaccess",
         sum = "h1:woHXXtnW8b9gLFdWO9HLPalAddBQ9V4LT+1vjKwR3W8=",
         version = "v1.5.0",
@@ -1635,12 +1757,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_webrisk",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/webrisk",
         sum = "h1:ypSnpGlJnZSXbN9a13PDmAYvVekBLnGKxQ3Q9SMwnYY=",
         version = "v1.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_websecurityscanner",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/websecurityscanner",
         sum = "h1:y7yIFg/h/mO+5Y5aCOtVAnpGUOgqCH5rXQ2Oc8Oq2+g=",
         version = "v1.4.0",
@@ -1648,6 +1772,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_workflows",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/workflows",
         sum = "h1:7Chpin9p50NTU8Tb7qk+I11U/IwVXmDhEoSsdccvInE=",
         version = "v1.9.0",
@@ -1668,6 +1793,7 @@ def go_repositories():

     go_repository(
         name = "in_gopkg_check_v1",
+        build_file_proto_mode = "disable_global",
         importpath = "gopkg.in/check.v1",
         sum = "h1:qIbj1fsPNlZgppZ+VLlY7N33q108Sa+fhmuc+sWQYwY=",
         version = "v1.0.0-20180628173108-788fd7840127",
@@ -1700,6 +1826,7 @@ def go_repositories():
     )
     go_repository(
         name = "in_gopkg_yaml_v3",
+        build_file_proto_mode = "disable_global",
         importpath = "gopkg.in/yaml.v3",
         sum = "h1:fxVm/GzAzEWqLHuvctI91KS9hhNmmWOoWu0XTYJS7CA=",
         version = "v3.0.1",
@@ -1750,18 +1877,21 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_google_appengine",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/appengine",
         sum = "h1:FZR1q0exgwxzPzp/aF+VccGrSfxfPpkBqjIIEq3ru6c=",
         version = "v1.6.7",
     )
     go_repository(
         name = "org_golang_google_genproto",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/genproto",
         sum = "h1:jCw9YRd2s40X9Vxi4zKsPRvSPlHWNqadVkpbMsCPzPQ=",
         version = "v0.0.0-20221114212237-e4508ebdbee1",
     )
     go_repository(
         name = "org_golang_google_grpc",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/grpc",
         sum = "h1:DS/BukOZWp8s6p4Dt/tOaJaTQyPyOoCcrjroHuCeLzY=",
         version = "v1.50.1",
@@ -1775,12 +1905,14 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_google_protobuf",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/protobuf",
         sum = "h1:d0NfwRgPtno5B1Wa6L2DAG+KivqkdutMf1UhdNx175w=",
         version = "v1.28.1",
     )
     go_repository(
         name = "org_golang_x_crypto",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/crypto",
         sum = "h1:VklqNMn3ovrHsnt90PveolxSbWFaJdECFbxSq0Mqo2M=",
         version = "v0.0.0-20190308221718-c2843e01d9a2",
@@ -1813,18 +1945,21 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_x_mod",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/mod",
         sum = "h1:6zppjxzCulZykYSLyVDYbneBfbaBIQPYMevg0bEwv2s=",
         version = "v0.6.0-dev.0.20220419223038-86c51ed26bb4",
     )
     go_repository(
         name = "org_golang_x_net",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/net",
         sum = "h1:sZfSu1wtKLGlWI4ZZayP0ck9Y73K1ynO6gqzTdBVdPU=",
         version = "v0.2.0",
     )
     go_repository(
         name = "org_golang_x_oauth2",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/oauth2",
         sum = "h1:GtQkldQ9m7yvzCL1V+LrYow3Khe0eJH0w7RbX/VbaIU=",
         version = "v0.2.0",
@@ -1837,12 +1972,14 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_x_sys",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/sys",
         sum = "h1:ljd4t30dBnAvMZaQCevtY0xLLD0A+bRZXbgLMLU1F/A=",
         version = "v0.2.0",
     )
     go_repository(
         name = "org_golang_x_term",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/term",
         sum = "h1:z85xZCsEl7bi/KwbNADeBYoOP0++7W1ipu+aGnpwzRM=",
         version = "v0.2.0",
@@ -1850,6 +1987,7 @@ def go_repositories():

     go_repository(
         name = "org_golang_x_text",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/text",
         sum = "h1:BrVqGRd7+k1DiOgtnFvAkoQEWQvBc25ouMJM6429SFg=",
         version = "v0.4.0",
@@ -1863,6 +2001,7 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_x_tools",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/tools",
         sum = "h1:VveCTK38A2rkS8ZqFY25HIDFscX5X9OoEhJd3quQmXU=",
         version = "v0.1.12",
@@ -1870,6 +2009,7 @@ def go_repositories():

     go_repository(
         name = "org_golang_x_xerrors",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/xerrors",
         sum = "h1:go1bK/D/BFZV2I8cIQd1NKEZ+0owSTG1fDTci4IqFcE=",
         version = "v0.0.0-20200804184101-5ec99f83aff1",
diff --git a/runtime/BUILD.bazel b/runtime/BUILD.bazel
index b5140a3..5ad24a9 100644
--- a/runtime/BUILD.bazel
+++ b/runtime/BUILD.bazel
@@ -1,7 +1,5 @@
 load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")

-package(default_visibility = ["//visibility:public"])
-
 go_library(
     name = "runtime",
     srcs = [
@@ -23,11 +21,12 @@ go_library(
         "query.go",
     ],
     importpath = "github.com/grpc-ecosystem/grpc-gateway/v2/runtime",
+    visibility = ["//visibility:public"],
     deps = [
         "//internal/httprule",
         "//utilities",
-        "@go_googleapis//google/api:httpbody_go_proto",
-        "@io_bazel_rules_go//proto/wkt:field_mask_go_proto",
+        "@org_golang_google_genproto//googleapis/api/httpbody",
+        "@org_golang_google_genproto//protobuf/field_mask",
         "@org_golang_google_grpc//codes",
         "@org_golang_google_grpc//grpclog",
         "@org_golang_google_grpc//health/grpc_health_v1",
@@ -44,9 +43,14 @@ go_library(
     ],
 )

+alias(
+    name = "go_default_library",
+    actual = ":runtime",
+    visibility = ["//visibility:public"],
+)
+
 go_test(
     name = "runtime_test",
-    size = "small",
     srcs = [
         "context_test.go",
         "convert_test.go",
@@ -70,11 +74,11 @@ go_test(
         "//utilities",
         "@com_github_google_go_cmp//cmp",
         "@com_github_google_go_cmp//cmp/cmpopts",
-        "@go_googleapis//google/api:httpbody_go_proto",
-        "@go_googleapis//google/rpc:errdetails_go_proto",
-        "@go_googleapis//google/rpc:status_go_proto",
-        "@io_bazel_rules_go//proto/wkt:field_mask_go_proto",
-        "@org_golang_google_grpc//:go_default_library",
+        "@org_golang_google_genproto//googleapis/api/httpbody",
+        "@org_golang_google_genproto//googleapis/rpc/errdetails",
+        "@org_golang_google_genproto//googleapis/rpc/status",
+        "@org_golang_google_genproto//protobuf/field_mask",
+        "@org_golang_google_grpc//:grpc",
         "@org_golang_google_grpc//codes",
         "@org_golang_google_grpc//health/grpc_health_v1",
         "@org_golang_google_grpc//metadata",
@@ -89,9 +93,3 @@ go_test(
         "@org_golang_google_protobuf//types/known/wrapperspb",
     ],
 )
-
-alias(
-    name = "go_default_library",
-    actual = ":runtime",
-    visibility = ["//visibility:public"],
-)

