diff --git a/repositories.bzl b/repositories.bzl
index b165cdd..2ff7081 100644
--- a/repositories.bzl
+++ b/repositories.bzl
@@ -22,6 +22,7 @@ def go_repositories():
 
     go_repository(
         name = "com_github_antihax_optional",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/antihax/optional",
         sum = "h1:xK2lYat7ZLaVVcIuj82J8kIro4V6kDe0AUDFboUCwcg=",
         version = "v1.0.0",
@@ -90,6 +91,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_census_instrumentation_opencensus_proto",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/census-instrumentation/opencensus-proto",
         sum = "h1:iKLQ0xPNFxR/2hzXZMrBo8f1j86j5WHzznCCQxV/b8g=",
         version = "v0.4.1",
@@ -102,6 +104,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_cespare_xxhash_v2",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/cespare/xxhash/v2",
         sum = "h1:DC2CZ1Ep5Y4k3ZQ899DldepgrayRUGE6BBZ/cd9Cj44=",
         version = "v2.2.0",
@@ -134,12 +137,14 @@ def go_repositories():
 
     go_repository(
         name = "com_github_cncf_udpa_go",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/cncf/udpa/go",
         sum = "h1:QQ3GSy+MqSHxm/d8nCtnAiZdYFd45cYZPs8vOOIYKfk=",
         version = "v0.0.0-20220112060539-c52dc94e7fbe",
     )
     go_repository(
         name = "com_github_cncf_xds_go",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/cncf/xds/go",
         sum = "h1:58f1tJ1ra+zFINPlwLWvQsR9CzAKt2e+EWV2yX9oXQ4=",
         version = "v0.0.0-20230310173818-32f1caf87195",
@@ -183,6 +188,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_creack_pty",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/creack/pty",
         sum = "h1:uDmaGzcdjhF4i/plgjmEsriH11Y0o7RKapEf/LDaM3w=",
         version = "v1.1.9",
@@ -209,12 +215,14 @@ def go_repositories():
 
     go_repository(
         name = "com_github_envoyproxy_go_control_plane",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/envoyproxy/go-control-plane",
         sum = "h1:jtLewhRR2vMRNnq2ZZUoCjUlgut+Y0+sDDWPOfwOi1o=",
         version = "v0.11.0",
     )
     go_repository(
         name = "com_github_envoyproxy_protoc_gen_validate",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/envoyproxy/protoc-gen-validate",
         sum = "h1:oIfnZFdC0YhpNNEX+SuIqko4cqqVZeN9IGTrhZje83Y=",
         version = "v0.10.0",
@@ -290,6 +298,7 @@ def go_repositories():
 
     go_repository(
         name = "com_github_golang_glog",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/golang/glog",
         sum = "h1:/d3pCKDPWNnvIWe0vVUpNP32qc8U3PDVxySP/y360qE=",
         version = "v1.1.0",
@@ -309,6 +318,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_golang_protobuf",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/golang/protobuf",
         sum = "h1:KhyjKVUg7Usr/dYsdSqoFveMYd5ko72D+zANwlG1mmg=",
         version = "v1.5.3",
@@ -322,6 +332,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_google_go_cmp",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/google/go-cmp",
         sum = "h1:O2Tfq5qg4qc4AmwVlvv0oLiVAGB7enBSJ2x2DqQFi38=",
         version = "v0.5.9",
@@ -353,6 +364,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_google_uuid",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/google/uuid",
         sum = "h1:t6JiXgmwXMjEs8VusXIJk2BXHsn+wx8BZdTaoZ5fu7I=",
         version = "v1.3.0",
@@ -599,18 +611,21 @@ def go_repositories():
 
     go_repository(
         name = "com_github_kr_pretty",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/kr/pretty",
         sum = "h1:flRD4NNwYAUpkphVc1HcthR4KEIFJ65n8Mw5qdRn3LE=",
         version = "v0.3.1",
     )
     go_repository(
         name = "com_github_kr_pty",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/kr/pty",
         sum = "h1:VkoXIwSboBpnk99O/KFauAEILuNHv5DVFKZMBN/gUgw=",
         version = "v1.1.1",
     )
     go_repository(
         name = "com_github_kr_text",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/kr/text",
         sum = "h1:5Nx0Ya0ZqY2ygV366QzturHI13Jq95ApcVaJBhpS+AY=",
         version = "v0.2.0",
@@ -731,6 +746,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_pkg_diff",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/pkg/diff",
         sum = "h1:aoZm08cpOy4WuID//EZDgcC4zIxODThtZNPirFr42+A=",
         version = "v0.0.0-20210226163009-20ebb0f2a09e",
@@ -795,6 +811,7 @@ def go_repositories():
 
     go_repository(
         name = "com_github_rogpeppe_fastuuid",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/rogpeppe/fastuuid",
         sum = "h1:Ppwyp6VYCF1nvBTXL3trRso7mXMlRrw9ooo375wvi2s=",
         version = "v1.2.0",
@@ -802,6 +819,7 @@ def go_repositories():
 
     go_repository(
         name = "com_github_rogpeppe_go_internal",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/rogpeppe/go-internal",
         sum = "h1:73kH8U+JUqXU8lRuOHeVHaa/SZPifC7BkcraZVejAe8=",
         version = "v1.9.0",
@@ -942,18 +960,21 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go",
         sum = "h1:Zc8gqp3+a9/Eyph2KDmcGaPtbKRIoqq4YTlL4NMD0Ys=",
         version = "v0.110.0",
     )
     go_repository(
         name = "com_google_cloud_go_accessapproval",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/accessapproval",
         sum = "h1:x0cEHro/JFPd7eS4BlEWNTMecIj2HdXjOVB5BtvwER0=",
         version = "v1.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_accesscontextmanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/accesscontextmanager",
         sum = "h1:MG60JgnEoawHJrbWw0jGdv6HLNSf6gQvYRiXpuzqgEA=",
         version = "v1.7.0",
@@ -961,6 +982,7 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_aiplatform",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/aiplatform",
         sum = "h1:zTw+suCVchgZyO+k847wjzdVjWmrAuehxdvcZvJwfGg=",
         version = "v1.37.0",
@@ -968,24 +990,28 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_analytics",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/analytics",
         sum = "h1:LqAo3tAh2FU9+w/r7vc3hBjU23Kv7GhO/PDIW7kIYgM=",
         version = "v0.19.0",
     )
     go_repository(
         name = "com_google_cloud_go_apigateway",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/apigateway",
         sum = "h1:ZI9mVO7x3E9RK/BURm2p1aw9YTBSCQe3klmyP1WxWEg=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_apigeeconnect",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/apigeeconnect",
         sum = "h1:sWOmgDyAsi1AZ48XRHcATC0tsi9SkPT7DA/+VCfkaeA=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_apigeeregistry",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/apigeeregistry",
         sum = "h1:E43RdhhCxdlV+I161gUY2rI4eOaMzHTA5kNkvRsFXvc=",
         version = "v0.6.0",
@@ -999,6 +1025,7 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_appengine",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/appengine",
         sum = "h1:aBGDKmRIaRRoWJ2tAoN0oVSHoWLhtO9aj/NvUyP4aYs=",
         version = "v1.7.1",
@@ -1006,12 +1033,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_area120",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/area120",
         sum = "h1:ugckkFh4XkHJMPhTIx0CyvdoBxmOpMe8rNs4Ok8GAag=",
         version = "v0.7.1",
     )
     go_repository(
         name = "com_google_cloud_go_artifactregistry",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/artifactregistry",
         sum = "h1:o1Q80vqEB6Qp8WLEH3b8FBLNUCrGQ4k5RFj0sn/sgO8=",
         version = "v1.13.0",
@@ -1019,36 +1048,42 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_asset",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/asset",
         sum = "h1:YAsssO08BqZ6mncbb6FPlj9h6ACS7bJQUOlzciSfbNk=",
         version = "v1.13.0",
     )
     go_repository(
         name = "com_google_cloud_go_assuredworkloads",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/assuredworkloads",
         sum = "h1:VLGnVFta+N4WM+ASHbhc14ZOItOabDLH1MSoDv+Xuag=",
         version = "v1.10.0",
     )
     go_repository(
         name = "com_google_cloud_go_automl",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/automl",
         sum = "h1:50VugllC+U4IGl3tDNcZaWvApHBTrn/TvyHDJ0wM+Uw=",
         version = "v1.12.0",
     )
     go_repository(
         name = "com_google_cloud_go_baremetalsolution",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/baremetalsolution",
         sum = "h1:2AipdYXL0VxMboelTTw8c1UJ7gYu35LZYUbuRv9Q28s=",
         version = "v0.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_batch",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/batch",
         sum = "h1:YbMt0E6BtqeD5FvSv1d56jbVsWEzlGm55lYte+M6Mzs=",
         version = "v0.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_beyondcorp",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/beyondcorp",
         sum = "h1:UkY2BTZkEUAVrgqnSdOJ4p3y9ZRBPEe1LkjgC8Bj/Pc=",
         version = "v0.5.0",
@@ -1056,42 +1091,49 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_bigquery",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/bigquery",
         sum = "h1:RscMV6LbnAmhAzD893Lv9nXXy2WCaJmbxYPWDLbGqNQ=",
         version = "v1.50.0",
     )
     go_repository(
         name = "com_google_cloud_go_billing",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/billing",
         sum = "h1:JYj28UYF5w6VBAh0gQYlgHJ/OD1oA+JgW29YZQU+UHM=",
         version = "v1.13.0",
     )
     go_repository(
         name = "com_google_cloud_go_binaryauthorization",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/binaryauthorization",
         sum = "h1:d3pMDBCCNivxt5a4eaV7FwL7cSH0H7RrEnFrTb1QKWs=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_certificatemanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/certificatemanager",
         sum = "h1:5C5UWeSt8Jkgp7OWn2rCkLmYurar/vIWIoSQ2+LaTOc=",
         version = "v1.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_channel",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/channel",
         sum = "h1:GpcQY5UJKeOekYgsX3QXbzzAc/kRGtBq43fTmyKe6Uw=",
         version = "v1.12.0",
     )
     go_repository(
         name = "com_google_cloud_go_cloudbuild",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/cloudbuild",
         sum = "h1:GHQCjV4WlPPVU/j3Rlpc8vNIDwThhd1U9qSY/NPZdko=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_clouddms",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/clouddms",
         sum = "h1:E7v4TpDGUyEm1C/4KIrpVSOCTm0P6vWdHT0I4mostRA=",
         version = "v1.5.0",
@@ -1099,6 +1141,7 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_cloudtasks",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/cloudtasks",
         sum = "h1:uK5k6abf4yligFgYFnG0ni8msai/dSv6mDmiBulU0hU=",
         version = "v1.10.0",
@@ -1106,12 +1149,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_compute",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/compute",
         sum = "h1:+9zda3WGgW1ZSTlVppLCYFIr48Pa35q1uG2N1itbCEQ=",
         version = "v1.19.0",
     )
     go_repository(
         name = "com_google_cloud_go_compute_metadata",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/compute/metadata",
         sum = "h1:mg4jlk7mCAj6xXp9UJ4fjI9VUI5rubuGBW5aJ7UnBMY=",
         version = "v0.2.3",
@@ -1119,12 +1164,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_contactcenterinsights",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/contactcenterinsights",
         sum = "h1:jXIpfcH/VYSE1SYcPzO0n1VVb+sAamiLOgCw45JbOQk=",
         version = "v1.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_container",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/container",
         sum = "h1:NKlY/wCDapfVZlbVVaeuu2UZZED5Dy1z4Zx1KhEzm8c=",
         version = "v1.15.0",
@@ -1132,6 +1179,7 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_containeranalysis",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/containeranalysis",
         sum = "h1:EQ4FFxNaEAg8PqQCO7bVQfWz9NVwZCUKaM1b3ycfx3U=",
         version = "v0.9.0",
@@ -1139,24 +1187,28 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_datacatalog",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datacatalog",
         sum = "h1:4H5IJiyUE0X6ShQBqgFFZvGGcrwGVndTwUSLP4c52gw=",
         version = "v1.13.0",
     )
     go_repository(
         name = "com_google_cloud_go_dataflow",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataflow",
         sum = "h1:eYyD9o/8Nm6EttsKZaEGD84xC17bNgSKCu0ZxwqUbpg=",
         version = "v0.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_dataform",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataform",
         sum = "h1:Dyk+fufup1FR6cbHjFpMuP4SfPiF3LI3JtoIIALoq48=",
         version = "v0.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_datafusion",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datafusion",
         sum = "h1:sZjRnS3TWkGsu1LjYPFD/fHeMLZNXDK6PDHi2s2s/bk=",
         version = "v1.6.0",
@@ -1164,18 +1216,21 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_datalabeling",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datalabeling",
         sum = "h1:ch4qA2yvddGRUrlfwrNJCr79qLqhS9QBwofPHfFlDIk=",
         version = "v0.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_dataplex",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataplex",
         sum = "h1:RvoZ5T7gySwm1CHzAw7yY1QwwqaGswunmqEssPxU/AM=",
         version = "v1.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_dataproc",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataproc",
         sum = "h1:W47qHL3W4BPkAIbk4SWmIERwsWBaNnWm0P2sdx3YgGU=",
         version = "v1.12.0",
@@ -1183,6 +1238,7 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_dataqna",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataqna",
         sum = "h1:yFzi/YU4YAdjyo7pXkBE2FeHbgz5OQQBVDdbErEHmVQ=",
         version = "v0.7.0",
@@ -1190,18 +1246,21 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_datastore",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datastore",
         sum = "h1:iF6I/HaLs3Ado8uRKMvZRvF/ZLkWaWE9i8AiHzbC774=",
         version = "v1.11.0",
     )
     go_repository(
         name = "com_google_cloud_go_datastream",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datastream",
         sum = "h1:BBCBTnWMDwwEzQQmipUXxATa7Cm7CA/gKjKcR2w35T0=",
         version = "v1.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_deploy",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/deploy",
         sum = "h1:otshdKEbmsi1ELYeCKNYppwV0UH5xD05drSdBm7ouTk=",
         version = "v1.8.0",
@@ -1209,12 +1268,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_dialogflow",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dialogflow",
         sum = "h1:uVlKKzp6G/VtSW0E7IH1Y5o0H48/UOCmqksG2riYCwQ=",
         version = "v1.32.0",
     )
     go_repository(
         name = "com_google_cloud_go_dlp",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dlp",
         sum = "h1:1JoJqezlgu6NWCroBxr4rOZnwNFILXr4cB9dMaSKO4A=",
         version = "v1.9.0",
@@ -1222,12 +1283,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_documentai",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/documentai",
         sum = "h1:KM3Xh0QQyyEdC8Gs2vhZfU+rt6OCPF0dwVwxKgLmWfI=",
         version = "v1.18.0",
     )
     go_repository(
         name = "com_google_cloud_go_domains",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/domains",
         sum = "h1:2ti/o9tlWL4N+wIuWUNH+LbfgpwxPr8J1sv9RHA4bYQ=",
         version = "v0.8.0",
@@ -1235,12 +1298,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_edgecontainer",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/edgecontainer",
         sum = "h1:O0YVE5v+O0Q/ODXYsQHmHb+sYM8KNjGZw2pjX2Ws41c=",
         version = "v1.0.0",
     )
     go_repository(
         name = "com_google_cloud_go_errorreporting",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/errorreporting",
         sum = "h1:kj1XEWMu8P0qlLhm3FwcaFsUvXChV/OraZwA70trRR0=",
         version = "v0.3.0",
@@ -1248,18 +1313,21 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_essentialcontacts",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/essentialcontacts",
         sum = "h1:gIzEhCoOT7bi+6QZqZIzX1Erj4SswMPIteNvYVlu+pM=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_eventarc",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/eventarc",
         sum = "h1:fsJmNeqvqtk74FsaVDU6cH79lyZNCYP8Rrv7EhaB/PU=",
         version = "v1.11.0",
     )
     go_repository(
         name = "com_google_cloud_go_filestore",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/filestore",
         sum = "h1:ckTEXN5towyTMu4q0uQ1Mde/JwTHur0gXs8oaIZnKfw=",
         version = "v1.6.0",
@@ -1267,24 +1335,28 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_firestore",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/firestore",
         sum = "h1:IBlRyxgGySXu5VuW0RgGFlTtLukSnNkpDiEOMkQkmpA=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_functions",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/functions",
         sum = "h1:pPDqtsXG2g9HeOQLoquLbmvmb82Y4Ezdo1GXuotFoWg=",
         version = "v1.13.0",
     )
     go_repository(
         name = "com_google_cloud_go_gaming",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gaming",
         sum = "h1:7vEhFnZmd931Mo7sZ6pJy7uQPDxF7m7v8xtBheG08tc=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_gkebackup",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkebackup",
         sum = "h1:za3QZvw6ujR0uyqkhomKKKNoXDyqYGPJies3voUK8DA=",
         version = "v0.4.0",
@@ -1292,54 +1364,63 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_gkeconnect",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkeconnect",
         sum = "h1:gXYKciHS/Lgq0GJ5Kc9SzPA35NGc3yqu6SkjonpEr2Q=",
         version = "v0.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_gkehub",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkehub",
         sum = "h1:TqCSPsEBQ6oZSJgEYZ3XT8x2gUadbvfwI32YB0kuHCs=",
         version = "v0.12.0",
     )
     go_repository(
         name = "com_google_cloud_go_gkemulticloud",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkemulticloud",
         sum = "h1:8I84Q4vl02rJRsFiinBxl7WCozfdLlUVBQuSrqr9Wtk=",
         version = "v0.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_gsuiteaddons",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gsuiteaddons",
         sum = "h1:1mvhXqJzV0Vg5Fa95QwckljODJJfDFXV4pn+iL50zzA=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_iam",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/iam",
         sum = "h1:+CmB+K0J/33d0zSQ9SlFWUeCCEn5XJA0ZMZ3pHE9u8k=",
         version = "v0.13.0",
     )
     go_repository(
         name = "com_google_cloud_go_iap",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/iap",
         sum = "h1:PxVHFuMxmSZyfntKXHXhd8bo82WJ+LcATenq7HLdVnU=",
         version = "v1.7.1",
     )
     go_repository(
         name = "com_google_cloud_go_ids",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/ids",
         sum = "h1:fodnCDtOXuMmS8LTC2y3h8t24U8F3eKWfhi+3LY6Qf0=",
         version = "v1.3.0",
     )
     go_repository(
         name = "com_google_cloud_go_iot",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/iot",
         sum = "h1:39W5BFSarRNZfVG0eXI5LYux+OVQT8GkgpHCnrZL2vM=",
         version = "v1.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_kms",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/kms",
         sum = "h1:7hm1bRqGCA1GBRQUrp831TwJ9TWhP+tvLuP497CQS2g=",
         version = "v1.10.1",
@@ -1347,18 +1428,21 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_language",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/language",
         sum = "h1:7Ulo2mDk9huBoBi8zCE3ONOoBrL6UXfAI71CLQ9GEIM=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_lifesciences",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/lifesciences",
         sum = "h1:uWrMjWTsGjLZpCTWEAzYvyXj+7fhiZST45u9AgasasI=",
         version = "v0.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_logging",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/logging",
         sum = "h1:CJYxlNNNNAMkHp9em/YEXcfJg+rPDg7YfwoRpMU+t5I=",
         version = "v1.7.0",
@@ -1366,18 +1450,21 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_longrunning",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/longrunning",
         sum = "h1:v+yFJOfKC3yZdY6ZUI933pIYdhyhV8S3NpWrXWmg7jM=",
         version = "v0.4.1",
     )
     go_repository(
         name = "com_google_cloud_go_managedidentities",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/managedidentities",
         sum = "h1:ZRQ4k21/jAhrHBVKl/AY7SjgzeJwG1iZa+mJ82P+VNg=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_maps",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/maps",
         sum = "h1:mv9YaczD4oZBZkM5XJl6fXQ984IkJNHPwkc8MUsdkBo=",
         version = "v0.7.0",
@@ -1385,24 +1472,28 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_mediatranslation",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/mediatranslation",
         sum = "h1:anPxH+/WWt8Yc3EdoEJhPMBRF7EhIdz426A+tuoA0OU=",
         version = "v0.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_memcache",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/memcache",
         sum = "h1:8/VEmWCpnETCrBwS3z4MhT+tIdKgR1Z4Tr2tvYH32rg=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_metastore",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/metastore",
         sum = "h1:QCFhZVe2289KDBQ7WxaHV2rAmPrmRAdLC6gbjUd3HPo=",
         version = "v1.10.0",
     )
     go_repository(
         name = "com_google_cloud_go_monitoring",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/monitoring",
         sum = "h1:2qsrgXGVoRXpP7otZ14eE1I568zAa92sJSDPyOJvwjM=",
         version = "v1.13.0",
@@ -1410,12 +1501,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_networkconnectivity",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/networkconnectivity",
         sum = "h1:ZD6b4Pk1jEtp/cx9nx0ZYcL3BKqDa+KixNDZ6Bjs1B8=",
         version = "v1.11.0",
     )
     go_repository(
         name = "com_google_cloud_go_networkmanagement",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/networkmanagement",
         sum = "h1:8KWEUNGcpSX9WwZXq7FtciuNGPdPdPN/ruDm769yAEM=",
         version = "v1.6.0",
@@ -1423,30 +1516,35 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_networksecurity",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/networksecurity",
         sum = "h1:sOc42Ig1K2LiKlzG71GUVloeSJ0J3mffEBYmvu+P0eo=",
         version = "v0.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_notebooks",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/notebooks",
         sum = "h1:Kg2K3K7CbSXYJHZ1aGQpf1xi5x2GUvQWf2sFVuiZh8M=",
         version = "v1.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_optimization",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/optimization",
         sum = "h1:dj8O4VOJRB4CUwZXdmwNViH1OtI0WtWL867/lnYH248=",
         version = "v1.3.1",
     )
     go_repository(
         name = "com_google_cloud_go_orchestration",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/orchestration",
         sum = "h1:Vw+CEXo8M/FZ1rb4EjcLv0gJqqw89b7+g+C/EmniTb8=",
         version = "v1.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_orgpolicy",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/orgpolicy",
         sum = "h1:XDriMWug7sd0kYT1QKofRpRHzjad0bK8Q8uA9q+XrU4=",
         version = "v1.10.0",
@@ -1454,24 +1552,28 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_osconfig",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/osconfig",
         sum = "h1:PkSQx4OHit5xz2bNyr11KGcaFccL5oqglFPdTboyqwQ=",
         version = "v1.11.0",
     )
     go_repository(
         name = "com_google_cloud_go_oslogin",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/oslogin",
         sum = "h1:whP7vhpmc+ufZa90eVpkfbgzJRK/Xomjz+XCD4aGwWw=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_phishingprotection",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/phishingprotection",
         sum = "h1:l6tDkT7qAEV49MNEJkEJTB6vOO/onbSOcNtAT09HPuA=",
         version = "v0.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_policytroubleshooter",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/policytroubleshooter",
         sum = "h1:yKAGC4p9O61ttZUswaq9GAn1SZnEzTd0vUYXD7ZBT7Y=",
         version = "v1.6.0",
@@ -1479,6 +1581,7 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_privatecatalog",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/privatecatalog",
         sum = "h1:EPEJ1DpEGXLDnmc7mnCAqFmkwUJbIsaLAiLHVOkkwtc=",
         version = "v0.8.0",
@@ -1486,12 +1589,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_pubsub",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/pubsub",
         sum = "h1:vCge8m7aUKBJYOgrZp7EsNDf6QMd2CAlXZqWTn3yq6s=",
         version = "v1.30.0",
     )
     go_repository(
         name = "com_google_cloud_go_pubsublite",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/pubsublite",
         sum = "h1:cb9fsrtpINtETHiJ3ECeaVzrfIVhcGjhhJEjybHXHao=",
         version = "v1.7.0",
@@ -1499,36 +1604,42 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_recaptchaenterprise_v2",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/recaptchaenterprise/v2",
         sum = "h1:6iOCujSNJ0YS7oNymI64hXsjGq60T4FK1zdLugxbzvU=",
         version = "v2.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_recommendationengine",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/recommendationengine",
         sum = "h1:VibRFCwWXrFebEWKHfZAt2kta6pS7Tlimsnms0fjv7k=",
         version = "v0.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_recommender",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/recommender",
         sum = "h1:ZnFRY5R6zOVk2IDS1Jbv5Bw+DExCI5rFumsTnMXiu/A=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_redis",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/redis",
         sum = "h1:JoAd3SkeDt3rLFAAxEvw6wV4t+8y4ZzfZcZmddqphQ8=",
         version = "v1.11.0",
     )
     go_repository(
         name = "com_google_cloud_go_resourcemanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/resourcemanager",
         sum = "h1:NRM0p+RJkaQF9Ee9JMnUV9BQ2QBIOq/v8M+Pbv/wmCs=",
         version = "v1.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_resourcesettings",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/resourcesettings",
         sum = "h1:8Dua37kQt27CCWHm4h/Q1XqCF6ByD7Ouu49xg95qJzI=",
         version = "v1.5.0",
@@ -1536,12 +1647,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_retail",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/retail",
         sum = "h1:1Dda2OpFNzIb4qWgFZjYlpP7sxX3aLeypKG6A3H4Yys=",
         version = "v1.12.0",
     )
     go_repository(
         name = "com_google_cloud_go_run",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/run",
         sum = "h1:ydJQo+k+MShYnBfhaRHSZYeD/SQKZzZLAROyfpeD9zw=",
         version = "v0.9.0",
@@ -1549,12 +1662,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_scheduler",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/scheduler",
         sum = "h1:NpQAHtx3sulByTLe2dMwWmah8PWgeoieFPpJpArwFV0=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_secretmanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/secretmanager",
         sum = "h1:pu03bha7ukxF8otyPKTFdDz+rr9sE3YauS5PliDXK60=",
         version = "v1.10.0",
@@ -1562,12 +1677,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_security",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/security",
         sum = "h1:PYvDxopRQBfYAXKAuDpFCKBvDOWPWzp9k/H5nB3ud3o=",
         version = "v1.13.0",
     )
     go_repository(
         name = "com_google_cloud_go_securitycenter",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/securitycenter",
         sum = "h1:AF3c2s3awNTMoBtMX3oCUoOMmGlYxGOeuXSYHNBkf14=",
         version = "v1.19.0",
@@ -1581,6 +1698,7 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_servicedirectory",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/servicedirectory",
         sum = "h1:SJwk0XX2e26o25ObYUORXx6torSFiYgsGkWSkZgkoSU=",
         version = "v1.9.0",
@@ -1599,12 +1717,14 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go_shell",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/shell",
         sum = "h1:wT0Uw7ib7+AgZST9eCDygwTJn4+bHMDtZo5fh7kGWDU=",
         version = "v1.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_spanner",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/spanner",
         sum = "h1:7VdjZ8zj4sHbDw55atp5dfY6kn1j9sam9DRNpPQhqR4=",
         version = "v1.45.0",
@@ -1612,6 +1732,7 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_speech",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/speech",
         sum = "h1:JEVoWGNnTF128kNty7T4aG4eqv2z86yiMJPT9Zjp+iw=",
         version = "v1.15.0",
@@ -1625,6 +1746,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go_storagetransfer",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/storagetransfer",
         sum = "h1:5T+PM+3ECU3EY2y9Brv0Sf3oka8pKmsCfpQ07+91G9o=",
         version = "v1.8.0",
@@ -1632,36 +1754,42 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_talent",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/talent",
         sum = "h1:nI9sVZPjMKiO2q3Uu0KhTDVov3Xrlpt63fghP9XjyEM=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_texttospeech",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/texttospeech",
         sum = "h1:H4g1ULStsbVtalbZGktyzXzw6jP26RjVGYx9RaYjBzc=",
         version = "v1.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_tpu",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/tpu",
         sum = "h1:/34T6CbSi+kTv5E19Q9zbU/ix8IviInZpzwz3rsFE+A=",
         version = "v1.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_trace",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/trace",
         sum = "h1:olxC0QHC59zgJVALtgqfD9tGk0lfeCP5/AGXL3Px/no=",
         version = "v1.9.0",
     )
     go_repository(
         name = "com_google_cloud_go_translate",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/translate",
         sum = "h1:GvLP4oQ4uPdChBmBaUSa/SaZxCdyWELtlAaKzpHsXdA=",
         version = "v1.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_video",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/video",
         sum = "h1:upIbnGI0ZgACm58HPjAeBMleW3sl5cT84AbYQ8PWOgM=",
         version = "v1.15.0",
@@ -1669,24 +1797,28 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_videointelligence",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/videointelligence",
         sum = "h1:Uh5BdoET8XXqXX2uXIahGb+wTKbLkGH7s4GXR58RrG8=",
         version = "v1.10.0",
     )
     go_repository(
         name = "com_google_cloud_go_vision_v2",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/vision/v2",
         sum = "h1:8C8RXUJoflCI4yVdqhTy9tRyygSHmp60aP363z23HKg=",
         version = "v2.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_vmmigration",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/vmmigration",
         sum = "h1:Azs5WKtfOC8pxvkyrDvt7J0/4DYBch0cVbuFfCCFt5k=",
         version = "v1.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_vmwareengine",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/vmwareengine",
         sum = "h1:b0NBu7S294l0gmtrT0nOJneMYgZapr5x9tVWvgDoVEM=",
         version = "v0.3.0",
@@ -1694,6 +1826,7 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_vpcaccess",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/vpcaccess",
         sum = "h1:FOe6CuiQD3BhHJWt7E8QlbBcaIzVRddupwJlp7eqmn4=",
         version = "v1.6.0",
@@ -1701,12 +1834,14 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_webrisk",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/webrisk",
         sum = "h1:IY+L2+UwxcVm2zayMAtBhZleecdIFLiC+QJMzgb0kT0=",
         version = "v1.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_websecurityscanner",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/websecurityscanner",
         sum = "h1:AHC1xmaNMOZtNqxI9Rmm87IJEyPaRkOxeI0gpAacXGk=",
         version = "v1.5.0",
@@ -1714,6 +1849,7 @@ def go_repositories():
 
     go_repository(
         name = "com_google_cloud_go_workflows",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/workflows",
         sum = "h1:FfGp9w0cYnaKZJhUOMqCOJCYT/WlvYBfTQhFWV3sRKI=",
         version = "v1.10.0",
@@ -1734,6 +1870,7 @@ def go_repositories():
 
     go_repository(
         name = "in_gopkg_check_v1",
+        build_file_proto_mode = "disable_global",
         importpath = "gopkg.in/check.v1",
         sum = "h1:Hei/4ADfdWqJk1ZMxUNpqntNwaWcugrBjAiHlqqRiVk=",
         version = "v1.0.0-20201130134442-10cb98267c6c",
@@ -1766,6 +1903,7 @@ def go_repositories():
     )
     go_repository(
         name = "in_gopkg_yaml_v3",
+        build_file_proto_mode = "disable_global",
         importpath = "gopkg.in/yaml.v3",
         sum = "h1:fxVm/GzAzEWqLHuvctI91KS9hhNmmWOoWu0XTYJS7CA=",
         version = "v3.0.1",
@@ -1816,24 +1954,28 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_google_appengine",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/appengine",
         sum = "h1:FZR1q0exgwxzPzp/aF+VccGrSfxfPpkBqjIIEq3ru6c=",
         version = "v1.6.7",
     )
     go_repository(
         name = "org_golang_google_genproto",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/genproto",
         sum = "h1:Ao9GzfUMPH3zjVfzXG5rlWlk+Q8MXWKwWpwVQE1MXfw=",
         version = "v0.0.0-20230526203410-71b5a4ffd15e",
     )
     go_repository(
         name = "org_golang_google_genproto_googleapis_api",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/genproto/googleapis/api",
         sum = "h1:kVKPf/IiYSBWEWtkIn6wZXwWGCnLKcC8oWfZvXjsGnM=",
         version = "v0.0.0-20230530153820-e85fd2cbaebc",
     )
     go_repository(
         name = "org_golang_google_genproto_googleapis_rpc",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/genproto/googleapis/rpc",
         sum = "h1:XSJ8Vk1SWuNr8S18z1NZSziL0CPIXLCCMDOEFtHBOFc=",
         version = "v0.0.0-20230530153820-e85fd2cbaebc",
@@ -1841,6 +1983,7 @@ def go_repositories():
 
     go_repository(
         name = "org_golang_google_grpc",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/grpc",
         sum = "h1:3Oj82/tFSCeUrRTg/5E/7d/W5A1tj6Ky1ABAuZuv5ag=",
         version = "v1.55.0",
@@ -1854,12 +1997,14 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_google_protobuf",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/protobuf",
         sum = "h1:kPPoIgf3TsEvrm0PFe15JQ+570QVxYzEvvHqChK+cng=",
         version = "v1.30.0",
     )
     go_repository(
         name = "org_golang_x_crypto",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/crypto",
         sum = "h1:VklqNMn3ovrHsnt90PveolxSbWFaJdECFbxSq0Mqo2M=",
         version = "v0.0.0-20190308221718-c2843e01d9a2",
@@ -1892,18 +2037,21 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_x_mod",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/mod",
         sum = "h1:LUYupSeNrTNCGzR/hVBk2NHZO4hXcVaW1k4Qx7rjPx8=",
         version = "v0.8.0",
     )
     go_repository(
         name = "org_golang_x_net",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/net",
         sum = "h1:X2//UzNDwYmtCLn7To6G58Wr6f5ahEAQgKNzv9Y951M=",
         version = "v0.10.0",
     )
     go_repository(
         name = "org_golang_x_oauth2",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/oauth2",
         sum = "h1:6dkIjl3j3LtZ/O3sTgZTMsLKSftL/B8Zgq4huOIIUu8=",
         version = "v0.8.0",
@@ -1916,12 +2064,14 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_x_sys",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/sys",
         sum = "h1:EBmGv8NaZBZTWvrbjNoL6HVt+IVy3QDQpJs7VRIw3tU=",
         version = "v0.8.0",
     )
     go_repository(
         name = "org_golang_x_term",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/term",
         sum = "h1:n5xxQn2i3PC0yLAbjTpNT85q/Kgzcr2gIoX9OrJUols=",
         version = "v0.8.0",
@@ -1929,6 +2079,7 @@ def go_repositories():
 
     go_repository(
         name = "org_golang_x_text",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/text",
         sum = "h1:2sjJmO8cDvYveuX97RDLsxlyUxLl+GHoLxBiRdHllBE=",
         version = "v0.9.0",
@@ -1942,6 +2093,7 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_x_tools",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/tools",
         sum = "h1:BOw41kyTf3PuCW1pVQf8+Cyg8pMlkYB1oo9iJ6D/lKM=",
         version = "v0.6.0",
@@ -1949,6 +2101,7 @@ def go_repositories():
 
     go_repository(
         name = "org_golang_x_xerrors",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/xerrors",
         sum = "h1:E7g+9GITq07hpfrRu66IVDexMakfv52eLZ2CXBWiKr4=",
         version = "v0.0.0-20191204190536-9bdfabe68543",
diff --git a/runtime/BUILD.bazel b/runtime/BUILD.bazel
index a8789f1..c3c37af 100644
--- a/runtime/BUILD.bazel
+++ b/runtime/BUILD.bazel
@@ -1,7 +1,5 @@
 load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 
-package(default_visibility = ["//visibility:public"])
-
 go_library(
     name = "runtime",
     srcs = [
@@ -23,6 +21,7 @@ go_library(
         "query.go",
     ],
     importpath = "github.com/grpc-ecosystem/grpc-gateway/v2/runtime",
+    visibility = ["//visibility:public"],
     deps = [
         "//internal/httprule",
         "//utilities",
@@ -44,9 +43,14 @@ go_library(
     ],
 )
 
+alias(
+    name = "go_default_library",
+    actual = ":runtime",
+    visibility = ["//visibility:public"],
+)
+
 go_test(
     name = "runtime_test",
-    size = "small",
     srcs = [
         "context_test.go",
         "convert_test.go",
@@ -89,9 +93,3 @@ go_test(
         "@org_golang_google_protobuf//types/known/wrapperspb",
     ],
 )
-
-alias(
-    name = "go_default_library",
-    actual = ":runtime",
-    visibility = ["//visibility:public"],
-)
diff --git a/runtime/internal/examplepb/BUILD.bazel b/runtime/internal/examplepb/BUILD.bazel
index cfed28f..794faad 100644
--- a/runtime/internal/examplepb/BUILD.bazel
+++ b/runtime/internal/examplepb/BUILD.bazel
@@ -2,10 +2,6 @@ load("@rules_proto//proto:defs.bzl", "proto_library")
 load("@io_bazel_rules_go//go:def.bzl", "go_library")
 load("@io_bazel_rules_go//proto:def.bzl", "go_proto_library")
 
-# gazelle:exclude non_standard_names_grpc.pb.go
-
-package(default_visibility = ["//visibility:public"])
-
 proto_library(
     name = "examplepb_proto",
     srcs = [
@@ -14,6 +10,7 @@ proto_library(
         "proto2.proto",
         "proto3.proto",
     ],
+    visibility = ["//runtime:__subpackages__"],
     deps = [
         "@com_google_protobuf//:any_proto",
         "@com_google_protobuf//:duration_proto",
@@ -34,6 +31,7 @@ go_proto_library(
     ],
     importpath = "github.com/grpc-ecosystem/grpc-gateway/v2/runtime/internal/examplepb",
     proto = ":examplepb_proto",
+    visibility = ["//runtime:__subpackages__"],
     deps = ["@go_googleapis//google/api:annotations_go_proto"],
 )
 
@@ -41,6 +39,7 @@ go_library(
     name = "examplepb",
     embed = [":examplepb_go_proto"],
     importpath = "github.com/grpc-ecosystem/grpc-gateway/v2/runtime/internal/examplepb",
+    visibility = ["//runtime:__subpackages__"],
 )
 
 alias(
