diff --git a/repositories.bzl b/repositories.bzl
index e633e062..4cacbc6e 100644
--- a/repositories.bzl
+++ b/repositories.bzl
@@ -3,6 +3,7 @@ load("@bazel_gazelle//:deps.bzl", "go_repository")
 def go_repositories():
     go_repository(
         name = "co_honnef_go_tools",
+        build_file_proto_mode = "disable_global",
         importpath = "honnef.co/go/tools",
         sum = "h1:/hemPrYIhOhy8zYrNj+069zDB68us2sMGsfkFJO0iZs=",
         version = "v0.0.0-20190523083050-ea95bdfd59fc",
