diff --git a/repositories.bzl b/repositories.bzl
index e633e062..f4a71f96 100644
--- a/repositories.bzl
+++ b/repositories.bzl
@@ -22,6 +22,7 @@ def go_repositories():

     go_repository(
         name = "com_github_antihax_optional",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/antihax/optional",
         sum = "h1:xK2lYat7ZLaVVcIuj82J8kIro4V6kDe0AUDFboUCwcg=",
         version = "v1.0.0",
@@ -90,6 +91,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_census_instrumentation_opencensus_proto",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/census-instrumentation/opencensus-proto",
         sum = "h1:iKLQ0xPNFxR/2hzXZMrBo8f1j86j5WHzznCCQxV/b8g=",
         version = "v0.4.1",
@@ -102,6 +104,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_cespare_xxhash_v2",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/cespare/xxhash/v2",
         sum = "h1:DC2CZ1Ep5Y4k3ZQ899DldepgrayRUGE6BBZ/cd9Cj44=",
         version = "v2.2.0",
@@ -134,12 +137,14 @@ def go_repositories():

     go_repository(
         name = "com_github_cncf_udpa_go",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/cncf/udpa/go",
         sum = "h1:QQ3GSy+MqSHxm/d8nCtnAiZdYFd45cYZPs8vOOIYKfk=",
         version = "v0.0.0-20220112060539-c52dc94e7fbe",
     )
     go_repository(
         name = "com_github_cncf_xds_go",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/cncf/xds/go",
         sum = "h1:/inchEIKaYC1Akx+H+gqO04wryn5h75LSazbRlnya1k=",
         version = "v0.0.0-20230607035331-e9ce68804cb4",
@@ -209,12 +214,14 @@ def go_repositories():

     go_repository(
         name = "com_github_envoyproxy_go_control_plane",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/envoyproxy/go-control-plane",
         sum = "h1:7T++XKzy4xg7PKy+bM+Sa9/oe1OC88yz2hXQUISoXfA=",
         version = "v0.11.1-0.20230524094728-9239064ad72f",
     )
     go_repository(
         name = "com_github_envoyproxy_protoc_gen_validate",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/envoyproxy/protoc-gen-validate",
         sum = "h1:c0g45+xCJhdgFGw7a5QAfdS4byAbud7miNWJ1WwEVf8=",
         version = "v0.10.1",
@@ -290,6 +297,7 @@ def go_repositories():

     go_repository(
         name = "com_github_golang_glog",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/golang/glog",
         sum = "h1:/d3pCKDPWNnvIWe0vVUpNP32qc8U3PDVxySP/y360qE=",
         version = "v1.1.0",
@@ -309,6 +317,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_golang_protobuf",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/golang/protobuf",
         sum = "h1:KhyjKVUg7Usr/dYsdSqoFveMYd5ko72D+zANwlG1mmg=",
         version = "v1.5.3",
@@ -322,6 +331,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_google_go_cmp",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/google/go-cmp",
         sum = "h1:O2Tfq5qg4qc4AmwVlvv0oLiVAGB7enBSJ2x2DqQFi38=",
         version = "v0.5.9",
@@ -353,6 +363,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_github_google_uuid",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/google/uuid",
         sum = "h1:t6JiXgmwXMjEs8VusXIJk2BXHsn+wx8BZdTaoZ5fu7I=",
         version = "v1.3.0",
@@ -599,12 +610,14 @@ def go_repositories():

     go_repository(
         name = "com_github_kr_pretty",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/kr/pretty",
         sum = "h1:flRD4NNwYAUpkphVc1HcthR4KEIFJ65n8Mw5qdRn3LE=",
         version = "v0.3.1",
     )
     go_repository(
         name = "com_github_kr_pty",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/kr/pty",
         sum = "h1:VkoXIwSboBpnk99O/KFauAEILuNHv5DVFKZMBN/gUgw=",
         version = "v1.1.1",
@@ -795,6 +808,7 @@ def go_repositories():

     go_repository(
         name = "com_github_rogpeppe_fastuuid",
+        build_file_proto_mode = "disable_global",
         importpath = "github.com/rogpeppe/fastuuid",
         sum = "h1:Ppwyp6VYCF1nvBTXL3trRso7mXMlRrw9ooo375wvi2s=",
         version = "v1.2.0",
@@ -942,18 +956,21 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go",
         sum = "h1:1JYyxKMN9hd5dR2MYTPWkGUgcoxVVhg0LKNKEo0qvmk=",
         version = "v0.110.4",
     )
     go_repository(
         name = "com_google_cloud_go_accessapproval",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/accessapproval",
         sum = "h1:/5YjNhR6lzCvmJZAnByYkfEgWjfAKwYP6nkuTk6nKFE=",
         version = "v1.7.1",
     )
     go_repository(
         name = "com_google_cloud_go_accesscontextmanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/accesscontextmanager",
         sum = "h1:WIAt9lW9AXtqw/bnvrEUaE8VG/7bAAeMzRCBGMkc4+w=",
         version = "v1.8.1",
@@ -961,6 +978,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_aiplatform",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/aiplatform",
         sum = "h1:FLTOQdXDqigyOPYrGGE5AiTpDyRROIZrPU1eXfKzKTY=",
         version = "v1.45.0",
@@ -968,18 +986,21 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_analytics",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/analytics",
         sum = "h1:T400N/hkELka6OsgK20JYoit0xvKnZtWoe36ft4wGBs=",
         version = "v0.21.2",
     )
     go_repository(
         name = "com_google_cloud_go_apigateway",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/apigateway",
         sum = "h1:aBSwCQPcp9rZ0zVEUeJbR623palnqtvxJlUyvzsKGQc=",
         version = "v1.6.1",
     )
     go_repository(
         name = "com_google_cloud_go_apigeeconnect",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/apigeeconnect",
         sum = "h1:6u/jj0P2c3Mcm+H9qLsXI7gYcTiG9ueyQL3n6vCmFJM=",
         version = "v1.6.1",
@@ -999,6 +1020,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_appengine",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/appengine",
         sum = "h1:J+aaUZ6IbTpBegXbmEsh8qZZy864ZVnOoWyfa1XSNbI=",
         version = "v1.8.1",
@@ -1006,12 +1028,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_area120",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/area120",
         sum = "h1:wiOq3KDpdqXmaHzvZwKdpoM+3lDcqsI2Lwhyac7stss=",
         version = "v0.8.1",
     )
     go_repository(
         name = "com_google_cloud_go_artifactregistry",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/artifactregistry",
         sum = "h1:k6hNqab2CubhWlGcSzunJ7kfxC7UzpAfQ1UPb9PDCKI=",
         version = "v1.14.1",
@@ -1019,36 +1043,42 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_asset",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/asset",
         sum = "h1:vlHdznX70eYW4V1y1PxocvF6tEwxJTTarwIGwOhFF3U=",
         version = "v1.14.1",
     )
     go_repository(
         name = "com_google_cloud_go_assuredworkloads",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/assuredworkloads",
         sum = "h1:yaO0kwS+SnhVSTF7BqTyVGt3DTocI6Jqo+S3hHmCwNk=",
         version = "v1.11.1",
     )
     go_repository(
         name = "com_google_cloud_go_automl",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/automl",
         sum = "h1:iP9iQurb0qbz+YOOMfKSEjhONA/WcoOIjt6/m+6pIgo=",
         version = "v1.13.1",
     )
     go_repository(
         name = "com_google_cloud_go_baremetalsolution",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/baremetalsolution",
         sum = "h1:2AipdYXL0VxMboelTTw8c1UJ7gYu35LZYUbuRv9Q28s=",
         version = "v0.5.0",
     )
     go_repository(
         name = "com_google_cloud_go_batch",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/batch",
         sum = "h1:YbMt0E6BtqeD5FvSv1d56jbVsWEzlGm55lYte+M6Mzs=",
         version = "v0.7.0",
     )
     go_repository(
         name = "com_google_cloud_go_beyondcorp",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/beyondcorp",
         sum = "h1:doSatyJyIY1ffqsHuv5DiPSYoXZRIUrJYLArWLZqE/E=",
         version = "v0.6.1",
@@ -1056,42 +1086,49 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_bigquery",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/bigquery",
         sum = "h1:JKLNdxI0N+TIUWD6t9KN646X27N5dQWq9dZbbTWZ8hc=",
         version = "v1.52.0",
     )
     go_repository(
         name = "com_google_cloud_go_billing",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/billing",
         sum = "h1:1iktEAIZ2uA6KpebC235zi/rCXDdDYQ0bTXTNetSL80=",
         version = "v1.16.0",
     )
     go_repository(
         name = "com_google_cloud_go_binaryauthorization",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/binaryauthorization",
         sum = "h1:cAkOhf1ic92zEN4U1zRoSupTmwmxHfklcp1X7CCBKvE=",
         version = "v1.6.1",
     )
     go_repository(
         name = "com_google_cloud_go_certificatemanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/certificatemanager",
         sum = "h1:uKsohpE0hiobx1Eak9jNcPCznwfB6gvyQCcS28Ah9E8=",
         version = "v1.7.1",
     )
     go_repository(
         name = "com_google_cloud_go_channel",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/channel",
         sum = "h1:dqRkK2k7Ll/HHeYGxv18RrfhozNxuTJRkspW0iaFZoY=",
         version = "v1.16.0",
     )
     go_repository(
         name = "com_google_cloud_go_cloudbuild",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/cloudbuild",
         sum = "h1:N6Tl7Xhi0+GWGdt0i2WwaLZKgKeGP4m9A/cERzZcU5k=",
         version = "v1.10.1",
     )
     go_repository(
         name = "com_google_cloud_go_clouddms",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/clouddms",
         sum = "h1:rjR1nV6oVf2aNNB7B5uz1PDIlBjlOiBgR+q5n7bbB7M=",
         version = "v1.6.1",
@@ -1099,6 +1136,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_cloudtasks",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/cloudtasks",
         sum = "h1:zyF35LjQyVQQnWbglmVDbsgOHqkbkaxTeRDisEJsXtE=",
         version = "v1.11.1",
@@ -1106,12 +1144,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_compute",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/compute",
         sum = "h1:6aKEtlUiwEpJzM001l0yFkpXmUVXaN8W+fbkb2AZNbg=",
         version = "v1.20.1",
     )
     go_repository(
         name = "com_google_cloud_go_compute_metadata",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/compute/metadata",
         sum = "h1:mg4jlk7mCAj6xXp9UJ4fjI9VUI5rubuGBW5aJ7UnBMY=",
         version = "v0.2.3",
@@ -1119,12 +1159,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_contactcenterinsights",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/contactcenterinsights",
         sum = "h1:hy4L0bc3fQNZZrhPjuoH62RiisD5B71/S1OZNunsTRk=",
         version = "v1.9.1",
     )
     go_repository(
         name = "com_google_cloud_go_container",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/container",
         sum = "h1:WKBegIfJJc+CL2PIgNpQuvLgGW/CoGJjge5Yjpc0YuU=",
         version = "v1.22.1",
@@ -1132,6 +1174,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_containeranalysis",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/containeranalysis",
         sum = "h1:SM/ibWHWp4TYyJMwrILtcBtYKObyupwOVeceI9pNblw=",
         version = "v0.10.1",
@@ -1139,24 +1182,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_datacatalog",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datacatalog",
         sum = "h1:cFPBt8V5V2T3mu/96tc4nhcMB+5cYcpwjBfn79bZDI8=",
         version = "v1.14.1",
     )
     go_repository(
         name = "com_google_cloud_go_dataflow",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataflow",
         sum = "h1:VzG2tqsk/HbmOtq/XSfdF4cBvUWRK+S+oL9k4eWkENQ=",
         version = "v0.9.1",
     )
     go_repository(
         name = "com_google_cloud_go_dataform",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataform",
         sum = "h1:xcWso0hKOoxeW72AjBSIp/UfkvpqHNzzS0/oygHlcqY=",
         version = "v0.8.1",
     )
     go_repository(
         name = "com_google_cloud_go_datafusion",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datafusion",
         sum = "h1:eX9CZoyhKQW6g1Xj7+RONeDj1mV8KQDKEB9KLELX9/8=",
         version = "v1.7.1",
@@ -1164,18 +1211,21 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_datalabeling",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datalabeling",
         sum = "h1:zxsCD/BLKXhNuRssen8lVXChUj8VxF3ofN06JfdWOXw=",
         version = "v0.8.1",
     )
     go_repository(
         name = "com_google_cloud_go_dataplex",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataplex",
         sum = "h1:RvUH/k3Qi5AOXUAmQVsNCcND9qwJJq3biMSPngO0TQY=",
         version = "v1.8.1",
     )
     go_repository(
         name = "com_google_cloud_go_dataproc",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataproc",
         sum = "h1:W47qHL3W4BPkAIbk4SWmIERwsWBaNnWm0P2sdx3YgGU=",
         version = "v1.12.0",
@@ -1183,6 +1233,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_dataqna",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dataqna",
         sum = "h1:ITpUJep04hC9V7C+gcK390HO++xesQFSUJ7S4nSnF3U=",
         version = "v0.8.1",
@@ -1190,18 +1241,21 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_datastore",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datastore",
         sum = "h1:e713dRZv/yyGRHJ4MQLhAcuxlSzqXlYvEYFMKM7q0u4=",
         version = "v1.12.0",
     )
     go_repository(
         name = "com_google_cloud_go_datastream",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/datastream",
         sum = "h1:BxAt9Tvi1hoEqI4fvyXh/Oc8vd7b5aCZb3bzewh8Dvg=",
         version = "v1.9.1",
     )
     go_repository(
         name = "com_google_cloud_go_deploy",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/deploy",
         sum = "h1:rp+Sf2bWuqJYBuygQl6diFAdvlR8kklhD+stDvyl1zM=",
         version = "v1.11.0",
@@ -1209,12 +1263,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_dialogflow",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dialogflow",
         sum = "h1:kP0t9SX0w3Fbs1q36mSZ3GQuyOgauVhdNXw0wK4cmOI=",
         version = "v1.38.0",
     )
     go_repository(
         name = "com_google_cloud_go_dlp",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/dlp",
         sum = "h1:tF3wsJ2QulRhRLWPzWVkeDz3FkOGVoMl6cmDUHtfYxw=",
         version = "v1.10.1",
@@ -1222,12 +1278,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_documentai",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/documentai",
         sum = "h1:DK9nDulPQgdy3pJIYjMIRrFSAe/Ch3TpfHVn83aV/Gk=",
         version = "v1.20.0",
     )
     go_repository(
         name = "com_google_cloud_go_domains",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/domains",
         sum = "h1:rqz6KY7mEg7Zs/69U6m6LMbB7PxFDWmT3QWNXIqhHm0=",
         version = "v0.9.1",
@@ -1235,12 +1293,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_edgecontainer",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/edgecontainer",
         sum = "h1:zhHWnLzg6AqzE+I3gzJqiIwHfjEBhWctNQEzqb+FaRo=",
         version = "v1.1.1",
     )
     go_repository(
         name = "com_google_cloud_go_errorreporting",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/errorreporting",
         sum = "h1:kj1XEWMu8P0qlLhm3FwcaFsUvXChV/OraZwA70trRR0=",
         version = "v0.3.0",
@@ -1248,18 +1308,21 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_essentialcontacts",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/essentialcontacts",
         sum = "h1:OEJ0MLXXCW/tX1fkxzEZOsv/wRfyFsvDVNaHWBAvoV0=",
         version = "v1.6.2",
     )
     go_repository(
         name = "com_google_cloud_go_eventarc",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/eventarc",
         sum = "h1:8ZAkv7MTnAhix5kSw+Cm/xVzG8+OhC+flZGL9iRdpQA=",
         version = "v1.12.1",
     )
     go_repository(
         name = "com_google_cloud_go_filestore",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/filestore",
         sum = "h1:Eiz8xZzMJc5ppBWkuaod/PUdUZGCFR8ku0uS+Ah2fRw=",
         version = "v1.7.1",
@@ -1267,24 +1330,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_firestore",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/firestore",
         sum = "h1:PPgtwcYUOXV2jFe1bV3nda3RCrOa8cvBjTOn2MQVfW8=",
         version = "v1.11.0",
     )
     go_repository(
         name = "com_google_cloud_go_functions",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/functions",
         sum = "h1:LtAyqvO1TFmNLcROzHZhV0agEJfBi+zfMZsF4RT/a7U=",
         version = "v1.15.1",
     )
     go_repository(
         name = "com_google_cloud_go_gaming",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gaming",
         sum = "h1:5qZmZEWzMf8GEFgm9NeC3bjFRpt7x4S6U7oLbxaf7N8=",
         version = "v1.10.1",
     )
     go_repository(
         name = "com_google_cloud_go_gkebackup",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkebackup",
         sum = "h1:za3QZvw6ujR0uyqkhomKKKNoXDyqYGPJies3voUK8DA=",
         version = "v0.4.0",
@@ -1292,54 +1359,63 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_gkeconnect",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkeconnect",
         sum = "h1:a1ckRvVznnuvDWESM2zZDzSVFvggeBaVY5+BVB8tbT0=",
         version = "v0.8.1",
     )
     go_repository(
         name = "com_google_cloud_go_gkehub",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkehub",
         sum = "h1:2BLSb8i+Co1P05IYCKATXy5yaaIw/ZqGvVSBTLdzCQo=",
         version = "v0.14.1",
     )
     go_repository(
         name = "com_google_cloud_go_gkemulticloud",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gkemulticloud",
         sum = "h1:vg81EW3GQ4RO4PT1MdNHE8aF87EiohZp/WwMWfUTTR0=",
         version = "v0.6.1",
     )
     go_repository(
         name = "com_google_cloud_go_gsuiteaddons",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/gsuiteaddons",
         sum = "h1:mi9jxZpzVjLQibTS/XfPZvl+Jr6D5Bs8pGqUjllRb00=",
         version = "v1.6.1",
     )
     go_repository(
         name = "com_google_cloud_go_iam",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/iam",
         sum = "h1:67gSqaPukx7O8WLLHMa0PNs3EBGd2eE4d+psbO/CO94=",
         version = "v1.1.0",
     )
     go_repository(
         name = "com_google_cloud_go_iap",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/iap",
         sum = "h1:X1tcp+EoJ/LGX6cUPt3W2D4H2Kbqq0pLAsldnsCjLlE=",
         version = "v1.8.1",
     )
     go_repository(
         name = "com_google_cloud_go_ids",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/ids",
         sum = "h1:khXYmSoDDhWGEVxHl4c4IgbwSRR+qE/L4hzP3vaU9Hc=",
         version = "v1.4.1",
     )
     go_repository(
         name = "com_google_cloud_go_iot",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/iot",
         sum = "h1:yrH0OSmicD5bqGBoMlWG8UltzdLkYzNUwNVUVz7OT54=",
         version = "v1.7.1",
     )
     go_repository(
         name = "com_google_cloud_go_kms",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/kms",
         sum = "h1:xZmZuwy2cwzsocmKDOPu4BL7umg8QXagQx6fKVmf45U=",
         version = "v1.12.1",
@@ -1347,18 +1423,21 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_language",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/language",
         sum = "h1:3MXeGEv8AlX+O2LyV4pO4NGpodanc26AmXwOuipEym0=",
         version = "v1.10.1",
     )
     go_repository(
         name = "com_google_cloud_go_lifesciences",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/lifesciences",
         sum = "h1:axkANGx1wiBXHiPcJZAE+TDjjYoJRIDzbHC/WYllCBU=",
         version = "v0.9.1",
     )
     go_repository(
         name = "com_google_cloud_go_logging",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/logging",
         sum = "h1:CJYxlNNNNAMkHp9em/YEXcfJg+rPDg7YfwoRpMU+t5I=",
         version = "v1.7.0",
@@ -1366,18 +1445,21 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_longrunning",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/longrunning",
         sum = "h1:Fr7TXftcqTudoyRJa113hyaqlGdiBQkp0Gq7tErFDWI=",
         version = "v0.5.1",
     )
     go_repository(
         name = "com_google_cloud_go_managedidentities",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/managedidentities",
         sum = "h1:2/qZuOeLgUHorSdxSQGtnOu9xQkBn37+j+oZQv/KHJY=",
         version = "v1.6.1",
     )
     go_repository(
         name = "com_google_cloud_go_maps",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/maps",
         sum = "h1:mv9YaczD4oZBZkM5XJl6fXQ984IkJNHPwkc8MUsdkBo=",
         version = "v0.7.0",
@@ -1385,24 +1467,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_mediatranslation",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/mediatranslation",
         sum = "h1:50cF7c1l3BanfKrpnTCaTvhf+Fo6kdF21DG0byG7gYU=",
         version = "v0.8.1",
     )
     go_repository(
         name = "com_google_cloud_go_memcache",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/memcache",
         sum = "h1:7lkLsF0QF+Mre0O/NvkD9Q5utUNwtzvIYjrOLOs0HO0=",
         version = "v1.10.1",
     )
     go_repository(
         name = "com_google_cloud_go_metastore",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/metastore",
         sum = "h1:sF2yYgo2P4b3hJP2LlIZoafZixtabF/fnORDDMkFeqQ=",
         version = "v1.11.1",
     )
     go_repository(
         name = "com_google_cloud_go_monitoring",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/monitoring",
         sum = "h1:65JhLMd+JiYnXr6j5Z63dUYCuOg770p8a/VC+gil/58=",
         version = "v1.15.1",
@@ -1410,12 +1496,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_networkconnectivity",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/networkconnectivity",
         sum = "h1:LnrYM6lBEeTq+9f2lR4DjBhv31EROSAQi/P5W4Q0AEc=",
         version = "v1.12.1",
     )
     go_repository(
         name = "com_google_cloud_go_networkmanagement",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/networkmanagement",
         sum = "h1:/3xP37eMxnyvkfLrsm1nv1b2FbMMSAEAOlECTvoeCq4=",
         version = "v1.8.0",
@@ -1423,30 +1511,35 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_networksecurity",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/networksecurity",
         sum = "h1:TBLEkMp3AE+6IV/wbIGRNTxnqLXHCTEQWoxRVC18TzY=",
         version = "v0.9.1",
     )
     go_repository(
         name = "com_google_cloud_go_notebooks",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/notebooks",
         sum = "h1:CUqMNEtv4EHFnbogV+yGHQH5iAQLmijOx191innpOcs=",
         version = "v1.9.1",
     )
     go_repository(
         name = "com_google_cloud_go_optimization",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/optimization",
         sum = "h1:pEwOAmO00mxdbesCRSsfj8Sd4rKY9kBrYW7Vd3Pq7cA=",
         version = "v1.4.1",
     )
     go_repository(
         name = "com_google_cloud_go_orchestration",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/orchestration",
         sum = "h1:KmN18kE/xa1n91cM5jhCh7s1/UfIguSCisw7nTMUzgE=",
         version = "v1.8.1",
     )
     go_repository(
         name = "com_google_cloud_go_orgpolicy",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/orgpolicy",
         sum = "h1:I/7dHICQkNwym9erHqmlb50LRU588NPCvkfIY0Bx9jI=",
         version = "v1.11.1",
@@ -1454,24 +1547,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_osconfig",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/osconfig",
         sum = "h1:dgyEHdfqML6cUW6/MkihNdTVc0INQst0qSE8Ou1ub9c=",
         version = "v1.12.1",
     )
     go_repository(
         name = "com_google_cloud_go_oslogin",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/oslogin",
         sum = "h1:LdSuG3xBYu2Sgr3jTUULL1XCl5QBx6xwzGqzoDUw1j0=",
         version = "v1.10.1",
     )
     go_repository(
         name = "com_google_cloud_go_phishingprotection",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/phishingprotection",
         sum = "h1:aK/lNmSd1vtbft/vLe2g7edXK72sIQbqr2QyrZN/iME=",
         version = "v0.8.1",
     )
     go_repository(
         name = "com_google_cloud_go_policytroubleshooter",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/policytroubleshooter",
         sum = "h1:AZ2n6dw6OnYpDZAUk6WK1drupzTWNMRk/uatXEIDAsU=",
         version = "v1.7.1",
@@ -1479,6 +1576,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_privatecatalog",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/privatecatalog",
         sum = "h1:B/18xGo+E0EMS9LOEQ0zXz7F2asMgmVgTYGSI89MHOA=",
         version = "v0.9.1",
@@ -1486,12 +1584,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_pubsub",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/pubsub",
         sum = "h1:JOEkgEYBuUTHSyHS4TcqOFuWr+vD6qO/imsFqShUCp4=",
         version = "v1.32.0",
     )
     go_repository(
         name = "com_google_cloud_go_pubsublite",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/pubsublite",
         sum = "h1:pX+idpWMIH30/K7c0epN6V703xpIcMXWRjKJsz0tYGY=",
         version = "v1.8.1",
@@ -1499,36 +1599,42 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_recaptchaenterprise_v2",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/recaptchaenterprise/v2",
         sum = "h1:IGkbudobsTXAwmkEYOzPCQPApUCsN4Gbq3ndGVhHQpI=",
         version = "v2.7.2",
     )
     go_repository(
         name = "com_google_cloud_go_recommendationengine",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/recommendationengine",
         sum = "h1:nMr1OEVHuDambRn+/y4RmNAmnR/pXCuHtH0Y4tCgGRQ=",
         version = "v0.8.1",
     )
     go_repository(
         name = "com_google_cloud_go_recommender",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/recommender",
         sum = "h1:UKp94UH5/Lv2WXSQe9+FttqV07x/2p1hFTMMYVFtilg=",
         version = "v1.10.1",
     )
     go_repository(
         name = "com_google_cloud_go_redis",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/redis",
         sum = "h1:YrjQnCC7ydk+k30op7DSjSHw1yAYhqYXFcOq1bSXRYA=",
         version = "v1.13.1",
     )
     go_repository(
         name = "com_google_cloud_go_resourcemanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/resourcemanager",
         sum = "h1:QIAMfndPOHR6yTmMUB0ZN+HSeRmPjR/21Smq5/xwghI=",
         version = "v1.9.1",
     )
     go_repository(
         name = "com_google_cloud_go_resourcesettings",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/resourcesettings",
         sum = "h1:Fdyq418U69LhvNPFdlEO29w+DRRjwDA4/pFamm4ksAg=",
         version = "v1.6.1",
@@ -1536,12 +1642,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_retail",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/retail",
         sum = "h1:gYBrb9u/Hc5s5lUTFXX1Vsbc/9BEvgtioY6ZKaK0DK8=",
         version = "v1.14.1",
     )
     go_repository(
         name = "com_google_cloud_go_run",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/run",
         sum = "h1:ydJQo+k+MShYnBfhaRHSZYeD/SQKZzZLAROyfpeD9zw=",
         version = "v0.9.0",
@@ -1549,12 +1657,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_scheduler",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/scheduler",
         sum = "h1:yoZbZR8880KgPGLmACOMCiY2tPk+iX4V/dkxqTirlz8=",
         version = "v1.10.1",
     )
     go_repository(
         name = "com_google_cloud_go_secretmanager",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/secretmanager",
         sum = "h1:cLTCwAjFh9fKvU6F13Y4L9vPcx9yiWPyWXE4+zkuEQs=",
         version = "v1.11.1",
@@ -1562,18 +1672,21 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_security",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/security",
         sum = "h1:jR3itwycg/TgGA0uIgTItcVhA55hKWiNJxaNNpQJaZE=",
         version = "v1.15.1",
     )
     go_repository(
         name = "com_google_cloud_go_securitycenter",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/securitycenter",
         sum = "h1:XOGJ9OpnDtqg8izd7gYk/XUhj8ytjIalyjjsR6oyG0M=",
         version = "v1.23.0",
     )
     go_repository(
         name = "com_google_cloud_go_servicecontrol",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/servicecontrol",
         sum = "h1:d0uV7Qegtfaa7Z2ClDzr9HJmnbJW7jn0WhZ7wOX6hLE=",
         version = "v1.11.1",
@@ -1581,30 +1694,35 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_servicedirectory",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/servicedirectory",
         sum = "h1:J/0csas97yAQ+dcc7i8HqbaOA4KOfPu7BPhJdxYRhCk=",
         version = "v1.10.1",
     )
     go_repository(
         name = "com_google_cloud_go_servicemanagement",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/servicemanagement",
         sum = "h1:fopAQI/IAzlxnVeiKn/8WiV6zKndjFkvi+gzu+NjywY=",
         version = "v1.8.0",
     )
     go_repository(
         name = "com_google_cloud_go_serviceusage",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/serviceusage",
         sum = "h1:rXyq+0+RSIm3HFypctp7WoXxIA563rn206CfMWdqXX4=",
         version = "v1.6.0",
     )
     go_repository(
         name = "com_google_cloud_go_shell",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/shell",
         sum = "h1:aHbwH9LSqs4r2rbay9f6fKEls61TAjT63jSyglsw7sI=",
         version = "v1.7.1",
     )
     go_repository(
         name = "com_google_cloud_go_spanner",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/spanner",
         sum = "h1:aqiMP8dhsEXgn9K5EZBWxPG7dxIiyM2VaikqeU4iteg=",
         version = "v1.47.0",
@@ -1612,6 +1730,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_speech",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/speech",
         sum = "h1:KIV99afoYTJqA2qi8Cjbl5DpjSRzvqFgKcptGXg6kxw=",
         version = "v1.17.1",
@@ -1625,6 +1744,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go_storagetransfer",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/storagetransfer",
         sum = "h1:+ZLkeXx0K0Pk5XdDmG0MnUVqIR18lllsihU/yq39I8Q=",
         version = "v1.10.0",
@@ -1632,12 +1752,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_talent",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/talent",
         sum = "h1:j46ZgD6N2YdpFPux9mc7OAf4YK3tiBCsbLKc8rQx+bU=",
         version = "v1.6.2",
     )
     go_repository(
         name = "com_google_cloud_go_texttospeech",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/texttospeech",
         sum = "h1:S/pR/GZT9p15R7Y2dk2OXD/3AufTct/NSxT4a7nxByw=",
         version = "v1.7.1",
@@ -1650,6 +1772,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go_trace",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/trace",
         sum = "h1:EwGdOLCNfYOOPtgqo+D2sDLZmRCEO1AagRTJCU6ztdg=",
         version = "v1.10.1",
@@ -1662,6 +1785,7 @@ def go_repositories():
     )
     go_repository(
         name = "com_google_cloud_go_video",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/video",
         sum = "h1:gWi0caJILQb9VwZPq28R1Wrg5YMsoLIvtvKDSglcQL8=",
         version = "v1.17.1",
@@ -1669,24 +1793,28 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_videointelligence",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/videointelligence",
         sum = "h1:MBMWnkQ78GQnRz5lfdTAbBq/8QMCF3wahgtHh3s/J+k=",
         version = "v1.11.1",
     )
     go_repository(
         name = "com_google_cloud_go_vision_v2",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/vision/v2",
         sum = "h1:ccK6/YgPfGHR/CyESz1mvIbsht5Y2xRsWCPqmTNydEw=",
         version = "v2.7.2",
     )
     go_repository(
         name = "com_google_cloud_go_vmmigration",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/vmmigration",
         sum = "h1:gnjIclgqbEMc+cF5IJuPxp53wjBIlqZ8h9hE8Rkwp7A=",
         version = "v1.7.1",
     )
     go_repository(
         name = "com_google_cloud_go_vmwareengine",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/vmwareengine",
         sum = "h1:roQrCAkaysVvXxFMuK26lORi+gablOY54htDtDDow0w=",
         version = "v0.4.1",
@@ -1694,6 +1822,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_vpcaccess",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/vpcaccess",
         sum = "h1:ram0GzjNWElmbxXMIzeOZUkQ9J8ZAahD6V8ilPGqX0Y=",
         version = "v1.7.1",
@@ -1701,12 +1830,14 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_webrisk",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/webrisk",
         sum = "h1:Ssy3MkOMOnyRV5H2bkMQ13Umv7CwB/kugo3qkAX83Fk=",
         version = "v1.9.1",
     )
     go_repository(
         name = "com_google_cloud_go_websecurityscanner",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/websecurityscanner",
         sum = "h1:CfEF/vZ+xXyAR3zC9iaC/QRdf1MEgS20r5UR17Q4gOg=",
         version = "v1.6.1",
@@ -1714,6 +1845,7 @@ def go_repositories():

     go_repository(
         name = "com_google_cloud_go_workflows",
+        build_file_proto_mode = "disable_global",
         importpath = "cloud.google.com/go/workflows",
         sum = "h1:2akeQ/PgtRhrNuD/n1WvJd5zb7YyuDZrlOanBj2ihPg=",
         version = "v1.11.1",
@@ -1734,6 +1866,7 @@ def go_repositories():

     go_repository(
         name = "in_gopkg_check_v1",
+        build_file_proto_mode = "disable_global",
         importpath = "gopkg.in/check.v1",
         sum = "h1:Hei/4ADfdWqJk1ZMxUNpqntNwaWcugrBjAiHlqqRiVk=",
         version = "v1.0.0-20201130134442-10cb98267c6c",
@@ -1766,6 +1899,7 @@ def go_repositories():
     )
     go_repository(
         name = "in_gopkg_yaml_v3",
+        build_file_proto_mode = "disable_global",
         importpath = "gopkg.in/yaml.v3",
         sum = "h1:fxVm/GzAzEWqLHuvctI91KS9hhNmmWOoWu0XTYJS7CA=",
         version = "v3.0.1",
@@ -1816,12 +1950,14 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_google_appengine",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/appengine",
         sum = "h1:FZR1q0exgwxzPzp/aF+VccGrSfxfPpkBqjIIEq3ru6c=",
         version = "v1.6.7",
     )
     go_repository(
         name = "org_golang_google_genproto",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/genproto",
         sum = "h1:Au6te5hbKUV8pIYWHqOUZ1pva5qK/rwbIhoXEUB9Lu8=",
         version = "v0.0.0-20230706204954-ccb25ca9f130",
@@ -1841,6 +1977,7 @@ def go_repositories():

     go_repository(
         name = "org_golang_google_grpc",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/grpc",
         sum = "h1:fVRFRnXvU+x6C4IlHZewvJOVHoOv1TUuQyoRsYnB4bI=",
         version = "v1.56.2",
@@ -1854,12 +1991,14 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_google_protobuf",
+        build_file_proto_mode = "disable_global",
         importpath = "google.golang.org/protobuf",
         sum = "h1:g0LDEJHgrBl9N9r17Ru3sqWhkIx2NB67okBHPwC7hs8=",
         version = "v1.31.0",
     )
     go_repository(
         name = "org_golang_x_crypto",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/crypto",
         sum = "h1:6Ewdq3tDic1mg5xRO4milcWCfMVQhI4NkqWWvqejpuA=",
         version = "v0.11.0",
@@ -1892,18 +2031,21 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_x_mod",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/mod",
         sum = "h1:LUYupSeNrTNCGzR/hVBk2NHZO4hXcVaW1k4Qx7rjPx8=",
         version = "v0.8.0",
     )
     go_repository(
         name = "org_golang_x_net",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/net",
         sum = "h1:cfawfvKITfUsFCeJIHJrbSxpeu/E81khclypR0GVT50=",
         version = "v0.12.0",
     )
     go_repository(
         name = "org_golang_x_oauth2",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/oauth2",
         sum = "h1:zHCpF2Khkwy4mMB4bv0U37YtJdTGW8jI0glAApi0Kh8=",
         version = "v0.10.0",
@@ -1916,12 +2058,14 @@ def go_repositories():
     )
     go_repository(
         name = "org_golang_x_sys",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/sys",
         sum = "h1:SqMFp9UcQJZa+pmYuAKjd9xq1f0j5rLcDIk0mj4qAsA=",
         version = "v0.10.0",
     )
     go_repository(
         name = "org_golang_x_term",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/term",
         sum = "h1:3R7pNqamzBraeqj/Tj8qt1aQ2HpmlC+Cx/qL/7hn4/c=",
         version = "v0.10.0",
@@ -1929,6 +2073,7 @@ def go_repositories():

     go_repository(
         name = "org_golang_x_text",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/text",
         sum = "h1:LAntKIrcmeSKERyiOh0XMV39LXS8IE9UL2yP7+f5ij4=",
         version = "v0.11.0",
@@ -1942,6 +2087,7 @@ def go_repositories():

     go_repository(
         name = "org_golang_x_tools",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/tools",
         sum = "h1:BOw41kyTf3PuCW1pVQf8+Cyg8pMlkYB1oo9iJ6D/lKM=",
         version = "v0.6.0",
@@ -1949,6 +2095,7 @@ def go_repositories():

     go_repository(
         name = "org_golang_x_xerrors",
+        build_file_proto_mode = "disable_global",
         importpath = "golang.org/x/xerrors",
         sum = "h1:E7g+9GITq07hpfrRu66IVDexMakfv52eLZ2CXBWiKr4=",
         version = "v0.0.0-20191204190536-9bdfabe68543",
diff --git a/runtime/BUILD.bazel b/runtime/BUILD.bazel
index b5140a3..5ad24a9 100644
--- a/runtime/BUILD.bazel
+++ b/runtime/BUILD.bazel
@@ -1,7 +1,5 @@
 load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")

-package(default_visibility = ["//visibility:public"])
-
 go_library(
     name = "runtime",
     srcs = [
@@ -23,11 +21,12 @@ go_library(
         "query.go",
     ],
     importpath = "github.com/grpc-ecosystem/grpc-gateway/v2/runtime",
+    visibility = ["//visibility:public"],
     deps = [
         "//internal/httprule",
         "//utilities",
-        "@go_googleapis//google/api:httpbody_go_proto",
-        "@io_bazel_rules_go//proto/wkt:field_mask_go_proto",
+        "@org_golang_google_genproto//googleapis/api/httpbody",
+        "@org_golang_google_genproto//protobuf/field_mask",
         "@org_golang_google_grpc//codes",
         "@org_golang_google_grpc//grpclog",
         "@org_golang_google_grpc//health/grpc_health_v1",
@@ -44,9 +43,14 @@ go_library(
     ],
 )

+alias(
+    name = "go_default_library",
+    actual = ":runtime",
+    visibility = ["//visibility:public"],
+)
+
 go_test(
     name = "runtime_test",
-    size = "small",
     srcs = [
         "context_test.go",
         "convert_test.go",
@@ -70,11 +74,11 @@ go_test(
         "//utilities",
         "@com_github_google_go_cmp//cmp",
         "@com_github_google_go_cmp//cmp/cmpopts",
-        "@go_googleapis//google/api:httpbody_go_proto",
-        "@go_googleapis//google/rpc:errdetails_go_proto",
-        "@go_googleapis//google/rpc:status_go_proto",
-        "@io_bazel_rules_go//proto/wkt:field_mask_go_proto",
-        "@org_golang_google_grpc//:go_default_library",
+        "@org_golang_google_genproto//googleapis/api/httpbody",
+        "@org_golang_google_genproto//googleapis/rpc/errdetails",
+        "@org_golang_google_genproto//googleapis/rpc/status",
+        "@org_golang_google_genproto//protobuf/field_mask",
+        "@org_golang_google_grpc//:grpc",
         "@org_golang_google_grpc//codes",
         "@org_golang_google_grpc//health/grpc_health_v1",
         "@org_golang_google_grpc//metadata",
@@ -89,9 +93,3 @@ go_test(
         "@org_golang_google_protobuf//types/known/wrapperspb",
     ],
 )
-
-alias(
-    name = "go_default_library",
-    actual = ":runtime",
-    visibility = ["//visibility:public"],
-)

