diff --git a/repositories.bzl b/repositories.bzl
index e633e062..4cacbc6e 100644
--- a/repositories.bzl
+++ b/repositories.bzl
@@ -3,6 +3,7 @@ load("@bazel_gazelle//:deps.bzl", "go_repository")
 def go_repositories():
     go_repository(
         name = "co_honnef_go_tools",
+        build_file_proto_mode = "disable_global",
         importpath = "honnef.co/go/tools",
         sum = "h1:/hemPrYIhOhy8zYrNj+069zDB68us2sMGsfkFJO0iZs=",
         version = "v0.0.0-20190523083050-ea95bdfd59fc",
diff --git a/runtime/BUILD.bazel b/runtime/BUILD.bazel
index b5140a3..5ad24a9 100644
--- a/runtime/BUILD.bazel
+++ b/runtime/BUILD.bazel
@@ -1,7 +1,5 @@
 load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")

-package(default_visibility = ["//visibility:public"])
-
 go_library(
     name = "runtime",
     srcs = [
@@ -23,11 +21,12 @@ go_library(
         "query.go",
     ],
     importpath = "github.com/grpc-ecosystem/grpc-gateway/v2/runtime",
+    visibility = ["//visibility:public"],
     deps = [
         "//internal/httprule",
         "//utilities",
-        "@go_googleapis//google/api:httpbody_go_proto",
-        "@io_bazel_rules_go//proto/wkt:field_mask_go_proto",
+        "@org_golang_google_genproto//googleapis/api/httpbody",
+        "@org_golang_google_genproto//protobuf/field_mask",
         "@org_golang_google_grpc//codes",
         "@org_golang_google_grpc//grpclog",
         "@org_golang_google_grpc//health/grpc_health_v1",
@@ -44,9 +43,14 @@ go_library(
     ],
 )

+alias(
+    name = "go_default_library",
+    actual = ":runtime",
+    visibility = ["//visibility:public"],
+)
+
 go_test(
     name = "runtime_test",
-    size = "small",
     srcs = [
         "context_test.go",
         "convert_test.go",
@@ -70,11 +74,11 @@ go_test(
         "//utilities",
         "@com_github_google_go_cmp//cmp",
         "@com_github_google_go_cmp//cmp/cmpopts",
-        "@go_googleapis//google/api:httpbody_go_proto",
-        "@go_googleapis//google/rpc:errdetails_go_proto",
-        "@go_googleapis//google/rpc:status_go_proto",
-        "@io_bazel_rules_go//proto/wkt:field_mask_go_proto",
-        "@org_golang_google_grpc//:go_default_library",
+        "@org_golang_google_genproto//googleapis/api/httpbody",
+        "@org_golang_google_genproto//googleapis/rpc/errdetails",
+        "@org_golang_google_genproto//googleapis/rpc/status",
+        "@org_golang_google_genproto//protobuf/field_mask",
+        "@org_golang_google_grpc//:grpc",
         "@org_golang_google_grpc//codes",
         "@org_golang_google_grpc//health/grpc_health_v1",
         "@org_golang_google_grpc//metadata",
@@ -89,9 +93,3 @@ go_test(
         "@org_golang_google_protobuf//types/known/wrapperspb",
     ],
 )
-
-alias(
-    name = "go_default_library",
-    actual = ":runtime",
-    visibility = ["//visibility:public"],
-)

