// vscode/src/vs/platform/workspace/common/workspace.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var event_1 = require("vs/base/common/event");
var uri_1 = require("vs/base/common/uri");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var paths = require("vs/base/common/paths");
var files_1 = require("vs/platform/files/common/files");
exports.IWorkspaceContextService = instantiation_1.createDecorator('contextService');
var WorkspaceContextService = (function () {
    function WorkspaceContextService(workspace) {
        this.workspaceRegistry = new Map();
        this.workspace = workspace;
        this.workspaceEmitter = new event_1.Emitter();
        var workspaceRegistryKey = workspace.resource.with({ fragment: '', query: '' }).toString();
        this.workspaceRegistry.set(workspaceRegistryKey, workspace);
    }
    WorkspaceContextService.prototype.getWorkspace = function () {
        return this.workspace;
    };
    WorkspaceContextService.prototype.hasWorkspace = function () {
        return !!this.workspace;
    };
    WorkspaceContextService.prototype.isInsideWorkspace = function (resource) {
        if (resource && this.workspace) {
            return files_1.isEqual(resource.fsPath, this.workspace.resource.fsPath) || files_1.isParent(resource.fsPath, this.workspace.resource.fsPath);
        }
        return false;
    };
    WorkspaceContextService.prototype.toWorkspaceRelativePath = function (resource, toOSPath) {
        if (this.isInsideWorkspace(resource)) {
            return paths.normalize(paths.relative(this.workspace.resource.fsPath, resource.fsPath), toOSPath);
        }
        return null;
    };
    WorkspaceContextService.prototype.toResource = function (workspaceRelativePath) {
        if (typeof workspaceRelativePath === 'string' && this.workspace) {
            return uri_1.default.file(paths.join(this.workspace.resource.fsPath, workspaceRelativePath));
        }
        return null;
    };
    WorkspaceContextService.prototype.tryGetWorkspaceFromRegistry = function (resource) {
        var workspaceRegistryKey = resource.with({ fragment: '', query: '' }).toString();
        return this.workspaceRegistry.get(workspaceRegistryKey);
    };
    WorkspaceContextService.prototype.setWorkspace = function (workspace) {
        this.workspace = workspace;
        // TODO that when @rothfels changes the URI scheme on Sourcegraph we no longer need to do .with({ fragment: '', query: '' })
        var workspaceRegistryKey = workspace.resource.with({ fragment: '', query: '' }).toString();
        this.workspaceRegistry.set(workspaceRegistryKey, workspace);
        this.workspaceEmitter.fire(workspace);
    };
    Object.defineProperty(WorkspaceContextService.prototype, "onWorkspaceUpdated", {
        get: function () {
            return this.workspaceEmitter.event;
        },
        enumerable: true,
        configurable: true
    });
    return WorkspaceContextService;
}());
exports.WorkspaceContextService = WorkspaceContextService;
