// vscode/src/vs/workbench/browser/part.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
require("./media/part.css");
var builder_1 = require("vs/base/browser/builder");
var component_1 = require("vs/workbench/common/component");
/**
 * Parts are layed out in the workbench and have their own layout that arranges an optional title
 * and mandatory content area to show content.
 */
var Part = (function (_super) {
    __extends(Part, _super);
    function Part(id, options) {
        var _this = _super.call(this, id) || this;
        _this.options = options;
        return _this;
    }
    /**
     * Note: Clients should not call this method, the workbench calls this
     * method. Calling it otherwise may result in unexpected behavior.
     *
     * Called to create title and content area of the part.
     */
    Part.prototype.create = function (parent) {
        this.parent = parent;
        this.titleArea = this.createTitleArea(parent);
        this.contentArea = this.createContentArea(parent);
        this.partLayout = new PartLayout(this.parent, this.options, this.titleArea, this.contentArea);
    };
    /**
     * Returns the overall part container.
     */
    Part.prototype.getContainer = function () {
        return this.parent;
    };
    /**
     * Subclasses override to provide a title area implementation.
     */
    Part.prototype.createTitleArea = function (parent) {
        return null;
    };
    /**
     * Subclasses override to provide a content area implementation.
     */
    Part.prototype.createContentArea = function (parent) {
        return null;
    };
    /**
     * Returns the content area container.
     */
    Part.prototype.getContentArea = function () {
        return this.contentArea;
    };
    /**
     * Layout title and content area in the given dimension.
     */
    Part.prototype.layout = function (dimension) {
        return this.partLayout.layout(dimension);
    };
    /**
     * Returns the part layout implementation.
     */
    Part.prototype.getLayout = function () {
        return this.partLayout;
    };
    return Part;
}(component_1.WorkbenchComponent));
exports.Part = Part;
var TITLE_HEIGHT = 35;
var PartLayout = (function () {
    function PartLayout(container, options, titleArea, contentArea) {
        this.container = container;
        this.options = options;
        this.titleArea = titleArea;
        this.contentArea = contentArea;
    }
    PartLayout.prototype.layout = function (dimension) {
        var width = dimension.width, height = dimension.height;
        // Return the applied sizes to title and content
        var sizes = [];
        // Title Size: Width (Fill), Height (Variable)
        var titleSize;
        if (this.options && this.options.hasTitle) {
            titleSize = new builder_1.Dimension(width, Math.min(height, TITLE_HEIGHT));
        }
        else {
            titleSize = new builder_1.Dimension(0, 0);
        }
        // Content Size: Width (Fill), Height (Variable)
        var contentSize = new builder_1.Dimension(width, height - titleSize.height);
        sizes.push(titleSize);
        sizes.push(contentSize);
        // Content
        if (this.contentArea) {
            this.contentArea.size(contentSize.width, contentSize.height);
        }
        return sizes;
    };
    return PartLayout;
}());
exports.PartLayout = PartLayout;
