diff --git a/js/private/js_binary.bzl b/js/private/js_binary.bzl
index b683eca..46b0ca0 100644
--- a/js/private/js_binary.bzl
+++ b/js/private/js_binary.bzl
@@ -130,6 +130,7 @@ _ATTRS = {
         and ["Make variable"](https://bazel.build/reference/be/make-variables) substitution.
         """,
     ),
+    "use_default_shell_env": attr.bool(),
     "fixed_args": attr.string_list(
         doc = """Fixed command line arguments to pass to the Node.js when this
         binary target is executed.
diff --git a/js/private/js_run_binary.bzl b/js/private/js_run_binary.bzl
index 6b7d109..27dd001 100644
--- a/js/private/js_run_binary.bzl
+++ b/js/private/js_run_binary.bzl
@@ -22,6 +22,7 @@ def js_run_binary(
         name,
         tool,
         env = {},
+        use_default_shell_env = False,
         srcs = [],
         outs = [],
         out_dirs = [],
@@ -358,6 +359,7 @@ See https://github.com/aspect-build/rules_js/tree/main/docs#using-binaries-publi
         name = name,
         tool = tool,
         env = dicts.add(fixed_env, env),
+        use_default_shell_env = use_default_shell_env,
         srcs = srcs + extra_srcs,
         outs = outs + extra_outs,
         out_dirs = out_dirs,
