// vscode/src/vs/platform/editor/common/editor.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
exports.IEditorService = instantiation_1.createDecorator('editorService');
/**
 * Possible locations for opening an editor.
 */
var Position;
(function (Position) {
    /** Opens the editor in the first position replacing the input currently showing */
    Position[Position["ONE"] = 0] = "ONE";
    /** Opens the editor in the second position replacing the input currently showing */
    Position[Position["TWO"] = 1] = "TWO";
    /** Opens the editor in the third most position replacing the input currently showing */
    Position[Position["THREE"] = 2] = "THREE";
})(Position = exports.Position || (exports.Position = {}));
exports.POSITIONS = [Position.ONE, Position.TWO, Position.THREE];
var Direction;
(function (Direction) {
    Direction[Direction["LEFT"] = 0] = "LEFT";
    Direction[Direction["RIGHT"] = 1] = "RIGHT";
})(Direction = exports.Direction || (exports.Direction = {}));
