// vscode/src/vs/workbench/parts/extensions/common/extensionsFileTemplate.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var nls_1 = require("vs/nls");
var extensionManagement_1 = require("vs/platform/extensionManagement/common/extensionManagement");
exports.ExtensionsConfigurationSchemaId = 'vscode://schemas/extensions';
exports.ExtensionsConfigurationSchema = {
    id: exports.ExtensionsConfigurationSchemaId,
    type: 'object',
    title: nls_1.localize('app.extensions.json.title', "Extensions"),
    properties: {
        recommendations: {
            type: 'array',
            description: nls_1.localize('app.extensions.json.recommendations', "List of extensions recommendations. The identifier of an extension is always '${publisher}.${name}'. For example: 'vscode.csharp'."),
            items: {
                type: 'string',
                defaultSnippets: [{ label: 'Example', body: 'vscode.csharp' }],
                pattern: extensionManagement_1.EXTENSION_IDENTIFIER_PATTERN,
                errorMessage: nls_1.localize('app.extension.identifier.errorMessage', "Expected format '${publisher}.${name}'. Example: 'vscode.csharp'.")
            },
        },
    }
};
exports.ExtensionsConfigurationInitialContent = [
    '{',
    '\t// See http://go.microsoft.com/fwlink/?LinkId=827846',
    '\t// for the documentation about the extensions.json format',
    '\t"recommendations": [',
    '\t\t// Extension identifier format: ${publisher}.${name}. Example: vscode.csharp',
    '\t\t',
    '\t]',
    '}'
].join('\n');
