// vscode/src/vs/workbench/parts/extensions/common/extensionQuery.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var Query = (function () {
    function Query(value, sortBy, sortOrder) {
        this.value = value;
        this.sortBy = sortBy;
        this.sortOrder = sortOrder;
        this.value = value.trim();
    }
    Query.parse = function (value) {
        var sortBy = '';
        var sortOrder = '';
        value = value.replace(/@sort:(\w+)(-\w*)?/g, function (match, by, order) {
            if (order === '-asc' || order === '-desc') {
                sortOrder = order.substr(1);
            }
            sortBy = by;
            return '';
        });
        return new Query(value, sortBy, sortOrder);
    };
    Query.prototype.toString = function () {
        var result = this.value;
        if (this.sortBy) {
            result = "" + result + (result ? ' ' : '') + "@sort:" + this.sortBy;
            if (this.sortOrder) {
                result = result + "-" + this.sortOrder;
            }
        }
        return result;
    };
    Query.prototype.isValid = function () {
        return !/@outdated/.test(this.value) && (!!this.sortBy || !this.sortOrder);
    };
    Query.prototype.equals = function (other) {
        return this.value === other.value && this.sortBy === other.sortBy && this.sortOrder === other.sortOrder;
    };
    return Query;
}());
exports.Query = Query;
