diff --git a/cmd/frontend/graphqlbackend/BUILD.bazel b/cmd/frontend/graphqlbackend/BUILD.bazel
index 96ae77fc0a..28a5be4f78 100644
--- a/cmd/frontend/graphqlbackend/BUILD.bazel
+++ b/cmd/frontend/graphqlbackend/BUILD.bazel
@@ -313,7 +313,7 @@ go_library(
         "@com_github_opentracing_opentracing_go//ext",
         "@com_github_prometheus_client_golang//prometheus",
         "@com_github_prometheus_client_golang//prometheus/promauto",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_go_diff//diff",
         "@com_github_sourcegraph_go_langserver//pkg/lsp",
         "@com_github_sourcegraph_jsonx//:jsonx",
diff --git a/cmd/frontend/internal/highlight/BUILD.bazel b/cmd/frontend/internal/highlight/BUILD.bazel
index 0b9a7928c6..95610b2462 100644
--- a/cmd/frontend/internal/highlight/BUILD.bazel
+++ b/cmd/frontend/internal/highlight/BUILD.bazel
@@ -31,7 +31,7 @@ go_library(
         "@com_github_opentracing_opentracing_go//log",
         "@com_github_prometheus_client_golang//prometheus",
         "@com_github_prometheus_client_golang//prometheus/promauto",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
         "@io_opentelemetry_go_otel//attribute",
         "@org_golang_google_protobuf//proto",
         "@org_golang_x_exp//slices",
@@ -59,7 +59,7 @@ go_test(
         "@com_github_google_go_cmp//cmp",
         "@com_github_grafana_regexp//:regexp",
         "@com_github_hexops_autogold_v2//:autogold",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
         "@com_github_stretchr_testify//require",
         "@org_golang_google_protobuf//proto",
     ],
diff --git a/cmd/gitserver/server/BUILD.bazel b/cmd/gitserver/server/BUILD.bazel
index cca31f5e02..ca93f4b0aa 100644
--- a/cmd/gitserver/server/BUILD.bazel
+++ b/cmd/gitserver/server/BUILD.bazel
@@ -84,7 +84,7 @@ go_library(
         "@com_github_opentracing_opentracing_go//log",
         "@com_github_prometheus_client_golang//prometheus",
         "@com_github_prometheus_client_golang//prometheus/promauto",
-        "@com_github_sourcegraph_conc//:conc",
+        "@back_compat_com_github_sourcegraph_conc//:conc",
         "@com_github_sourcegraph_log//:log",
         "@com_github_sourcegraph_mountinfo//:mountinfo",
         "@io_opentelemetry_go_otel//attribute",
diff --git a/cmd/searcher/internal/search/BUILD.bazel b/cmd/searcher/internal/search/BUILD.bazel
index 4f87600d25..29cd8703ab 100644
--- a/cmd/searcher/internal/search/BUILD.bazel
+++ b/cmd/searcher/internal/search/BUILD.bazel
@@ -49,7 +49,7 @@ go_library(
         "@com_github_prometheus_client_golang//prometheus",
         "@com_github_prometheus_client_golang//prometheus/promauto",
         "@com_github_roaringbitmap_roaring//:roaring",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//:log",
         "@com_github_sourcegraph_mountinfo//:mountinfo",
         "@com_github_sourcegraph_zoekt//:zoekt",
@@ -97,7 +97,7 @@ go_test(
         "@com_github_google_go_cmp//cmp",
         "@com_github_grafana_regexp//:regexp",
         "@com_github_roaringbitmap_roaring//:roaring",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//logtest",
         "@com_github_sourcegraph_zoekt//:zoekt",
         "@com_github_sourcegraph_zoekt//query",
diff --git a/cmd/worker/internal/outboundwebhooks/BUILD.bazel b/cmd/worker/internal/outboundwebhooks/BUILD.bazel
index feffca7e7b..2371dae606 100644
--- a/cmd/worker/internal/outboundwebhooks/BUILD.bazel
+++ b/cmd/worker/internal/outboundwebhooks/BUILD.bazel
@@ -31,7 +31,7 @@ go_library(
         "//internal/workerutil/dbworker/store",
         "//lib/errors",
         "@com_github_keegancsmith_sqlf//:sqlf",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//:log",
     ],
 )
diff --git a/dev/codeintel-qa/cmd/clone-and-index/BUILD.bazel b/dev/codeintel-qa/cmd/clone-and-index/BUILD.bazel
index 6cd25601b6..af985f3f3b 100644
--- a/dev/codeintel-qa/cmd/clone-and-index/BUILD.bazel
+++ b/dev/codeintel-qa/cmd/clone-and-index/BUILD.bazel
@@ -9,7 +9,7 @@ go_library(
         "//dev/codeintel-qa/internal",
         "//dev/sg/root",
         "//lib/errors",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_run//:run",
     ],
 )
diff --git a/dev/codeintel-qa/cmd/download/BUILD.bazel b/dev/codeintel-qa/cmd/download/BUILD.bazel
index 8b13a10ef9..f22d52962d 100644
--- a/dev/codeintel-qa/cmd/download/BUILD.bazel
+++ b/dev/codeintel-qa/cmd/download/BUILD.bazel
@@ -9,7 +9,7 @@ go_library(
         "//dev/codeintel-qa/internal",
         "//dev/sg/root",
         "//lib/errors",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_google_cloud_go_storage//:storage",
         "@org_golang_google_api//iterator",
     ],
diff --git a/dev/codeintel-qa/cmd/upload-gcs/BUILD.bazel b/dev/codeintel-qa/cmd/upload-gcs/BUILD.bazel
index de9600e71e..33c30c5c8d 100644
--- a/dev/codeintel-qa/cmd/upload-gcs/BUILD.bazel
+++ b/dev/codeintel-qa/cmd/upload-gcs/BUILD.bazel
@@ -9,7 +9,7 @@ go_library(
         "//dev/codeintel-qa/internal",
         "//dev/sg/root",
         "//lib/errors",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_google_cloud_go_storage//:storage",
     ],
 )
diff --git a/dev/scaletesting/bulkrepocreate/BUILD.bazel b/dev/scaletesting/bulkrepocreate/BUILD.bazel
index 9287ce5727..1f743d5782 100644
--- a/dev/scaletesting/bulkrepocreate/BUILD.bazel
+++ b/dev/scaletesting/bulkrepocreate/BUILD.bazel
@@ -13,7 +13,7 @@ go_library(
         "//lib/output",
         "@com_github_google_go_github_v41//github",
         "@com_github_mattn_go_sqlite3//:go-sqlite3",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_run//:run",
         "@org_golang_x_oauth2//:oauth2",
     ],
diff --git a/dev/scaletesting/bulkreposettings/BUILD.bazel b/dev/scaletesting/bulkreposettings/BUILD.bazel
index 4871637dd6..0af99e49e3 100644
--- a/dev/scaletesting/bulkreposettings/BUILD.bazel
+++ b/dev/scaletesting/bulkreposettings/BUILD.bazel
@@ -10,7 +10,7 @@ go_library(
         "//lib/errors",
         "//lib/output",
         "@com_github_google_go_github//github",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//:log",
         "@com_github_urfave_cli_v2//:cli",
         "@org_golang_x_oauth2//:oauth2",
diff --git a/dev/scaletesting/codehostcopy/BUILD.bazel b/dev/scaletesting/codehostcopy/BUILD.bazel
index 0ad7f74f0e..9fd5f6ac58 100644
--- a/dev/scaletesting/codehostcopy/BUILD.bazel
+++ b/dev/scaletesting/codehostcopy/BUILD.bazel
@@ -20,7 +20,7 @@ go_library(
         "//lib/errors",
         "//lib/output",
         "@com_github_google_go_github//github",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//:log",
         "@com_github_sourcegraph_run//:run",
         "@com_github_urfave_cli_v2//:cli",
diff --git a/dev/scaletesting/syntheticgithubdata/BUILD.bazel b/dev/scaletesting/syntheticgithubdata/BUILD.bazel
index 869199c837..36c2cc0f4f 100644
--- a/dev/scaletesting/syntheticgithubdata/BUILD.bazel
+++ b/dev/scaletesting/syntheticgithubdata/BUILD.bazel
@@ -17,7 +17,7 @@ go_library(
         "//lib/output",
         "@com_github_google_go_github_v41//github",
         "@com_github_mattn_go_sqlite3//:go-sqlite3",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@org_golang_x_oauth2//:oauth2",
     ],
 )
diff --git a/dev/sg/BUILD.bazel b/dev/sg/BUILD.bazel
index 2dfc8f01c6..646643dee7 100644
--- a/dev/sg/BUILD.bazel
+++ b/dev/sg/BUILD.bazel
@@ -112,7 +112,7 @@ go_library(
         "@com_github_opsgenie_opsgenie_go_sdk_v2//alert",
         "@com_github_opsgenie_opsgenie_go_sdk_v2//client",
         "@com_github_slack_go_slack//:slack",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//:log",
         "@com_github_sourcegraph_run//:run",
         "@com_github_urfave_cli_v2//:cli",
diff --git a/dev/sg/internal/check/BUILD.bazel b/dev/sg/internal/check/BUILD.bazel
index 7b1dcd09f0..4e6675908d 100644
--- a/dev/sg/internal/check/BUILD.bazel
+++ b/dev/sg/internal/check/BUILD.bazel
@@ -22,8 +22,8 @@ go_library(
         "//lib/errors",
         "//lib/output",
         "@com_github_masterminds_semver//:semver",
-        "@com_github_sourcegraph_conc//pool",
-        "@com_github_sourcegraph_conc//stream",
+        "@back_compat_com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//stream",
         "@io_opentelemetry_go_otel//attribute",
         "@io_opentelemetry_go_otel_trace//:trace",
         "@org_uber_go_atomic//:atomic",
diff --git a/dev/sg/internal/generate/golang/BUILD.bazel b/dev/sg/internal/generate/golang/BUILD.bazel
index 16d4e48db0..89bc45e76a 100644
--- a/dev/sg/internal/generate/golang/BUILD.bazel
+++ b/dev/sg/internal/generate/golang/BUILD.bazel
@@ -12,7 +12,7 @@ go_library(
         "//lib/errors",
         "//lib/output",
         "@com_github_grafana_regexp//:regexp",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_run//:run",
     ],
 )
diff --git a/dev/sg/internal/images/BUILD.bazel b/dev/sg/internal/images/BUILD.bazel
index a3b622d1cc..7884e5ca03 100644
--- a/dev/sg/internal/images/BUILD.bazel
+++ b/dev/sg/internal/images/BUILD.bazel
@@ -17,7 +17,7 @@ go_library(
         "@com_github_docker_docker_credential_helpers//credentials",
         "@com_github_masterminds_semver//:semver",
         "@com_github_opencontainers_go_digest//:go-digest",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@in_gopkg_yaml_v3//:yaml_v3",
         "@io_k8s_sigs_kustomize_kyaml//kio",
         "@io_k8s_sigs_kustomize_kyaml//yaml",
diff --git a/dev/sg/internal/run/BUILD.bazel b/dev/sg/internal/run/BUILD.bazel
index 3a1b9fed14..1ee1ee1ee1 100644
--- a/dev/sg/internal/run/BUILD.bazel
+++ b/dev/sg/internal/run/BUILD.bazel
@@ -28,7 +28,7 @@ go_library(
         "//lib/process",
         "@com_github_grafana_regexp//:regexp",
         "@com_github_rjeczalik_notify//:notify",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@org_golang_x_sync//errgroup",
         "@org_golang_x_sync//semaphore",
     ],
diff --git a/enterprise/cmd/frontend/internal/compute/streaming/BUILD.bazel b/enterprise/cmd/frontend/internal/compute/streaming/BUILD.bazel
index be1604e9cb..1730c38ddd 100644
--- a/enterprise/cmd/frontend/internal/compute/streaming/BUILD.bazel
+++ b/enterprise/cmd/frontend/internal/compute/streaming/BUILD.bazel
@@ -25,7 +25,7 @@ go_library(
         "//internal/trace",
         "//lib/errors",
         "@com_github_opentracing_opentracing_go//log",
-        "@com_github_sourcegraph_conc//stream",
+        "@back_compat_com_github_sourcegraph_conc//stream",
         "@com_github_sourcegraph_log//:log",
     ],
 )
diff --git a/enterprise/cmd/repo-updater/internal/authz/BUILD.bazel b/enterprise/cmd/repo-updater/internal/authz/BUILD.bazel
index b6053845df..c97bd2486a 100644
--- a/enterprise/cmd/repo-updater/internal/authz/BUILD.bazel
+++ b/enterprise/cmd/repo-updater/internal/authz/BUILD.bazel
@@ -38,7 +38,7 @@ go_library(
         "@com_github_keegancsmith_sqlf//:sqlf",
         "@com_github_prometheus_client_golang//prometheus",
         "@com_github_prometheus_client_golang//prometheus/promauto",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//:log",
         "@io_opentelemetry_go_otel//attribute",
         "@org_golang_x_exp//maps",
@@ -86,7 +86,7 @@ go_test(
         "@com_github_google_go_cmp//cmp",
         "@com_github_grafana_regexp//:regexp",
         "@com_github_inconshreveable_log15//:log15",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//logtest",
         "@com_github_stretchr_testify//assert",
         "@com_github_stretchr_testify//require",
diff --git a/enterprise/internal/codeintel/codenav/internal/lsifstore/BUILD.bazel b/enterprise/internal/codeintel/codenav/internal/lsifstore/BUILD.bazel
index d12f4da8bf..c1fac876de 100644
--- a/enterprise/internal/codeintel/codenav/internal/lsifstore/BUILD.bazel
+++ b/enterprise/internal/codeintel/codenav/internal/lsifstore/BUILD.bazel
@@ -31,7 +31,7 @@ go_library(
         "@com_github_keegancsmith_sqlf//:sqlf",
         "@com_github_lib_pq//:pq",
         "@com_github_opentracing_opentracing_go//log",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
         "@io_opentelemetry_go_otel//attribute",
         "@org_golang_google_protobuf//proto",
     ],
@@ -64,6 +64,6 @@ go_test(
         "//lib/codeintel/precise",
         "@com_github_google_go_cmp//cmp",
         "@com_github_sourcegraph_log//logtest",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
     ],
 )
diff --git a/enterprise/internal/codeintel/ranking/internal/background/BUILD.bazel b/enterprise/internal/codeintel/ranking/internal/background/BUILD.bazel
index 7595dd712f..71780bdba7 100644
--- a/enterprise/internal/codeintel/ranking/internal/background/BUILD.bazel
+++ b/enterprise/internal/codeintel/ranking/internal/background/BUILD.bazel
@@ -18,6 +18,6 @@ go_library(
         "//internal/goroutine",
         "//internal/observation",
         "@com_github_sourcegraph_log//:log",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
     ],
 )
diff --git a/enterprise/internal/codeintel/ranking/internal/lsifstore/BUILD.bazel b/enterprise/internal/codeintel/ranking/internal/lsifstore/BUILD.bazel
index b18e8e40eb..a45a1b03a7 100644
--- a/enterprise/internal/codeintel/ranking/internal/lsifstore/BUILD.bazel
+++ b/enterprise/internal/codeintel/ranking/internal/lsifstore/BUILD.bazel
@@ -19,7 +19,7 @@ go_library(
         "//internal/observation",
         "@com_github_keegancsmith_sqlf//:sqlf",
         "@com_github_opentracing_opentracing_go//log",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
         "@org_golang_google_protobuf//proto",
     ],
 )
diff --git a/enterprise/internal/codeintel/shared/types/BUILD.bazel b/enterprise/internal/codeintel/shared/types/BUILD.bazel
index 13c7acd38d..7f45de70f1 100644
--- a/enterprise/internal/codeintel/shared/types/BUILD.bazel
+++ b/enterprise/internal/codeintel/shared/types/BUILD.bazel
@@ -22,7 +22,7 @@ go_library(
     deps = [
         "//internal/executor",
         "//lib/errors",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
     ],
 )
 
@@ -35,6 +35,6 @@ go_test(
     embed = [":types"],
     deps = [
         "@com_github_google_go_cmp//cmp",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
     ],
 )
diff --git a/enterprise/internal/codeintel/uploads/BUILD.bazel b/enterprise/internal/codeintel/uploads/BUILD.bazel
index 75b7e1d4d0..e07a664b5f 100644
--- a/enterprise/internal/codeintel/uploads/BUILD.bazel
+++ b/enterprise/internal/codeintel/uploads/BUILD.bazel
@@ -67,6 +67,6 @@ go_test(
         "//internal/workerutil/dbworker/store",
         "//lib/codeintel/precise",
         "@com_github_keegancsmith_sqlf//:sqlf",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
     ],
 )
diff --git a/enterprise/internal/codeintel/uploads/internal/background/BUILD.bazel b/enterprise/internal/codeintel/uploads/internal/background/BUILD.bazel
index 6e1f0b539f..e3837ffc83 100644
--- a/enterprise/internal/codeintel/uploads/internal/background/BUILD.bazel
+++ b/enterprise/internal/codeintel/uploads/internal/background/BUILD.bazel
@@ -49,7 +49,7 @@ go_library(
         "@com_github_opentracing_opentracing_go//log",
         "@com_github_prometheus_client_golang//prometheus",
         "@com_github_sourcegraph_log//:log",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
         "@io_opentelemetry_go_otel//attribute",
         "@org_golang_google_protobuf//proto",
     ],
@@ -92,7 +92,7 @@ go_test(
         "@com_github_google_go_cmp//cmp",
         "@com_github_keegancsmith_sqlf//:sqlf",
         "@com_github_sourcegraph_log//logtest",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
         "@org_golang_google_protobuf//proto",
     ],
 )
diff --git a/enterprise/internal/codeintel/uploads/internal/lsifstore/BUILD.bazel b/enterprise/internal/codeintel/uploads/internal/lsifstore/BUILD.bazel
index 07b9e864e8..4b24dcc93a 100644
--- a/enterprise/internal/codeintel/uploads/internal/lsifstore/BUILD.bazel
+++ b/enterprise/internal/codeintel/uploads/internal/lsifstore/BUILD.bazel
@@ -30,7 +30,7 @@ go_library(
         "@com_github_keegancsmith_sqlf//:sqlf",
         "@com_github_lib_pq//:pq",
         "@com_github_opentracing_opentracing_go//log",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
         "@org_golang_google_protobuf//proto",
     ],
 )
@@ -59,6 +59,6 @@ go_test(
         "@com_github_keegancsmith_sqlf//:sqlf",
         "@com_github_sourcegraph_log//:log",
         "@com_github_sourcegraph_log//logtest",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
     ],
 )
diff --git a/enterprise/internal/embeddings/BUILD.bazel b/enterprise/internal/embeddings/BUILD.bazel
index f773dba863..e350b55f6f 100644
--- a/enterprise/internal/embeddings/BUILD.bazel
+++ b/enterprise/internal/embeddings/BUILD.bazel
@@ -23,7 +23,7 @@ go_library(
         "//internal/observation",
         "//internal/uploadstore",
         "//lib/errors",
-        "@com_github_sourcegraph_conc//:conc",
+        "@back_compat_com_github_sourcegraph_conc//:conc",
     ],
 )
 
diff --git a/enterprise/internal/insights/pipeline/BUILD.bazel b/enterprise/internal/insights/pipeline/BUILD.bazel
index 13ad10fc2a..fc766ef7c8 100644
--- a/enterprise/internal/insights/pipeline/BUILD.bazel
+++ b/enterprise/internal/insights/pipeline/BUILD.bazel
@@ -22,7 +22,7 @@ go_library(
         "@com_github_derision_test_glock//:glock",
         "@com_github_prometheus_client_golang//prometheus",
         "@com_github_prometheus_client_golang//prometheus/promauto",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//:log",
     ],
 )
diff --git a/enterprise/internal/insights/scheduler/BUILD.bazel b/enterprise/internal/insights/scheduler/BUILD.bazel
index 4a33196e09..44333e8760 100644
--- a/enterprise/internal/insights/scheduler/BUILD.bazel
+++ b/enterprise/internal/insights/scheduler/BUILD.bazel
@@ -41,7 +41,7 @@ go_library(
         "@com_github_derision_test_glock//:glock",
         "@com_github_keegancsmith_sqlf//:sqlf",
         "@com_github_lib_pq//:pq",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//:log",
     ],
 )
diff --git a/enterprise/internal/oobmigration/migrations/codeintel/BUILD.bazel b/enterprise/internal/oobmigration/migrations/codeintel/BUILD.bazel
index cd0a71ee95..a76695e4aa 100644
--- a/enterprise/internal/oobmigration/migrations/codeintel/BUILD.bazel
+++ b/enterprise/internal/oobmigration/migrations/codeintel/BUILD.bazel
@@ -25,8 +25,8 @@ go_library(
         "//lib/errors",
         "@com_github_keegancsmith_sqlf//:sqlf",
         "@com_github_lib_pq//:pq",
-        "@com_github_sourcegraph_conc//pool",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
         "@io_k8s_utils//lru",
         "@org_golang_google_protobuf//proto",
     ],
diff --git a/internal/comby/BUILD.bazel b/internal/comby/BUILD.bazel
index c223389d66..b1533dbaaa 100644
--- a/internal/comby/BUILD.bazel
+++ b/internal/comby/BUILD.bazel
@@ -16,7 +16,7 @@ go_library(
         "//lib/errors",
         "@com_github_grafana_regexp//:regexp",
         "@com_github_inconshreveable_log15//:log15",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
     ],
 )
 
diff --git a/internal/search/commit/BUILD.bazel b/internal/search/commit/BUILD.bazel
index 3b0bdde0c5..1cee482379 100644
--- a/internal/search/commit/BUILD.bazel
+++ b/internal/search/commit/BUILD.bazel
@@ -21,7 +21,7 @@ go_library(
         "//internal/types",
         "@com_github_grafana_regexp//:regexp",
         "@com_github_opentracing_opentracing_go//log",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
     ],
 )
 
diff --git a/internal/search/job/jobutil/BUILD.bazel b/internal/search/job/jobutil/BUILD.bazel
index 76f3addfbe..5054f4db45 100644
--- a/internal/search/job/jobutil/BUILD.bazel
+++ b/internal/search/job/jobutil/BUILD.bazel
@@ -51,7 +51,7 @@ go_library(
         "//schema",
         "@com_github_grafana_regexp//:regexp",
         "@com_github_opentracing_opentracing_go//log",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//:log",
         "@com_github_sourcegraph_zoekt//query",
         "@io_opentelemetry_go_otel//attribute",
diff --git a/internal/search/repos/BUILD.bazel b/internal/search/repos/BUILD.bazel
index d7c2c2383a..456a520c89 100644
--- a/internal/search/repos/BUILD.bazel
+++ b/internal/search/repos/BUILD.bazel
@@ -33,7 +33,7 @@ go_library(
         "@com_github_grafana_regexp//:regexp",
         "@com_github_grafana_regexp//syntax",
         "@com_github_opentracing_opentracing_go//log",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//:log",
         "@com_github_sourcegraph_zoekt//:zoekt",
         "@com_github_sourcegraph_zoekt//query",
diff --git a/internal/search/searcher/BUILD.bazel b/internal/search/searcher/BUILD.bazel
index 1d0bd0680f..beab7960bf 100644
--- a/internal/search/searcher/BUILD.bazel
+++ b/internal/search/searcher/BUILD.bazel
@@ -37,7 +37,7 @@ go_library(
         "@com_github_opentracing_contrib_go_stdlib//nethttp",
         "@com_github_opentracing_opentracing_go//ext",
         "@com_github_opentracing_opentracing_go//log",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_sourcegraph_log//:log",
         "@io_opentelemetry_go_otel//attribute",
         "@org_golang_google_grpc//codes",
diff --git a/internal/search/streaming/BUILD.bazel b/internal/search/streaming/BUILD.bazel
index 3f181e56cb..1a7a5095ca 100644
--- a/internal/search/streaming/BUILD.bazel
+++ b/internal/search/streaming/BUILD.bazel
@@ -33,7 +33,7 @@ go_test(
         "//internal/search/result",
         "//internal/types",
         "@com_github_google_go_cmp//cmp",
-        "@com_github_sourcegraph_conc//pool",
+        "@back_compat_com_github_sourcegraph_conc//pool",
         "@com_github_stretchr_testify//require",
         "@org_uber_go_atomic//:atomic",
     ],
diff --git a/lib/codeintel/lsif/scip/BUILD.bazel b/lib/codeintel/lsif/scip/BUILD.bazel
index de94759a39..2d6cd65fae 100644
--- a/lib/codeintel/lsif/scip/BUILD.bazel
+++ b/lib/codeintel/lsif/scip/BUILD.bazel
@@ -12,7 +12,7 @@ go_library(
         "//lib/codeintel/lsif/conversion",
         "//lib/codeintel/precise",
         "//lib/codeintel/upload",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
     ],
 )
 
@@ -28,6 +28,6 @@ go_test(
         "//lib/codeintel/precise",
         "@com_github_google_go_cmp//cmp",
         "@com_github_google_go_cmp//cmp/cmpopts",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
     ],
 )
diff --git a/lib/codeintel/upload/BUILD.bazel b/lib/codeintel/upload/BUILD.bazel
index d224ac7f39..147da41a86 100644
--- a/lib/codeintel/upload/BUILD.bazel
+++ b/lib/codeintel/upload/BUILD.bazel
@@ -18,7 +18,7 @@ go_library(
         "//lib/errors",
         "//lib/output",
         "@com_github_klauspost_pgzip//:pgzip",
-        "@com_github_sourcegraph_scip//bindings/go/scip",
+        "@back_compat_com_github_sourcegraph_scip//bindings/go/scip",
         "@org_golang_google_protobuf//proto",
     ],
 )
