// vscode/src/vs/workbench/parts/search/common/search.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var winjs_base_1 = require("vs/base/common/winjs.base");
var errors_1 = require("vs/base/common/errors");
var editorCommonExtensions_1 = require("vs/editor/common/editorCommonExtensions");
var SymbolProvider = (function () {
    function SymbolProvider(provider, workspace) {
        this.provider = provider;
        this.workspace = workspace;
        this.resolveWorkspaceSymbol = provider.resolveWorkspaceSymbol;
    }
    SymbolProvider.prototype.provideWorkspaceSymbols = function (search) {
        return this.provider.provideWorkspaceSymbols(search);
    };
    return SymbolProvider;
}());
var WorkspaceSymbolProviderRegistry;
(function (WorkspaceSymbolProviderRegistry) {
    var _supports = [];
    function register(support, workspace) {
        var supportProvider;
        if (support) {
            supportProvider = new SymbolProvider(support, workspace);
            _supports.push(supportProvider);
        }
        return {
            dispose: function () {
                if (supportProvider) {
                    var idx = _supports.indexOf(supportProvider);
                    if (idx >= 0) {
                        _supports.splice(idx, 1);
                        support = undefined;
                    }
                }
            }
        };
    }
    WorkspaceSymbolProviderRegistry.register = register;
    function all(workspace) {
        if (workspace) {
            return _supports.filter(function (support) { return support.workspace && support.workspace.resource && support.workspace.resource.toString() === workspace.resource.toString(); });
        }
        return _supports.slice(0); // make a copy
    }
    WorkspaceSymbolProviderRegistry.all = all;
})(WorkspaceSymbolProviderRegistry = exports.WorkspaceSymbolProviderRegistry || (exports.WorkspaceSymbolProviderRegistry = {}));
function getWorkspaceSymbols(query, workspace) {
    var result = [];
    var promises = WorkspaceSymbolProviderRegistry.all(workspace).map(function (support) {
        return support.provideWorkspaceSymbols(query).then(function (value) {
            if (Array.isArray(value)) {
                result.push([support, value]);
            }
        }, errors_1.onUnexpectedError);
    });
    return winjs_base_1.TPromise.join(promises).then(function (_) { return result; });
}
exports.getWorkspaceSymbols = getWorkspaceSymbols;
editorCommonExtensions_1.CommonEditorRegistry.registerLanguageCommand('_executeWorkspaceSymbolProvider', function (accessor, args) {
    var query = args.query;
    if (typeof query !== 'string') {
        throw errors_1.illegalArgument();
    }
    // TODO(john): send current workspace for query
    return getWorkspaceSymbols(query);
});
