// vscode/src/vs/workbench/parts/files/browser/fileActions.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
require('./media/fileactions.css');
var winjs_base_1 = require('vs/base/common/winjs.base');
var nls = require('vs/nls');
var platform_1 = require('vs/base/common/platform');
var async_1 = require('vs/base/common/async');
var paths = require('vs/base/common/paths');
var uri_1 = require('vs/base/common/uri');
var errors = require('vs/base/common/errors');
var errorMessage_1 = require('vs/base/common/errorMessage');
var strings = require('vs/base/common/strings');
var events_1 = require('vs/base/common/events');
var severity_1 = require('vs/base/common/severity');
var diagnostics = require('vs/base/common/diagnostics');
var textEditor_1 = require('vs/workbench/browser/parts/editor/textEditor');
var actions_1 = require('vs/base/common/actions');
var inputBox_1 = require('vs/base/browser/ui/inputbox/inputBox');
var lifecycle_1 = require('vs/base/common/lifecycle');
var files_1 = require('vs/workbench/parts/files/common/files');
var textfiles_1 = require('vs/workbench/services/textfile/common/textfiles');
var files_2 = require('vs/platform/files/common/files');
var diffEditorInput_1 = require('vs/workbench/common/editor/diffEditorInput');
var editor_1 = require('vs/workbench/common/editor');
var fileEditorInput_1 = require('vs/workbench/parts/files/common/editors/fileEditorInput');
var explorerViewModel_1 = require('vs/workbench/parts/files/common/explorerViewModel');
var untitledEditorService_1 = require('vs/workbench/services/untitled/common/untitledEditorService');
var editorService_1 = require('vs/workbench/services/editor/common/editorService');
var viewlet_1 = require('vs/workbench/browser/viewlet');
var groupService_1 = require('vs/workbench/services/group/common/groupService');
var quickOpenService_1 = require('vs/workbench/services/quickopen/common/quickOpenService');
var viewletService_1 = require('vs/workbench/services/viewlet/common/viewletService');
var editor_2 = require('vs/platform/editor/common/editor');
var event_1 = require('vs/platform/event/common/event');
var instantiation_1 = require('vs/platform/instantiation/common/instantiation');
var message_1 = require('vs/platform/message/common/message');
var workspace_1 = require('vs/platform/workspace/common/workspace');
var keybinding_1 = require('vs/base/common/keybinding');
var BaseFileAction = (function (_super) {
    __extends(BaseFileAction, _super);
    function BaseFileAction(id, label, _contextService, _editorService, _fileService, _messageService, _textFileService, _eventService) {
        _super.call(this, id, label);
        this._contextService = _contextService;
        this._editorService = _editorService;
        this._fileService = _fileService;
        this._messageService = _messageService;
        this._textFileService = _textFileService;
        this._eventService = _eventService;
        this.enabled = false;
    }
    Object.defineProperty(BaseFileAction.prototype, "contextService", {
        get: function () {
            return this._contextService;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BaseFileAction.prototype, "messageService", {
        get: function () {
            return this._messageService;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BaseFileAction.prototype, "editorService", {
        get: function () {
            return this._editorService;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BaseFileAction.prototype, "fileService", {
        get: function () {
            return this._fileService;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BaseFileAction.prototype, "eventService", {
        get: function () {
            return this._eventService;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BaseFileAction.prototype, "textFileService", {
        get: function () {
            return this._textFileService;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(BaseFileAction.prototype, "element", {
        get: function () {
            return this._element;
        },
        set: function (element) {
            this._element = element;
        },
        enumerable: true,
        configurable: true
    });
    BaseFileAction.prototype._isEnabled = function () {
        return true;
    };
    BaseFileAction.prototype._updateEnablement = function () {
        this.enabled = !!(this._contextService && this._fileService && this._editorService && this._isEnabled());
    };
    BaseFileAction.prototype.onError = function (error) {
        this._messageService.show(message_1.Severity.Error, error);
    };
    BaseFileAction.prototype.onWarning = function (warning) {
        this._messageService.show(message_1.Severity.Warning, warning);
    };
    BaseFileAction.prototype.onErrorWithRetry = function (error, retry, extraAction) {
        var actions = [
            new actions_1.Action(this.id, nls.localize('retry', "Retry"), null, true, function () { return retry(); }),
            message_1.CancelAction
        ];
        if (extraAction) {
            actions.unshift(extraAction);
        }
        var errorWithRetry = {
            actions: actions,
            message: errorMessage_1.toErrorMessage(error, false)
        };
        this._messageService.show(message_1.Severity.Error, errorWithRetry);
    };
    BaseFileAction = __decorate([
        __param(2, workspace_1.IWorkspaceContextService),
        __param(3, editorService_1.IWorkbenchEditorService),
        __param(4, files_2.IFileService),
        __param(5, message_1.IMessageService),
        __param(6, textfiles_1.ITextFileService),
        __param(7, event_1.IEventService)
    ], BaseFileAction);
    return BaseFileAction;
}(actions_1.Action));
exports.BaseFileAction = BaseFileAction;
var TriggerRenameFileAction = (function (_super) {
    __extends(TriggerRenameFileAction, _super);
    function TriggerRenameFileAction(tree, element, contextService, editorService, fileService, messageService, textFileService, eventService, instantiationService) {
        _super.call(this, TriggerRenameFileAction.ID, nls.localize('rename', "Rename"), contextService, editorService, fileService, messageService, textFileService, eventService);
        this.tree = tree;
        this.element = element;
        this.renameAction = instantiationService.createInstance(RenameFileAction, element);
        this._updateEnablement();
    }
    TriggerRenameFileAction.prototype.validateFileName = function (parent, name) {
        return this.renameAction.validateFileName(this.element.parent, name);
    };
    TriggerRenameFileAction.prototype.run = function (context) {
        var _this = this;
        if (!context) {
            return winjs_base_1.TPromise.wrapError('No context provided to BaseEnableFileRenameAction.');
        }
        var viewletState = context.viewletState;
        if (!viewletState) {
            return winjs_base_1.TPromise.wrapError('Invalid viewlet state provided to BaseEnableFileRenameAction.');
        }
        var stat = context.stat;
        if (!stat) {
            return winjs_base_1.TPromise.wrapError('Invalid stat provided to BaseEnableFileRenameAction.');
        }
        viewletState.setEditable(stat, {
            action: this.renameAction,
            validator: function (value) {
                var message = _this.validateFileName(_this.element.parent, value);
                if (!message) {
                    return null;
                }
                return {
                    content: message,
                    formatContent: true,
                    type: inputBox_1.MessageType.ERROR
                };
            }
        });
        this.tree.refresh(stat, false).then(function () {
            _this.tree.setHighlight(stat);
            var unbind = _this.tree.addListener2(events_1.EventType.HIGHLIGHT, function (e) {
                if (!e.highlight) {
                    viewletState.clearEditable(stat);
                    _this.tree.refresh(stat).done(null, errors.onUnexpectedError);
                    unbind.dispose();
                }
            });
        }).done(null, errors.onUnexpectedError);
    };
    TriggerRenameFileAction.ID = 'workbench.files.action.triggerRename';
    TriggerRenameFileAction = __decorate([
        __param(2, workspace_1.IWorkspaceContextService),
        __param(3, editorService_1.IWorkbenchEditorService),
        __param(4, files_2.IFileService),
        __param(5, message_1.IMessageService),
        __param(6, textfiles_1.ITextFileService),
        __param(7, event_1.IEventService),
        __param(8, instantiation_1.IInstantiationService)
    ], TriggerRenameFileAction);
    return TriggerRenameFileAction;
}(BaseFileAction));
exports.TriggerRenameFileAction = TriggerRenameFileAction;
var BaseRenameAction = (function (_super) {
    __extends(BaseRenameAction, _super);
    function BaseRenameAction(id, label, element, contextService, editorService, fileService, messageService, textFileService, eventService) {
        _super.call(this, id, label, contextService, editorService, fileService, messageService, textFileService, eventService);
        this.element = element;
    }
    BaseRenameAction.prototype.run = function (context) {
        var _this = this;
        if (!context) {
            return winjs_base_1.TPromise.wrapError('No context provided to BaseRenameFileAction.');
        }
        var name = context.value;
        if (!name) {
            return winjs_base_1.TPromise.wrapError('No new name provided to BaseRenameFileAction.');
        }
        // Automatically trim whitespaces and trailing dots to produce nice file names
        name = getWellFormedFileName(name);
        var existingName = getWellFormedFileName(this.element.name);
        // Return early if name is invalid or didn't change
        if (name === existingName || this.validateFileName(this.element.parent, name)) {
            return winjs_base_1.TPromise.as(null);
        }
        // Call function and Emit Event through viewer
        var promise = this.runAction(name).then(function (stat) {
            if (stat) {
                _this.onSuccess(stat);
            }
        }, function (error) {
            _this.onError(error);
        });
        return promise;
    };
    BaseRenameAction.prototype.validateFileName = function (parent, name) {
        var source = this.element.name;
        var target = name;
        if (!platform_1.isLinux) {
            source = source.toLowerCase();
            target = target.toLowerCase();
        }
        if (getWellFormedFileName(source) === getWellFormedFileName(target)) {
            return null;
        }
        return validateFileName(parent, name, false);
    };
    BaseRenameAction.prototype.onSuccess = function (stat) {
        var before = null;
        if (!(this.element instanceof explorerViewModel_1.NewStatPlaceholder)) {
            before = this.element.clone(); // Clone element to not expose viewers element to listeners
        }
        this.eventService.emit('files.internal:fileChanged', new textfiles_1.LocalFileChangeEvent(before, stat));
    };
    BaseRenameAction = __decorate([
        __param(3, workspace_1.IWorkspaceContextService),
        __param(4, editorService_1.IWorkbenchEditorService),
        __param(5, files_2.IFileService),
        __param(6, message_1.IMessageService),
        __param(7, textfiles_1.ITextFileService),
        __param(8, event_1.IEventService)
    ], BaseRenameAction);
    return BaseRenameAction;
}(BaseFileAction));
exports.BaseRenameAction = BaseRenameAction;
var RenameFileAction = (function (_super) {
    __extends(RenameFileAction, _super);
    function RenameFileAction(element, contextService, editorService, fileService, messageService, textFileService, eventService) {
        _super.call(this, RenameFileAction.ID, nls.localize('rename', "Rename"), element, contextService, editorService, fileService, messageService, textFileService, eventService);
        this._updateEnablement();
    }
    RenameFileAction.prototype.runAction = function (newName) {
        var _this = this;
        // Handle dirty
        var revertPromise = winjs_base_1.TPromise.as(null);
        var dirty = this.textFileService.getDirty().filter(function (d) { return paths.isEqualOrParent(d.fsPath, _this.element.resource.fsPath); });
        if (dirty.length) {
            var message = void 0;
            if (this.element.isDirectory) {
                if (dirty.length === 1) {
                    message = nls.localize('dirtyMessageFolderOne', "You are renaming a folder with unsaved changes in 1 file. Do you want to continue?");
                }
                else {
                    message = nls.localize('dirtyMessageFolder', "You are renaming a folder with unsaved changes in {0} files. Do you want to continue?", dirty.length);
                }
            }
            else {
                message = nls.localize('dirtyMessageFile', "You are renaming a file with unsaved changes. Do you want to continue?");
            }
            var res = this.messageService.confirm({
                message: message,
                type: 'warning',
                detail: nls.localize('dirtyWarning', "Your changes will be lost if you don't save them."),
                primaryButton: nls.localize({ key: 'renameLabel', comment: ['&& denotes a mnemonic'] }, "&&Rename")
            });
            if (!res) {
                return winjs_base_1.TPromise.as(null);
            }
            revertPromise = this.textFileService.revertAll(dirty);
        }
        return revertPromise.then(function () {
            return _this.fileService.rename(_this.element.resource, newName).then(null, function (error) {
                _this.onErrorWithRetry(error, function () { return _this.runAction(newName); });
            });
        });
    };
    RenameFileAction.ID = 'workbench.files.action.renameFile';
    RenameFileAction = __decorate([
        __param(1, workspace_1.IWorkspaceContextService),
        __param(2, editorService_1.IWorkbenchEditorService),
        __param(3, files_2.IFileService),
        __param(4, message_1.IMessageService),
        __param(5, textfiles_1.ITextFileService),
        __param(6, event_1.IEventService)
    ], RenameFileAction);
    return RenameFileAction;
}(BaseRenameAction));
exports.RenameFileAction = RenameFileAction;
/* Base New File/Folder Action */
var BaseNewAction = (function (_super) {
    __extends(BaseNewAction, _super);
    function BaseNewAction(id, label, tree, isFile, editableAction, element, contextService, editorService, fileService, messageService, textFileService, eventService) {
        _super.call(this, id, label, contextService, editorService, fileService, messageService, textFileService, eventService);
        if (element) {
            this.presetFolder = element.isDirectory ? element : element.parent;
        }
        this.tree = tree;
        this.isFile = isFile;
        this.renameAction = editableAction;
    }
    BaseNewAction.prototype.run = function (context) {
        var _this = this;
        if (!context) {
            return winjs_base_1.TPromise.wrapError('No context provided to BaseNewAction.');
        }
        var viewletState = context.viewletState;
        if (!viewletState) {
            return winjs_base_1.TPromise.wrapError('Invalid viewlet state provided to BaseNewAction.');
        }
        var folder = this.presetFolder;
        if (!folder) {
            var focus_1 = this.tree.getFocus();
            if (focus_1) {
                folder = focus_1.isDirectory ? focus_1 : focus_1.parent;
            }
            else {
                folder = this.tree.getInput();
            }
        }
        if (!folder) {
            return winjs_base_1.TPromise.wrapError('Invalid parent folder to create.');
        }
        return this.tree.reveal(folder, 0.5).then(function () {
            return _this.tree.expand(folder).then(function () {
                var stat = explorerViewModel_1.NewStatPlaceholder.addNewStatPlaceholder(folder, !_this.isFile);
                _this.renameAction.element = stat;
                viewletState.setEditable(stat, {
                    action: _this.renameAction,
                    validator: function (value) {
                        var message = _this.renameAction.validateFileName(folder, value);
                        if (!message) {
                            return null;
                        }
                        return {
                            content: message,
                            formatContent: true,
                            type: inputBox_1.MessageType.ERROR
                        };
                    }
                });
                return _this.tree.refresh(folder).then(function () {
                    return _this.tree.expand(folder).then(function () {
                        return _this.tree.reveal(stat, 0.5).then(function () {
                            _this.tree.setHighlight(stat);
                            var unbind = _this.tree.addListener2(events_1.EventType.HIGHLIGHT, function (e) {
                                if (!e.highlight) {
                                    stat.destroy();
                                    _this.tree.refresh(folder).done(null, errors.onUnexpectedError);
                                    unbind.dispose();
                                }
                            });
                        });
                    });
                });
            });
        });
    };
    BaseNewAction = __decorate([
        __param(6, workspace_1.IWorkspaceContextService),
        __param(7, editorService_1.IWorkbenchEditorService),
        __param(8, files_2.IFileService),
        __param(9, message_1.IMessageService),
        __param(10, textfiles_1.ITextFileService),
        __param(11, event_1.IEventService)
    ], BaseNewAction);
    return BaseNewAction;
}(BaseFileAction));
exports.BaseNewAction = BaseNewAction;
/* New File */
var NewFileAction = (function (_super) {
    __extends(NewFileAction, _super);
    function NewFileAction(tree, element, contextService, editorService, fileService, messageService, textFileService, eventService, instantiationService) {
        _super.call(this, 'workbench.action.files.newFile', nls.localize('newFile', "New File"), tree, true, instantiationService.createInstance(CreateFileAction, element), null, contextService, editorService, fileService, messageService, textFileService, eventService);
        this.class = 'explorer-action new-file';
        this._updateEnablement();
    }
    NewFileAction = __decorate([
        __param(2, workspace_1.IWorkspaceContextService),
        __param(3, editorService_1.IWorkbenchEditorService),
        __param(4, files_2.IFileService),
        __param(5, message_1.IMessageService),
        __param(6, textfiles_1.ITextFileService),
        __param(7, event_1.IEventService),
        __param(8, instantiation_1.IInstantiationService)
    ], NewFileAction);
    return NewFileAction;
}(BaseNewAction));
exports.NewFileAction = NewFileAction;
/* New Folder */
var NewFolderAction = (function (_super) {
    __extends(NewFolderAction, _super);
    function NewFolderAction(tree, element, contextService, editorService, fileService, messageService, textFileService, eventService, instantiationService) {
        _super.call(this, 'workbench.action.files.newFolder', nls.localize('newFolder', "New Folder"), tree, false, instantiationService.createInstance(CreateFolderAction, element), null, contextService, editorService, fileService, messageService, textFileService, eventService);
        this.class = 'explorer-action new-folder';
        this._updateEnablement();
    }
    NewFolderAction = __decorate([
        __param(2, workspace_1.IWorkspaceContextService),
        __param(3, editorService_1.IWorkbenchEditorService),
        __param(4, files_2.IFileService),
        __param(5, message_1.IMessageService),
        __param(6, textfiles_1.ITextFileService),
        __param(7, event_1.IEventService),
        __param(8, instantiation_1.IInstantiationService)
    ], NewFolderAction);
    return NewFolderAction;
}(BaseNewAction));
exports.NewFolderAction = NewFolderAction;
var BaseGlobalNewAction = (function (_super) {
    __extends(BaseGlobalNewAction, _super);
    function BaseGlobalNewAction(id, label, viewletService, instantiationService, messageService) {
        _super.call(this, id, label);
        this.viewletService = viewletService;
        this.instantiationService = instantiationService;
        this.messageService = messageService;
    }
    BaseGlobalNewAction.prototype.run = function () {
        var _this = this;
        return this.viewletService.openViewlet(files_1.VIEWLET_ID, true).then(function (viewlet) {
            return winjs_base_1.TPromise.timeout(100).then(function () {
                viewlet.focus();
                var explorer = viewlet;
                var explorerView = explorer.getExplorerView();
                // Not having a folder opened
                if (!explorerView) {
                    return _this.messageService.show(message_1.Severity.Info, nls.localize('openFolderFirst', "Open a folder first to create files or folders within."));
                }
                if (!explorerView.isExpanded()) {
                    explorerView.expand();
                }
                var action = _this.toDispose = _this.instantiationService.createInstance(_this.getAction(), explorerView.getViewer(), null);
                return explorer.getActionRunner().run(action);
            });
        });
    };
    BaseGlobalNewAction.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
        if (this.toDispose) {
            this.toDispose.dispose();
            this.toDispose = null;
        }
    };
    BaseGlobalNewAction = __decorate([
        __param(2, viewletService_1.IViewletService),
        __param(3, instantiation_1.IInstantiationService),
        __param(4, message_1.IMessageService)
    ], BaseGlobalNewAction);
    return BaseGlobalNewAction;
}(actions_1.Action));
exports.BaseGlobalNewAction = BaseGlobalNewAction;
/* Create new file from anywhere: Open untitled */
var GlobalNewUntitledFileAction = (function (_super) {
    __extends(GlobalNewUntitledFileAction, _super);
    function GlobalNewUntitledFileAction(id, label, editorService, untitledEditorService) {
        _super.call(this, id, label);
        this.editorService = editorService;
        this.untitledEditorService = untitledEditorService;
    }
    GlobalNewUntitledFileAction.prototype.run = function () {
        var input = this.untitledEditorService.createOrGet();
        return this.editorService.openEditor(input, { pinned: true }); // untitled are always pinned
    };
    GlobalNewUntitledFileAction.ID = 'workbench.action.files.newUntitledFile';
    GlobalNewUntitledFileAction.LABEL = nls.localize('newUntitledFile', "New Untitled File");
    GlobalNewUntitledFileAction = __decorate([
        __param(2, editorService_1.IWorkbenchEditorService),
        __param(3, untitledEditorService_1.IUntitledEditorService)
    ], GlobalNewUntitledFileAction);
    return GlobalNewUntitledFileAction;
}(actions_1.Action));
exports.GlobalNewUntitledFileAction = GlobalNewUntitledFileAction;
/* Create new file from anywhere */
var GlobalNewFileAction = (function (_super) {
    __extends(GlobalNewFileAction, _super);
    function GlobalNewFileAction() {
        _super.apply(this, arguments);
    }
    GlobalNewFileAction.prototype.getAction = function () {
        return NewFileAction;
    };
    GlobalNewFileAction.ID = 'workbench.action.files.newFile';
    GlobalNewFileAction.LABEL = nls.localize('newFile', "New File");
    return GlobalNewFileAction;
}(BaseGlobalNewAction));
exports.GlobalNewFileAction = GlobalNewFileAction;
/* Create new folder from anywhere */
var GlobalNewFolderAction = (function (_super) {
    __extends(GlobalNewFolderAction, _super);
    function GlobalNewFolderAction() {
        _super.apply(this, arguments);
    }
    GlobalNewFolderAction.prototype.getAction = function () {
        return NewFolderAction;
    };
    GlobalNewFolderAction.ID = 'workbench.action.files.newFolder';
    GlobalNewFolderAction.LABEL = nls.localize('newFolder', "New Folder");
    return GlobalNewFolderAction;
}(BaseGlobalNewAction));
exports.GlobalNewFolderAction = GlobalNewFolderAction;
/* Create New File/Folder (only used internally by explorerViewer) */
var BaseCreateAction = (function (_super) {
    __extends(BaseCreateAction, _super);
    function BaseCreateAction() {
        _super.apply(this, arguments);
    }
    BaseCreateAction.prototype.validateFileName = function (parent, name) {
        if (this.element instanceof explorerViewModel_1.NewStatPlaceholder) {
            return validateFileName(parent, name, false);
        }
        return _super.prototype.validateFileName.call(this, parent, name);
    };
    return BaseCreateAction;
}(BaseRenameAction));
exports.BaseCreateAction = BaseCreateAction;
/* Create New File (only used internally by explorerViewer) */
var CreateFileAction = (function (_super) {
    __extends(CreateFileAction, _super);
    function CreateFileAction(element, contextService, editorService, fileService, messageService, textFileService, eventService) {
        _super.call(this, CreateFileAction.ID, CreateFileAction.LABEL, element, contextService, editorService, fileService, messageService, textFileService, eventService);
        this._updateEnablement();
    }
    CreateFileAction.prototype.runAction = function (fileName) {
        var _this = this;
        return this.fileService.createFile(uri_1.default.file(paths.join(this.element.parent.resource.fsPath, fileName))).then(null, function (error) {
            _this.onErrorWithRetry(error, function () { return _this.runAction(fileName); });
        });
    };
    CreateFileAction.ID = 'workbench.files.action.createFileFromExplorer';
    CreateFileAction.LABEL = nls.localize('createNewFile', "New File");
    CreateFileAction = __decorate([
        __param(1, workspace_1.IWorkspaceContextService),
        __param(2, editorService_1.IWorkbenchEditorService),
        __param(3, files_2.IFileService),
        __param(4, message_1.IMessageService),
        __param(5, textfiles_1.ITextFileService),
        __param(6, event_1.IEventService)
    ], CreateFileAction);
    return CreateFileAction;
}(BaseCreateAction));
exports.CreateFileAction = CreateFileAction;
/* Create New Folder (only used internally by explorerViewer) */
var CreateFolderAction = (function (_super) {
    __extends(CreateFolderAction, _super);
    function CreateFolderAction(element, contextService, editorService, fileService, messageService, textFileService, eventService) {
        _super.call(this, CreateFolderAction.ID, CreateFolderAction.LABEL, null, contextService, editorService, fileService, messageService, textFileService, eventService);
        this._updateEnablement();
    }
    CreateFolderAction.prototype.runAction = function (fileName) {
        var _this = this;
        return this.fileService.createFolder(uri_1.default.file(paths.join(this.element.parent.resource.fsPath, fileName))).then(null, function (error) {
            _this.onErrorWithRetry(error, function () { return _this.runAction(fileName); });
        });
    };
    CreateFolderAction.ID = 'workbench.files.action.createFolderFromExplorer';
    CreateFolderAction.LABEL = nls.localize('createNewFolder', "New Folder");
    CreateFolderAction = __decorate([
        __param(1, workspace_1.IWorkspaceContextService),
        __param(2, editorService_1.IWorkbenchEditorService),
        __param(3, files_2.IFileService),
        __param(4, message_1.IMessageService),
        __param(5, textfiles_1.ITextFileService),
        __param(6, event_1.IEventService)
    ], CreateFolderAction);
    return CreateFolderAction;
}(BaseCreateAction));
exports.CreateFolderAction = CreateFolderAction;
var BaseDeleteFileAction = (function (_super) {
    __extends(BaseDeleteFileAction, _super);
    function BaseDeleteFileAction(id, label, tree, element, useTrash, contextService, editorService, fileService, messageService, textFileService, eventService) {
        _super.call(this, id, label, contextService, editorService, fileService, messageService, textFileService, eventService);
        this.tree = tree;
        this.element = element;
        this.useTrash = useTrash && !paths.isUNC(element.resource.fsPath); // on UNC shares there is no trash
        this._updateEnablement();
    }
    BaseDeleteFileAction.prototype.run = function (context) {
        var _this = this;
        // Remove highlight
        if (this.tree) {
            this.tree.clearHighlight();
        }
        // Read context
        if (context && context.event) {
            var bypassTrash = (platform_1.isMacintosh && context.event.altKey) || (!platform_1.isMacintosh && context.event.shiftKey);
            if (bypassTrash) {
                this.useTrash = false;
            }
        }
        var primaryButton;
        if (this.useTrash) {
            primaryButton = platform_1.isWindows ? nls.localize('deleteButtonLabelRecycleBin', "&&Move to Recycle Bin") : nls.localize({ key: 'deleteButtonLabelTrash', comment: ['&& denotes a mnemonic'] }, "&&Move to Trash");
        }
        else {
            primaryButton = nls.localize({ key: 'deleteButtonLabel', comment: ['&& denotes a mnemonic'] }, "&&Delete");
        }
        // Handle dirty
        var revertPromise = winjs_base_1.TPromise.as(null);
        var dirty = this.textFileService.getDirty().filter(function (d) { return paths.isEqualOrParent(d.fsPath, _this.element.resource.fsPath); });
        if (dirty.length) {
            var message = void 0;
            if (this.element.isDirectory) {
                if (dirty.length === 1) {
                    message = nls.localize('dirtyMessageFolderOneDelete', "You are deleting a folder with unsaved changes in 1 file. Do you want to continue?");
                }
                else {
                    message = nls.localize('dirtyMessageFolderDelete', "You are deleting a folder with unsaved changes in {0} files. Do you want to continue?", dirty.length);
                }
            }
            else {
                message = nls.localize('dirtyMessageFileDelete', "You are deleting a file with unsaved changes. Do you want to continue?");
            }
            var res = this.messageService.confirm({
                message: message,
                type: 'warning',
                detail: nls.localize('dirtyWarning', "Your changes will be lost if you don't save them."),
                primaryButton: primaryButton
            });
            if (!res) {
                return winjs_base_1.TPromise.as(null);
            }
            this.skipConfirm = true; // since we already asked for confirmation
            revertPromise = this.textFileService.revertAll(dirty);
        }
        // Check if file is dirty in editor and save it to avoid data loss
        return revertPromise.then(function () {
            // Ask for Confirm
            if (!_this.skipConfirm) {
                var confirm_1;
                if (_this.useTrash) {
                    confirm_1 = {
                        message: _this.element.isDirectory ? nls.localize('confirmMoveTrashMessageFolder', "Are you sure you want to delete '{0}' and its contents?", _this.element.name) : nls.localize('confirmMoveTrashMessageFile', "Are you sure you want to delete '{0}'?", _this.element.name),
                        detail: platform_1.isWindows ? nls.localize('undoBin', "You can restore from the recycle bin.") : nls.localize('undoTrash', "You can restore from the trash."),
                        primaryButton: primaryButton
                    };
                }
                else {
                    confirm_1 = {
                        message: _this.element.isDirectory ? nls.localize('confirmDeleteMessageFolder', "Are you sure you want to permanently delete '{0}' and its contents?", _this.element.name) : nls.localize('confirmDeleteMessageFile', "Are you sure you want to permanently delete '{0}'?", _this.element.name),
                        detail: nls.localize('irreversible', "This action is irreversible!"),
                        primaryButton: primaryButton
                    };
                }
                if (!_this.messageService.confirm(confirm_1)) {
                    return winjs_base_1.TPromise.as(null);
                }
            }
            // Since a delete operation can take a while we want to emit the event proactively to avoid issues
            // with stale entries in the explorer tree.
            _this.eventService.emit('files.internal:fileChanged', new textfiles_1.LocalFileChangeEvent(_this.element.clone(), null));
            // Call function
            var servicePromise = _this.fileService.del(_this.element.resource, _this.useTrash).then(function () {
                if (_this.element.parent) {
                    _this.tree.setFocus(_this.element.parent); // move focus to parent
                }
            }, function (error) {
                // Allow to retry
                var extraAction;
                if (_this.useTrash) {
                    extraAction = new actions_1.Action('permanentDelete', nls.localize('permDelete', "Delete Permanently"), null, true, function () { _this.useTrash = false; _this.skipConfirm = true; return _this.run(); });
                }
                _this.onErrorWithRetry(error, function () { return _this.run(); }, extraAction);
                // Since the delete failed, best we can do is to refresh the explorer from the root to show the current state of files.
                var event = new textfiles_1.LocalFileChangeEvent(new explorerViewModel_1.FileStat(_this.contextService.getWorkspace().resource, true, true), new explorerViewModel_1.FileStat(_this.contextService.getWorkspace().resource, true, true));
                _this.eventService.emit('files.internal:fileChanged', event);
                // Focus back to tree
                _this.tree.DOMFocus();
            });
            return servicePromise;
        });
    };
    BaseDeleteFileAction = __decorate([
        __param(5, workspace_1.IWorkspaceContextService),
        __param(6, editorService_1.IWorkbenchEditorService),
        __param(7, files_2.IFileService),
        __param(8, message_1.IMessageService),
        __param(9, textfiles_1.ITextFileService),
        __param(10, event_1.IEventService)
    ], BaseDeleteFileAction);
    return BaseDeleteFileAction;
}(BaseFileAction));
exports.BaseDeleteFileAction = BaseDeleteFileAction;
/* Move File/Folder to trash */
var MoveFileToTrashAction = (function (_super) {
    __extends(MoveFileToTrashAction, _super);
    function MoveFileToTrashAction(tree, element, contextService, editorService, fileService, messageService, textFileService, eventService) {
        _super.call(this, MoveFileToTrashAction.ID, nls.localize('delete', "Delete"), tree, element, true, contextService, editorService, fileService, messageService, textFileService, eventService);
    }
    MoveFileToTrashAction.ID = 'workbench.files.action.moveFileToTrash';
    MoveFileToTrashAction = __decorate([
        __param(2, workspace_1.IWorkspaceContextService),
        __param(3, editorService_1.IWorkbenchEditorService),
        __param(4, files_2.IFileService),
        __param(5, message_1.IMessageService),
        __param(6, textfiles_1.ITextFileService),
        __param(7, event_1.IEventService)
    ], MoveFileToTrashAction);
    return MoveFileToTrashAction;
}(BaseDeleteFileAction));
exports.MoveFileToTrashAction = MoveFileToTrashAction;
/* Import File */
var ImportFileAction = (function (_super) {
    __extends(ImportFileAction, _super);
    function ImportFileAction(tree, element, clazz, contextService, editorService, fileService, messageService, textFileService, eventService) {
        _super.call(this, ImportFileAction.ID, nls.localize('importFiles', "Import Files"), contextService, editorService, fileService, messageService, textFileService, eventService);
        this.tree = tree;
        this.element = element;
        if (clazz) {
            this.class = clazz;
        }
        this._updateEnablement();
    }
    ImportFileAction.prototype.getViewer = function () {
        return this.tree;
    };
    ImportFileAction.prototype.run = function (context) {
        var _this = this;
        var importPromise = winjs_base_1.TPromise.as(null).then(function () {
            var input = context.input;
            if (input.files && input.files.length > 0) {
                // Find parent for import
                var targetElement_1;
                if (_this.element) {
                    targetElement_1 = _this.element;
                }
                else {
                    targetElement_1 = _this.tree.getFocus() || _this.tree.getInput();
                }
                if (!targetElement_1.isDirectory) {
                    targetElement_1 = targetElement_1.parent;
                }
                // Create real files array
                var filesArray_1 = [];
                for (var i = 0; i < input.files.length; i++) {
                    var file = input.files[i];
                    filesArray_1.push(file);
                }
                // Resolve target to check for name collisions and ask user
                return _this.fileService.resolveFile(targetElement_1.resource).then(function (targetStat) {
                    // Check for name collisions
                    var targetNames = {};
                    targetStat.children.forEach(function (child) {
                        targetNames[platform_1.isLinux ? child.name : child.name.toLowerCase()] = child;
                    });
                    var overwrite = true;
                    if (filesArray_1.some(function (file) {
                        return !!targetNames[platform_1.isLinux ? file.name : file.name.toLowerCase()];
                    })) {
                        var confirm_2 = {
                            message: nls.localize('confirmOverwrite', "A file or folder with the same name already exists in the destination folder. Do you want to replace it?"),
                            detail: nls.localize('irreversible', "This action is irreversible!"),
                            primaryButton: nls.localize({ key: 'replaceButtonLabel', comment: ['&& denotes a mnemonic'] }, "&&Replace")
                        };
                        overwrite = _this.messageService.confirm(confirm_2);
                    }
                    if (!overwrite) {
                        return;
                    }
                    // Run import in sequence
                    var importPromisesFactory = [];
                    filesArray_1.forEach(function (file) {
                        importPromisesFactory.push(function () {
                            var sourceFile = uri_1.default.file(file.path);
                            return _this.fileService.importFile(sourceFile, targetElement_1.resource).then(function (result) {
                                if (result.stat) {
                                    // Emit Deleted Event if file gets replaced unless it is the same file
                                    var oldFile = targetNames[platform_1.isLinux ? file.name : file.name.toLowerCase()];
                                    if (oldFile && oldFile.resource.fsPath !== result.stat.resource.fsPath) {
                                        _this.eventService.emit('files.internal:fileChanged', new textfiles_1.LocalFileChangeEvent(oldFile, null));
                                    }
                                    // Emit Import Event
                                    _this.eventService.emit('files.internal:fileChanged', new FileImportedEvent(result.stat, result.isNew, context.event));
                                }
                            }, function (error) {
                                _this.messageService.show(message_1.Severity.Error, error);
                            });
                        });
                    });
                    return async_1.sequence(importPromisesFactory);
                });
            }
        });
        return importPromise.then(function () {
            _this.tree.clearHighlight();
        }, function (error) {
            _this.onError(error);
            _this.tree.clearHighlight();
        });
    };
    ImportFileAction.ID = 'workbench.files.action.importFile';
    ImportFileAction = __decorate([
        __param(3, workspace_1.IWorkspaceContextService),
        __param(4, editorService_1.IWorkbenchEditorService),
        __param(5, files_2.IFileService),
        __param(6, message_1.IMessageService),
        __param(7, textfiles_1.ITextFileService),
        __param(8, event_1.IEventService)
    ], ImportFileAction);
    return ImportFileAction;
}(BaseFileAction));
exports.ImportFileAction = ImportFileAction;
/** File import event is emitted when a file is import into the workbench. */
var FileImportedEvent = (function (_super) {
    __extends(FileImportedEvent, _super);
    function FileImportedEvent(stat, isNew, originalEvent) {
        _super.call(this, null, stat, originalEvent);
        this.isNew = isNew;
    }
    FileImportedEvent.prototype.gotAdded = function () {
        return this.isNew;
    };
    FileImportedEvent.prototype.gotMoved = function () {
        return false;
    };
    FileImportedEvent.prototype.gotUpdated = function () {
        return !this.isNew;
    };
    FileImportedEvent.prototype.gotDeleted = function () {
        return false;
    };
    return FileImportedEvent;
}(textfiles_1.LocalFileChangeEvent));
exports.FileImportedEvent = FileImportedEvent;
// Copy File/Folder
var fileToCopy;
var CopyFileAction = (function (_super) {
    __extends(CopyFileAction, _super);
    function CopyFileAction(tree, element, contextService, editorService, fileService, messageService, textFileService, eventService) {
        _super.call(this, CopyFileAction.ID, nls.localize('copyFile', "Copy"), contextService, editorService, fileService, messageService, textFileService, eventService);
        this.tree = tree;
        this.element = element;
        this._updateEnablement();
    }
    CopyFileAction.prototype.run = function () {
        // Remember as file/folder to copy
        fileToCopy = this.element;
        // Remove highlight
        if (this.tree) {
            this.tree.clearHighlight();
        }
        this.tree.DOMFocus();
        return winjs_base_1.TPromise.as(null);
    };
    CopyFileAction.ID = 'workbench.files.action.copyFile';
    CopyFileAction = __decorate([
        __param(2, workspace_1.IWorkspaceContextService),
        __param(3, editorService_1.IWorkbenchEditorService),
        __param(4, files_2.IFileService),
        __param(5, message_1.IMessageService),
        __param(6, textfiles_1.ITextFileService),
        __param(7, event_1.IEventService)
    ], CopyFileAction);
    return CopyFileAction;
}(BaseFileAction));
exports.CopyFileAction = CopyFileAction;
// Paste File/Folder
var PasteFileAction = (function (_super) {
    __extends(PasteFileAction, _super);
    function PasteFileAction(tree, element, contextService, editorService, fileService, messageService, textFileService, eventService, instantiationService) {
        _super.call(this, PasteFileAction.ID, nls.localize('pasteFile', "Paste"), contextService, editorService, fileService, messageService, textFileService, eventService);
        this.instantiationService = instantiationService;
        this.tree = tree;
        this.element = element;
        this._updateEnablement();
    }
    PasteFileAction.prototype._isEnabled = function () {
        // Need at least a file to copy
        if (!fileToCopy) {
            return false;
        }
        // Check if file was deleted or moved meanwhile
        var root = this.tree.getInput();
        var exists = root.find(fileToCopy.resource);
        if (!exists) {
            fileToCopy = null;
            return false;
        }
        // Check if target is ancestor of pasted folder
        if (this.element.resource.toString() !== fileToCopy.resource.toString() && paths.isEqualOrParent(this.element.resource.fsPath, fileToCopy.resource.fsPath)) {
            return false;
        }
        return true;
    };
    PasteFileAction.prototype.run = function () {
        var _this = this;
        // Find target
        var target;
        if (this.element.resource.toString() === fileToCopy.resource.toString()) {
            target = this.element.parent;
        }
        else {
            target = this.element.isDirectory ? this.element : this.element.parent;
        }
        // Reuse duplicate action
        var pasteAction = this.instantiationService.createInstance(DuplicateFileAction, this.tree, fileToCopy, target);
        return pasteAction.run().then(function () {
            _this.tree.DOMFocus();
        });
    };
    PasteFileAction.ID = 'workbench.files.action.pasteFile';
    PasteFileAction = __decorate([
        __param(2, workspace_1.IWorkspaceContextService),
        __param(3, editorService_1.IWorkbenchEditorService),
        __param(4, files_2.IFileService),
        __param(5, message_1.IMessageService),
        __param(6, textfiles_1.ITextFileService),
        __param(7, event_1.IEventService),
        __param(8, instantiation_1.IInstantiationService)
    ], PasteFileAction);
    return PasteFileAction;
}(BaseFileAction));
exports.PasteFileAction = PasteFileAction;
// Duplicate File/Folder
var DuplicateFileAction = (function (_super) {
    __extends(DuplicateFileAction, _super);
    function DuplicateFileAction(tree, element, target, contextService, editorService, fileService, messageService, textFileService, eventService) {
        _super.call(this, 'workbench.files.action.duplicateFile', nls.localize('duplicateFile', "Duplicate"), contextService, editorService, fileService, messageService, textFileService, eventService);
        this.tree = tree;
        this.element = element;
        this.target = (target && target.isDirectory) ? target : element.parent;
        this._updateEnablement();
    }
    DuplicateFileAction.prototype.run = function () {
        var _this = this;
        // Remove highlight
        if (this.tree) {
            this.tree.clearHighlight();
        }
        // Copy File and emit event
        var result = this.fileService.copyFile(this.element.resource, this.findTarget()).then(function (stat) {
            _this.eventService.emit('files.internal:fileChanged', new textfiles_1.LocalFileChangeEvent(null, stat));
        }, function (error) {
            _this.onError(error);
        });
        return result;
    };
    DuplicateFileAction.prototype.onError = function (error) {
        this.messageService.show(message_1.Severity.Error, error);
    };
    DuplicateFileAction.prototype.findTarget = function () {
        var root = this.tree.getInput();
        var name = this.element.name;
        var candidate = uri_1.default.file(paths.join(this.target.resource.fsPath, name));
        while (true) {
            if (!root.find(candidate)) {
                break;
            }
            name = this.toCopyName(name, this.element.isDirectory);
            candidate = uri_1.default.file(paths.join(this.target.resource.fsPath, name));
        }
        return candidate;
    };
    DuplicateFileAction.prototype.toCopyName = function (name, isFolder) {
        // file.1.txt=>file.2.txt
        if (!isFolder && name.match(/(\d+)(\..*)$/)) {
            return name.replace(/(\d+)(\..*)$/, function (match, g1, g2) { return (parseInt(g1) + 1) + g2; });
        }
        // file.txt=>file.1.txt
        var lastIndexOfDot = name.lastIndexOf('.');
        if (!isFolder && lastIndexOfDot >= 0) {
            return strings.format('{0}.1{1}', name.substr(0, lastIndexOfDot), name.substr(lastIndexOfDot));
        }
        // folder.1=>folder.2
        if (isFolder && name.match(/(\d+)$/)) {
            return name.replace(/(\d+)$/, function (match) {
                var groups = [];
                for (var _i = 1; _i < arguments.length; _i++) {
                    groups[_i - 1] = arguments[_i];
                }
                return String(parseInt(groups[0]) + 1);
            });
        }
        // file/folder=>file.1/folder.1
        return strings.format('{0}.1', name);
    };
    DuplicateFileAction = __decorate([
        __param(3, workspace_1.IWorkspaceContextService),
        __param(4, editorService_1.IWorkbenchEditorService),
        __param(5, files_2.IFileService),
        __param(6, message_1.IMessageService),
        __param(7, textfiles_1.ITextFileService),
        __param(8, event_1.IEventService)
    ], DuplicateFileAction);
    return DuplicateFileAction;
}(BaseFileAction));
exports.DuplicateFileAction = DuplicateFileAction;
// Open to the side
var OpenToSideAction = (function (_super) {
    __extends(OpenToSideAction, _super);
    function OpenToSideAction(tree, resource, preserveFocus, editorService) {
        _super.call(this, OpenToSideAction.ID, OpenToSideAction.LABEL);
        this.editorService = editorService;
        this.tree = tree;
        this.preserveFocus = preserveFocus;
        this.resource = resource;
        this.updateEnablement();
    }
    OpenToSideAction.prototype.updateEnablement = function () {
        var activeEditor = this.editorService.getActiveEditor();
        this.enabled = (!activeEditor || activeEditor.position !== editor_2.Position.RIGHT);
    };
    OpenToSideAction.prototype.run = function () {
        // Remove highlight
        this.tree.clearHighlight();
        // Set side input
        return this.editorService.openEditor({
            resource: this.resource,
            options: {
                preserveFocus: this.preserveFocus
            }
        }, true);
    };
    OpenToSideAction.ID = 'workbench.files.action.openToSide';
    OpenToSideAction.LABEL = nls.localize('openToSide', "Open to the Side");
    OpenToSideAction = __decorate([
        __param(3, editorService_1.IWorkbenchEditorService)
    ], OpenToSideAction);
    return OpenToSideAction;
}(actions_1.Action));
exports.OpenToSideAction = OpenToSideAction;
var globalResourceToCompare;
var SelectResourceForCompareAction = (function (_super) {
    __extends(SelectResourceForCompareAction, _super);
    function SelectResourceForCompareAction(resource, tree) {
        _super.call(this, 'workbench.files.action.selectForCompare', nls.localize('compareSource', "Select for Compare"));
        this.tree = tree;
        this.resource = resource;
        this.enabled = true;
    }
    SelectResourceForCompareAction.prototype.run = function () {
        // Remember as source file to compare
        globalResourceToCompare = this.resource;
        // Remove highlight
        if (this.tree) {
            this.tree.clearHighlight();
            this.tree.DOMFocus();
        }
        return winjs_base_1.TPromise.as(null);
    };
    return SelectResourceForCompareAction;
}(actions_1.Action));
exports.SelectResourceForCompareAction = SelectResourceForCompareAction;
// Global Compare with
var GlobalCompareResourcesAction = (function (_super) {
    __extends(GlobalCompareResourcesAction, _super);
    function GlobalCompareResourcesAction(id, label, quickOpenService, instantiationService, editorService, editorGroupService, messageService) {
        _super.call(this, id, label);
        this.quickOpenService = quickOpenService;
        this.instantiationService = instantiationService;
        this.editorService = editorService;
        this.editorGroupService = editorGroupService;
        this.messageService = messageService;
    }
    GlobalCompareResourcesAction.prototype.run = function () {
        var _this = this;
        var fileInput = editor_1.asFileEditorInput(this.editorService.getActiveEditorInput());
        if (fileInput) {
            // Keep as resource to compare
            globalResourceToCompare = fileInput.getResource();
            // Listen for next editor to open
            var unbind_1 = this.editorGroupService.onEditorOpening(function (e) {
                unbind_1.dispose(); // listen once
                var otherFileInput = editor_1.asFileEditorInput(e.editorInput);
                if (otherFileInput) {
                    var compareAction_1 = _this.instantiationService.createInstance(CompareResourcesAction, otherFileInput.getResource(), null);
                    if (compareAction_1._isEnabled()) {
                        e.prevent();
                        compareAction_1.run().done(function () { return compareAction_1.dispose(); });
                    }
                    else {
                        _this.messageService.show(message_1.Severity.Info, nls.localize('unableToFileToCompare', "The selected file can not be compared with '{0}'.", paths.basename(globalResourceToCompare.fsPath)));
                    }
                }
            });
            // Bring up quick open
            this.quickOpenService.show().then(function () {
                unbind_1.dispose(); // make sure to unbind if quick open is closing
            });
        }
        else {
            this.messageService.show(message_1.Severity.Info, nls.localize('openFileToCompare', "Open a file first to compare it with another file."));
        }
        return winjs_base_1.TPromise.as(true);
    };
    GlobalCompareResourcesAction.ID = 'workbench.files.action.compareFileWith';
    GlobalCompareResourcesAction.LABEL = nls.localize('globalCompareFile', "Compare Active File With...");
    GlobalCompareResourcesAction = __decorate([
        __param(2, quickOpenService_1.IQuickOpenService),
        __param(3, instantiation_1.IInstantiationService),
        __param(4, editorService_1.IWorkbenchEditorService),
        __param(5, groupService_1.IEditorGroupService),
        __param(6, message_1.IMessageService)
    ], GlobalCompareResourcesAction);
    return GlobalCompareResourcesAction;
}(actions_1.Action));
exports.GlobalCompareResourcesAction = GlobalCompareResourcesAction;
// Compare with Resource
var CompareResourcesAction = (function (_super) {
    __extends(CompareResourcesAction, _super);
    function CompareResourcesAction(resource, tree, contextService, instantiationService, editorService) {
        _super.call(this, 'workbench.files.action.compareFiles', CompareResourcesAction.computeLabel());
        this.contextService = contextService;
        this.instantiationService = instantiationService;
        this.editorService = editorService;
        this.tree = tree;
        this.resource = resource;
    }
    CompareResourcesAction.computeLabel = function () {
        if (globalResourceToCompare) {
            return nls.localize('compareWith', "Compare with '{0}'", paths.basename(globalResourceToCompare.fsPath));
        }
        return nls.localize('compareFiles', "Compare Files");
    };
    CompareResourcesAction.prototype.getLabel = function () {
        return CompareResourcesAction.computeLabel();
    };
    CompareResourcesAction.prototype._isEnabled = function () {
        // Need at least a resource to compare
        if (!globalResourceToCompare) {
            return false;
        }
        // Check if file was deleted or moved meanwhile (explorer only)
        if (this.tree) {
            var root = this.tree.getInput();
            if (root instanceof explorerViewModel_1.FileStat) {
                var exists = root.find(globalResourceToCompare);
                if (!exists) {
                    globalResourceToCompare = null;
                    return false;
                }
            }
        }
        // Check if target is identical to source
        if (this.resource.toString() === globalResourceToCompare.toString()) {
            return false;
        }
        return true;
    };
    CompareResourcesAction.prototype.run = function () {
        // Remove highlight
        if (this.tree) {
            this.tree.clearHighlight();
        }
        var leftInput = this.instantiationService.createInstance(fileEditorInput_1.FileEditorInput, globalResourceToCompare, void 0);
        var rightInput = this.instantiationService.createInstance(fileEditorInput_1.FileEditorInput, this.resource, void 0);
        return this.editorService.openEditor(new diffEditorInput_1.DiffEditorInput(diffEditorInput_1.toDiffLabel(globalResourceToCompare, this.resource, this.contextService), null, leftInput, rightInput));
    };
    CompareResourcesAction = __decorate([
        __param(2, workspace_1.IWorkspaceContextService),
        __param(3, instantiation_1.IInstantiationService),
        __param(4, editorService_1.IWorkbenchEditorService)
    ], CompareResourcesAction);
    return CompareResourcesAction;
}(actions_1.Action));
exports.CompareResourcesAction = CompareResourcesAction;
// Refresh Explorer Viewer
var RefreshViewExplorerAction = (function (_super) {
    __extends(RefreshViewExplorerAction, _super);
    function RefreshViewExplorerAction(explorerView, clazz) {
        _super.call(this, 'workbench.files.action.refreshExplorer', nls.localize('refresh', "Refresh"), clazz, true, function (context) { return explorerView.refresh(); });
    }
    return RefreshViewExplorerAction;
}(actions_1.Action));
exports.RefreshViewExplorerAction = RefreshViewExplorerAction;
var BaseActionWithErrorReporting = (function (_super) {
    __extends(BaseActionWithErrorReporting, _super);
    function BaseActionWithErrorReporting(id, label, messageService) {
        _super.call(this, id, label);
        this.messageService = messageService;
    }
    BaseActionWithErrorReporting.prototype.run = function (context) {
        var _this = this;
        return this.doRun(context).then(function () { return true; }, function (error) {
            _this.messageService.show(message_1.Severity.Error, errorMessage_1.toErrorMessage(error, false));
        });
    };
    return BaseActionWithErrorReporting;
}(actions_1.Action));
exports.BaseActionWithErrorReporting = BaseActionWithErrorReporting;
var BaseSaveFileAction = (function (_super) {
    __extends(BaseSaveFileAction, _super);
    function BaseSaveFileAction(id, label, editorService, textFileService, untitledEditorService, messageService) {
        _super.call(this, id, label, messageService);
        this.editorService = editorService;
        this.textFileService = textFileService;
        this.untitledEditorService = untitledEditorService;
        this.enabled = true;
    }
    BaseSaveFileAction.prototype.setResource = function (resource) {
        this.resource = resource;
    };
    BaseSaveFileAction.prototype.doRun = function (context) {
        var _this = this;
        var source;
        if (this.resource) {
            source = this.resource;
        }
        else {
            source = editor_1.getUntitledOrFileResource(this.editorService.getActiveEditorInput(), true);
        }
        if (source) {
            // Save As (or Save untitled with associated path)
            if (this.isSaveAs() || source.scheme === 'untitled') {
                var encodingOfSource_1;
                if (source.scheme === 'untitled') {
                    encodingOfSource_1 = this.untitledEditorService.get(source).getEncoding();
                }
                else if (source.scheme === 'file') {
                    var textModel = this.textFileService.models.get(source);
                    encodingOfSource_1 = textModel && textModel.getEncoding(); // text model can be null e.g. if this is a binary file!
                }
                var selectionOfSource_1;
                var activeEditor = this.editorService.getActiveEditor();
                if (activeEditor instanceof textEditor_1.BaseTextEditor) {
                    var activeResource = editor_1.getUntitledOrFileResource(activeEditor.input, true);
                    if (activeResource && activeResource.toString() === source.toString()) {
                        selectionOfSource_1 = activeEditor.getSelection();
                    }
                }
                // Special case: an untitled file with associated path gets saved directly unless "saveAs" is true
                var savePromise = void 0;
                if (!this.isSaveAs() && source.scheme === 'untitled' && this.untitledEditorService.hasAssociatedFilePath(source)) {
                    savePromise = this.textFileService.save(source).then(function (result) {
                        if (result) {
                            return uri_1.default.file(source.fsPath);
                        }
                        return null;
                    });
                }
                else {
                    savePromise = this.textFileService.saveAs(source);
                }
                return savePromise.then(function (target) {
                    if (!target || target.toString() === source.toString()) {
                        return; // save canceled or same resource used
                    }
                    var replaceWith = {
                        resource: target,
                        encoding: encodingOfSource_1,
                        options: {
                            pinned: true,
                            selection: selectionOfSource_1
                        }
                    };
                    return _this.editorService.replaceEditors([{
                            toReplace: { resource: source },
                            replaceWith: replaceWith
                        }]).then(function () { return true; });
                });
            }
            // Just save
            return this.textFileService.save(source, { force: true /* force a change to the file to trigger external watchers if any */ });
        }
        return winjs_base_1.TPromise.as(false);
    };
    BaseSaveFileAction = __decorate([
        __param(2, editorService_1.IWorkbenchEditorService),
        __param(3, textfiles_1.ITextFileService),
        __param(4, untitledEditorService_1.IUntitledEditorService),
        __param(5, message_1.IMessageService)
    ], BaseSaveFileAction);
    return BaseSaveFileAction;
}(BaseActionWithErrorReporting));
exports.BaseSaveFileAction = BaseSaveFileAction;
var SaveFileAction = (function (_super) {
    __extends(SaveFileAction, _super);
    function SaveFileAction() {
        _super.apply(this, arguments);
    }
    SaveFileAction.prototype.isSaveAs = function () {
        return false;
    };
    SaveFileAction.ID = 'workbench.action.files.save';
    SaveFileAction.LABEL = nls.localize('save', "Save");
    return SaveFileAction;
}(BaseSaveFileAction));
exports.SaveFileAction = SaveFileAction;
var SaveFileAsAction = (function (_super) {
    __extends(SaveFileAsAction, _super);
    function SaveFileAsAction() {
        _super.apply(this, arguments);
    }
    SaveFileAsAction.prototype.isSaveAs = function () {
        return true;
    };
    SaveFileAsAction.ID = 'workbench.action.files.saveAs';
    SaveFileAsAction.LABEL = nls.localize('saveAs', "Save As...");
    return SaveFileAsAction;
}(BaseSaveFileAction));
exports.SaveFileAsAction = SaveFileAsAction;
var BaseSaveAllAction = (function (_super) {
    __extends(BaseSaveAllAction, _super);
    function BaseSaveAllAction(id, label, editorService, editorGroupService, textFileService, untitledEditorService, messageService) {
        _super.call(this, id, label, messageService);
        this.editorService = editorService;
        this.editorGroupService = editorGroupService;
        this.textFileService = textFileService;
        this.untitledEditorService = untitledEditorService;
        this.toDispose = [];
        this.lastIsDirty = this.textFileService.isDirty();
        this.enabled = this.lastIsDirty;
        this.registerListeners();
    }
    BaseSaveAllAction.prototype.registerListeners = function () {
        var _this = this;
        // listen to files being changed locally
        this.toDispose.push(this.textFileService.models.onModelDirty(function (e) { return _this.updateEnablement(true); }));
        this.toDispose.push(this.textFileService.models.onModelSaved(function (e) { return _this.updateEnablement(false); }));
        this.toDispose.push(this.textFileService.models.onModelReverted(function (e) { return _this.updateEnablement(false); }));
        this.toDispose.push(this.textFileService.models.onModelSaveError(function (e) { return _this.updateEnablement(true); }));
        if (this.includeUntitled()) {
            this.toDispose.push(this.untitledEditorService.onDidChangeDirty(function (resource) { return _this.updateEnablement(_this.untitledEditorService.isDirty(resource)); }));
        }
    };
    BaseSaveAllAction.prototype.updateEnablement = function (isDirty) {
        if (this.lastIsDirty !== isDirty) {
            this.enabled = this.textFileService.isDirty();
            this.lastIsDirty = this.enabled;
        }
    };
    BaseSaveAllAction.prototype.doRun = function (context) {
        var _this = this;
        var stacks = this.editorGroupService.getStacksModel();
        // Store some properties per untitled file to restore later after save is completed
        var mapUntitledToProperties = Object.create(null);
        this.textFileService.getDirty()
            .filter(function (r) { return r.scheme === 'untitled'; }) // All untitled resources
            .map(function (r) { return _this.untitledEditorService.get(r); }) // Mapped to their inputs
            .filter(function (input) { return !!input; }) // If possible :)
            .forEach(function (input) {
            mapUntitledToProperties[input.getResource().toString()] = {
                encoding: input.getEncoding(),
                indexInGroups: stacks.groups.map(function (g) { return g.indexOf(input); }),
                activeInGroups: stacks.groups.map(function (g) { return g.isActive(input); })
            };
        });
        // Save all
        return this.textFileService.saveAll(this.getSaveAllArguments(context)).then(function (results) {
            // Reopen saved untitled editors
            var untitledToReopen = [];
            results.results.forEach(function (result) {
                if (!result.success || result.source.scheme !== 'untitled') {
                    return;
                }
                var untitledProps = mapUntitledToProperties[result.source.toString()];
                if (!untitledProps) {
                    return;
                }
                // For each position where the untitled file was opened
                untitledProps.indexInGroups.forEach(function (indexInGroup, index) {
                    if (indexInGroup >= 0) {
                        untitledToReopen.push({
                            input: {
                                resource: result.target,
                                encoding: untitledProps.encoding,
                                options: {
                                    pinned: true,
                                    index: indexInGroup,
                                    preserveFocus: true,
                                    inactive: !untitledProps.activeInGroups[index]
                                }
                            },
                            position: index
                        });
                    }
                });
            });
            if (untitledToReopen.length) {
                return _this.editorService.openEditors(untitledToReopen).then(function () { return true; });
            }
        });
    };
    BaseSaveAllAction.prototype.dispose = function () {
        this.toDispose = lifecycle_1.dispose(this.toDispose);
        _super.prototype.dispose.call(this);
    };
    BaseSaveAllAction = __decorate([
        __param(2, editorService_1.IWorkbenchEditorService),
        __param(3, groupService_1.IEditorGroupService),
        __param(4, textfiles_1.ITextFileService),
        __param(5, untitledEditorService_1.IUntitledEditorService),
        __param(6, message_1.IMessageService)
    ], BaseSaveAllAction);
    return BaseSaveAllAction;
}(BaseActionWithErrorReporting));
exports.BaseSaveAllAction = BaseSaveAllAction;
var SaveAllAction = (function (_super) {
    __extends(SaveAllAction, _super);
    function SaveAllAction() {
        _super.apply(this, arguments);
    }
    Object.defineProperty(SaveAllAction.prototype, "class", {
        get: function () {
            return 'explorer-action save-all';
        },
        enumerable: true,
        configurable: true
    });
    SaveAllAction.prototype.getSaveAllArguments = function () {
        return this.includeUntitled();
    };
    SaveAllAction.prototype.includeUntitled = function () {
        return true;
    };
    SaveAllAction.ID = 'workbench.action.files.saveAll';
    SaveAllAction.LABEL = nls.localize('saveAll', "Save All");
    return SaveAllAction;
}(BaseSaveAllAction));
exports.SaveAllAction = SaveAllAction;
var SaveAllInGroupAction = (function (_super) {
    __extends(SaveAllInGroupAction, _super);
    function SaveAllInGroupAction() {
        _super.apply(this, arguments);
    }
    Object.defineProperty(SaveAllInGroupAction.prototype, "class", {
        get: function () {
            return 'explorer-action save-all';
        },
        enumerable: true,
        configurable: true
    });
    SaveAllInGroupAction.prototype.getSaveAllArguments = function (editorIdentifier) {
        if (!editorIdentifier) {
            return this.includeUntitled();
        }
        var editorGroup = editorIdentifier.group;
        var resourcesToSave = [];
        editorGroup.getEditors().forEach(function (editor) {
            var resource = editor_1.getUntitledOrFileResource(editor, true);
            if (resource) {
                resourcesToSave.push(resource);
            }
        });
        return resourcesToSave;
    };
    SaveAllInGroupAction.prototype.includeUntitled = function () {
        return true;
    };
    SaveAllInGroupAction.ID = 'workbench.files.action.saveAllInGroup';
    SaveAllInGroupAction.LABEL = nls.localize('saveAllInGroup', "Save All in Group");
    return SaveAllInGroupAction;
}(BaseSaveAllAction));
exports.SaveAllInGroupAction = SaveAllInGroupAction;
var SaveFilesAction = (function (_super) {
    __extends(SaveFilesAction, _super);
    function SaveFilesAction() {
        _super.apply(this, arguments);
    }
    SaveFilesAction.prototype.getSaveAllArguments = function () {
        return this.includeUntitled();
    };
    SaveFilesAction.prototype.includeUntitled = function () {
        return false;
    };
    SaveFilesAction.ID = 'workbench.action.files.saveFiles';
    SaveFilesAction.LABEL = nls.localize('saveFiles', "Save Dirty Files");
    return SaveFilesAction;
}(BaseSaveAllAction));
exports.SaveFilesAction = SaveFilesAction;
var RevertFileAction = (function (_super) {
    __extends(RevertFileAction, _super);
    function RevertFileAction(id, label, editorService, textFileService) {
        _super.call(this, id, label);
        this.editorService = editorService;
        this.textFileService = textFileService;
        this.enabled = true;
    }
    RevertFileAction.prototype.setResource = function (resource) {
        this.resource = resource;
    };
    RevertFileAction.prototype.run = function () {
        var resource;
        if (this.resource) {
            resource = this.resource;
        }
        else {
            var activeFileInput = editor_1.asFileEditorInput(this.editorService.getActiveEditorInput(), true);
            if (activeFileInput) {
                resource = activeFileInput.getResource();
            }
        }
        if (resource && resource.scheme !== 'untitled') {
            return this.textFileService.revert(resource, true /* force */);
        }
        return winjs_base_1.TPromise.as(true);
    };
    RevertFileAction.ID = 'workbench.action.files.revert';
    RevertFileAction.LABEL = nls.localize('revert', "Revert File");
    RevertFileAction = __decorate([
        __param(2, editorService_1.IWorkbenchEditorService),
        __param(3, textfiles_1.ITextFileService)
    ], RevertFileAction);
    return RevertFileAction;
}(actions_1.Action));
exports.RevertFileAction = RevertFileAction;
var FocusOpenEditorsView = (function (_super) {
    __extends(FocusOpenEditorsView, _super);
    function FocusOpenEditorsView(id, label, viewletService) {
        _super.call(this, id, label);
        this.viewletService = viewletService;
    }
    FocusOpenEditorsView.prototype.run = function () {
        return this.viewletService.openViewlet(files_1.VIEWLET_ID, true).then(function (viewlet) {
            var openEditorsView = viewlet.getOpenEditorsView();
            if (openEditorsView) {
                openEditorsView.expand();
                openEditorsView.getViewer().DOMFocus();
            }
        });
    };
    FocusOpenEditorsView.ID = 'workbench.files.action.focusOpenEditorsView';
    FocusOpenEditorsView.LABEL = nls.localize({ key: 'focusOpenEditors', comment: ['Open is an adjective'] }, "Focus on Open Editors View");
    FocusOpenEditorsView = __decorate([
        __param(2, viewletService_1.IViewletService)
    ], FocusOpenEditorsView);
    return FocusOpenEditorsView;
}(actions_1.Action));
exports.FocusOpenEditorsView = FocusOpenEditorsView;
var FocusFilesExplorer = (function (_super) {
    __extends(FocusFilesExplorer, _super);
    function FocusFilesExplorer(id, label, viewletService) {
        _super.call(this, id, label);
        this.viewletService = viewletService;
    }
    FocusFilesExplorer.prototype.run = function () {
        return this.viewletService.openViewlet(files_1.VIEWLET_ID, true).then(function (viewlet) {
            var view = viewlet.getExplorerView();
            if (view) {
                view.expand();
                view.getViewer().DOMFocus();
            }
        });
    };
    FocusFilesExplorer.ID = 'workbench.files.action.focusFilesExplorer';
    FocusFilesExplorer.LABEL = nls.localize('focusFilesExplorer', "Focus on Files Explorer");
    FocusFilesExplorer = __decorate([
        __param(2, viewletService_1.IViewletService)
    ], FocusFilesExplorer);
    return FocusFilesExplorer;
}(actions_1.Action));
exports.FocusFilesExplorer = FocusFilesExplorer;
var ShowActiveFileInExplorer = (function (_super) {
    __extends(ShowActiveFileInExplorer, _super);
    function ShowActiveFileInExplorer(id, label, editorService, viewletService, contextService, messageService) {
        _super.call(this, id, label);
        this.editorService = editorService;
        this.viewletService = viewletService;
        this.contextService = contextService;
        this.messageService = messageService;
    }
    ShowActiveFileInExplorer.prototype.run = function () {
        var _this = this;
        var fileInput = editor_1.asFileEditorInput(this.editorService.getActiveEditorInput(), true);
        if (fileInput) {
            return this.viewletService.openViewlet(files_1.VIEWLET_ID, false).then(function (viewlet) {
                var isInsideWorkspace = _this.contextService.isInsideWorkspace(fileInput.getResource());
                if (isInsideWorkspace) {
                    var explorerView = viewlet.getExplorerView();
                    if (explorerView) {
                        explorerView.expand();
                        explorerView.select(fileInput.getResource(), true);
                    }
                }
                else {
                    var openEditorsView = viewlet.getOpenEditorsView();
                    if (openEditorsView) {
                        openEditorsView.expand();
                    }
                }
            });
        }
        else {
            this.messageService.show(severity_1.default.Info, nls.localize('openFileToShow', "Open a file first to show it in the explorer"));
        }
        return winjs_base_1.TPromise.as(true);
    };
    ShowActiveFileInExplorer.ID = 'workbench.files.action.showActiveFileInExplorer';
    ShowActiveFileInExplorer.LABEL = nls.localize('showInExplorer', "Show Active File in Explorer");
    ShowActiveFileInExplorer = __decorate([
        __param(2, editorService_1.IWorkbenchEditorService),
        __param(3, viewletService_1.IViewletService),
        __param(4, workspace_1.IWorkspaceContextService),
        __param(5, message_1.IMessageService)
    ], ShowActiveFileInExplorer);
    return ShowActiveFileInExplorer;
}(actions_1.Action));
exports.ShowActiveFileInExplorer = ShowActiveFileInExplorer;
var CollapseExplorerView = (function (_super) {
    __extends(CollapseExplorerView, _super);
    function CollapseExplorerView(id, label, viewletService) {
        _super.call(this, id, label);
        this.viewletService = viewletService;
    }
    CollapseExplorerView.prototype.run = function () {
        return this.viewletService.openViewlet(files_1.VIEWLET_ID, true).then(function (viewlet) {
            var explorerView = viewlet.getExplorerView();
            if (explorerView) {
                var viewer = explorerView.getViewer();
                if (viewer) {
                    var action = new viewlet_1.CollapseAction(viewer, true, null);
                    action.run().done();
                    action.dispose();
                }
            }
        });
    };
    CollapseExplorerView.ID = 'workbench.files.action.collapseFilesExplorerFolders';
    CollapseExplorerView.LABEL = nls.localize('collapseExplorerFolders', "Collapse Folders in Explorer");
    CollapseExplorerView = __decorate([
        __param(2, viewletService_1.IViewletService)
    ], CollapseExplorerView);
    return CollapseExplorerView;
}(actions_1.Action));
exports.CollapseExplorerView = CollapseExplorerView;
var RefreshExplorerView = (function (_super) {
    __extends(RefreshExplorerView, _super);
    function RefreshExplorerView(id, label, viewletService) {
        _super.call(this, id, label);
        this.viewletService = viewletService;
    }
    RefreshExplorerView.prototype.run = function () {
        return this.viewletService.openViewlet(files_1.VIEWLET_ID, true).then(function (viewlet) {
            var explorerView = viewlet.getExplorerView();
            if (explorerView) {
                explorerView.refresh();
            }
        });
    };
    RefreshExplorerView.ID = 'workbench.files.action.refreshFilesExplorer';
    RefreshExplorerView.LABEL = nls.localize('refreshExplorer', "Refresh Explorer");
    RefreshExplorerView = __decorate([
        __param(2, viewletService_1.IViewletService)
    ], RefreshExplorerView);
    return RefreshExplorerView;
}(actions_1.Action));
exports.RefreshExplorerView = RefreshExplorerView;
function keybindingForAction(id) {
    switch (id) {
        case GlobalNewUntitledFileAction.ID:
            return new keybinding_1.Keybinding(32768 /* CtrlCmd */ | 44 /* KEY_N */);
        case TriggerRenameFileAction.ID:
            return new keybinding_1.Keybinding(platform_1.isMacintosh ? 3 /* Enter */ : 60 /* F2 */);
        case SaveFileAction.ID:
            return new keybinding_1.Keybinding(32768 /* CtrlCmd */ | 49 /* KEY_S */);
        case MoveFileToTrashAction.ID:
            return new keybinding_1.Keybinding(platform_1.isMacintosh ? 32768 /* CtrlCmd */ | 1 /* Backspace */ : 20 /* Delete */);
        case CopyFileAction.ID:
            return new keybinding_1.Keybinding(32768 /* CtrlCmd */ | 33 /* KEY_C */);
        case PasteFileAction.ID:
            return new keybinding_1.Keybinding(32768 /* CtrlCmd */ | 52 /* KEY_V */);
        case OpenToSideAction.ID:
            if (platform_1.isMacintosh) {
                return new keybinding_1.Keybinding(4096 /* WinCtrl */ | 3 /* Enter */);
            }
            else {
                return new keybinding_1.Keybinding(32768 /* CtrlCmd */ | 3 /* Enter */);
            }
    }
    return null;
}
exports.keybindingForAction = keybindingForAction;
function validateFileName(parent, name, allowOverwriting) {
    if (allowOverwriting === void 0) { allowOverwriting = false; }
    // Produce a well formed file name
    name = getWellFormedFileName(name);
    // Name not provided
    if (!name || name.length === 0 || /^\s+$/.test(name)) {
        return nls.localize('emptyFileNameError', "A file or folder name must be provided.");
    }
    // Do not allow to overwrite existing file
    if (!allowOverwriting) {
        if (parent.children && parent.children.some(function (c) {
            if (platform_1.isLinux) {
                return c.name === name;
            }
            return c.name.toLowerCase() === name.toLowerCase();
        })) {
            return nls.localize('fileNameExistsError', "A file or folder **{0}** already exists at this location. Please choose a different name.", name);
        }
    }
    // Invalid File name
    if (!paths.isValidBasename(name)) {
        return nls.localize('invalidFileNameError', "The name **{0}** is not valid as a file or folder name. Please choose a different name.", name);
    }
    // Max length restriction (on Windows)
    if (platform_1.isWindows) {
        var fullPathLength = name.length + parent.resource.fsPath.length + 1;
        if (fullPathLength > 255) {
            return nls.localize('filePathTooLongError', "The name **{0}** results in a path that is too long. Please choose a shorter name.", name);
        }
    }
    return null;
}
exports.validateFileName = validateFileName;
function getWellFormedFileName(filename) {
    if (!filename) {
        return filename;
    }
    // Trim whitespaces
    filename = strings.trim(strings.trim(filename, ' '), '\t');
    // Remove trailing dots
    filename = strings.rtrim(filename, '.');
    return filename;
}
exports.getWellFormedFileName = getWellFormedFileName;
// Diagnostics support
var diag;
if (!diag) {
    diag = diagnostics.register('FileActionsDiagnostics', function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i - 0] = arguments[_i];
        }
        console.log(args[1] + ' - ' + args[0] + ' (time: ' + args[2].getTime() + ' [' + args[2].toUTCString() + '])');
    });
}
