// vscode/src/vs/workbench/parts/debug/browser/debugEditorInputs.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var winjs_base_1 = require("vs/base/common/winjs.base");
var modesRegistry_1 = require("vs/editor/common/modes/modesRegistry");
var strinput = require("vs/workbench/common/editor/stringEditorInput");
var editor_1 = require("vs/workbench/common/editor");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var DebugStringEditorInput = (function (_super) {
    __extends(DebugStringEditorInput, _super);
    function DebugStringEditorInput(name, resourceUrl, description, value, modeId, singleton, instantiationService) {
        var _this = _super.call(this, name, description, value, modeId || modesRegistry_1.PLAINTEXT_MODE_ID, singleton, instantiationService) || this;
        _this.resourceUrl = resourceUrl;
        return _this;
    }
    DebugStringEditorInput.prototype.getResource = function () {
        return this.resourceUrl;
    };
    return DebugStringEditorInput;
}(strinput.StringEditorInput));
DebugStringEditorInput = __decorate([
    __param(6, instantiation_1.IInstantiationService)
], DebugStringEditorInput);
exports.DebugStringEditorInput = DebugStringEditorInput;
var DebugErrorEditorInput = (function (_super) {
    __extends(DebugErrorEditorInput, _super);
    function DebugErrorEditorInput(name, value) {
        var _this = _super.call(this) || this;
        _this.name = name;
        _this.value = value;
        return _this;
    }
    DebugErrorEditorInput.prototype.getTypeId = function () {
        return DebugErrorEditorInput.ID;
    };
    DebugErrorEditorInput.prototype.resolve = function (refresh) {
        return winjs_base_1.TPromise.as(null);
    };
    DebugErrorEditorInput.prototype.getName = function () {
        return this.name;
    };
    return DebugErrorEditorInput;
}(editor_1.EditorInput));
DebugErrorEditorInput.ID = 'workbench.editors.debugErrorEditorInput';
exports.DebugErrorEditorInput = DebugErrorEditorInput;
