// d3-scale/index.js
export {
  default as scaleBand,
  point as scalePoint
} from "./src/band";

export {
  default as scaleIdentity
} from "./src/identity";

export {
  default as scaleLinear
} from "./src/linear";

export {
  default as scaleLog
} from "./src/log";

export {
  default as scaleOrdinal,
  implicit as scaleImplicit
} from "./src/ordinal";

export {
  default as scalePow,
  sqrt as scaleSqrt
} from "./src/pow";

export {
  default as scaleQuantile
} from "./src/quantile";

export {
  default as scaleQuantize
} from "./src/quantize";

export {
  default as scaleThreshold
} from "./src/threshold";

export {
  default as scaleTime
} from "./src/time";

export {
  default as scaleUtc
} from "./src/utcTime";

export {
  default as scaleCategory10
} from "./src/category10";

export {
  default as scaleCategory20b
} from "./src/category20b";

export {
  default as scaleCategory20c
} from "./src/category20c";

export {
  default as scaleCategory20
} from "./src/category20";

export {
  default as scaleCubehelix
} from "./src/cubehelix";

export {
  default as scaleRainbow,
  warm as scaleWarm,
  cool as scaleCool
} from "./src/rainbow";

export {
  default as scaleViridis,
  magma as scaleMagma,
  inferno as scaleInferno,
  plasma as scalePlasma
} from "./src/viridis";
