// vscode/src/vs/workbench/parts/update/electron-browser/releaseNotesInput.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var nls_1 = require('vs/nls');
var winjs_base_1 = require('vs/base/common/winjs.base');
var editor_1 = require('vs/workbench/common/editor');
var ReleaseNotesInput = (function (_super) {
    __extends(ReleaseNotesInput, _super);
    function ReleaseNotesInput(_version) {
        _super.call(this);
        this._version = _version;
    }
    Object.defineProperty(ReleaseNotesInput, "ID", {
        get: function () { return 'workbench.releaseNotes.input'; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ReleaseNotesInput.prototype, "version", {
        get: function () { return this._version; },
        enumerable: true,
        configurable: true
    });
    ReleaseNotesInput.prototype.getTypeId = function () {
        return ReleaseNotesInput.ID;
    };
    ReleaseNotesInput.prototype.getName = function () {
        return nls_1.localize('releaseNotesInputName', "Release Notes: {0}", this.version);
    };
    ReleaseNotesInput.prototype.matches = function (other) {
        if (!(other instanceof ReleaseNotesInput)) {
            return false;
        }
        var otherInput = other;
        return this.version === otherInput.version;
    };
    ReleaseNotesInput.prototype.resolve = function (refresh) {
        return winjs_base_1.TPromise.as(null);
    };
    ReleaseNotesInput.prototype.supportsSplitEditor = function () {
        return false;
    };
    return ReleaseNotesInput;
}(editor_1.EditorInput));
exports.ReleaseNotesInput = ReleaseNotesInput;
