// vscode/src/vs/platform/telemetry/node/appInsightsAppender.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var appInsights = require("applicationinsights");
var types_1 = require("vs/base/common/types");
var objects_1 = require("vs/base/common/objects");
var winjs_base_1 = require("vs/base/common/winjs.base");
var _initialized = false;
function ensureAIEngineIsInitialized() {
    if (_initialized === false) {
        // we need to pass some fake key, otherwise AI throws an exception
        appInsights.setup('2588e01f-f6c9-4cd6-a348-143741f8d702')
            .setAutoCollectConsole(false)
            .setAutoCollectExceptions(false)
            .setAutoCollectPerformance(false)
            .setAutoCollectRequests(false);
        _initialized = true;
    }
}
function getClient(aiKey) {
    ensureAIEngineIsInitialized();
    var client = appInsights.getClient(aiKey);
    client.channel.setOfflineMode(true);
    client.context.tags[client.context.keys.deviceMachineName] = ''; //prevent App Insights from reporting machine name
    if (aiKey.indexOf('AIF-') === 0) {
        client.config.endpointUrl = 'https://vortex.data.microsoft.com/collect/v1';
    }
    return client;
}
var AppInsightsAppender = (function () {
    function AppInsightsAppender(_eventPrefix, _defaultData, aiKeyOrClientFactory // allow factory function for testing
    ) {
        this._eventPrefix = _eventPrefix;
        this._defaultData = _defaultData;
        if (!this._defaultData) {
            this._defaultData = Object.create(null);
        }
        if (typeof aiKeyOrClientFactory === 'string') {
            this._aiClient = getClient(aiKeyOrClientFactory);
        }
        else if (typeof aiKeyOrClientFactory === 'function') {
            this._aiClient = aiKeyOrClientFactory();
        }
    }
    AppInsightsAppender._getData = function (data) {
        var properties = Object.create(null);
        var measurements = Object.create(null);
        var flat = Object.create(null);
        AppInsightsAppender._flaten(data, flat);
        for (var prop in flat) {
            // enforce property names less than 150 char, take the last 150 char
            prop = prop.length > 150 ? prop.substr(prop.length - 149) : prop;
            var value = flat[prop];
            if (typeof value === 'number') {
                measurements[prop] = value;
            }
            else if (typeof value === 'boolean') {
                measurements[prop] = value ? 1 : 0;
            }
            else if (typeof value === 'string') {
                //enforce property value to be less than 1024 char, take the first 1024 char
                properties[prop] = value.substring(0, 1023);
            }
            else if (typeof value !== 'undefined' && value !== null) {
                properties[prop] = value;
            }
        }
        return {
            properties: properties,
            measurements: measurements
        };
    };
    AppInsightsAppender._flaten = function (obj, result, order, prefix) {
        if (order === void 0) { order = 0; }
        if (!obj) {
            return;
        }
        for (var _i = 0, _a = Object.getOwnPropertyNames(obj); _i < _a.length; _i++) {
            var item = _a[_i];
            var value = obj[item];
            var index = prefix ? prefix + item : item;
            if (Array.isArray(value)) {
                result[index] = objects_1.safeStringify(value);
            }
            else if (value instanceof Date) {
                // TODO unsure why this is here and not in _getData
                result[index] = value.toISOString();
            }
            else if (types_1.isObject(value)) {
                if (order < 2) {
                    AppInsightsAppender._flaten(value, result, order + 1, index + '.');
                }
                else {
                    result[index] = objects_1.safeStringify(value);
                }
            }
            else {
                result[index] = value;
            }
        }
    };
    AppInsightsAppender.prototype.log = function (eventName, data) {
        if (!this._aiClient) {
            return;
        }
        data = objects_1.mixin(data, this._defaultData);
        var _a = AppInsightsAppender._getData(data), properties = _a.properties, measurements = _a.measurements;
        this._aiClient.trackEvent(this._eventPrefix + '/' + eventName, properties, measurements);
    };
    AppInsightsAppender.prototype.dispose = function () {
        var _this = this;
        if (this._aiClient) {
            return new winjs_base_1.TPromise(function (resolve) {
                _this._aiClient.sendPendingData(function () {
                    // all data flushed
                    _this._aiClient = undefined;
                    resolve(void 0);
                });
            });
        }
    };
    return AppInsightsAppender;
}());
exports.AppInsightsAppender = AppInsightsAppender;
