// flow-status-webpack-plugin/index.js
const colors = require('colors');
const shell = require('shelljs');

function FlowStatusWebpackPlugin(options) {
    this.options = options || {};
}

FlowStatusWebpackPlugin.prototype.apply = function(compiler) {
    const options = this.options;
    const flowArgs = options.flowArgs || '';

    function startFlow(cb) {
        if (options.restartFlow === false) {
            shell.exec('./node_modules/.bin/flow start ' + flowArgs, () => cb());
        }
        else {
            shell.exec('./node_modules/.bin/flow stop', () => {
                shell.exec('./node_modules/.bin/flow start ' + flowArgs, () => cb());
            });
        }
    }

    var firstRun = true;
    function startFlowIfFirstRun(compiler, cb) {
        if (firstRun) {
            firstRun = false;
            startFlow(cb);
        }
        else {
            cb();
        }
    }

    // restart flow if interfacesPath was provided regardless of whether webpack is in normal or watch mode
    compiler.plugin('run', startFlowIfFirstRun);
    compiler.plugin('watch-run', startFlowIfFirstRun);

    var waitingForFlow = false;

    function flowStatus() {
        if (!waitingForFlow) {
            waitingForFlow = true;

            // this will start a flow server if it was not running
            shell.exec('./node_modules/.bin/flow status --color always', {silent: true}, (code, stdout, stderr) => {
                const hasErrors = code !== 0;

                if (hasErrors) {
                    console.log('\n----------------'.red);
                    console.log('Flow has errors!');
                    console.log('----------------\n'.red);
                } else {
                    console.log('\n-----------------------------'.green);
                    console.log('Everything is fine with Flow!');
                    console.log('-----------------------------\n'.green);
                }
                console.log(stdout);
                console.error(stderr);

                waitingForFlow = false;
            });
        }
    }

    // When Webpack compilation is done, we should run Flow Status.
    compiler.plugin('done', flowStatus);
};

module.exports = FlowStatusWebpackPlugin;
