// vscode/src/vs/workbench/parts/debug/browser/debugViewlet.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
require("./media/debugViewlet.css");
var nls = require("vs/nls");
var builder_1 = require("vs/base/browser/builder");
var winjs_base_1 = require("vs/base/common/winjs.base");
var lifecycle = require("vs/base/common/lifecycle");
var splitview_1 = require("vs/base/browser/ui/splitview/splitview");
var memento_1 = require("vs/workbench/common/memento");
var viewlet_1 = require("vs/workbench/browser/viewlet");
var debug_1 = require("vs/workbench/parts/debug/common/debug");
var debugViewRegistry_1 = require("vs/workbench/parts/debug/browser/debugViewRegistry");
var debugActions_1 = require("vs/workbench/parts/debug/browser/debugActions");
var debugActionItems_1 = require("vs/workbench/parts/debug/browser/debugActionItems");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var progress_1 = require("vs/platform/progress/common/progress");
var workspace_1 = require("vs/platform/workspace/common/workspace");
var telemetry_1 = require("vs/platform/telemetry/common/telemetry");
var storage_1 = require("vs/platform/storage/common/storage");
var DebugViewlet = (function (_super) {
    __extends(DebugViewlet, _super);
    function DebugViewlet(telemetryService, progressService, debugService, instantiationService, contextService, storageService) {
        var _this = _super.call(this, debug_1.VIEWLET_ID, telemetryService) || this;
        _this.progressService = progressService;
        _this.debugService = debugService;
        _this.instantiationService = instantiationService;
        _this.contextService = contextService;
        _this.progressRunner = null;
        _this.viewletSettings = _this.getMemento(storageService, memento_1.Scope.WORKSPACE);
        _this.toDispose = [];
        _this.views = [];
        _this.toDispose.push(_this.debugService.onDidChangeState(function () {
            _this.onDebugServiceStateChange();
        }));
        return _this;
    }
    // viewlet
    DebugViewlet.prototype.create = function (parent) {
        var _this = this;
        _super.prototype.create.call(this, parent);
        this.$el = parent.div().addClass('debug-viewlet');
        if (this.contextService.getWorkspace()) {
            var actionRunner_1 = this.getActionRunner();
            this.views = debugViewRegistry_1.DebugViewRegistry.getDebugViews().map(function (viewConstructor) { return _this.instantiationService.createInstance(viewConstructor, actionRunner_1, _this.viewletSettings); });
            this.splitView = new splitview_1.SplitView(this.$el.getHTMLElement());
            this.toDispose.push(this.splitView);
            this.views.forEach(function (v) { return _this.splitView.addView(v); });
            // Track focus
            this.toDispose.push(this.splitView.onFocus(function (view) {
                _this.lastFocusedView = view;
            }));
        }
        else {
            this.$el.append(builder_1.$([
                '<div class="noworkspace-view">',
                '<p>', nls.localize('noWorkspace', "There is no currently opened folder."), '</p>',
                '<p>', nls.localize('pleaseRestartToDebug', "Open a folder in order to start debugging."), '</p>',
                '</div>'
            ].join('')));
        }
        return winjs_base_1.TPromise.as(null);
    };
    DebugViewlet.prototype.setVisible = function (visible) {
        var _this = this;
        return _super.prototype.setVisible.call(this, visible).then(function () {
            return winjs_base_1.TPromise.join(_this.views.map(function (view) { return view.setVisible(visible); }));
        });
    };
    DebugViewlet.prototype.layout = function (dimension) {
        if (this.splitView) {
            this.splitView.layout(dimension.height);
        }
    };
    DebugViewlet.prototype.focus = function () {
        _super.prototype.focus.call(this);
        if (this.lastFocusedView && this.lastFocusedView.isExpanded()) {
            this.lastFocusedView.focusBody();
            return;
        }
        if (this.views.length > 0) {
            this.views[0].focusBody();
        }
    };
    DebugViewlet.prototype.getActions = function () {
        var _this = this;
        if (this.debugService.state === debug_1.State.Disabled) {
            return [];
        }
        if (!this.actions) {
            this.actions = [
                this.instantiationService.createInstance(debugActions_1.StartAction, debugActions_1.StartAction.ID, debugActions_1.StartAction.LABEL),
                this.instantiationService.createInstance(debugActions_1.ConfigureAction, debugActions_1.ConfigureAction.ID, debugActions_1.ConfigureAction.LABEL),
                this.instantiationService.createInstance(debugActions_1.ToggleReplAction, debugActions_1.ToggleReplAction.ID, debugActions_1.ToggleReplAction.LABEL)
            ];
            this.actions.forEach(function (a) {
                _this.toDispose.push(a);
            });
        }
        return this.actions;
    };
    DebugViewlet.prototype.getActionItem = function (action) {
        if (action.id === debugActions_1.StartAction.ID) {
            return this.instantiationService.createInstance(debugActionItems_1.StartDebugActionItem, null, action);
        }
        return null;
    };
    DebugViewlet.prototype.onDebugServiceStateChange = function () {
        if (this.progressRunner) {
            this.progressRunner.done();
        }
        if (this.debugService.state === debug_1.State.Initializing) {
            this.progressRunner = this.progressService.show(true);
        }
        else {
            this.progressRunner = null;
        }
    };
    DebugViewlet.prototype.dispose = function () {
        this.toDispose = lifecycle.dispose(this.toDispose);
        _super.prototype.dispose.call(this);
    };
    DebugViewlet.prototype.shutdown = function () {
        this.views.forEach(function (v) { return v.shutdown(); });
        _super.prototype.shutdown.call(this);
    };
    return DebugViewlet;
}(viewlet_1.Viewlet));
DebugViewlet = __decorate([
    __param(0, telemetry_1.ITelemetryService),
    __param(1, progress_1.IProgressService),
    __param(2, debug_1.IDebugService),
    __param(3, instantiation_1.IInstantiationService),
    __param(4, workspace_1.IWorkspaceContextService),
    __param(5, storage_1.IStorageService)
], DebugViewlet);
exports.DebugViewlet = DebugViewlet;
