// vscode/src/vs/workbench/parts/extensions/browser/dependenciesViewer.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var dom = require("vs/base/browser/dom");
var nls_1 = require("vs/nls");
var lifecycle_1 = require("vs/base/common/lifecycle");
var winjs_base_1 = require("vs/base/common/winjs.base");
var treeDefaults_1 = require("vs/base/parts/tree/browser/treeDefaults");
var actions_1 = require("vs/base/common/actions");
var extensions_1 = require("vs/workbench/parts/extensions/common/extensions");
var event_1 = require("vs/base/common/event");
var event_2 = require("vs/base/browser/event");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var DataSource = (function () {
    function DataSource() {
    }
    DataSource.prototype.getId = function (tree, element) {
        var _this = this;
        var id = element.identifier;
        this.getParent(tree, element).then(function (parent) {
            id = parent ? _this.getId(tree, parent) + '/' + id : id;
        });
        return id;
    };
    DataSource.prototype.hasChildren = function (tree, element) {
        return element.hasDependencies;
    };
    DataSource.prototype.getChildren = function (tree, element) {
        return winjs_base_1.TPromise.as(element.dependencies);
    };
    DataSource.prototype.getParent = function (tree, element) {
        return winjs_base_1.TPromise.as(element.dependent);
    };
    return DataSource;
}());
exports.DataSource = DataSource;
var Renderer = (function () {
    function Renderer(instantiationService) {
        this.instantiationService = instantiationService;
    }
    Renderer.prototype.getHeight = function (tree, element) {
        return 62;
    };
    Renderer.prototype.getTemplateId = function (tree, element) {
        return element.extension ? Renderer.EXTENSION_TEMPLATE_ID : Renderer.UNKNOWN_EXTENSION_TEMPLATE_ID;
    };
    Renderer.prototype.renderTemplate = function (tree, templateId, container) {
        if (Renderer.EXTENSION_TEMPLATE_ID === templateId) {
            return this.renderExtensionTemplate(tree, container);
        }
        return this.renderUnknownExtensionTemplate(tree, container);
    };
    Renderer.prototype.renderExtensionTemplate = function (tree, container) {
        dom.addClass(container, 'dependency');
        var icon = dom.append(container, dom.$('img.icon'));
        var details = dom.append(container, dom.$('.details'));
        var header = dom.append(details, dom.$('.header'));
        var name = dom.append(header, dom.$('span.name'));
        var openExtensionAction = this.instantiationService.createInstance(OpenExtensionAction);
        var extensionDisposables = [dom.addDisposableListener(name, 'click', function (e) {
                tree.setFocus(openExtensionAction.extensionDependencies);
                tree.setSelection([openExtensionAction.extensionDependencies]);
                openExtensionAction.run(e.ctrlKey || e.metaKey);
                e.stopPropagation();
                e.preventDefault();
            })];
        var identifier = dom.append(header, dom.$('span.identifier'));
        var footer = dom.append(details, dom.$('.footer'));
        var author = dom.append(footer, dom.$('.author'));
        return {
            icon: icon,
            name: name,
            identifier: identifier,
            author: author,
            extensionDisposables: extensionDisposables,
            set extensionDependencies(e) {
                openExtensionAction.extensionDependencies = e;
            }
        };
    };
    Renderer.prototype.renderUnknownExtensionTemplate = function (tree, container) {
        var messageContainer = dom.append(container, dom.$('div.unknown-dependency'));
        dom.append(messageContainer, dom.$('span.error-marker')).textContent = nls_1.localize('error', "Error");
        dom.append(messageContainer, dom.$('span.message')).textContent = nls_1.localize('Unknown Dependency', "Unknown Dependency:");
        var identifier = dom.append(messageContainer, dom.$('span.message'));
        return { identifier: identifier };
    };
    Renderer.prototype.renderElement = function (tree, element, templateId, templateData) {
        if (templateId === Renderer.EXTENSION_TEMPLATE_ID) {
            this.renderExtension(tree, element, templateData);
            return;
        }
        this.renderUnknownExtension(tree, element, templateData);
    };
    Renderer.prototype.renderExtension = function (tree, element, data) {
        var extension = element.extension;
        var onError = event_1.once(event_2.domEvent(data.icon, 'error'));
        onError(function () { return data.icon.src = extension.iconUrlFallback; }, null, data.extensionDisposables);
        data.icon.src = extension.iconUrl;
        if (!data.icon.complete) {
            data.icon.style.visibility = 'hidden';
            data.icon.onload = function () { return data.icon.style.visibility = 'inherit'; };
        }
        else {
            data.icon.style.visibility = 'inherit';
        }
        data.name.textContent = extension.displayName;
        data.identifier.textContent = extension.identifier;
        data.author.textContent = extension.publisherDisplayName;
        data.extensionDependencies = element;
    };
    Renderer.prototype.renderUnknownExtension = function (tree, element, data) {
        data.identifier.textContent = element.identifier;
    };
    Renderer.prototype.disposeTemplate = function (tree, templateId, templateData) {
        if (templateId === Renderer.EXTENSION_TEMPLATE_ID) {
            templateData.extensionDisposables = lifecycle_1.dispose(templateData.extensionDisposables);
        }
    };
    return Renderer;
}());
Renderer.EXTENSION_TEMPLATE_ID = 'extension-template';
Renderer.UNKNOWN_EXTENSION_TEMPLATE_ID = 'unknown-extension-template';
Renderer = __decorate([
    __param(0, instantiation_1.IInstantiationService)
], Renderer);
exports.Renderer = Renderer;
var Controller = (function (_super) {
    __extends(Controller, _super);
    function Controller(extensionsWorkdbenchService) {
        var _this = _super.call(this) || this;
        _this.extensionsWorkdbenchService = extensionsWorkdbenchService;
        _this.downKeyBindingDispatcher.set(32768 /* CtrlCmd */ | 3 /* Enter */, function (tree, event) { return _this.openExtension(tree, true); });
        return _this;
    }
    Controller.prototype.onLeftClick = function (tree, element, event) {
        var currentFoucssed = tree.getFocus();
        if (_super.prototype.onLeftClick.call(this, tree, element, event)) {
            if (element.dependent === null) {
                if (currentFoucssed) {
                    tree.setFocus(currentFoucssed);
                }
                else {
                    tree.focusFirst();
                }
                return true;
            }
        }
        return false;
    };
    Controller.prototype.onEnter = function (tree, event) {
        if (_super.prototype.onEnter.call(this, tree, event)) {
            return this.openExtension(tree, false);
        }
        return false;
    };
    Controller.prototype.openExtension = function (tree, sideByside) {
        var element = tree.getFocus();
        if (element.extension) {
            this.extensionsWorkdbenchService.open(element.extension, sideByside);
            return true;
        }
        return false;
    };
    return Controller;
}(treeDefaults_1.DefaultController));
Controller = __decorate([
    __param(0, extensions_1.IExtensionsWorkbenchService)
], Controller);
exports.Controller = Controller;
var OpenExtensionAction = (function (_super) {
    __extends(OpenExtensionAction, _super);
    function OpenExtensionAction(extensionsWorkdbenchService) {
        var _this = _super.call(this, 'extensions.action.openDependency', '') || this;
        _this.extensionsWorkdbenchService = extensionsWorkdbenchService;
        return _this;
    }
    Object.defineProperty(OpenExtensionAction.prototype, "extensionDependencies", {
        get: function () {
            return this._extensionDependencies;
        },
        set: function (extensionDependencies) {
            this._extensionDependencies = extensionDependencies;
        },
        enumerable: true,
        configurable: true
    });
    OpenExtensionAction.prototype.run = function (sideByside) {
        return this.extensionsWorkdbenchService.open(this._extensionDependencies.extension, sideByside);
    };
    return OpenExtensionAction;
}(actions_1.Action));
OpenExtensionAction = __decorate([
    __param(0, extensions_1.IExtensionsWorkbenchService)
], OpenExtensionAction);
