// vscode/src/vs/workbench/common/editor/rangeDecorations.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var editorCommon = require("vs/editor/common/editorCommon");
var editorService_1 = require("vs/workbench/services/editor/common/editorService");
var editor_1 = require("vs/workbench/common/editor");
var RangeHighlightDecorations = (function () {
    function RangeHighlightDecorations(editorService) {
        this.editorService = editorService;
        this.rangeHighlightDecorationId = null;
        this.editor = null;
        this.editorDisposables = [];
    }
    RangeHighlightDecorations.prototype.removeHighlightRange = function () {
        if (this.editor && this.rangeHighlightDecorationId) {
            this.doRemoveRangeHighlight(this.editor, this.rangeHighlightDecorationId);
        }
        this.rangeHighlightDecorationId = null;
    };
    RangeHighlightDecorations.prototype.highlightRange = function (range, editor) {
        editor = editor ? editor : this.getEditor(range);
        if (editor) {
            this.doHighlightRange(editor.getControl(), range);
        }
    };
    RangeHighlightDecorations.prototype.doHighlightRange = function (editor, selectionRange) {
        var _this = this;
        this.removeHighlightRange();
        editor.changeDecorations(function (changeAccessor) {
            _this.rangeHighlightDecorationId = changeAccessor.addDecoration(selectionRange.range, _this.createRangeHighlightDecoration());
        });
        this.setEditor(editor);
    };
    RangeHighlightDecorations.prototype.getEditor = function (resourceRange) {
        var editorInput = editor_1.asFileEditorInput(this.editorService.getActiveEditorInput());
        if (editorInput) {
            if (editorInput.getResource().fsPath === resourceRange.resource.fsPath) {
                return this.editorService.getActiveEditor();
            }
        }
        return null;
    };
    RangeHighlightDecorations.prototype.setEditor = function (editor) {
        var _this = this;
        if (this.editor !== editor) {
            this.disposeEditorListeners();
            this.editor = editor;
            this.editorDisposables.push(this.editor.onDidChangeCursorPosition(function (e) {
                if (e.reason === editorCommon.CursorChangeReason.Explicit
                    || e.reason === editorCommon.CursorChangeReason.Undo
                    || e.reason === editorCommon.CursorChangeReason.Redo) {
                    _this.doRemoveRangeHighlight(_this.editor, _this.rangeHighlightDecorationId);
                }
            }));
            this.editorDisposables.push(this.editor.onDidChangeModel(function () { _this.doRemoveRangeHighlight(_this.editor, _this.rangeHighlightDecorationId); }));
            this.editorDisposables.push(this.editor.onDidDispose(function () {
                _this.doRemoveRangeHighlight(_this.editor, _this.rangeHighlightDecorationId);
                _this.editor = null;
            }));
        }
    };
    RangeHighlightDecorations.prototype.disposeEditorListeners = function () {
        this.editorDisposables.forEach(function (disposable) { return disposable.dispose(); });
        this.editorDisposables = [];
    };
    RangeHighlightDecorations.prototype.doRemoveRangeHighlight = function (model, rangeHighlightDecorationId) {
        model.deltaDecorations([rangeHighlightDecorationId], []);
    };
    RangeHighlightDecorations.prototype.createRangeHighlightDecoration = function () {
        return {
            stickiness: editorCommon.TrackedRangeStickiness.NeverGrowsWhenTypingAtEdges,
            className: 'rangeHighlight',
            isWholeLine: true
        };
    };
    RangeHighlightDecorations.prototype.dispose = function () {
        this.removeHighlightRange();
        this.disposeEditorListeners();
        this.editor = null;
    };
    return RangeHighlightDecorations;
}());
RangeHighlightDecorations = __decorate([
    __param(0, editorService_1.IWorkbenchEditorService)
], RangeHighlightDecorations);
exports.RangeHighlightDecorations = RangeHighlightDecorations;
