// vscode/src/vs/workbench/parts/files/browser/views/emptyView.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var nls = require("vs/nls");
var errors = require("vs/base/common/errors");
var env = require("vs/base/common/platform");
var DOM = require("vs/base/browser/dom");
var winjs_base_1 = require("vs/base/common/winjs.base");
var button_1 = require("vs/base/browser/ui/button/button");
var builder_1 = require("vs/base/browser/builder");
var splitview_1 = require("vs/base/browser/ui/splitview/splitview");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var fileActions_1 = require("vs/workbench/browser/actions/fileActions");
var EmptyView = (function (_super) {
    __extends(EmptyView, _super);
    function EmptyView(actionRunner, instantiationService) {
        var _this = _super.call(this, {
            minimumSize: 2 * 22,
            ariaHeaderLabel: nls.localize('explorerSection', "Files Explorer Section")
        }) || this;
        _this.actionRunner = actionRunner;
        _this.instantiationService = instantiationService;
        return _this;
    }
    EmptyView.prototype.renderHeader = function (container) {
        var titleDiv = builder_1.$('div.title').appendTo(container);
        builder_1.$('span').text(nls.localize('noWorkspace', "No Folder Opened")).appendTo(titleDiv);
    };
    EmptyView.prototype.renderBody = function (container) {
        var _this = this;
        DOM.addClass(container, 'explorer-empty-view');
        var titleDiv = builder_1.$('div.section').appendTo(container);
        builder_1.$('p').text(nls.localize('noWorkspaceHelp', "You have not yet opened a folder.")).appendTo(titleDiv);
        var section = builder_1.$('div.section').appendTo(container);
        this.openFolderButton = new button_1.Button(section);
        this.openFolderButton.label = nls.localize('openFolder', "Open Folder");
        this.openFolderButton.addListener2('click', function () {
            var actionClass = env.isMacintosh ? fileActions_1.OpenFileFolderAction : fileActions_1.OpenFolderAction;
            var action = _this.instantiationService.createInstance(actionClass, actionClass.ID, actionClass.LABEL);
            _this.actionRunner.run(action).done(function () {
                action.dispose();
            }, function (err) {
                action.dispose();
                errors.onUnexpectedError(err);
            });
        });
    };
    EmptyView.prototype.layoutBody = function (size) {
        // no-op
    };
    EmptyView.prototype.create = function () {
        return winjs_base_1.TPromise.as(null);
    };
    EmptyView.prototype.setVisible = function (visible) {
        return winjs_base_1.TPromise.as(null);
    };
    EmptyView.prototype.focusBody = function () {
        if (this.openFolderButton) {
            this.openFolderButton.getElement().focus();
        }
    };
    EmptyView.prototype.reveal = function (element, relativeTop) {
        return winjs_base_1.TPromise.as(null);
    };
    EmptyView.prototype.getActions = function () {
        return [];
    };
    EmptyView.prototype.getSecondaryActions = function () {
        return [];
    };
    EmptyView.prototype.getActionItem = function (action) {
        return null;
    };
    EmptyView.prototype.shutdown = function () {
        // Subclass to implement
    };
    return EmptyView;
}(splitview_1.CollapsibleView));
EmptyView = __decorate([
    __param(1, instantiation_1.IInstantiationService)
], EmptyView);
exports.EmptyView = EmptyView;
