// vscode/src/vs/workbench/parts/files/browser/views/emptyView.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var nls = require("vs/nls");
var env = require("vs/base/common/platform");
var DOM = require("vs/base/browser/dom");
var winjs_base_1 = require("vs/base/common/winjs.base");
var button_1 = require("vs/base/browser/ui/button/button");
var builder_1 = require("vs/base/browser/builder");
var splitview_1 = require("vs/base/browser/ui/splitview/splitview");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var platform_1 = require("vs/platform/platform");
var telemetry_1 = require("vs/platform/telemetry/common/telemetry");
var actionRegistry_1 = require("vs/workbench/common/actionRegistry");
var EmptyView = (function (_super) {
    __extends(EmptyView, _super);
    function EmptyView(telemetryService, instantiationService) {
        var _this = _super.call(this, {
            minimumSize: 2 * 22,
            ariaHeaderLabel: nls.localize('explorerSection', "Files Explorer Section")
        }) || this;
        _this.telemetryService = telemetryService;
        _this.instantiationService = instantiationService;
        return _this;
    }
    EmptyView.prototype.renderHeader = function (container) {
        var titleDiv = builder_1.$('div.title').appendTo(container);
        builder_1.$('span').text(nls.localize('noWorkspace', "No Folder Opened")).appendTo(titleDiv);
    };
    EmptyView.prototype.renderBody = function (container) {
        var _this = this;
        DOM.addClass(container, 'explorer-empty-view');
        var titleDiv = builder_1.$('div.section').appendTo(container);
        builder_1.$('p').text(nls.localize('noWorkspaceHelp', "You have not yet opened a folder.")).appendTo(titleDiv);
        var section = builder_1.$('div.section').appendTo(container);
        this.openFolderButton = new button_1.Button(section);
        this.openFolderButton.label = nls.localize('openFolder', "Open Folder");
        this.openFolderButton.addListener2('click', function () {
            _this.runWorkbenchAction(env.isMacintosh ? 'workbench.action.files.openFileFolder' : 'workbench.action.files.openFolder');
        });
    };
    EmptyView.prototype.layoutBody = function (size) {
        // no-op
    };
    EmptyView.prototype.runWorkbenchAction = function (actionId) {
        this.telemetryService.publicLog('workbenchActionExecuted', { id: actionId, from: 'explorer' });
        var actionRegistry = platform_1.Registry.as(actionRegistry_1.Extensions.WorkbenchActions);
        var actionDescriptor = actionRegistry.getWorkbenchAction(actionId);
        var action = this.instantiationService.createInstance(actionDescriptor.syncDescriptor);
        return action.run().done(function () { return action.dispose(); });
    };
    EmptyView.prototype.create = function () {
        return winjs_base_1.TPromise.as(null);
    };
    EmptyView.prototype.setVisible = function (visible) {
        return winjs_base_1.TPromise.as(null);
    };
    EmptyView.prototype.focusBody = function () {
        if (this.openFolderButton) {
            this.openFolderButton.getElement().focus();
        }
    };
    EmptyView.prototype.reveal = function (element, relativeTop) {
        return winjs_base_1.TPromise.as(null);
    };
    EmptyView.prototype.getActions = function () {
        return [];
    };
    EmptyView.prototype.getSecondaryActions = function () {
        return [];
    };
    EmptyView.prototype.getActionItem = function (action) {
        return null;
    };
    EmptyView.prototype.shutdown = function () {
        // Subclass to implement
    };
    return EmptyView;
}(splitview_1.CollapsibleView));
EmptyView = __decorate([
    __param(0, telemetry_1.ITelemetryService),
    __param(1, instantiation_1.IInstantiationService)
], EmptyView);
exports.EmptyView = EmptyView;
