// vscode/src/vs/editor/contrib/toggleTabFocusMode/common/toggleTabFocusMode.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var nls = require("vs/nls");
var editorCommonExtensions_1 = require("vs/editor/common/editorCommonExtensions");
var commonEditorConfig_1 = require("vs/editor/common/config/commonEditorConfig");
var ToggleTabFocusModeAction = ToggleTabFocusModeAction_1 = (function (_super) {
    __extends(ToggleTabFocusModeAction, _super);
    function ToggleTabFocusModeAction() {
        return _super.call(this, {
            id: ToggleTabFocusModeAction_1.ID,
            label: nls.localize('toggle.tabfocusmode', "Toggle Use of Tab Key for Setting Focus"),
            alias: 'Toggle Use of Tab Key for Setting Focus',
            precondition: null,
            kbOpts: {
                kbExpr: null,
                primary: 32768 /* CtrlCmd */ | 43 /* KEY_M */,
                mac: { primary: 4096 /* WinCtrl */ | 16384 /* Shift */ | 43 /* KEY_M */ }
            }
        }) || this;
    }
    ToggleTabFocusModeAction.prototype.run = function (accessor, editor) {
        var oldValue = commonEditorConfig_1.TabFocus.getTabFocusMode();
        commonEditorConfig_1.TabFocus.setTabFocusMode(!oldValue);
    };
    return ToggleTabFocusModeAction;
}(editorCommonExtensions_1.EditorAction));
ToggleTabFocusModeAction.ID = 'editor.action.toggleTabFocusMode';
ToggleTabFocusModeAction = ToggleTabFocusModeAction_1 = __decorate([
    editorCommonExtensions_1.editorAction
], ToggleTabFocusModeAction);
exports.ToggleTabFocusModeAction = ToggleTabFocusModeAction;
var ToggleTabFocusModeAction_1;
