// vscode/src/vs/editor/contrib/quickOpen/browser/quickCommand.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var nls = require('vs/nls');
var errors_1 = require('vs/base/common/errors');
var filters_1 = require('vs/base/common/filters');
var winjs_base_1 = require('vs/base/common/winjs.base');
var quickOpenModel_1 = require('vs/base/parts/quickopen/browser/quickOpenModel');
var quickOpen_1 = require('vs/base/parts/quickopen/common/quickOpen');
var keybinding_1 = require('vs/platform/keybinding/common/keybinding');
var editorCommon_1 = require('vs/editor/common/editorCommon');
var editorQuickOpen_1 = require('./editorQuickOpen');
var editorCommonExtensions_1 = require('vs/editor/common/editorCommonExtensions');
var browser = require('vs/base/browser/browser');
var EditorActionCommandEntry = (function (_super) {
    __extends(EditorActionCommandEntry, _super);
    function EditorActionCommandEntry(key, highlights, action, editor) {
        _super.call(this);
        this.key = key;
        this.setHighlights(highlights);
        this.action = action;
        this.editor = editor;
    }
    EditorActionCommandEntry.prototype.getLabel = function () {
        return this.action.label;
    };
    EditorActionCommandEntry.prototype.getAriaLabel = function () {
        return nls.localize('ariaLabelEntry', "{0}, commands", this.getLabel());
    };
    EditorActionCommandEntry.prototype.getGroupLabel = function () {
        return this.key;
    };
    EditorActionCommandEntry.prototype.run = function (mode, context) {
        var _this = this;
        if (mode === quickOpen_1.Mode.OPEN) {
            // Use a timeout to give the quick open widget a chance to close itself first
            winjs_base_1.TPromise.timeout(50).done(function () {
                // Some actions are enabled only when editor has focus
                _this.editor.focus();
                try {
                    var promise = _this.action.run() || winjs_base_1.TPromise.as(null);
                    promise.done(null, errors_1.onUnexpectedError);
                }
                catch (error) {
                    errors_1.onUnexpectedError(error);
                }
            }, errors_1.onUnexpectedError);
            return true;
        }
        return false;
    };
    return EditorActionCommandEntry;
}(quickOpenModel_1.QuickOpenEntryGroup));
exports.EditorActionCommandEntry = EditorActionCommandEntry;
var QuickCommandAction = (function (_super) {
    __extends(QuickCommandAction, _super);
    function QuickCommandAction() {
        _super.call(this, nls.localize('quickCommandActionInput', "Type the name of an action you want to execute"), {
            id: 'editor.action.quickCommand',
            label: nls.localize('QuickCommandAction.label', "Command Palette"),
            alias: 'Command Palette',
            precondition: null,
            kbOpts: {
                kbExpr: editorCommon_1.EditorContextKeys.Focus,
                primary: (browser.isIE11orEarlier ? 8192 /* Alt */ | 59 /* F1 */ : 59 /* F1 */)
            },
            menuOpts: {}
        });
    }
    QuickCommandAction.prototype.run = function (accessor, editor) {
        var _this = this;
        var keybindingService = accessor.get(keybinding_1.IKeybindingService);
        this._show(this.getController(editor), {
            getModel: function (value) {
                return new quickOpenModel_1.QuickOpenModel(_this._editorActionsToEntries(keybindingService, editor, value));
            },
            getAutoFocus: function (searchValue) {
                return {
                    autoFocusFirstEntry: true,
                    autoFocusPrefixMatch: searchValue
                };
            }
        });
    };
    QuickCommandAction.prototype._sort = function (elementA, elementB) {
        var elementAName = elementA.getLabel().toLowerCase();
        var elementBName = elementB.getLabel().toLowerCase();
        return elementAName.localeCompare(elementBName);
    };
    QuickCommandAction.prototype._editorActionsToEntries = function (keybindingService, editor, searchValue) {
        var actions = editor.getSupportedActions();
        var entries = [];
        for (var i = 0; i < actions.length; i++) {
            var action = actions[i];
            var keys = keybindingService.lookupKeybindings(action.id).map(function (k) { return keybindingService.getLabelFor(k); });
            if (action.label) {
                var highlights = filters_1.matchesFuzzy(searchValue, action.label);
                if (highlights) {
                    entries.push(new EditorActionCommandEntry(keys.length > 0 ? keys.join(', ') : '', highlights, action, editor));
                }
            }
        }
        // Sort by name
        entries = entries.sort(this._sort);
        return entries;
    };
    QuickCommandAction = __decorate([
        editorCommonExtensions_1.editorAction
    ], QuickCommandAction);
    return QuickCommandAction;
}(editorQuickOpen_1.BaseEditorQuickOpenAction));
exports.QuickCommandAction = QuickCommandAction;
