diff --git a/cmd/frontend/internal/app/otlpadapter/BUILD.bazel b/cmd/frontend/internal/app/otlpadapter/BUILD.bazel
index 9c05e6d96c..6700175c0d 100644
--- a/cmd/frontend/internal/app/otlpadapter/BUILD.bazel
+++ b/cmd/frontend/internal/app/otlpadapter/BUILD.bazel
@@ -16,10 +16,10 @@ go_library(
         "@com_github_gorilla_mux//:mux",
         "@com_github_sourcegraph_log//:log",
         "@com_github_sourcegraph_log//std",
-        "@io_opentelemetry_go_collector//config/confighttp",
-        "@io_opentelemetry_go_collector//config/configtelemetry",
-        "@io_opentelemetry_go_collector//config/configtls",
         "@io_opentelemetry_go_collector_component//:component",
+        "@io_opentelemetry_go_collector_config_confighttp//:confighttp",
+        "@io_opentelemetry_go_collector_config_configtelemetry//:configtelemetry",
+        "@io_opentelemetry_go_collector_config_configtls//:configtls",
         "@io_opentelemetry_go_collector_exporter//:exporter",
         "@io_opentelemetry_go_collector_exporter_otlpexporter//:otlpexporter",
         "@io_opentelemetry_go_collector_exporter_otlphttpexporter//:otlphttpexporter",
