diff --git a/BUILD.bazel b/BUILD.bazel
index b0957560b9..9bc63273e8 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -325,7 +325,7 @@ exports_files([
     "go.mod",
     # Used for when copy_to_directory might reference an empty filegroup
     # under certain conditions. See //ui/assets/...
-    "CHANGELOG.md",
+    "CONTRIBUTING.md",
 ])

 # stamp_tags(
diff --git a/ui/assets/enterprise/BUILD.bazel b/ui/assets/enterprise/BUILD.bazel
index 180a828683..c3a34519ad 100644
--- a/ui/assets/enterprise/BUILD.bazel
+++ b/ui/assets/enterprise/BUILD.bazel
@@ -16,9 +16,7 @@ go_library(
 copy_to_directory(
     name = "copy_bundle",
     srcs = [
-        "//:CHANGELOG.md",
-        "//client/browser:integration-assets",
-        "//client/web:bundle-enterprise",
+        "//:CONTRIBUTING.md",
         "//ui/assets/img",
     ],
     out = "dist",
diff --git a/ui/assets/oss/BUILD.bazel b/ui/assets/oss/BUILD.bazel
index 8fa2db0f75..82b18e25a8 100644
--- a/ui/assets/oss/BUILD.bazel
+++ b/ui/assets/oss/BUILD.bazel
@@ -16,8 +16,7 @@ go_library(
 copy_to_directory(
     name = "copy_bundle",
     srcs = [
-        "//:CHANGELOG.md",
-        "//client/web:bundle",
+        "//:CONTRIBUTING.md",
         "//ui/assets/img",
     ],
     out = "dist",
diff --git a/ui/assets/enterprise/assets.go b/ui/assets/enterprise/assets.go
index 0936b92a8c..c395c54bf2 100644
--- a/ui/assets/enterprise/assets.go
+++ b/ui/assets/enterprise/assets.go
@@ -12,7 +12,6 @@ import (
 	"github.com/sourcegraph/sourcegraph/ui/assets"
 )

-//go:embed *
 var assetsFS embed.FS
 var afs fs.FS = assetsFS

diff --git a/ui/assets/oss/assets.go b/ui/assets/oss/assets.go
index ea4012cee1..3444294b9a 100644
--- a/ui/assets/oss/assets.go
+++ b/ui/assets/oss/assets.go
@@ -12,7 +12,6 @@ import (
 	"github.com/sourcegraph/sourcegraph/ui/assets"
 )

-//go:embed *
 var assetsFS embed.FS
 var afs fs.FS = assetsFS

