// vscode/src/vs/editor/common/config/defaultConfig.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var nls = require('vs/nls');
var platform = require('vs/base/common/platform');
var wordHelper_1 = require('vs/editor/common/model/wordHelper');
exports.DEFAULT_INDENTATION = {
    tabSize: 4,
    insertSpaces: true,
    detectIndentation: true
};
exports.DEFAULT_TRIM_AUTO_WHITESPACE = true;
var DEFAULT_WINDOWS_FONT_FAMILY = 'Consolas, \'Courier New\', monospace';
var DEFAULT_MAC_FONT_FAMILY = 'Menlo, Monaco, \'Courier New\', monospace';
var DEFAULT_LINUX_FONT_FAMILY = '\'Droid Sans Mono\', \'Courier New\', monospace, \'Droid Sans Fallback\'';
/**
 * Determined from empirical observations.
 */
exports.GOLDEN_LINE_HEIGHT_RATIO = platform.isMacintosh ? 1.5 : 1.35;
var ConfigClass = (function () {
    function ConfigClass() {
        this.editor = {
            experimentalScreenReader: true,
            rulers: [],
            wordSeparators: wordHelper_1.USUAL_WORD_SEPARATORS,
            selectionClipboard: true,
            ariaLabel: nls.localize('editorViewAccessibleLabel', "Editor content"),
            lineNumbers: 'on',
            selectOnLineNumbers: true,
            lineNumbersMinChars: 5,
            glyphMargin: false,
            lineDecorationsWidth: 10,
            revealHorizontalRightPadding: 30,
            roundedSelection: true,
            theme: 'vs',
            readOnly: false,
            scrollbar: {
                verticalScrollbarSize: 14,
                horizontal: 'auto',
                useShadows: true,
                verticalHasArrows: false,
                horizontalHasArrows: false
            },
            overviewRulerLanes: 2,
            cursorBlinking: 'blink',
            mouseWheelZoom: false,
            cursorStyle: 'line',
            fontLigatures: false,
            disableTranslate3d: false,
            hideCursorInOverviewRuler: false,
            scrollBeyondLastLine: true,
            automaticLayout: false,
            wrappingColumn: 300,
            wordWrap: false,
            wrappingIndent: 'same',
            wordWrapBreakBeforeCharacters: '([{‘“〈《「『【〔（［｛｢£¥＄￡￥+＋',
            wordWrapBreakAfterCharacters: ' \t})]?|&,;¢°′″‰℃、。｡､￠，．：；？！％・･ゝゞヽヾーァィゥェォッャュョヮヵヶぁぃぅぇぉっゃゅょゎゕゖㇰㇱㇲㇳㇴㇵㇶㇷㇸㇹㇺㇻㇼㇽㇾㇿ々〻ｧｨｩｪｫｬｭｮｯｰ’”〉》」』】〕）］｝｣',
            wordWrapBreakObtrusiveCharacters: '.',
            // Features
            hover: true,
            contextmenu: true,
            mouseWheelScrollSensitivity: 1,
            quickSuggestions: true,
            quickSuggestionsDelay: 10,
            parameterHints: true,
            iconsInSuggestions: true,
            autoClosingBrackets: true,
            formatOnType: false,
            suggestOnTriggerCharacters: true,
            acceptSuggestionOnEnter: true,
            snippetSuggestions: 'bottom',
            tabCompletion: false,
            wordBasedSuggestions: true,
            suggestFontSize: 0,
            suggestLineHeight: 0,
            selectionHighlight: true,
            codeLens: true,
            referenceInfos: true,
            folding: true,
            renderWhitespace: 'none',
            renderControlCharacters: false,
            renderIndentGuides: false,
            renderLineHighlight: true,
            useTabStops: true,
            fontFamily: (platform.isMacintosh ? DEFAULT_MAC_FONT_FAMILY : (platform.isLinux ? DEFAULT_LINUX_FONT_FAMILY : DEFAULT_WINDOWS_FONT_FAMILY)),
            fontWeight: 'normal',
            fontSize: (platform.isMacintosh ? 12 : 14),
            lineHeight: 0
        };
    }
    return ConfigClass;
}());
exports.DefaultConfig = new ConfigClass();
