// vscode/src/vs/workbench/services/editor/browser/editorService.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var winjs_base_1 = require("vs/base/common/winjs.base");
var uri_1 = require("vs/base/common/uri");
var network = require("vs/base/common/network");
var platform_1 = require("vs/platform/platform");
var paths_1 = require("vs/base/common/paths");
var editor_1 = require("vs/workbench/common/editor");
var resourceEditorInput_1 = require("vs/workbench/common/editor/resourceEditorInput");
var untitledEditorInput_1 = require("vs/workbench/common/editor/untitledEditorInput");
var untitledEditorService_1 = require("vs/workbench/services/untitled/common/untitledEditorService");
var editorService_1 = require("vs/workbench/services/editor/common/editorService");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var diffEditorInput_1 = require("vs/workbench/common/editor/diffEditorInput");
var workspace_1 = require("vs/platform/workspace/common/workspace");
var nls = require("vs/nls");
var labels_1 = require("vs/base/common/labels");
var WorkbenchEditorService = (function () {
    function WorkbenchEditorService(editorPart, untitledEditorService, workspaceContextService, instantiationService) {
        this.untitledEditorService = untitledEditorService;
        this.workspaceContextService = workspaceContextService;
        this.instantiationService = instantiationService;
        this.editorPart = editorPart;
        this.fileInputDescriptor = platform_1.Registry.as(editor_1.Extensions.Editors).getDefaultFileInput();
    }
    WorkbenchEditorService.prototype.getActiveEditor = function () {
        return this.editorPart.getActiveEditor();
    };
    WorkbenchEditorService.prototype.getActiveEditorInput = function () {
        return this.editorPart.getActiveEditorInput();
    };
    WorkbenchEditorService.prototype.getVisibleEditors = function () {
        return this.editorPart.getVisibleEditors();
    };
    WorkbenchEditorService.prototype.isVisible = function (input, includeSideBySide) {
        if (!input) {
            return false;
        }
        return this.getVisibleEditors().some(function (editor) {
            if (!editor.input) {
                return false;
            }
            if (input.matches(editor.input)) {
                return true;
            }
            if (includeSideBySide && editor.input instanceof editor_1.SideBySideEditorInput) {
                var sideBySideInput = editor.input;
                return input.matches(sideBySideInput.master) || input.matches(sideBySideInput.details);
            }
            return false;
        });
    };
    WorkbenchEditorService.prototype.openEditor = function (input, arg2, arg3) {
        var _this = this;
        if (!input) {
            return winjs_base_1.TPromise.as(null);
        }
        // Workbench Input Support
        if (input instanceof editor_1.EditorInput) {
            return this.doOpenEditor(input, this.toOptions(arg2), arg3);
        }
        // Support opening foreign resources (such as a http link that points outside of the workbench)
        var resourceInput = input;
        if (resourceInput.resource instanceof uri_1.default) {
            var schema = resourceInput.resource.scheme;
            if (schema === network.Schemas.http || schema === network.Schemas.https) {
                window.open(resourceInput.resource.toString(true));
                return winjs_base_1.TPromise.as(null);
            }
        }
        // Untyped Text Editor Support (required for code that uses this service below workbench level)
        var textInput = input;
        return this.createInput(textInput).then(function (typedInput) {
            if (typedInput) {
                return _this.doOpenEditor(typedInput, editor_1.TextEditorOptions.from(textInput), arg2);
            }
            return winjs_base_1.TPromise.as(null);
        });
    };
    WorkbenchEditorService.prototype.toOptions = function (arg1) {
        if (!arg1 || arg1 instanceof editor_1.EditorOptions) {
            return arg1;
        }
        var textOptions = arg1;
        if (!!textOptions.selection) {
            return editor_1.TextEditorOptions.create(arg1);
        }
        return editor_1.EditorOptions.create(arg1);
    };
    WorkbenchEditorService.prototype.doOpenEditor = function (input, options, arg3) {
        return this.editorPart.openEditor(input, options, arg3);
    };
    WorkbenchEditorService.prototype.openEditors = function (editors) {
        var _this = this;
        return winjs_base_1.TPromise.join(editors.map(function (editor) { return _this.createInput(editor.input); })).then(function (inputs) {
            var typedInputs = inputs.map(function (input, index) {
                var options = editors[index].input instanceof editor_1.EditorInput ? _this.toOptions(editors[index].options) : editor_1.TextEditorOptions.from(editors[index].input);
                return {
                    input: input,
                    options: options,
                    position: editors[index].position
                };
            });
            return _this.editorPart.openEditors(typedInputs);
        });
    };
    WorkbenchEditorService.prototype.replaceEditors = function (editors, position) {
        var _this = this;
        return winjs_base_1.TPromise.join(editors.map(function (editor) { return _this.createInput(editor.toReplace); })).then(function (toReplaceInputs) {
            return winjs_base_1.TPromise.join(editors.map(function (editor) { return _this.createInput(editor.replaceWith); })).then(function (replaceWithInputs) {
                var typedReplacements = editors.map(function (editor, index) {
                    var options = editor.toReplace instanceof editor_1.EditorInput ? _this.toOptions(editor.options) : editor_1.TextEditorOptions.from(editor.replaceWith);
                    return {
                        toReplace: toReplaceInputs[index],
                        replaceWith: replaceWithInputs[index],
                        options: options
                    };
                });
                return _this.editorPart.replaceEditors(typedReplacements, position);
            });
        });
    };
    WorkbenchEditorService.prototype.closeEditor = function (position, input) {
        return this.doCloseEditor(position, input);
    };
    WorkbenchEditorService.prototype.doCloseEditor = function (position, input) {
        return this.editorPart.closeEditor(position, input);
    };
    WorkbenchEditorService.prototype.closeEditors = function (position, except, direction) {
        return this.editorPart.closeEditors(position, except, direction);
    };
    WorkbenchEditorService.prototype.closeAllEditors = function (except) {
        return this.editorPart.closeAllEditors(except);
    };
    WorkbenchEditorService.prototype.createInput = function (input) {
        var _this = this;
        // Workbench Input Support
        if (input instanceof editor_1.EditorInput) {
            return winjs_base_1.TPromise.as(input);
        }
        // Side by Side Support
        var resourceSideBySideInput = input;
        if (resourceSideBySideInput.masterResource && resourceSideBySideInput.detailResource) {
            return this.createInput({ resource: resourceSideBySideInput.masterResource }).then(function (masterInput) {
                return _this.createInput({ resource: resourceSideBySideInput.detailResource }).then(function (detailInput) {
                    return new editor_1.SideBySideEditorInput(resourceSideBySideInput.label || masterInput.getName(), typeof resourceSideBySideInput.description === 'string' ? resourceSideBySideInput.description : masterInput.getDescription(), detailInput, masterInput);
                });
            });
        }
        // Diff Editor Support
        var resourceDiffInput = input;
        if (resourceDiffInput.leftResource && resourceDiffInput.rightResource) {
            return this.createInput({ resource: resourceDiffInput.leftResource }).then(function (leftInput) {
                return _this.createInput({ resource: resourceDiffInput.rightResource }).then(function (rightInput) {
                    var label = resourceDiffInput.label || toDiffLabel(resourceDiffInput.leftResource, resourceDiffInput.rightResource, _this.workspaceContextService);
                    return new diffEditorInput_1.DiffEditorInput(label, resourceDiffInput.description, leftInput, rightInput);
                });
            });
        }
        // Base Text Editor Support for inmemory resources
        var resourceInput = input;
        // Untitled file support
        if (resourceInput.resource instanceof uri_1.default && (resourceInput.resource.scheme === untitledEditorInput_1.UntitledEditorInput.SCHEMA)) {
            return winjs_base_1.TPromise.as(this.untitledEditorService.createOrGet(resourceInput.resource));
        }
        else if (this.fileInputDescriptor && resourceInput.resource instanceof uri_1.default && resourceInput.resource.scheme === network.Schemas.file) {
            return this.createFileInput(resourceInput.resource, resourceInput.encoding);
        }
        else if (resourceInput.resource instanceof uri_1.default) {
            return winjs_base_1.TPromise.as(this.instantiationService.createInstance(resourceEditorInput_1.ResourceEditorInput, resourceInput.label || paths_1.basename(resourceInput.resource.fsPath), typeof resourceInput.description === 'string' ? resourceInput.description : paths_1.dirname(resourceInput.resource.fsPath), resourceInput.resource));
        }
        return winjs_base_1.TPromise.as(null);
    };
    WorkbenchEditorService.prototype.createFileInput = function (resource, encoding) {
        return this.instantiationService.createInstance(this.fileInputDescriptor).then(function (typedFileInput) {
            typedFileInput.setResource(resource);
            typedFileInput.setPreferredEncoding(encoding);
            return typedFileInput;
        });
    };
    return WorkbenchEditorService;
}());
WorkbenchEditorService = __decorate([
    __param(1, untitledEditorService_1.IUntitledEditorService),
    __param(2, workspace_1.IWorkspaceContextService),
    __param(3, instantiation_1.IInstantiationService)
], WorkbenchEditorService);
exports.WorkbenchEditorService = WorkbenchEditorService;
function toDiffLabel(res1, res2, context) {
    var leftName = labels_1.getPathLabel(res1.fsPath, context);
    var rightName = labels_1.getPathLabel(res2.fsPath, context);
    return nls.localize('compareLabels', "{0} ↔ {1}", leftName, rightName);
}
/**
 * Subclass of workbench editor service that delegates all calls to the provided editor service. Subclasses can choose to override the behavior
 * of openEditor() and closeEditor() by providing a handler.
 *
 * This gives clients a chance to override the behavior of openEditor() and closeEditor().
 */
var DelegatingWorkbenchEditorService = (function (_super) {
    __extends(DelegatingWorkbenchEditorService, _super);
    function DelegatingWorkbenchEditorService(untitledEditorService, instantiationService, workspaceContextService, editorService) {
        return _super.call(this, editorService, untitledEditorService, workspaceContextService, instantiationService) || this;
    }
    DelegatingWorkbenchEditorService.prototype.setEditorOpenHandler = function (handler) {
        this.editorOpenHandler = handler;
    };
    DelegatingWorkbenchEditorService.prototype.setEditorCloseHandler = function (handler) {
        this.editorCloseHandler = handler;
    };
    DelegatingWorkbenchEditorService.prototype.doOpenEditor = function (input, options, arg3) {
        var _this = this;
        var handleOpen = this.editorOpenHandler ? this.editorOpenHandler(input, options, arg3) : winjs_base_1.TPromise.as(void 0);
        return handleOpen.then(function (editor) {
            if (editor) {
                return winjs_base_1.TPromise.as(editor);
            }
            return _super.prototype.doOpenEditor.call(_this, input, options, arg3);
        });
    };
    DelegatingWorkbenchEditorService.prototype.doCloseEditor = function (position, input) {
        var _this = this;
        var handleClose = this.editorCloseHandler ? this.editorCloseHandler(position, input) : winjs_base_1.TPromise.as(void 0);
        return handleClose.then(function () {
            return _super.prototype.doCloseEditor.call(_this, position, input);
        });
    };
    return DelegatingWorkbenchEditorService;
}(WorkbenchEditorService));
DelegatingWorkbenchEditorService = __decorate([
    __param(0, untitledEditorService_1.IUntitledEditorService),
    __param(1, instantiation_1.IInstantiationService),
    __param(2, workspace_1.IWorkspaceContextService),
    __param(3, editorService_1.IWorkbenchEditorService)
], DelegatingWorkbenchEditorService);
exports.DelegatingWorkbenchEditorService = DelegatingWorkbenchEditorService;
