// vscode/src/vs/platform/lifecycle/common/lifecycle.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
exports.ILifecycleService = instantiation_1.createDecorator('lifecycleService');
var ShutdownReason;
(function (ShutdownReason) {
    /** Window is closed */
    ShutdownReason[ShutdownReason["CLOSE"] = 0] = "CLOSE";
    /** Application is quit */
    ShutdownReason[ShutdownReason["QUIT"] = 1] = "QUIT";
    /** Window is reloaded */
    ShutdownReason[ShutdownReason["RELOAD"] = 2] = "RELOAD";
    /** Other configuration loaded into window */
    ShutdownReason[ShutdownReason["LOAD"] = 3] = "LOAD";
})(ShutdownReason = exports.ShutdownReason || (exports.ShutdownReason = {}));
exports.NullLifecycleService = {
    _serviceBrand: null,
    willShutdown: false,
    onWillShutdown: function () { return ({ dispose: function () { } }); },
    onShutdown: function (reason) { return ({ dispose: function () { } }); }
};
