// vscode/src/vs/workbench/parts/files/common/files.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var events_1 = require('vs/base/common/events');
var editor_1 = require('vs/workbench/common/editor');
var instantiation_1 = require('vs/platform/instantiation/common/instantiation');
var explorerViewModel_1 = require('vs/workbench/parts/files/common/explorerViewModel');
var contextkey_1 = require('vs/platform/contextkey/common/contextkey');
/**
 * Explorer viewlet id.
 */
exports.VIEWLET_ID = 'workbench.view.explorer';
exports.ExplorerViewletVisible = new contextkey_1.RawContextKey('explorerViewletVisible', true);
/**
 * File editor input id.
 */
exports.FILE_EDITOR_INPUT_ID = 'workbench.editors.files.fileEditorInput';
/**
 * Text file editor id.
 */
exports.TEXT_FILE_EDITOR_ID = 'workbench.editors.files.textFileEditor';
/**
 * Binary file editor id.
 */
exports.BINARY_FILE_EDITOR_ID = 'workbench.editors.files.binaryFileEditor';
/**
 * API class to denote file editor inputs. Internal implementation is provided.
 *
 * Note: This class is not intended to be instantiated.
 */
var FileEditorInput = (function (_super) {
    __extends(FileEditorInput, _super);
    function FileEditorInput() {
        _super.apply(this, arguments);
    }
    return FileEditorInput;
}(editor_1.EditorInput));
exports.FileEditorInput = FileEditorInput;
/**
 * Helper to get a file resource from an object.
 */
function asFileResource(obj) {
    if (obj instanceof explorerViewModel_1.FileStat) {
        var stat = obj;
        return {
            resource: stat.resource,
            mimes: stat.mime ? stat.mime.split(', ') : [],
            isDirectory: stat.isDirectory
        };
    }
    return null;
}
exports.asFileResource = asFileResource;
/**
 * States the text text file editor model can be in.
 */
(function (ModelState) {
    ModelState[ModelState["SAVED"] = 0] = "SAVED";
    ModelState[ModelState["DIRTY"] = 1] = "DIRTY";
    ModelState[ModelState["PENDING_SAVE"] = 2] = "PENDING_SAVE";
    ModelState[ModelState["CONFLICT"] = 3] = "CONFLICT";
    ModelState[ModelState["ERROR"] = 4] = "ERROR";
})(exports.ModelState || (exports.ModelState = {}));
var ModelState = exports.ModelState;
/**
 * Local file change events are being emitted when a file is added, removed, moved or its contents got updated. These events
 * are being emitted from within the workbench and are not reflecting the truth on the disk file system. For that, please
 * use FileChangesEvent instead.
 */
var LocalFileChangeEvent = (function (_super) {
    __extends(LocalFileChangeEvent, _super);
    function LocalFileChangeEvent(before, after, originalEvent) {
        _super.call(this, null, before, after, originalEvent);
    }
    /**
     * Returns the meta information of the file before the event occurred or null if the file is new.
     */
    LocalFileChangeEvent.prototype.getBefore = function () {
        return this.oldValue;
    };
    /**
     * Returns the meta information of the file after the event occurred or null if the file got deleted.
     */
    LocalFileChangeEvent.prototype.getAfter = function () {
        return this.newValue;
    };
    /**
     * Indicates if the file was added as a new file.
     */
    LocalFileChangeEvent.prototype.gotAdded = function () {
        return !this.oldValue && !!this.newValue;
    };
    /**
     * Indicates if the file was moved to a different path.
     */
    LocalFileChangeEvent.prototype.gotMoved = function () {
        return !!this.oldValue && !!this.newValue && this.oldValue.resource.toString() !== this.newValue.resource.toString();
    };
    /**
     * Indicates if the files metadata was updated.
     */
    LocalFileChangeEvent.prototype.gotUpdated = function () {
        return !!this.oldValue && !!this.newValue && !this.gotMoved() && this.oldValue !== this.newValue;
    };
    /**
     * Indicates if the file was deleted.
     */
    LocalFileChangeEvent.prototype.gotDeleted = function () {
        return !!this.oldValue && !this.newValue;
    };
    return LocalFileChangeEvent;
}(events_1.PropertyChangeEvent));
exports.LocalFileChangeEvent = LocalFileChangeEvent;
(function (StateChange) {
    StateChange[StateChange["DIRTY"] = 0] = "DIRTY";
    StateChange[StateChange["SAVING"] = 1] = "SAVING";
    StateChange[StateChange["SAVE_ERROR"] = 2] = "SAVE_ERROR";
    StateChange[StateChange["SAVED"] = 3] = "SAVED";
    StateChange[StateChange["REVERTED"] = 4] = "REVERTED";
    StateChange[StateChange["ENCODING"] = 5] = "ENCODING";
})(exports.StateChange || (exports.StateChange = {}));
var StateChange = exports.StateChange;
var TextFileModelChangeEvent = (function () {
    function TextFileModelChangeEvent(model, kind) {
        this._resource = model.getResource();
        this._kind = kind;
    }
    Object.defineProperty(TextFileModelChangeEvent.prototype, "resource", {
        get: function () {
            return this._resource;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TextFileModelChangeEvent.prototype, "kind", {
        get: function () {
            return this._kind;
        },
        enumerable: true,
        configurable: true
    });
    return TextFileModelChangeEvent;
}());
exports.TextFileModelChangeEvent = TextFileModelChangeEvent;
exports.TEXT_FILE_SERVICE_ID = 'textFileService';
(function (AutoSaveMode) {
    AutoSaveMode[AutoSaveMode["OFF"] = 0] = "OFF";
    AutoSaveMode[AutoSaveMode["AFTER_SHORT_DELAY"] = 1] = "AFTER_SHORT_DELAY";
    AutoSaveMode[AutoSaveMode["AFTER_LONG_DELAY"] = 2] = "AFTER_LONG_DELAY";
    AutoSaveMode[AutoSaveMode["ON_FOCUS_CHANGE"] = 3] = "ON_FOCUS_CHANGE";
    AutoSaveMode[AutoSaveMode["ON_WINDOW_CHANGE"] = 4] = "ON_WINDOW_CHANGE";
})(exports.AutoSaveMode || (exports.AutoSaveMode = {}));
var AutoSaveMode = exports.AutoSaveMode;
(function (SaveReason) {
    SaveReason[SaveReason["EXPLICIT"] = 1] = "EXPLICIT";
    SaveReason[SaveReason["AUTO"] = 2] = "AUTO";
    SaveReason[SaveReason["FOCUS_CHANGE"] = 3] = "FOCUS_CHANGE";
    SaveReason[SaveReason["WINDOW_CHANGE"] = 4] = "WINDOW_CHANGE";
})(exports.SaveReason || (exports.SaveReason = {}));
var SaveReason = exports.SaveReason;
exports.ITextFileService = instantiation_1.createDecorator(exports.TEXT_FILE_SERVICE_ID);
