// vscode/src/vs/base/common/lifecycle.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var functional_1 = require("vs/base/common/functional");
exports.empty = Object.freeze({
    dispose: function () { }
});
function dispose(first) {
    var rest = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        rest[_i - 1] = arguments[_i];
    }
    if (Array.isArray(first)) {
        first.forEach(function (d) { return d && d.dispose(); });
        return [];
    }
    else if (rest.length === 0) {
        if (first) {
            first.dispose();
            return first;
        }
        return undefined;
    }
    else {
        dispose(first);
        dispose(rest);
        return [];
    }
}
exports.dispose = dispose;
function combinedDisposable(disposables) {
    return { dispose: function () { return dispose(disposables); } };
}
exports.combinedDisposable = combinedDisposable;
function toDisposable() {
    var fns = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        fns[_i] = arguments[_i];
    }
    return combinedDisposable(fns.map(function (fn) { return ({ dispose: fn }); }));
}
exports.toDisposable = toDisposable;
var Disposable = (function () {
    function Disposable() {
        this._toDispose = [];
    }
    Disposable.prototype.dispose = function () {
        this._toDispose = dispose(this._toDispose);
    };
    Disposable.prototype._register = function (t) {
        this._toDispose.push(t);
        return t;
    };
    return Disposable;
}());
exports.Disposable = Disposable;
var Disposables = (function (_super) {
    __extends(Disposables, _super);
    function Disposables() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Disposables.prototype.add = function (arg) {
        if (!Array.isArray(arg)) {
            return this._register(arg);
        }
        else {
            for (var _i = 0, arg_1 = arg; _i < arg_1.length; _i++) {
                var element = arg_1[_i];
                return this._register(element);
            }
            return undefined;
        }
    };
    return Disposables;
}(Disposable));
exports.Disposables = Disposables;
var OneDisposable = (function () {
    function OneDisposable() {
    }
    Object.defineProperty(OneDisposable.prototype, "value", {
        set: function (value) {
            if (this._value) {
                this._value.dispose();
            }
            this._value = value;
        },
        enumerable: true,
        configurable: true
    });
    OneDisposable.prototype.dispose = function () {
        this.value = null;
    };
    return OneDisposable;
}());
exports.OneDisposable = OneDisposable;
var ReferenceCollection = (function () {
    function ReferenceCollection() {
        this.references = Object.create(null);
    }
    ReferenceCollection.prototype.acquire = function (key) {
        var _this = this;
        var reference = this.references[key];
        if (!reference) {
            reference = this.references[key] = { counter: 0, object: this.createReferencedObject(key) };
        }
        var object = reference.object;
        var dispose = functional_1.once(function () {
            if (--reference.counter === 0) {
                _this.destroyReferencedObject(reference.object);
                delete _this.references[key];
            }
        });
        reference.counter++;
        return { object: object, dispose: dispose };
    };
    return ReferenceCollection;
}());
exports.ReferenceCollection = ReferenceCollection;
var ImmortalReference = (function () {
    function ImmortalReference(object) {
        this.object = object;
    }
    ImmortalReference.prototype.dispose = function () { };
    return ImmortalReference;
}());
exports.ImmortalReference = ImmortalReference;
