// vscode/src/vs/workbench/parts/git/common/gitContentProvider.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var modelService_1 = require("vs/editor/common/services/modelService");
var lifecycle_1 = require("vs/base/common/lifecycle");
var paths = require("vs/base/common/paths");
var async_1 = require("vs/base/common/async");
var winjs_base_1 = require("vs/base/common/winjs.base");
var resolverService_1 = require("vs/editor/common/services/resolverService");
var scm_1 = require("vs/workbench/services/scm/common/scm");
var git_1 = require("vs/workbench/parts/git/common/git");
var GitContentProvider = (function () {
    function GitContentProvider(textModelResolverService, modelService, scmService, gitService) {
        this.modelService = modelService;
        this.scmService = scmService;
        this.gitService = gitService;
        this.scmService.registerBaselineResourceProvider(this);
        textModelResolverService.registerTextModelContentProvider('git-index', this);
    }
    GitContentProvider.prototype.getBaselineResource = function (resource) {
        return winjs_base_1.TPromise.as(resource.with({ scheme: 'git-index' }));
    };
    GitContentProvider.prototype.provideTextContent = function (uri) {
        var _this = this;
        var model = this.modelService.createModel('', null, uri);
        var throttler = new async_1.Throttler();
        var setModelContents = function (contents) {
            if (model.isDisposed()) {
                return;
            }
            model.setValue(contents || '');
        };
        var updateModel = function () {
            var gitModel = _this.gitService.getModel();
            var root = gitModel.getRepositoryRoot();
            if (!root) {
                return winjs_base_1.TPromise.as(null);
            }
            var path = uri.fsPath;
            var relativePath = paths.relative(root, path).replace(/\\/g, '/');
            if (/^\.\./.test(relativePath)) {
                return winjs_base_1.TPromise.as(null);
            }
            var treeish = gitModel.getStatus().find(relativePath, git_1.StatusType.INDEX) ? '~' : 'HEAD';
            return _this.gitService.buffer(path, treeish).then(setModelContents);
        };
        var triggerModelUpdate = function () {
            if (_this.gitService.getState() !== git_1.ServiceState.OK) {
                return;
            }
            throttler.queue(updateModel);
        };
        var disposables = [
            this.gitService.addListener2(git_1.ServiceEvents.STATE_CHANGED, triggerModelUpdate),
            this.gitService.addListener2(git_1.ServiceEvents.OPERATION_END, function (e) {
                if (e.operation.id !== git_1.ServiceOperations.BACKGROUND_FETCH) {
                    triggerModelUpdate();
                }
            })
        ];
        model.onWillDispose(function () { return lifecycle_1.dispose(disposables); });
        triggerModelUpdate();
        return winjs_base_1.TPromise.as(model);
    };
    GitContentProvider.prototype.getId = function () {
        return 'git.contentprovider';
    };
    return GitContentProvider;
}());
GitContentProvider = __decorate([
    __param(0, resolverService_1.ITextModelResolverService),
    __param(1, modelService_1.IModelService),
    __param(2, scm_1.ISCMService),
    __param(3, git_1.IGitService)
], GitContentProvider);
exports.GitContentProvider = GitContentProvider;
