// vscode/src/vs/workbench/parts/search/common/constants.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var contextkey_1 = require('vs/platform/contextkey/common/contextkey');
exports.VIEWLET_ID = 'workbench.view.search';
exports.FindInFilesActionId = 'workbench.action.findInFiles';
exports.FocusActiveEditorActionId = 'search.action.focusActiveEditor';
exports.ToggleCaseSensitiveActionId = 'toggleSearchCaseSensitive';
exports.ToggleWholeWordActionId = 'toggleSearchWholeWord';
exports.ToggleRegexActionId = 'toggleSearchRegex';
exports.CloseReplaceWidgetActionId = 'closeReplaceInFilesWidget';
exports.SearchViewletVisibleKey = new contextkey_1.RawContextKey('searchViewletVisible', true);
exports.InputBoxFocussedKey = new contextkey_1.RawContextKey('inputBoxFocus', false);
exports.SearchInputBoxFocussedKey = new contextkey_1.RawContextKey('searchInputBoxFocus', false);
exports.ReplaceInputBoxFocussedKey = new contextkey_1.RawContextKey('replaceInputBoxFocus', false);
exports.ReplaceActiveKey = new contextkey_1.RawContextKey('replaceActive', false);
