// vscode/src/vs/platform/keybinding/common/abstractKeybindingService.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var nls = require("vs/nls");
var keybinding_1 = require("vs/base/common/keybinding");
var lifecycle_1 = require("vs/base/common/lifecycle");
var severity_1 = require("vs/base/common/severity");
var arrays_1 = require("vs/base/common/arrays");
var commands_1 = require("vs/platform/commands/common/commands");
var event_1 = require("vs/base/common/event");
var AbstractKeybindingService = (function () {
    function AbstractKeybindingService(contextKeyService, commandService, messageService, statusService) {
        this.toDispose = [];
        this._contextKeyService = contextKeyService;
        this._commandService = commandService;
        this._statusService = statusService;
        this._messageService = messageService;
        this._currentChord = 0;
        this._currentChordStatusMessage = null;
        this._onDidUpdateKeybindings = new event_1.Emitter();
        this.toDispose.push(this._onDidUpdateKeybindings);
    }
    AbstractKeybindingService.prototype.dispose = function () {
        this.toDispose = lifecycle_1.dispose(this.toDispose);
    };
    Object.defineProperty(AbstractKeybindingService.prototype, "onDidUpdateKeybindings", {
        get: function () {
            return this._onDidUpdateKeybindings ? this._onDidUpdateKeybindings.event : event_1.default.None; // Sinon stubbing walks properties on prototype
        },
        enumerable: true,
        configurable: true
    });
    AbstractKeybindingService.prototype.getLabelFor = function (keybinding) {
        return keybinding._toUSLabel();
    };
    AbstractKeybindingService.prototype.getHTMLLabelFor = function (keybinding) {
        return keybinding._toUSHTMLLabel();
    };
    AbstractKeybindingService.prototype.getAriaLabelFor = function (keybinding) {
        return keybinding._toUSAriaLabel();
    };
    AbstractKeybindingService.prototype.getElectronAcceleratorFor = function (keybinding) {
        return keybinding._toElectronAccelerator();
    };
    AbstractKeybindingService.prototype.getDefaultKeybindings = function () {
        return this._getResolver().getDefaultKeybindings() + '\n\n' + this._getAllCommandsAsComment();
    };
    AbstractKeybindingService.prototype.customKeybindingsCount = function () {
        return 0;
    };
    AbstractKeybindingService.prototype.lookupKeybindings = function (commandId) {
        return this._getResolver().lookupKeybinding(commandId);
    };
    AbstractKeybindingService.prototype._getAllCommandsAsComment = function () {
        var commands = commands_1.CommandsRegistry.getCommands();
        var unboundCommands = [];
        var boundCommands = this._getResolver().getDefaultBoundCommands();
        for (var id in commands) {
            if (id[0] === '_' || id.indexOf('vscode.') === 0) {
                continue;
            }
            if (typeof commands[id].description === 'object'
                && !arrays_1.isFalsyOrEmpty(commands[id].description.args)) {
                continue;
            }
            if (boundCommands[id]) {
                continue;
            }
            unboundCommands.push(id);
        }
        var pretty = unboundCommands.sort().join('\n// - ');
        return '// ' + nls.localize('unboundCommands', "Here are other available commands: ") + '\n// - ' + pretty;
    };
    AbstractKeybindingService.prototype.resolve = function (keybinding, target) {
        if (keybinding.isModifierKey()) {
            return null;
        }
        var contextValue = this._contextKeyService.getContextValue(target);
        return this._getResolver().resolve(contextValue, this._currentChord, keybinding.value);
    };
    AbstractKeybindingService.prototype._dispatch = function (keybinding, target) {
        var _this = this;
        // Check modifier key here and cancel early, it's also checked in resolve as the function
        // is used externally.
        var shouldPreventDefault = false;
        if (keybinding.isModifierKey()) {
            return shouldPreventDefault;
        }
        var resolveResult = this.resolve(keybinding, target);
        if (resolveResult && resolveResult.enterChord) {
            shouldPreventDefault = true;
            this._currentChord = resolveResult.enterChord;
            if (this._statusService) {
                var firstPartLabel = this.getLabelFor(new keybinding_1.Keybinding(this._currentChord));
                this._currentChordStatusMessage = this._statusService.setStatusMessage(nls.localize('first.chord', "({0}) was pressed. Waiting for second key of chord...", firstPartLabel));
            }
            return shouldPreventDefault;
        }
        if (this._statusService && this._currentChord) {
            if (!resolveResult || !resolveResult.commandId) {
                var firstPartLabel = this.getLabelFor(new keybinding_1.Keybinding(this._currentChord));
                var chordPartLabel = this.getLabelFor(keybinding);
                this._statusService.setStatusMessage(nls.localize('missing.chord', "The key combination ({0}, {1}) is not a command.", firstPartLabel, chordPartLabel), 10 * 1000 /* 10s */);
                shouldPreventDefault = true;
            }
        }
        if (this._currentChordStatusMessage) {
            this._currentChordStatusMessage.dispose();
            this._currentChordStatusMessage = null;
        }
        this._currentChord = 0;
        if (resolveResult && resolveResult.commandId) {
            if (!/^\^/.test(resolveResult.commandId)) {
                shouldPreventDefault = true;
            }
            var commandId = resolveResult.commandId.replace(/^\^/, '');
            this._commandService.executeCommand(commandId, resolveResult.commandArgs || {}).done(undefined, function (err) {
                _this._messageService.show(severity_1.default.Warning, err);
            });
        }
        return shouldPreventDefault;
    };
    return AbstractKeybindingService;
}());
exports.AbstractKeybindingService = AbstractKeybindingService;
