// vscode/src/vs/workbench/parts/output/browser/outputEditorInput.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var nls = require("vs/nls");
var lifecycle = require("vs/base/common/lifecycle");
var strings = require("vs/base/common/strings");
var async_1 = require("vs/base/common/async");
var stringEditorInput_1 = require("vs/workbench/common/editor/stringEditorInput");
var output_1 = require("vs/workbench/parts/output/common/output");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var panelService_1 = require("vs/workbench/services/panel/common/panelService");
/**
 * Output Editor Input
 */
var OutputEditorInput = (function (_super) {
    __extends(OutputEditorInput, _super);
    function OutputEditorInput(outputChannel, instantiationService, outputService, panelService) {
        var _this = _super.call(this, nls.localize('output', "Output"), outputChannel ? nls.localize('outputChannel', "for '{0}'", outputChannel.label) : '', '', output_1.OUTPUT_MIME, true, instantiationService) || this;
        _this.outputChannel = outputChannel;
        _this.outputService = outputService;
        _this.panelService = panelService;
        _this.bufferedOutput = '';
        _this.toDispose = [];
        _this.toDispose.push(_this.outputService.onOutput(_this.onOutputReceived, _this));
        _this.toDispose.push(_this.outputService.onActiveOutputChannel(function () { return _this.scheduleOutputAppend(); }));
        _this.toDispose.push(_this.panelService.onDidPanelOpen(function (panel) {
            if (panel.getId() === output_1.OUTPUT_PANEL_ID) {
                _this.appendOutput();
            }
        }));
        _this.appendOutputScheduler = new async_1.RunOnceScheduler(function () {
            if (_this.isVisible()) {
                _this.appendOutput();
            }
        }, OutputEditorInput.OUTPUT_DELAY);
        return _this;
    }
    OutputEditorInput.getInstances = function () {
        return Object.keys(OutputEditorInput.instances).map(function (key) { return OutputEditorInput.instances[key]; });
    };
    OutputEditorInput.getInstance = function (instantiationService, channel) {
        if (OutputEditorInput.instances[channel.id]) {
            return OutputEditorInput.instances[channel.id];
        }
        OutputEditorInput.instances[channel.id] = instantiationService.createInstance(OutputEditorInput, channel);
        return OutputEditorInput.instances[channel.id];
    };
    OutputEditorInput.prototype.appendOutput = function () {
        if (this.bufferedOutput.length === 0) {
            return;
        }
        if (this.value.length + this.bufferedOutput.length > output_1.MAX_OUTPUT_LENGTH) {
            this.setValue(this.outputChannel.output);
        }
        else {
            this.append(this.bufferedOutput);
        }
        this.bufferedOutput = '';
        var panel = this.panelService.getActivePanel();
        panel.revealLastLine(true);
    };
    OutputEditorInput.prototype.onOutputReceived = function (e) {
        if (this.outputSet && e.channelId === this.outputChannel.id) {
            if (e.output) {
                this.bufferedOutput = strings.appendWithLimit(this.bufferedOutput, e.output, output_1.MAX_OUTPUT_LENGTH);
                this.scheduleOutputAppend();
            }
            else if (e.output === null) {
                this.bufferedOutput = '';
                this.clearValue(); // special output indicates we should clear
            }
        }
    };
    OutputEditorInput.prototype.isVisible = function () {
        var panel = this.panelService.getActivePanel();
        return panel && panel.getId() === output_1.OUTPUT_PANEL_ID && this.outputService.getActiveChannel().id === this.outputChannel.id;
    };
    OutputEditorInput.prototype.scheduleOutputAppend = function () {
        if (this.isVisible() && this.bufferedOutput && !this.appendOutputScheduler.isScheduled()) {
            this.appendOutputScheduler.schedule();
        }
    };
    OutputEditorInput.prototype.getTypeId = function () {
        return output_1.OUTPUT_EDITOR_INPUT_ID;
    };
    OutputEditorInput.prototype.resolve = function (refresh) {
        var _this = this;
        return _super.prototype.resolve.call(this, refresh).then(function (model) {
            // Just return model if output already set
            if (_this.outputSet) {
                return model;
            }
            _this.setValue(_this.outputChannel.output);
            _this.outputSet = true;
            return model;
        });
    };
    OutputEditorInput.prototype.matches = function (otherInput) {
        if (otherInput instanceof OutputEditorInput) {
            var otherOutputEditorInput = otherInput;
            if (otherOutputEditorInput.outputChannel.id === this.outputChannel.id) {
                return _super.prototype.matches.call(this, otherInput);
            }
        }
        return false;
    };
    OutputEditorInput.prototype.dispose = function () {
        this.appendOutputScheduler.dispose();
        this.toDispose = lifecycle.dispose(this.toDispose);
        _super.prototype.dispose.call(this);
    };
    return OutputEditorInput;
}(stringEditorInput_1.StringEditorInput));
OutputEditorInput.OUTPUT_DELAY = 300; // delay in ms to accumulate output before emitting an event about it
OutputEditorInput.instances = Object.create(null);
OutputEditorInput = __decorate([
    __param(1, instantiation_1.IInstantiationService),
    __param(2, output_1.IOutputService),
    __param(3, panelService_1.IPanelService)
], OutputEditorInput);
exports.OutputEditorInput = OutputEditorInput;
