// vscode/src/vs/editor/contrib/suggest/browser/suggestController.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var nls = require("vs/nls");
var errors_1 = require("vs/base/common/errors");
var arrays_1 = require("vs/base/common/arrays");
var lifecycle_1 = require("vs/base/common/lifecycle");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var telemetry_1 = require("vs/platform/telemetry/common/telemetry");
var contextkey_1 = require("vs/platform/contextkey/common/contextkey");
var commands_1 = require("vs/platform/commands/common/commands");
var editorCommon_1 = require("vs/editor/common/editorCommon");
var editorCommonExtensions_1 = require("vs/editor/common/editorCommonExtensions");
var editorBrowserExtensions_1 = require("vs/editor/browser/editorBrowserExtensions");
var editOperation_1 = require("vs/editor/common/core/editOperation");
var range_1 = require("vs/editor/common/core/range");
var snippet_1 = require("vs/editor/contrib/snippet/common/snippet");
var snippetController_1 = require("vs/editor/contrib/snippet/common/snippetController");
var suggest_1 = require("vs/editor/contrib/suggest/common/suggest");
var suggestModel_1 = require("../common/suggestModel");
var suggestWidget_1 = require("./suggestWidget");
var AcceptOnCharacterOracle = (function () {
    function AcceptOnCharacterOracle(editor, widget, accept) {
        var _this = this;
        this._disposables = [];
        this._activeAcceptCharacters = new Set();
        this._disposables.push(widget.onDidShow(function () { return _this._onItem(widget.getFocusedItem()); }));
        this._disposables.push(widget.onDidFocus(this._onItem, this));
        this._disposables.push(widget.onDidHide(this.reset, this));
        this._disposables.push(editor.onWillType(function (text) {
            if (_this._activeItem) {
                var ch = text[text.length - 1];
                if (_this._activeAcceptCharacters.has(ch) && editor.getConfiguration().contribInfo.acceptSuggestionOnCommitCharacter) {
                    accept(_this._activeItem);
                }
            }
        }));
    }
    AcceptOnCharacterOracle.prototype._onItem = function (item) {
        if (!item || arrays_1.isFalsyOrEmpty(item.suggestion.commitCharacters)) {
            this.reset();
            return;
        }
        this._activeItem = item;
        this._activeAcceptCharacters.clear();
        for (var _i = 0, _a = item.suggestion.commitCharacters; _i < _a.length; _i++) {
            var ch = _a[_i];
            if (ch.length > 0) {
                this._activeAcceptCharacters.add(ch[0]);
            }
        }
    };
    AcceptOnCharacterOracle.prototype.reset = function () {
        this._activeItem = undefined;
    };
    AcceptOnCharacterOracle.prototype.dispose = function () {
        lifecycle_1.dispose(this._disposables);
    };
    return AcceptOnCharacterOracle;
}());
var SuggestController = SuggestController_1 = (function () {
    function SuggestController(editor, commandService, telemetryService, contextKeyService, instantiationService) {
        var _this = this;
        this.editor = editor;
        this.commandService = commandService;
        this.telemetryService = telemetryService;
        this.toDispose = [];
        this.model = new suggestModel_1.SuggestModel(this.editor);
        this.toDispose.push(this.model.onDidTrigger(function (e) { return _this.widget.showTriggered(e.auto); }));
        this.toDispose.push(this.model.onDidSuggest(function (e) { return _this.widget.showSuggestions(e.completionModel, e.isFrozen, e.auto); }));
        this.toDispose.push(this.model.onDidCancel(function (e) { return !e.retrigger && _this.widget.hideWidget(); }));
        // Manage the acceptSuggestionsOnEnter context key
        var acceptSuggestionsOnEnter = suggest_1.Context.AcceptSuggestionsOnEnter.bindTo(contextKeyService);
        var updateFromConfig = function () {
            acceptSuggestionsOnEnter.set(_this.editor.getConfiguration().contribInfo.acceptSuggestionOnEnter);
        };
        this.toDispose.push(this.editor.onDidChangeConfiguration(function (e) { return updateFromConfig(); }));
        updateFromConfig();
        this.widget = instantiationService.createInstance(suggestWidget_1.SuggestWidget, this.editor);
        this.toDispose.push(this.widget.onDidSelect(this.onDidSelectItem, this));
        // Wire up logic to accept a suggestion on certain characters
        var autoAcceptOracle = new AcceptOnCharacterOracle(editor, this.widget, function (item) { return _this.onDidSelectItem(item); });
        this.toDispose.push(autoAcceptOracle, this.model.onDidSuggest(function (e) {
            if (e.completionModel.items.length === 0) {
                autoAcceptOracle.reset();
            }
        }));
    }
    SuggestController.get = function (editor) {
        return editor.getContribution(SuggestController_1.ID);
    };
    SuggestController.prototype.getId = function () {
        return SuggestController_1.ID;
    };
    SuggestController.prototype.dispose = function () {
        this.toDispose = lifecycle_1.dispose(this.toDispose);
        if (this.widget) {
            this.widget.dispose();
            this.widget = null;
        }
        if (this.model) {
            this.model.dispose();
            this.model = null;
        }
    };
    SuggestController.prototype.onDidSelectItem = function (item) {
        if (item) {
            var suggestion = item.suggestion, position = item.position;
            var columnDelta = this.editor.getPosition().column - position.column;
            if (Array.isArray(suggestion.additionalTextEdits)) {
                this.editor.pushUndoStop();
                this.editor.executeEdits('suggestController.additionalTextEdits', suggestion.additionalTextEdits.map(function (edit) { return editOperation_1.EditOperation.replace(range_1.Range.lift(edit.range), edit.text); }));
                this.editor.pushUndoStop();
            }
            if (suggestion.snippetType === 'textmate') {
                snippetController_1.SnippetController.get(this.editor).insertSnippet(suggestion.insertText, suggestion.overwriteBefore + columnDelta, suggestion.overwriteAfter);
            }
            else {
                snippetController_1.SnippetController.get(this.editor).run(snippet_1.CodeSnippet.fromInternal(suggestion.insertText), suggestion.overwriteBefore + columnDelta, suggestion.overwriteAfter);
            }
            if (suggestion.command) {
                (_a = this.commandService).executeCommand.apply(_a, [suggestion.command.id].concat(suggestion.command.arguments)).done(undefined, errors_1.onUnexpectedError);
            }
            this.telemetryService.publicLog('suggestSnippetInsert', __assign({}, this.editor.getTelemetryData(), { suggestionType: suggestion.type }));
        }
        this.model.cancel();
        var _a;
    };
    SuggestController.prototype.triggerSuggest = function () {
        this.model.trigger(false, false);
        this.editor.revealLine(this.editor.getPosition().lineNumber);
        this.editor.focus();
    };
    SuggestController.prototype.acceptSelectedSuggestion = function () {
        if (this.widget) {
            var item = this.widget.getFocusedItem();
            this.onDidSelectItem(item);
        }
    };
    SuggestController.prototype.cancelSuggestWidget = function () {
        if (this.widget) {
            this.model.cancel();
            this.widget.hideDetailsOrHideWidget();
        }
    };
    SuggestController.prototype.selectNextSuggestion = function () {
        if (this.widget) {
            this.widget.selectNext();
        }
    };
    SuggestController.prototype.selectNextPageSuggestion = function () {
        if (this.widget) {
            this.widget.selectNextPage();
        }
    };
    SuggestController.prototype.selectPrevSuggestion = function () {
        if (this.widget) {
            this.widget.selectPrevious();
        }
    };
    SuggestController.prototype.selectPrevPageSuggestion = function () {
        if (this.widget) {
            this.widget.selectPreviousPage();
        }
    };
    SuggestController.prototype.toggleSuggestionDetails = function () {
        if (this.widget) {
            this.widget.toggleDetails();
        }
    };
    return SuggestController;
}());
SuggestController.ID = 'editor.contrib.suggestController';
SuggestController = SuggestController_1 = __decorate([
    editorBrowserExtensions_1.editorContribution,
    __param(1, commands_1.ICommandService),
    __param(2, telemetry_1.ITelemetryService),
    __param(3, contextkey_1.IContextKeyService),
    __param(4, instantiation_1.IInstantiationService)
], SuggestController);
exports.SuggestController = SuggestController;
var TriggerSuggestAction = (function (_super) {
    __extends(TriggerSuggestAction, _super);
    function TriggerSuggestAction() {
        return _super.call(this, {
            id: 'editor.action.triggerSuggest',
            label: nls.localize('suggest.trigger.label', "Trigger Suggest"),
            alias: 'Trigger Suggest',
            precondition: contextkey_1.ContextKeyExpr.and(editorCommon_1.EditorContextKeys.Writable, editorCommon_1.ModeContextKeys.hasCompletionItemProvider),
            kbOpts: {
                kbExpr: editorCommon_1.EditorContextKeys.TextFocus,
                primary: 2048 /* CtrlCmd */ | 10 /* Space */,
                mac: { primary: 256 /* WinCtrl */ | 10 /* Space */ }
            }
        }) || this;
    }
    TriggerSuggestAction.prototype.run = function (accessor, editor) {
        var controller = SuggestController.get(editor);
        if (!controller) {
            return;
        }
        controller.triggerSuggest();
    };
    return TriggerSuggestAction;
}(editorCommonExtensions_1.EditorAction));
TriggerSuggestAction = __decorate([
    editorCommonExtensions_1.editorAction
], TriggerSuggestAction);
exports.TriggerSuggestAction = TriggerSuggestAction;
var weight = editorCommonExtensions_1.CommonEditorRegistry.commandWeight(90);
var SuggestCommand = editorCommonExtensions_1.EditorCommand.bindToContribution(SuggestController.get);
editorCommonExtensions_1.CommonEditorRegistry.registerEditorCommand(new SuggestCommand({
    id: 'acceptSelectedSuggestion',
    precondition: suggest_1.Context.Visible,
    handler: function (x) { return x.acceptSelectedSuggestion(); },
    kbOpts: {
        weight: weight,
        kbExpr: editorCommon_1.EditorContextKeys.TextFocus,
        primary: 2 /* Tab */
    }
}));
editorCommonExtensions_1.CommonEditorRegistry.registerEditorCommand(new SuggestCommand({
    id: 'acceptSelectedSuggestionOnEnter',
    precondition: suggest_1.Context.Visible,
    handler: function (x) { return x.acceptSelectedSuggestion(); },
    kbOpts: {
        weight: weight,
        kbExpr: contextkey_1.ContextKeyExpr.and(editorCommon_1.EditorContextKeys.TextFocus, suggest_1.Context.AcceptSuggestionsOnEnter),
        primary: 3 /* Enter */
    }
}));
editorCommonExtensions_1.CommonEditorRegistry.registerEditorCommand(new SuggestCommand({
    id: 'hideSuggestWidget',
    precondition: suggest_1.Context.Visible,
    handler: function (x) { return x.cancelSuggestWidget(); },
    kbOpts: {
        weight: weight,
        kbExpr: editorCommon_1.EditorContextKeys.TextFocus,
        primary: 9 /* Escape */,
        secondary: [1024 /* Shift */ | 9 /* Escape */]
    }
}));
editorCommonExtensions_1.CommonEditorRegistry.registerEditorCommand(new SuggestCommand({
    id: 'selectNextSuggestion',
    precondition: contextkey_1.ContextKeyExpr.and(suggest_1.Context.Visible, suggest_1.Context.MultipleSuggestions),
    handler: function (c) { return c.selectNextSuggestion(); },
    kbOpts: {
        weight: weight,
        kbExpr: editorCommon_1.EditorContextKeys.TextFocus,
        primary: 18 /* DownArrow */,
        secondary: [512 /* Alt */ | 18 /* DownArrow */],
        mac: { primary: 18 /* DownArrow */, secondary: [512 /* Alt */ | 18 /* DownArrow */, 256 /* WinCtrl */ | 44 /* KEY_N */] }
    }
}));
editorCommonExtensions_1.CommonEditorRegistry.registerEditorCommand(new SuggestCommand({
    id: 'selectNextPageSuggestion',
    precondition: contextkey_1.ContextKeyExpr.and(suggest_1.Context.Visible, suggest_1.Context.MultipleSuggestions),
    handler: function (c) { return c.selectNextPageSuggestion(); },
    kbOpts: {
        weight: weight,
        kbExpr: editorCommon_1.EditorContextKeys.TextFocus,
        primary: 12 /* PageDown */,
        secondary: [512 /* Alt */ | 12 /* PageDown */]
    }
}));
editorCommonExtensions_1.CommonEditorRegistry.registerEditorCommand(new SuggestCommand({
    id: 'selectPrevSuggestion',
    precondition: contextkey_1.ContextKeyExpr.and(suggest_1.Context.Visible, suggest_1.Context.MultipleSuggestions),
    handler: function (c) { return c.selectPrevSuggestion(); },
    kbOpts: {
        weight: weight,
        kbExpr: editorCommon_1.EditorContextKeys.TextFocus,
        primary: 16 /* UpArrow */,
        secondary: [512 /* Alt */ | 16 /* UpArrow */],
        mac: { primary: 16 /* UpArrow */, secondary: [512 /* Alt */ | 16 /* UpArrow */, 256 /* WinCtrl */ | 46 /* KEY_P */] }
    }
}));
editorCommonExtensions_1.CommonEditorRegistry.registerEditorCommand(new SuggestCommand({
    id: 'selectPrevPageSuggestion',
    precondition: contextkey_1.ContextKeyExpr.and(suggest_1.Context.Visible, suggest_1.Context.MultipleSuggestions),
    handler: function (c) { return c.selectPrevPageSuggestion(); },
    kbOpts: {
        weight: weight,
        kbExpr: editorCommon_1.EditorContextKeys.TextFocus,
        primary: 11 /* PageUp */,
        secondary: [512 /* Alt */ | 11 /* PageUp */]
    }
}));
editorCommonExtensions_1.CommonEditorRegistry.registerEditorCommand(new SuggestCommand({
    id: 'toggleSuggestionDetails',
    precondition: suggest_1.Context.Visible,
    handler: function (x) { return x.toggleSuggestionDetails(); },
    kbOpts: {
        weight: weight,
        kbExpr: editorCommon_1.EditorContextKeys.TextFocus,
        primary: 2048 /* CtrlCmd */ | 10 /* Space */,
        mac: { primary: 256 /* WinCtrl */ | 10 /* Space */ }
    }
}));
var SuggestController_1;
