// vscode/src/vs/platform/extensionManagement/common/extensionTelemetry.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
function getLocalExtensionTelemetryData(extension) {
    return {
        id: extension.manifest.publisher + "." + extension.manifest.name,
        name: extension.manifest.name,
        galleryId: extension.metadata ? extension.metadata.id : null,
        publisherId: extension.metadata ? extension.metadata.publisherId : null,
        publisherName: extension.manifest.publisher,
        publisherDisplayName: extension.metadata ? extension.metadata.publisherDisplayName : null
    };
}
exports.getLocalExtensionTelemetryData = getLocalExtensionTelemetryData;
function getGalleryExtensionTelemetryData(extension) {
    return {
        id: extension.publisher + "." + extension.name,
        name: extension.name,
        galleryId: extension.id,
        publisherId: extension.publisherId,
        publisherName: extension.publisher,
        publisherDisplayName: extension.publisherDisplayName
    };
}
exports.getGalleryExtensionTelemetryData = getGalleryExtensionTelemetryData;
