// vscode/src/vs/platform/extensionManagement/common/extensionManagement.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var nls = require('vs/nls');
var instantiation_1 = require('vs/platform/instantiation/common/instantiation');
exports.EXTENSION_IDENTIFIER_PATTERN = '^[a-z0-9A-Z][a-z0-9\-A-Z]*\\.[a-z0-9A-Z][a-z0-9\-A-Z]*$';
(function (LocalExtensionType) {
    LocalExtensionType[LocalExtensionType["System"] = 0] = "System";
    LocalExtensionType[LocalExtensionType["User"] = 1] = "User";
})(exports.LocalExtensionType || (exports.LocalExtensionType = {}));
var LocalExtensionType = exports.LocalExtensionType;
exports.IExtensionManagementService = instantiation_1.createDecorator('extensionManagementService');
exports.IExtensionGalleryService = instantiation_1.createDecorator('extensionGalleryService');
(function (SortBy) {
    SortBy[SortBy["NoneOrRelevance"] = 0] = "NoneOrRelevance";
    SortBy[SortBy["LastUpdatedDate"] = 1] = "LastUpdatedDate";
    SortBy[SortBy["Title"] = 2] = "Title";
    SortBy[SortBy["PublisherName"] = 3] = "PublisherName";
    SortBy[SortBy["InstallCount"] = 4] = "InstallCount";
    SortBy[SortBy["PublishedDate"] = 5] = "PublishedDate";
    SortBy[SortBy["AverageRating"] = 6] = "AverageRating";
})(exports.SortBy || (exports.SortBy = {}));
var SortBy = exports.SortBy;
(function (SortOrder) {
    SortOrder[SortOrder["Default"] = 0] = "Default";
    SortOrder[SortOrder["Ascending"] = 1] = "Ascending";
    SortOrder[SortOrder["Descending"] = 2] = "Descending";
})(exports.SortOrder || (exports.SortOrder = {}));
var SortOrder = exports.SortOrder;
exports.IExtensionTipsService = instantiation_1.createDecorator('extensionTipsService');
exports.ExtensionsLabel = nls.localize('extensions', "Extensions");
exports.ExtensionsChannelId = 'extensions';
