// glamor/lib/CSSPropertyOperations.js
"use strict";

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

!function (f) {
  if ("object" == (typeof exports === "undefined" ? "undefined" : _typeof(exports)) && "undefined" != typeof module) module.exports = f();else if ("function" == typeof define && define.amd) define([], f);else {
    var g;g = "undefined" != typeof window ? window : "undefined" != typeof global ? global : "undefined" != typeof self ? self : this, g.CSSOps = f();
  }
}(function () {
  return function e(t, n, r) {
    function s(o, u) {
      if (!n[o]) {
        if (!t[o]) {
          var a = "function" == typeof require && require;if (!u && a) return a(o, !0);if (i) return i(o, !0);var f = new Error("Cannot find module '" + o + "'");throw f.code = "MODULE_NOT_FOUND", f;
        }var l = n[o] = { exports: {} };t[o][0].call(l.exports, function (e) {
          var n = t[o][1][e];return s(n ? n : e);
        }, l, l.exports, e, t, n, r);
      }return n[o].exports;
    }for (var i = "function" == typeof require && require, o = 0; o < r.length; o++) {
      s(r[o]);
    }return s;
  }({ 1: [function (_dereq_, module, exports) {
      module.exports = _dereq_("react/lib/CSSPropertyOperations");
    }, { "react/lib/CSSPropertyOperations": 14 }], 2: [function (_dereq_, module, exports) {
      "use strict";
      var canUseDOM = !("undefined" == typeof window || !window.document || !window.document.createElement),
          ExecutionEnvironment = { canUseDOM: canUseDOM, canUseWorkers: "undefined" != typeof Worker, canUseEventListeners: canUseDOM && !(!window.addEventListener && !window.attachEvent), canUseViewport: canUseDOM && !!window.screen, isInWorker: !canUseDOM };module.exports = ExecutionEnvironment;
    }, {}], 3: [function (_dereq_, module, exports) {
      "use strict";
      function camelize(string) {
        return string.replace(_hyphenPattern, function (_, character) {
          return character.toUpperCase();
        });
      }var _hyphenPattern = /-(.)/g;module.exports = camelize;
    }, {}], 4: [function (_dereq_, module, exports) {
      "use strict";
      function camelizeStyleName(string) {
        return camelize(string.replace(msPattern, "ms-"));
      }var camelize = _dereq_("./camelize"),
          msPattern = /^-ms-/;module.exports = camelizeStyleName;
    }, { "./camelize": 3 }], 5: [function (_dereq_, module, exports) {
      "use strict";
      function makeEmptyFunction(arg) {
        return function () {
          return arg;
        };
      }var emptyFunction = function emptyFunction() {};emptyFunction.thatReturns = makeEmptyFunction, emptyFunction.thatReturnsFalse = makeEmptyFunction(!1), emptyFunction.thatReturnsTrue = makeEmptyFunction(!0), emptyFunction.thatReturnsNull = makeEmptyFunction(null), emptyFunction.thatReturnsThis = function () {
        return this;
      }, emptyFunction.thatReturnsArgument = function (arg) {
        return arg;
      }, module.exports = emptyFunction;
    }, {}], 6: [function (_dereq_, module, exports) {
      "use strict";
      function hyphenate(string) {
        return string.replace(_uppercasePattern, "-$1").toLowerCase();
      }var _uppercasePattern = /([A-Z])/g;module.exports = hyphenate;
    }, {}], 7: [function (_dereq_, module, exports) {
      "use strict";
      function hyphenateStyleName(string) {
        return hyphenate(string).replace(msPattern, "-ms-");
      }var hyphenate = _dereq_("./hyphenate"),
          msPattern = /^ms-/;module.exports = hyphenateStyleName;
    }, { "./hyphenate": 6 }], 8: [function (_dereq_, module, exports) {
      "use strict";
      function invariant(condition, format, a, b, c, d, e, f) {
        if (!condition) {
          var error;if (void 0 === format) error = new Error("Minified exception occurred; use the non-minified dev environment for the full error message and additional helpful warnings.");else {
            var args = [a, b, c, d, e, f],
                argIndex = 0;error = new Error(format.replace(/%s/g, function () {
              return args[argIndex++];
            })), error.name = "Invariant Violation";
          }throw error.framesToPop = 1, error;
        }
      }module.exports = invariant;
    }, {}], 9: [function (_dereq_, module, exports) {
      "use strict";
      function memoizeStringOnly(callback) {
        var cache = {};return function (string) {
          return cache.hasOwnProperty(string) || (cache[string] = callback.call(this, string)), cache[string];
        };
      }module.exports = memoizeStringOnly;
    }, {}], 10: [function (_dereq_, module, exports) {
      "use strict";
      var performance,
          ExecutionEnvironment = _dereq_("./ExecutionEnvironment");ExecutionEnvironment.canUseDOM && (performance = window.performance || window.msPerformance || window.webkitPerformance), module.exports = performance || {};
    }, { "./ExecutionEnvironment": 2 }], 11: [function (_dereq_, module, exports) {
      "use strict";
      var performanceNow,
          performance = _dereq_("./performance");performanceNow = performance.now ? function () {
        return performance.now();
      } : function () {
        return Date.now();
      }, module.exports = performanceNow;
    }, { "./performance": 10 }], 12: [function (_dereq_, module, exports) {
      "use strict";
      var emptyFunction = _dereq_("./emptyFunction"),
          warning = emptyFunction;module.exports = warning;
    }, { "./emptyFunction": 5 }], 13: [function (_dereq_, module, exports) {
      "use strict";
      function prefixKey(prefix, key) {
        return prefix + key.charAt(0).toUpperCase() + key.substring(1);
      }var isUnitlessNumber = { animationIterationCount: !0, borderImageOutset: !0, borderImageSlice: !0, borderImageWidth: !0, boxFlex: !0, boxFlexGroup: !0, boxOrdinalGroup: !0, columnCount: !0, flex: !0, flexGrow: !0, flexPositive: !0, flexShrink: !0, flexNegative: !0, flexOrder: !0, gridRow: !0, gridColumn: !0, fontWeight: !0, lineClamp: !0, lineHeight: !0, opacity: !0, order: !0, orphans: !0, tabSize: !0, widows: !0, zIndex: !0, zoom: !0, fillOpacity: !0, floodOpacity: !0, stopOpacity: !0, strokeDasharray: !0, strokeDashoffset: !0, strokeMiterlimit: !0, strokeOpacity: !0, strokeWidth: !0 },
          prefixes = ["Webkit", "ms", "Moz", "O"];Object.keys(isUnitlessNumber).forEach(function (prop) {
        prefixes.forEach(function (prefix) {
          isUnitlessNumber[prefixKey(prefix, prop)] = isUnitlessNumber[prop];
        });
      });var shorthandPropertyExpansions = { background: { backgroundAttachment: !0, backgroundColor: !0, backgroundImage: !0, backgroundPositionX: !0, backgroundPositionY: !0, backgroundRepeat: !0 }, backgroundPosition: { backgroundPositionX: !0, backgroundPositionY: !0 }, border: { borderWidth: !0, borderStyle: !0, borderColor: !0 }, borderBottom: { borderBottomWidth: !0, borderBottomStyle: !0, borderBottomColor: !0 }, borderLeft: { borderLeftWidth: !0, borderLeftStyle: !0, borderLeftColor: !0 }, borderRight: { borderRightWidth: !0, borderRightStyle: !0, borderRightColor: !0 }, borderTop: { borderTopWidth: !0, borderTopStyle: !0, borderTopColor: !0 }, font: { fontStyle: !0, fontVariant: !0, fontWeight: !0, fontSize: !0, lineHeight: !0, fontFamily: !0 }, outline: { outlineWidth: !0, outlineStyle: !0, outlineColor: !0 } },
          CSSProperty = { isUnitlessNumber: isUnitlessNumber, shorthandPropertyExpansions: shorthandPropertyExpansions };module.exports = CSSProperty;
    }, {}], 14: [function (_dereq_, module, exports) {
      "use strict";
      var CSSProperty = _dereq_("./CSSProperty"),
          ExecutionEnvironment = _dereq_("fbjs/lib/ExecutionEnvironment"),
          dangerousStyleValue = (_dereq_("./ReactInstrumentation"), _dereq_("fbjs/lib/camelizeStyleName"), _dereq_("./dangerousStyleValue")),
          hyphenateStyleName = _dereq_("fbjs/lib/hyphenateStyleName"),
          memoizeStringOnly = _dereq_("fbjs/lib/memoizeStringOnly"),
          processStyleName = (_dereq_("fbjs/lib/warning"), memoizeStringOnly(function (styleName) {
        return hyphenateStyleName(styleName);
      })),
          hasShorthandPropertyBug = !1,
          styleFloatAccessor = "cssFloat";if (ExecutionEnvironment.canUseDOM) {
        var tempStyle = document.createElement("div").style;try {
          tempStyle.font = "";
        } catch (e) {
          hasShorthandPropertyBug = !0;
        }void 0 === document.documentElement.style.cssFloat && (styleFloatAccessor = "styleFloat");
      }var CSSPropertyOperations = { createMarkupForStyles: function createMarkupForStyles(styles, component) {
          var serialized = "";for (var styleName in styles) {
            if (styles.hasOwnProperty(styleName)) {
              var styleValue = styles[styleName];null != styleValue && (serialized += processStyleName(styleName) + ":", serialized += dangerousStyleValue(styleName, styleValue, component) + ";");
            }
          }return serialized || null;
        }, setValueForStyles: function setValueForStyles(node, styles, component) {
          var style = node.style;for (var styleName in styles) {
            if (styles.hasOwnProperty(styleName)) {
              var styleValue = dangerousStyleValue(styleName, styles[styleName], component);if ("float" !== styleName && "cssFloat" !== styleName || (styleName = styleFloatAccessor), styleValue) style[styleName] = styleValue;else {
                var expansion = hasShorthandPropertyBug && CSSProperty.shorthandPropertyExpansions[styleName];if (expansion) for (var individualStyleName in expansion) {
                  style[individualStyleName] = "";
                } else style[styleName] = "";
              }
            }
          }
        } };module.exports = CSSPropertyOperations;
    }, { "./CSSProperty": 13, "./ReactInstrumentation": 20, "./dangerousStyleValue": 22, "fbjs/lib/ExecutionEnvironment": 2, "fbjs/lib/camelizeStyleName": 4, "fbjs/lib/hyphenateStyleName": 7, "fbjs/lib/memoizeStringOnly": 9, "fbjs/lib/warning": 12 }], 15: [function (_dereq_, module, exports) {
      "use strict";
      function handleElement(debugID, element) {
        if (null != element && void 0 !== element._shadowChildren && element._shadowChildren !== element.props.children) {
          var isMutated = !1;if (Array.isArray(element._shadowChildren)) if (element._shadowChildren.length === element.props.children.length) for (var i = 0; i < element._shadowChildren.length; i++) {
            element._shadowChildren[i] !== element.props.children[i] && (isMutated = !0);
          } else isMutated = !0;!Array.isArray(element._shadowChildren) || isMutated;
        }
      }var ReactComponentTreeHook = _dereq_("./ReactComponentTreeHook"),
          ReactChildrenMutationWarningHook = (_dereq_("fbjs/lib/warning"), { onMountComponent: function onMountComponent(debugID) {
          handleElement(debugID, ReactComponentTreeHook.getElement(debugID));
        }, onUpdateComponent: function onUpdateComponent(debugID) {
          handleElement(debugID, ReactComponentTreeHook.getElement(debugID));
        } });module.exports = ReactChildrenMutationWarningHook;
    }, { "./ReactComponentTreeHook": 16, "fbjs/lib/warning": 12 }], 16: [function (_dereq_, module, exports) {
      "use strict";
      function isNative(fn) {
        var funcToString = Function.prototype.toString,
            hasOwnProperty = Object.prototype.hasOwnProperty,
            reIsNative = RegExp("^" + funcToString.call(hasOwnProperty).replace(/[\\^$.*+?()[\]{}|]/g, "\\$&").replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$");try {
          var source = funcToString.call(fn);return reIsNative.test(source);
        } catch (err) {
          return !1;
        }
      }function getKeyFromID(id) {
        return "." + id;
      }function getIDFromKey(key) {
        return parseInt(key.substr(1), 10);
      }function get(id) {
        if (canUseCollections) return itemMap.get(id);var key = getKeyFromID(id);return itemByKey[key];
      }function remove(id) {
        if (canUseCollections) itemMap.delete(id);else {
          var key = getKeyFromID(id);delete itemByKey[key];
        }
      }function create(id, element, parentID) {
        var item = { element: element, parentID: parentID, text: null, childIDs: [], isMounted: !1, updateCount: 0 };if (canUseCollections) itemMap.set(id, item);else {
          var key = getKeyFromID(id);itemByKey[key] = item;
        }
      }function addRoot(id) {
        if (canUseCollections) rootIDSet.add(id);else {
          var key = getKeyFromID(id);rootByKey[key] = !0;
        }
      }function removeRoot(id) {
        if (canUseCollections) rootIDSet.delete(id);else {
          var key = getKeyFromID(id);delete rootByKey[key];
        }
      }function getRegisteredIDs() {
        return canUseCollections ? Array.from(itemMap.keys()) : Object.keys(itemByKey).map(getIDFromKey);
      }function getRootIDs() {
        return canUseCollections ? Array.from(rootIDSet.keys()) : Object.keys(rootByKey).map(getIDFromKey);
      }function purgeDeep(id) {
        var item = get(id);if (item) {
          var childIDs = item.childIDs;remove(id), childIDs.forEach(purgeDeep);
        }
      }function describeComponentFrame(name, source, ownerName) {
        return "\n    in " + name + (source ? " (at " + source.fileName.replace(/^.*[\\\/]/, "") + ":" + source.lineNumber + ")" : ownerName ? " (created by " + ownerName + ")" : "");
      }function _getDisplayName(element) {
        return null == element ? "#empty" : "string" == typeof element || "number" == typeof element ? "#text" : "string" == typeof element.type ? element.type : element.type.displayName || element.type.name || "Unknown";
      }function describeID(id) {
        var ownerName,
            name = ReactComponentTreeHook.getDisplayName(id),
            element = ReactComponentTreeHook.getElement(id),
            ownerID = ReactComponentTreeHook.getOwnerID(id);return ownerID && (ownerName = ReactComponentTreeHook.getDisplayName(ownerID)), describeComponentFrame(name, element && element._source, ownerName);
      }var itemMap,
          rootIDSet,
          itemByKey,
          rootByKey,
          _prodInvariant = _dereq_("./reactProdInvariant"),
          ReactCurrentOwner = _dereq_("./ReactCurrentOwner"),
          canUseCollections = (_dereq_("fbjs/lib/invariant"), _dereq_("fbjs/lib/warning"), "function" == typeof Array.from && "function" == typeof Map && isNative(Map) && null != Map.prototype && "function" == typeof Map.prototype.keys && isNative(Map.prototype.keys) && "function" == typeof Set && isNative(Set) && null != Set.prototype && "function" == typeof Set.prototype.keys && isNative(Set.prototype.keys));canUseCollections ? (itemMap = new Map(), rootIDSet = new Set()) : (itemByKey = {}, rootByKey = {});var unmountedIDs = [],
          ReactComponentTreeHook = { onSetChildren: function onSetChildren(id, nextChildIDs) {
          var item = get(id);item.childIDs = nextChildIDs;for (var i = 0; i < nextChildIDs.length; i++) {
            var nextChildID = nextChildIDs[i],
                nextChild = get(nextChildID);nextChild ? void 0 : _prodInvariant("140"), null == nextChild.childIDs && "object" == _typeof(nextChild.element) && null != nextChild.element ? _prodInvariant("141") : void 0, nextChild.isMounted ? void 0 : _prodInvariant("71"), null == nextChild.parentID && (nextChild.parentID = id), nextChild.parentID !== id ? _prodInvariant("142", nextChildID, nextChild.parentID, id) : void 0;
          }
        }, onBeforeMountComponent: function onBeforeMountComponent(id, element, parentID) {
          create(id, element, parentID);
        }, onBeforeUpdateComponent: function onBeforeUpdateComponent(id, element) {
          var item = get(id);item && item.isMounted && (item.element = element);
        }, onMountComponent: function onMountComponent(id) {
          var item = get(id);item.isMounted = !0;var isRoot = 0 === item.parentID;isRoot && addRoot(id);
        }, onUpdateComponent: function onUpdateComponent(id) {
          var item = get(id);item && item.isMounted && item.updateCount++;
        }, onUnmountComponent: function onUnmountComponent(id) {
          var item = get(id);if (item) {
            item.isMounted = !1;var isRoot = 0 === item.parentID;isRoot && removeRoot(id);
          }unmountedIDs.push(id);
        }, purgeUnmountedComponents: function purgeUnmountedComponents() {
          if (!ReactComponentTreeHook._preventPurging) {
            for (var i = 0; i < unmountedIDs.length; i++) {
              var id = unmountedIDs[i];purgeDeep(id);
            }unmountedIDs.length = 0;
          }
        }, isMounted: function isMounted(id) {
          var item = get(id);return !!item && item.isMounted;
        }, getCurrentStackAddendum: function getCurrentStackAddendum(topElement) {
          var info = "";if (topElement) {
            var type = topElement.type,
                name = "function" == typeof type ? type.displayName || type.name : type,
                owner = topElement._owner;info += describeComponentFrame(name || "Unknown", topElement._source, owner && owner.getName());
          }var currentOwner = ReactCurrentOwner.current,
              id = currentOwner && currentOwner._debugID;return info += ReactComponentTreeHook.getStackAddendumByID(id);
        }, getStackAddendumByID: function getStackAddendumByID(id) {
          for (var info = ""; id;) {
            info += describeID(id), id = ReactComponentTreeHook.getParentID(id);
          }return info;
        }, getChildIDs: function getChildIDs(id) {
          var item = get(id);return item ? item.childIDs : [];
        }, getDisplayName: function getDisplayName(id) {
          var element = ReactComponentTreeHook.getElement(id);return element ? _getDisplayName(element) : null;
        }, getElement: function getElement(id) {
          var item = get(id);return item ? item.element : null;
        }, getOwnerID: function getOwnerID(id) {
          var element = ReactComponentTreeHook.getElement(id);return element && element._owner ? element._owner._debugID : null;
        }, getParentID: function getParentID(id) {
          var item = get(id);return item ? item.parentID : null;
        }, getSource: function getSource(id) {
          var item = get(id),
              element = item ? item.element : null,
              source = null != element ? element._source : null;return source;
        }, getText: function getText(id) {
          var element = ReactComponentTreeHook.getElement(id);return "string" == typeof element ? element : "number" == typeof element ? "" + element : null;
        }, getUpdateCount: function getUpdateCount(id) {
          var item = get(id);return item ? item.updateCount : 0;
        }, getRegisteredIDs: getRegisteredIDs, getRootIDs: getRootIDs };module.exports = ReactComponentTreeHook;
    }, { "./ReactCurrentOwner": 17, "./reactProdInvariant": 23, "fbjs/lib/invariant": 8, "fbjs/lib/warning": 12 }], 17: [function (_dereq_, module, exports) {
      "use strict";
      var ReactCurrentOwner = { current: null };module.exports = ReactCurrentOwner;
    }, {}], 18: [function (_dereq_, module, exports) {
      "use strict";
      function callHook(event, fn, context, arg1, arg2, arg3, arg4, arg5) {
        try {
          fn.call(context, arg1, arg2, arg3, arg4, arg5);
        } catch (e) {
          didHookThrowForEvent[event] = !0;
        }
      }function emitEvent(event, arg1, arg2, arg3, arg4, arg5) {
        for (var i = 0; i < hooks.length; i++) {
          var hook = hooks[i],
              fn = hook[event];fn && callHook(event, fn, hook, arg1, arg2, arg3, arg4, arg5);
        }
      }function clearHistory() {
        ReactComponentTreeHook.purgeUnmountedComponents(), ReactHostOperationHistoryHook.clearHistory();
      }function getTreeSnapshot(registeredIDs) {
        return registeredIDs.reduce(function (tree, id) {
          var ownerID = ReactComponentTreeHook.getOwnerID(id),
              parentID = ReactComponentTreeHook.getParentID(id);return tree[id] = { displayName: ReactComponentTreeHook.getDisplayName(id), text: ReactComponentTreeHook.getText(id), updateCount: ReactComponentTreeHook.getUpdateCount(id), childIDs: ReactComponentTreeHook.getChildIDs(id), ownerID: ownerID || ReactComponentTreeHook.getOwnerID(parentID), parentID: parentID }, tree;
        }, {});
      }function resetMeasurements() {
        var previousStartTime = currentFlushStartTime,
            previousMeasurements = currentFlushMeasurements || [],
            previousOperations = ReactHostOperationHistoryHook.getHistory();if (0 === currentFlushNesting) return currentFlushStartTime = null, currentFlushMeasurements = null, void clearHistory();if (previousMeasurements.length || previousOperations.length) {
          var registeredIDs = ReactComponentTreeHook.getRegisteredIDs();flushHistory.push({ duration: performanceNow() - previousStartTime, measurements: previousMeasurements || [], operations: previousOperations || [], treeSnapshot: getTreeSnapshot(registeredIDs) });
        }clearHistory(), currentFlushStartTime = performanceNow(), currentFlushMeasurements = [];
      }function checkDebugID(debugID) {
        var allowRoot = !(arguments.length <= 1 || void 0 === arguments[1]) && arguments[1];
      }function beginLifeCycleTimer(debugID, timerType) {
        0 !== currentFlushNesting && (currentTimerType && !lifeCycleTimerHasWarned && (lifeCycleTimerHasWarned = !0), currentTimerStartTime = performanceNow(), currentTimerNestedFlushDuration = 0, currentTimerDebugID = debugID, currentTimerType = timerType);
      }function endLifeCycleTimer(debugID, timerType) {
        0 !== currentFlushNesting && (currentTimerType === timerType || lifeCycleTimerHasWarned || (lifeCycleTimerHasWarned = !0), _isProfiling && currentFlushMeasurements.push({ timerType: timerType, instanceID: debugID, duration: performanceNow() - currentTimerStartTime - currentTimerNestedFlushDuration }), currentTimerStartTime = null, currentTimerNestedFlushDuration = null, currentTimerDebugID = null, currentTimerType = null);
      }function pauseCurrentLifeCycleTimer() {
        var currentTimer = { startTime: currentTimerStartTime, nestedFlushStartTime: performanceNow(), debugID: currentTimerDebugID, timerType: currentTimerType };lifeCycleTimerStack.push(currentTimer), currentTimerStartTime = null, currentTimerNestedFlushDuration = null, currentTimerDebugID = null, currentTimerType = null;
      }function resumeCurrentLifeCycleTimer() {
        var _lifeCycleTimerStack$ = lifeCycleTimerStack.pop(),
            startTime = _lifeCycleTimerStack$.startTime,
            nestedFlushStartTime = _lifeCycleTimerStack$.nestedFlushStartTime,
            debugID = _lifeCycleTimerStack$.debugID,
            timerType = _lifeCycleTimerStack$.timerType,
            nestedFlushDuration = performanceNow() - nestedFlushStartTime;currentTimerStartTime = startTime, currentTimerNestedFlushDuration += nestedFlushDuration, currentTimerDebugID = debugID, currentTimerType = timerType;
      }var ReactInvalidSetStateWarningHook = _dereq_("./ReactInvalidSetStateWarningHook"),
          ReactHostOperationHistoryHook = _dereq_("./ReactHostOperationHistoryHook"),
          ReactComponentTreeHook = _dereq_("./ReactComponentTreeHook"),
          ReactChildrenMutationWarningHook = _dereq_("./ReactChildrenMutationWarningHook"),
          ExecutionEnvironment = _dereq_("fbjs/lib/ExecutionEnvironment"),
          performanceNow = _dereq_("fbjs/lib/performanceNow"),
          hooks = (_dereq_("fbjs/lib/warning"), []),
          didHookThrowForEvent = {},
          _isProfiling = !1,
          flushHistory = [],
          lifeCycleTimerStack = [],
          currentFlushNesting = 0,
          currentFlushMeasurements = null,
          currentFlushStartTime = null,
          currentTimerDebugID = null,
          currentTimerStartTime = null,
          currentTimerNestedFlushDuration = null,
          currentTimerType = null,
          lifeCycleTimerHasWarned = !1,
          ReactDebugTool = { addHook: function addHook(hook) {
          hooks.push(hook);
        }, removeHook: function removeHook(hook) {
          for (var i = 0; i < hooks.length; i++) {
            hooks[i] === hook && (hooks.splice(i, 1), i--);
          }
        }, isProfiling: function isProfiling() {
          return _isProfiling;
        }, beginProfiling: function beginProfiling() {
          _isProfiling || (_isProfiling = !0, flushHistory.length = 0, resetMeasurements(), ReactDebugTool.addHook(ReactHostOperationHistoryHook));
        }, endProfiling: function endProfiling() {
          _isProfiling && (_isProfiling = !1, resetMeasurements(), ReactDebugTool.removeHook(ReactHostOperationHistoryHook));
        }, getFlushHistory: function getFlushHistory() {
          return flushHistory;
        }, onBeginFlush: function onBeginFlush() {
          currentFlushNesting++, resetMeasurements(), pauseCurrentLifeCycleTimer(), emitEvent("onBeginFlush");
        }, onEndFlush: function onEndFlush() {
          resetMeasurements(), currentFlushNesting--, resumeCurrentLifeCycleTimer(), emitEvent("onEndFlush");
        }, onBeginLifeCycleTimer: function onBeginLifeCycleTimer(debugID, timerType) {
          checkDebugID(debugID), emitEvent("onBeginLifeCycleTimer", debugID, timerType), beginLifeCycleTimer(debugID, timerType);
        }, onEndLifeCycleTimer: function onEndLifeCycleTimer(debugID, timerType) {
          checkDebugID(debugID), endLifeCycleTimer(debugID, timerType), emitEvent("onEndLifeCycleTimer", debugID, timerType);
        }, onBeginProcessingChildContext: function onBeginProcessingChildContext() {
          emitEvent("onBeginProcessingChildContext");
        }, onEndProcessingChildContext: function onEndProcessingChildContext() {
          emitEvent("onEndProcessingChildContext");
        }, onHostOperation: function onHostOperation(debugID, type, payload) {
          checkDebugID(debugID), emitEvent("onHostOperation", debugID, type, payload);
        }, onSetState: function onSetState() {
          emitEvent("onSetState");
        }, onSetChildren: function onSetChildren(debugID, childDebugIDs) {
          checkDebugID(debugID), childDebugIDs.forEach(checkDebugID), emitEvent("onSetChildren", debugID, childDebugIDs);
        }, onBeforeMountComponent: function onBeforeMountComponent(debugID, element, parentDebugID) {
          checkDebugID(debugID), checkDebugID(parentDebugID, !0), emitEvent("onBeforeMountComponent", debugID, element, parentDebugID);
        }, onMountComponent: function onMountComponent(debugID) {
          checkDebugID(debugID), emitEvent("onMountComponent", debugID);
        }, onBeforeUpdateComponent: function onBeforeUpdateComponent(debugID, element) {
          checkDebugID(debugID), emitEvent("onBeforeUpdateComponent", debugID, element);
        }, onUpdateComponent: function onUpdateComponent(debugID) {
          checkDebugID(debugID), emitEvent("onUpdateComponent", debugID);
        }, onBeforeUnmountComponent: function onBeforeUnmountComponent(debugID) {
          checkDebugID(debugID), emitEvent("onBeforeUnmountComponent", debugID);
        }, onUnmountComponent: function onUnmountComponent(debugID) {
          checkDebugID(debugID), emitEvent("onUnmountComponent", debugID);
        }, onTestEvent: function onTestEvent() {
          emitEvent("onTestEvent");
        } };ReactDebugTool.addDevtool = ReactDebugTool.addHook, ReactDebugTool.removeDevtool = ReactDebugTool.removeHook, ReactDebugTool.addHook(ReactInvalidSetStateWarningHook), ReactDebugTool.addHook(ReactComponentTreeHook), ReactDebugTool.addHook(ReactChildrenMutationWarningHook);var url = ExecutionEnvironment.canUseDOM && window.location.href || "";/[?&]react_perf\b/.test(url) && ReactDebugTool.beginProfiling(), module.exports = ReactDebugTool;
    }, { "./ReactChildrenMutationWarningHook": 15, "./ReactComponentTreeHook": 16, "./ReactHostOperationHistoryHook": 19, "./ReactInvalidSetStateWarningHook": 21, "fbjs/lib/ExecutionEnvironment": 2, "fbjs/lib/performanceNow": 11, "fbjs/lib/warning": 12 }], 19: [function (_dereq_, module, exports) {
      "use strict";
      var history = [],
          ReactHostOperationHistoryHook = { onHostOperation: function onHostOperation(debugID, type, payload) {
          history.push({ instanceID: debugID, type: type, payload: payload });
        }, clearHistory: function clearHistory() {
          ReactHostOperationHistoryHook._preventClearing || (history = []);
        }, getHistory: function getHistory() {
          return history;
        } };module.exports = ReactHostOperationHistoryHook;
    }, {}], 20: [function (_dereq_, module, exports) {
      "use strict";
      var debugTool = null;module.exports = { debugTool: debugTool };
    }, { "./ReactDebugTool": 18 }], 21: [function (_dereq_, module, exports) {
      "use strict";
      var processingChildContext,
          warnInvalidSetState,
          ReactInvalidSetStateWarningHook = (_dereq_("fbjs/lib/warning"), { onBeginProcessingChildContext: function onBeginProcessingChildContext() {
          processingChildContext = !0;
        }, onEndProcessingChildContext: function onEndProcessingChildContext() {
          processingChildContext = !1;
        }, onSetState: function onSetState() {
          warnInvalidSetState();
        } });module.exports = ReactInvalidSetStateWarningHook;
    }, { "fbjs/lib/warning": 12 }], 22: [function (_dereq_, module, exports) {
      "use strict";
      function dangerousStyleValue(name, value, component) {
        var isEmpty = null == value || "boolean" == typeof value || "" === value;if (isEmpty) return "";var isNonNumeric = isNaN(value);if (isNonNumeric || 0 === value || isUnitlessNumber.hasOwnProperty(name) && isUnitlessNumber[name]) return "" + value;if ("string" == typeof value) {
          value = value.trim();
        }return value + "px";
      }var CSSProperty = _dereq_("./CSSProperty"),
          isUnitlessNumber = (_dereq_("fbjs/lib/warning"), CSSProperty.isUnitlessNumber);module.exports = dangerousStyleValue;
    }, { "./CSSProperty": 13, "fbjs/lib/warning": 12 }], 23: [function (_dereq_, module, exports) {
      "use strict";
      function reactProdInvariant(code) {
        for (var argCount = arguments.length - 1, message = "Minified React error #" + code + "; visit http://facebook.github.io/react/docs/error-decoder.html?invariant=" + code, argIdx = 0; argIdx < argCount; argIdx++) {
          message += "&args[]=" + encodeURIComponent(arguments[argIdx + 1]);
        }message += " for the full message or use the non-minified dev environment for full errors and additional helpful warnings.";var error = new Error(message);throw error.name = "Invariant Violation", error.framesToPop = 1, error;
      }module.exports = reactProdInvariant;
    }, {}] }, {}, [1])(1);
});
