// vscode/src/vs/platform/instantiation/common/descriptors.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var errors_1 = require('vs/base/common/errors');
var AbstractDescriptor = (function () {
    function AbstractDescriptor(_staticArguments) {
        this._staticArguments = _staticArguments;
        // empty
    }
    AbstractDescriptor.prototype.appendStaticArguments = function (more) {
        this._staticArguments.push.apply(this._staticArguments, more);
    };
    AbstractDescriptor.prototype.staticArguments = function (nth) {
        if (isNaN(nth)) {
            return this._staticArguments.slice(0);
        }
        else {
            return this._staticArguments[nth];
        }
    };
    AbstractDescriptor.prototype._validate = function (type) {
        if (!type) {
            throw errors_1.illegalArgument('can not be falsy');
        }
    };
    return AbstractDescriptor;
}());
exports.AbstractDescriptor = AbstractDescriptor;
var SyncDescriptor = (function (_super) {
    __extends(SyncDescriptor, _super);
    function SyncDescriptor(_ctor) {
        var staticArguments = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            staticArguments[_i - 1] = arguments[_i];
        }
        _super.call(this, staticArguments);
        this._ctor = _ctor;
    }
    Object.defineProperty(SyncDescriptor.prototype, "ctor", {
        get: function () {
            return this._ctor;
        },
        enumerable: true,
        configurable: true
    });
    SyncDescriptor.prototype.bind = function () {
        var moreStaticArguments = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            moreStaticArguments[_i - 0] = arguments[_i];
        }
        var allArgs = [];
        allArgs = allArgs.concat(this.staticArguments());
        allArgs = allArgs.concat(moreStaticArguments);
        return new (SyncDescriptor.bind.apply(SyncDescriptor, [void 0].concat([this._ctor], allArgs)))();
    };
    return SyncDescriptor;
}(AbstractDescriptor));
exports.SyncDescriptor = SyncDescriptor;
exports.createSyncDescriptor = function (ctor) {
    var staticArguments = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        staticArguments[_i - 1] = arguments[_i];
    }
    return new (SyncDescriptor.bind.apply(SyncDescriptor, [void 0].concat([ctor], staticArguments)))();
};
var AsyncDescriptor = (function (_super) {
    __extends(AsyncDescriptor, _super);
    function AsyncDescriptor(_moduleName, _ctorName) {
        var staticArguments = [];
        for (var _i = 2; _i < arguments.length; _i++) {
            staticArguments[_i - 2] = arguments[_i];
        }
        _super.call(this, staticArguments);
        this._moduleName = _moduleName;
        this._ctorName = _ctorName;
        if (typeof _moduleName !== 'string') {
            throw new Error('Invalid AsyncDescriptor arguments, expected `moduleName` to be a string!');
        }
    }
    AsyncDescriptor.create = function (moduleName, ctorName) {
        return new AsyncDescriptor(moduleName, ctorName);
    };
    Object.defineProperty(AsyncDescriptor.prototype, "moduleName", {
        get: function () {
            return this._moduleName;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(AsyncDescriptor.prototype, "ctorName", {
        get: function () {
            return this._ctorName;
        },
        enumerable: true,
        configurable: true
    });
    AsyncDescriptor.prototype.bind = function () {
        var moreStaticArguments = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            moreStaticArguments[_i - 0] = arguments[_i];
        }
        var allArgs = [];
        allArgs = allArgs.concat(this.staticArguments());
        allArgs = allArgs.concat(moreStaticArguments);
        return new (AsyncDescriptor.bind.apply(AsyncDescriptor, [void 0].concat([this.moduleName, this.ctorName], allArgs)))();
    };
    return AsyncDescriptor;
}(AbstractDescriptor));
exports.AsyncDescriptor = AsyncDescriptor;
