// vscode/src/vs/workbench/parts/debug/node/debugConfigurationManager.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var nls = require("vs/nls");
var winjs_base_1 = require("vs/base/common/winjs.base");
var strings = require("vs/base/common/strings");
var types = require("vs/base/common/types");
var platform_1 = require("vs/base/common/platform");
var objects = require("vs/base/common/objects");
var uri_1 = require("vs/base/common/uri");
var network_1 = require("vs/base/common/network");
var paths = require("vs/base/common/paths");
var extensionsRegistry = require("vs/platform/extensions/common/extensionsRegistry");
var platform_2 = require("vs/platform/platform");
var jsonContributionRegistry_1 = require("vs/platform/jsonschemas/common/jsonContributionRegistry");
var configuration_1 = require("vs/platform/configuration/common/configuration");
var files_1 = require("vs/platform/files/common/files");
var telemetry_1 = require("vs/platform/telemetry/common/telemetry");
var workspace_1 = require("vs/platform/workspace/common/workspace");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var debugAdapter_1 = require("vs/workbench/parts/debug/node/debugAdapter");
var editorService_1 = require("vs/workbench/services/editor/common/editorService");
var quickOpenService_1 = require("vs/workbench/services/quickopen/common/quickOpenService");
var configurationResolver_1 = require("vs/workbench/services/configurationResolver/common/configurationResolver");
// debuggers extension point
exports.debuggersExtPoint = extensionsRegistry.ExtensionsRegistry.registerExtensionPoint('debuggers', [], {
    description: nls.localize('vscode.extension.contributes.debuggers', 'Contributes debug adapters.'),
    type: 'array',
    defaultSnippets: [{ body: [{ type: '', extensions: [] }] }],
    items: {
        type: 'object',
        defaultSnippets: [{ body: { type: '', program: '', runtime: '', enableBreakpointsFor: { languageIds: [''] } } }],
        properties: {
            type: {
                description: nls.localize('vscode.extension.contributes.debuggers.type', "Unique identifier for this debug adapter."),
                type: 'string'
            },
            label: {
                description: nls.localize('vscode.extension.contributes.debuggers.label', "Display name for this debug adapter."),
                type: 'string'
            },
            program: {
                description: nls.localize('vscode.extension.contributes.debuggers.program', "Path to the debug adapter program. Path is either absolute or relative to the extension folder."),
                type: 'string'
            },
            args: {
                description: nls.localize('vscode.extension.contributes.debuggers.args', "Optional arguments to pass to the adapter."),
                type: 'array'
            },
            runtime: {
                description: nls.localize('vscode.extension.contributes.debuggers.runtime', "Optional runtime in case the program attribute is not an executable but requires a runtime."),
                type: 'string'
            },
            runtimeArgs: {
                description: nls.localize('vscode.extension.contributes.debuggers.runtimeArgs', "Optional runtime arguments."),
                type: 'array'
            },
            variables: {
                description: nls.localize('vscode.extension.contributes.debuggers.variables', "Mapping from interactive variables (e.g ${action.pickProcess}) in `launch.json` to a command."),
                type: 'object'
            },
            initialConfigurations: {
                description: nls.localize('vscode.extension.contributes.debuggers.initialConfigurations', "Configurations for generating the initial \'launch.json\'."),
                type: ['array', 'string'],
            },
            configurationSnippets: {
                description: nls.localize('vscode.extension.contributes.debuggers.configurationSnippets', "Snippets for adding new configurations in \'launch.json\'."),
                type: 'array'
            },
            configurationAttributes: {
                description: nls.localize('vscode.extension.contributes.debuggers.configurationAttributes', "JSON schema configurations for validating \'launch.json\'."),
                type: 'object'
            },
            windows: {
                description: nls.localize('vscode.extension.contributes.debuggers.windows', "Windows specific settings."),
                type: 'object',
                properties: {
                    runtime: {
                        description: nls.localize('vscode.extension.contributes.debuggers.windows.runtime', "Runtime used for Windows."),
                        type: 'string'
                    }
                }
            },
            osx: {
                description: nls.localize('vscode.extension.contributes.debuggers.osx', "OS X specific settings."),
                type: 'object',
                properties: {
                    runtime: {
                        description: nls.localize('vscode.extension.contributes.debuggers.osx.runtime', "Runtime used for OSX."),
                        type: 'string'
                    }
                }
            },
            linux: {
                description: nls.localize('vscode.extension.contributes.debuggers.linux', "Linux specific settings."),
                type: 'object',
                properties: {
                    runtime: {
                        description: nls.localize('vscode.extension.contributes.debuggers.linux.runtime', "Runtime used for Linux."),
                        type: 'string'
                    }
                }
            }
        }
    }
});
// breakpoints extension point #9037
exports.breakpointsExtPoint = extensionsRegistry.ExtensionsRegistry.registerExtensionPoint('breakpoints', [], {
    description: nls.localize('vscode.extension.contributes.breakpoints', 'Contributes breakpoints.'),
    type: 'array',
    defaultSnippets: [{ body: [{ language: '' }] }],
    items: {
        type: 'object',
        defaultSnippets: [{ body: { language: '' } }],
        properties: {
            language: {
                description: nls.localize('vscode.extension.contributes.breakpoints.language', "Allow breakpoints for this language."),
                type: 'string'
            },
        }
    }
});
// debug general schema
exports.schemaId = 'vscode://schemas/launch';
var defaultCompound = { name: 'Compound', configurations: [] };
var schema = {
    id: exports.schemaId,
    type: 'object',
    title: nls.localize('app.launch.json.title', "Launch"),
    required: ['version', 'configurations'],
    default: { version: '0.2.0', configurations: [], compounds: [] },
    properties: {
        version: {
            type: 'string',
            description: nls.localize('app.launch.json.version', "Version of this file format."),
            default: '0.2.0'
        },
        configurations: {
            type: 'array',
            description: nls.localize('app.launch.json.configurations', "List of configurations. Add new configurations or edit existing ones by using IntelliSense."),
            items: {
                defaultSnippets: [],
                'type': 'object',
                oneOf: []
            }
        },
        // TODO@Isidor remove support for this in December
        debugServer: {
            type: 'number',
            description: nls.localize('app.launch.json.debugServer', "DEPRECATED: please move debugServer inside a configuration.")
        },
        compounds: {
            type: 'array',
            description: nls.localize('app.launch.json.compounds', "List of compounds. Each compound references multiple configurations which will get launched together."),
            items: {
                type: 'object',
                required: ['name', 'configurations'],
                properties: {
                    name: {
                        type: 'string',
                        description: nls.localize('app.launch.json.compound.name', "Name of compound. Appears in the launch configuration drop down menu.")
                    },
                    configurations: {
                        type: 'array',
                        default: [],
                        items: {
                            type: 'string'
                        },
                        description: nls.localize('app.launch.json.compounds.configurations', "Names of configurations that will be started as part of this compound.")
                    }
                },
                default: defaultCompound
            },
            default: [
                defaultCompound
            ]
        }
    }
};
var jsonRegistry = platform_2.Registry.as(jsonContributionRegistry_1.Extensions.JSONContribution);
jsonRegistry.registerSchema(exports.schemaId, schema);
var ConfigurationManager = (function () {
    function ConfigurationManager(contextService, fileService, telemetryService, editorService, configurationService, quickOpenService, configurationResolverService, instantiationService) {
        this.contextService = contextService;
        this.fileService = fileService;
        this.telemetryService = telemetryService;
        this.editorService = editorService;
        this.configurationService = configurationService;
        this.quickOpenService = quickOpenService;
        this.configurationResolverService = configurationResolverService;
        this.instantiationService = instantiationService;
        this.adapters = [];
        this.registerListeners();
        this.allModeIdsForBreakpoints = {};
    }
    ConfigurationManager.prototype.registerListeners = function () {
        var _this = this;
        exports.debuggersExtPoint.setHandler(function (extensions) {
            extensions.forEach(function (extension) {
                extension.value.forEach(function (rawAdapter) {
                    if (!rawAdapter.type || (typeof rawAdapter.type !== 'string')) {
                        extension.collector.error(nls.localize('debugNoType', "Debug adapter 'type' can not be omitted and must be of type 'string'."));
                    }
                    if (rawAdapter.enableBreakpointsFor) {
                        rawAdapter.enableBreakpointsFor.languageIds.forEach(function (modeId) {
                            _this.allModeIdsForBreakpoints[modeId] = true;
                        });
                    }
                    var duplicate = _this.adapters.filter(function (a) { return a.type === rawAdapter.type; }).pop();
                    if (duplicate) {
                        duplicate.merge(rawAdapter, extension.description);
                    }
                    else {
                        _this.adapters.push(_this.instantiationService.createInstance(debugAdapter_1.Adapter, rawAdapter, extension.description));
                    }
                });
            });
            // update the schema to include all attributes, snippets and types from extensions.
            _this.adapters.forEach(function (adapter) {
                var items = schema.properties['configurations'].items;
                var schemaAttributes = adapter.getSchemaAttributes();
                if (schemaAttributes) {
                    (_a = items.oneOf).push.apply(_a, schemaAttributes);
                }
                var configurationSnippets = adapter.configurationSnippets;
                if (configurationSnippets) {
                    (_b = items.defaultSnippets).push.apply(_b, configurationSnippets);
                }
                var _a, _b;
            });
        });
        exports.breakpointsExtPoint.setHandler(function (extensions) {
            extensions.forEach(function (ext) {
                ext.value.forEach(function (breakpoints) {
                    _this.allModeIdsForBreakpoints[breakpoints.language] = true;
                });
            });
        });
    };
    ConfigurationManager.prototype.getAdapter = function (type) {
        return this.adapters.filter(function (adapter) { return strings.equalsIgnoreCase(adapter.type, type); }).pop();
    };
    ConfigurationManager.prototype.getCompound = function (name) {
        var config = this.configurationService.getConfiguration('launch');
        if (!config || !config.compounds) {
            return null;
        }
        return config.compounds.filter(function (compound) { return compound.name === name; }).pop();
    };
    ConfigurationManager.prototype.getConfiguration = function (nameOrConfig) {
        var _this = this;
        var config = this.configurationService.getConfiguration('launch');
        var result = null;
        if (types.isObject(nameOrConfig)) {
            result = objects.deepClone(nameOrConfig);
        }
        else {
            if (!config || !config.configurations) {
                return winjs_base_1.TPromise.as(null);
            }
            // if the configuration name is not set yet, take the first launch config (can happen if debug viewlet has not been opened yet).
            var filtered = config.configurations.filter(function (cfg) { return cfg.name === nameOrConfig; });
            result = filtered.length === 1 ? filtered[0] : config.configurations[0];
            result = objects.deepClone(result);
            if (config && result && config.debugServer) {
                result.debugServer = config.debugServer;
            }
        }
        if (result) {
            // Set operating system specific properties #1873
            var setOSProperties = function (flag, osConfig) {
                if (flag && osConfig) {
                    Object.keys(osConfig).forEach(function (key) {
                        result[key] = osConfig[key];
                    });
                }
            };
            setOSProperties(platform_1.isWindows, result.windows);
            setOSProperties(platform_1.isMacintosh, result.osx);
            setOSProperties(platform_1.isLinux, result.linux);
            // massage configuration attributes - append workspace path to relatvie paths, substitute variables in paths.
            Object.keys(result).forEach(function (key) {
                result[key] = _this.configurationResolverService.resolveAny(result[key]);
            });
            var adapter = this.getAdapter(result.type);
            return this.configurationResolverService.resolveInteractiveVariables(result, adapter ? adapter.variables : null);
        }
        return winjs_base_1.TPromise.as(null);
    };
    ConfigurationManager.prototype.openConfigFile = function (sideBySide) {
        var _this = this;
        var resource = uri_1.default.file(paths.join(this.contextService.getWorkspace().resource.fsPath, '/.vscode/launch.json'));
        var configFileCreated = false;
        return this.fileService.resolveContent(resource).then(function (content) { return true; }, function (err) {
            return _this.quickOpenService.pick(_this.adapters, { placeHolder: nls.localize('selectDebug', "Select Environment") })
                .then(function (adapter) { return adapter ? adapter.getInitialConfigurationContent() : null; })
                .then(function (content) {
                if (!content) {
                    return false;
                }
                configFileCreated = true;
                return _this.fileService.updateContent(resource, content).then(function () { return true; });
            });
        })
            .then(function (errorFree) {
            if (!errorFree) {
                return false;
            }
            _this.telemetryService.publicLog('debugConfigure');
            return _this.editorService.openEditor({
                resource: resource,
                options: {
                    forceOpen: true,
                    pinned: configFileCreated // pin only if config file is created #8727
                },
            }, sideBySide).then(function () { return true; });
        }, function (error) {
            throw new Error(nls.localize('DebugConfig.failed', "Unable to create 'launch.json' file inside the '.vscode' folder ({0}).", error));
        });
    };
    ConfigurationManager.prototype.canSetBreakpointsIn = function (model) {
        if (model.uri.scheme === network_1.Schemas.inMemory) {
            return false;
        }
        if (this.configurationService.getConfiguration('debug').allowBreakpointsEverywhere) {
            return true;
        }
        var mode = model ? model.getMode() : null;
        var modeId = mode ? mode.getId() : null;
        return !!this.allModeIdsForBreakpoints[modeId];
    };
    return ConfigurationManager;
}());
ConfigurationManager = __decorate([
    __param(0, workspace_1.IWorkspaceContextService),
    __param(1, files_1.IFileService),
    __param(2, telemetry_1.ITelemetryService),
    __param(3, editorService_1.IWorkbenchEditorService),
    __param(4, configuration_1.IConfigurationService),
    __param(5, quickOpenService_1.IQuickOpenService),
    __param(6, configurationResolver_1.IConfigurationResolverService),
    __param(7, instantiation_1.IInstantiationService)
], ConfigurationManager);
exports.ConfigurationManager = ConfigurationManager;
