// vscode/src/vs/workbench/parts/terminal/electron-browser/terminalActions.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var nls = require('vs/nls');
var os = require('os');
var actions_1 = require('vs/base/common/actions');
var editorCommon_1 = require('vs/editor/common/editorCommon');
var codeEditorService_1 = require('vs/editor/common/services/codeEditorService');
var terminal_1 = require('vs/workbench/parts/terminal/electron-browser/terminal');
var actionbar_1 = require('vs/base/browser/ui/actionbar/actionbar');
var winjs_base_1 = require('vs/base/common/winjs.base');
var panel_1 = require('vs/workbench/browser/panel');
var partService_1 = require('vs/workbench/services/part/common/partService');
var panelService_1 = require('vs/workbench/services/panel/common/panelService');
var ToggleTerminalAction = (function (_super) {
    __extends(ToggleTerminalAction, _super);
    function ToggleTerminalAction(id, label, panelService, partService) {
        _super.call(this, id, label, terminal_1.TERMINAL_PANEL_ID, panelService, partService);
    }
    ToggleTerminalAction.ID = 'workbench.action.terminal.toggleTerminal';
    ToggleTerminalAction.LABEL = nls.localize('workbench.action.terminal.toggleTerminal', "Toggle Integrated Terminal");
    ToggleTerminalAction = __decorate([
        __param(2, panelService_1.IPanelService),
        __param(3, partService_1.IPartService)
    ], ToggleTerminalAction);
    return ToggleTerminalAction;
}(panel_1.TogglePanelAction));
exports.ToggleTerminalAction = ToggleTerminalAction;
var KillTerminalAction = (function (_super) {
    __extends(KillTerminalAction, _super);
    function KillTerminalAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
        this.class = 'terminal-action kill';
    }
    KillTerminalAction.prototype.run = function (event) {
        var terminalInstance = this.terminalService.getActiveInstance();
        if (terminalInstance) {
            this.terminalService.getActiveInstance().dispose();
            if (this.terminalService.terminalInstances.length > 0) {
                this.terminalService.showPanel(true);
            }
        }
        return winjs_base_1.TPromise.as(void 0);
    };
    KillTerminalAction.ID = 'workbench.action.terminal.kill';
    KillTerminalAction.LABEL = nls.localize('workbench.action.terminal.kill', "Kill the Active Terminal Instance");
    KillTerminalAction.PANEL_LABEL = nls.localize('workbench.action.terminal.kill.short', "Kill Terminal");
    KillTerminalAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], KillTerminalAction);
    return KillTerminalAction;
}(actions_1.Action));
exports.KillTerminalAction = KillTerminalAction;
/**
 * Copies the terminal selection. Note that since the command palette takes focus from the terminal,
 * this can only be triggered via a keybinding.
 */
var CopyTerminalSelectionAction = (function (_super) {
    __extends(CopyTerminalSelectionAction, _super);
    function CopyTerminalSelectionAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
    }
    CopyTerminalSelectionAction.prototype.run = function (event) {
        var terminalInstance = this.terminalService.getActiveInstance();
        if (terminalInstance) {
            terminalInstance.copySelection();
        }
        return winjs_base_1.TPromise.as(void 0);
    };
    CopyTerminalSelectionAction.ID = 'workbench.action.terminal.copySelection';
    CopyTerminalSelectionAction.LABEL = nls.localize('workbench.action.terminal.copySelection', "Copy Selection");
    CopyTerminalSelectionAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], CopyTerminalSelectionAction);
    return CopyTerminalSelectionAction;
}(actions_1.Action));
exports.CopyTerminalSelectionAction = CopyTerminalSelectionAction;
var CreateNewTerminalAction = (function (_super) {
    __extends(CreateNewTerminalAction, _super);
    function CreateNewTerminalAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
        this.class = 'terminal-action new';
    }
    CreateNewTerminalAction.prototype.run = function (event) {
        this.terminalService.setActiveInstance(this.terminalService.createInstance());
        return this.terminalService.showPanel(true);
    };
    CreateNewTerminalAction.ID = 'workbench.action.terminal.new';
    CreateNewTerminalAction.LABEL = nls.localize('workbench.action.terminal.new', "Create New Integrated Terminal");
    CreateNewTerminalAction.PANEL_LABEL = nls.localize('workbench.action.terminal.new.short', "New Terminal");
    CreateNewTerminalAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], CreateNewTerminalAction);
    return CreateNewTerminalAction;
}(actions_1.Action));
exports.CreateNewTerminalAction = CreateNewTerminalAction;
var FocusTerminalAction = (function (_super) {
    __extends(FocusTerminalAction, _super);
    function FocusTerminalAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
    }
    FocusTerminalAction.prototype.run = function (event) {
        var terminalInstance = this.terminalService.getActiveInstance();
        if (!terminalInstance) {
            terminalInstance = this.terminalService.createInstance();
        }
        this.terminalService.setActiveInstance(terminalInstance);
        return this.terminalService.showPanel(true);
    };
    FocusTerminalAction.ID = 'workbench.action.terminal.focus';
    FocusTerminalAction.LABEL = nls.localize('workbench.action.terminal.focus', "Focus Terminal");
    FocusTerminalAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], FocusTerminalAction);
    return FocusTerminalAction;
}(actions_1.Action));
exports.FocusTerminalAction = FocusTerminalAction;
var FocusNextTerminalAction = (function (_super) {
    __extends(FocusNextTerminalAction, _super);
    function FocusNextTerminalAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
    }
    FocusNextTerminalAction.prototype.run = function (event) {
        this.terminalService.setActiveInstanceToNext();
        return this.terminalService.showPanel(true);
    };
    FocusNextTerminalAction.ID = 'workbench.action.terminal.focusNext';
    FocusNextTerminalAction.LABEL = nls.localize('workbench.action.terminal.focusNext', "Focus Next Terminal");
    FocusNextTerminalAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], FocusNextTerminalAction);
    return FocusNextTerminalAction;
}(actions_1.Action));
exports.FocusNextTerminalAction = FocusNextTerminalAction;
var FocusPreviousTerminalAction = (function (_super) {
    __extends(FocusPreviousTerminalAction, _super);
    function FocusPreviousTerminalAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
    }
    FocusPreviousTerminalAction.prototype.run = function (event) {
        this.terminalService.setActiveInstanceToPrevious();
        return this.terminalService.showPanel(true);
    };
    FocusPreviousTerminalAction.ID = 'workbench.action.terminal.focusPrevious';
    FocusPreviousTerminalAction.LABEL = nls.localize('workbench.action.terminal.focusPrevious', "Focus Previous Terminal");
    FocusPreviousTerminalAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], FocusPreviousTerminalAction);
    return FocusPreviousTerminalAction;
}(actions_1.Action));
exports.FocusPreviousTerminalAction = FocusPreviousTerminalAction;
var TerminalPasteAction = (function (_super) {
    __extends(TerminalPasteAction, _super);
    function TerminalPasteAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
    }
    TerminalPasteAction.prototype.run = function (event) {
        var terminalInstance = this.terminalService.getActiveInstance();
        if (!terminalInstance) {
            terminalInstance = this.terminalService.createInstance();
        }
        terminalInstance.paste();
        return winjs_base_1.TPromise.as(void 0);
    };
    TerminalPasteAction.ID = 'workbench.action.terminal.paste';
    TerminalPasteAction.LABEL = nls.localize('workbench.action.terminal.paste', "Paste into Active Terminal");
    TerminalPasteAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], TerminalPasteAction);
    return TerminalPasteAction;
}(actions_1.Action));
exports.TerminalPasteAction = TerminalPasteAction;
var RunSelectedTextInTerminalAction = (function (_super) {
    __extends(RunSelectedTextInTerminalAction, _super);
    function RunSelectedTextInTerminalAction(id, label, codeEditorService, terminalService) {
        _super.call(this, id, label);
        this.codeEditorService = codeEditorService;
        this.terminalService = terminalService;
    }
    RunSelectedTextInTerminalAction.prototype.run = function (event) {
        var terminalInstance = this.terminalService.getActiveInstance();
        if (!terminalInstance) {
            terminalInstance = this.terminalService.createInstance();
        }
        var editor = this.codeEditorService.getFocusedCodeEditor();
        var selection = editor.getSelection();
        var text;
        if (selection.isEmpty()) {
            text = editor.getValue();
        }
        else {
            var endOfLinePreference = os.EOL === '\n' ? editorCommon_1.EndOfLinePreference.LF : editorCommon_1.EndOfLinePreference.CRLF;
            text = editor.getModel().getValueInRange(selection, endOfLinePreference);
        }
        terminalInstance.sendText(text, true);
        return winjs_base_1.TPromise.as(void 0);
    };
    RunSelectedTextInTerminalAction.ID = 'workbench.action.terminal.runSelectedText';
    RunSelectedTextInTerminalAction.LABEL = nls.localize('workbench.action.terminal.runSelectedText', "Run Selected Text In Active Terminal");
    RunSelectedTextInTerminalAction = __decorate([
        __param(2, codeEditorService_1.ICodeEditorService),
        __param(3, terminal_1.ITerminalService)
    ], RunSelectedTextInTerminalAction);
    return RunSelectedTextInTerminalAction;
}(actions_1.Action));
exports.RunSelectedTextInTerminalAction = RunSelectedTextInTerminalAction;
var SwitchTerminalInstanceAction = (function (_super) {
    __extends(SwitchTerminalInstanceAction, _super);
    function SwitchTerminalInstanceAction(id, label, terminalService) {
        _super.call(this, SwitchTerminalInstanceAction.ID, SwitchTerminalInstanceAction.LABEL);
        this.terminalService = terminalService;
        this.class = 'terminal-action switch-terminal-instance';
    }
    SwitchTerminalInstanceAction.prototype.run = function (item) {
        var selectedTerminalIndex = parseInt(item.split(':')[0], 10) - 1;
        this.terminalService.setActiveInstanceByIndex(selectedTerminalIndex);
        return this.terminalService.showPanel(true);
    };
    SwitchTerminalInstanceAction.ID = 'workbench.action.terminal.switchTerminalInstance';
    SwitchTerminalInstanceAction.LABEL = nls.localize('workbench.action.terminal.switchTerminalInstance', "Switch Terminal Instance");
    SwitchTerminalInstanceAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], SwitchTerminalInstanceAction);
    return SwitchTerminalInstanceAction;
}(actions_1.Action));
exports.SwitchTerminalInstanceAction = SwitchTerminalInstanceAction;
var SwitchTerminalInstanceActionItem = (function (_super) {
    __extends(SwitchTerminalInstanceActionItem, _super);
    function SwitchTerminalInstanceActionItem(action, terminalService) {
        _super.call(this, null, action, terminalService.getInstanceLabels(), terminalService.activeTerminalInstanceIndex);
        this.terminalService = terminalService;
        this.toDispose.push(terminalService.onInstancesChanged(this._updateItems, this));
        this.toDispose.push(terminalService.onActiveInstanceChanged(this._updateItems, this));
        this.toDispose.push(terminalService.onInstanceTitleChanged(this._updateItems, this));
    }
    SwitchTerminalInstanceActionItem.prototype._updateItems = function () {
        this.setOptions(this.terminalService.getInstanceLabels(), this.terminalService.activeTerminalInstanceIndex);
    };
    SwitchTerminalInstanceActionItem = __decorate([
        __param(1, terminal_1.ITerminalService)
    ], SwitchTerminalInstanceActionItem);
    return SwitchTerminalInstanceActionItem;
}(actionbar_1.SelectActionItem));
exports.SwitchTerminalInstanceActionItem = SwitchTerminalInstanceActionItem;
var ScrollDownTerminalAction = (function (_super) {
    __extends(ScrollDownTerminalAction, _super);
    function ScrollDownTerminalAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
    }
    ScrollDownTerminalAction.prototype.run = function (event) {
        var terminalInstance = this.terminalService.getActiveInstance();
        if (terminalInstance) {
            terminalInstance.scrollDownLine();
        }
        return winjs_base_1.TPromise.as(void 0);
    };
    ScrollDownTerminalAction.ID = 'workbench.action.terminal.scrollDown';
    ScrollDownTerminalAction.LABEL = nls.localize('workbench.action.terminal.scrollDown', "Scroll Down (Line)");
    ScrollDownTerminalAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], ScrollDownTerminalAction);
    return ScrollDownTerminalAction;
}(actions_1.Action));
exports.ScrollDownTerminalAction = ScrollDownTerminalAction;
var ScrollDownPageTerminalAction = (function (_super) {
    __extends(ScrollDownPageTerminalAction, _super);
    function ScrollDownPageTerminalAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
    }
    ScrollDownPageTerminalAction.prototype.run = function (event) {
        var terminalInstance = this.terminalService.getActiveInstance();
        if (terminalInstance) {
            terminalInstance.scrollDownPage();
        }
        return winjs_base_1.TPromise.as(void 0);
    };
    ScrollDownPageTerminalAction.ID = 'workbench.action.terminal.scrollDownPage';
    ScrollDownPageTerminalAction.LABEL = nls.localize('workbench.action.terminal.scrollDownPage', "Scroll Down (Page)");
    ScrollDownPageTerminalAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], ScrollDownPageTerminalAction);
    return ScrollDownPageTerminalAction;
}(actions_1.Action));
exports.ScrollDownPageTerminalAction = ScrollDownPageTerminalAction;
var ScrollUpTerminalAction = (function (_super) {
    __extends(ScrollUpTerminalAction, _super);
    function ScrollUpTerminalAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
    }
    ScrollUpTerminalAction.prototype.run = function (event) {
        var terminalInstance = this.terminalService.getActiveInstance();
        if (terminalInstance) {
            terminalInstance.scrollUpLine();
        }
        return winjs_base_1.TPromise.as(void 0);
    };
    ScrollUpTerminalAction.ID = 'workbench.action.terminal.scrollUp';
    ScrollUpTerminalAction.LABEL = nls.localize('workbench.action.terminal.scrollUp', "Scroll Up (Line)");
    ScrollUpTerminalAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], ScrollUpTerminalAction);
    return ScrollUpTerminalAction;
}(actions_1.Action));
exports.ScrollUpTerminalAction = ScrollUpTerminalAction;
var ScrollUpPageTerminalAction = (function (_super) {
    __extends(ScrollUpPageTerminalAction, _super);
    function ScrollUpPageTerminalAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
    }
    ScrollUpPageTerminalAction.prototype.run = function (event) {
        var terminalInstance = this.terminalService.getActiveInstance();
        if (terminalInstance) {
            terminalInstance.scrollUpPage();
        }
        return winjs_base_1.TPromise.as(void 0);
    };
    ScrollUpPageTerminalAction.ID = 'workbench.action.terminal.scrollUpPage';
    ScrollUpPageTerminalAction.LABEL = nls.localize('workbench.action.terminal.scrollUpPage', "Scroll Up (Page)");
    ScrollUpPageTerminalAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], ScrollUpPageTerminalAction);
    return ScrollUpPageTerminalAction;
}(actions_1.Action));
exports.ScrollUpPageTerminalAction = ScrollUpPageTerminalAction;
var ClearTerminalAction = (function (_super) {
    __extends(ClearTerminalAction, _super);
    function ClearTerminalAction(id, label, terminalService) {
        _super.call(this, id, label);
        this.terminalService = terminalService;
    }
    ClearTerminalAction.prototype.run = function (event) {
        var terminalInstance = this.terminalService.getActiveInstance();
        if (terminalInstance) {
            terminalInstance.clear();
        }
        return winjs_base_1.TPromise.as(void 0);
    };
    ClearTerminalAction.ID = 'workbench.action.terminal.clear';
    ClearTerminalAction.LABEL = nls.localize('workbench.action.terminal.clear', "Clear");
    ClearTerminalAction = __decorate([
        __param(2, terminal_1.ITerminalService)
    ], ClearTerminalAction);
    return ClearTerminalAction;
}(actions_1.Action));
exports.ClearTerminalAction = ClearTerminalAction;
