// vscode/src/vs/workbench/parts/terminal/electron-browser/terminal.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var platform = require('vs/base/common/platform');
var processes = require('vs/base/node/processes');
var contextkey_1 = require('vs/platform/contextkey/common/contextkey');
var instantiation_1 = require('vs/platform/instantiation/common/instantiation');
exports.TERMINAL_PANEL_ID = 'workbench.panel.terminal';
exports.TERMINAL_SERVICE_ID = 'terminalService';
exports.TERMINAL_DEFAULT_SHELL_LINUX = !platform.isWindows ? (process.env.SHELL || 'sh') : 'sh';
exports.TERMINAL_DEFAULT_SHELL_OSX = !platform.isWindows ? (process.env.SHELL || 'sh') : 'sh';
exports.TERMINAL_DEFAULT_SHELL_WINDOWS = processes.getWindowsShell();
/**
 * A context key that is set when the integrated terminal has focus.
 */
exports.KEYBINDING_CONTEXT_TERMINAL_FOCUS = new contextkey_1.RawContextKey('terminalFocus', undefined);
exports.KEYBINDING_CONTEXT_TERMINAL_NOT_FOCUSED = exports.KEYBINDING_CONTEXT_TERMINAL_FOCUS.toNegated();
exports.ITerminalService = instantiation_1.createDecorator(exports.TERMINAL_SERVICE_ID);
