// vscode/src/vs/workbench/api/node/extHostTextEditor.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var errors_1 = require("vs/base/common/errors");
var idGenerator_1 = require("vs/base/common/idGenerator");
var winjs_base_1 = require("vs/base/common/winjs.base");
var extHostTypes_1 = require("./extHostTypes");
var TypeConverters = require("./extHostTypeConverters");
var TextEditorDecorationType = (function () {
    function TextEditorDecorationType(proxy, options) {
        this.key = TextEditorDecorationType._Keys.nextId();
        this._proxy = proxy;
        this._proxy.$registerTextEditorDecorationType(this.key, options);
    }
    TextEditorDecorationType.prototype.dispose = function () {
        this._proxy.$removeTextEditorDecorationType(this.key);
    };
    return TextEditorDecorationType;
}());
TextEditorDecorationType._Keys = new idGenerator_1.IdGenerator('TextEditorDecorationType');
exports.TextEditorDecorationType = TextEditorDecorationType;
var TextEditorEdit = (function () {
    function TextEditorEdit(document, options) {
        this._documentVersionId = document.version;
        this._collectedEdits = [];
        this._setEndOfLine = 0;
        this._undoStopBefore = options.undoStopBefore;
        this._undoStopAfter = options.undoStopAfter;
    }
    TextEditorEdit.prototype.finalize = function () {
        return {
            documentVersionId: this._documentVersionId,
            edits: this._collectedEdits,
            setEndOfLine: this._setEndOfLine,
            undoStopBefore: this._undoStopBefore,
            undoStopAfter: this._undoStopAfter
        };
    };
    TextEditorEdit.prototype.replace = function (location, value) {
        var range = null;
        if (location instanceof extHostTypes_1.Position) {
            range = new extHostTypes_1.Range(location, location);
        }
        else if (location instanceof extHostTypes_1.Range) {
            range = location;
        }
        else {
            throw new Error('Unrecognized location');
        }
        this._collectedEdits.push({
            range: range,
            text: value,
            forceMoveMarkers: false
        });
    };
    TextEditorEdit.prototype.insert = function (location, value) {
        this._collectedEdits.push({
            range: new extHostTypes_1.Range(location, location),
            text: value,
            forceMoveMarkers: true
        });
    };
    TextEditorEdit.prototype.delete = function (location) {
        var range = null;
        if (location instanceof extHostTypes_1.Range) {
            range = location;
        }
        else {
            throw new Error('Unrecognized location');
        }
        this._collectedEdits.push({
            range: range,
            text: null,
            forceMoveMarkers: true
        });
    };
    TextEditorEdit.prototype.setEndOfLine = function (endOfLine) {
        if (endOfLine !== extHostTypes_1.EndOfLine.LF && endOfLine !== extHostTypes_1.EndOfLine.CRLF) {
            throw errors_1.illegalArgument('endOfLine');
        }
        this._setEndOfLine = endOfLine;
    };
    return TextEditorEdit;
}());
exports.TextEditorEdit = TextEditorEdit;
function deprecated(name, message) {
    if (message === void 0) { message = 'Refer to the documentation for further details.'; }
    return function (target, key, descriptor) {
        var originalMethod = descriptor.value;
        descriptor.value = function () {
            var args = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                args[_i] = arguments[_i];
            }
            console.warn("[Deprecation Warning] method '" + name + "' is deprecated and should no longer be used. " + message);
            return originalMethod.apply(this, args);
        };
        return descriptor;
    };
}
var ExtHostTextEditorOptions = (function () {
    function ExtHostTextEditorOptions(proxy, id, source) {
        this._proxy = proxy;
        this._id = id;
        this._accept(source);
    }
    ExtHostTextEditorOptions.prototype._accept = function (source) {
        this._tabSize = source.tabSize;
        this._insertSpaces = source.insertSpaces;
        this._cursorStyle = source.cursorStyle;
        this._lineNumbers = source.lineNumbers;
    };
    Object.defineProperty(ExtHostTextEditorOptions.prototype, "tabSize", {
        get: function () {
            return this._tabSize;
        },
        set: function (value) {
            var tabSize = this._validateTabSize(value);
            if (tabSize === null) {
                // ignore invalid call
                return;
            }
            if (typeof tabSize === 'number') {
                if (this._tabSize === tabSize) {
                    // nothing to do
                    return;
                }
                // reflect the new tabSize value immediately
                this._tabSize = tabSize;
            }
            warnOnError(this._proxy.$trySetOptions(this._id, {
                tabSize: tabSize
            }));
        },
        enumerable: true,
        configurable: true
    });
    ExtHostTextEditorOptions.prototype._validateTabSize = function (value) {
        if (value === 'auto') {
            return 'auto';
        }
        if (typeof value === 'number') {
            var r = Math.floor(value);
            return (r > 0 ? r : null);
        }
        if (typeof value === 'string') {
            var r = parseInt(value, 10);
            if (isNaN(r)) {
                return null;
            }
            return (r > 0 ? r : null);
        }
        return null;
    };
    Object.defineProperty(ExtHostTextEditorOptions.prototype, "insertSpaces", {
        get: function () {
            return this._insertSpaces;
        },
        set: function (value) {
            var insertSpaces = this._validateInsertSpaces(value);
            if (typeof insertSpaces === 'boolean') {
                if (this._insertSpaces === insertSpaces) {
                    // nothing to do
                    return;
                }
                // reflect the new insertSpaces value immediately
                this._insertSpaces = insertSpaces;
            }
            warnOnError(this._proxy.$trySetOptions(this._id, {
                insertSpaces: insertSpaces
            }));
        },
        enumerable: true,
        configurable: true
    });
    ExtHostTextEditorOptions.prototype._validateInsertSpaces = function (value) {
        if (value === 'auto') {
            return 'auto';
        }
        return (value === 'false' ? false : Boolean(value));
    };
    Object.defineProperty(ExtHostTextEditorOptions.prototype, "cursorStyle", {
        get: function () {
            return this._cursorStyle;
        },
        set: function (value) {
            if (this._cursorStyle === value) {
                // nothing to do
                return;
            }
            this._cursorStyle = value;
            warnOnError(this._proxy.$trySetOptions(this._id, {
                cursorStyle: value
            }));
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExtHostTextEditorOptions.prototype, "lineNumbers", {
        get: function () {
            return this._lineNumbers;
        },
        set: function (value) {
            if (this._lineNumbers === value) {
                // nothing to do
                return;
            }
            this._lineNumbers = value;
            warnOnError(this._proxy.$trySetOptions(this._id, {
                lineNumbers: value
            }));
        },
        enumerable: true,
        configurable: true
    });
    ExtHostTextEditorOptions.prototype.assign = function (newOptions) {
        var bulkConfigurationUpdate = {};
        var hasUpdate = false;
        if (typeof newOptions.tabSize !== 'undefined') {
            var tabSize = this._validateTabSize(newOptions.tabSize);
            if (tabSize === 'auto') {
                hasUpdate = true;
                bulkConfigurationUpdate.tabSize = tabSize;
            }
            else if (typeof tabSize === 'number' && this._tabSize !== tabSize) {
                // reflect the new tabSize value immediately
                this._tabSize = tabSize;
                hasUpdate = true;
                bulkConfigurationUpdate.tabSize = tabSize;
            }
        }
        if (typeof newOptions.insertSpaces !== 'undefined') {
            var insertSpaces = this._validateInsertSpaces(newOptions.insertSpaces);
            if (insertSpaces === 'auto') {
                hasUpdate = true;
                bulkConfigurationUpdate.insertSpaces = insertSpaces;
            }
            else if (this._insertSpaces !== insertSpaces) {
                // reflect the new insertSpaces value immediately
                this._insertSpaces = insertSpaces;
                hasUpdate = true;
                bulkConfigurationUpdate.insertSpaces = insertSpaces;
            }
        }
        if (typeof newOptions.cursorStyle !== 'undefined') {
            if (this._cursorStyle !== newOptions.cursorStyle) {
                this._cursorStyle = newOptions.cursorStyle;
                hasUpdate = true;
                bulkConfigurationUpdate.cursorStyle = newOptions.cursorStyle;
            }
        }
        if (typeof newOptions.lineNumbers !== 'undefined') {
            if (this._lineNumbers !== newOptions.lineNumbers) {
                this._lineNumbers = newOptions.lineNumbers;
                hasUpdate = true;
                bulkConfigurationUpdate.lineNumbers = newOptions.lineNumbers;
            }
        }
        if (hasUpdate) {
            warnOnError(this._proxy.$trySetOptions(this._id, bulkConfigurationUpdate));
        }
    };
    return ExtHostTextEditorOptions;
}());
exports.ExtHostTextEditorOptions = ExtHostTextEditorOptions;
var ExtHostTextEditor = (function () {
    function ExtHostTextEditor(proxy, id, document, selections, options, viewColumn) {
        this._proxy = proxy;
        this._id = id;
        this._documentData = document;
        this._selections = selections;
        this._options = new ExtHostTextEditorOptions(this._proxy, this._id, options);
        this._viewColumn = viewColumn;
    }
    ExtHostTextEditor.prototype.dispose = function () {
        this._documentData = null;
    };
    ExtHostTextEditor.prototype.show = function (column) {
        this._proxy.$tryShowEditor(this._id, TypeConverters.fromViewColumn(column));
    };
    ExtHostTextEditor.prototype.hide = function () {
        this._proxy.$tryHideEditor(this._id);
    };
    Object.defineProperty(ExtHostTextEditor.prototype, "document", {
        // ---- the document
        get: function () {
            return this._documentData
                ? this._documentData.document
                : undefined;
        },
        set: function (value) {
            throw errors_1.readonly('document');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExtHostTextEditor.prototype, "options", {
        // ---- options
        get: function () {
            return this._options;
        },
        set: function (value) {
            this._options.assign(value);
        },
        enumerable: true,
        configurable: true
    });
    ExtHostTextEditor.prototype._acceptOptions = function (options) {
        this._options._accept(options);
    };
    Object.defineProperty(ExtHostTextEditor.prototype, "viewColumn", {
        // ---- view column
        get: function () {
            return this._viewColumn;
        },
        set: function (value) {
            throw errors_1.readonly('viewColumn');
        },
        enumerable: true,
        configurable: true
    });
    ExtHostTextEditor.prototype._acceptViewColumn = function (value) {
        this._viewColumn = value;
    };
    Object.defineProperty(ExtHostTextEditor.prototype, "selection", {
        // ---- selections
        get: function () {
            return this._selections && this._selections[0];
        },
        set: function (value) {
            if (!(value instanceof extHostTypes_1.Selection)) {
                throw errors_1.illegalArgument('selection');
            }
            this._selections = [value];
            this._trySetSelection(true);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ExtHostTextEditor.prototype, "selections", {
        get: function () {
            return this._selections;
        },
        set: function (value) {
            if (!Array.isArray(value) || value.some(function (a) { return !(a instanceof extHostTypes_1.Selection); })) {
                throw errors_1.illegalArgument('selections');
            }
            this._selections = value;
            this._trySetSelection(true);
        },
        enumerable: true,
        configurable: true
    });
    ExtHostTextEditor.prototype.setDecorations = function (decorationType, ranges) {
        var _this = this;
        this._runOnProxy(function () { return _this._proxy.$trySetDecorations(_this._id, decorationType.key, TypeConverters.fromRangeOrRangeWithMessage(ranges)); }, true);
    };
    ExtHostTextEditor.prototype.revealRange = function (range, revealType) {
        var _this = this;
        this._runOnProxy(function () { return _this._proxy.$tryRevealRange(_this._id, TypeConverters.fromRange(range), (revealType || extHostTypes_1.TextEditorRevealType.Default)); }, true);
    };
    ExtHostTextEditor.prototype._trySetSelection = function (silent) {
        var _this = this;
        var selection = this._selections.map(TypeConverters.fromSelection);
        return this._runOnProxy(function () { return _this._proxy.$trySetSelections(_this._id, selection); }, silent);
    };
    ExtHostTextEditor.prototype._acceptSelections = function (selections) {
        this._selections = selections;
    };
    // ---- editing
    ExtHostTextEditor.prototype.edit = function (callback, options) {
        if (options === void 0) { options = { undoStopBefore: true, undoStopAfter: true }; }
        var edit = new TextEditorEdit(this._documentData.document, options);
        callback(edit);
        return this._applyEdit(edit);
    };
    ExtHostTextEditor.prototype._applyEdit = function (editBuilder) {
        var editData = editBuilder.finalize();
        // prepare data for serialization
        var edits = editData.edits.map(function (edit) {
            return {
                range: TypeConverters.fromRange(edit.range),
                text: edit.text,
                forceMoveMarkers: edit.forceMoveMarkers
            };
        });
        return this._proxy.$tryApplyEdits(this._id, editData.documentVersionId, edits, {
            setEndOfLine: editData.setEndOfLine,
            undoStopBefore: editData.undoStopBefore,
            undoStopAfter: editData.undoStopAfter
        });
    };
    ExtHostTextEditor.prototype.insertSnippet = function (snippet, where, options) {
        if (options === void 0) { options = { undoStopBefore: true, undoStopAfter: true }; }
        var ranges;
        if (!where || (Array.isArray(where) && where.length === 0)) {
            ranges = this._selections.map(TypeConverters.fromRange);
        }
        else if (where instanceof extHostTypes_1.Position) {
            var _a = TypeConverters.fromPosition(where), lineNumber = _a.lineNumber, column = _a.column;
            ranges = [{ startLineNumber: lineNumber, startColumn: column, endLineNumber: lineNumber, endColumn: column }];
        }
        else if (where instanceof extHostTypes_1.Range) {
            ranges = [TypeConverters.fromRange(where)];
        }
        else {
            ranges = [];
            for (var _i = 0, where_1 = where; _i < where_1.length; _i++) {
                var posOrRange = where_1[_i];
                if (posOrRange instanceof extHostTypes_1.Range) {
                    ranges.push(TypeConverters.fromRange(posOrRange));
                }
                else {
                    var _b = TypeConverters.fromPosition(posOrRange), lineNumber = _b.lineNumber, column = _b.column;
                    ranges.push({ startLineNumber: lineNumber, startColumn: column, endLineNumber: lineNumber, endColumn: column });
                }
            }
        }
        return this._proxy.$tryInsertSnippet(this._id, snippet.value, ranges, options);
    };
    // ---- util
    ExtHostTextEditor.prototype._runOnProxy = function (callback, silent) {
        var _this = this;
        return callback().then(function () { return _this; }, function (err) {
            if (!silent) {
                return winjs_base_1.TPromise.wrapError(silent);
            }
            console.warn(err);
            return undefined;
        });
    };
    return ExtHostTextEditor;
}());
__decorate([
    deprecated('TextEditor.show')
], ExtHostTextEditor.prototype, "show", null);
__decorate([
    deprecated('TextEditor.hide')
], ExtHostTextEditor.prototype, "hide", null);
exports.ExtHostTextEditor = ExtHostTextEditor;
function warnOnError(promise) {
    promise.then(null, function (err) {
        console.warn(err);
    });
}
