// vscode/src/vs/workbench/api/node/mainThreadDocuments.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var errorMessage_1 = require("vs/base/common/errorMessage");
var modelService_1 = require("vs/editor/common/services/modelService");
var editorCommon = require("vs/editor/common/editorCommon");
var threadService_1 = require("vs/workbench/services/thread/common/threadService");
var uri_1 = require("vs/base/common/uri");
var lifecycle_1 = require("vs/base/common/lifecycle");
var event_1 = require("vs/platform/event/common/event");
var textfiles_1 = require("vs/workbench/services/textfile/common/textfiles");
var winjs_base_1 = require("vs/base/common/winjs.base");
var files_1 = require("vs/platform/files/common/files");
var modeService_1 = require("vs/editor/common/services/modeService");
var untitledEditorService_1 = require("vs/workbench/services/untitled/common/untitledEditorService");
var extHost_protocol_1 = require("./extHost.protocol");
var resolverService_1 = require("vs/editor/common/services/resolverService");
var codeEditorService_1 = require("vs/editor/common/services/codeEditorService");
var MainThreadDocuments = (function (_super) {
    __extends(MainThreadDocuments, _super);
    function MainThreadDocuments(threadService, modelService, modeService, eventService, textFileService, codeEditorService, fileService, textModelResolverService, untitledEditorService) {
        var _this = _super.call(this) || this;
        _this._modelService = modelService;
        _this._modeService = modeService;
        _this._textModelResolverService = textModelResolverService;
        _this._textFileService = textFileService;
        _this._codeEditorService = codeEditorService;
        _this._fileService = fileService;
        _this._untitledEditorService = untitledEditorService;
        _this._proxy = threadService.get(extHost_protocol_1.ExtHostContext.ExtHostDocuments);
        _this._modelIsSynced = {};
        _this._toDispose = [];
        modelService.onModelAdded(_this._onModelAdded, _this, _this._toDispose);
        modelService.onModelRemoved(_this._onModelRemoved, _this, _this._toDispose);
        modelService.onModelModeChanged(_this._onModelModeChanged, _this, _this._toDispose);
        _this._toDispose.push(textFileService.models.onModelSaved(function (e) {
            if (_this._shouldHandleFileEvent(e)) {
                _this._proxy.$acceptModelSaved(e.resource.toString());
            }
        }));
        _this._toDispose.push(textFileService.models.onModelReverted(function (e) {
            if (_this._shouldHandleFileEvent(e)) {
                _this._proxy.$acceptModelReverted(e.resource.toString());
            }
        }));
        _this._toDispose.push(textFileService.models.onModelDirty(function (e) {
            if (_this._shouldHandleFileEvent(e)) {
                _this._proxy.$acceptModelDirty(e.resource.toString());
            }
        }));
        _this._modelToDisposeMap = Object.create(null);
        _this._resourceContentProvider = Object.create(null);
        return _this;
    }
    MainThreadDocuments.prototype.dispose = function () {
        var _this = this;
        Object.keys(this._modelToDisposeMap).forEach(function (modelUrl) {
            _this._modelToDisposeMap[modelUrl].dispose();
        });
        this._modelToDisposeMap = Object.create(null);
        this._toDispose = lifecycle_1.dispose(this._toDispose);
    };
    MainThreadDocuments.prototype._shouldHandleFileEvent = function (e) {
        var model = this._modelService.getModel(e.resource);
        return model && !model.isTooLargeForHavingARichMode();
    };
    MainThreadDocuments.prototype._onModelAdded = function (model) {
        var _this = this;
        // Same filter as in mainThreadEditorsTracker
        if (model.isTooLargeForHavingARichMode()) {
            // don't synchronize too large models
            return null;
        }
        var modelUrl = model.uri;
        this._modelIsSynced[modelUrl.toString()] = true;
        this._modelToDisposeMap[modelUrl.toString()] = model.addBulkListener(function (events) { return _this._onModelEvents(modelUrl, events); });
        this._proxy.$acceptModelAdd({
            url: model.uri,
            versionId: model.getVersionId(),
            value: model.toRawText(),
            modeId: model.getMode().getId(),
            isDirty: this._textFileService.isDirty(modelUrl)
        });
    };
    MainThreadDocuments.prototype._onModelModeChanged = function (event) {
        var model = event.model, oldModeId = event.oldModeId;
        var modelUrl = model.uri;
        if (!this._modelIsSynced[modelUrl.toString()]) {
            return;
        }
        this._proxy.$acceptModelModeChanged(model.uri.toString(), oldModeId, model.getMode().getId());
    };
    MainThreadDocuments.prototype._onModelRemoved = function (model) {
        var modelUrl = model.uri;
        if (!this._modelIsSynced[modelUrl.toString()]) {
            return;
        }
        delete this._modelIsSynced[modelUrl.toString()];
        this._modelToDisposeMap[modelUrl.toString()].dispose();
        delete this._modelToDisposeMap[modelUrl.toString()];
        this._proxy.$acceptModelRemoved(modelUrl.toString());
    };
    MainThreadDocuments.prototype._onModelEvents = function (modelUrl, events) {
        var changedEvents = [];
        for (var i = 0, len = events.length; i < len; i++) {
            var e = events[i];
            switch (e.getType()) {
                case editorCommon.EventType.ModelContentChanged2:
                    changedEvents.push(e.getData());
                    break;
            }
        }
        if (changedEvents.length > 0) {
            this._proxy.$acceptModelChanged(modelUrl.toString(), changedEvents, this._textFileService.isDirty(modelUrl));
        }
    };
    // --- from extension host process
    MainThreadDocuments.prototype.$trySaveDocument = function (uri) {
        return this._textFileService.save(uri);
    };
    MainThreadDocuments.prototype.$tryOpenDocument = function (uri) {
        if (!uri.scheme || !(uri.fsPath || uri.authority)) {
            return winjs_base_1.TPromise.wrapError("Invalid uri. Scheme and authority or path must be set.");
        }
        var promise;
        switch (uri.scheme) {
            case 'untitled':
                promise = this._handleUnititledScheme(uri);
                break;
            case 'file':
            default:
                promise = this._handleAsResourceInput(uri);
                break;
        }
        return promise.then(function (success) {
            if (!success) {
                return winjs_base_1.TPromise.wrapError('cannot open ' + uri.toString());
            }
        }, function (err) {
            return winjs_base_1.TPromise.wrapError('cannot open ' + uri.toString() + '. Detail: ' + errorMessage_1.toErrorMessage(err));
        });
    };
    MainThreadDocuments.prototype._handleAsResourceInput = function (uri) {
        return this._textModelResolverService.createModelReference(uri).then(function (ref) {
            var result = !!ref.object;
            // TODO@Joao TODO@Joh when should this model reference be disposed?
            // ref.dispose();
            return result;
        });
    };
    MainThreadDocuments.prototype._handleUnititledScheme = function (uri) {
        var _this = this;
        var asFileUri = uri_1.default.file(uri.fsPath);
        return this._fileService.resolveFile(asFileUri).then(function (stats) {
            // don't create a new file ontop of an existing file
            return winjs_base_1.TPromise.wrapError('file already exists on disk');
        }, function (err) {
            var input = _this._untitledEditorService.createOrGet(asFileUri);
            return input.resolve(true).then(function (model) {
                if (input.getResource().toString() !== uri.toString()) {
                    throw new Error("expected URI " + uri.toString() + " BUT GOT " + input.getResource().toString());
                }
                if (!_this._modelIsSynced[uri.toString()]) {
                    throw new Error("expected URI " + uri.toString() + " to have come to LIFE");
                }
                return _this._proxy.$acceptModelDirty(uri.toString()); // mark as dirty
            }).then(function () {
                return true;
            });
        });
    };
    // --- virtual document logic
    MainThreadDocuments.prototype.$registerTextContentProvider = function (handle, scheme) {
        var _this = this;
        this._resourceContentProvider[handle] = this._textModelResolverService.registerTextModelContentProvider(scheme, {
            provideTextContent: function (uri) {
                return _this._proxy.$provideTextDocumentContent(handle, uri).then(function (value) {
                    if (typeof value === 'string') {
                        var firstLineText = value.substr(0, 1 + value.search(/\r?\n/));
                        var mode = _this._modeService.getOrCreateModeByFilenameOrFirstLine(uri.fsPath, firstLineText);
                        return _this._modelService.createModel(value, mode, uri);
                    }
                });
            }
        });
    };
    MainThreadDocuments.prototype.$unregisterTextContentProvider = function (handle) {
        var registration = this._resourceContentProvider[handle];
        if (registration) {
            registration.dispose();
            delete this._resourceContentProvider[handle];
        }
    };
    MainThreadDocuments.prototype.$onVirtualDocumentChange = function (uri, value) {
        var model = this._modelService.getModel(uri);
        if (model) {
            model.setValue(value);
        }
    };
    return MainThreadDocuments;
}(extHost_protocol_1.MainThreadDocumentsShape));
MainThreadDocuments = __decorate([
    __param(0, threadService_1.IThreadService),
    __param(1, modelService_1.IModelService),
    __param(2, modeService_1.IModeService),
    __param(3, event_1.IEventService),
    __param(4, textfiles_1.ITextFileService),
    __param(5, codeEditorService_1.ICodeEditorService),
    __param(6, files_1.IFileService),
    __param(7, resolverService_1.ITextModelResolverService),
    __param(8, untitledEditorService_1.IUntitledEditorService)
], MainThreadDocuments);
exports.MainThreadDocuments = MainThreadDocuments;
