// vscode/src/vs/workbench/api/node/mainThreadDocuments.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var uri_1 = require("vs/base/common/uri");
var errorMessage_1 = require("vs/base/common/errorMessage");
var async_1 = require("vs/base/common/async");
var modelService_1 = require("vs/editor/common/services/modelService");
var threadService_1 = require("vs/workbench/services/thread/common/threadService");
var lifecycle_1 = require("vs/base/common/lifecycle");
var textfiles_1 = require("vs/workbench/services/textfile/common/textfiles");
var winjs_base_1 = require("vs/base/common/winjs.base");
var files_1 = require("vs/platform/files/common/files");
var modeService_1 = require("vs/editor/common/services/modeService");
var untitledEditorService_1 = require("vs/workbench/services/untitled/common/untitledEditorService");
var groupService_1 = require("vs/workbench/services/group/common/groupService");
var extHost_protocol_1 = require("./extHost.protocol");
var resolverService_1 = require("vs/editor/common/services/resolverService");
var codeEditorService_1 = require("vs/editor/common/services/codeEditorService");
var editorCommon = require("vs/editor/common/editorCommon");
var TimeoutReference = (function () {
    function TimeoutReference(codeEditorService, editorGroupService, reference) {
        var _this = this;
        this.codeEditorService = codeEditorService;
        this.editorGroupService = editorGroupService;
        this.reference = reference;
        this._disposed = false;
        var check = function () {
            if (!_this.isUsed()) {
                _this.dispose();
            }
            else {
                _this._timer = async_1.setDisposableTimeout(check, TimeoutReference._delay);
            }
        };
        this._timer = async_1.setDisposableTimeout(check, TimeoutReference._delay);
    }
    TimeoutReference.prototype.dispose = function () {
        if (!this._disposed) {
            this._disposed = true;
            lifecycle_1.dispose(this.reference, this._timer);
        }
    };
    TimeoutReference.prototype.isUsed = function () {
        for (var _i = 0, _a = this.codeEditorService.listCodeEditors(); _i < _a.length; _i++) {
            var editor = _a[_i];
            if (editor.getModel() === this.reference.object.textEditorModel) {
                return true;
            }
        }
        for (var _b = 0, _c = this.editorGroupService.getStacksModel().groups; _b < _c.length; _b++) {
            var group = _c[_b];
            if (group.contains(this.reference.object.textEditorModel.uri)) {
                return true;
            }
        }
        return false;
    };
    return TimeoutReference;
}());
TimeoutReference._delay = 1000 * 60 * 3;
var MainThreadDocuments = (function (_super) {
    __extends(MainThreadDocuments, _super);
    function MainThreadDocuments(documentsAndEditors, threadService, modelService, modeService, textFileService, codeEditorService, fileService, textModelResolverService, untitledEditorService, editorGroupService) {
        var _this = _super.call(this) || this;
        _this._modelService = modelService;
        _this._modeService = modeService;
        _this._textModelResolverService = textModelResolverService;
        _this._textFileService = textFileService;
        _this._codeEditorService = codeEditorService;
        _this._fileService = fileService;
        _this._untitledEditorService = untitledEditorService;
        _this._editorGroupService = editorGroupService;
        _this._proxy = threadService.get(extHost_protocol_1.ExtHostContext.ExtHostDocuments);
        _this._modelIsSynced = {};
        _this._toDispose = [];
        _this._toDispose.push(documentsAndEditors.onDocumentAdd(function (models) { return models.forEach(_this._onModelAdded, _this); }));
        _this._toDispose.push(documentsAndEditors.onDocumentRemove(function (urls) { return urls.forEach(_this._onModelRemoved, _this); }));
        modelService.onModelModeChanged(_this._onModelModeChanged, _this, _this._toDispose);
        _this._toDispose.push(textFileService.models.onModelSaved(function (e) {
            if (_this._shouldHandleFileEvent(e)) {
                _this._proxy.$acceptModelSaved(e.resource.toString());
            }
        }));
        _this._toDispose.push(textFileService.models.onModelReverted(function (e) {
            if (_this._shouldHandleFileEvent(e)) {
                _this._proxy.$acceptModelReverted(e.resource.toString());
            }
        }));
        _this._toDispose.push(textFileService.models.onModelDirty(function (e) {
            if (_this._shouldHandleFileEvent(e)) {
                _this._proxy.$acceptModelDirty(e.resource.toString());
            }
        }));
        _this._modelToDisposeMap = Object.create(null);
        _this._resourceContentProvider = Object.create(null);
        return _this;
    }
    MainThreadDocuments.prototype.dispose = function () {
        var _this = this;
        Object.keys(this._modelToDisposeMap).forEach(function (modelUrl) {
            _this._modelToDisposeMap[modelUrl].dispose();
        });
        this._modelToDisposeMap = Object.create(null);
        this._toDispose = lifecycle_1.dispose(this._toDispose);
    };
    MainThreadDocuments.prototype._shouldHandleFileEvent = function (e) {
        var model = this._modelService.getModel(e.resource);
        return model && !model.isTooLargeForHavingARichMode();
    };
    MainThreadDocuments.prototype._onModelAdded = function (model) {
        var _this = this;
        // Same filter as in mainThreadEditorsTracker
        if (model.isTooLargeForHavingARichMode()) {
            // don't synchronize too large models
            return null;
        }
        var modelUrl = model.uri;
        this._modelIsSynced[modelUrl.toString()] = true;
        this._modelToDisposeMap[modelUrl.toString()] = model.addBulkListener(function (events) { return _this._onModelEvents(modelUrl, events); });
    };
    MainThreadDocuments.prototype._onModelModeChanged = function (event) {
        var model = event.model, oldModeId = event.oldModeId;
        var modelUrl = model.uri;
        if (!this._modelIsSynced[modelUrl.toString()]) {
            return;
        }
        this._proxy.$acceptModelModeChanged(model.uri.toString(), oldModeId, model.getLanguageIdentifier().language);
    };
    MainThreadDocuments.prototype._onModelRemoved = function (modelUrl) {
        if (!this._modelIsSynced[modelUrl]) {
            return;
        }
        delete this._modelIsSynced[modelUrl];
        this._modelToDisposeMap[modelUrl].dispose();
        delete this._modelToDisposeMap[modelUrl];
    };
    MainThreadDocuments.prototype._onModelEvents = function (modelUrl, events) {
        var changedEvents = [];
        for (var i = 0, len = events.length; i < len; i++) {
            var e = events[i];
            switch (e.getType()) {
                case editorCommon.EventType.ModelContentChanged2:
                    changedEvents.push(e.getData());
                    break;
            }
        }
        if (changedEvents.length > 0) {
            this._proxy.$acceptModelChanged(modelUrl.toString(), changedEvents, this._textFileService.isDirty(modelUrl));
        }
    };
    // --- from extension host process
    MainThreadDocuments.prototype.$trySaveDocument = function (uri) {
        return this._textFileService.save(uri);
    };
    MainThreadDocuments.prototype.$tryDeleteDocument = function (uri) {
        return this._textFileService.delete(uri);
    };
    MainThreadDocuments.prototype.$tryRevertAll = function (uris) {
        return this._textFileService.revertAll(uris, { force: true }).then(function () { return true; }, function () { return false; });
    };
    MainThreadDocuments.prototype.$tryOpenDocument = function (uri) {
        if (!uri.scheme || !(uri.fsPath || uri.authority)) {
            return winjs_base_1.TPromise.wrapError("Invalid uri. Scheme and authority or path must be set.");
        }
        var promise;
        switch (uri.scheme) {
            case 'untitled':
                promise = this._handleUnititledScheme(uri);
                break;
            case 'file':
            default:
                promise = this._handleAsResourceInput(uri);
                break;
        }
        return promise.then(function (success) {
            if (!success) {
                return winjs_base_1.TPromise.wrapError('cannot open ' + uri.toString());
            }
            return undefined;
        }, function (err) {
            return winjs_base_1.TPromise.wrapError('cannot open ' + uri.toString() + '. Detail: ' + errorMessage_1.toErrorMessage(err));
        });
    };
    MainThreadDocuments.prototype.$tryCreateDocument = function (options) {
        return this._doCreateUntitled(void 0, options ? options.language : void 0, options ? options.content : void 0);
    };
    MainThreadDocuments.prototype._handleAsResourceInput = function (uri) {
        var _this = this;
        return this._textModelResolverService.createModelReference(uri).then(function (ref) {
            // TimeoutReference will check every 3 min if the
            // reference is still in use. This is quite harsh to
            // extensions but we don't want them to make us hold
            // on to model indefinitely
            _this._toDispose.push(new TimeoutReference(_this._codeEditorService, _this._editorGroupService, ref));
            var result = !!ref.object;
            return result;
        });
    };
    MainThreadDocuments.prototype._handleUnititledScheme = function (uri) {
        var _this = this;
        var asFileUri = uri_1.default.file(uri.fsPath);
        return this._fileService.resolveFile(asFileUri).then(function (stats) {
            // don't create a new file ontop of an existing file
            return winjs_base_1.TPromise.wrapError('file already exists on disk');
        }, function (err) { return _this._doCreateUntitled(asFileUri).then(function (resource) { return !!resource; }); });
    };
    MainThreadDocuments.prototype._doCreateUntitled = function (uri, modeId, initialValue) {
        var _this = this;
        var input = this._untitledEditorService.createOrGet(uri, modeId, initialValue);
        return input.resolve(true).then(function (model) {
            if (!_this._modelIsSynced[input.getResource().toString()]) {
                throw new Error("expected URI " + input.getResource().toString() + " to have come to LIFE");
            }
            return _this._proxy.$acceptModelDirty(input.getResource().toString()); // mark as dirty
        }).then(function () {
            return input.getResource();
        });
    };
    // --- virtual document logic
    MainThreadDocuments.prototype.$registerTextContentProvider = function (handle, scheme) {
        var _this = this;
        this._resourceContentProvider[handle] = this._textModelResolverService.registerTextModelContentProvider(scheme, {
            provideTextContent: function (uri) {
                return _this._proxy.$provideTextDocumentContent(handle, uri).then(function (value) {
                    if (typeof value === 'string') {
                        var firstLineText = value.substr(0, 1 + value.search(/\r?\n/));
                        var mode = _this._modeService.getOrCreateModeByFilenameOrFirstLine(uri.fsPath, firstLineText);
                        return _this._modelService.createModel(value, mode, uri);
                    }
                    return undefined;
                });
            }
        });
    };
    MainThreadDocuments.prototype.$unregisterTextContentProvider = function (handle) {
        var registration = this._resourceContentProvider[handle];
        if (registration) {
            registration.dispose();
            delete this._resourceContentProvider[handle];
        }
    };
    MainThreadDocuments.prototype.$onVirtualDocumentChange = function (uri, value) {
        var model = this._modelService.getModel(uri);
        if (!model) {
            return;
        }
        var raw = {
            lines: value.lines,
            length: value.length,
            BOM: value.BOM,
            EOL: value.EOL,
            containsRTL: value.containsRTL,
            isBasicASCII: value.isBasicASCII,
        };
        if (!model.equals(raw)) {
            model.setValueFromTextSource(raw);
        }
    };
    return MainThreadDocuments;
}(extHost_protocol_1.MainThreadDocumentsShape));
MainThreadDocuments = __decorate([
    __param(1, threadService_1.IThreadService),
    __param(2, modelService_1.IModelService),
    __param(3, modeService_1.IModeService),
    __param(4, textfiles_1.ITextFileService),
    __param(5, codeEditorService_1.ICodeEditorService),
    __param(6, files_1.IFileService),
    __param(7, resolverService_1.ITextModelResolverService),
    __param(8, untitledEditorService_1.IUntitledEditorService),
    __param(9, groupService_1.IEditorGroupService)
], MainThreadDocuments);
exports.MainThreadDocuments = MainThreadDocuments;
