// vscode/src/vs/editor/browser/viewParts/currentLineMarginHighlight/currentLineMarginHighlight.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
require("./currentLineMarginHighlight.css");
var dynamicViewOverlay_1 = require("vs/editor/browser/view/dynamicViewOverlay");
var CurrentLineMarginHighlightOverlay = (function (_super) {
    __extends(CurrentLineMarginHighlightOverlay, _super);
    function CurrentLineMarginHighlightOverlay(context, layoutProvider) {
        var _this = _super.call(this) || this;
        _this._context = context;
        _this._lineHeight = _this._context.configuration.editor.lineHeight;
        _this._renderLineHighlight = _this._context.configuration.editor.viewInfo.renderLineHighlight;
        _this._layoutProvider = layoutProvider;
        _this._primaryCursorIsInEditableRange = true;
        _this._primaryCursorLineNumber = 1;
        _this._contentLeft = _this._context.configuration.editor.layoutInfo.contentLeft;
        _this._context.addEventHandler(_this);
        return _this;
    }
    CurrentLineMarginHighlightOverlay.prototype.dispose = function () {
        this._context.removeEventHandler(this);
        this._context = null;
    };
    // --- begin event handlers
    CurrentLineMarginHighlightOverlay.prototype.onModelFlushed = function () {
        this._primaryCursorIsInEditableRange = true;
        this._primaryCursorLineNumber = 1;
        return true;
    };
    CurrentLineMarginHighlightOverlay.prototype.onModelLinesDeleted = function (e) {
        return true;
    };
    CurrentLineMarginHighlightOverlay.prototype.onModelLinesInserted = function (e) {
        return true;
    };
    CurrentLineMarginHighlightOverlay.prototype.onCursorPositionChanged = function (e) {
        var hasChanged = false;
        if (this._primaryCursorIsInEditableRange !== e.isInEditableRange) {
            this._primaryCursorIsInEditableRange = e.isInEditableRange;
            hasChanged = true;
        }
        if (this._primaryCursorLineNumber !== e.position.lineNumber) {
            this._primaryCursorLineNumber = e.position.lineNumber;
            hasChanged = true;
        }
        return hasChanged;
    };
    CurrentLineMarginHighlightOverlay.prototype.onConfigurationChanged = function (e) {
        if (e.lineHeight) {
            this._lineHeight = this._context.configuration.editor.lineHeight;
        }
        if (e.viewInfo.renderLineHighlight) {
            this._renderLineHighlight = this._context.configuration.editor.viewInfo.renderLineHighlight;
        }
        if (e.layoutInfo) {
            this._contentLeft = this._context.configuration.editor.layoutInfo.contentLeft;
        }
        return true;
    };
    CurrentLineMarginHighlightOverlay.prototype.onLayoutChanged = function (layoutInfo) {
        return true;
    };
    CurrentLineMarginHighlightOverlay.prototype.onZonesChanged = function () {
        return true;
    };
    // --- end event handlers
    CurrentLineMarginHighlightOverlay.prototype.prepareRender = function (ctx) {
    };
    CurrentLineMarginHighlightOverlay.prototype.render = function (startLineNumber, lineNumber) {
        if (lineNumber === this._primaryCursorLineNumber) {
            if (this._shouldShowCurrentLine()) {
                return ('<div class="current-line" style="width:'
                    + String(this._contentLeft)
                    + 'px; height:'
                    + String(this._lineHeight)
                    + 'px;"></div>');
            }
            else {
                return '';
            }
        }
        return '';
    };
    CurrentLineMarginHighlightOverlay.prototype._shouldShowCurrentLine = function () {
        return (this._renderLineHighlight === 'gutter' || this._renderLineHighlight === 'all') && this._primaryCursorIsInEditableRange;
    };
    return CurrentLineMarginHighlightOverlay;
}(dynamicViewOverlay_1.DynamicViewOverlay));
exports.CurrentLineMarginHighlightOverlay = CurrentLineMarginHighlightOverlay;
