// vscode/src/vs/platform/platform.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var Types = require('vs/base/common/types');
var Assert = require('vs/base/common/assert');
var RegistryImpl = (function () {
    function RegistryImpl() {
        this.data = {};
    }
    RegistryImpl.prototype.add = function (id, data) {
        Assert.ok(Types.isString(id));
        Assert.ok(Types.isObject(data));
        Assert.ok(!this.data.hasOwnProperty(id), 'There is already an extension with this id');
        this.data[id] = data;
    };
    RegistryImpl.prototype.knows = function (id) {
        return this.data.hasOwnProperty(id);
    };
    RegistryImpl.prototype.as = function (id) {
        return this.data[id] || null;
    };
    return RegistryImpl;
}());
exports.Registry = new RegistryImpl();
/**
 * A base class for registries that leverage the instantiation service to create instances.
 */
var BaseRegistry = (function () {
    function BaseRegistry() {
        this.toBeInstantiated = [];
        this.instances = [];
    }
    BaseRegistry.prototype.setInstantiationService = function (service) {
        this.instantiationService = service;
        while (this.toBeInstantiated.length > 0) {
            var entry = this.toBeInstantiated.shift();
            this.instantiate(entry);
        }
    };
    BaseRegistry.prototype.instantiate = function (ctor) {
        var instance = this.instantiationService.createInstance(ctor);
        this.instances.push(instance);
    };
    BaseRegistry.prototype._register = function (ctor) {
        if (this.instantiationService) {
            this.instantiate(ctor);
        }
        else {
            this.toBeInstantiated.push(ctor);
        }
    };
    BaseRegistry.prototype._getInstances = function () {
        return this.instances.slice(0);
    };
    BaseRegistry.prototype._setInstances = function (instances) {
        this.instances = instances;
    };
    return BaseRegistry;
}());
exports.BaseRegistry = BaseRegistry;
