// vscode/src/vs/base/node/config.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var fs = require("fs");
var path = require("path");
var objects = require("vs/base/common/objects");
var lifecycle_1 = require("vs/base/common/lifecycle");
var event_1 = require("vs/base/common/event");
var json = require("vs/base/common/json");
/**
 * A simple helper to watch a configured file for changes and process its contents as JSON object.
 * Supports:
 * - comments in JSON files and errors
 * - symlinks for the config file itself
 * - delayed processing of changes to accomodate for lots of changes
 * - configurable defaults
 */
var ConfigWatcher = (function () {
    function ConfigWatcher(_path, options) {
        if (options === void 0) { options = { changeBufferDelay: 0, defaultConfig: Object.create(null) }; }
        this._path = _path;
        this.options = options;
        this.disposables = [];
        this._onDidUpdateConfiguration = new event_1.Emitter();
        this.disposables.push(this._onDidUpdateConfiguration);
        this.registerWatcher();
        this.initAsync();
    }
    Object.defineProperty(ConfigWatcher.prototype, "path", {
        get: function () {
            return this._path;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ConfigWatcher.prototype, "hasParseErrors", {
        get: function () {
            return this.parseErrors && this.parseErrors.length > 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ConfigWatcher.prototype, "onDidUpdateConfiguration", {
        get: function () {
            return this._onDidUpdateConfiguration.event;
        },
        enumerable: true,
        configurable: true
    });
    ConfigWatcher.prototype.initAsync = function () {
        var _this = this;
        this.loadAsync(function (config) {
            if (!_this.loaded) {
                _this.updateCache(config); // prevent race condition if config was loaded sync already
            }
        });
    };
    ConfigWatcher.prototype.updateCache = function (value) {
        this.cache = value;
        this.loaded = true;
    };
    ConfigWatcher.prototype.loadSync = function () {
        try {
            return this.parse(fs.readFileSync(this._path).toString());
        }
        catch (error) {
            return this.options.defaultConfig;
        }
    };
    ConfigWatcher.prototype.loadAsync = function (callback) {
        var _this = this;
        fs.readFile(this._path, function (error, raw) {
            if (error) {
                return callback(_this.options.defaultConfig);
            }
            return callback(_this.parse(raw.toString()));
        });
    };
    ConfigWatcher.prototype.parse = function (raw) {
        var res;
        try {
            this.parseErrors = [];
            res = json.parse(raw, this.parseErrors);
        }
        catch (error) {
        }
        return res || this.options.defaultConfig;
    };
    ConfigWatcher.prototype.registerWatcher = function () {
        var _this = this;
        // Watch the parent of the path so that we detect ADD and DELETES
        var parentFolder = path.dirname(this._path);
        this.watch(parentFolder);
        // Check if the path is a symlink and watch its target if so
        fs.lstat(this._path, function (err, stat) {
            if (err || stat.isDirectory()) {
                return; // path is not a valid file
            }
            // We found a symlink
            if (stat.isSymbolicLink()) {
                fs.readlink(_this._path, function (err, realPath) {
                    if (err) {
                        return; // path is not a valid symlink
                    }
                    _this.watch(realPath);
                });
            }
        });
    };
    ConfigWatcher.prototype.watch = function (path) {
        var _this = this;
        if (this.disposed) {
            return; // avoid watchers that will never get disposed by checking for being disposed
        }
        try {
            var watcher_1 = fs.watch(path);
            watcher_1.on('change', function () { return _this.onConfigFileChange(); });
            this.disposables.push(lifecycle_1.toDisposable(function () {
                watcher_1.removeAllListeners();
                watcher_1.close();
            }));
        }
        catch (error) {
            fs.exists(path, function (exists) {
                if (exists) {
                    console.warn("Failed to watch " + path + " for configuration changes (" + error.toString() + ")");
                }
            });
        }
    };
    ConfigWatcher.prototype.onConfigFileChange = function () {
        var _this = this;
        if (this.timeoutHandle) {
            global.clearTimeout(this.timeoutHandle);
            this.timeoutHandle = null;
        }
        // we can get multiple change events for one change, so we buffer through a timeout
        this.timeoutHandle = global.setTimeout(function () { return _this.reload(); }, this.options.changeBufferDelay);
    };
    ConfigWatcher.prototype.reload = function (callback) {
        var _this = this;
        this.loadAsync(function (currentConfig) {
            if (!objects.equals(currentConfig, _this.cache)) {
                _this.updateCache(currentConfig);
                _this._onDidUpdateConfiguration.fire({ config: _this.cache });
            }
            if (callback) {
                return callback(currentConfig);
            }
        });
    };
    ConfigWatcher.prototype.getConfig = function () {
        this.ensureLoaded();
        return this.cache;
    };
    ConfigWatcher.prototype.getValue = function (key, fallback) {
        this.ensureLoaded();
        if (!key) {
            return fallback;
        }
        var value = this.cache ? this.cache[key] : void 0;
        return typeof value !== 'undefined' ? value : fallback;
    };
    ConfigWatcher.prototype.ensureLoaded = function () {
        if (!this.loaded) {
            this.updateCache(this.loadSync());
        }
    };
    ConfigWatcher.prototype.dispose = function () {
        this.disposed = true;
        this.disposables = lifecycle_1.dispose(this.disposables);
    };
    return ConfigWatcher;
}());
exports.ConfigWatcher = ConfigWatcher;
