// vscode/src/vs/base/node/mime.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var paths = require('path');
var mime = require('vs/base/common/mime');
var stream = require('vs/base/node/stream');
var encoding = require('vs/base/node/encoding');
/**
 * Lots of binary file types exists where the type can be determined by matching the first few bytes against some "magic patterns".
 * E.g. PDF files always start with %PDF- and the rest of the file contains mostly text, but sometimes binary data (for fonts and images).
 * In order to detect these types correctly (and independently from the file's extension), the content base mime type detection must be performed
 * on any file, not only on text files.
 *
 * Here is the original mime type detection in pseudocode:
 *
 * let mimes = [];
 *
 * read file extension
 *
 * if (file extension matches) {
 * 	if (file extension is bogus) {
 * 		// ignore.
 * 		// this covers *.manifest files which can contain arbitrary content, so the extension is of no value.
 * 		// a consequence of this is that the content based mime type becomes the most specific type in the array
 * 	} else {
 * 		mimes.push(associated mime type)	  // first element: most specific
 * 	}
 * }
 *
 * read file contents
 *
 * if (content based match found) {	// this is independent from text or binary
 * 	mimes.push(associated mime type)
 * 	if (a second mime exists for the match) {   // should be rare; text/plain should never be included here
 * 		// e.g. for svg: ['image/svg+xml', 'application/xml']
 * 		mimes.push(second mime)
 * 	}
 * }
 *
 * if (content == text)
 * 	mimes.push('text/plain')   // last element: least specific
 * else
 * 	mimes.push('application/octet-stream')    // last element: least specific
 */
var BUFFER_READ_MAX_LEN = 512; // max buffer len to use when detecting encoding/mime
var ASAR_EXT = '.asar';
function doDetectMimesFromStream(instream, callback) {
    stream.readExactlyByStream(instream, BUFFER_READ_MAX_LEN, function (err, buffer, bytesRead) {
        handleReadResult(err, buffer, bytesRead, callback);
    });
}
function doDetectMimesFromFile(absolutePath, callback) {
    if (paths.extname(absolutePath) === ASAR_EXT) {
        return callback(null, { encoding: encoding.UTF8, mimes: [mime.MIME_BINARY] }); // https://github.com/Microsoft/vscode/issues/646
    }
    stream.readExactlyByFile(absolutePath, BUFFER_READ_MAX_LEN, function (err, buffer, bytesRead) {
        handleReadResult(err, buffer, bytesRead, callback);
    });
}
function handleReadResult(err, buffer, bytesRead, callback) {
    if (err) {
        return callback(err, null);
    }
    return callback(null, detectMimeAndEncodingFromBuffer(buffer, bytesRead));
}
function detectMimeAndEncodingFromBuffer(buffer, bytesRead) {
    var enc = encoding.detectEncodingByBOMFromBuffer(buffer, bytesRead);
    // Detect 0 bytes to see if file is binary (ignore for UTF 16 though)
    var isText = true;
    if (enc !== encoding.UTF16be && enc !== encoding.UTF16le) {
        for (var i = 0; i < bytesRead && i < BUFFER_READ_MAX_LEN; i++) {
            if (buffer.readInt8(i) === 0) {
                isText = false;
                break;
            }
        }
    }
    return {
        mimes: isText ? [mime.MIME_TEXT] : [mime.MIME_BINARY],
        encoding: enc
    };
}
exports.detectMimeAndEncodingFromBuffer = detectMimeAndEncodingFromBuffer;
function filterAndSortMimes(detectedMimes, guessedMimes) {
    var mimes = detectedMimes;
    // Add extension based mime as first element as this is the desire of whoever created the file.
    // Never care about application/octet-stream or application/unknown as guessed mime, as this is the fallback of the guess which is never accurate
    var guessedMime = guessedMimes[0];
    if (guessedMime !== mime.MIME_BINARY && guessedMime !== mime.MIME_UNKNOWN) {
        mimes.unshift(guessedMime);
    }
    // Remove duplicate elements from array and sort unspecific mime to the end
    var uniqueSortedMimes = mimes.filter(function (element, position) {
        return element && mimes.indexOf(element) === position;
    }).sort(function (mimeA, mimeB) {
        if (mimeA === mime.MIME_BINARY) {
            return 1;
        }
        if (mimeB === mime.MIME_BINARY) {
            return -1;
        }
        if (mimeA === mime.MIME_TEXT) {
            return 1;
        }
        if (mimeB === mime.MIME_TEXT) {
            return -1;
        }
        return 0;
    });
    return uniqueSortedMimes;
}
/**
 * Opens the given stream to detect its mime type. Returns an array of mime types sorted from most specific to unspecific.
 * @param instream the readable stream to detect the mime types from.
 * @param nameHint an additional hint that can be used to detect a mime from a file extension.
 */
function detectMimesFromStream(instream, nameHint, callback) {
    doDetectMimesFromStream(instream, function (error, result) {
        handleMimeResult(nameHint, error, result, callback);
    });
}
exports.detectMimesFromStream = detectMimesFromStream;
/**
 * Opens the given file to detect its mime type. Returns an array of mime types sorted from most specific to unspecific.
 * @param absolutePath the absolute path of the file.
 */
function detectMimesFromFile(absolutePath, callback) {
    doDetectMimesFromFile(absolutePath, function (error, result) {
        handleMimeResult(absolutePath, error, result, callback);
    });
}
exports.detectMimesFromFile = detectMimesFromFile;
function handleMimeResult(nameHint, error, result, callback) {
    if (error) {
        return callback(error, null);
    }
    var filterAndSortedMimes = filterAndSortMimes(result.mimes, mime.guessMimeTypes(nameHint));
    result.mimes = filterAndSortedMimes;
    callback(null, result);
}
