// vscode/src/vs/base/browser/ui/countBadge/countBadge.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
require('./countBadge.css');
var builder_1 = require('vs/base/browser/builder');
var strings = require('vs/base/common/strings');
var CountBadge = (function () {
    function CountBadge(container, count, titleFormat) {
        this.$el = builder_1.$('.monaco-count-badge').appendTo(container);
        this.titleFormat = titleFormat || '';
        this.setCount(count || 0);
    }
    CountBadge.prototype.setCount = function (count) {
        this.count = count;
        this.render();
    };
    CountBadge.prototype.setTitleFormat = function (titleFormat) {
        this.titleFormat = titleFormat;
        this.render();
    };
    CountBadge.prototype.render = function () {
        this.$el.text('' + this.count);
        this.$el.title(strings.format(this.titleFormat, this.count));
    };
    CountBadge.prototype.dispose = function () {
        if (this.$el) {
            this.$el.destroy();
            this.$el = null;
        }
    };
    return CountBadge;
}());
exports.CountBadge = CountBadge;
