// vscode/src/vs/editor/common/services/abstractCodeEditorService.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var event_1 = require("vs/base/common/event");
var AbstractCodeEditorService = (function () {
    function AbstractCodeEditorService() {
        this._codeEditors = Object.create(null);
        this._onCodeEditorAdd = new event_1.Emitter();
        this._onCodeEditorRemove = new event_1.Emitter();
    }
    AbstractCodeEditorService.prototype.addCodeEditor = function (editor) {
        this._codeEditors[editor.getId()] = editor;
        this._onCodeEditorAdd.fire(editor);
    };
    Object.defineProperty(AbstractCodeEditorService.prototype, "onCodeEditorAdd", {
        get: function () {
            return this._onCodeEditorAdd.event;
        },
        enumerable: true,
        configurable: true
    });
    AbstractCodeEditorService.prototype.removeCodeEditor = function (editor) {
        if (delete this._codeEditors[editor.getId()]) {
            this._onCodeEditorRemove.fire(editor);
        }
    };
    Object.defineProperty(AbstractCodeEditorService.prototype, "onCodeEditorRemove", {
        get: function () {
            return this._onCodeEditorRemove.event;
        },
        enumerable: true,
        configurable: true
    });
    AbstractCodeEditorService.prototype.getCodeEditor = function (editorId) {
        return this._codeEditors[editorId] || null;
    };
    AbstractCodeEditorService.prototype.listCodeEditors = function () {
        var _this = this;
        return Object.keys(this._codeEditors).map(function (id) { return _this._codeEditors[id]; });
    };
    AbstractCodeEditorService.prototype.getFocusedCodeEditor = function () {
        var editorWithWidgetFocus = null;
        var editors = this.listCodeEditors();
        for (var i = 0; i < editors.length; i++) {
            var editor = editors[i];
            if (editor.isFocused()) {
                // bingo!
                return editor;
            }
            if (editor.hasWidgetFocus()) {
                editorWithWidgetFocus = editor;
            }
        }
        return editorWithWidgetFocus;
    };
    return AbstractCodeEditorService;
}());
exports.AbstractCodeEditorService = AbstractCodeEditorService;
