// vscode/src/vs/editor/common/services/languagesRegistry.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var errors_1 = require("vs/base/common/errors");
var event_1 = require("vs/base/common/event");
var mime = require("vs/base/common/mime");
var strings = require("vs/base/common/strings");
var modesRegistry_1 = require("vs/editor/common/modes/modesRegistry");
var hasOwnProperty = Object.prototype.hasOwnProperty;
var LanguagesRegistry = (function () {
    function LanguagesRegistry(useModesRegistry) {
        if (useModesRegistry === void 0) { useModesRegistry = true; }
        var _this = this;
        this._onDidAddModes = new event_1.Emitter();
        this.onDidAddModes = this._onDidAddModes.event;
        this._languages = {};
        this.mime2LanguageId = {};
        this.name2LanguageId = {};
        this.lowerName2Id = {};
        if (useModesRegistry) {
            this._registerLanguages(modesRegistry_1.ModesRegistry.getLanguages());
            modesRegistry_1.ModesRegistry.onDidAddLanguages(function (m) { return _this._registerLanguages(m); });
        }
    }
    LanguagesRegistry.prototype._registerLanguages = function (desc) {
        var _this = this;
        if (desc.length === 0) {
            return;
        }
        var addedModes = [];
        for (var i = 0; i < desc.length; i++) {
            this._registerLanguage(desc[i]);
            addedModes.push(desc[i].id);
        }
        // Rebuild fast path maps
        this.mime2LanguageId = {};
        this.name2LanguageId = {};
        this.lowerName2Id = {};
        Object.keys(this._languages).forEach(function (langId) {
            var language = _this._languages[langId];
            if (language.name) {
                _this.name2LanguageId[language.name] = langId;
            }
            language.aliases.forEach(function (alias) {
                _this.lowerName2Id[alias.toLowerCase()] = langId;
            });
            language.mimetypes.forEach(function (mimetype) {
                _this.mime2LanguageId[mimetype] = langId;
            });
        });
        this._onDidAddModes.fire(addedModes);
    };
    LanguagesRegistry.prototype._registerLanguage = function (lang) {
        var langId = lang.id;
        var resolvedLanguage = null;
        if (hasOwnProperty.call(this._languages, langId)) {
            resolvedLanguage = this._languages[langId];
        }
        else {
            resolvedLanguage = {
                name: null,
                mimetypes: [],
                aliases: [],
                extensions: [],
                filenames: [],
                configurationFiles: []
            };
            this._languages[langId] = resolvedLanguage;
        }
        LanguagesRegistry._mergeLanguage(resolvedLanguage, lang);
    };
    LanguagesRegistry._mergeLanguage = function (resolvedLanguage, lang) {
        var langId = lang.id;
        var primaryMime = null;
        if (typeof lang.mimetypes !== 'undefined' && Array.isArray(lang.mimetypes)) {
            for (var i = 0; i < lang.mimetypes.length; i++) {
                if (!primaryMime) {
                    primaryMime = lang.mimetypes[i];
                }
                resolvedLanguage.mimetypes.push(lang.mimetypes[i]);
            }
        }
        if (!primaryMime) {
            primaryMime = "text/x-" + langId;
            resolvedLanguage.mimetypes.push(primaryMime);
        }
        if (Array.isArray(lang.extensions)) {
            for (var _i = 0, _a = lang.extensions; _i < _a.length; _i++) {
                var extension = _a[_i];
                mime.registerTextMime({ id: langId, mime: primaryMime, extension: extension });
                resolvedLanguage.extensions.push(extension);
            }
        }
        if (Array.isArray(lang.filenames)) {
            for (var _b = 0, _c = lang.filenames; _b < _c.length; _b++) {
                var filename = _c[_b];
                mime.registerTextMime({ id: langId, mime: primaryMime, filename: filename });
                resolvedLanguage.filenames.push(filename);
            }
        }
        if (Array.isArray(lang.filenamePatterns)) {
            for (var _d = 0, _e = lang.filenamePatterns; _d < _e.length; _d++) {
                var filenamePattern = _e[_d];
                mime.registerTextMime({ id: langId, mime: primaryMime, filepattern: filenamePattern });
            }
        }
        if (typeof lang.firstLine === 'string' && lang.firstLine.length > 0) {
            var firstLineRegexStr = lang.firstLine;
            if (firstLineRegexStr.charAt(0) !== '^') {
                firstLineRegexStr = '^' + firstLineRegexStr;
            }
            try {
                var firstLineRegex = new RegExp(firstLineRegexStr);
                if (!strings.regExpLeadsToEndlessLoop(firstLineRegex)) {
                    mime.registerTextMime({ id: langId, mime: primaryMime, firstline: firstLineRegex });
                }
            }
            catch (err) {
                // Most likely, the regex was bad
                errors_1.onUnexpectedError(err);
            }
        }
        resolvedLanguage.aliases.push(langId);
        if (typeof lang.aliases !== 'undefined' && Array.isArray(lang.aliases)) {
            for (var i = 0; i < lang.aliases.length; i++) {
                if (!lang.aliases[i] || lang.aliases[i].length === 0) {
                    continue;
                }
                resolvedLanguage.aliases.push(lang.aliases[i]);
            }
        }
        var containsAliases = (typeof lang.aliases !== 'undefined' && Array.isArray(lang.aliases) && lang.aliases.length > 0);
        if (containsAliases && lang.aliases[0] === null) {
        }
        else {
            var bestName = (containsAliases ? lang.aliases[0] : null) || langId;
            if (containsAliases || !resolvedLanguage.name) {
                resolvedLanguage.name = bestName;
            }
        }
        if (typeof lang.configuration === 'string') {
            resolvedLanguage.configurationFiles.push(lang.configuration);
        }
    };
    LanguagesRegistry.prototype.isRegisteredMode = function (mimetypeOrModeId) {
        // Is this a known mime type ?
        if (hasOwnProperty.call(this.mime2LanguageId, mimetypeOrModeId)) {
            return true;
        }
        // Is this a known mode id ?
        return hasOwnProperty.call(this._languages, mimetypeOrModeId);
    };
    LanguagesRegistry.prototype.getRegisteredModes = function () {
        return Object.keys(this._languages);
    };
    LanguagesRegistry.prototype.getRegisteredLanguageNames = function () {
        return Object.keys(this.name2LanguageId);
    };
    LanguagesRegistry.prototype.getLanguageName = function (modeId) {
        if (!hasOwnProperty.call(this._languages, modeId)) {
            return null;
        }
        return this._languages[modeId].name;
    };
    LanguagesRegistry.prototype.getModeIdForLanguageNameLowercase = function (languageNameLower) {
        return this.lowerName2Id[languageNameLower] || null;
    };
    LanguagesRegistry.prototype.getConfigurationFiles = function (modeId) {
        if (!hasOwnProperty.call(this._languages, modeId)) {
            return [];
        }
        return this._languages[modeId].configurationFiles || [];
    };
    LanguagesRegistry.prototype.getMimeForMode = function (theModeId) {
        var keys = Object.keys(this.mime2LanguageId);
        for (var i = 0, len = keys.length; i < len; i++) {
            var _mime = keys[i];
            var modeId = this.mime2LanguageId[_mime];
            if (modeId === theModeId) {
                return _mime;
            }
        }
        return null;
    };
    LanguagesRegistry.prototype.extractModeIds = function (commaSeparatedMimetypesOrCommaSeparatedIdsOrName) {
        var _this = this;
        if (!commaSeparatedMimetypesOrCommaSeparatedIdsOrName) {
            return [];
        }
        return (commaSeparatedMimetypesOrCommaSeparatedIdsOrName.
            split(',').
            map(function (mimeTypeOrIdOrName) { return mimeTypeOrIdOrName.trim(); }).
            map(function (mimeTypeOrIdOrName) {
            return _this.mime2LanguageId[mimeTypeOrIdOrName] || mimeTypeOrIdOrName;
        }).
            filter(function (modeId) {
            return hasOwnProperty.call(_this._languages, modeId);
        }));
    };
    LanguagesRegistry.prototype.getModeIdsFromLanguageName = function (languageName) {
        if (!languageName) {
            return [];
        }
        if (hasOwnProperty.call(this.name2LanguageId, languageName)) {
            return [this.name2LanguageId[languageName]];
        }
        return [];
    };
    LanguagesRegistry.prototype.getModeIdsFromFilenameOrFirstLine = function (filename, firstLine) {
        if (!filename && !firstLine) {
            return [];
        }
        var mimeTypes = mime.guessMimeTypes(filename, firstLine);
        return this.extractModeIds(mimeTypes.join(','));
    };
    LanguagesRegistry.prototype.getExtensions = function (languageName) {
        var languageId = this.name2LanguageId[languageName];
        if (!languageId) {
            return [];
        }
        if (!hasOwnProperty.call(this._languages, languageId)) {
            return [];
        }
        return this._languages[languageId].extensions;
    };
    LanguagesRegistry.prototype.getFilenames = function (languageName) {
        var languageId = this.name2LanguageId[languageName];
        if (!languageId) {
            return [];
        }
        if (!hasOwnProperty.call(this._languages, languageId)) {
            return [];
        }
        return this._languages[languageId].filenames;
    };
    return LanguagesRegistry;
}());
exports.LanguagesRegistry = LanguagesRegistry;
