// vscode/src/vs/workbench/electron-browser/workbench.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
require("./media/workbench.css");
var winjs_base_1 = require("vs/base/common/winjs.base");
var lifecycle_1 = require("vs/base/common/lifecycle");
var strings = require("vs/base/common/strings");
var event_1 = require("vs/base/common/event");
var DOM = require("vs/base/browser/dom");
var builder_1 = require("vs/base/browser/builder");
var async_1 = require("vs/base/common/async");
var browser = require("vs/base/browser/browser");
var assert = require("vs/base/common/assert");
var timer = require("vs/base/common/timer");
var stopwatch_1 = require("vs/base/common/stopwatch");
var errors = require("vs/base/common/errors");
var backupModelService_1 = require("vs/workbench/services/backup/node/backupModelService");
var backupFileService_1 = require("vs/workbench/services/backup/node/backupFileService");
var backup_1 = require("vs/workbench/services/backup/common/backup");
var errorMessage_1 = require("vs/base/common/errorMessage");
var platform_1 = require("vs/platform/platform");
var platform_2 = require("vs/base/common/platform");
var editor_1 = require("vs/platform/editor/common/editor");
var contributions_1 = require("vs/workbench/common/contributions");
var editor_2 = require("vs/workbench/common/editor");
var history_1 = require("vs/workbench/services/history/browser/history");
var activitybarPart_1 = require("vs/workbench/browser/parts/activitybar/activitybarPart");
var editorPart_1 = require("vs/workbench/browser/parts/editor/editorPart");
var sidebarPart_1 = require("vs/workbench/browser/parts/sidebar/sidebarPart");
var panelPart_1 = require("vs/workbench/browser/parts/panel/panelPart");
var statusbarPart_1 = require("vs/workbench/browser/parts/statusbar/statusbarPart");
var titlebarPart_1 = require("vs/workbench/browser/parts/titlebar/titlebarPart");
var layout_1 = require("vs/workbench/browser/layout");
var actionBarRegistry_1 = require("vs/workbench/browser/actionBarRegistry");
var panel_1 = require("vs/workbench/browser/panel");
var quickOpenController_1 = require("vs/workbench/browser/parts/quickopen/quickOpenController");
var diffEditorInput_1 = require("vs/workbench/common/editor/diffEditorInput");
var extensions_1 = require("vs/platform/instantiation/common/extensions");
var untitledEditorService_1 = require("vs/workbench/services/untitled/common/untitledEditorService");
var editorService_1 = require("vs/workbench/services/editor/browser/editorService");
var partService_1 = require("vs/workbench/services/part/common/partService");
var workspace_1 = require("vs/platform/workspace/common/workspace");
var storage_1 = require("vs/platform/storage/common/storage");
var contextmenuService_1 = require("vs/workbench/services/contextview/electron-browser/contextmenuService");
var keybindingService_1 = require("vs/workbench/services/keybinding/electron-browser/keybindingService");
var configuration_1 = require("vs/platform/configuration/common/configuration");
var configurationEditing_1 = require("vs/workbench/services/configuration/common/configurationEditing");
var configurationEditingService_1 = require("vs/workbench/services/configuration/node/configurationEditingService");
var contextKeyService_1 = require("vs/platform/contextkey/browser/contextKeyService");
var keybinding_1 = require("vs/platform/keybinding/common/keybinding");
var contextkey_1 = require("vs/platform/contextkey/common/contextkey");
var activityBarService_1 = require("vs/workbench/services/activity/common/activityBarService");
var viewlet_1 = require("vs/workbench/services/viewlet/browser/viewlet");
var viewletService_1 = require("vs/workbench/services/viewlet/browser/viewletService");
var fileService_1 = require("vs/workbench/services/files/electron-browser/fileService");
var files_1 = require("vs/platform/files/common/files");
var configurationResolver_1 = require("vs/workbench/services/configurationResolver/common/configurationResolver");
var configurationResolverService_1 = require("vs/workbench/services/configurationResolver/node/configurationResolverService");
var panelService_1 = require("vs/workbench/services/panel/common/panelService");
var titleService_1 = require("vs/workbench/services/title/common/titleService");
var editorService_2 = require("vs/workbench/services/editor/common/editorService");
var quickOpenService_1 = require("vs/workbench/services/quickopen/common/quickOpenService");
var groupService_1 = require("vs/workbench/services/group/common/groupService");
var history_2 = require("vs/workbench/services/history/common/history");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var descriptors_1 = require("vs/platform/instantiation/common/descriptors");
var textFileService_1 = require("vs/workbench/services/textfile/electron-browser/textFileService");
var textfiles_1 = require("vs/workbench/services/textfile/common/textfiles");
var scm_1 = require("vs/workbench/services/scm/common/scm");
var scmService_1 = require("vs/workbench/services/scm/common/scmService");
var textModelResolverService_1 = require("vs/workbench/services/textmodelResolver/common/textModelResolverService");
var resolverService_1 = require("vs/editor/common/services/resolverService");
var lifecycle_2 = require("vs/platform/lifecycle/common/lifecycle");
var windows_1 = require("vs/platform/windows/common/windows");
var message_1 = require("vs/platform/message/common/message");
var statusbar_1 = require("vs/platform/statusbar/common/statusbar");
var actions_1 = require("vs/platform/actions/common/actions");
var menuService_1 = require("vs/platform/actions/common/menuService");
var contextView_1 = require("vs/platform/contextview/browser/contextView");
var environment_1 = require("vs/platform/environment/common/environment");
var telemetry_1 = require("vs/platform/telemetry/common/telemetry");
exports.MessagesVisibleContext = new contextkey_1.RawContextKey('globalMessageVisible', false);
exports.EditorsVisibleContext = new contextkey_1.RawContextKey('editorIsOpen', false);
exports.InZenModeContext = new contextkey_1.RawContextKey('inZenMode', false);
exports.NoEditorsVisibleContext = exports.EditorsVisibleContext.toNegated();
var Identifiers = {
    WORKBENCH_CONTAINER: 'workbench.main.container',
    TITLEBAR_PART: 'workbench.parts.titlebar',
    ACTIVITYBAR_PART: 'workbench.parts.activitybar',
    SIDEBAR_PART: 'workbench.parts.sidebar',
    PANEL_PART: 'workbench.parts.panel',
    EDITOR_PART: 'workbench.parts.editor',
    STATUSBAR_PART: 'workbench.parts.statusbar'
};
/**
 * The workbench creates and lays out all parts that make up the workbench.
 */
var Workbench = (function () {
    function Workbench(parent, container, workspace, options, serviceCollection, instantiationService, untitledEditorService, contextService, storageService, lifecycleService, messageService, configurationService, telemetryService, environmentService, windowService) {
        var _this = this;
        this.instantiationService = instantiationService;
        this.untitledEditorService = untitledEditorService;
        this.contextService = contextService;
        this.storageService = storageService;
        this.lifecycleService = lifecycleService;
        this.messageService = messageService;
        this.configurationService = configurationService;
        this.telemetryService = telemetryService;
        this.environmentService = environmentService;
        this.windowService = windowService;
        this.parent = parent;
        this.container = container;
        this.workbenchParams = {
            workspace: workspace,
            options: options,
            serviceCollection: serviceCollection
        };
        this.hasFilesToCreateOpenOrDiff =
            (options.filesToCreate && options.filesToCreate.length > 0) ||
                (options.filesToOpen && options.filesToOpen.length > 0) ||
                (options.filesToDiff && options.filesToDiff.length > 0);
        this.toDispose = [];
        this.toShutdown = [];
        this.editorBackgroundDelayer = new async_1.Delayer(50);
        this._onTitleBarVisibilityChange = new event_1.Emitter();
        this.creationPromise = new winjs_base_1.TPromise(function (c) {
            _this.creationPromiseComplete = c;
        });
    }
    Object.defineProperty(Workbench.prototype, "onTitleBarVisibilityChange", {
        get: function () {
            return this._onTitleBarVisibilityChange.event;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Starts the workbench and creates the HTML elements on the container. A workbench can only be started
     * once. Use the shutdown function to free up resources created by the workbench on startup.
     */
    Workbench.prototype.startup = function (callbacks) {
        var _this = this;
        assert.ok(!this.workbenchStarted, 'Can not start a workbench that was already started');
        assert.ok(!this.workbenchShutdown, 'Can not start a workbench that was shutdown');
        try {
            this.workbenchStarted = true;
            this.callbacks = callbacks;
            // Create Workbench
            this.createWorkbench();
            // Services
            this.initServices();
            if (this.callbacks && this.callbacks.onServicesCreated) {
                this.callbacks.onServicesCreated();
            }
            // Contexts
            this.messagesVisibleContext = exports.MessagesVisibleContext.bindTo(this.contextKeyService);
            this.editorsVisibleContext = exports.EditorsVisibleContext.bindTo(this.contextKeyService);
            this.inZenMode = exports.InZenModeContext.bindTo(this.contextKeyService);
            // Register Listeners
            this.registerListeners();
            // Settings
            this.initSettings();
            // Create Workbench and Parts
            this.renderWorkbench();
            // Workbench Layout
            this.createWorkbenchLayout();
            // Load composits and editors in parallel
            var compositeAndEditorPromises = [];
            // Restore last opened viewlet
            var viewletRestoreStopWatch_1;
            if (!this.sideBarHidden) {
                var viewletIdToRestore = void 0;
                if (this.shouldRestoreLastOpenedViewlet()) {
                    viewletIdToRestore = this.storageService.get(sidebarPart_1.SidebarPart.activeViewletSettingsKey, storage_1.StorageScope.WORKSPACE);
                }
                if (!viewletIdToRestore) {
                    viewletIdToRestore = this.viewletService.getDefaultViewletId();
                }
                viewletRestoreStopWatch_1 = stopwatch_1.StopWatch.create();
                var viewletTimerEvent_1 = timer.start(timer.Topic.STARTUP, strings.format('[renderer] open viewlet {0}', viewletIdToRestore));
                compositeAndEditorPromises.push(this.viewletService.openViewlet(viewletIdToRestore).then(function () {
                    viewletTimerEvent_1.stop();
                    viewletRestoreStopWatch_1.stop();
                }));
            }
            // Load Panel
            var panelRegistry = platform_1.Registry.as(panel_1.Extensions.Panels);
            var panelId = this.storageService.get(panelPart_1.PanelPart.activePanelSettingsKey, storage_1.StorageScope.WORKSPACE, panelRegistry.getDefaultPanelId());
            if (!this.panelHidden && !!panelId) {
                compositeAndEditorPromises.push(this.panelPart.openPanel(panelId, false));
            }
            // Load Editors
            var editorRestoreStopWatch_1 = stopwatch_1.StopWatch.create();
            var editorTimerEvent_1 = timer.start(timer.Topic.STARTUP, '[renderer] restoring editor view state');
            compositeAndEditorPromises.push(this.resolveEditorsToOpen().then(function (inputsWithOptions) {
                var editorOpenPromise;
                if (inputsWithOptions.length) {
                    var editors = inputsWithOptions.map(function (inputWithOptions) {
                        return {
                            input: inputWithOptions.input,
                            options: inputWithOptions.options,
                            position: editor_1.Position.ONE
                        };
                    });
                    editorOpenPromise = _this.editorPart.openEditors(editors);
                }
                else {
                    editorOpenPromise = _this.editorPart.restoreEditors();
                }
                return editorOpenPromise.then(function () {
                    _this.onEditorsChanged(); // make sure we show the proper background in the editor area
                    editorTimerEvent_1.stop();
                    editorRestoreStopWatch_1.stop();
                });
            }));
            // Flag workbench as created once done
            var workbenchDone_1 = function (error) {
                _this.workbenchCreated = true;
                _this.creationPromiseComplete(true);
                if (_this.callbacks && _this.callbacks.onWorkbenchStarted) {
                    _this.callbacks.onWorkbenchStarted({
                        customKeybindingsCount: _this.keybindingService.customKeybindingsCount(),
                        restoreViewletDuration: viewletRestoreStopWatch_1 ? viewletRestoreStopWatch_1.elapsed() : 0,
                        restoreEditorsDuration: editorRestoreStopWatch_1.elapsed(),
                        pinnedViewlets: _this.activitybarPart.getPinned()
                    });
                }
                if (error) {
                    errors.onUnexpectedError(error);
                }
            };
            // Join viewlet, panel and editor promises
            winjs_base_1.TPromise.join(compositeAndEditorPromises).then(function () { return workbenchDone_1(); }, function (error) { return workbenchDone_1(error); });
        }
        catch (error) {
            // Print out error
            console.error(errorMessage_1.toErrorMessage(error, true));
            // Rethrow
            throw error;
        }
    };
    Workbench.prototype.resolveEditorsToOpen = function () {
        var _this = this;
        // Files to open, diff or create
        if (this.hasFilesToCreateOpenOrDiff) {
            var wbopt = this.workbenchParams.options;
            var filesToCreate = wbopt.filesToCreate || [];
            var filesToOpen_1 = wbopt.filesToOpen || [];
            var filesToDiff_1 = wbopt.filesToDiff;
            // Files to diff is exclusive
            if (filesToDiff_1 && filesToDiff_1.length) {
                return winjs_base_1.TPromise.join(filesToDiff_1.map(function (resourceInput) { return _this.editorService.createInput(resourceInput); })).then(function (inputsToDiff) {
                    return [{ input: new diffEditorInput_1.DiffEditorInput(diffEditorInput_1.toDiffLabel(filesToDiff_1[0].resource, filesToDiff_1[1].resource, _this.contextService), null, inputsToDiff[0], inputsToDiff[1]) }];
                });
            }
            else {
                var inputs_1 = [];
                var options_1 = [];
                // Files to create
                inputs_1.push.apply(inputs_1, filesToCreate.map(function (resourceInput) { return _this.untitledEditorService.createOrGet(resourceInput.resource); }));
                options_1.push.apply(options_1, filesToCreate.map(function (r) { return null; })); // fill empty options for files to create because we dont have options there
                // Files to open
                var filesToOpenInputPromise = filesToOpen_1.map(function (resourceInput) { return _this.editorService.createInput(resourceInput); });
                return winjs_base_1.TPromise.join(filesToOpenInputPromise).then(function (inputsToOpen) {
                    inputs_1.push.apply(inputs_1, inputsToOpen);
                    options_1.push.apply(options_1, filesToOpen_1.map(function (resourceInput) { return editor_2.TextEditorOptions.from(resourceInput); }));
                    return inputs_1.map(function (input, index) { return { input: input, options: options_1[index] }; });
                });
            }
        }
        else if (!this.workbenchParams.workspace && this.telemetryService.getExperiments().openUntitledFile) {
            return winjs_base_1.TPromise.as([{ input: this.untitledEditorService.createOrGet() }]);
        }
        return winjs_base_1.TPromise.as([]);
    };
    Workbench.prototype.initServices = function () {
        var serviceCollection = this.workbenchParams.serviceCollection;
        this.toDispose.push(this.lifecycleService.onShutdown(this.shutdownComponents, this));
        // Services we contribute
        serviceCollection.set(partService_1.IPartService, this);
        // Status bar
        this.statusbarPart = this.instantiationService.createInstance(statusbarPart_1.StatusbarPart, Identifiers.STATUSBAR_PART);
        this.toDispose.push(this.statusbarPart);
        this.toShutdown.push(this.statusbarPart);
        serviceCollection.set(statusbar_1.IStatusbarService, this.statusbarPart);
        // Keybindings
        this.contextKeyService = this.instantiationService.createInstance(contextKeyService_1.ContextKeyService);
        serviceCollection.set(contextkey_1.IContextKeyService, this.contextKeyService);
        this.keybindingService = this.instantiationService.createInstance(keybindingService_1.WorkbenchKeybindingService, window);
        this.toDispose.push(this.keybindingService);
        serviceCollection.set(keybinding_1.IKeybindingService, this.keybindingService);
        // Context Menu
        serviceCollection.set(contextView_1.IContextMenuService, this.instantiationService.createInstance(contextmenuService_1.ContextMenuService));
        // Menus/Actions
        serviceCollection.set(actions_1.IMenuService, new descriptors_1.SyncDescriptor(menuService_1.MenuService));
        // Title bar
        this.titlebarPart = this.instantiationService.createInstance(titlebarPart_1.TitlebarPart, Identifiers.TITLEBAR_PART);
        this.toDispose.push(this.titlebarPart);
        this.toShutdown.push(this.titlebarPart);
        serviceCollection.set(titleService_1.ITitleService, this.titlebarPart);
        // Sidebar part
        this.sidebarPart = this.instantiationService.createInstance(sidebarPart_1.SidebarPart, Identifiers.SIDEBAR_PART);
        this.toDispose.push(this.sidebarPart);
        this.toShutdown.push(this.sidebarPart);
        // Viewlet service
        this.viewletService = this.instantiationService.createInstance(viewletService_1.ViewletService, this.sidebarPart);
        serviceCollection.set(viewlet_1.IViewletService, this.viewletService);
        // Panel service (panel part)
        this.panelPart = this.instantiationService.createInstance(panelPart_1.PanelPart, Identifiers.PANEL_PART);
        this.toDispose.push(this.panelPart);
        this.toShutdown.push(this.panelPart);
        serviceCollection.set(panelService_1.IPanelService, this.panelPart);
        // Activity service (activitybar part)
        this.activitybarPart = this.instantiationService.createInstance(activitybarPart_1.ActivitybarPart, Identifiers.ACTIVITYBAR_PART);
        this.toDispose.push(this.activitybarPart);
        this.toShutdown.push(this.activitybarPart);
        serviceCollection.set(activityBarService_1.IActivityBarService, this.activitybarPart);
        // Editor service (editor part)
        this.editorPart = this.instantiationService.createInstance(editorPart_1.EditorPart, Identifiers.EDITOR_PART, !this.hasFilesToCreateOpenOrDiff);
        this.toDispose.push(this.editorPart);
        this.toShutdown.push(this.editorPart);
        this.editorService = this.instantiationService.createInstance(editorService_1.WorkbenchEditorService, this.editorPart);
        serviceCollection.set(editorService_2.IWorkbenchEditorService, this.editorService);
        serviceCollection.set(groupService_1.IEditorGroupService, this.editorPart);
        // File Service
        var fileService = this.instantiationService.createInstance(fileService_1.FileService);
        serviceCollection.set(files_1.IFileService, fileService);
        // History
        serviceCollection.set(history_2.IHistoryService, this.instantiationService.createInstance(history_1.HistoryService));
        // Backup File Service
        var workspace = this.contextService.getWorkspace();
        serviceCollection.set(backup_1.IBackupFileService, this.instantiationService.createInstance(backupFileService_1.BackupFileService, this.windowService.getCurrentWindowId()));
        // Backup Service
        serviceCollection.set(backup_1.IBackupModelService, this.instantiationService.createInstance(backupModelService_1.BackupModelService));
        // Text File Service
        serviceCollection.set(textfiles_1.ITextFileService, this.instantiationService.createInstance(textFileService_1.TextFileService));
        // SCM Service
        serviceCollection.set(scm_1.ISCMService, this.instantiationService.createInstance(scmService_1.SCMService));
        // Text Model Resolver Service
        serviceCollection.set(resolverService_1.ITextModelResolverService, this.instantiationService.createInstance(textModelResolverService_1.TextModelResolverService));
        // Configuration Editing
        this.configurationEditingService = this.instantiationService.createInstance(configurationEditingService_1.ConfigurationEditingService);
        serviceCollection.set(configurationEditing_1.IConfigurationEditingService, this.configurationEditingService);
        // Configuration Resolver
        var configurationResolverService = this.instantiationService.createInstance(configurationResolverService_1.ConfigurationResolverService, workspace ? workspace.resource : null, process.env);
        serviceCollection.set(configurationResolver_1.IConfigurationResolverService, configurationResolverService);
        // Quick open service (quick open controller)
        this.quickOpen = this.instantiationService.createInstance(quickOpenController_1.QuickOpenController);
        this.toDispose.push(this.quickOpen);
        this.toShutdown.push(this.quickOpen);
        serviceCollection.set(quickOpenService_1.IQuickOpenService, this.quickOpen);
        // Contributed services
        var contributedServices = extensions_1.getServices();
        for (var _i = 0, contributedServices_1 = contributedServices; _i < contributedServices_1.length; _i++) {
            var contributedService = contributedServices_1[_i];
            serviceCollection.set(contributedService.id, contributedService.descriptor);
        }
        // Set the some services to registries that have been created eagerly
        platform_1.Registry.as(actionBarRegistry_1.Extensions.Actionbar).setInstantiationService(this.instantiationService);
        platform_1.Registry.as(contributions_1.Extensions.Workbench).setInstantiationService(this.instantiationService);
        platform_1.Registry.as(editor_2.Extensions.Editors).setInstantiationService(this.instantiationService);
    };
    Workbench.prototype.initSettings = function () {
        // Sidebar visibility
        this.sideBarHidden = this.storageService.getBoolean(Workbench.sidebarHiddenSettingKey, storage_1.StorageScope.WORKSPACE, false);
        if (!this.contextService.getWorkspace()) {
            this.sideBarHidden = true; // we hide sidebar in single-file-mode
        }
        // Panel part visibility
        var panelRegistry = platform_1.Registry.as(panel_1.Extensions.Panels);
        this.panelHidden = this.storageService.getBoolean(Workbench.panelHiddenSettingKey, storage_1.StorageScope.WORKSPACE, true);
        if (!this.contextService.getWorkspace() || !panelRegistry.getDefaultPanelId()) {
            this.panelHidden = true; // we hide panel part in single-file-mode or if there is no default panel
        }
        // Sidebar position
        var sideBarPosition = this.configurationService.lookup(Workbench.sidebarPositionConfigurationKey).value;
        this.sideBarPosition = (sideBarPosition === 'right') ? partService_1.Position.RIGHT : partService_1.Position.LEFT;
        // Statusbar visibility
        var statusBarVisible = this.configurationService.lookup(Workbench.statusbarVisibleConfigurationKey).value;
        this.statusBarHidden = !statusBarVisible;
        // Activity bar visibility
        var activityBarVisible = this.configurationService.lookup(Workbench.activityBarVisibleConfigurationKey).value;
        this.activityBarHidden = !activityBarVisible;
        // Zen mode
        this.zenMode = {
            active: false,
            transitionedToFullScreen: false,
            wasSideBarVisible: false,
            wasPanelVisible: false
        };
    };
    /**
     * Returns whether the workbench has been started.
     */
    Workbench.prototype.isStarted = function () {
        return this.workbenchStarted && !this.workbenchShutdown;
    };
    /**
     * Returns whether the workbench has been fully created.
     */
    Workbench.prototype.isCreated = function () {
        return this.workbenchCreated && this.workbenchStarted;
    };
    Workbench.prototype.joinCreation = function () {
        return this.creationPromise;
    };
    Workbench.prototype.hasFocus = function (part) {
        var activeElement = document.activeElement;
        if (!activeElement) {
            return false;
        }
        var container = this.getContainer(part);
        return DOM.isAncestor(activeElement, container);
    };
    Workbench.prototype.getContainer = function (part) {
        var container = null;
        switch (part) {
            case partService_1.Parts.TITLEBAR_PART:
                container = this.titlebarPart.getContainer();
                break;
            case partService_1.Parts.ACTIVITYBAR_PART:
                container = this.activitybarPart.getContainer();
                break;
            case partService_1.Parts.SIDEBAR_PART:
                container = this.sidebarPart.getContainer();
                break;
            case partService_1.Parts.PANEL_PART:
                container = this.panelPart.getContainer();
                break;
            case partService_1.Parts.EDITOR_PART:
                container = this.editorPart.getContainer();
                break;
            case partService_1.Parts.STATUSBAR_PART:
                container = this.statusbarPart.getContainer();
                break;
        }
        return container && container.getHTMLElement();
    };
    Workbench.prototype.isVisible = function (part) {
        switch (part) {
            case partService_1.Parts.TITLEBAR_PART:
                return this.getCustomTitleBarStyle() && !browser.isFullscreen();
            case partService_1.Parts.SIDEBAR_PART:
                return !this.sideBarHidden;
            case partService_1.Parts.PANEL_PART:
                return !this.panelHidden;
            case partService_1.Parts.STATUSBAR_PART:
                return !this.statusBarHidden;
            case partService_1.Parts.ACTIVITYBAR_PART:
                return !this.activityBarHidden;
        }
        return true; // any other part cannot be hidden
    };
    Workbench.prototype.getTitleBarOffset = function () {
        var offset = 0;
        if (this.isVisible(partService_1.Parts.TITLEBAR_PART)) {
            offset = 22 / browser.getZoomFactor(); // adjust the position based on title bar size and zoom factor
        }
        return offset;
    };
    Workbench.prototype.getCustomTitleBarStyle = function () {
        if (!platform_2.isMacintosh) {
            return null; // custom title bar is only supported on Mac currently
        }
        var isDev = !this.environmentService.isBuilt || this.environmentService.isExtensionDevelopment;
        if (isDev) {
            return null; // not enabled when developing due to https://github.com/electron/electron/issues/3647
        }
        var windowConfig = this.configurationService.getConfiguration();
        var style = windowConfig && windowConfig.window && windowConfig.window.titleBarStyle;
        if (style === 'custom') {
            return style;
        }
        return null;
    };
    Workbench.prototype.setStatusBarHidden = function (hidden, skipLayout) {
        this.statusBarHidden = hidden;
        // Layout
        if (!skipLayout) {
            this.workbenchLayout.layout({ forceStyleRecompute: true });
        }
    };
    Workbench.prototype.setActivityBarHidden = function (hidden, skipLayout) {
        this.activityBarHidden = hidden;
        // Layout
        if (!skipLayout) {
            this.workbenchLayout.layout({ forceStyleRecompute: true });
        }
    };
    Workbench.prototype.setSideBarHidden = function (hidden, skipLayout) {
        this.sideBarHidden = hidden;
        // Adjust CSS
        if (hidden) {
            this.workbench.addClass('nosidebar');
        }
        else {
            this.workbench.removeClass('nosidebar');
        }
        // Layout
        if (!skipLayout) {
            this.workbenchLayout.layout({ forceStyleRecompute: true });
        }
        // If sidebar becomes hidden, also hide the current active Viewlet if any
        if (hidden && this.sidebarPart.getActiveViewlet()) {
            this.sidebarPart.hideActiveViewlet();
            var activeEditor = this.editorPart.getActiveEditor();
            var activePanel = this.panelPart.getActivePanel();
            // Pass Focus to Editor or Panel if Sidebar is now hidden
            if (this.hasFocus(partService_1.Parts.PANEL_PART) && activePanel) {
                activePanel.focus();
            }
            else if (activeEditor) {
                activeEditor.focus();
            }
        }
        else if (!hidden && !this.sidebarPart.getActiveViewlet()) {
            var viewletToOpen = this.sidebarPart.getLastActiveViewletId() || this.viewletService.getDefaultViewletId();
            if (viewletToOpen) {
                this.sidebarPart.openViewlet(viewletToOpen, true).done(null, errors.onUnexpectedError);
            }
        }
        // Remember in settings
        this.storageService.store(Workbench.sidebarHiddenSettingKey, hidden ? 'true' : 'false', storage_1.StorageScope.WORKSPACE);
    };
    Workbench.prototype.setPanelHidden = function (hidden, skipLayout) {
        this.panelHidden = hidden;
        // Adjust CSS
        if (hidden) {
            this.workbench.addClass('nopanel');
        }
        else {
            this.workbench.removeClass('nopanel');
        }
        // Layout
        if (!skipLayout) {
            this.workbenchLayout.layout({ forceStyleRecompute: true });
        }
        // If panel part becomes hidden, also hide the current active panel if any
        if (hidden && this.panelPart.getActivePanel()) {
            this.panelPart.hideActivePanel();
            // Pass Focus to Editor if Panel part is now hidden
            var editor = this.editorPart.getActiveEditor();
            if (editor) {
                editor.focus();
            }
        }
        else if (!hidden && !this.panelPart.getActivePanel()) {
            var registry = platform_1.Registry.as(panel_1.Extensions.Panels);
            var panelToOpen = this.panelPart.getLastActivePanelId() || registry.getDefaultPanelId();
            if (panelToOpen) {
                this.panelPart.openPanel(panelToOpen, true).done(null, errors.onUnexpectedError);
            }
        }
        // Remember in settings
        this.storageService.store(Workbench.panelHiddenSettingKey, hidden ? 'true' : 'false', storage_1.StorageScope.WORKSPACE);
    };
    Workbench.prototype.toggleMaximizedPanel = function () {
        this.workbenchLayout.layout({ forceStyleRecompute: true, toggleMaximizedPanel: true });
    };
    Workbench.prototype.getSideBarPosition = function () {
        return this.sideBarPosition;
    };
    Workbench.prototype.setSideBarPosition = function (position) {
        if (this.sideBarHidden) {
            this.setSideBarHidden(false, true /* Skip Layout */);
        }
        var newPositionValue = (position === partService_1.Position.LEFT) ? 'left' : 'right';
        var oldPositionValue = (this.sideBarPosition === partService_1.Position.LEFT) ? 'left' : 'right';
        this.sideBarPosition = position;
        // Adjust CSS
        this.activitybarPart.getContainer().removeClass(oldPositionValue);
        this.sidebarPart.getContainer().removeClass(oldPositionValue);
        this.activitybarPart.getContainer().addClass(newPositionValue);
        this.sidebarPart.getContainer().addClass(newPositionValue);
        // Layout
        this.workbenchLayout.layout({ forceStyleRecompute: true });
    };
    Workbench.prototype.dispose = function () {
        if (this.isStarted()) {
            this.shutdownComponents();
            this.workbenchShutdown = true;
        }
        this.toDispose = lifecycle_1.dispose(this.toDispose);
    };
    /**
     * Asks the workbench and all its UI components inside to lay out according to
     * the containers dimension the workbench is living in.
     */
    Workbench.prototype.layout = function (options) {
        if (this.isStarted()) {
            this.workbenchLayout.layout(options);
        }
    };
    Workbench.prototype.shutdownComponents = function (reason) {
        if (reason === void 0) { reason = lifecycle_2.ShutdownReason.QUIT; }
        // Restore sidebar if we are being shutdown as a matter of a reload
        if (reason === lifecycle_2.ShutdownReason.RELOAD) {
            this.storageService.store(Workbench.sidebarRestoreSettingKey, 'true', storage_1.StorageScope.WORKSPACE);
        }
        // Pass shutdown on to each participant
        this.toShutdown.forEach(function (s) { return s.shutdown(); });
    };
    Workbench.prototype.registerListeners = function () {
        var _this = this;
        // Listen to editor changes
        this.toDispose.push(this.editorPart.onEditorsChanged(function () { return _this.onEditorsChanged(); }));
        // Handle message service and quick open events
        this.toDispose.push(this.messageService.onMessagesShowing(function () { return _this.messagesVisibleContext.set(true); }));
        this.toDispose.push(this.messageService.onMessagesCleared(function () { return _this.messagesVisibleContext.reset(); }));
        this.toDispose.push(this.quickOpen.onShow(function () { return _this.messageService.suspend(); })); // when quick open is open, don't show messages behind
        this.toDispose.push(this.quickOpen.onHide(function () { return _this.messageService.resume(); })); // resume messages once quick open is closed again
        // Configuration changes
        this.toDispose.push(this.configurationService.onDidUpdateConfiguration(function () { return _this.onDidUpdateConfiguration(); }));
        // Fullscreen changes
        this.toDispose.push(browser.onDidChangeFullscreen(function () { return _this.onFullscreenChanged(); }));
    };
    Workbench.prototype.onFullscreenChanged = function () {
        if (!this.isCreated) {
            return; // we need to be ready
        }
        // Apply as CSS class
        var isFullscreen = browser.isFullscreen();
        if (isFullscreen) {
            this.addClass('fullscreen');
        }
        else {
            this.removeClass('fullscreen');
            if (this.zenMode.transitionedToFullScreen && this.zenMode.active) {
                this.toggleZenMode();
            }
        }
        // Changing fullscreen state of the window has an impact on custom title bar visibility, so we need to update
        var hasCustomTitle = this.getCustomTitleBarStyle() === 'custom';
        if (hasCustomTitle) {
            this._onTitleBarVisibilityChange.fire();
            this.layout(); // handle title bar when fullscreen changes
        }
    };
    Workbench.prototype.onEditorsChanged = function () {
        var visibleEditors = this.editorService.getVisibleEditors().length;
        // We update the editorpart class to indicate if an editor is opened or not
        // through a delay to accomodate for fast editor switching
        var editorContainer = this.editorPart.getContainer();
        if (visibleEditors === 0) {
            this.editorsVisibleContext.reset();
            this.editorBackgroundDelayer.trigger(function () { return editorContainer.addClass('empty'); });
        }
        else {
            this.editorsVisibleContext.set(true);
            this.editorBackgroundDelayer.trigger(function () { return editorContainer.removeClass('empty'); });
        }
    };
    Workbench.prototype.onDidUpdateConfiguration = function (skipLayout) {
        var newSidebarPositionValue = this.configurationService.lookup(Workbench.sidebarPositionConfigurationKey).value;
        var newSidebarPosition = (newSidebarPositionValue === 'right') ? partService_1.Position.RIGHT : partService_1.Position.LEFT;
        if (newSidebarPosition !== this.getSideBarPosition()) {
            this.setSideBarPosition(newSidebarPosition);
        }
        if (!this.zenMode.active) {
            var newStatusbarHiddenValue = !this.configurationService.lookup(Workbench.statusbarVisibleConfigurationKey).value;
            if (newStatusbarHiddenValue !== this.statusBarHidden) {
                this.setStatusBarHidden(newStatusbarHiddenValue, skipLayout);
            }
            var newActivityBarHiddenValue = !this.configurationService.lookup(Workbench.activityBarVisibleConfigurationKey).value;
            if (newActivityBarHiddenValue !== this.activityBarHidden) {
                this.setActivityBarHidden(newActivityBarHiddenValue, skipLayout);
            }
        }
    };
    Workbench.prototype.createWorkbenchLayout = function () {
        this.workbenchLayout = this.instantiationService.createInstance(layout_1.WorkbenchLayout, builder_1.$(this.container), // Parent
        this.workbench, // Workbench Container
        {
            titlebar: this.titlebarPart,
            activitybar: this.activitybarPart,
            editor: this.editorPart,
            sidebar: this.sidebarPart,
            panel: this.panelPart,
            statusbar: this.statusbarPart,
        }, this.quickOpen // Quickopen
        );
        this.toDispose.push(this.workbenchLayout);
    };
    Workbench.prototype.createWorkbench = function () {
        // Create Workbench DIV Off-DOM
        this.workbenchContainer = builder_1.$('.monaco-workbench-container');
        this.workbench = builder_1.$().div({ 'class': 'monaco-workbench ' + (platform_2.isWindows ? 'windows' : platform_2.isLinux ? 'linux' : 'mac'), id: Identifiers.WORKBENCH_CONTAINER }).appendTo(this.workbenchContainer);
    };
    Workbench.prototype.renderWorkbench = function () {
        // Apply sidebar state as CSS class
        if (this.sideBarHidden) {
            this.workbench.addClass('nosidebar');
        }
        if (this.panelHidden) {
            this.workbench.addClass('nopanel');
        }
        // Apply no-workspace state as CSS class
        if (!this.workbenchParams.workspace) {
            this.workbench.addClass('no-workspace');
        }
        // Apply title style if shown
        var titleStyle = this.getCustomTitleBarStyle();
        if (titleStyle) {
            DOM.addClass(this.parent, "titlebar-style-" + titleStyle);
        }
        // Apply fullscreen state
        if (browser.isFullscreen()) {
            this.workbench.addClass('fullscreen');
        }
        // Create Parts
        this.createTitlebarPart();
        this.createActivityBarPart();
        this.createSidebarPart();
        this.createEditorPart();
        this.createPanelPart();
        this.createStatusbarPart();
        // Add Workbench to DOM
        this.workbenchContainer.build(this.container);
    };
    Workbench.prototype.createTitlebarPart = function () {
        var titlebarContainer = builder_1.$(this.workbench).div({
            'class': ['part', 'titlebar'],
            id: Identifiers.TITLEBAR_PART,
            role: 'contentinfo'
        });
        this.titlebarPart.create(titlebarContainer);
    };
    Workbench.prototype.createActivityBarPart = function () {
        var activitybarPartContainer = builder_1.$(this.workbench)
            .div({
            'class': ['part', 'activitybar', this.sideBarPosition === partService_1.Position.LEFT ? 'left' : 'right'],
            id: Identifiers.ACTIVITYBAR_PART,
            role: 'navigation'
        });
        this.activitybarPart.create(activitybarPartContainer);
    };
    Workbench.prototype.createSidebarPart = function () {
        var sidebarPartContainer = builder_1.$(this.workbench)
            .div({
            'class': ['part', 'sidebar', this.sideBarPosition === partService_1.Position.LEFT ? 'left' : 'right'],
            id: Identifiers.SIDEBAR_PART,
            role: 'complementary'
        });
        this.sidebarPart.create(sidebarPartContainer);
    };
    Workbench.prototype.createPanelPart = function () {
        var panelPartContainer = builder_1.$(this.workbench)
            .div({
            'class': ['part', 'panel', 'monaco-editor-background'],
            id: Identifiers.PANEL_PART,
            role: 'complementary'
        });
        this.panelPart.create(panelPartContainer);
    };
    Workbench.prototype.createEditorPart = function () {
        var editorContainer = builder_1.$(this.workbench)
            .div({
            'class': ['part', 'editor', 'monaco-editor-background', 'empty'],
            id: Identifiers.EDITOR_PART,
            role: 'main'
        });
        this.editorPart.create(editorContainer);
    };
    Workbench.prototype.createStatusbarPart = function () {
        var statusbarContainer = builder_1.$(this.workbench).div({
            'class': ['part', 'statusbar'],
            id: Identifiers.STATUSBAR_PART,
            role: 'contentinfo'
        });
        this.statusbarPart.create(statusbarContainer);
    };
    Workbench.prototype.getEditorPart = function () {
        assert.ok(this.workbenchStarted, 'Workbench is not started. Call startup() first.');
        return this.editorPart;
    };
    Workbench.prototype.getSidebarPart = function () {
        assert.ok(this.workbenchStarted, 'Workbench is not started. Call startup() first.');
        return this.sidebarPart;
    };
    Workbench.prototype.getPanelPart = function () {
        assert.ok(this.workbenchStarted, 'Workbench is not started. Call startup() first.');
        return this.panelPart;
    };
    Workbench.prototype.getInstantiationService = function () {
        assert.ok(this.workbenchStarted, 'Workbench is not started. Call startup() first.');
        return this.instantiationService;
    };
    Workbench.prototype.addClass = function (clazz) {
        if (this.workbench) {
            this.workbench.addClass(clazz);
        }
    };
    Workbench.prototype.removeClass = function (clazz) {
        if (this.workbench) {
            this.workbench.removeClass(clazz);
        }
    };
    Workbench.prototype.getWorkbenchElementId = function () {
        return Identifiers.WORKBENCH_CONTAINER;
    };
    Workbench.prototype.toggleZenMode = function () {
        var _this = this;
        this.zenMode.active = !this.zenMode.active;
        // Check if zen mode transitioned to full screen and if now we are out of zen mode -> we need to go out of full screen
        var toggleFullScreen = false;
        if (this.zenMode.active) {
            var windowConfig = this.configurationService.getConfiguration();
            toggleFullScreen = !browser.isFullscreen() && windowConfig.window.fullScreenZenMode;
            this.zenMode.transitionedToFullScreen = toggleFullScreen;
            this.zenMode.wasSideBarVisible = this.isVisible(partService_1.Parts.SIDEBAR_PART);
            this.zenMode.wasPanelVisible = this.isVisible(partService_1.Parts.PANEL_PART);
            this.setPanelHidden(true, true);
            this.setSideBarHidden(true, true);
            this.setActivityBarHidden(true, true);
            this.setStatusBarHidden(true, true);
        }
        else {
            if (this.zenMode.wasPanelVisible) {
                this.setPanelHidden(false, true);
            }
            if (this.zenMode.wasSideBarVisible) {
                this.setSideBarHidden(false, true);
            }
            // Status bar and activity bar visibility come from settings -> update their visibility.
            this.onDidUpdateConfiguration(true);
            toggleFullScreen = this.zenMode.transitionedToFullScreen && browser.isFullscreen();
        }
        this.inZenMode.set(this.zenMode.active);
        (toggleFullScreen ? this.windowService.toggleFullScreen() : winjs_base_1.TPromise.as(null))
            .done(function () { return _this.layout(); }, errors.onUnexpectedError);
    };
    Workbench.prototype.shouldRestoreLastOpenedViewlet = function () {
        if (!this.environmentService.isBuilt) {
            return true; // always restore sidebar when we are in development mode
        }
        var restore = this.storageService.getBoolean(Workbench.sidebarRestoreSettingKey, storage_1.StorageScope.WORKSPACE);
        if (restore) {
            this.storageService.remove(Workbench.sidebarRestoreSettingKey, storage_1.StorageScope.WORKSPACE); // only support once
        }
        return restore;
    };
    return Workbench;
}());
Workbench.sidebarHiddenSettingKey = 'workbench.sidebar.hidden';
Workbench.sidebarRestoreSettingKey = 'workbench.sidebar.restore';
Workbench.panelHiddenSettingKey = 'workbench.panel.hidden';
Workbench.sidebarPositionConfigurationKey = 'workbench.sideBar.location';
Workbench.statusbarVisibleConfigurationKey = 'workbench.statusBar.visible';
Workbench.activityBarVisibleConfigurationKey = 'workbench.activityBar.visible';
Workbench = __decorate([
    __param(5, instantiation_1.IInstantiationService),
    __param(6, untitledEditorService_1.IUntitledEditorService),
    __param(7, workspace_1.IWorkspaceContextService),
    __param(8, storage_1.IStorageService),
    __param(9, lifecycle_2.ILifecycleService),
    __param(10, message_1.IMessageService),
    __param(11, configuration_1.IConfigurationService),
    __param(12, telemetry_1.ITelemetryService),
    __param(13, environment_1.IEnvironmentService),
    __param(14, windows_1.IWindowService)
], Workbench);
exports.Workbench = Workbench;
