// vscode/src/vs/workbench/browser/parts/quickopen/quickOpenController.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
require('./media/quickopen.css');
require('vs/workbench/browser/parts/quickopen/quickopen.contribution');
var winjs_base_1 = require('vs/base/common/winjs.base');
var nls = require('vs/nls');
var builder_1 = require('vs/base/browser/builder');
var strings = require('vs/base/common/strings');
var filters = require('vs/base/common/filters');
var DOM = require('vs/base/browser/dom');
var uuid = require('vs/base/common/uuid');
var types = require('vs/base/common/types');
var cancellation_1 = require('vs/base/common/cancellation');
var quickOpen_1 = require('vs/base/parts/quickopen/common/quickOpen');
var quickOpenModel_1 = require('vs/base/parts/quickopen/browser/quickOpenModel');
var quickOpenWidget_1 = require('vs/base/parts/quickopen/browser/quickOpenWidget');
var actionBarRegistry_1 = require('vs/workbench/browser/actionBarRegistry');
var labels = require('vs/base/common/labels');
var paths = require('vs/base/common/paths');
var textfiles_1 = require('vs/workbench/services/textfile/common/textfiles');
var platform_1 = require('vs/platform/platform');
var modeService_1 = require('vs/editor/common/services/modeService');
var labels_1 = require('vs/workbench/browser/labels');
var modelService_1 = require('vs/editor/common/services/modelService');
var editor_1 = require('vs/workbench/common/editor');
var component_1 = require('vs/workbench/common/component');
var event_1 = require('vs/base/common/event');
var partService_1 = require('vs/workbench/services/part/common/partService');
var quickopen_1 = require('vs/workbench/browser/quickopen');
var errors = require('vs/base/common/errors');
var editorService_1 = require('vs/workbench/services/editor/common/editorService');
var configuration_1 = require('vs/platform/configuration/common/configuration');
var instantiation_1 = require('vs/platform/instantiation/common/instantiation');
var message_1 = require('vs/platform/message/common/message');
var telemetry_1 = require('vs/platform/telemetry/common/telemetry');
var workspace_1 = require('vs/platform/workspace/common/workspace');
var contextkey_1 = require('vs/platform/contextkey/common/contextkey');
var history_1 = require('vs/workbench/services/history/common/history');
var HELP_PREFIX = '?';
var QUICK_OPEN_MODE = new contextkey_1.RawContextKey('inQuickOpen', false);
var QuickOpenController = (function (_super) {
    __extends(QuickOpenController, _super);
    function QuickOpenController(editorService, messageService, telemetryService, contextService, contextKeyService, configurationService, historyService, instantiationService, partService) {
        _super.call(this, QuickOpenController.ID);
        this.editorService = editorService;
        this.messageService = messageService;
        this.telemetryService = telemetryService;
        this.contextService = contextService;
        this.configurationService = configurationService;
        this.historyService = historyService;
        this.instantiationService = instantiationService;
        this.partService = partService;
        this.actionProvider = new actionBarRegistry_1.ContributableActionProvider();
        this.previousValue = '';
        this.mapResolvedHandlersToPrefix = {};
        this.handlerOnOpenCalled = {};
        this.promisesToCompleteOnHide = [];
        this.inQuickOpenMode = QUICK_OPEN_MODE.bindTo(contextKeyService);
        this._onShow = new event_1.Emitter();
        this._onHide = new event_1.Emitter();
        this.updateConfiguration(this.configurationService.getConfiguration());
        this.registerListeners();
    }
    QuickOpenController.prototype.registerListeners = function () {
        var _this = this;
        this.configurationService.onDidUpdateConfiguration(function (e) { return _this.updateConfiguration(e.config); });
    };
    QuickOpenController.prototype.updateConfiguration = function (settings) {
        this.closeOnFocusLost = settings.workbench.quickOpen.closeOnFocusLost;
    };
    Object.defineProperty(QuickOpenController.prototype, "onShow", {
        get: function () {
            return this._onShow.event;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(QuickOpenController.prototype, "onHide", {
        get: function () {
            return this._onHide.event;
        },
        enumerable: true,
        configurable: true
    });
    QuickOpenController.prototype.quickNavigate = function (configuration, next) {
        if (this.quickOpenWidget) {
            this.quickOpenWidget.quickNavigate(configuration, next);
        }
    };
    QuickOpenController.prototype.input = function (options, token) {
        var _this = this;
        if (options === void 0) { options = {}; }
        if (token === void 0) { token = cancellation_1.CancellationToken.None; }
        if (this.pickOpenWidget && this.pickOpenWidget.isVisible()) {
            this.pickOpenWidget.hide(quickOpenWidget_1.HideReason.CANCELED);
        }
        var defaultMessage = options.prompt
            ? nls.localize('inputModeEntryDescription', "{0} (Press 'Enter' to confirm or 'Escape' to cancel)", options.prompt)
            : nls.localize('inputModeEntry', "Press 'Enter' to confirm your input or 'Escape' to cancel");
        var currentPick = defaultMessage;
        var currentValidation = winjs_base_1.TPromise.as(true);
        var currentDecoration;
        var lastValue;
        var init = function (resolve, reject) {
            // open quick pick with just one choice. we will recurse whenever
            // the validation/success message changes
            _this.doPick(winjs_base_1.TPromise.as([{ label: currentPick }]), {
                ignoreFocusLost: options.ignoreFocusLost,
                autoFocus: { autoFocusFirstEntry: true },
                password: options.password,
                placeHolder: options.placeHolder,
                value: lastValue === void 0 ? options.value : lastValue,
                valueSelect: lastValue === void 0,
                inputDecoration: currentDecoration,
                onDidType: function (value) {
                    lastValue = value;
                    if (options.validateInput) {
                        if (currentValidation) {
                            currentValidation.cancel();
                        }
                        currentValidation = winjs_base_1.TPromise.timeout(100).then(function () {
                            return options.validateInput(value).then(function (message) {
                                currentDecoration = !!message ? message_1.Severity.Error : void 0;
                                var newPick = message || defaultMessage;
                                if (newPick !== currentPick) {
                                    currentPick = newPick;
                                    resolve(new winjs_base_1.TPromise(init));
                                }
                                return !message;
                            });
                        }, function (err) {
                            // ignore
                        });
                    }
                }
            }, token).then(resolve, reject);
        };
        return new winjs_base_1.TPromise(init).then(function (item) {
            return currentValidation.then(function (valid) {
                if (valid && item) {
                    return lastValue || options.value || '';
                }
            });
        });
    };
    QuickOpenController.prototype.pick = function (arg1, options, token) {
        var _this = this;
        if (!options) {
            options = Object.create(null);
        }
        var arrayPromise;
        if (Array.isArray(arg1)) {
            arrayPromise = winjs_base_1.TPromise.as(arg1);
        }
        else if (winjs_base_1.TPromise.is(arg1)) {
            arrayPromise = arg1;
        }
        else {
            throw new Error('illegal input');
        }
        var isAboutStrings = false;
        var entryPromise = arrayPromise.then(function (elements) {
            return elements.map(function (element) {
                if (typeof element === 'string') {
                    isAboutStrings = true;
                    return { label: element };
                }
                else {
                    return element;
                }
            });
        });
        if (this.pickOpenWidget && this.pickOpenWidget.isVisible()) {
            this.pickOpenWidget.hide(quickOpenWidget_1.HideReason.CANCELED);
        }
        return new winjs_base_1.TPromise(function (resolve, reject, progress) {
            function onItem(item) {
                return item && isAboutStrings ? item.label : item;
            }
            _this.doPick(entryPromise, options, token).then(function (item) { return resolve(onItem(item)); }, function (err) { return reject(err); }, function (item) { return progress(onItem(item)); });
        });
    };
    QuickOpenController.prototype.doPick = function (picksPromise, options, token) {
        var _this = this;
        if (token === void 0) { token = cancellation_1.CancellationToken.None; }
        var autoFocus = options.autoFocus;
        // Use a generated token to avoid race conditions from long running promises
        var currentPickerToken = uuid.generateUuid();
        this.currentPickerToken = currentPickerToken;
        // Create upon first open
        if (!this.pickOpenWidget) {
            this.pickOpenWidget = new quickOpenWidget_1.QuickOpenWidget(builder_1.withElementById(this.partService.getWorkbenchElementId()).getHTMLElement(), {
                onOk: function () { },
                onCancel: function () { },
                onType: function (value) { },
                onShow: function () { return _this.handleOnShow(true); },
                onHide: function (reason) { return _this.handleOnHide(true, reason); }
            }, {
                inputPlaceHolder: options.placeHolder || ''
            }, this.telemetryService);
            var pickOpenContainer = this.pickOpenWidget.create();
            DOM.addClass(pickOpenContainer, 'show-file-icons');
        }
        else {
            this.pickOpenWidget.setPlaceHolder(options.placeHolder || '');
        }
        // Respect input value
        if (options.value) {
            this.pickOpenWidget.setValue(options.value, options.valueSelect);
        }
        // Respect password
        this.pickOpenWidget.setPassword(options.password);
        // Input decoration
        if (!types.isUndefinedOrNull(options.inputDecoration)) {
            this.pickOpenWidget.showInputDecoration(options.inputDecoration);
        }
        else {
            this.pickOpenWidget.clearInputDecoration();
        }
        // Layout
        if (this.layoutDimensions) {
            this.pickOpenWidget.layout(this.layoutDimensions);
        }
        return new winjs_base_1.TPromise(function (complete, error, progress) {
            // hide widget when being cancelled
            token.onCancellationRequested(function (e) {
                if (_this.currentPickerToken === currentPickerToken) {
                    _this.pickOpenWidget.hide(quickOpenWidget_1.HideReason.CANCELED);
                }
            });
            var picksPromiseDone = false;
            // Resolve picks
            picksPromise.then(function (picks) {
                if (_this.currentPickerToken !== currentPickerToken) {
                    return; // Return if another request came after
                }
                picksPromiseDone = true;
                // Reset Progress
                _this.pickOpenWidget.getProgressBar().stop().getContainer().hide();
                // Model
                var model = new quickOpenModel_1.QuickOpenModel();
                var entries = picks.map(function (e) { return _this.instantiationService.createInstance(PickOpenEntry, e, function () { return progress(e); }); });
                if (picks.length === 0) {
                    entries.push(_this.instantiationService.createInstance(PickOpenEntry, { label: nls.localize('emptyPicks', "There are no entries to pick from") }, null));
                }
                model.setEntries(entries);
                // Handlers
                _this.pickOpenWidget.setCallbacks({
                    onOk: function () {
                        if (picks.length === 0) {
                            return complete(null);
                        }
                        var index = -1;
                        var context;
                        entries.forEach(function (entry, i) {
                            if (entry.shouldRunWithContext) {
                                index = i;
                                context = entry.shouldRunWithContext;
                            }
                        });
                        var selectedPick = picks[index];
                        if (selectedPick && typeof selectedPick.run === 'function') {
                            selectedPick.run(context);
                        }
                        complete(selectedPick || null);
                    },
                    onCancel: function () { return complete(void 0); },
                    onFocusLost: function () { return !_this.closeOnFocusLost || options.ignoreFocusLost; },
                    onType: function (value) {
                        // the caller takes care of all input
                        if (options.onDidType) {
                            options.onDidType(value);
                            return;
                        }
                        if (picks.length === 0) {
                            return;
                        }
                        value = value ? strings.trim(value) : value;
                        // Reset filtering
                        if (!value) {
                            entries.forEach(function (e) {
                                e.setHighlights(null);
                                e.setHidden(false);
                            });
                        }
                        else {
                            entries.forEach(function (entry) {
                                var labelHighlights = filters.matchesFuzzy(value, entry.getLabel());
                                var descriptionHighlights = options.matchOnDescription && filters.matchesFuzzy(value, entry.getDescription());
                                var detailHighlights = options.matchOnDetail && entry.getDetail() && filters.matchesFuzzy(value, entry.getDetail());
                                if (entry.shouldAlwaysShow() || labelHighlights || descriptionHighlights || detailHighlights) {
                                    entry.setHighlights(labelHighlights, descriptionHighlights, detailHighlights);
                                    entry.setHidden(false);
                                }
                                else {
                                    entry.setHighlights(null, null, null);
                                    entry.setHidden(true);
                                }
                            });
                        }
                        _this.pickOpenWidget.refresh(model, value ? { autoFocusFirstEntry: true } : autoFocus);
                    },
                    onShow: function () { return _this.handleOnShow(true); },
                    onHide: function (reason) { return _this.handleOnHide(true, reason); }
                });
                // Set input
                if (!_this.pickOpenWidget.isVisible()) {
                    _this.pickOpenWidget.show(model, { autoFocus: autoFocus });
                }
                else {
                    _this.pickOpenWidget.setInput(model, autoFocus);
                }
            }, function (err) {
                _this.pickOpenWidget.hide();
                error(err);
            });
            // Progress if task takes a long time
            winjs_base_1.TPromise.timeout(800).then(function () {
                if (!picksPromiseDone && _this.currentPickerToken === currentPickerToken) {
                    _this.pickOpenWidget.getProgressBar().infinite().getContainer().show();
                }
            });
            // Show picker empty if resolving takes a while
            if (!picksPromiseDone) {
                _this.pickOpenWidget.show(new quickOpenModel_1.QuickOpenModel());
            }
        });
    };
    QuickOpenController.prototype.accept = function () {
        [this.quickOpenWidget, this.pickOpenWidget].forEach(function (w) {
            if (w && w.isVisible()) {
                w.accept();
            }
        });
    };
    QuickOpenController.prototype.focus = function () {
        [this.quickOpenWidget, this.pickOpenWidget].forEach(function (w) {
            if (w && w.isVisible()) {
                w.focus();
            }
        });
    };
    QuickOpenController.prototype.close = function () {
        [this.quickOpenWidget, this.pickOpenWidget].forEach(function (w) {
            if (w && w.isVisible()) {
                w.hide(quickOpenWidget_1.HideReason.CANCELED);
            }
        });
    };
    QuickOpenController.prototype.emitQuickOpenVisibilityChange = function (isVisible) {
        var _this = this;
        if (this.visibilityChangeTimeoutHandle) {
            window.clearTimeout(this.visibilityChangeTimeoutHandle);
        }
        this.visibilityChangeTimeoutHandle = setTimeout(function () {
            if (isVisible) {
                _this._onShow.fire();
            }
            else {
                _this._onHide.fire();
            }
            _this.visibilityChangeTimeoutHandle = void 0;
        }, 100 /* to prevent flashing, we accumulate visibility changes over a timeout of 100ms */);
    };
    QuickOpenController.prototype.show = function (prefix, options) {
        var _this = this;
        var quickNavigateConfiguration = options ? options.quickNavigateConfiguration : void 0;
        this.previousValue = prefix;
        var promiseCompletedOnHide = new winjs_base_1.TPromise(function (c) {
            _this.promisesToCompleteOnHide.push(c);
        });
        // Telemetry: log that quick open is shown and log the mode
        var registry = platform_1.Registry.as(quickopen_1.Extensions.Quickopen);
        var handlerDescriptor = registry.getQuickOpenHandler(prefix) || registry.getDefaultQuickOpenHandler();
        this.telemetryService.publicLog('quickOpenWidgetShown', { mode: handlerDescriptor.getId(), quickNavigate: quickNavigateConfiguration });
        // Trigger onOpen
        this.resolveHandler(handlerDescriptor)
            .done(null, errors.onUnexpectedError);
        // Create upon first open
        if (!this.quickOpenWidget) {
            this.quickOpenWidget = new quickOpenWidget_1.QuickOpenWidget(builder_1.withElementById(this.partService.getWorkbenchElementId()).getHTMLElement(), {
                onOk: function () { },
                onCancel: function () { },
                onType: function (value) { return _this.onType(value || ''); },
                onShow: function () { return _this.handleOnShow(false); },
                onHide: function (reason) { return _this.handleOnHide(false, reason); },
                onFocusLost: function () { return !_this.closeOnFocusLost; }
            }, {
                inputPlaceHolder: this.hasHandler(HELP_PREFIX) ? nls.localize('quickOpenInput', "Type '?' to get help on the actions you can take from here") : ''
            }, this.telemetryService);
            var quickOpenContainer = this.quickOpenWidget.create();
            DOM.addClass(quickOpenContainer, 'show-file-icons');
        }
        // Layout
        if (this.layoutDimensions) {
            this.quickOpenWidget.layout(this.layoutDimensions);
        }
        // Show quick open with prefix or editor history
        if (!this.quickOpenWidget.isVisible() || quickNavigateConfiguration) {
            if (prefix) {
                this.quickOpenWidget.show(prefix, { quickNavigateConfiguration: quickNavigateConfiguration });
            }
            else {
                var editorHistory = this.getEditorHistoryWithGroupLabel();
                if (editorHistory.getEntries().length < 2) {
                    quickNavigateConfiguration = null; // If no entries can be shown, default to normal quick open mode
                }
                var autoFocus = void 0;
                if (!quickNavigateConfiguration) {
                    autoFocus = { autoFocusFirstEntry: true };
                }
                else {
                    var visibleEditorCount = this.editorService.getVisibleEditors().length;
                    autoFocus = { autoFocusFirstEntry: visibleEditorCount === 0, autoFocusSecondEntry: visibleEditorCount !== 0 };
                }
                this.quickOpenWidget.show(editorHistory, { quickNavigateConfiguration: quickNavigateConfiguration, autoFocus: autoFocus });
            }
        }
        else {
            this.quickOpenWidget.show(prefix || '');
        }
        return promiseCompletedOnHide;
    };
    QuickOpenController.prototype.handleOnShow = function (isPicker) {
        if (isPicker && this.quickOpenWidget) {
            this.quickOpenWidget.hide(quickOpenWidget_1.HideReason.FOCUS_LOST);
        }
        else if (!isPicker && this.pickOpenWidget) {
            this.pickOpenWidget.hide(quickOpenWidget_1.HideReason.FOCUS_LOST);
        }
        this.inQuickOpenMode.set(true);
        this.emitQuickOpenVisibilityChange(true);
    };
    QuickOpenController.prototype.handleOnHide = function (isPicker, reason) {
        var _this = this;
        if (!isPicker) {
            // Clear state
            this.previousActiveHandlerDescriptor = null;
            // Pass to handlers
            var _loop_1 = function(prefix) {
                if (this_1.mapResolvedHandlersToPrefix.hasOwnProperty(prefix)) {
                    var promise = this_1.mapResolvedHandlersToPrefix[prefix];
                    promise.then(function (handler) {
                        _this.handlerOnOpenCalled[prefix] = false;
                        // Don't check if onOpen was called to preserve old behaviour for now
                        handler.onClose(reason === quickOpenWidget_1.HideReason.CANCELED);
                    });
                }
            };
            var this_1 = this;
            for (var prefix in this.mapResolvedHandlersToPrefix) {
                _loop_1(prefix);
            }
            // Complete promises that are waiting
            while (this.promisesToCompleteOnHide.length) {
                this.promisesToCompleteOnHide.pop()(true);
            }
        }
        if (reason !== quickOpenWidget_1.HideReason.FOCUS_LOST) {
            this.restoreFocus(); // focus back to editor unless user clicked somewhere else
        }
        this.inQuickOpenMode.reset();
        this.emitQuickOpenVisibilityChange(false);
    };
    QuickOpenController.prototype.hasHandler = function (prefix) {
        return !!platform_1.Registry.as(quickopen_1.Extensions.Quickopen).getQuickOpenHandler(prefix);
    };
    QuickOpenController.prototype.getEditorHistoryWithGroupLabel = function () {
        var entries = this.getEditorHistoryEntries();
        // Apply label to first entry
        if (entries.length > 0) {
            entries[0] = new EditorHistoryEntryGroup(entries[0], nls.localize('historyMatches', "recently opened"), false);
        }
        return new quickOpenModel_1.QuickOpenModel(entries, this.actionProvider);
    };
    QuickOpenController.prototype.restoreFocus = function () {
        // Try to focus active editor
        var editor = this.editorService.getActiveEditor();
        if (editor) {
            editor.focus();
        }
    };
    QuickOpenController.prototype.onType = function (value) {
        var _this = this;
        this.previousValue = value;
        // look for a handler
        var registry = platform_1.Registry.as(quickopen_1.Extensions.Quickopen);
        var handlerDescriptor = registry.getQuickOpenHandler(value);
        var defaultHandlerDescriptor = registry.getDefaultQuickOpenHandler();
        var instantProgress = handlerDescriptor && handlerDescriptor.instantProgress;
        // Use a generated token to avoid race conditions from long running promises
        var currentResultToken = uuid.generateUuid();
        this.currentResultToken = currentResultToken;
        // Reset Progress
        if (!instantProgress) {
            this.quickOpenWidget.getProgressBar().stop().getContainer().hide();
        }
        // Reset Extra Class
        this.quickOpenWidget.setExtraClass(null);
        // Remove leading and trailing whitespace
        var trimmedValue = strings.trim(value);
        // If no value provided, default to editor history
        if (!trimmedValue) {
            // Trigger onOpen
            this.resolveHandler(handlerDescriptor || defaultHandlerDescriptor)
                .done(null, errors.onUnexpectedError);
            this.quickOpenWidget.setInput(this.getEditorHistoryWithGroupLabel(), { autoFocusFirstEntry: true });
            return;
        }
        var resultPromise;
        var resultPromiseDone = false;
        if (handlerDescriptor) {
            resultPromise = this.handleSpecificHandler(handlerDescriptor, value, currentResultToken);
        }
        else {
            resultPromise = this.handleDefaultHandler(defaultHandlerDescriptor, value, currentResultToken);
        }
        // Remember as the active one
        this.previousActiveHandlerDescriptor = handlerDescriptor;
        // Progress if task takes a long time
        winjs_base_1.TPromise.timeout(instantProgress ? 0 : 800).then(function () {
            if (!resultPromiseDone && currentResultToken === _this.currentResultToken) {
                _this.quickOpenWidget.getProgressBar().infinite().getContainer().show();
            }
        });
        // Promise done handling
        resultPromise.done(function () {
            resultPromiseDone = true;
            if (currentResultToken === _this.currentResultToken) {
                _this.quickOpenWidget.getProgressBar().getContainer().hide();
            }
        }, function (error) {
            resultPromiseDone = true;
            errors.onUnexpectedError(error);
            _this.messageService.show(message_1.Severity.Error, types.isString(error) ? new Error(error) : error);
        });
    };
    QuickOpenController.prototype.handleDefaultHandler = function (handler, value, currentResultToken) {
        var _this = this;
        // Fill in history results if matching
        var matchingHistoryEntries = this.getEditorHistoryEntries(value);
        if (matchingHistoryEntries.length > 0) {
            matchingHistoryEntries[0] = new EditorHistoryEntryGroup(matchingHistoryEntries[0], nls.localize('historyMatches', "recently opened"), false);
        }
        // Resolve
        return this.resolveHandler(handler).then(function (resolvedHandler) {
            var quickOpenModel = new quickOpenModel_1.QuickOpenModel(matchingHistoryEntries, _this.actionProvider);
            var inputSet = false;
            // If we have matching entries from history we want to show them directly and not wait for the other results to come in
            // This also applies when we used to have entries from a previous run and now there are no more history results matching
            var previousInput = _this.quickOpenWidget.getInput();
            var wasShowingHistory = previousInput && previousInput.entries && previousInput.entries.some(function (e) { return e instanceof EditorHistoryEntry || e instanceof EditorHistoryEntryGroup; });
            if (wasShowingHistory || matchingHistoryEntries.length > 0) {
                (resolvedHandler.hasShortResponseTime() ? winjs_base_1.TPromise.timeout(QuickOpenController.MAX_SHORT_RESPONSE_TIME) : winjs_base_1.TPromise.as(undefined)).then(function () {
                    if (_this.currentResultToken === currentResultToken && !inputSet) {
                        _this.quickOpenWidget.setInput(quickOpenModel, { autoFocusFirstEntry: true });
                        inputSet = true;
                    }
                });
            }
            // Get results
            return resolvedHandler.getResults(value).then(function (result) {
                if (_this.currentResultToken === currentResultToken) {
                    // now is the time to show the input if we did not have set it before
                    if (!inputSet) {
                        _this.quickOpenWidget.setInput(quickOpenModel, { autoFocusFirstEntry: true });
                        inputSet = true;
                    }
                    // merge history and default handler results
                    var handlerResults = (result && result.entries) || [];
                    _this.mergeResults(quickOpenModel, handlerResults, resolvedHandler.getGroupLabel());
                }
            });
        });
    };
    QuickOpenController.prototype.getEditorHistoryEntries = function (searchValue) {
        var _this = this;
        if (searchValue) {
            searchValue = searchValue.replace(/ /g, ''); // get rid of all whitespace
        }
        // Just return all if we are not searching
        var history = this.historyService.getHistory();
        if (!searchValue) {
            return history.map(function (input) { return _this.instantiationService.createInstance(EditorHistoryEntry, input); });
        }
        var searchInPath = searchValue.indexOf(paths.nativeSep) >= 0;
        var results = [];
        history.forEach(function (input) {
            var resource;
            if (input instanceof editor_1.EditorInput) {
                resource = editor_1.getUntitledOrFileResource(input);
            }
            else {
                resource = input.resource;
            }
            if (!resource) {
                return; //For now, only support to match on inputs that provide resource information
            }
            var searchTargetToMatch;
            if (searchInPath) {
                searchTargetToMatch = labels.getPathLabel(resource, _this.contextService);
            }
            else if (input instanceof editor_1.EditorInput) {
                searchTargetToMatch = input.getName();
            }
            else {
                searchTargetToMatch = paths.basename(input.resource.fsPath);
            }
            // Check if this entry is a match for the search value
            if (!filters.matchesFuzzy(searchValue, searchTargetToMatch)) {
                return;
            }
            var entry = _this.instantiationService.createInstance(EditorHistoryEntry, input);
            var _a = quickOpenModel_1.QuickOpenEntry.highlight(entry, searchValue), labelHighlights = _a.labelHighlights, descriptionHighlights = _a.descriptionHighlights;
            entry.setHighlights(labelHighlights, descriptionHighlights);
            results.push(entry);
        });
        // Sort
        return results.sort(function (elementA, elementB) { return quickOpenModel_1.QuickOpenEntry.compare(elementA, elementB, searchValue); });
    };
    QuickOpenController.prototype.mergeResults = function (quickOpenModel, handlerResults, groupLabel) {
        // Remove results already showing by checking for a "resource" property
        var mapEntryToResource = this.mapEntriesToResource(quickOpenModel);
        var additionalHandlerResults = [];
        for (var i = 0; i < handlerResults.length; i++) {
            var result = handlerResults[i];
            var resource = result.getResource();
            if (!result.isFile() || !resource || !mapEntryToResource[resource.toString()]) {
                additionalHandlerResults.push(result);
            }
        }
        // Show additional handler results below any existing results
        if (additionalHandlerResults.length > 0) {
            var useTopBorder = quickOpenModel.getEntries().length > 0;
            additionalHandlerResults[0] = new quickOpenModel_1.QuickOpenEntryGroup(additionalHandlerResults[0], groupLabel, useTopBorder);
            quickOpenModel.addEntries(additionalHandlerResults);
            this.quickOpenWidget.refresh(quickOpenModel, { autoFocusFirstEntry: true });
        }
        else if (quickOpenModel.getEntries().length === 0) {
            quickOpenModel.addEntries([new PlaceholderQuickOpenEntry(nls.localize('noResultsFound1', "No results found"))]);
            this.quickOpenWidget.refresh(quickOpenModel, { autoFocusFirstEntry: true });
        }
    };
    QuickOpenController.prototype.handleSpecificHandler = function (handlerDescriptor, value, currentResultToken) {
        var _this = this;
        return this.resolveHandler(handlerDescriptor).then(function (resolvedHandler) {
            // Remove handler prefix from search value
            value = value.substr(handlerDescriptor.prefix.length);
            // Return early if the handler can not run in the current environment and inform the user
            var canRun = resolvedHandler.canRun();
            if (types.isUndefinedOrNull(canRun) || (typeof canRun === 'boolean' && !canRun) || typeof canRun === 'string') {
                var placeHolderLabel = (typeof canRun === 'string') ? canRun : nls.localize('canNotRunPlaceholder', "This quick open handler can not be used in the current context");
                var model = new quickOpenModel_1.QuickOpenModel([new PlaceholderQuickOpenEntry(placeHolderLabel)], _this.actionProvider);
                _this.showModel(model, resolvedHandler.getAutoFocus(value, _this.quickOpenWidget.getQuickNavigateConfiguration()), resolvedHandler.getAriaLabel());
                return winjs_base_1.TPromise.as(null);
            }
            // Support extra class from handler
            var extraClass = resolvedHandler.getClass();
            if (extraClass) {
                _this.quickOpenWidget.setExtraClass(extraClass);
            }
            // When handlers change, clear the result list first before loading the new results
            if (_this.previousActiveHandlerDescriptor !== handlerDescriptor) {
                _this.clearModel();
            }
            // Receive Results from Handler and apply
            return resolvedHandler.getResults(value).then(function (result) {
                if (_this.currentResultToken === currentResultToken) {
                    if (!result || !result.entries.length) {
                        var model = new quickOpenModel_1.QuickOpenModel([new PlaceholderQuickOpenEntry(resolvedHandler.getEmptyLabel(value))]);
                        _this.showModel(model, resolvedHandler.getAutoFocus(value, _this.quickOpenWidget.getQuickNavigateConfiguration()), resolvedHandler.getAriaLabel());
                    }
                    else {
                        _this.showModel(result, resolvedHandler.getAutoFocus(value, _this.quickOpenWidget.getQuickNavigateConfiguration()), resolvedHandler.getAriaLabel());
                    }
                }
            });
        });
    };
    QuickOpenController.prototype.showModel = function (model, autoFocus, ariaLabel) {
        // If the given model is already set in the widget, refresh and return early
        if (this.quickOpenWidget.getInput() === model) {
            this.quickOpenWidget.refresh(model, autoFocus);
            return;
        }
        // Otherwise just set it
        this.quickOpenWidget.setInput(model, autoFocus, ariaLabel);
    };
    QuickOpenController.prototype.clearModel = function () {
        this.showModel(new quickOpenModel_1.QuickOpenModel(), null);
    };
    QuickOpenController.prototype.mapEntriesToResource = function (model) {
        var entries = model.getEntries();
        var mapEntryToPath = {};
        entries.forEach(function (entry) {
            if (entry.getResource()) {
                mapEntryToPath[entry.getResource().toString()] = entry;
            }
        });
        return mapEntryToPath;
    };
    QuickOpenController.prototype.resolveHandler = function (handler) {
        var _this = this;
        var result = this._resolveHandler(handler);
        var id = handler.getId();
        if (!this.handlerOnOpenCalled[id]) {
            var original_1 = result;
            this.handlerOnOpenCalled[id] = true;
            result = this.mapResolvedHandlersToPrefix[id] = original_1.then(function (resolved) {
                _this.mapResolvedHandlersToPrefix[id] = original_1;
                resolved.onOpen();
                return resolved;
            });
        }
        return result.then(null, function (error) {
            delete _this.mapResolvedHandlersToPrefix[id];
            return winjs_base_1.TPromise.wrapError('Unable to instantiate quick open handler ' + handler.moduleName + ' - ' + handler.ctorName + ': ' + JSON.stringify(error));
        });
    };
    QuickOpenController.prototype._resolveHandler = function (handler) {
        var id = handler.getId();
        // Return Cached
        if (this.mapResolvedHandlersToPrefix[id]) {
            return this.mapResolvedHandlersToPrefix[id];
        }
        // Otherwise load and create
        return this.mapResolvedHandlersToPrefix[id] = this.instantiationService.createInstance(handler);
    };
    QuickOpenController.prototype.layout = function (dimension) {
        this.layoutDimensions = dimension;
        if (this.quickOpenWidget) {
            this.quickOpenWidget.layout(this.layoutDimensions);
        }
        if (this.pickOpenWidget) {
            this.pickOpenWidget.layout(this.layoutDimensions);
        }
    };
    QuickOpenController.prototype.dispose = function () {
        if (this.quickOpenWidget) {
            this.quickOpenWidget.dispose();
        }
        if (this.pickOpenWidget) {
            this.pickOpenWidget.dispose();
        }
        _super.prototype.dispose.call(this);
    };
    QuickOpenController.MAX_SHORT_RESPONSE_TIME = 500;
    QuickOpenController.ID = 'workbench.component.quickopen';
    QuickOpenController = __decorate([
        __param(0, editorService_1.IWorkbenchEditorService),
        __param(1, message_1.IMessageService),
        __param(2, telemetry_1.ITelemetryService),
        __param(3, workspace_1.IWorkspaceContextService),
        __param(4, contextkey_1.IContextKeyService),
        __param(5, configuration_1.IConfigurationService),
        __param(6, history_1.IHistoryService),
        __param(7, instantiation_1.IInstantiationService),
        __param(8, partService_1.IPartService)
    ], QuickOpenController);
    return QuickOpenController;
}(component_1.WorkbenchComponent));
exports.QuickOpenController = QuickOpenController;
var PlaceholderQuickOpenEntry = (function (_super) {
    __extends(PlaceholderQuickOpenEntry, _super);
    function PlaceholderQuickOpenEntry(placeHolderLabel) {
        _super.call(this);
        this.placeHolderLabel = placeHolderLabel;
    }
    PlaceholderQuickOpenEntry.prototype.getLabel = function () {
        return this.placeHolderLabel;
    };
    return PlaceholderQuickOpenEntry;
}(quickOpenModel_1.QuickOpenEntryGroup));
var PickOpenEntry = (function (_super) {
    __extends(PickOpenEntry, _super);
    function PickOpenEntry(item, onPreview, modeService, modelService) {
        _super.call(this, item.label);
        this.onPreview = onPreview;
        this.modeService = modeService;
        this.modelService = modelService;
        this.description = item.description;
        this.detail = item.detail;
        this.hasSeparator = item.separator && item.separator.border;
        this.separatorLabel = item.separator && item.separator.label;
        this.alwaysShow = item.alwaysShow;
        var fileItem = item;
        this.resource = fileItem.resource;
        this.isFolder = fileItem.isFolder;
    }
    PickOpenEntry.prototype.getLabelOptions = function () {
        return {
            extraClasses: this.resource ? labels_1.getIconClasses(this.modelService, this.modeService, this.resource, this.isFolder) : []
        };
    };
    Object.defineProperty(PickOpenEntry.prototype, "shouldRunWithContext", {
        get: function () {
            return this._shouldRunWithContext;
        },
        enumerable: true,
        configurable: true
    });
    PickOpenEntry.prototype.getDescription = function () {
        return this.description;
    };
    PickOpenEntry.prototype.getDetail = function () {
        return this.detail;
    };
    PickOpenEntry.prototype.showBorder = function () {
        return this.hasSeparator;
    };
    PickOpenEntry.prototype.getGroupLabel = function () {
        return this.separatorLabel;
    };
    PickOpenEntry.prototype.shouldAlwaysShow = function () {
        return this.alwaysShow;
    };
    PickOpenEntry.prototype.run = function (mode, context) {
        if (mode === quickOpen_1.Mode.OPEN) {
            this._shouldRunWithContext = context;
            return true;
        }
        if (mode === quickOpen_1.Mode.PREVIEW && this.onPreview) {
            this.onPreview();
        }
        return false;
    };
    PickOpenEntry = __decorate([
        __param(2, modeService_1.IModeService),
        __param(3, modelService_1.IModelService)
    ], PickOpenEntry);
    return PickOpenEntry;
}(PlaceholderQuickOpenEntry));
var EditorHistoryEntryGroup = (function (_super) {
    __extends(EditorHistoryEntryGroup, _super);
    function EditorHistoryEntryGroup() {
        _super.apply(this, arguments);
    }
    return EditorHistoryEntryGroup;
}(quickOpenModel_1.QuickOpenEntryGroup));
exports.EditorHistoryEntryGroup = EditorHistoryEntryGroup;
var EditorHistoryEntry = (function (_super) {
    __extends(EditorHistoryEntry, _super);
    function EditorHistoryEntry(input, editorService, modeService, modelService, textFileService, contextService, configurationService) {
        _super.call(this, editorService);
        this.modeService = modeService;
        this.modelService = modelService;
        this.textFileService = textFileService;
        this.configurationService = configurationService;
        this.input = input;
        if (input instanceof editor_1.EditorInput) {
            this.resource = editor_1.getUntitledOrFileResource(input);
            this.label = input.getName();
            this.description = input.getDescription();
        }
        else {
            var resourceInput = input;
            this.resource = resourceInput.resource;
            this.label = paths.basename(resourceInput.resource.fsPath);
            this.description = labels.getPathLabel(paths.dirname(this.resource.fsPath), contextService);
        }
    }
    EditorHistoryEntry.prototype.getIcon = function () {
        return this.resource && this.textFileService.isDirty(this.resource) ? 'dirty' : '';
    };
    EditorHistoryEntry.prototype.getLabel = function () {
        return this.label;
    };
    EditorHistoryEntry.prototype.getLabelOptions = function () {
        return {
            extraClasses: labels_1.getIconClasses(this.modelService, this.modeService, this.resource)
        };
    };
    EditorHistoryEntry.prototype.getAriaLabel = function () {
        return nls.localize('entryAriaLabel', "{0}, recently opened", this.getLabel());
    };
    EditorHistoryEntry.prototype.getDescription = function () {
        return this.description;
    };
    EditorHistoryEntry.prototype.getResource = function () {
        return this.resource;
    };
    EditorHistoryEntry.prototype.getInput = function () {
        return this.input;
    };
    EditorHistoryEntry.prototype.run = function (mode, context) {
        if (mode === quickOpen_1.Mode.OPEN) {
            var sideBySide = !context.quickNavigateConfiguration && context.keymods.indexOf(32768 /* CtrlCmd */) >= 0;
            var pinned = !this.configurationService.getConfiguration().workbench.editor.enablePreviewFromQuickOpen;
            if (this.input instanceof editor_1.EditorInput) {
                this.editorService.openEditor(this.input, { pinned: pinned }, sideBySide).done(null, errors.onUnexpectedError);
            }
            else {
                this.editorService.openEditor({ resource: this.input.resource, options: { pinned: true } }, sideBySide);
            }
            return true;
        }
        return false;
    };
    EditorHistoryEntry = __decorate([
        __param(1, editorService_1.IWorkbenchEditorService),
        __param(2, modeService_1.IModeService),
        __param(3, modelService_1.IModelService),
        __param(4, textfiles_1.ITextFileService),
        __param(5, workspace_1.IWorkspaceContextService),
        __param(6, configuration_1.IConfigurationService)
    ], EditorHistoryEntry);
    return EditorHistoryEntry;
}(quickopen_1.EditorQuickOpenEntry));
exports.EditorHistoryEntry = EditorHistoryEntry;
