// vscode/src/vs/workbench/parts/content/dynamicContentInput.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var strings = require("vs/base/common/strings");
var winjs_base_1 = require("vs/base/common/winjs.base");
var editor_1 = require("vs/workbench/common/editor");
var lifecycle_1 = require("vs/base/common/lifecycle");
var telemetryUtils_1 = require("vs/platform/telemetry/common/telemetryUtils");
var resolverService_1 = require("vs/editor/common/services/resolverService");
var marked_1 = require("vs/base/common/marked/marked");
var network_1 = require("vs/base/common/network");
var telemetry_1 = require("vs/platform/telemetry/common/telemetry");
var lifecycle_2 = require("vs/platform/lifecycle/common/lifecycle");
var DynamicContentModel = (function (_super) {
    __extends(DynamicContentModel, _super);
    function DynamicContentModel(mainRef, snippetRefs) {
        var _this = _super.call(this) || this;
        _this.mainRef = mainRef;
        _this.snippetRefs = snippetRefs;
        return _this;
    }
    Object.defineProperty(DynamicContentModel.prototype, "main", {
        get: function () {
            return this.mainRef.object;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(DynamicContentModel.prototype, "snippets", {
        get: function () {
            return this.snippetRefs.map(function (snippet) { return snippet.object; });
        },
        enumerable: true,
        configurable: true
    });
    DynamicContentModel.prototype.dispose = function () {
        this.snippetRefs.forEach(function (ref) { return ref.dispose(); });
        this.mainRef.dispose();
        _super.prototype.dispose.call(this);
    };
    return DynamicContentModel;
}(editor_1.EditorModel));
exports.DynamicContentModel = DynamicContentModel;
var DynamicContentInput = (function (_super) {
    __extends(DynamicContentInput, _super);
    function DynamicContentInput(name, description, resource, staticContent, telemetryFrom, onReady, telemetryService, lifecycleService, textModelResolverService) {
        var _this = _super.call(this) || this;
        _this.name = name;
        _this.description = description;
        _this.resource = resource;
        _this.staticContent = staticContent;
        _this.telemetryFrom = telemetryFrom;
        _this.onReady = onReady;
        _this.telemetryService = telemetryService;
        _this.textModelResolverService = textModelResolverService;
        _this.disposables = [];
        _this.maxTopScroll = 0;
        _this.maxBottomScroll = 0;
        _this.disposables.push(lifecycleService.onShutdown(function (e) { return _this.disposeTelemetry(e); }));
        return _this;
    }
    DynamicContentInput.prototype.getResource = function () {
        return this.resource;
    };
    DynamicContentInput.prototype.getTypeId = function () {
        return DynamicContentInput.ID;
    };
    DynamicContentInput.prototype.getName = function () {
        return this.name;
    };
    DynamicContentInput.prototype.getDescription = function () {
        return this.description;
    };
    DynamicContentInput.prototype.getTelemetryFrom = function () {
        return this.telemetryFrom || 'dynamicContent';
    };
    DynamicContentInput.prototype.getTelemetryDescriptor = function () {
        var descriptor = _super.prototype.getTelemetryDescriptor.call(this);
        descriptor['target'] = this.getTelemetryFrom();
        descriptor['resource'] = telemetryUtils_1.telemetryURIDescriptor(this.resource);
        return descriptor;
    };
    DynamicContentInput.prototype.resolve = function (refresh) {
        var _this = this;
        console.log("resolving!");
        if (!this.promise) {
            this.resolveTelemetry();
            this.promise = this.textModelResolverService.createModelReference(this.resource)
                .then(function (ref) {
                if (strings.endsWith(_this.getResource().path, '.html')) {
                    return new DynamicContentModel(ref, []);
                }
                var snippets = [];
                var i = 0;
                var renderer = new marked_1.marked.Renderer();
                renderer.code = function (code, lang) {
                    var resource = _this.resource.with({ scheme: network_1.Schemas.walkThroughSnippet, fragment: i++ + "." + lang });
                    snippets.push(_this.textModelResolverService.createModelReference(resource));
                    return '';
                };
                var markdown = _this.staticContent;
                marked_1.marked(markdown, { renderer: renderer });
                return winjs_base_1.TPromise.join(snippets)
                    .then(function (refs) { return new DynamicContentModel(ref, refs); });
            });
        }
        return this.promise;
    };
    DynamicContentInput.prototype.matches = function (otherInput) {
        if (_super.prototype.matches.call(this, otherInput) === true) {
            return true;
        }
        if (otherInput instanceof DynamicContentInput) {
            var otherResourceEditorInput = otherInput;
            // Compare by properties
            return otherResourceEditorInput.resource.toString() === this.resource.toString();
        }
        return false;
    };
    DynamicContentInput.prototype.dispose = function () {
        this.disposables = lifecycle_1.dispose(this.disposables);
        if (this.promise) {
            this.promise.then(function (model) { return model.dispose(); });
            this.promise = null;
        }
        this.disposeTelemetry();
        _super.prototype.dispose.call(this);
    };
    DynamicContentInput.prototype.relativeScrollPosition = function (topScroll, bottomScroll) {
        this.maxTopScroll = Math.max(this.maxTopScroll, topScroll);
        this.maxBottomScroll = Math.max(this.maxBottomScroll, bottomScroll);
    };
    DynamicContentInput.prototype.resolveTelemetry = function () {
        if (!this.resolveTime) {
            this.resolveTime = Date.now();
            this.telemetryService.publicLog('resolvingInput', {
                target: this.getTelemetryFrom(),
            });
        }
    };
    DynamicContentInput.prototype.disposeTelemetry = function (reason) {
        if (this.resolveTime) {
            this.telemetryService.publicLog('disposingInput', {
                target: this.getTelemetryFrom(),
                timeSpent: (Date.now() - this.resolveTime) / 60,
                reason: reason ? lifecycle_2.ShutdownReason[reason] : 'DISPOSE',
                maxTopScroll: this.maxTopScroll,
                maxBottomScroll: this.maxBottomScroll,
            });
            this.resolveTime = null;
        }
    };
    return DynamicContentInput;
}(editor_1.EditorInput));
DynamicContentInput.ID = 'workbench.editors.dynamicContentInput';
DynamicContentInput = __decorate([
    __param(6, telemetry_1.ITelemetryService),
    __param(7, lifecycle_2.ILifecycleService),
    __param(8, resolverService_1.ITextModelResolverService)
], DynamicContentInput);
exports.DynamicContentInput = DynamicContentInput;
