// vscode/src/vs/workbench/parts/git/browser/views/notroot/notrootView.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
require("./notrootView.css");
var nls = require("vs/nls");
var winjs = require("vs/base/common/winjs.base");
var ee = require("vs/base/common/eventEmitter");
var builder = require("vs/base/browser/builder");
var $ = builder.$;
var NotRootView = (function (_super) {
    __extends(NotRootView, _super);
    function NotRootView() {
        var _this = _super.apply(this, arguments) || this;
        _this.ID = 'notroot';
        return _this;
    }
    Object.defineProperty(NotRootView.prototype, "element", {
        get: function () {
            if (!this._element) {
                this.render();
            }
            return this._element;
        },
        enumerable: true,
        configurable: true
    });
    NotRootView.prototype.render = function () {
        this._element = $([
            '<div class="notroot-view">',
            '<p>', nls.localize('wrongRoot', "This directory seems to be contained in a git repository."), '</p>',
            '<p>', nls.localize('pleaseRestart', "Open the repository's root directory in order to access Git features."), '</p>',
            '</div>'
        ].join('')).getHTMLElement();
    };
    NotRootView.prototype.focus = function () {
        return;
    };
    NotRootView.prototype.layout = function (dimension) {
        return;
    };
    NotRootView.prototype.setVisible = function (visible) {
        return winjs.TPromise.as(null);
    };
    NotRootView.prototype.getControl = function () {
        return null;
    };
    NotRootView.prototype.getActions = function () {
        return [];
    };
    NotRootView.prototype.getSecondaryActions = function () {
        return [];
    };
    return NotRootView;
}(ee.EventEmitter));
exports.NotRootView = NotRootView;
