// vscode/src/vs/editor/common/services/editorSimpleWorker.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var uri_1 = require('vs/base/common/uri');
var winjs_base_1 = require('vs/base/common/winjs.base');
var range_1 = require('vs/editor/common/core/range');
var filters_1 = require('vs/base/common/filters');
var diffComputer_1 = require('vs/editor/common/diff/diffComputer');
var mirrorModel2_1 = require('vs/editor/common/model/mirrorModel2');
var linkComputer_1 = require('vs/editor/common/modes/linkComputer');
var inplaceReplaceSupport_1 = require('vs/editor/common/modes/supports/inplaceReplaceSupport');
var wordHelper_1 = require('vs/editor/common/model/wordHelper');
var standaloneBase_1 = require('vs/editor/common/standalone/standaloneBase');
/**
 * @internal
 */
var MirrorModel = (function (_super) {
    __extends(MirrorModel, _super);
    function MirrorModel() {
        _super.apply(this, arguments);
    }
    Object.defineProperty(MirrorModel.prototype, "uri", {
        get: function () {
            return this._uri;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MirrorModel.prototype, "version", {
        get: function () {
            return this._versionId;
        },
        enumerable: true,
        configurable: true
    });
    MirrorModel.prototype.getValue = function () {
        return this.getText();
    };
    MirrorModel.prototype.getLinesContent = function () {
        return this._lines.slice(0);
    };
    MirrorModel.prototype.getLineCount = function () {
        return this._lines.length;
    };
    MirrorModel.prototype.getLineContent = function (lineNumber) {
        return this._lines[lineNumber - 1];
    };
    MirrorModel.prototype.getWordAtPosition = function (position, wordDefinition) {
        var wordAtText = wordHelper_1.getWordAtText(position.column, wordHelper_1.ensureValidWordDefinition(wordDefinition), this._lines[position.lineNumber - 1], 0);
        if (wordAtText) {
            return new range_1.Range(position.lineNumber, wordAtText.startColumn, position.lineNumber, wordAtText.endColumn);
        }
        return null;
    };
    MirrorModel.prototype.getWordUntilPosition = function (position, wordDefinition) {
        var wordAtPosition = this.getWordAtPosition(position, wordDefinition);
        if (!wordAtPosition) {
            return {
                word: '',
                startColumn: position.column,
                endColumn: position.column
            };
        }
        return {
            word: this._lines[position.lineNumber - 1].substring(wordAtPosition.startColumn - 1, position.column - 1),
            startColumn: wordAtPosition.startColumn,
            endColumn: position.column
        };
    };
    MirrorModel.prototype._getAllWords = function (wordDefinition) {
        var _this = this;
        var result = [];
        this._lines.forEach(function (line) {
            _this._wordenize(line, wordDefinition).forEach(function (info) {
                result.push(line.substring(info.start, info.end));
            });
        });
        return result;
    };
    MirrorModel.prototype.getAllUniqueWords = function (wordDefinition, skipWordOnce) {
        var foundSkipWord = false;
        var uniqueWords = Object.create(null);
        return this._getAllWords(wordDefinition).filter(function (word) {
            if (skipWordOnce && !foundSkipWord && skipWordOnce === word) {
                foundSkipWord = true;
                return false;
            }
            else if (uniqueWords[word]) {
                return false;
            }
            else {
                uniqueWords[word] = true;
                return true;
            }
        });
    };
    // TODO@Joh, TODO@Alex - remove these and make sure the super-things work
    MirrorModel.prototype._wordenize = function (content, wordDefinition) {
        var result = [];
        var match;
        while (match = wordDefinition.exec(content)) {
            if (match[0].length === 0) {
                // it did match the empty string
                break;
            }
            result.push({ start: match.index, end: match.index + match[0].length });
        }
        return result;
    };
    MirrorModel.prototype.getValueInRange = function (range) {
        if (range.startLineNumber === range.endLineNumber) {
            return this._lines[range.startLineNumber - 1].substring(range.startColumn - 1, range.endColumn - 1);
        }
        var lineEnding = this._eol, startLineIndex = range.startLineNumber - 1, endLineIndex = range.endLineNumber - 1, resultLines = [];
        resultLines.push(this._lines[startLineIndex].substring(range.startColumn - 1));
        for (var i = startLineIndex + 1; i < endLineIndex; i++) {
            resultLines.push(this._lines[i]);
        }
        resultLines.push(this._lines[endLineIndex].substring(0, range.endColumn - 1));
        return resultLines.join(lineEnding);
    };
    return MirrorModel;
}(mirrorModel2_1.MirrorModel2));
exports.MirrorModel = MirrorModel;
/**
 * @internal
 */
var BaseEditorSimpleWorker = (function () {
    function BaseEditorSimpleWorker() {
        this._foreignModule = null;
    }
    // ---- BEGIN diff --------------------------------------------------------------------------
    BaseEditorSimpleWorker.prototype.computeDiff = function (originalUrl, modifiedUrl, ignoreTrimWhitespace) {
        var original = this._getModel(originalUrl);
        var modified = this._getModel(modifiedUrl);
        if (!original || !modified) {
            return null;
        }
        var originalLines = original.getLinesContent();
        var modifiedLines = modified.getLinesContent();
        var diffComputer = new diffComputer_1.DiffComputer(originalLines, modifiedLines, {
            shouldPostProcessCharChanges: true,
            shouldIgnoreTrimWhitespace: ignoreTrimWhitespace,
            shouldConsiderTrimWhitespaceInEmptyCase: true
        });
        return winjs_base_1.TPromise.as(diffComputer.computeDiff());
    };
    BaseEditorSimpleWorker.prototype.computeDirtyDiff = function (originalUrl, modifiedUrl, ignoreTrimWhitespace) {
        var original = this._getModel(originalUrl);
        var modified = this._getModel(modifiedUrl);
        if (!original || !modified) {
            return null;
        }
        var originalLines = original.getLinesContent();
        var modifiedLines = modified.getLinesContent();
        var diffComputer = new diffComputer_1.DiffComputer(originalLines, modifiedLines, {
            shouldPostProcessCharChanges: false,
            shouldIgnoreTrimWhitespace: ignoreTrimWhitespace,
            shouldConsiderTrimWhitespaceInEmptyCase: false
        });
        return winjs_base_1.TPromise.as(diffComputer.computeDiff());
    };
    // ---- END diff --------------------------------------------------------------------------
    BaseEditorSimpleWorker.prototype.computeLinks = function (modelUrl) {
        var model = this._getModel(modelUrl);
        if (!model) {
            return null;
        }
        return winjs_base_1.TPromise.as(linkComputer_1.computeLinks(model));
    };
    // ---- BEGIN suggest --------------------------------------------------------------------------
    BaseEditorSimpleWorker.prototype.textualSuggest = function (modelUrl, position, wordDef, wordDefFlags) {
        var model = this._getModel(modelUrl);
        if (!model) {
            return null;
        }
        return winjs_base_1.TPromise.as(this._suggestFiltered(model, position, new RegExp(wordDef, wordDefFlags)));
    };
    BaseEditorSimpleWorker.prototype._suggestFiltered = function (model, position, wordDefRegExp) {
        var value = this._suggestUnfiltered(model, position, wordDefRegExp);
        // filter suggestions
        return {
            currentWord: value.currentWord,
            suggestions: value.suggestions.filter(function (element) { return !!filters_1.fuzzyContiguousFilter(value.currentWord, element.label); }),
            incomplete: value.incomplete
        };
    };
    BaseEditorSimpleWorker.prototype._suggestUnfiltered = function (model, position, wordDefRegExp) {
        var currentWord = model.getWordUntilPosition(position, wordDefRegExp).word;
        var allWords = model.getAllUniqueWords(wordDefRegExp, currentWord);
        var suggestions = allWords.filter(function (word) {
            return !(/^-?\d*\.?\d/.test(word)); // filter out numbers
        }).map(function (word) {
            return {
                type: 'text',
                label: word,
                insertText: word,
                noAutoAccept: true
            };
        });
        return {
            currentWord: currentWord,
            suggestions: suggestions
        };
    };
    // ---- END suggest --------------------------------------------------------------------------
    BaseEditorSimpleWorker.prototype.navigateValueSet = function (modelUrl, range, up, wordDef, wordDefFlags) {
        var model = this._getModel(modelUrl);
        if (!model) {
            return null;
        }
        var wordDefRegExp = new RegExp(wordDef, wordDefFlags);
        if (range.startColumn === range.endColumn) {
            range.endColumn += 1;
        }
        var selectionText = model.getValueInRange(range);
        var wordRange = model.getWordAtPosition({ lineNumber: range.startLineNumber, column: range.startColumn }, wordDefRegExp);
        var word = null;
        if (wordRange !== null) {
            word = model.getValueInRange(wordRange);
        }
        var result = inplaceReplaceSupport_1.BasicInplaceReplace.INSTANCE.navigateValueSet(range, selectionText, wordRange, word, up);
        return winjs_base_1.TPromise.as(result);
    };
    // ---- BEGIN foreign module support --------------------------------------------------------------------------
    BaseEditorSimpleWorker.prototype.loadForeignModule = function (moduleId, createData) {
        var _this = this;
        return new winjs_base_1.TPromise(function (c, e) {
            // Use the global require to be sure to get the global config
            self.require([moduleId], function (foreignModule) {
                var ctx = {
                    getMirrorModels: function () {
                        return _this._getModels();
                    }
                };
                _this._foreignModule = foreignModule.create(ctx, createData);
                var methods = [];
                for (var prop in _this._foreignModule) {
                    if (typeof _this._foreignModule[prop] === 'function') {
                        methods.push(prop);
                    }
                }
                c(methods);
            }, e);
        });
    };
    // foreign method request
    BaseEditorSimpleWorker.prototype.fmr = function (method, args) {
        if (!this._foreignModule || typeof this._foreignModule[method] !== 'function') {
            return winjs_base_1.TPromise.wrapError(new Error('Missing requestHandler or method: ' + method));
        }
        try {
            return winjs_base_1.TPromise.as(this._foreignModule[method].apply(this._foreignModule, args));
        }
        catch (e) {
            return winjs_base_1.TPromise.wrapError(e);
        }
    };
    return BaseEditorSimpleWorker;
}());
exports.BaseEditorSimpleWorker = BaseEditorSimpleWorker;
/**
 * @internal
 */
var EditorSimpleWorkerImpl = (function (_super) {
    __extends(EditorSimpleWorkerImpl, _super);
    function EditorSimpleWorkerImpl() {
        _super.call(this);
        this._models = Object.create(null);
    }
    EditorSimpleWorkerImpl.prototype.dispose = function () {
        this._models = Object.create(null);
    };
    EditorSimpleWorkerImpl.prototype._getModel = function (uri) {
        return this._models[uri];
    };
    EditorSimpleWorkerImpl.prototype._getModels = function () {
        var _this = this;
        var all = [];
        Object.keys(this._models).forEach(function (key) { return all.push(_this._models[key]); });
        return all;
    };
    EditorSimpleWorkerImpl.prototype.acceptNewModel = function (data) {
        this._models[data.url] = new MirrorModel(uri_1.default.parse(data.url), data.value.lines, data.value.EOL, data.versionId);
    };
    EditorSimpleWorkerImpl.prototype.acceptModelChanged = function (strURL, events) {
        if (!this._models[strURL]) {
            return;
        }
        var model = this._models[strURL];
        model.onEvents(events);
    };
    EditorSimpleWorkerImpl.prototype.acceptRemovedModel = function (strURL) {
        if (!this._models[strURL]) {
            return;
        }
        delete this._models[strURL];
    };
    return EditorSimpleWorkerImpl;
}(BaseEditorSimpleWorker));
exports.EditorSimpleWorkerImpl = EditorSimpleWorkerImpl;
/**
 * Called on the worker side
 * @internal
 */
function create() {
    return new EditorSimpleWorkerImpl();
}
exports.create = create;
var global = self;
var isWebWorker = (typeof global.importScripts === 'function');
if (isWebWorker) {
    global.monaco = standaloneBase_1.createMonacoBaseAPI();
}
