// vscode/src/vs/editor/common/config/commonEditorConfig.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var nls = require("vs/nls");
var event_1 = require("vs/base/common/event");
var lifecycle_1 = require("vs/base/common/lifecycle");
var objects = require("vs/base/common/objects");
var platform = require("vs/base/common/platform");
var configurationRegistry_1 = require("vs/platform/configuration/common/configurationRegistry");
var platform_1 = require("vs/platform/platform");
var defaultConfig_1 = require("vs/editor/common/config/defaultConfig");
var editorCommon = require("vs/editor/common/editorCommon");
var editorLayoutProvider_1 = require("vs/editor/common/viewLayout/editorLayoutProvider");
var scrollable_1 = require("vs/base/common/scrollable");
var fontInfo_1 = require("vs/editor/common/config/fontInfo");
var editorZoom_1 = require("vs/editor/common/config/editorZoom");
exports.TabFocus = new (function () {
    function class_1() {
        this._tabFocus = false;
        this._onDidChangeTabFocus = new event_1.Emitter();
        this.onDidChangeTabFocus = this._onDidChangeTabFocus.event;
    }
    class_1.prototype.getTabFocusMode = function () {
        return this._tabFocus;
    };
    class_1.prototype.setTabFocusMode = function (tabFocusMode) {
        if (this._tabFocus === tabFocusMode) {
            return;
        }
        this._tabFocus = tabFocusMode;
        this._onDidChangeTabFocus.fire(this._tabFocus);
    };
    return class_1;
}());
/**
 * Experimental screen reader support toggle
 */
var GlobalScreenReaderNVDA = (function () {
    function GlobalScreenReaderNVDA() {
    }
    GlobalScreenReaderNVDA.getValue = function () {
        return this._value;
    };
    GlobalScreenReaderNVDA.setValue = function (value) {
        if (this._value === value) {
            return;
        }
        this._value = value;
        this._onChange.fire(this._value);
    };
    return GlobalScreenReaderNVDA;
}());
GlobalScreenReaderNVDA._value = false;
GlobalScreenReaderNVDA._onChange = new event_1.Emitter();
GlobalScreenReaderNVDA.onChange = GlobalScreenReaderNVDA._onChange.event;
exports.GlobalScreenReaderNVDA = GlobalScreenReaderNVDA;
var ConfigurationWithDefaults = (function () {
    function ConfigurationWithDefaults(options) {
        this._editor = objects.clone(defaultConfig_1.DefaultConfig.editor);
        this._mergeOptionsIn(options);
    }
    ConfigurationWithDefaults.prototype.getEditorOptions = function () {
        return this._editor;
    };
    ConfigurationWithDefaults.prototype._mergeOptionsIn = function (newOptions) {
        this._editor = objects.mixin(this._editor, newOptions || {});
    };
    ConfigurationWithDefaults.prototype.updateOptions = function (newOptions) {
        // Apply new options
        this._mergeOptionsIn(newOptions);
    };
    return ConfigurationWithDefaults;
}());
exports.ConfigurationWithDefaults = ConfigurationWithDefaults;
var InternalEditorOptionsHelper = (function () {
    function InternalEditorOptionsHelper() {
    }
    InternalEditorOptionsHelper.createInternalEditorOptions = function (outerWidth, outerHeight, opts, fontInfo, editorClassName, isDominatedByLongLines, lineNumbersDigitCount, canUseTranslate3d, pixelRatio) {
        var stopRenderingLineAfter;
        if (typeof opts.stopRenderingLineAfter !== 'undefined') {
            stopRenderingLineAfter = toInteger(opts.stopRenderingLineAfter, -1);
        }
        else {
            stopRenderingLineAfter = 10000;
        }
        var scrollbar = this._sanitizeScrollbarOpts(opts.scrollbar, toFloat(opts.mouseWheelScrollSensitivity, 1));
        var minimap = this._sanitizeMinimapOpts(opts.minimap);
        var glyphMargin = toBoolean(opts.glyphMargin);
        var lineNumbersMinChars = toInteger(opts.lineNumbersMinChars, 1);
        var lineDecorationsWidth;
        if (typeof opts.lineDecorationsWidth === 'string' && /^\d+(\.\d+)?ch$/.test(opts.lineDecorationsWidth)) {
            var multiple = parseFloat(opts.lineDecorationsWidth.substr(0, opts.lineDecorationsWidth.length - 2));
            lineDecorationsWidth = multiple * fontInfo.typicalHalfwidthCharacterWidth;
        }
        else {
            lineDecorationsWidth = toInteger(opts.lineDecorationsWidth, 0);
        }
        if (opts.folding) {
            lineDecorationsWidth += 16;
        }
        var renderLineNumbers;
        var renderCustomLineNumbers;
        var renderRelativeLineNumbers;
        {
            var lineNumbers = opts.lineNumbers;
            // Compatibility with old true or false values
            if (lineNumbers === true) {
                lineNumbers = 'on';
            }
            else if (lineNumbers === false) {
                lineNumbers = 'off';
            }
            if (typeof lineNumbers === 'function') {
                renderLineNumbers = true;
                renderCustomLineNumbers = lineNumbers;
                renderRelativeLineNumbers = false;
            }
            else if (lineNumbers === 'relative') {
                renderLineNumbers = true;
                renderCustomLineNumbers = null;
                renderRelativeLineNumbers = true;
            }
            else if (lineNumbers === 'on') {
                renderLineNumbers = true;
                renderCustomLineNumbers = null;
                renderRelativeLineNumbers = false;
            }
            else {
                renderLineNumbers = false;
                renderCustomLineNumbers = null;
                renderRelativeLineNumbers = false;
            }
        }
        var layoutInfo = editorLayoutProvider_1.EditorLayoutProvider.compute({
            outerWidth: outerWidth,
            outerHeight: outerHeight,
            showGlyphMargin: glyphMargin,
            lineHeight: fontInfo.lineHeight,
            showLineNumbers: renderLineNumbers,
            lineNumbersMinChars: lineNumbersMinChars,
            lineNumbersDigitCount: lineNumbersDigitCount,
            lineDecorationsWidth: lineDecorationsWidth,
            typicalHalfwidthCharacterWidth: fontInfo.typicalHalfwidthCharacterWidth,
            maxDigitWidth: fontInfo.maxDigitWidth,
            verticalScrollbarWidth: scrollbar.verticalScrollbarSize,
            horizontalScrollbarHeight: scrollbar.horizontalScrollbarSize,
            scrollbarArrowSize: scrollbar.arrowSize,
            verticalScrollbarHasArrows: scrollbar.verticalHasArrows,
            minimap: minimap.enabled,
            minimapRenderCharacters: minimap.renderCharacters,
            minimapMaxColumn: minimap.maxColumn,
            pixelRatio: pixelRatio
        });
        var bareWrappingInfo = null;
        {
            var wordWrap = opts.wordWrap;
            var wordWrapColumn = toInteger(opts.wordWrapColumn, 1);
            var wordWrapMinified = toBoolean(opts.wordWrapMinified);
            // Compatibility with old true or false values
            if (wordWrap === true) {
                wordWrap = 'on';
            }
            else if (wordWrap === false) {
                wordWrap = 'off';
            }
            if (wordWrapMinified && isDominatedByLongLines) {
                // Force viewport width wrapping if model is dominated by long lines
                bareWrappingInfo = {
                    isViewportWrapping: true,
                    wrappingColumn: Math.max(1, layoutInfo.viewportColumn)
                };
            }
            else if (wordWrap === 'on') {
                bareWrappingInfo = {
                    isViewportWrapping: true,
                    wrappingColumn: Math.max(1, layoutInfo.viewportColumn)
                };
            }
            else if (wordWrap === 'bounded') {
                bareWrappingInfo = {
                    isViewportWrapping: true,
                    wrappingColumn: Math.min(Math.max(1, layoutInfo.viewportColumn), wordWrapColumn)
                };
            }
            else if (wordWrap === 'wordWrapColumn') {
                bareWrappingInfo = {
                    isViewportWrapping: false,
                    wrappingColumn: wordWrapColumn
                };
            }
            else {
                bareWrappingInfo = {
                    isViewportWrapping: false,
                    wrappingColumn: -1
                };
            }
        }
        var wrappingInfo = new editorCommon.EditorWrappingInfo({
            isViewportWrapping: bareWrappingInfo.isViewportWrapping,
            wrappingColumn: bareWrappingInfo.wrappingColumn,
            wrappingIndent: wrappingIndentFromString(opts.wrappingIndent),
            wordWrapBreakBeforeCharacters: String(opts.wordWrapBreakBeforeCharacters),
            wordWrapBreakAfterCharacters: String(opts.wordWrapBreakAfterCharacters),
            wordWrapBreakObtrusiveCharacters: String(opts.wordWrapBreakObtrusiveCharacters),
        });
        var readOnly = toBoolean(opts.readOnly);
        var tabFocusMode = exports.TabFocus.getTabFocusMode();
        if (readOnly) {
            tabFocusMode = true;
        }
        var renderWhitespace = opts.renderWhitespace;
        // Compatibility with old true or false values
        if (renderWhitespace === true) {
            renderWhitespace = 'boundary';
        }
        else if (renderWhitespace === false) {
            renderWhitespace = 'none';
        }
        var renderLineHighlight = opts.renderLineHighlight;
        // Compatibility with old true or false values
        if (renderLineHighlight === true) {
            renderLineHighlight = 'line';
        }
        else if (renderLineHighlight === false) {
            renderLineHighlight = 'none';
        }
        var viewInfo = new editorCommon.InternalEditorViewOptions({
            theme: opts.theme,
            canUseTranslate3d: canUseTranslate3d,
            disableMonospaceOptimizations: (toBoolean(opts.disableMonospaceOptimizations) || toBoolean(opts.fontLigatures)),
            experimentalScreenReader: toBoolean(opts.experimentalScreenReader),
            rulers: toSortedIntegerArray(opts.rulers),
            ariaLabel: String(opts.ariaLabel),
            renderLineNumbers: renderLineNumbers,
            renderCustomLineNumbers: renderCustomLineNumbers,
            renderRelativeLineNumbers: renderRelativeLineNumbers,
            selectOnLineNumbers: toBoolean(opts.selectOnLineNumbers),
            glyphMargin: glyphMargin,
            revealHorizontalRightPadding: toInteger(opts.revealHorizontalRightPadding, 0),
            roundedSelection: toBoolean(opts.roundedSelection),
            overviewRulerLanes: toInteger(opts.overviewRulerLanes, 0, 3),
            overviewRulerBorder: toBoolean(opts.overviewRulerBorder),
            cursorBlinking: cursorBlinkingStyleFromString(opts.cursorBlinking),
            mouseWheelZoom: toBoolean(opts.mouseWheelZoom),
            cursorStyle: cursorStyleFromString(opts.cursorStyle),
            hideCursorInOverviewRuler: toBoolean(opts.hideCursorInOverviewRuler),
            scrollBeyondLastLine: toBoolean(opts.scrollBeyondLastLine),
            editorClassName: editorClassName,
            stopRenderingLineAfter: stopRenderingLineAfter,
            renderWhitespace: renderWhitespace,
            renderControlCharacters: toBoolean(opts.renderControlCharacters),
            renderIndentGuides: toBoolean(opts.renderIndentGuides),
            renderLineHighlight: renderLineHighlight,
            scrollbar: scrollbar,
            minimap: minimap,
            fixedOverflowWidgets: toBoolean(opts.fixedOverflowWidgets)
        });
        var contribInfo = new editorCommon.EditorContribOptions({
            selectionClipboard: toBoolean(opts.selectionClipboard),
            hover: toBoolean(opts.hover),
            contextmenu: toBoolean(opts.contextmenu),
            quickSuggestions: toBoolean(opts.quickSuggestions),
            quickSuggestionsDelay: toInteger(opts.quickSuggestionsDelay),
            parameterHints: toBoolean(opts.parameterHints),
            iconsInSuggestions: toBoolean(opts.iconsInSuggestions),
            formatOnType: toBoolean(opts.formatOnType),
            formatOnPaste: toBoolean(opts.formatOnPaste),
            suggestOnTriggerCharacters: toBoolean(opts.suggestOnTriggerCharacters),
            acceptSuggestionOnEnter: toBoolean(opts.acceptSuggestionOnEnter),
            acceptSuggestionOnCommitCharacter: toBoolean(opts.acceptSuggestionOnCommitCharacter),
            snippetSuggestions: opts.snippetSuggestions,
            emptySelectionClipboard: opts.emptySelectionClipboard,
            wordBasedSuggestions: opts.wordBasedSuggestions,
            suggestFontSize: opts.suggestFontSize,
            suggestLineHeight: opts.suggestLineHeight,
            selectionHighlight: toBoolean(opts.selectionHighlight),
            occurrencesHighlight: toBoolean(opts.occurrencesHighlight),
            codeLens: opts.referenceInfos && opts.codeLens,
            folding: toBoolean(opts.folding),
            matchBrackets: toBoolean(opts.matchBrackets),
        });
        return new editorCommon.InternalEditorOptions({
            lineHeight: fontInfo.lineHeight,
            readOnly: readOnly,
            wordSeparators: String(opts.wordSeparators),
            autoClosingBrackets: toBoolean(opts.autoClosingBrackets),
            useTabStops: toBoolean(opts.useTabStops),
            tabFocusMode: tabFocusMode,
            dragAndDrop: toBoolean(opts.dragAndDrop),
            layoutInfo: layoutInfo,
            fontInfo: fontInfo,
            viewInfo: viewInfo,
            wrappingInfo: wrappingInfo,
            contribInfo: contribInfo,
        });
    };
    InternalEditorOptionsHelper._sanitizeScrollbarOpts = function (raw, mouseWheelScrollSensitivity) {
        var visibilityFromString = function (visibility) {
            switch (visibility) {
                case 'hidden':
                    return scrollable_1.ScrollbarVisibility.Hidden;
                case 'visible':
                    return scrollable_1.ScrollbarVisibility.Visible;
                default:
                    return scrollable_1.ScrollbarVisibility.Auto;
            }
        };
        var horizontalScrollbarSize = toIntegerWithDefault(raw.horizontalScrollbarSize, 10);
        var verticalScrollbarSize = toIntegerWithDefault(raw.verticalScrollbarSize, 14);
        return new editorCommon.InternalEditorScrollbarOptions({
            vertical: visibilityFromString(raw.vertical),
            horizontal: visibilityFromString(raw.horizontal),
            arrowSize: toIntegerWithDefault(raw.arrowSize, 11),
            useShadows: toBooleanWithDefault(raw.useShadows, true),
            verticalHasArrows: toBooleanWithDefault(raw.verticalHasArrows, false),
            horizontalHasArrows: toBooleanWithDefault(raw.horizontalHasArrows, false),
            horizontalScrollbarSize: horizontalScrollbarSize,
            horizontalSliderSize: toIntegerWithDefault(raw.horizontalSliderSize, horizontalScrollbarSize),
            verticalScrollbarSize: verticalScrollbarSize,
            verticalSliderSize: toIntegerWithDefault(raw.verticalSliderSize, verticalScrollbarSize),
            handleMouseWheel: toBooleanWithDefault(raw.handleMouseWheel, true),
            mouseWheelScrollSensitivity: mouseWheelScrollSensitivity
        });
    };
    InternalEditorOptionsHelper._sanitizeMinimapOpts = function (raw) {
        var maxColumn = toIntegerWithDefault(raw.maxColumn, defaultConfig_1.DefaultConfig.editor.minimap.maxColumn);
        if (maxColumn < 1) {
            maxColumn = 1;
        }
        return new editorCommon.InternalEditorMinimapOptions({
            enabled: toBooleanWithDefault(raw.enabled, defaultConfig_1.DefaultConfig.editor.minimap.enabled),
            renderCharacters: toBooleanWithDefault(raw.renderCharacters, defaultConfig_1.DefaultConfig.editor.minimap.renderCharacters),
            maxColumn: maxColumn,
        });
    };
    return InternalEditorOptionsHelper;
}());
function toBoolean(value) {
    return value === 'false' ? false : Boolean(value);
}
function toBooleanWithDefault(value, defaultValue) {
    if (typeof value === 'undefined') {
        return defaultValue;
    }
    return toBoolean(value);
}
function toFloat(source, defaultValue) {
    var r = parseFloat(source);
    if (isNaN(r)) {
        r = defaultValue;
    }
    return r;
}
function toInteger(source, minimum, maximum) {
    if (minimum === void 0) { minimum = -1073741824 /* MIN_SAFE_SMALL_INTEGER */; }
    if (maximum === void 0) { maximum = 1073741824 /* MAX_SAFE_SMALL_INTEGER */; }
    var r = parseInt(source, 10);
    if (isNaN(r)) {
        r = 0;
    }
    r = Math.max(minimum, r);
    r = Math.min(maximum, r);
    return r | 0;
}
function toSortedIntegerArray(source) {
    if (!Array.isArray(source)) {
        return [];
    }
    var arrSource = source;
    var r = arrSource.map(function (el) { return toInteger(el); });
    r.sort();
    return r;
}
function wrappingIndentFromString(wrappingIndent) {
    if (wrappingIndent === 'indent') {
        return editorCommon.WrappingIndent.Indent;
    }
    else if (wrappingIndent === 'same') {
        return editorCommon.WrappingIndent.Same;
    }
    else {
        return editorCommon.WrappingIndent.None;
    }
}
function cursorStyleFromString(cursorStyle) {
    if (cursorStyle === 'line') {
        return editorCommon.TextEditorCursorStyle.Line;
    }
    else if (cursorStyle === 'block') {
        return editorCommon.TextEditorCursorStyle.Block;
    }
    else if (cursorStyle === 'underline') {
        return editorCommon.TextEditorCursorStyle.Underline;
    }
    else if (cursorStyle === 'line-thin') {
        return editorCommon.TextEditorCursorStyle.LineThin;
    }
    else if (cursorStyle === 'block-outline') {
        return editorCommon.TextEditorCursorStyle.BlockOutline;
    }
    else if (cursorStyle === 'underline-thin') {
        return editorCommon.TextEditorCursorStyle.UnderlineThin;
    }
    return editorCommon.TextEditorCursorStyle.Line;
}
function cursorBlinkingStyleFromString(cursorBlinkingStyle) {
    switch (cursorBlinkingStyle) {
        case 'blink':
            return editorCommon.TextEditorCursorBlinkingStyle.Blink;
        case 'smooth':
            return editorCommon.TextEditorCursorBlinkingStyle.Smooth;
        case 'phase':
            return editorCommon.TextEditorCursorBlinkingStyle.Phase;
        case 'expand':
            return editorCommon.TextEditorCursorBlinkingStyle.Expand;
        case 'visible': // maintain compatibility
        case 'solid':
            return editorCommon.TextEditorCursorBlinkingStyle.Solid;
    }
    return editorCommon.TextEditorCursorBlinkingStyle.Blink;
}
function toIntegerWithDefault(source, defaultValue) {
    if (typeof source === 'undefined') {
        return defaultValue;
    }
    return toInteger(source);
}
var CommonEditorConfiguration = (function (_super) {
    __extends(CommonEditorConfiguration, _super);
    function CommonEditorConfiguration(options, elementSizeObserver) {
        if (elementSizeObserver === void 0) { elementSizeObserver = null; }
        var _this = _super.call(this) || this;
        _this._onDidChange = _this._register(new event_1.Emitter());
        _this.onDidChange = _this._onDidChange.event;
        _this._configWithDefaults = new ConfigurationWithDefaults(options);
        _this._elementSizeObserver = elementSizeObserver;
        _this._isDominatedByLongLines = false;
        _this._lineNumbersDigitCount = 1;
        _this.editor = _this._computeInternalOptions();
        _this.editorClone = _this.editor.clone();
        _this._register(editorZoom_1.EditorZoom.onDidChangeZoomLevel(function (_) { return _this._recomputeOptions(); }));
        _this._register(exports.TabFocus.onDidChangeTabFocus(function (_) { return _this._recomputeOptions(); }));
        return _this;
    }
    CommonEditorConfiguration.prototype.dispose = function () {
        _super.prototype.dispose.call(this);
    };
    CommonEditorConfiguration.prototype._recomputeOptions = function () {
        this._setOptions(this._computeInternalOptions());
    };
    CommonEditorConfiguration.prototype._setOptions = function (newOptions) {
        if (this.editor && this.editor.equals(newOptions)) {
            return;
        }
        var changeEvent = this.editor.createChangeEvent(newOptions);
        this.editor = newOptions;
        this.editorClone = this.editor.clone();
        this._onDidChange.fire(changeEvent);
    };
    CommonEditorConfiguration.prototype.getRawOptions = function () {
        return this._configWithDefaults.getEditorOptions();
    };
    CommonEditorConfiguration.prototype._computeInternalOptions = function () {
        var opts = this._configWithDefaults.getEditorOptions();
        var editorClassName = this._getEditorClassName(opts.theme, toBoolean(opts.fontLigatures), opts.mouseStyle);
        var disableTranslate3d = toBoolean(opts.disableTranslate3d);
        var canUseTranslate3d = this._getCanUseTranslate3d();
        if (disableTranslate3d) {
            canUseTranslate3d = false;
        }
        var bareFontInfo = fontInfo_1.BareFontInfo.createFromRawSettings(opts, this.getZoomLevel());
        return InternalEditorOptionsHelper.createInternalEditorOptions(this.getOuterWidth(), this.getOuterHeight(), opts, this.readConfiguration(bareFontInfo), editorClassName, this._isDominatedByLongLines, this._lineNumbersDigitCount, canUseTranslate3d, this._getPixelRatio());
    };
    CommonEditorConfiguration.prototype.updateOptions = function (newOptions) {
        this._configWithDefaults.updateOptions(newOptions);
        this._recomputeOptions();
    };
    CommonEditorConfiguration.prototype.setIsDominatedByLongLines = function (isDominatedByLongLines) {
        this._isDominatedByLongLines = isDominatedByLongLines;
        this._recomputeOptions();
    };
    CommonEditorConfiguration.prototype.setMaxLineNumber = function (maxLineNumber) {
        var digitCount = CommonEditorConfiguration.digitCount(maxLineNumber);
        if (this._lineNumbersDigitCount === digitCount) {
            return;
        }
        this._lineNumbersDigitCount = digitCount;
        this._recomputeOptions();
    };
    CommonEditorConfiguration.digitCount = function (n) {
        var r = 0;
        while (n) {
            n = Math.floor(n / 10);
            r++;
        }
        return r ? r : 1;
    };
    return CommonEditorConfiguration;
}(lifecycle_1.Disposable));
exports.CommonEditorConfiguration = CommonEditorConfiguration;
var configurationRegistry = platform_1.Registry.as(configurationRegistry_1.Extensions.Configuration);
var editorConfiguration = {
    'id': 'editor',
    'order': 5,
    'type': 'object',
    'title': nls.localize('editorConfigurationTitle', "Editor"),
    'overridable': true,
    'properties': {
        'editor.fontFamily': {
            'type': 'string',
            'default': defaultConfig_1.DefaultConfig.editor.fontFamily,
            'description': nls.localize('fontFamily', "Controls the font family.")
        },
        'editor.fontWeight': {
            'type': 'string',
            'enum': ['normal', 'bold', '100', '200', '300', '400', '500', '600', '700', '800', '900'],
            'default': defaultConfig_1.DefaultConfig.editor.fontWeight,
            'description': nls.localize('fontWeight', "Controls the font weight.")
        },
        'editor.fontSize': {
            'type': 'number',
            'default': defaultConfig_1.DefaultConfig.editor.fontSize,
            'description': nls.localize('fontSize', "Controls the font size in pixels.")
        },
        'editor.lineHeight': {
            'type': 'number',
            'default': defaultConfig_1.DefaultConfig.editor.lineHeight,
            'description': nls.localize('lineHeight', "Controls the line height. Use 0 to compute the lineHeight from the fontSize.")
        },
        'editor.lineNumbers': {
            'type': 'string',
            'enum': ['off', 'on', 'relative'],
            'default': defaultConfig_1.DefaultConfig.editor.lineNumbers,
            'description': nls.localize('lineNumbers', "Controls the display of line numbers. Possible values are 'on', 'off', and 'relative'. 'relative' shows the line count from the current cursor position.")
        },
        'editor.rulers': {
            'type': 'array',
            'items': {
                'type': 'number'
            },
            'default': defaultConfig_1.DefaultConfig.editor.rulers,
            'description': nls.localize('rulers', "Columns at which to show vertical rulers")
        },
        'editor.wordSeparators': {
            'type': 'string',
            'default': defaultConfig_1.DefaultConfig.editor.wordSeparators,
            'description': nls.localize('wordSeparators', "Characters that will be used as word separators when doing word related navigations or operations")
        },
        'editor.tabSize': {
            'type': 'number',
            'default': defaultConfig_1.DEFAULT_INDENTATION.tabSize,
            'minimum': 1,
            'description': nls.localize('tabSize', "The number of spaces a tab is equal to. This setting is overriden based on the file contents when `editor.detectIndentation` is on."),
            'errorMessage': nls.localize('tabSize.errorMessage', "Expected 'number'. Note that the value \"auto\" has been replaced by the `editor.detectIndentation` setting.")
        },
        'editor.insertSpaces': {
            'type': 'boolean',
            'default': defaultConfig_1.DEFAULT_INDENTATION.insertSpaces,
            'description': nls.localize('insertSpaces', "Insert spaces when pressing Tab. This setting is overriden based on the file contents when `editor.detectIndentation` is on."),
            'errorMessage': nls.localize('insertSpaces.errorMessage', "Expected 'boolean'. Note that the value \"auto\" has been replaced by the `editor.detectIndentation` setting.")
        },
        'editor.detectIndentation': {
            'type': 'boolean',
            'default': defaultConfig_1.DEFAULT_INDENTATION.detectIndentation,
            'description': nls.localize('detectIndentation', "When opening a file, `editor.tabSize` and `editor.insertSpaces` will be detected based on the file contents.")
        },
        'editor.roundedSelection': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.roundedSelection,
            'description': nls.localize('roundedSelection', "Controls if selections have rounded corners")
        },
        'editor.scrollBeyondLastLine': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.scrollBeyondLastLine,
            'description': nls.localize('scrollBeyondLastLine', "Controls if the editor will scroll beyond the last line")
        },
        'editor.minimap.enabled': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.minimap.enabled,
            'description': nls.localize('minimap.enabled', "Controls if the minimap is shown")
        },
        'editor.minimap.renderCharacters': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.minimap.renderCharacters,
            'description': nls.localize('minimap.renderCharacters', "Render the actual characters on a line (as opposed to color blocks)")
        },
        'editor.minimap.maxColumn': {
            'type': 'number',
            'default': defaultConfig_1.DefaultConfig.editor.minimap.maxColumn,
            'description': nls.localize('minimap.maxColumn', "Limit the width of the minimap to render at most a certain number of columns")
        },
        'editor.wordWrap': {
            'type': 'string',
            'enum': ['off', 'on', 'wordWrapColumn', 'bounded'],
            'enumDescriptions': [
                nls.localize('wordWrap.off', "Lines will never wrap."),
                nls.localize('wordWrap.on', "Lines will wrap at the viewport width."),
                nls.localize('wordWrap.wordWrapColumn', "Lines will wrap at `editor.wordWrapColumn`."),
                nls.localize('wordWrap.bounded', "Lines will wrap at the minimum of viewport and `editor.wordWrapColumn`."),
            ],
            'default': defaultConfig_1.DefaultConfig.editor.wordWrap,
            'description': nls.localize('wordWrap', "Controls how lines should wrap. Can be:\n - 'off' (disable wrapping),\n - 'on' (viewport wrapping),\n - 'wordWrapColumn' (wrap at `editor.wordWrapColumn`) or\n - 'bounded' (wrap at minimum of viewport and `editor.wordWrapColumn`).")
        },
        'editor.wordWrapColumn': {
            'type': 'integer',
            'default': defaultConfig_1.DefaultConfig.editor.wordWrapColumn,
            'minimum': 1,
            'description': nls.localize('wordWrapColumn', "Controls the wrapping column of the editor when `editor.wordWrap` is 'wordWrapColumn' or 'bounded'.")
        },
        'editor.wrappingIndent': {
            'type': 'string',
            'enum': ['none', 'same', 'indent'],
            'default': defaultConfig_1.DefaultConfig.editor.wrappingIndent,
            'description': nls.localize('wrappingIndent', "Controls the indentation of wrapped lines. Can be one of 'none', 'same' or 'indent'.")
        },
        'editor.mouseWheelScrollSensitivity': {
            'type': 'number',
            'default': defaultConfig_1.DefaultConfig.editor.mouseWheelScrollSensitivity,
            'description': nls.localize('mouseWheelScrollSensitivity', "A multiplier to be used on the `deltaX` and `deltaY` of mouse wheel scroll events")
        },
        'editor.quickSuggestions': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.quickSuggestions,
            'description': nls.localize('quickSuggestions', "Controls if quick suggestions should show up or not while typing")
        },
        'editor.quickSuggestionsDelay': {
            'type': 'integer',
            'default': defaultConfig_1.DefaultConfig.editor.quickSuggestionsDelay,
            'minimum': 0,
            'description': nls.localize('quickSuggestionsDelay', "Controls the delay in ms after which quick suggestions will show up")
        },
        'editor.parameterHints': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.parameterHints,
            'description': nls.localize('parameterHints', "Enables parameter hints")
        },
        'editor.autoClosingBrackets': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.autoClosingBrackets,
            'description': nls.localize('autoClosingBrackets', "Controls if the editor should automatically close brackets after opening them")
        },
        'editor.formatOnType': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.formatOnType,
            'description': nls.localize('formatOnType', "Controls if the editor should automatically format the line after typing")
        },
        'editor.formatOnPaste': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.formatOnPaste,
            'description': nls.localize('formatOnPaste', "Controls if the editor should automatically format the pasted content. A formatter must be available and the formatter should be able to format a range in a document.")
        },
        'editor.suggestOnTriggerCharacters': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.suggestOnTriggerCharacters,
            'description': nls.localize('suggestOnTriggerCharacters', "Controls if suggestions should automatically show up when typing trigger characters")
        },
        'editor.acceptSuggestionOnEnter': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.acceptSuggestionOnEnter,
            'description': nls.localize('acceptSuggestionOnEnter', "Controls if suggestions should be accepted on 'Enter' - in addition to 'Tab'. Helps to avoid ambiguity between inserting new lines or accepting suggestions.")
        },
        'editor.acceptSuggestionOnCommitCharacter': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.acceptSuggestionOnCommitCharacter,
            'description': nls.localize('acceptSuggestionOnCommitCharacter', "Controls if suggestions should be accepted on commit characters. For instance in JavaScript the semi-colon (';') can be a commit character that accepts a suggestion and types that character.")
        },
        'editor.snippetSuggestions': {
            'type': 'string',
            'enum': ['top', 'bottom', 'inline', 'none'],
            'default': defaultConfig_1.DefaultConfig.editor.snippetSuggestions,
            'description': nls.localize('snippetSuggestions', "Controls whether snippets are shown with other suggestions and how they are sorted.")
        },
        'editor.emptySelectionClipboard': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.emptySelectionClipboard,
            'description': nls.localize('emptySelectionClipboard', "Controls whether copying without a selection copies the current line.")
        },
        'editor.wordBasedSuggestions': {
            'anyOf': [
                'boolean',
                {
                    type: 'object',
                    properties: {
                        strings: {
                            type: 'boolean',
                            description: nls.localize('wordBasedSuggestions.strings', "Enable word based suggestions inside strings.")
                        },
                        comments: {
                            type: 'boolean',
                            description: nls.localize('wordBasedSuggestions.comments', "Enable word based suggestions inside comments.")
                        },
                        default: {
                            type: 'boolean',
                            description: nls.localize('wordBasedSuggestions.default', "Enable word based suggestions outside of strings and comments.")
                        },
                    }
                }
            ],
            'default': defaultConfig_1.DefaultConfig.editor.wordBasedSuggestions,
            'description': nls.localize('wordBasedSuggestions', "Enable word based suggestions.")
        },
        'editor.suggestFontSize': {
            'type': 'integer',
            'default': 0,
            'minimum': 0,
            'description': nls.localize('suggestFontSize', "Font size for the suggest widget")
        },
        'editor.suggestLineHeight': {
            'type': 'integer',
            'default': 0,
            'minimum': 0,
            'description': nls.localize('suggestLineHeight', "Line height for the suggest widget")
        },
        'editor.selectionHighlight': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.selectionHighlight,
            'description': nls.localize('selectionHighlight', "Controls whether the editor should highlight similar matches to the selection")
        },
        'editor.occurrencesHighlight': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.occurrencesHighlight,
            'description': nls.localize('occurrencesHighlight', "Controls whether the editor should highlight semantic symbol occurrences")
        },
        'editor.overviewRulerLanes': {
            'type': 'integer',
            'default': 3,
            'description': nls.localize('overviewRulerLanes', "Controls the number of decorations that can show up at the same position in the overview ruler")
        },
        'editor.overviewRulerBorder': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.overviewRulerBorder,
            'description': nls.localize('overviewRulerBorder', "Controls if a border should be drawn around the overview ruler.")
        },
        'editor.cursorBlinking': {
            'type': 'string',
            'enum': ['blink', 'smooth', 'phase', 'expand', 'solid'],
            'default': defaultConfig_1.DefaultConfig.editor.cursorBlinking,
            'description': nls.localize('cursorBlinking', "Control the cursor animation style, possible values are 'blink', 'smooth', 'phase', 'expand' and 'solid'")
        },
        'editor.mouseWheelZoom': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.mouseWheelZoom,
            'description': nls.localize('mouseWheelZoom', "Zoom the font of the editor when using mouse wheel and holding Ctrl")
        },
        'editor.cursorStyle': {
            'type': 'string',
            'enum': ['block', 'block-outline', 'line', 'line-thin', 'underline', 'underline-thin'],
            'default': defaultConfig_1.DefaultConfig.editor.cursorStyle,
            'description': nls.localize('cursorStyle', "Controls the cursor style, accepted values are 'block', 'block-outline', 'line', 'line-thin', 'underline' and 'underline-thin'")
        },
        'editor.fontLigatures': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.fontLigatures,
            'description': nls.localize('fontLigatures', "Enables font ligatures")
        },
        'editor.hideCursorInOverviewRuler': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.hideCursorInOverviewRuler,
            'description': nls.localize('hideCursorInOverviewRuler', "Controls if the cursor should be hidden in the overview ruler.")
        },
        'editor.renderWhitespace': {
            'type': 'string',
            'enum': ['none', 'boundary', 'all'],
            default: defaultConfig_1.DefaultConfig.editor.renderWhitespace,
            description: nls.localize('renderWhitespace', "Controls how the editor should render whitespace characters, possibilities are 'none', 'boundary', and 'all'. The 'boundary' option does not render single spaces between words.")
        },
        'editor.renderControlCharacters': {
            'type': 'boolean',
            default: defaultConfig_1.DefaultConfig.editor.renderControlCharacters,
            description: nls.localize('renderControlCharacters', "Controls whether the editor should render control characters")
        },
        'editor.renderIndentGuides': {
            'type': 'boolean',
            default: defaultConfig_1.DefaultConfig.editor.renderIndentGuides,
            description: nls.localize('renderIndentGuides', "Controls whether the editor should render indent guides")
        },
        'editor.renderLineHighlight': {
            'type': 'string',
            'enum': ['none', 'gutter', 'line', 'all'],
            default: defaultConfig_1.DefaultConfig.editor.renderLineHighlight,
            description: nls.localize('renderLineHighlight', "Controls how the editor should render the current line highlight, possibilities are 'none', 'gutter', 'line', and 'all'.")
        },
        'editor.codeLens': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.codeLens,
            'description': nls.localize('codeLens', "Controls if the editor shows code lenses")
        },
        'editor.folding': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.folding,
            'description': nls.localize('folding', "Controls whether the editor has code folding enabled")
        },
        'editor.matchBrackets': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.matchBrackets,
            'description': nls.localize('matchBrackets', "Highlight matching brackets when one of them is selected.")
        },
        'editor.glyphMargin': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.glyphMargin,
            'description': nls.localize('glyphMargin', "Controls whether the editor should render the vertical glyph margin. Glyph margin is mostly used for debugging.")
        },
        'editor.useTabStops': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.useTabStops,
            'description': nls.localize('useTabStops', "Inserting and deleting whitespace follows tab stops")
        },
        'editor.trimAutoWhitespace': {
            'type': 'boolean',
            'default': defaultConfig_1.DEFAULT_TRIM_AUTO_WHITESPACE,
            'description': nls.localize('trimAutoWhitespace', "Remove trailing auto inserted whitespace")
        },
        'editor.stablePeek': {
            'type': 'boolean',
            'default': false,
            'description': nls.localize('stablePeek', "Keep peek editors open even when double clicking their content or when hitting Escape.")
        },
        'editor.dragAndDrop': {
            'type': 'boolean',
            'default': defaultConfig_1.DefaultConfig.editor.dragAndDrop,
            'description': nls.localize('dragAndDrop', "Controls if the editor should allow to move selections via drag and drop.")
        },
        'diffEditor.renderSideBySide': {
            'type': 'boolean',
            'default': true,
            'description': nls.localize('sideBySide', "Controls if the diff editor shows the diff side by side or inline")
        },
        'diffEditor.ignoreTrimWhitespace': {
            'type': 'boolean',
            'default': true,
            'description': nls.localize('ignoreTrimWhitespace', "Controls if the diff editor shows changes in leading or trailing whitespace as diffs")
        },
        'diffEditor.renderIndicators': {
            'type': 'boolean',
            'default': true,
            'description': nls.localize('renderIndicators', "Controls if the diff editor shows +/- indicators for added/removed changes")
        }
    }
};
if (platform.isLinux) {
    editorConfiguration['properties']['editor.selectionClipboard'] = {
        'type': 'boolean',
        'default': defaultConfig_1.DefaultConfig.editor.selectionClipboard,
        'description': nls.localize('selectionClipboard', "Controls if the Linux primary clipboard should be supported.")
    };
}
configurationRegistry.registerConfiguration(editorConfiguration);
