// vscode/src/vs/editor/browser/viewParts/marginDecorations/marginDecorations.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
require("./marginDecorations.css");
var glyphMargin_1 = require("vs/editor/browser/viewParts/glyphMargin/glyphMargin");
var MarginViewLineDecorationsOverlay = (function (_super) {
    __extends(MarginViewLineDecorationsOverlay, _super);
    function MarginViewLineDecorationsOverlay(context) {
        var _this = _super.call(this) || this;
        _this._context = context;
        _this._renderResult = null;
        _this._context.addEventHandler(_this);
        return _this;
    }
    MarginViewLineDecorationsOverlay.prototype.dispose = function () {
        this._context.removeEventHandler(this);
        this._context = null;
        this._renderResult = null;
    };
    // --- begin event handlers
    MarginViewLineDecorationsOverlay.prototype.onModelFlushed = function () {
        return true;
    };
    MarginViewLineDecorationsOverlay.prototype.onModelDecorationsChanged = function (e) {
        return true;
    };
    MarginViewLineDecorationsOverlay.prototype.onModelLinesDeleted = function (e) {
        return true;
    };
    MarginViewLineDecorationsOverlay.prototype.onModelLineChanged = function (e) {
        return true;
    };
    MarginViewLineDecorationsOverlay.prototype.onModelLinesInserted = function (e) {
        return true;
    };
    MarginViewLineDecorationsOverlay.prototype.onCursorPositionChanged = function (e) {
        return false;
    };
    MarginViewLineDecorationsOverlay.prototype.onCursorSelectionChanged = function (e) {
        return false;
    };
    MarginViewLineDecorationsOverlay.prototype.onCursorRevealRange = function (e) {
        return false;
    };
    MarginViewLineDecorationsOverlay.prototype.onConfigurationChanged = function (e) {
        return true;
    };
    MarginViewLineDecorationsOverlay.prototype.onLayoutChanged = function (layoutInfo) {
        return true;
    };
    MarginViewLineDecorationsOverlay.prototype.onScrollChanged = function (e) {
        return e.scrollTopChanged;
    };
    MarginViewLineDecorationsOverlay.prototype.onZonesChanged = function () {
        return true;
    };
    // --- end event handlers
    MarginViewLineDecorationsOverlay.prototype._getDecorations = function (ctx) {
        var decorations = ctx.getDecorationsInViewport();
        var r = [];
        for (var i = 0, len = decorations.length; i < len; i++) {
            var d = decorations[i];
            var marginClassName = d.source.options.marginClassName;
            if (marginClassName) {
                r.push(new glyphMargin_1.DecorationToRender(d.range.startLineNumber, d.range.endLineNumber, marginClassName));
            }
        }
        return r;
    };
    MarginViewLineDecorationsOverlay.prototype.prepareRender = function (ctx) {
        var visibleStartLineNumber = ctx.visibleRange.startLineNumber;
        var visibleEndLineNumber = ctx.visibleRange.endLineNumber;
        var toRender = this._render(visibleStartLineNumber, visibleEndLineNumber, this._getDecorations(ctx));
        var output = [];
        for (var lineNumber = visibleStartLineNumber; lineNumber <= visibleEndLineNumber; lineNumber++) {
            var lineIndex = lineNumber - visibleStartLineNumber;
            var classNames_1 = toRender[lineIndex];
            var lineOutput = '';
            for (var i = 0, len = classNames_1.length; i < len; i++) {
                lineOutput += '<div class="cmdr ' + classNames_1[i] + '" style=""></div>';
            }
            output[lineIndex] = lineOutput;
        }
        this._renderResult = output;
    };
    MarginViewLineDecorationsOverlay.prototype.render = function (startLineNumber, lineNumber) {
        if (!this._renderResult) {
            return '';
        }
        return this._renderResult[lineNumber - startLineNumber];
    };
    return MarginViewLineDecorationsOverlay;
}(glyphMargin_1.DedupOverlay));
exports.MarginViewLineDecorationsOverlay = MarginViewLineDecorationsOverlay;
