// vscode/src/vs/base/browser/ui/toolbar/toolbar.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
require("./toolbar.css");
var nls = require("vs/nls");
var winjs_base_1 = require("vs/base/common/winjs.base");
var builder_1 = require("vs/base/browser/builder");
var types = require("vs/base/common/types");
var actions_1 = require("vs/base/common/actions");
var actionbar_1 = require("vs/base/browser/ui/actionbar/actionbar");
var dropdown_1 = require("vs/base/browser/ui/dropdown/dropdown");
exports.CONTEXT = 'context.toolbar';
/**
 * A widget that combines an action bar for primary actions and a dropdown for secondary actions.
 */
var ToolBar = (function () {
    function ToolBar(container, contextMenuProvider, options) {
        if (options === void 0) { options = { orientation: actionbar_1.ActionsOrientation.HORIZONTAL }; }
        var _this = this;
        this.options = options;
        this.lookupKeybindings = typeof this.options.getKeyBinding === 'function' && typeof this.options.getKeyBindingLabel === 'function';
        this.toggleMenuAction = new ToggleMenuAction(function () { return _this.toggleMenuActionItem && _this.toggleMenuActionItem.show(); });
        var element = document.createElement('div');
        element.className = 'monaco-toolbar';
        container.appendChild(element);
        this.actionBar = new actionbar_1.ActionBar(builder_1.$(element), {
            orientation: options.orientation,
            ariaLabel: options.ariaLabel,
            actionItemProvider: function (action) {
                // Return special action item for the toggle menu action
                if (action.id === ToggleMenuAction.ID) {
                    // Dispose old
                    if (_this.toggleMenuActionItem) {
                        _this.toggleMenuActionItem.dispose();
                    }
                    // Create new
                    _this.toggleMenuActionItem = new DropdownMenuActionItem(action, action.menuActions, contextMenuProvider, _this.options.actionItemProvider, _this.actionRunner, _this.options.getKeyBinding, 'toolbar-toggle-more');
                    _this.toggleMenuActionItem.setActionContext(_this.actionBar.context);
                    return _this.toggleMenuActionItem;
                }
                return options.actionItemProvider ? options.actionItemProvider(action) : null;
            }
        });
    }
    Object.defineProperty(ToolBar.prototype, "actionRunner", {
        get: function () {
            return this.actionBar.actionRunner;
        },
        set: function (actionRunner) {
            this.actionBar.actionRunner = actionRunner;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ToolBar.prototype, "context", {
        set: function (context) {
            this.actionBar.context = context;
            if (this.toggleMenuActionItem) {
                this.toggleMenuActionItem.setActionContext(context);
            }
        },
        enumerable: true,
        configurable: true
    });
    ToolBar.prototype.getContainer = function () {
        return this.actionBar.getContainer();
    };
    ToolBar.prototype.setAriaLabel = function (label) {
        this.actionBar.setAriaLabel(label);
    };
    ToolBar.prototype.setActions = function (primaryActions, secondaryActions) {
        var _this = this;
        return function () {
            var primaryActionsToSet = primaryActions ? primaryActions.slice(0) : [];
            // Inject additional action to open secondary actions if present
            _this.hasSecondaryActions = secondaryActions && secondaryActions.length > 0;
            if (_this.hasSecondaryActions) {
                _this.toggleMenuAction.menuActions = secondaryActions.slice(0);
                primaryActionsToSet.push(_this.toggleMenuAction);
            }
            _this.actionBar.clear();
            primaryActionsToSet.forEach(function (action) {
                _this.actionBar.push(action, { icon: true, label: false, keybinding: _this.getKeybindingLabel(action) });
            });
        };
    };
    ToolBar.prototype.getKeybindingLabel = function (action) {
        var key = this.lookupKeybindings ? this.options.getKeyBinding(action) : void 0;
        return key ? this.options.getKeyBindingLabel(key) : void 0;
    };
    ToolBar.prototype.addPrimaryAction = function (primaryAction) {
        var _this = this;
        return function () {
            // Add after the "..." action if we have secondary actions
            if (_this.hasSecondaryActions) {
                var itemCount = _this.actionBar.length();
                _this.actionBar.push(primaryAction, { icon: true, label: false, index: itemCount, keybinding: _this.getKeybindingLabel(primaryAction) });
            }
            else {
                _this.actionBar.push(primaryAction, { icon: true, label: false, keybinding: _this.getKeybindingLabel(primaryAction) });
            }
        };
    };
    ToolBar.prototype.dispose = function () {
        this.actionBar.dispose();
        this.toggleMenuAction.dispose();
        if (this.toggleMenuActionItem) {
            this.toggleMenuActionItem.dispose();
        }
    };
    return ToolBar;
}());
exports.ToolBar = ToolBar;
var ToggleMenuAction = (function (_super) {
    __extends(ToggleMenuAction, _super);
    function ToggleMenuAction(toggleDropdownMenu) {
        var _this = _super.call(this, ToggleMenuAction.ID, nls.localize('more', "More"), null, true) || this;
        _this.toggleDropdownMenu = toggleDropdownMenu;
        return _this;
    }
    ToggleMenuAction.prototype.run = function () {
        this.toggleDropdownMenu();
        return winjs_base_1.TPromise.as(true);
    };
    Object.defineProperty(ToggleMenuAction.prototype, "menuActions", {
        get: function () {
            return this._menuActions;
        },
        set: function (actions) {
            this._menuActions = actions;
        },
        enumerable: true,
        configurable: true
    });
    return ToggleMenuAction;
}(actions_1.Action));
ToggleMenuAction.ID = 'toolbar.toggle.more';
var DropdownMenuActionItem = (function (_super) {
    __extends(DropdownMenuActionItem, _super);
    function DropdownMenuActionItem(action, menuActionsOrProvider, contextMenuProvider, actionItemProvider, actionRunner, keybindings, clazz) {
        var _this = _super.call(this, null, action) || this;
        _this.menuActionsOrProvider = menuActionsOrProvider;
        _this.contextMenuProvider = contextMenuProvider;
        _this.actionItemProvider = actionItemProvider;
        _this.actionRunner = actionRunner;
        _this.keybindings = keybindings;
        _this.clazz = clazz;
        return _this;
    }
    DropdownMenuActionItem.prototype.render = function (container) {
        var _this = this;
        var labelRenderer = function (el) {
            _this.builder = builder_1.$('a.action-label').attr({
                tabIndex: '0',
                role: 'button',
                'aria-haspopup': 'true',
                title: _this._action.label || '',
                class: _this.clazz
            });
            _this.builder.appendTo(el);
            return null;
        };
        var options = {
            contextMenuProvider: this.contextMenuProvider,
            labelRenderer: labelRenderer
        };
        // Render the DropdownMenu around a simple action to toggle it
        if (types.isArray(this.menuActionsOrProvider)) {
            options.actions = this.menuActionsOrProvider;
        }
        else {
            options.actionProvider = this.menuActionsOrProvider;
        }
        this.dropdownMenu = new dropdown_1.DropdownMenu(container, options);
        this.dropdownMenu.menuOptions = {
            actionItemProvider: this.actionItemProvider,
            actionRunner: this.actionRunner,
            getKeyBinding: this.keybindings,
            context: this._context
        };
        // Reemit events for running actions
        this.toUnbind = this.addEmitter2(this.dropdownMenu);
    };
    DropdownMenuActionItem.prototype.setActionContext = function (newContext) {
        _super.prototype.setActionContext.call(this, newContext);
        if (this.dropdownMenu) {
            this.dropdownMenu.menuOptions.context = newContext;
        }
    };
    DropdownMenuActionItem.prototype.show = function () {
        if (this.dropdownMenu) {
            this.dropdownMenu.show();
        }
    };
    DropdownMenuActionItem.prototype.dispose = function () {
        this.toUnbind.dispose();
        this.dropdownMenu.dispose();
        _super.prototype.dispose.call(this);
    };
    return DropdownMenuActionItem;
}(actionbar_1.BaseActionItem));
exports.DropdownMenuActionItem = DropdownMenuActionItem;
