diff --git a/migrations/codeinsights/BUILD.bazel b/migrations/codeinsights/BUILD.bazel
index 840ded1e83..ba8b6af798 100644
--- a/migrations/codeinsights/BUILD.bazel
+++ b/migrations/codeinsights/BUILD.bazel
@@ -20,6 +20,6 @@ write_source_files(
     files = {
         "squashed.sql": ":squashed",
     },
-    suggested_update_target = "//dev:write_all_generated",
+    # suggested_update_target = "//dev:write_all_generated",
     visibility = ["//visibility:public"],
 )
diff --git a/migrations/codeintel/BUILD.bazel b/migrations/codeintel/BUILD.bazel
index f369388a73..132d6407d5 100644
--- a/migrations/codeintel/BUILD.bazel
+++ b/migrations/codeintel/BUILD.bazel
@@ -19,6 +19,6 @@ write_source_files(
     files = {
         "squashed.sql": ":squashed",
     },
-    suggested_update_target = "//dev:write_all_generated",
+    # suggested_update_target = "//dev:write_all_generated",
     visibility = ["//visibility:public"],
 )
diff --git a/migrations/frontend/BUILD.bazel b/migrations/frontend/BUILD.bazel
index a089860ea3..a2c994886a 100644
--- a/migrations/frontend/BUILD.bazel
+++ b/migrations/frontend/BUILD.bazel
@@ -19,6 +19,6 @@ write_source_files(
     files = {
         "squashed.sql": ":squashed",
     },
-    suggested_update_target = "//dev:write_all_generated",
+    # suggested_update_target = "//dev:write_all_generated",
     visibility = ["//visibility:public"],
 )
