// vscode/src/vs/base/common/collections.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
function createStringDictionary() {
    return Object.create(null);
}
exports.createStringDictionary = createStringDictionary;
function createNumberDictionary() {
    return Object.create(null);
}
exports.createNumberDictionary = createNumberDictionary;
function lookup(from, what, alternate) {
    if (alternate === void 0) { alternate = null; }
    var key = String(what);
    if (contains(from, key)) {
        return from[key];
    }
    return alternate;
}
exports.lookup = lookup;
function lookupOrInsert(from, stringOrNumber, alternate) {
    var key = String(stringOrNumber);
    if (contains(from, key)) {
        return from[key];
    }
    else {
        if (typeof alternate === 'function') {
            alternate = alternate();
        }
        from[key] = alternate;
        return alternate;
    }
}
exports.lookupOrInsert = lookupOrInsert;
function insert(into, data, hashFn) {
    into[hashFn(data)] = data;
}
exports.insert = insert;
var hasOwnProperty = Object.prototype.hasOwnProperty;
function contains(from, what) {
    return hasOwnProperty.call(from, what);
}
exports.contains = contains;
function values(from) {
    var result = [];
    for (var key in from) {
        if (hasOwnProperty.call(from, key)) {
            result.push(from[key]);
        }
    }
    return result;
}
exports.values = values;
function forEach(from, callback) {
    for (var key in from) {
        if (hasOwnProperty.call(from, key)) {
            var result = callback({ key: key, value: from[key] }, function () {
                delete from[key];
            });
            if (result === false) {
                return;
            }
        }
    }
}
exports.forEach = forEach;
function remove(from, key) {
    if (!hasOwnProperty.call(from, key)) {
        return false;
    }
    delete from[key];
    return true;
}
exports.remove = remove;
/**
 * Groups the collection into a dictionary based on the provided
 * group function.
 */
function groupBy(data, groupFn) {
    var result = createStringDictionary();
    data.forEach(function (element) { return lookupOrInsert(result, groupFn(element), []).push(element); });
    return result;
}
exports.groupBy = groupBy;
