// vscode/src/vs/base/common/events.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var Event = (function () {
    function Event(originalEvent) {
        this.time = (new Date()).getTime();
        this.originalEvent = originalEvent;
        this.source = null;
    }
    return Event;
}());
exports.Event = Event;
var PropertyChangeEvent = (function (_super) {
    __extends(PropertyChangeEvent, _super);
    function PropertyChangeEvent(key, oldValue, newValue, originalEvent) {
        var _this = _super.call(this, originalEvent) || this;
        _this.key = key;
        _this.oldValue = oldValue;
        _this.newValue = newValue;
        return _this;
    }
    return PropertyChangeEvent;
}(Event));
exports.PropertyChangeEvent = PropertyChangeEvent;
var ViewerEvent = (function (_super) {
    __extends(ViewerEvent, _super);
    function ViewerEvent(element, originalEvent) {
        var _this = _super.call(this, originalEvent) || this;
        _this.element = element;
        return _this;
    }
    return ViewerEvent;
}(Event));
exports.ViewerEvent = ViewerEvent;
exports.EventType = {
    PROPERTY_CHANGED: 'propertyChanged',
    SELECTION: 'selection',
    FOCUS: 'focus',
    BLUR: 'blur',
    HIGHLIGHT: 'highlight',
    EXPAND: 'expand',
    COLLAPSE: 'collapse',
    TOGGLE: 'toggle',
    BEFORE_RUN: 'beforeRun',
    RUN: 'run',
    EDIT: 'edit',
    SAVE: 'save',
    CANCEL: 'cancel',
    CHANGE: 'change',
    DISPOSE: 'dispose',
};
