// vscode/src/vs/base/common/mime.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var paths = require('vs/base/common/paths');
var types = require('vs/base/common/types');
var strings = require('vs/base/common/strings');
var glob_1 = require('vs/base/common/glob');
exports.MIME_TEXT = 'text/plain';
exports.MIME_BINARY = 'application/octet-stream';
exports.MIME_UNKNOWN = 'application/unknown';
var registeredAssociations = [];
var nonUserRegisteredAssociations = [];
var userRegisteredAssociations = [];
/**
 * Associate a text mime to the registry.
 */
function registerTextMime(association) {
    // Register
    var associationItem = toTextMimeAssociationItem(association);
    registeredAssociations.push(associationItem);
    if (!associationItem.userConfigured) {
        nonUserRegisteredAssociations.push(associationItem);
    }
    else {
        userRegisteredAssociations.push(associationItem);
    }
    // Check for conflicts unless this is a user configured association
    if (!associationItem.userConfigured) {
        registeredAssociations.forEach(function (a) {
            if (a.mime === associationItem.mime || a.userConfigured) {
                return; // same mime or userConfigured is ok
            }
            if (associationItem.extension && a.extension === associationItem.extension) {
                console.warn("Overwriting extension <<" + associationItem.extension + ">> to now point to mime <<" + associationItem.mime + ">>");
            }
            if (associationItem.filename && a.filename === associationItem.filename) {
                console.warn("Overwriting filename <<" + associationItem.filename + ">> to now point to mime <<" + associationItem.mime + ">>");
            }
            if (associationItem.filepattern && a.filepattern === associationItem.filepattern) {
                console.warn("Overwriting filepattern <<" + associationItem.filepattern + ">> to now point to mime <<" + associationItem.mime + ">>");
            }
            if (associationItem.firstline && a.firstline === associationItem.firstline) {
                console.warn("Overwriting firstline <<" + associationItem.firstline + ">> to now point to mime <<" + associationItem.mime + ">>");
            }
        });
    }
}
exports.registerTextMime = registerTextMime;
function toTextMimeAssociationItem(association) {
    return {
        id: association.id,
        mime: association.mime,
        filename: association.filename,
        extension: association.extension,
        filepattern: association.filepattern,
        firstline: association.firstline,
        userConfigured: association.userConfigured,
        filenameLowercase: association.filename ? association.filename.toLowerCase() : void 0,
        extensionLowercase: association.extension ? association.extension.toLowerCase() : void 0,
        filepatternLowercase: association.filepattern ? association.filepattern.toLowerCase() : void 0,
        filepatternOnPath: association.filepattern ? association.filepattern.indexOf(paths.sep) >= 0 : false
    };
}
/**
 * Clear text mimes from the registry.
 */
function clearTextMimes(onlyUserConfigured) {
    if (!onlyUserConfigured) {
        registeredAssociations = [];
        nonUserRegisteredAssociations = [];
        userRegisteredAssociations = [];
    }
    else {
        registeredAssociations = registeredAssociations.filter(function (a) { return !a.userConfigured; });
        userRegisteredAssociations = [];
    }
}
exports.clearTextMimes = clearTextMimes;
/**
 * Given a file, return the best matching mime type for it
 */
function guessMimeTypes(path, firstLine) {
    if (!path) {
        return [exports.MIME_UNKNOWN];
    }
    path = path.toLowerCase();
    var filename = paths.basename(path);
    // 1.) User configured mappings have highest priority
    var configuredMime = guessMimeTypeByPath(path, filename, userRegisteredAssociations);
    if (configuredMime) {
        return [configuredMime, exports.MIME_TEXT];
    }
    // 2.) Registered mappings have middle priority
    var registeredMime = guessMimeTypeByPath(path, filename, nonUserRegisteredAssociations);
    if (registeredMime) {
        return [registeredMime, exports.MIME_TEXT];
    }
    // 3.) Firstline has lowest priority
    if (firstLine) {
        var firstlineMime = guessMimeTypeByFirstline(firstLine);
        if (firstlineMime) {
            return [firstlineMime, exports.MIME_TEXT];
        }
    }
    return [exports.MIME_UNKNOWN];
}
exports.guessMimeTypes = guessMimeTypes;
function guessMimeTypeByPath(path, filename, associations) {
    var filenameMatch;
    var patternMatch;
    var extensionMatch;
    for (var i = 0; i < associations.length; i++) {
        var association = associations[i];
        // First exact name match
        if (filename === association.filenameLowercase) {
            filenameMatch = association;
            break; // take it!
        }
        // Longest pattern match
        if (association.filepattern) {
            if (!patternMatch || association.filepattern.length > patternMatch.filepattern.length) {
                var target = association.filepatternOnPath ? path : filename; // match on full path if pattern contains path separator
                if (glob_1.match(association.filepatternLowercase, target)) {
                    patternMatch = association;
                }
            }
        }
        // Longest extension match
        if (association.extension) {
            if (!extensionMatch || association.extension.length > extensionMatch.extension.length) {
                if (strings.endsWith(filename, association.extensionLowercase)) {
                    extensionMatch = association;
                }
            }
        }
    }
    // 1.) Exact name match has second highest prio
    if (filenameMatch) {
        return filenameMatch.mime;
    }
    // 2.) Match on pattern
    if (patternMatch) {
        return patternMatch.mime;
    }
    // 3.) Match on extension comes next
    if (extensionMatch) {
        return extensionMatch.mime;
    }
    return null;
}
function guessMimeTypeByFirstline(firstLine) {
    if (strings.startsWithUTF8BOM(firstLine)) {
        firstLine = firstLine.substr(1);
    }
    if (firstLine.length > 0) {
        for (var i = 0; i < registeredAssociations.length; ++i) {
            var association = registeredAssociations[i];
            if (!association.firstline) {
                continue;
            }
            // Make sure the entire line matches, not just a subpart.
            var matches = firstLine.match(association.firstline);
            if (matches && matches.length > 0 && matches[0].length === firstLine.length) {
                return association.mime;
            }
        }
    }
    return null;
}
function isBinaryMime(mimes) {
    if (!mimes) {
        return false;
    }
    var mimeVals;
    if (types.isArray(mimes)) {
        mimeVals = mimes;
    }
    else {
        mimeVals = mimes.split(',').map(function (mime) { return mime.trim(); });
    }
    return mimeVals.indexOf(exports.MIME_BINARY) >= 0;
}
exports.isBinaryMime = isBinaryMime;
function isUnspecific(mime) {
    if (!mime) {
        return true;
    }
    if (typeof mime === 'string') {
        return mime === exports.MIME_BINARY || mime === exports.MIME_TEXT || mime === exports.MIME_UNKNOWN;
    }
    return mime.length === 1 && isUnspecific(mime[0]);
}
exports.isUnspecific = isUnspecific;
function suggestFilename(langId, prefix) {
    for (var i = 0; i < registeredAssociations.length; i++) {
        var association = registeredAssociations[i];
        if (association.userConfigured) {
            continue; // only support registered ones
        }
        if (association.id === langId && association.extension) {
            return prefix + association.extension;
        }
    }
    return prefix; // without any known extension, just return the prefix
}
exports.suggestFilename = suggestFilename;
