// vscode/src/vs/workbench/browser/parts/editor/editorActions.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var winjs_base_1 = require("vs/base/common/winjs.base");
var nls = require("vs/nls");
var actions_1 = require("vs/base/common/actions");
var objects_1 = require("vs/base/common/objects");
var codeEditorService_1 = require("vs/editor/common/services/codeEditorService");
var editor_1 = require("vs/workbench/common/editor");
var quickOpenModel_1 = require("vs/base/parts/quickopen/browser/quickOpenModel");
var quickopen_1 = require("vs/workbench/browser/quickopen");
var editorService_1 = require("vs/workbench/services/editor/common/editorService");
var quickOpenService_1 = require("vs/workbench/services/quickopen/common/quickOpenService");
var partService_1 = require("vs/workbench/services/part/common/partService");
var editor_2 = require("vs/platform/editor/common/editor");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var history_1 = require("vs/workbench/services/history/common/history");
var keybinding_1 = require("vs/platform/keybinding/common/keybinding");
var groupService_1 = require("vs/workbench/services/group/common/groupService");
var commands_1 = require("vs/platform/commands/common/commands");
var SplitEditorAction = (function (_super) {
    __extends(SplitEditorAction, _super);
    function SplitEditorAction(id, label, editorService, editorGroupService) {
        var _this = _super.call(this, id, label, 'split-editor-action') || this;
        _this.editorService = editorService;
        _this.editorGroupService = editorGroupService;
        return _this;
    }
    SplitEditorAction.prototype.run = function (context) {
        var _this = this;
        var editorToSplit;
        if (context) {
            editorToSplit = this.editorService.getVisibleEditors()[this.editorGroupService.getStacksModel().positionOfGroup(context.group)];
        }
        else {
            editorToSplit = this.editorService.getActiveEditor();
        }
        // Can only split with target editor
        if (!editorToSplit) {
            return winjs_base_1.TPromise.as(true);
        }
        // Return if the editor to split does not support split editing
        if (editorToSplit.input instanceof editor_1.EditorInput && !editorToSplit.input.supportsSplitEditor()) {
            return winjs_base_1.TPromise.as(true);
        }
        // Options
        var options;
        var codeEditor = codeEditorService_1.getCodeEditor(editorToSplit);
        if (codeEditor) {
            options = new editor_1.TextEditorOptions();
            options.fromEditor(codeEditor);
        }
        else {
            options = new editor_1.EditorOptions();
        }
        options.pinned = true;
        // Count editors
        var visibleEditors = this.editorService.getVisibleEditors();
        var editorCount = visibleEditors.length;
        var targetPosition;
        switch (editorCount) {
            // Open split editor to the right/bottom of left/top one
            case 1:
                targetPosition = editor_2.Position.TWO;
                break;
            // Special case two editors opened
            case 2:
                // Continue splitting to the right/bottom
                if (editorToSplit.position === editor_2.Position.TWO) {
                    targetPosition = editor_2.Position.THREE;
                }
                else if (editorToSplit.position === editor_2.Position.ONE) {
                    options.preserveFocus = true;
                    return this.editorService.openEditor(editorToSplit.input, options, editor_2.Position.THREE).then(function () {
                        _this.editorGroupService.moveGroup(editor_2.Position.THREE, editor_2.Position.TWO);
                        _this.editorGroupService.focusGroup(editor_2.Position.TWO);
                    });
                }
        }
        // Only split if we have a target position to split to
        if (typeof targetPosition === 'number') {
            return this.editorService.openEditor(editorToSplit.input, options, targetPosition);
        }
        return winjs_base_1.TPromise.as(true);
    };
    return SplitEditorAction;
}(actions_1.Action));
SplitEditorAction.ID = 'workbench.action.splitEditor';
SplitEditorAction.LABEL = nls.localize('splitEditor', "Split Editor");
SplitEditorAction = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService),
    __param(3, groupService_1.IEditorGroupService)
], SplitEditorAction);
exports.SplitEditorAction = SplitEditorAction;
var NavigateBetweenGroupsAction = (function (_super) {
    __extends(NavigateBetweenGroupsAction, _super);
    function NavigateBetweenGroupsAction(id, label, editorService, editorGroupService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorService = editorService;
        _this.editorGroupService = editorGroupService;
        return _this;
    }
    NavigateBetweenGroupsAction.prototype.run = function () {
        // Can cycle split with active editor
        var activeEditor = this.editorService.getActiveEditor();
        if (!activeEditor) {
            return winjs_base_1.TPromise.as(false);
        }
        // Cycle to the left/top and use module to start at 0 again
        var visibleEditors = this.editorService.getVisibleEditors();
        var editorCount = visibleEditors.length;
        var newIndex = (activeEditor.position + 1) % editorCount;
        this.editorGroupService.focusGroup(newIndex);
        return winjs_base_1.TPromise.as(true);
    };
    return NavigateBetweenGroupsAction;
}(actions_1.Action));
NavigateBetweenGroupsAction.ID = 'workbench.action.navigateEditorGroups';
NavigateBetweenGroupsAction.LABEL = nls.localize('navigateEditorGroups', "Navigate Between Editor Groups");
NavigateBetweenGroupsAction = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService),
    __param(3, groupService_1.IEditorGroupService)
], NavigateBetweenGroupsAction);
exports.NavigateBetweenGroupsAction = NavigateBetweenGroupsAction;
var FocusActiveGroupAction = (function (_super) {
    __extends(FocusActiveGroupAction, _super);
    function FocusActiveGroupAction(id, label, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorService = editorService;
        return _this;
    }
    FocusActiveGroupAction.prototype.run = function () {
        var activeEditor = this.editorService.getActiveEditor();
        if (activeEditor) {
            activeEditor.focus();
        }
        return winjs_base_1.TPromise.as(true);
    };
    return FocusActiveGroupAction;
}(actions_1.Action));
FocusActiveGroupAction.ID = 'workbench.action.focusActiveEditorGroup';
FocusActiveGroupAction.LABEL = nls.localize('focusActiveEditorGroup', "Focus Active Editor Group");
FocusActiveGroupAction = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService)
], FocusActiveGroupAction);
exports.FocusActiveGroupAction = FocusActiveGroupAction;
var FocusFirstGroupAction = (function (_super) {
    __extends(FocusFirstGroupAction, _super);
    function FocusFirstGroupAction(id, label, editorService, editorGroupService, historyService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorService = editorService;
        _this.editorGroupService = editorGroupService;
        _this.historyService = historyService;
        return _this;
    }
    FocusFirstGroupAction.prototype.run = function () {
        // Find left/top editor and focus it
        var editors = this.editorService.getVisibleEditors();
        for (var _i = 0, editors_1 = editors; _i < editors_1.length; _i++) {
            var editor = editors_1[_i];
            if (editor.position === editor_2.Position.ONE) {
                this.editorGroupService.focusGroup(editor_2.Position.ONE);
                return winjs_base_1.TPromise.as(true);
            }
        }
        // Since no editor is currently opened, try to open last history entry to the target side
        var history = this.historyService.getHistory();
        for (var _a = 0, history_2 = history; _a < history_2.length; _a++) {
            var input = history_2[_a];
            // For now only support to open files from history to the side
            if (input instanceof editor_1.EditorInput) {
                if (!!editor_1.getUntitledOrFileResource(input)) {
                    return this.editorService.openEditor(input, null, editor_2.Position.ONE);
                }
            }
            else {
                return this.editorService.openEditor(input, editor_2.Position.ONE);
            }
        }
        return winjs_base_1.TPromise.as(true);
    };
    return FocusFirstGroupAction;
}(actions_1.Action));
FocusFirstGroupAction.ID = 'workbench.action.focusFirstEditorGroup';
FocusFirstGroupAction.LABEL = nls.localize('focusFirstEditorGroup', "Focus First Editor Group");
FocusFirstGroupAction = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService),
    __param(3, groupService_1.IEditorGroupService),
    __param(4, history_1.IHistoryService)
], FocusFirstGroupAction);
exports.FocusFirstGroupAction = FocusFirstGroupAction;
var BaseFocusSideGroupAction = (function (_super) {
    __extends(BaseFocusSideGroupAction, _super);
    function BaseFocusSideGroupAction(id, label, editorService, editorGroupService, historyService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorService = editorService;
        _this.editorGroupService = editorGroupService;
        _this.historyService = historyService;
        return _this;
    }
    BaseFocusSideGroupAction.prototype.run = function () {
        // Require at least the reference editor to be visible
        var editors = this.editorService.getVisibleEditors();
        var referenceEditor;
        for (var i = 0; i < editors.length; i++) {
            var editor = editors[i];
            // Target editor exists so focus it
            if (editor.position === this.getTargetEditorSide()) {
                this.editorGroupService.focusGroup(editor.position);
                return winjs_base_1.TPromise.as(true);
            }
            // Remember reference editor
            if (editor.position === this.getReferenceEditorSide()) {
                referenceEditor = editor;
            }
        }
        // Require the reference editor to be visible and supporting split editor
        if (referenceEditor && referenceEditor.input.supportsSplitEditor()) {
            // Options
            var options = void 0;
            var codeEditor = codeEditorService_1.getCodeEditor(referenceEditor);
            if (codeEditor) {
                options = new editor_1.TextEditorOptions();
                options.pinned = true;
                options.fromEditor(codeEditor);
            }
            else {
                options = editor_1.EditorOptions.create({ pinned: true });
            }
            return this.editorService.openEditor(referenceEditor.input, options, this.getTargetEditorSide());
        }
        else if (referenceEditor) {
            var history_3 = this.historyService.getHistory();
            for (var _i = 0, history_4 = history_3; _i < history_4.length; _i++) {
                var input = history_4[_i];
                // For now only support to open files from history to the side
                if (input instanceof editor_1.EditorInput) {
                    if (!!editor_1.getUntitledOrFileResource(input)) {
                        return this.editorService.openEditor(input, { pinned: true }, this.getTargetEditorSide());
                    }
                }
                else {
                    return this.editorService.openEditor({ resource: input.resource, options: { pinned: true } }, this.getTargetEditorSide());
                }
            }
        }
        return winjs_base_1.TPromise.as(true);
    };
    return BaseFocusSideGroupAction;
}(actions_1.Action));
BaseFocusSideGroupAction = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService),
    __param(3, groupService_1.IEditorGroupService),
    __param(4, history_1.IHistoryService)
], BaseFocusSideGroupAction);
exports.BaseFocusSideGroupAction = BaseFocusSideGroupAction;
var FocusSecondGroupAction = (function (_super) {
    __extends(FocusSecondGroupAction, _super);
    function FocusSecondGroupAction(id, label, editorService, editorGroupService, historyService) {
        return _super.call(this, id, label, editorService, editorGroupService, historyService) || this;
    }
    FocusSecondGroupAction.prototype.getReferenceEditorSide = function () {
        return editor_2.Position.ONE;
    };
    FocusSecondGroupAction.prototype.getTargetEditorSide = function () {
        return editor_2.Position.TWO;
    };
    return FocusSecondGroupAction;
}(BaseFocusSideGroupAction));
FocusSecondGroupAction.ID = 'workbench.action.focusSecondEditorGroup';
FocusSecondGroupAction.LABEL = nls.localize('focusSecondEditorGroup', "Focus Second Editor Group");
FocusSecondGroupAction = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService),
    __param(3, groupService_1.IEditorGroupService),
    __param(4, history_1.IHistoryService)
], FocusSecondGroupAction);
exports.FocusSecondGroupAction = FocusSecondGroupAction;
var FocusThirdGroupAction = (function (_super) {
    __extends(FocusThirdGroupAction, _super);
    function FocusThirdGroupAction(id, label, editorService, editorGroupService, historyService) {
        return _super.call(this, id, label, editorService, editorGroupService, historyService) || this;
    }
    FocusThirdGroupAction.prototype.getReferenceEditorSide = function () {
        return editor_2.Position.TWO;
    };
    FocusThirdGroupAction.prototype.getTargetEditorSide = function () {
        return editor_2.Position.THREE;
    };
    return FocusThirdGroupAction;
}(BaseFocusSideGroupAction));
FocusThirdGroupAction.ID = 'workbench.action.focusThirdEditorGroup';
FocusThirdGroupAction.LABEL = nls.localize('focusThirdEditorGroup', "Focus Third Editor Group");
FocusThirdGroupAction = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService),
    __param(3, groupService_1.IEditorGroupService),
    __param(4, history_1.IHistoryService)
], FocusThirdGroupAction);
exports.FocusThirdGroupAction = FocusThirdGroupAction;
var FocusPreviousGroup = (function (_super) {
    __extends(FocusPreviousGroup, _super);
    function FocusPreviousGroup(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    FocusPreviousGroup.prototype.run = function () {
        // Require an active editor
        var activeEditor = this.editorService.getActiveEditor();
        if (!activeEditor) {
            return winjs_base_1.TPromise.as(true);
        }
        // Find the next position to the left/top
        var nextPosition = editor_2.Position.ONE;
        if (activeEditor.position === editor_2.Position.THREE) {
            nextPosition = editor_2.Position.TWO;
        }
        // Focus next position if provided
        this.editorGroupService.focusGroup(nextPosition);
        return winjs_base_1.TPromise.as(true);
    };
    return FocusPreviousGroup;
}(actions_1.Action));
FocusPreviousGroup.ID = 'workbench.action.focusPreviousGroup';
FocusPreviousGroup.LABEL = nls.localize('focusPreviousGroup', "Focus Previous Group");
FocusPreviousGroup = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], FocusPreviousGroup);
exports.FocusPreviousGroup = FocusPreviousGroup;
var FocusNextGroup = (function (_super) {
    __extends(FocusNextGroup, _super);
    function FocusNextGroup(id, label, editorService, instantiationService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorService = editorService;
        _this.navigateActions = [];
        _this.navigateActions[editor_2.Position.ONE] = instantiationService.createInstance(FocusFirstGroupAction, FocusFirstGroupAction.ID, FocusFirstGroupAction.LABEL);
        _this.navigateActions[editor_2.Position.TWO] = instantiationService.createInstance(FocusSecondGroupAction, FocusSecondGroupAction.ID, FocusSecondGroupAction.LABEL);
        _this.navigateActions[editor_2.Position.THREE] = instantiationService.createInstance(FocusThirdGroupAction, FocusThirdGroupAction.ID, FocusThirdGroupAction.LABEL);
        return _this;
    }
    FocusNextGroup.prototype.run = function (event) {
        // Find the next position to the right/bottom to use
        var nextPosition;
        var activeEditor = this.editorService.getActiveEditor();
        if (!activeEditor) {
            nextPosition = editor_2.Position.ONE;
        }
        else if (activeEditor.position === editor_2.Position.ONE) {
            nextPosition = editor_2.Position.TWO;
        }
        else if (activeEditor.position === editor_2.Position.TWO) {
            nextPosition = editor_2.Position.THREE;
        }
        // Run the action for the target next position
        if (typeof nextPosition === 'number' && this.navigateActions[nextPosition]) {
            return this.navigateActions[nextPosition].run(event);
        }
        return winjs_base_1.TPromise.as(true);
    };
    return FocusNextGroup;
}(actions_1.Action));
FocusNextGroup.ID = 'workbench.action.focusNextGroup';
FocusNextGroup.LABEL = nls.localize('focusNextGroup', "Focus Next Group");
FocusNextGroup = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService),
    __param(3, instantiation_1.IInstantiationService)
], FocusNextGroup);
exports.FocusNextGroup = FocusNextGroup;
var OpenToSideAction = (function (_super) {
    __extends(OpenToSideAction, _super);
    function OpenToSideAction(editorService, editorGroupService) {
        var _this = _super.call(this, OpenToSideAction.OPEN_TO_SIDE_ID, OpenToSideAction.OPEN_TO_SIDE_LABEL) || this;
        _this.editorService = editorService;
        _this.editorGroupService = editorGroupService;
        _this.updateEnablement();
        _this.updateClass();
        return _this;
    }
    OpenToSideAction.prototype.updateClass = function () {
        var editorGroupLayoutVertical = (this.editorGroupService.getGroupOrientation() !== 'horizontal');
        this.class = editorGroupLayoutVertical ? 'quick-open-sidebyside-vertical' : 'quick-open-sidebyside-horizontal';
    };
    OpenToSideAction.prototype.updateEnablement = function () {
        var activeEditor = this.editorService.getActiveEditor();
        this.enabled = (!activeEditor || activeEditor.position !== editor_2.Position.THREE);
    };
    OpenToSideAction.prototype.run = function (context) {
        var entry = toEditorQuickOpenEntry(context);
        if (entry) {
            var input = entry.getInput();
            if (input instanceof editor_1.EditorInput) {
                return this.editorService.openEditor(input, entry.getOptions(), true);
            }
            var resourceInput = input;
            resourceInput.options = objects_1.mixin(resourceInput.options, entry.getOptions());
            return this.editorService.openEditor(resourceInput, true);
        }
        return winjs_base_1.TPromise.as(false);
    };
    return OpenToSideAction;
}(actions_1.Action));
OpenToSideAction.OPEN_TO_SIDE_ID = 'workbench.action.openToSide';
OpenToSideAction.OPEN_TO_SIDE_LABEL = nls.localize('openToSide', "Open to the Side");
OpenToSideAction = __decorate([
    __param(0, editorService_1.IWorkbenchEditorService),
    __param(1, groupService_1.IEditorGroupService)
], OpenToSideAction);
exports.OpenToSideAction = OpenToSideAction;
function toEditorQuickOpenEntry(element) {
    // QuickOpenEntryGroup
    if (element instanceof quickOpenModel_1.QuickOpenEntryGroup) {
        var group = element;
        if (group.getEntry()) {
            element = group.getEntry();
        }
    }
    // EditorQuickOpenEntry or EditorQuickOpenEntryGroup both implement IEditorQuickOpenEntry
    if (element instanceof quickopen_1.EditorQuickOpenEntry || element instanceof quickopen_1.EditorQuickOpenEntryGroup) {
        return element;
    }
    return null;
}
exports.toEditorQuickOpenEntry = toEditorQuickOpenEntry;
var CloseEditorAction = (function (_super) {
    __extends(CloseEditorAction, _super);
    function CloseEditorAction(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label, 'close-editor-action') || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    CloseEditorAction.prototype.run = function (context) {
        var position = context ? this.editorGroupService.getStacksModel().positionOfGroup(context.group) : null;
        // Close Active Editor
        if (typeof position !== 'number') {
            var activeEditor = this.editorService.getActiveEditor();
            if (activeEditor) {
                return this.editorService.closeEditor(activeEditor.position, activeEditor.input);
            }
        }
        var input = context ? context.editor : null;
        if (!input) {
            // Get Top Editor at Position
            var visibleEditors = this.editorService.getVisibleEditors();
            if (visibleEditors[position]) {
                input = visibleEditors[position].input;
            }
        }
        if (input) {
            return this.editorService.closeEditor(position, input);
        }
        return winjs_base_1.TPromise.as(false);
    };
    return CloseEditorAction;
}(actions_1.Action));
CloseEditorAction.ID = 'workbench.action.closeActiveEditor';
CloseEditorAction.LABEL = nls.localize('closeEditor', "Close Editor");
CloseEditorAction = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], CloseEditorAction);
exports.CloseEditorAction = CloseEditorAction;
var CloseLeftEditorsInGroupAction = (function (_super) {
    __extends(CloseLeftEditorsInGroupAction, _super);
    function CloseLeftEditorsInGroupAction(id, label, editorService, groupService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorService = editorService;
        _this.groupService = groupService;
        return _this;
    }
    CloseLeftEditorsInGroupAction.prototype.run = function (context) {
        var editor = getTarget(this.editorService, this.groupService, context);
        if (editor) {
            return this.editorService.closeEditors(editor.position, editor.input, editor_2.Direction.LEFT);
        }
        return winjs_base_1.TPromise.as(false);
    };
    return CloseLeftEditorsInGroupAction;
}(actions_1.Action));
CloseLeftEditorsInGroupAction.ID = 'workbench.action.closeEditorsToTheLeft';
CloseLeftEditorsInGroupAction.LABEL = nls.localize('closeEditorsToTheLeft', "Close Editors to the Left");
CloseLeftEditorsInGroupAction = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService),
    __param(3, groupService_1.IEditorGroupService)
], CloseLeftEditorsInGroupAction);
exports.CloseLeftEditorsInGroupAction = CloseLeftEditorsInGroupAction;
var CloseRightEditorsInGroupAction = (function (_super) {
    __extends(CloseRightEditorsInGroupAction, _super);
    function CloseRightEditorsInGroupAction(id, label, editorService, groupService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorService = editorService;
        _this.groupService = groupService;
        return _this;
    }
    CloseRightEditorsInGroupAction.prototype.run = function (context) {
        var editor = getTarget(this.editorService, this.groupService, context);
        if (editor) {
            return this.editorService.closeEditors(editor.position, editor.input, editor_2.Direction.RIGHT);
        }
        return winjs_base_1.TPromise.as(false);
    };
    return CloseRightEditorsInGroupAction;
}(actions_1.Action));
CloseRightEditorsInGroupAction.ID = 'workbench.action.closeEditorsToTheRight';
CloseRightEditorsInGroupAction.LABEL = nls.localize('closeEditorsToTheRight', "Close Editors to the Right");
CloseRightEditorsInGroupAction = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService),
    __param(3, groupService_1.IEditorGroupService)
], CloseRightEditorsInGroupAction);
exports.CloseRightEditorsInGroupAction = CloseRightEditorsInGroupAction;
var CloseAllEditorsAction = (function (_super) {
    __extends(CloseAllEditorsAction, _super);
    function CloseAllEditorsAction(id, label, editorService) {
        var _this = _super.call(this, id, label, 'action-close-all-files') || this;
        _this.editorService = editorService;
        return _this;
    }
    CloseAllEditorsAction.prototype.run = function () {
        return this.editorService.closeAllEditors();
    };
    return CloseAllEditorsAction;
}(actions_1.Action));
CloseAllEditorsAction.ID = 'workbench.action.closeAllEditors';
CloseAllEditorsAction.LABEL = nls.localize('closeAllEditors', "Close All Editors");
CloseAllEditorsAction = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService)
], CloseAllEditorsAction);
exports.CloseAllEditorsAction = CloseAllEditorsAction;
var CloseEditorsInOtherGroupsAction = (function (_super) {
    __extends(CloseEditorsInOtherGroupsAction, _super);
    function CloseEditorsInOtherGroupsAction(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    CloseEditorsInOtherGroupsAction.prototype.run = function (context) {
        var position = context ? this.editorGroupService.getStacksModel().positionOfGroup(context.group) : null;
        if (typeof position !== 'number') {
            var activeEditor = this.editorService.getActiveEditor();
            if (activeEditor) {
                position = activeEditor.position;
            }
        }
        if (typeof position === 'number') {
            return this.editorService.closeAllEditors(position);
        }
        return winjs_base_1.TPromise.as(false);
    };
    return CloseEditorsInOtherGroupsAction;
}(actions_1.Action));
CloseEditorsInOtherGroupsAction.ID = 'workbench.action.closeEditorsInOtherGroups';
CloseEditorsInOtherGroupsAction.LABEL = nls.localize('closeEditorsInOtherGroups', "Close Editors in Other Groups");
CloseEditorsInOtherGroupsAction = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], CloseEditorsInOtherGroupsAction);
exports.CloseEditorsInOtherGroupsAction = CloseEditorsInOtherGroupsAction;
var CloseOtherEditorsInGroupAction = (function (_super) {
    __extends(CloseOtherEditorsInGroupAction, _super);
    function CloseOtherEditorsInGroupAction(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    CloseOtherEditorsInGroupAction.prototype.run = function (context) {
        var position = context ? this.editorGroupService.getStacksModel().positionOfGroup(context.group) : null;
        var input = context ? context.editor : null;
        // If position or input are not passed in take the position and input of the active editor.
        var active = this.editorService.getActiveEditor();
        if (active) {
            position = typeof position === 'number' ? position : active.position;
            input = input ? input : active.input;
        }
        if (typeof position === 'number' && input) {
            return this.editorService.closeEditors(position, input);
        }
        return winjs_base_1.TPromise.as(false);
    };
    return CloseOtherEditorsInGroupAction;
}(actions_1.Action));
CloseOtherEditorsInGroupAction.ID = 'workbench.action.closeOtherEditors';
CloseOtherEditorsInGroupAction.LABEL = nls.localize('closeOtherEditorsInGroup', "Close Other Editors");
CloseOtherEditorsInGroupAction = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], CloseOtherEditorsInGroupAction);
exports.CloseOtherEditorsInGroupAction = CloseOtherEditorsInGroupAction;
var CloseEditorsInGroupAction = (function (_super) {
    __extends(CloseEditorsInGroupAction, _super);
    function CloseEditorsInGroupAction(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    CloseEditorsInGroupAction.prototype.run = function (context) {
        var position = context ? this.editorGroupService.getStacksModel().positionOfGroup(context.group) : null;
        if (typeof position !== 'number') {
            var activeEditor = this.editorService.getActiveEditor();
            if (activeEditor) {
                position = activeEditor.position;
            }
        }
        if (typeof position === 'number') {
            return this.editorService.closeEditors(position);
        }
        return winjs_base_1.TPromise.as(false);
    };
    return CloseEditorsInGroupAction;
}(actions_1.Action));
CloseEditorsInGroupAction.ID = 'workbench.action.closeEditorsInGroup';
CloseEditorsInGroupAction.LABEL = nls.localize('closeEditorsInGroup', "Close All Editors in Group");
CloseEditorsInGroupAction = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], CloseEditorsInGroupAction);
exports.CloseEditorsInGroupAction = CloseEditorsInGroupAction;
var MoveGroupLeftAction = (function (_super) {
    __extends(MoveGroupLeftAction, _super);
    function MoveGroupLeftAction(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    MoveGroupLeftAction.prototype.run = function (context) {
        var position = context ? this.editorGroupService.getStacksModel().positionOfGroup(context.group) : null;
        if (typeof position !== 'number') {
            var activeEditor = this.editorService.getActiveEditor();
            if (activeEditor && (activeEditor.position === editor_2.Position.TWO || activeEditor.position === editor_2.Position.THREE)) {
                position = activeEditor.position;
            }
        }
        if (typeof position === 'number') {
            var newPosition = (position === editor_2.Position.TWO) ? editor_2.Position.ONE : editor_2.Position.TWO;
            // Move group
            this.editorGroupService.moveGroup(position, newPosition);
        }
        return winjs_base_1.TPromise.as(false);
    };
    return MoveGroupLeftAction;
}(actions_1.Action));
MoveGroupLeftAction.ID = 'workbench.action.moveActiveEditorGroupLeft';
MoveGroupLeftAction.LABEL = nls.localize('moveActiveGroupLeft', "Move Editor Group Left");
MoveGroupLeftAction = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], MoveGroupLeftAction);
exports.MoveGroupLeftAction = MoveGroupLeftAction;
var MoveGroupRightAction = (function (_super) {
    __extends(MoveGroupRightAction, _super);
    function MoveGroupRightAction(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    MoveGroupRightAction.prototype.run = function (context) {
        var position = context ? this.editorGroupService.getStacksModel().positionOfGroup(context.group) : null;
        if (typeof position !== 'number') {
            var activeEditor = this.editorService.getActiveEditor();
            var editors = this.editorService.getVisibleEditors();
            if ((editors.length === 2 && activeEditor.position === editor_2.Position.ONE) || (editors.length === 3 && activeEditor.position !== editor_2.Position.THREE)) {
                position = activeEditor.position;
            }
        }
        if (typeof position === 'number') {
            var newPosition = (position === editor_2.Position.ONE) ? editor_2.Position.TWO : editor_2.Position.THREE;
            // Move group
            this.editorGroupService.moveGroup(position, newPosition);
        }
        return winjs_base_1.TPromise.as(false);
    };
    return MoveGroupRightAction;
}(actions_1.Action));
MoveGroupRightAction.ID = 'workbench.action.moveActiveEditorGroupRight';
MoveGroupRightAction.LABEL = nls.localize('moveActiveGroupRight', "Move Editor Group Right");
MoveGroupRightAction = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], MoveGroupRightAction);
exports.MoveGroupRightAction = MoveGroupRightAction;
var MinimizeOtherGroupsAction = (function (_super) {
    __extends(MinimizeOtherGroupsAction, _super);
    function MinimizeOtherGroupsAction(id, label, editorGroupService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        return _this;
    }
    MinimizeOtherGroupsAction.prototype.run = function () {
        this.editorGroupService.arrangeGroups(groupService_1.GroupArrangement.MINIMIZE_OTHERS);
        return winjs_base_1.TPromise.as(false);
    };
    return MinimizeOtherGroupsAction;
}(actions_1.Action));
MinimizeOtherGroupsAction.ID = 'workbench.action.minimizeOtherEditors';
MinimizeOtherGroupsAction.LABEL = nls.localize('minimizeOtherEditorGroups', "Minimize Other Editor Groups");
MinimizeOtherGroupsAction = __decorate([
    __param(2, groupService_1.IEditorGroupService)
], MinimizeOtherGroupsAction);
exports.MinimizeOtherGroupsAction = MinimizeOtherGroupsAction;
var EvenGroupWidthsAction = (function (_super) {
    __extends(EvenGroupWidthsAction, _super);
    function EvenGroupWidthsAction(id, label, editorGroupService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        return _this;
    }
    EvenGroupWidthsAction.prototype.run = function () {
        this.editorGroupService.arrangeGroups(groupService_1.GroupArrangement.EVEN);
        return winjs_base_1.TPromise.as(false);
    };
    return EvenGroupWidthsAction;
}(actions_1.Action));
EvenGroupWidthsAction.ID = 'workbench.action.evenEditorWidths';
EvenGroupWidthsAction.LABEL = nls.localize('evenEditorGroups', "Even Editor Group Widths");
EvenGroupWidthsAction = __decorate([
    __param(2, groupService_1.IEditorGroupService)
], EvenGroupWidthsAction);
exports.EvenGroupWidthsAction = EvenGroupWidthsAction;
var MaximizeGroupAction = (function (_super) {
    __extends(MaximizeGroupAction, _super);
    function MaximizeGroupAction(id, label, editorService, editorGroupService, partService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorService = editorService;
        _this.editorGroupService = editorGroupService;
        _this.partService = partService;
        return _this;
    }
    MaximizeGroupAction.prototype.run = function () {
        if (this.editorService.getActiveEditor()) {
            this.editorGroupService.arrangeGroups(groupService_1.GroupArrangement.MINIMIZE_OTHERS);
            this.partService.setSideBarHidden(true);
        }
        return winjs_base_1.TPromise.as(false);
    };
    return MaximizeGroupAction;
}(actions_1.Action));
MaximizeGroupAction.ID = 'workbench.action.maximizeEditor';
MaximizeGroupAction.LABEL = nls.localize('maximizeEditor', "Maximize Editor Group and Hide Sidebar");
MaximizeGroupAction = __decorate([
    __param(2, editorService_1.IWorkbenchEditorService),
    __param(3, groupService_1.IEditorGroupService),
    __param(4, partService_1.IPartService)
], MaximizeGroupAction);
exports.MaximizeGroupAction = MaximizeGroupAction;
var KeepEditorAction = (function (_super) {
    __extends(KeepEditorAction, _super);
    function KeepEditorAction(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    KeepEditorAction.prototype.run = function (context) {
        var target = getTarget(this.editorService, this.editorGroupService, context);
        if (target) {
            this.editorGroupService.pinEditor(target.position, target.input);
        }
        return winjs_base_1.TPromise.as(true);
    };
    return KeepEditorAction;
}(actions_1.Action));
KeepEditorAction.ID = 'workbench.action.keepEditor';
KeepEditorAction.LABEL = nls.localize('keepEditor', "Keep Editor");
KeepEditorAction = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], KeepEditorAction);
exports.KeepEditorAction = KeepEditorAction;
function getTarget(editorService, editorGroupService, context) {
    if (context) {
        return { input: context.editor, position: editorGroupService.getStacksModel().positionOfGroup(context.group) };
    }
    var activeEditor = editorService.getActiveEditor();
    if (activeEditor) {
        return { input: activeEditor.input, position: activeEditor.position };
    }
    return null;
}
var BaseNavigateEditorAction = (function (_super) {
    __extends(BaseNavigateEditorAction, _super);
    function BaseNavigateEditorAction(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    BaseNavigateEditorAction.prototype.run = function () {
        var model = this.editorGroupService.getStacksModel();
        var result = this.navigate();
        if (result) {
            return this.editorService.openEditor(result.editor, null, model.positionOfGroup(result.group));
        }
        return winjs_base_1.TPromise.as(false);
    };
    return BaseNavigateEditorAction;
}(actions_1.Action));
exports.BaseNavigateEditorAction = BaseNavigateEditorAction;
var OpenNextEditor = (function (_super) {
    __extends(OpenNextEditor, _super);
    function OpenNextEditor(id, label, editorGroupService, editorService) {
        return _super.call(this, id, label, editorGroupService, editorService) || this;
    }
    OpenNextEditor.prototype.navigate = function () {
        return this.editorGroupService.getStacksModel().next(true /* jump groups */);
    };
    return OpenNextEditor;
}(BaseNavigateEditorAction));
OpenNextEditor.ID = 'workbench.action.nextEditor';
OpenNextEditor.LABEL = nls.localize('openNextEditor', "Open Next Editor");
OpenNextEditor = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], OpenNextEditor);
exports.OpenNextEditor = OpenNextEditor;
var OpenPreviousEditor = (function (_super) {
    __extends(OpenPreviousEditor, _super);
    function OpenPreviousEditor(id, label, editorGroupService, editorService) {
        return _super.call(this, id, label, editorGroupService, editorService) || this;
    }
    OpenPreviousEditor.prototype.navigate = function () {
        return this.editorGroupService.getStacksModel().previous(true /* jump groups */);
    };
    return OpenPreviousEditor;
}(BaseNavigateEditorAction));
OpenPreviousEditor.ID = 'workbench.action.previousEditor';
OpenPreviousEditor.LABEL = nls.localize('openPreviousEditor', "Open Previous Editor");
OpenPreviousEditor = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], OpenPreviousEditor);
exports.OpenPreviousEditor = OpenPreviousEditor;
var OpenNextEditorInGroup = (function (_super) {
    __extends(OpenNextEditorInGroup, _super);
    function OpenNextEditorInGroup(id, label, editorGroupService, editorService) {
        return _super.call(this, id, label, editorGroupService, editorService) || this;
    }
    OpenNextEditorInGroup.prototype.navigate = function () {
        return this.editorGroupService.getStacksModel().next(false /* do NOT jump groups */);
    };
    return OpenNextEditorInGroup;
}(BaseNavigateEditorAction));
OpenNextEditorInGroup.ID = 'workbench.action.nextEditorInGroup';
OpenNextEditorInGroup.LABEL = nls.localize('nextEditorInGroup', "Open Next Editor in Group");
OpenNextEditorInGroup = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], OpenNextEditorInGroup);
exports.OpenNextEditorInGroup = OpenNextEditorInGroup;
var OpenPreviousEditorInGroup = (function (_super) {
    __extends(OpenPreviousEditorInGroup, _super);
    function OpenPreviousEditorInGroup(id, label, editorGroupService, editorService) {
        return _super.call(this, id, label, editorGroupService, editorService) || this;
    }
    OpenPreviousEditorInGroup.prototype.navigate = function () {
        return this.editorGroupService.getStacksModel().previous(false /* do NOT jump groups */);
    };
    return OpenPreviousEditorInGroup;
}(BaseNavigateEditorAction));
OpenPreviousEditorInGroup.ID = 'workbench.action.previousEditorInGroup';
OpenPreviousEditorInGroup.LABEL = nls.localize('openPreviousEditorInGroup', "Open Previous Editor in Group");
OpenPreviousEditorInGroup = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], OpenPreviousEditorInGroup);
exports.OpenPreviousEditorInGroup = OpenPreviousEditorInGroup;
var NavigateForwardAction = (function (_super) {
    __extends(NavigateForwardAction, _super);
    function NavigateForwardAction(id, label, historyService) {
        var _this = _super.call(this, id, label) || this;
        _this.historyService = historyService;
        return _this;
    }
    NavigateForwardAction.prototype.run = function () {
        this.historyService.forward();
        return winjs_base_1.TPromise.as(null);
    };
    return NavigateForwardAction;
}(actions_1.Action));
NavigateForwardAction.ID = 'workbench.action.navigateForward';
NavigateForwardAction.LABEL = nls.localize('navigateNext', "Go Forward");
NavigateForwardAction = __decorate([
    __param(2, history_1.IHistoryService)
], NavigateForwardAction);
exports.NavigateForwardAction = NavigateForwardAction;
var NavigateBackwardsAction = (function (_super) {
    __extends(NavigateBackwardsAction, _super);
    function NavigateBackwardsAction(id, label, historyService) {
        var _this = _super.call(this, id, label) || this;
        _this.historyService = historyService;
        return _this;
    }
    NavigateBackwardsAction.prototype.run = function () {
        this.historyService.back();
        return winjs_base_1.TPromise.as(null);
    };
    return NavigateBackwardsAction;
}(actions_1.Action));
NavigateBackwardsAction.ID = 'workbench.action.navigateBack';
NavigateBackwardsAction.LABEL = nls.localize('navigatePrevious', "Go Back");
NavigateBackwardsAction = __decorate([
    __param(2, history_1.IHistoryService)
], NavigateBackwardsAction);
exports.NavigateBackwardsAction = NavigateBackwardsAction;
var ReopenClosedEditorAction = (function (_super) {
    __extends(ReopenClosedEditorAction, _super);
    function ReopenClosedEditorAction(id, label, historyService) {
        var _this = _super.call(this, id, label) || this;
        _this.historyService = historyService;
        return _this;
    }
    ReopenClosedEditorAction.prototype.run = function () {
        this.historyService.reopenLastClosedEditor();
        return winjs_base_1.TPromise.as(false);
    };
    return ReopenClosedEditorAction;
}(actions_1.Action));
ReopenClosedEditorAction.ID = 'workbench.action.reopenClosedEditor';
ReopenClosedEditorAction.LABEL = nls.localize('reopenClosedEditor', "Reopen Closed Editor");
ReopenClosedEditorAction = __decorate([
    __param(2, history_1.IHistoryService)
], ReopenClosedEditorAction);
exports.ReopenClosedEditorAction = ReopenClosedEditorAction;
exports.NAVIGATE_IN_GROUP_ONE_PREFIX = 'edt one ';
var ShowEditorsInGroupOneAction = (function (_super) {
    __extends(ShowEditorsInGroupOneAction, _super);
    function ShowEditorsInGroupOneAction(actionId, actionLabel, quickOpenService) {
        var _this = _super.call(this, actionId, actionLabel, exports.NAVIGATE_IN_GROUP_ONE_PREFIX, quickOpenService) || this;
        _this.class = 'show-group-editors-action';
        return _this;
    }
    return ShowEditorsInGroupOneAction;
}(quickopen_1.QuickOpenAction));
ShowEditorsInGroupOneAction.ID = 'workbench.action.showEditorsInFirstGroup';
ShowEditorsInGroupOneAction.LABEL = nls.localize('showEditorsInFirstGroup', "Show Editors in First Group");
ShowEditorsInGroupOneAction = __decorate([
    __param(2, quickOpenService_1.IQuickOpenService)
], ShowEditorsInGroupOneAction);
exports.ShowEditorsInGroupOneAction = ShowEditorsInGroupOneAction;
exports.NAVIGATE_IN_GROUP_TWO_PREFIX = 'edt two ';
var ShowEditorsInGroupTwoAction = (function (_super) {
    __extends(ShowEditorsInGroupTwoAction, _super);
    function ShowEditorsInGroupTwoAction(actionId, actionLabel, quickOpenService) {
        var _this = _super.call(this, actionId, actionLabel, exports.NAVIGATE_IN_GROUP_TWO_PREFIX, quickOpenService) || this;
        _this.class = 'show-group-editors-action';
        return _this;
    }
    return ShowEditorsInGroupTwoAction;
}(quickopen_1.QuickOpenAction));
ShowEditorsInGroupTwoAction.ID = 'workbench.action.showEditorsInSecondGroup';
ShowEditorsInGroupTwoAction.LABEL = nls.localize('showEditorsInSecondGroup', "Show Editors in Second Group");
ShowEditorsInGroupTwoAction = __decorate([
    __param(2, quickOpenService_1.IQuickOpenService)
], ShowEditorsInGroupTwoAction);
exports.ShowEditorsInGroupTwoAction = ShowEditorsInGroupTwoAction;
exports.NAVIGATE_IN_GROUP_THREE_PREFIX = 'edt three ';
var ShowEditorsInGroupThreeAction = (function (_super) {
    __extends(ShowEditorsInGroupThreeAction, _super);
    function ShowEditorsInGroupThreeAction(actionId, actionLabel, quickOpenService) {
        var _this = _super.call(this, actionId, actionLabel, exports.NAVIGATE_IN_GROUP_THREE_PREFIX, quickOpenService) || this;
        _this.class = 'show-group-editors-action';
        return _this;
    }
    return ShowEditorsInGroupThreeAction;
}(quickopen_1.QuickOpenAction));
ShowEditorsInGroupThreeAction.ID = 'workbench.action.showEditorsInThirdGroup';
ShowEditorsInGroupThreeAction.LABEL = nls.localize('showEditorsInThirdGroup', "Show Editors in Third Group");
ShowEditorsInGroupThreeAction = __decorate([
    __param(2, quickOpenService_1.IQuickOpenService)
], ShowEditorsInGroupThreeAction);
exports.ShowEditorsInGroupThreeAction = ShowEditorsInGroupThreeAction;
var ShowEditorsInGroupAction = (function (_super) {
    __extends(ShowEditorsInGroupAction, _super);
    function ShowEditorsInGroupAction(id, label, quickOpenService, editorGroupService) {
        var _this = _super.call(this, id, label) || this;
        _this.quickOpenService = quickOpenService;
        _this.editorGroupService = editorGroupService;
        return _this;
    }
    ShowEditorsInGroupAction.prototype.run = function (context) {
        var stacks = this.editorGroupService.getStacksModel();
        var groupCount = stacks.groups.length;
        if (groupCount <= 1 || !context) {
            return this.quickOpenService.show(exports.NAVIGATE_ALL_EDITORS_GROUP_PREFIX);
        }
        switch (stacks.positionOfGroup(context.group)) {
            case editor_2.Position.TWO:
                return this.quickOpenService.show(exports.NAVIGATE_IN_GROUP_TWO_PREFIX);
            case editor_2.Position.THREE:
                return this.quickOpenService.show(exports.NAVIGATE_IN_GROUP_THREE_PREFIX);
        }
        return this.quickOpenService.show(exports.NAVIGATE_IN_GROUP_ONE_PREFIX);
    };
    return ShowEditorsInGroupAction;
}(actions_1.Action));
ShowEditorsInGroupAction.ID = 'workbench.action.showEditorsInGroup';
ShowEditorsInGroupAction.LABEL = nls.localize('showEditorsInGroup', "Show Editors in Group");
ShowEditorsInGroupAction = __decorate([
    __param(2, quickOpenService_1.IQuickOpenService),
    __param(3, groupService_1.IEditorGroupService)
], ShowEditorsInGroupAction);
exports.ShowEditorsInGroupAction = ShowEditorsInGroupAction;
exports.NAVIGATE_ALL_EDITORS_GROUP_PREFIX = 'edt ';
var ShowAllEditorsAction = (function (_super) {
    __extends(ShowAllEditorsAction, _super);
    function ShowAllEditorsAction(actionId, actionLabel, quickOpenService) {
        return _super.call(this, actionId, actionLabel, exports.NAVIGATE_ALL_EDITORS_GROUP_PREFIX, quickOpenService) || this;
    }
    return ShowAllEditorsAction;
}(quickopen_1.QuickOpenAction));
ShowAllEditorsAction.ID = 'workbench.action.showAllEditors';
ShowAllEditorsAction.LABEL = nls.localize('showAllEditors', "Show All Editors");
ShowAllEditorsAction = __decorate([
    __param(2, quickOpenService_1.IQuickOpenService)
], ShowAllEditorsAction);
exports.ShowAllEditorsAction = ShowAllEditorsAction;
var BaseQuickOpenEditorInGroupAction = (function (_super) {
    __extends(BaseQuickOpenEditorInGroupAction, _super);
    function BaseQuickOpenEditorInGroupAction(id, label, quickOpenService, keybindingService, editorGroupService) {
        var _this = _super.call(this, id, label) || this;
        _this.quickOpenService = quickOpenService;
        _this.keybindingService = keybindingService;
        _this.editorGroupService = editorGroupService;
        return _this;
    }
    BaseQuickOpenEditorInGroupAction.prototype.run = function () {
        var keys = this.keybindingService.lookupKeybindings(this.id);
        var stacks = this.editorGroupService.getStacksModel();
        if (stacks.activeGroup) {
            var activePosition = stacks.positionOfGroup(stacks.activeGroup);
            var prefix = exports.NAVIGATE_IN_GROUP_ONE_PREFIX;
            if (activePosition === editor_2.Position.TWO) {
                prefix = exports.NAVIGATE_IN_GROUP_TWO_PREFIX;
            }
            else if (activePosition === editor_2.Position.THREE) {
                prefix = exports.NAVIGATE_IN_GROUP_THREE_PREFIX;
            }
            this.quickOpenService.show(prefix, { quickNavigateConfiguration: { keybindings: keys } });
        }
        return winjs_base_1.TPromise.as(true);
    };
    return BaseQuickOpenEditorInGroupAction;
}(actions_1.Action));
BaseQuickOpenEditorInGroupAction = __decorate([
    __param(2, quickOpenService_1.IQuickOpenService),
    __param(3, keybinding_1.IKeybindingService),
    __param(4, groupService_1.IEditorGroupService)
], BaseQuickOpenEditorInGroupAction);
exports.BaseQuickOpenEditorInGroupAction = BaseQuickOpenEditorInGroupAction;
var OpenPreviousRecentlyUsedEditorInGroupAction = (function (_super) {
    __extends(OpenPreviousRecentlyUsedEditorInGroupAction, _super);
    function OpenPreviousRecentlyUsedEditorInGroupAction(id, label, quickOpenService, keybindingService, editorGroupService) {
        return _super.call(this, id, label, quickOpenService, keybindingService, editorGroupService) || this;
    }
    return OpenPreviousRecentlyUsedEditorInGroupAction;
}(BaseQuickOpenEditorInGroupAction));
OpenPreviousRecentlyUsedEditorInGroupAction.ID = 'workbench.action.openPreviousRecentlyUsedEditorInGroup';
OpenPreviousRecentlyUsedEditorInGroupAction.LABEL = nls.localize('openPreviousRecentlyUsedEditorInGroup', "Open Previous Recently Used Editor in Group");
OpenPreviousRecentlyUsedEditorInGroupAction = __decorate([
    __param(2, quickOpenService_1.IQuickOpenService),
    __param(3, keybinding_1.IKeybindingService),
    __param(4, groupService_1.IEditorGroupService)
], OpenPreviousRecentlyUsedEditorInGroupAction);
exports.OpenPreviousRecentlyUsedEditorInGroupAction = OpenPreviousRecentlyUsedEditorInGroupAction;
var OpenNextRecentlyUsedEditorInGroupAction = (function (_super) {
    __extends(OpenNextRecentlyUsedEditorInGroupAction, _super);
    function OpenNextRecentlyUsedEditorInGroupAction(id, label, quickOpenService, keybindingService, editorGroupService) {
        return _super.call(this, id, label, quickOpenService, keybindingService, editorGroupService) || this;
    }
    return OpenNextRecentlyUsedEditorInGroupAction;
}(BaseQuickOpenEditorInGroupAction));
OpenNextRecentlyUsedEditorInGroupAction.ID = 'workbench.action.openNextRecentlyUsedEditorInGroup';
OpenNextRecentlyUsedEditorInGroupAction.LABEL = nls.localize('openNextRecentlyUsedEditorInGroup', "Open Next Recently Used Editor in Group");
OpenNextRecentlyUsedEditorInGroupAction = __decorate([
    __param(2, quickOpenService_1.IQuickOpenService),
    __param(3, keybinding_1.IKeybindingService),
    __param(4, groupService_1.IEditorGroupService)
], OpenNextRecentlyUsedEditorInGroupAction);
exports.OpenNextRecentlyUsedEditorInGroupAction = OpenNextRecentlyUsedEditorInGroupAction;
var OpenPreviousEditorFromHistoryAction = (function (_super) {
    __extends(OpenPreviousEditorFromHistoryAction, _super);
    function OpenPreviousEditorFromHistoryAction(id, label, quickOpenService, keybindingService) {
        var _this = _super.call(this, id, label) || this;
        _this.quickOpenService = quickOpenService;
        _this.keybindingService = keybindingService;
        return _this;
    }
    OpenPreviousEditorFromHistoryAction.prototype.run = function () {
        var keys = this.keybindingService.lookupKeybindings(this.id);
        this.quickOpenService.show(null, { quickNavigateConfiguration: { keybindings: keys } });
        return winjs_base_1.TPromise.as(true);
    };
    return OpenPreviousEditorFromHistoryAction;
}(actions_1.Action));
OpenPreviousEditorFromHistoryAction.ID = 'workbench.action.openPreviousEditorFromHistory';
OpenPreviousEditorFromHistoryAction.LABEL = nls.localize('navigateEditorHistoryByInput', "Open Previous Editor from History");
OpenPreviousEditorFromHistoryAction = __decorate([
    __param(2, quickOpenService_1.IQuickOpenService),
    __param(3, keybinding_1.IKeybindingService)
], OpenPreviousEditorFromHistoryAction);
exports.OpenPreviousEditorFromHistoryAction = OpenPreviousEditorFromHistoryAction;
var ClearEditorHistoryAction = (function (_super) {
    __extends(ClearEditorHistoryAction, _super);
    function ClearEditorHistoryAction(id, label, historyService) {
        var _this = _super.call(this, id, label) || this;
        _this.historyService = historyService;
        return _this;
    }
    ClearEditorHistoryAction.prototype.run = function () {
        // Editor history
        this.historyService.clear();
        return winjs_base_1.TPromise.as(true);
    };
    return ClearEditorHistoryAction;
}(actions_1.Action));
ClearEditorHistoryAction.ID = 'workbench.action.clearEditorHistory';
ClearEditorHistoryAction.LABEL = nls.localize('clearEditorHistory', "Clear Editor History");
ClearEditorHistoryAction = __decorate([
    __param(2, history_1.IHistoryService)
], ClearEditorHistoryAction);
exports.ClearEditorHistoryAction = ClearEditorHistoryAction;
var FocusLastEditorInStackAction = (function (_super) {
    __extends(FocusLastEditorInStackAction, _super);
    function FocusLastEditorInStackAction(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    FocusLastEditorInStackAction.prototype.run = function () {
        var active = this.editorService.getActiveEditor();
        if (active) {
            var group = this.editorGroupService.getStacksModel().groupAt(active.position);
            var editor = group.getEditor(group.count - 1);
            if (editor) {
                return this.editorService.openEditor(editor);
            }
        }
        return winjs_base_1.TPromise.as(true);
    };
    return FocusLastEditorInStackAction;
}(actions_1.Action));
FocusLastEditorInStackAction.ID = 'workbench.action.openLastEditorInGroup';
FocusLastEditorInStackAction.LABEL = nls.localize('focusLastEditorInStack', "Open Last Editor in Group");
FocusLastEditorInStackAction = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], FocusLastEditorInStackAction);
exports.FocusLastEditorInStackAction = FocusLastEditorInStackAction;
var MoveEditorLeftInGroupAction = (function (_super) {
    __extends(MoveEditorLeftInGroupAction, _super);
    function MoveEditorLeftInGroupAction(id, label, commandService) {
        var _this = _super.call(this, id, label) || this;
        _this.commandService = commandService;
        return _this;
    }
    MoveEditorLeftInGroupAction.prototype.run = function () {
        var args = {
            to: editor_1.ActiveEditorMovePositioning.LEFT
        };
        this.commandService.executeCommand(editor_1.EditorCommands.MoveActiveEditor, args);
        return winjs_base_1.TPromise.as(true);
    };
    return MoveEditorLeftInGroupAction;
}(actions_1.Action));
MoveEditorLeftInGroupAction.ID = 'workbench.action.moveEditorLeftInGroup';
MoveEditorLeftInGroupAction.LABEL = nls.localize('moveEditorLeft', "Move Editor Left");
MoveEditorLeftInGroupAction = __decorate([
    __param(2, commands_1.ICommandService)
], MoveEditorLeftInGroupAction);
exports.MoveEditorLeftInGroupAction = MoveEditorLeftInGroupAction;
var MoveEditorRightInGroupAction = (function (_super) {
    __extends(MoveEditorRightInGroupAction, _super);
    function MoveEditorRightInGroupAction(id, label, commandService) {
        var _this = _super.call(this, id, label) || this;
        _this.commandService = commandService;
        return _this;
    }
    MoveEditorRightInGroupAction.prototype.run = function () {
        var args = {
            to: editor_1.ActiveEditorMovePositioning.RIGHT
        };
        this.commandService.executeCommand(editor_1.EditorCommands.MoveActiveEditor, args);
        return winjs_base_1.TPromise.as(true);
    };
    return MoveEditorRightInGroupAction;
}(actions_1.Action));
MoveEditorRightInGroupAction.ID = 'workbench.action.moveEditorRightInGroup';
MoveEditorRightInGroupAction.LABEL = nls.localize('moveEditorRight', "Move Editor Right");
MoveEditorRightInGroupAction = __decorate([
    __param(2, commands_1.ICommandService)
], MoveEditorRightInGroupAction);
exports.MoveEditorRightInGroupAction = MoveEditorRightInGroupAction;
var MoveEditorToPreviousGroupAction = (function (_super) {
    __extends(MoveEditorToPreviousGroupAction, _super);
    function MoveEditorToPreviousGroupAction(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    MoveEditorToPreviousGroupAction.prototype.run = function () {
        var activeEditor = this.editorService.getActiveEditor();
        if (activeEditor && activeEditor.position !== editor_2.Position.ONE) {
            this.editorGroupService.moveEditor(activeEditor.input, activeEditor.position, activeEditor.position - 1);
        }
        return winjs_base_1.TPromise.as(true);
    };
    return MoveEditorToPreviousGroupAction;
}(actions_1.Action));
MoveEditorToPreviousGroupAction.ID = 'workbench.action.moveEditorToPreviousGroup';
MoveEditorToPreviousGroupAction.LABEL = nls.localize('moveEditorToPreviousGroup', "Move Editor into Previous Group");
MoveEditorToPreviousGroupAction = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], MoveEditorToPreviousGroupAction);
exports.MoveEditorToPreviousGroupAction = MoveEditorToPreviousGroupAction;
var MoveEditorToNextGroupAction = (function (_super) {
    __extends(MoveEditorToNextGroupAction, _super);
    function MoveEditorToNextGroupAction(id, label, editorGroupService, editorService) {
        var _this = _super.call(this, id, label) || this;
        _this.editorGroupService = editorGroupService;
        _this.editorService = editorService;
        return _this;
    }
    MoveEditorToNextGroupAction.prototype.run = function () {
        var activeEditor = this.editorService.getActiveEditor();
        if (activeEditor && activeEditor.position !== editor_2.Position.THREE) {
            this.editorGroupService.moveEditor(activeEditor.input, activeEditor.position, activeEditor.position + 1);
        }
        return winjs_base_1.TPromise.as(true);
    };
    return MoveEditorToNextGroupAction;
}(actions_1.Action));
MoveEditorToNextGroupAction.ID = 'workbench.action.moveEditorToNextGroup';
MoveEditorToNextGroupAction.LABEL = nls.localize('moveEditorToNextGroup', "Move Editor into Next Group");
MoveEditorToNextGroupAction = __decorate([
    __param(2, groupService_1.IEditorGroupService),
    __param(3, editorService_1.IWorkbenchEditorService)
], MoveEditorToNextGroupAction);
exports.MoveEditorToNextGroupAction = MoveEditorToNextGroupAction;
