diff --git a/cmd/zoekt-sourcegraph-indexserver/BUILD.bazel b/cmd/zoekt-sourcegraph-indexserver/BUILD.bazel
index 503fd67..6989a2d 100644
--- a/cmd/zoekt-sourcegraph-indexserver/BUILD.bazel
+++ b/cmd/zoekt-sourcegraph-indexserver/BUILD.bazel
@@ -50,6 +50,8 @@ go_binary(
     name = "zoekt-sourcegraph-indexserver",
     embed = [":zoekt-sourcegraph-indexserver_lib"],
     visibility = ["//visibility:public"],
+    pure = "on",
+    static = "on",
 )

 go_test(
