// vscode/src/vs/platform/message/common/messageIpc.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var winjs_base_1 = require("vs/base/common/winjs.base");
var ChoiceChannel = (function () {
    function ChoiceChannel(service) {
        this.service = service;
    }
    ChoiceChannel.prototype.call = function (command, args) {
        switch (command) {
            case 'choose': return this.service.choose(args[0], args[1], args[2], args[3]);
        }
        return winjs_base_1.TPromise.wrapError('invalid command');
    };
    return ChoiceChannel;
}());
exports.ChoiceChannel = ChoiceChannel;
var ChoiceChannelClient = (function () {
    function ChoiceChannelClient(channel) {
        this.channel = channel;
    }
    ChoiceChannelClient.prototype.choose = function (severity, message, options, modal) {
        return this.channel.call('choose', [severity, message, options, modal]);
    };
    return ChoiceChannelClient;
}());
exports.ChoiceChannelClient = ChoiceChannelClient;
