diff --git a/lib/private/run_binary.bzl b/lib/private/run_binary.bzl
index 7ba4210..55bd984 100644
--- a/lib/private/run_binary.bzl
+++ b/lib/private/run_binary.bzl
@@ -76,7 +76,7 @@ Possible fixes:
         mnemonic = ctx.attr.mnemonic if ctx.attr.mnemonic else None,
         progress_message = ctx.attr.progress_message if ctx.attr.progress_message else None,
         execution_requirements = ctx.attr.execution_requirements if ctx.attr.execution_requirements else None,
-        use_default_shell_env = False,
+        use_default_shell_env = True,
         env = dicts.add(ctx.configuration.default_shell_env, envs),
         input_manifests = tool_input_mfs,
     )
@@ -95,6 +95,7 @@ _run_binary = rule(
             cfg = "exec",
         ),
         "env": attr.string_dict(),
+        "use_default_shell_env": attr.bool(default = False),
         "srcs": attr.label_list(
             allow_files = True,
         ),
@@ -113,6 +114,7 @@ def run_binary(
         srcs = [],
         args = [],
         env = {},
+        use_default_shell_env = False,
         outs = [],
         out_dirs = [],
         mnemonic = "RunBinary",
@@ -220,6 +222,7 @@ def run_binary(
         srcs = srcs,
         args = args,
         env = env,
+        use_default_shell_env = use_default_shell_env,
         outs = outs,
         out_dirs = out_dirs + ([name] if output_dir else []),
         mnemonic = mnemonic,
