// vscode/src/vs/workbench/browser/parts/panel/panelActions.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
require("./media/panelpart.css");
var nls = require("vs/nls");
var winjs_base_1 = require("vs/base/common/winjs.base");
var actions_1 = require("vs/base/common/actions");
var platform_1 = require("vs/platform/platform");
var actions_2 = require("vs/platform/actions/common/actions");
var actionRegistry_1 = require("vs/workbench/common/actionRegistry");
var panelService_1 = require("vs/workbench/services/panel/common/panelService");
var partService_1 = require("vs/workbench/services/part/common/partService");
var keybinding_1 = require("vs/platform/keybinding/common/keybinding");
var PanelAction = (function (_super) {
    __extends(PanelAction, _super);
    function PanelAction(panel, keybindingService, panelService) {
        var _this = _super.call(this, panel.id, panel.name) || this;
        _this.panel = panel;
        _this.keybindingService = keybindingService;
        _this.panelService = panelService;
        _this.tooltip = nls.localize('panelActionTooltip', "{0} ({1})", panel.name, _this.getKeybindingLabel(panel.commandId));
        return _this;
    }
    PanelAction.prototype.run = function (event) {
        var _this = this;
        return this.panelService.openPanel(this.panel.id, true).then(function () { return _this.activate(); });
    };
    PanelAction.prototype.activate = function () {
        if (!this.checked) {
            this._setChecked(true);
        }
    };
    PanelAction.prototype.deactivate = function () {
        if (this.checked) {
            this._setChecked(false);
        }
    };
    PanelAction.prototype.getKeybindingLabel = function (id) {
        var keys = this.keybindingService.lookupKeybinding(id);
        return keys ? keys.getLabel() : '';
    };
    return PanelAction;
}(actions_1.Action));
PanelAction = __decorate([
    __param(1, keybinding_1.IKeybindingService),
    __param(2, panelService_1.IPanelService)
], PanelAction);
exports.PanelAction = PanelAction;
var ClosePanelAction = (function (_super) {
    __extends(ClosePanelAction, _super);
    function ClosePanelAction(id, name, partService) {
        var _this = _super.call(this, id, name, 'hide-panel-action') || this;
        _this.partService = partService;
        return _this;
    }
    ClosePanelAction.prototype.run = function () {
        return this.partService.setPanelHidden(true);
    };
    return ClosePanelAction;
}(actions_1.Action));
ClosePanelAction.ID = 'workbench.action.closePanel';
ClosePanelAction.LABEL = nls.localize('closePanel', "Close Panel");
ClosePanelAction = __decorate([
    __param(2, partService_1.IPartService)
], ClosePanelAction);
exports.ClosePanelAction = ClosePanelAction;
var TogglePanelAction = (function (_super) {
    __extends(TogglePanelAction, _super);
    function TogglePanelAction(id, name, partService) {
        var _this = _super.call(this, id, name, partService.isVisible(partService_1.Parts.PANEL_PART) ? 'panel expanded' : 'panel') || this;
        _this.partService = partService;
        return _this;
    }
    TogglePanelAction.prototype.run = function () {
        return this.partService.setPanelHidden(this.partService.isVisible(partService_1.Parts.PANEL_PART));
    };
    return TogglePanelAction;
}(actions_1.Action));
TogglePanelAction.ID = 'workbench.action.togglePanel';
TogglePanelAction.LABEL = nls.localize('togglePanel', "Toggle Panel");
TogglePanelAction = __decorate([
    __param(2, partService_1.IPartService)
], TogglePanelAction);
exports.TogglePanelAction = TogglePanelAction;
var FocusPanelAction = (function (_super) {
    __extends(FocusPanelAction, _super);
    function FocusPanelAction(id, label, panelService, partService) {
        var _this = _super.call(this, id, label) || this;
        _this.panelService = panelService;
        _this.partService = partService;
        return _this;
    }
    FocusPanelAction.prototype.run = function () {
        // Show panel
        if (!this.partService.isVisible(partService_1.Parts.PANEL_PART)) {
            return this.partService.setPanelHidden(false);
        }
        // Focus into active panel
        var panel = this.panelService.getActivePanel();
        if (panel) {
            panel.focus();
        }
        return winjs_base_1.TPromise.as(true);
    };
    return FocusPanelAction;
}(actions_1.Action));
FocusPanelAction.ID = 'workbench.action.focusPanel';
FocusPanelAction.LABEL = nls.localize('focusPanel', "Focus into Panel");
FocusPanelAction = __decorate([
    __param(2, panelService_1.IPanelService),
    __param(3, partService_1.IPartService)
], FocusPanelAction);
var ToggleMaximizedPanelAction = (function (_super) {
    __extends(ToggleMaximizedPanelAction, _super);
    function ToggleMaximizedPanelAction(id, label, partService) {
        var _this = _super.call(this, id, label) || this;
        _this.partService = partService;
        return _this;
    }
    ToggleMaximizedPanelAction.prototype.run = function () {
        var _this = this;
        // Show panel
        return this.partService.setPanelHidden(false)
            .then(function () { return _this.partService.toggleMaximizedPanel(); });
    };
    return ToggleMaximizedPanelAction;
}(actions_1.Action));
ToggleMaximizedPanelAction.ID = 'workbench.action.toggleMaximizedPanel';
ToggleMaximizedPanelAction.LABEL = nls.localize('toggleMaximizedPanel', "Toggle Maximized Panel");
ToggleMaximizedPanelAction = __decorate([
    __param(2, partService_1.IPartService)
], ToggleMaximizedPanelAction);
var actionRegistry = platform_1.Registry.as(actionRegistry_1.Extensions.WorkbenchActions);
actionRegistry.registerWorkbenchAction(new actions_2.SyncActionDescriptor(TogglePanelAction, TogglePanelAction.ID, TogglePanelAction.LABEL, { primary: 2048 /* CtrlCmd */ | 40 /* KEY_J */ }), 'View: Toggle Panel Visibility', nls.localize('view', "View"));
actionRegistry.registerWorkbenchAction(new actions_2.SyncActionDescriptor(FocusPanelAction, FocusPanelAction.ID, FocusPanelAction.LABEL), 'View: Focus into Panel', nls.localize('view', "View"));
actionRegistry.registerWorkbenchAction(new actions_2.SyncActionDescriptor(ToggleMaximizedPanelAction, ToggleMaximizedPanelAction.ID, ToggleMaximizedPanelAction.LABEL), 'View: Toggle Maximized Panel', nls.localize('view', "View"));
actionRegistry.registerWorkbenchAction(new actions_2.SyncActionDescriptor(ClosePanelAction, ClosePanelAction.ID, ClosePanelAction.LABEL), 'View: Close Panel', nls.localize('view', "View"));
