// commandpost/lib/command.js
"use strict";
var option_1 = require("./option");
var argument_1 = require("./argument");
var utils = require("./utils");
// jsdoc, see constructor.
var Command = (function () {
    /**
     * class of command.
     * ```
     * cmd -v sub --path foo/bar buzz.txt
     *        ↑ this one!
     * ```
     * @param name name and flags pass flags pass 'foo'(sub command) or 'foo <bar>'(sub command & required argument) or 'foo [bar]'(sub command & optional argument) or 'foo <bar...>'(sub command & required variadic argument) or 'foo [bar...]'(sub command & optional variadic argument).
     * @class
     */
    function Command(name) {
        var _this = this;
        /**
         * @private
         */
        this._help = new option_1.default("-h, --help", "display help");
        /**
         * e.g. bar
         * @private
         */
        this._rest = [];
        /**
         * e.g.
         * ```
         *   git -p clone git@github.com:vvakame/dtsm.git
         *       ↑ this!
         * ```
         * @type {Array}
         */
        this.options = [];
        /**
         * e.g.
         * ```
         *   git -p clone git@github.com:vvakame/dtsm.git
         *          ↑ this!
         * ```
         * @type {Array}
         */
        this.subCommands = [];
        /**
         * parsed option values.
         * @type {any}
         */
        this.parsedOpts = {};
        /**
         * parsed option arguments.
         * @type {any}
         */
        this.parsedArgs = {};
        /**
         * unknown options.
         * @type {Array}
         */
        this.unknownOptions = [];
        var args = name.split(/\s+/);
        this.name = args.shift();
        var findOptional = false;
        var findVariadic = false;
        this.args = args.map(function (argStr) {
            if (findVariadic) {
                throw new Error("parameter can not placed after variadic parameter");
            }
            var arg = new argument_1.default(argStr);
            if (arg.required && findOptional) {
                throw new Error("required parameter is not placed after optional parameter");
            }
            if (!arg.required) {
                findOptional = true;
            }
            if (arg.variadic) {
                findVariadic = true;
            }
            return arg;
        });
        this._action = function () {
            process.stdout.write(_this.helpText() + "\n");
        };
    }
    /**
     * set description for this command.
     * @param desc
     * @returns {Command}
     * @method
     */
    Command.prototype.description = function (desc) {
        this._description = desc;
        return this;
    };
    /**
     * set usage for this command.
     * @param usage
     * @returns {Command}
     * @method
     */
    Command.prototype.usage = function (usage) {
        this._usage = usage;
        return this;
    };
    /**
     * add option for this command.
     * see {@link Option}.
     * @param flags
     * @param description
     * @param defaultValue
     * @returns {Command}
     */
    Command.prototype.option = function (flags, description, defaultValue) {
        var option = new option_1.default(flags, description, defaultValue);
        this.options.push(option);
        return this;
    };
    /**
     * allow unknown option.
     * by default, An error occured if unknown option is included.
     * @param flag
     * @returns {Command}
     */
    Command.prototype.allowUnknownOption = function (flag) {
        if (flag === void 0) { flag = true; }
        this._allowUnknownOption = flag;
        return this;
    };
    /**
     * add action at this command selected.
     * @param fn
     * @returns {Command}
     */
    Command.prototype.action = function (fn) {
        this._action = fn;
        return this;
    };
    /**
     * create sub command.
     * @param name
     * @returns {Command<Opt2, Arg2>} new command instance
     */
    Command.prototype.subCommand = function (name) {
        var command = new Command(name);
        command.parent = this;
        this.subCommands.push(command);
        return command;
    };
    /**
     * check arg is matches this command.
     * @param arg
     * @returns {boolean}
     */
    Command.prototype.is = function (arg) {
        return this.name === arg;
    };
    /**
     * add help this option.
     * in general case, use default help option.
     * @param flags
     * @param description
     * @returns {Command}
     */
    Command.prototype.help = function (flags, description) {
        this._help = new option_1.default(flags, description);
        return this;
    };
    /**
     * add show version option to this command.
     * @param version
     * @param flags
     * @param description
     * @returns {Command}
     */
    Command.prototype.version = function (version, flags, description) {
        if (description === void 0) { description = "output the version number"; }
        this._version = new option_1.default(flags, description);
        this._versionStr = version;
        return this;
    };
    /**
     * exec action of command.
     * this method MUST call after parse process.
     * @returns {Promise<{}>}
     */
    Command.prototype.exec = function () {
        return Promise.resolve(this._action(this.parsedOpts, this.parsedArgs, this._rest));
    };
    /**
     * parse argv.
     * @param argv
     * @returns {Promise<{}>}
     */
    Command.prototype.parse = function (argv) {
        var _this = this;
        return Promise
            .resolve(null)
            .then(function () {
            var rest = _this._parseRawArgs(argv);
            // resolve help action
            if (_this._args.some(function (arg) { return _this._help.is(arg); })) {
                // include help option. (help for this command
                process.stdout.write(_this.helpText() + "\n");
                process.exit(0);
                return Promise.resolve({});
            }
            var subCommand;
            if (_this.parent == null) {
                // only for top level (why? because I can't decide which is natural syntax between `foo help bar buzz` and `foo bar help buzz`.
                if (_this._rest.some(function (arg) { return _this._help.name() === arg; })) {
                    // include help sub command. (help for deeper level sub command
                    if (rest[0]) {
                        subCommand = _this.subCommands.filter(function (cmd) { return cmd.is(rest[0]); })[0];
                        if (subCommand) {
                            process.stdout.write(subCommand.helpText() + "\n");
                            process.exit(0);
                            return Promise.resolve({});
                        }
                    }
                }
            }
            // resolve version option
            if (_this._version && _this._args.some(function (arg) { return _this._version.is(arg); })) {
                process.stdout.write((_this._versionStr || "unknown") + "\n");
                process.exit(0);
                return Promise.resolve({});
            }
            if (rest[0]) {
                subCommand = _this.subCommands.filter(function (cmd) { return cmd.is(rest[0]); })[0];
                if (subCommand) {
                    return subCommand.parse(rest.slice(1));
                }
            }
            return _this.exec();
        });
    };
    /**
     * @returns {*}
     * @private
     */
    Command.prototype._getAncestorsAndMe = function () {
        if (!this.parent) {
            return [this];
        }
        else {
            return this.parent._getAncestorsAndMe().concat([this]);
        }
    };
    /**
     * @param args
     * @returns {string[]}
     * @private
     */
    Command.prototype._parseRawArgs = function (args) {
        args = args.slice(0);
        var target = [];
        var rest = [];
        var _loop_1 = function(i) {
            var arg = args[i];
            if (arg === "--") {
                // Honor option terminator
                target = target.concat(args.slice(i));
                return "break";
            }
            var cmd = this_1.subCommands.filter(function (cmd) { return cmd.is(arg); })[0];
            if (cmd) {
                rest = args.slice(i);
                return "break";
            }
            target.push(arg);
        };
        var this_1 = this;
        for (var i = 0; i < args.length; i++) {
            var state_1 = _loop_1(i);
            if (state_1 === "break") break;
        }
        this._rawArgs = target.slice(0);
        this._args = this._normalize(target);
        this._rest = this._parseOptions(this._args);
        var cmds = this._getAncestorsAndMe();
        var allowUnknownOption = cmds.reverse().map(function (cmd) { return cmd._allowUnknownOption; }).filter(function (allowUnknownOption) { return typeof allowUnknownOption !== "undefined"; })[0];
        if (this.unknownOptions.length !== 0 && !allowUnknownOption) {
            var errMsg = "unknown option";
            errMsg += this.unknownOptions.length === 1 ? " " : "s ";
            errMsg += this.unknownOptions.join(", ") + "\n";
            errMsg += this.helpText();
            throw new Error(errMsg);
        }
        if (this._matchSubCommand(rest)) {
            return rest;
        }
        this._rest = this._parseArgs(this._rest);
        return rest;
    };
    /**
     * @param rest
     * @returns {boolean}
     * @private
     */
    Command.prototype._matchSubCommand = function (rest) {
        if (rest == null || !rest[0]) {
            return false;
        }
        var subCommand = this.subCommands.filter(function (cmd) { return cmd.is(rest[0]); })[0];
        return !!subCommand;
    };
    /**
     * @param args
     * @returns {string[]}
     * @private
     */
    Command.prototype._parseOptions = function (args) {
        var _this = this;
        args = args.slice(0);
        var rest = [];
        var processedOptions = [];
        var _loop_2 = function() {
            var arg = args.shift();
            if (arg === "--") {
                rest = rest.concat(args);
                return "break";
            }
            var opt = this_2.options.filter(function (opt) { return opt.is(arg); })[0];
            if (!opt) {
                rest.push(arg);
                if (arg.indexOf("-") === 0 && !this_2._help.is(arg) && (!this_2._version || !this_2._version.is(arg))) {
                    this_2.unknownOptions.push(arg);
                }
                return "continue";
            }
            args = opt.parse(this_2.parsedOpts, [arg].concat(args));
            processedOptions.push(opt);
        };
        var this_2 = this;
        while (args.length !== 0) {
            var state_2 = _loop_2();
            if (state_2 === "break") break;
        }
        this.options
            .filter(function (opt) { return processedOptions.indexOf(opt) === -1; })
            .forEach(function (opt) {
            if (opt.required || opt.optional) {
                // string[]
                _this.parsedOpts[opt.name()] = _this.parsedOpts[opt.name()] || [];
                if (opt.defaultValue) {
                    _this.parsedOpts[opt.name()].push(opt.defaultValue);
                }
            }
            else {
                _this.parsedOpts[opt.name()] = opt.defaultValue;
            }
        });
        return rest;
    };
    /**
     * @param rest
     * @returns {string[]}
     * @private
     */
    Command.prototype._parseArgs = function (rest) {
        var _this = this;
        rest = rest.slice(0);
        this.args.forEach(function (argInfo) {
            rest = argInfo.parse(_this.parsedArgs, rest);
        });
        return rest;
    };
    /**
     * @param args
     * @returns {string[]}
     * @private
     */
    Command.prototype._normalize = function (args) {
        var result = [];
        var _loop_3 = function(i) {
            var arg = args[i];
            var lastOpt = void 0;
            if (0 < i) {
                lastOpt = this_3.options.filter(function (opt) { return opt.is(args[i - 1]); })[0];
            }
            if (arg === "--") {
                // Honor option terminator
                result = result.concat(args.slice(i));
                return "break";
            }
            else if (lastOpt && lastOpt.required) {
                result.push(arg);
            }
            else if (/^-[^-]/.test(arg)) {
                // expand combined short hand option. "-abc" to "-a -b -c"
                arg.slice(1).split("").forEach(function (c) { return result.push("-" + c); });
            }
            else if (/^--/.test(arg) && arg.indexOf("=") !== -1) {
                result.push(arg.slice(0, arg.indexOf("=")), arg.slice(arg.indexOf("=") + 1));
            }
            else {
                result.push(arg);
            }
        };
        var this_3 = this;
        for (var i = 0; i < args.length; i++) {
            var state_3 = _loop_3(i);
            if (state_3 === "break") break;
        }
        return result;
    };
    /**
     * generate help text.
     * @returns {string}
     */
    Command.prototype.helpText = function () {
        var result = "";
        // usage part
        result += "  Usage: ";
        if (this._usage != null) {
            result += this._usage;
        }
        else {
            result += this._getAncestorsAndMe().map(function (cmd) { return cmd.name; }).join(" ") + " ";
            if (this.options.length !== 0) {
                result += "[options] ";
            }
            if (this.subCommands.length !== 0) {
                result += "[command] ";
            }
            if (this.args.length !== 0) {
                result += "[--] ";
                result += this.args.map(function (arg) {
                    if (arg.required) {
                        return "<" + arg.name + (arg.variadic ? "..." : "") + ">";
                    }
                    else {
                        return "[" + arg.name + (arg.variadic ? "..." : "") + "]";
                    }
                }).join(" ");
            }
        }
        result += "\n\n";
        // options part
        if (this.options.length !== 0) {
            result += "  Options:\n\n";
            var optionsMaxLength_1 = utils.maxLength(this.options.map(function (opt) { return opt.flags; }));
            result += this.options.map(function (opt) {
                var result = "    ";
                result += utils.pad(opt.flags, optionsMaxLength_1);
                result += "  ";
                result += opt.description || "";
                result += "\n";
                return result;
            }).join("");
            result += "\n\n";
        }
        // sub commands part
        if (this.subCommands.length !== 0) {
            result += "  Commands:\n\n";
            var subCommandsMaxLength_1 = utils.maxLength(this.subCommands.map(function (cmd) { return cmd.name; }));
            result += this.subCommands.map(function (cmd) {
                var result = "    ";
                result += utils.pad(cmd.name, subCommandsMaxLength_1);
                result += "  ";
                result += cmd._description || "";
                result += "\n";
                return result;
            }).join("");
            result += "\n\n";
        }
        return result;
    };
    return Command;
}());
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = Command;
