// vscode/src/vs/platform/backup/common/backupIpc.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var BackupChannel = (function () {
    function BackupChannel(service) {
        this.service = service;
    }
    BackupChannel.prototype.call = function (command, arg) {
        switch (command) {
            case 'getBackupPath': return this.service.getBackupPath(arg);
        }
    };
    return BackupChannel;
}());
exports.BackupChannel = BackupChannel;
var BackupChannelClient = (function () {
    function BackupChannelClient(channel) {
        this.channel = channel;
    }
    BackupChannelClient.prototype.getBackupPath = function (windowId) {
        return this.channel.call('getBackupPath', windowId);
    };
    return BackupChannelClient;
}());
exports.BackupChannelClient = BackupChannelClient;
