diff --git a/toolchain/zig-wrapper.zig b/toolchain/zig-wrapper.zig
index 4a0fab4..8a740f7 100644
--- a/toolchain/zig-wrapper.zig
+++ b/toolchain/zig-wrapper.zig
@@ -328,7 +328,7 @@ test "zig-wrapper:parseArgs" {
     // not using testing.allocator, because parseArgs is designed to be used
     // with an arena.
     var gpa = std.heap.GeneralPurposeAllocator(.{}){};
-    var allocator = gpa.allocator();
+    const allocator = gpa.allocator();

     const tests = [_]struct {
         args: []const [:0]const u8,
@@ -428,7 +428,7 @@ test "zig-wrapper:parseArgs" {
             try tmp.dir.makePath(dir);

         var argv_it = TestArgIterator{ .argv = tt.args };
-        var res = try parseArgs(allocator, tmp.dir, &argv_it);
+        const res = try parseArgs(allocator, tmp.dir, &argv_it);

         switch (tt.want_result) {
             .err => |want_msg| try testing.expectEqualStrings(
