// vscode/src/vs/workbench/parts/git/common/noopGitService.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var git_1 = require('vs/workbench/parts/git/common/git');
var winjs_base_1 = require('vs/base/common/winjs.base');
var event_1 = require('vs/base/common/event');
var NoOpGitService = (function () {
    function NoOpGitService() {
        this._onOutput = new event_1.Emitter();
    }
    Object.defineProperty(NoOpGitService.prototype, "onOutput", {
        get: function () { return this._onOutput.event; },
        enumerable: true,
        configurable: true
    });
    NoOpGitService.prototype.getVersion = function () {
        return winjs_base_1.TPromise.as(null);
    };
    NoOpGitService.prototype.serviceState = function () {
        return winjs_base_1.TPromise.as(git_1.RawServiceState.OK);
    };
    NoOpGitService.prototype.statusCount = function () {
        return winjs_base_1.TPromise.as(0);
    };
    NoOpGitService.prototype.status = function () {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.init = function () {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.add = function (filesPaths) {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.stage = function (filePath, content) {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.branch = function (name, checkout) {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.checkout = function (treeish, filePaths) {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.clean = function (filePaths) {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.undo = function () {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.reset = function (treeish, hard) {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.revertFiles = function (treeish, filePaths) {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.fetch = function () {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.pull = function (rebase) {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.push = function () {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.sync = function () {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.commit = function (message, amend, stage, signoff) {
        return winjs_base_1.TPromise.as(NoOpGitService.STATUS);
    };
    NoOpGitService.prototype.detectMimetypes = function (path, treeish) {
        return winjs_base_1.TPromise.as([]);
    };
    NoOpGitService.prototype.show = function (path, treeish) {
        return winjs_base_1.TPromise.as(null);
    };
    NoOpGitService.prototype.getCommitTemplate = function () {
        return winjs_base_1.TPromise.as(null);
    };
    NoOpGitService.prototype.getCommit = function (ref) {
        return winjs_base_1.TPromise.as(null);
    };
    NoOpGitService.STATUS = {
        repositoryRoot: null,
        state: git_1.ServiceState.NotAWorkspace,
        status: [],
        HEAD: null,
        refs: [],
        remotes: []
    };
    return NoOpGitService;
}());
exports.NoOpGitService = NoOpGitService;
