// vscode/src/vs/workbench/parts/files/common/editors/fileEditorInput.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var winjs_base_1 = require("vs/base/common/winjs.base");
var paths = require("vs/base/common/paths");
var labels = require("vs/base/common/labels");
var binaryEditorModel_1 = require("vs/workbench/common/editor/binaryEditorModel");
var files_1 = require("vs/platform/files/common/files");
var files_2 = require("vs/workbench/parts/files/common/files");
var textfiles_1 = require("vs/workbench/services/textfile/common/textfiles");
var workspace_1 = require("vs/platform/workspace/common/workspace");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var lifecycle_1 = require("vs/base/common/lifecycle");
/**
 * A file editor input is the input type for the file editor of file system resources.
 */
var FileEditorInput = (function (_super) {
    __extends(FileEditorInput, _super);
    /**
     * An editor input who's contents are retrieved from file services.
     */
    function FileEditorInput(resource, preferredEncoding, instantiationService, contextService, textFileService) {
        var _this = _super.call(this) || this;
        _this.instantiationService = instantiationService;
        _this.contextService = contextService;
        _this.textFileService = textFileService;
        _this.toUnbind = [];
        if (resource) {
            _this.setResource(resource);
            _this.preferredEncoding = preferredEncoding;
        }
        _this.registerListeners();
        return _this;
    }
    FileEditorInput.prototype.registerListeners = function () {
        var _this = this;
        // Model changes
        this.toUnbind.push(this.textFileService.models.onModelDirty(function (e) { return _this.onDirtyStateChange(e); }));
        this.toUnbind.push(this.textFileService.models.onModelSaveError(function (e) { return _this.onDirtyStateChange(e); }));
        this.toUnbind.push(this.textFileService.models.onModelSaved(function (e) { return _this.onDirtyStateChange(e); }));
        this.toUnbind.push(this.textFileService.models.onModelReverted(function (e) { return _this.onDirtyStateChange(e); }));
    };
    FileEditorInput.prototype.onDirtyStateChange = function (e) {
        if (e.resource.toString() === this.resource.toString()) {
            this._onDidChangeDirty.fire();
        }
    };
    FileEditorInput.prototype.setResource = function (resource) {
        if (resource.scheme !== 'file') {
            throw new Error('FileEditorInput can only handle file:// resources.');
        }
        this.resource = resource;
        // Reset resource dependent properties
        this.name = null;
        this.description = null;
        this.verboseDescription = null;
    };
    FileEditorInput.prototype.getResource = function () {
        return this.resource;
    };
    FileEditorInput.prototype.setPreferredEncoding = function (encoding) {
        this.preferredEncoding = encoding;
    };
    FileEditorInput.prototype.getEncoding = function () {
        var textModel = this.textFileService.models.get(this.resource);
        if (textModel) {
            return textModel.getEncoding();
        }
        return this.preferredEncoding;
    };
    FileEditorInput.prototype.setEncoding = function (encoding, mode) {
        this.preferredEncoding = encoding;
        var textModel = this.textFileService.models.get(this.resource);
        if (textModel) {
            textModel.setEncoding(encoding, mode);
        }
    };
    FileEditorInput.prototype.setForceOpenAsBinary = function () {
        this.forceOpenAsBinary = true;
    };
    FileEditorInput.prototype.getTypeId = function () {
        return files_2.FILE_EDITOR_INPUT_ID;
    };
    FileEditorInput.prototype.getName = function () {
        if (!this.name) {
            this.name = paths.basename(this.resource.fsPath);
        }
        return this.name;
    };
    FileEditorInput.prototype.getDescription = function (verbose) {
        if (!verbose) {
            if (!this.description) {
                this.description = labels.getPathLabel(paths.dirname(this.resource.fsPath), this.contextService);
            }
            return this.description;
        }
        if (!this.verboseDescription) {
            this.verboseDescription = labels.getPathLabel(this.resource.fsPath);
        }
        return this.verboseDescription;
    };
    FileEditorInput.prototype.isDirty = function () {
        var model = this.textFileService.models.get(this.resource);
        if (!model) {
            return false;
        }
        var state = model.getState();
        if (state === textfiles_1.ModelState.CONFLICT || state === textfiles_1.ModelState.ERROR) {
            return true; // always indicate dirty state if we are in conflict or error state
        }
        if (this.textFileService.getAutoSaveMode() === textfiles_1.AutoSaveMode.AFTER_SHORT_DELAY) {
            return false; // fast auto save enabled so we do not declare dirty
        }
        return model.isDirty();
    };
    FileEditorInput.prototype.confirmSave = function () {
        return this.textFileService.confirmSave([this.resource]);
    };
    FileEditorInput.prototype.save = function () {
        return this.textFileService.save(this.resource);
    };
    FileEditorInput.prototype.revert = function () {
        return this.textFileService.revert(this.resource);
    };
    FileEditorInput.prototype.getPreferredEditorId = function (candidates) {
        return this.forceOpenAsBinary ? files_2.BINARY_FILE_EDITOR_ID : files_2.TEXT_FILE_EDITOR_ID;
    };
    FileEditorInput.prototype.resolve = function (refresh) {
        var _this = this;
        return this.textFileService.models.loadOrCreate(this.resource, this.preferredEncoding, refresh).then(null, function (error) {
            // In case of an error that indicates that the file is binary or too large, just return with the binary editor model
            if (error.fileOperationResult === files_1.FileOperationResult.FILE_IS_BINARY || error.fileOperationResult === files_1.FileOperationResult.FILE_TOO_LARGE) {
                return _this.instantiationService.createInstance(binaryEditorModel_1.BinaryEditorModel, _this.resource, _this.getName()).load();
            }
            // Bubble any other error up
            return winjs_base_1.TPromise.wrapError(error);
        });
    };
    FileEditorInput.prototype.dispose = function () {
        // Listeners
        lifecycle_1.dispose(this.toUnbind);
        _super.prototype.dispose.call(this);
    };
    FileEditorInput.prototype.matches = function (otherInput) {
        if (_super.prototype.matches.call(this, otherInput) === true) {
            return true;
        }
        if (otherInput) {
            return otherInput instanceof FileEditorInput && otherInput.resource.toString() === this.resource.toString();
        }
        return false;
    };
    return FileEditorInput;
}(files_2.FileEditorInput));
FileEditorInput = __decorate([
    __param(2, instantiation_1.IInstantiationService),
    __param(3, workspace_1.IWorkspaceContextService),
    __param(4, textfiles_1.ITextFileService)
], FileEditorInput);
exports.FileEditorInput = FileEditorInput;
