// vscode/src/vs/workbench/parts/preferences/browser/preferencesWidgets.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var nls_1 = require("vs/nls");
var DOM = require("vs/base/browser/dom");
var winjs_base_1 = require("vs/base/common/winjs.base");
var widget_1 = require("vs/base/browser/ui/widget");
var event_1 = require("vs/base/common/event");
var editorBrowser_1 = require("vs/editor/browser/editorBrowser");
var inputBox_1 = require("vs/base/browser/ui/inputbox/inputBox");
var instantiation_1 = require("vs/platform/instantiation/common/instantiation");
var contextView_1 = require("vs/platform/contextview/browser/contextView");
var keybinding_1 = require("vs/platform/keybinding/common/keybinding");
var workspace_1 = require("vs/platform/workspace/common/workspace");
var configurationEditing_1 = require("vs/workbench/services/configuration/common/configurationEditing");
var actionbar_1 = require("vs/base/browser/ui/actionbar/actionbar");
var actions_1 = require("vs/base/common/actions");
var SettingsGroupTitleWidget = (function (_super) {
    __extends(SettingsGroupTitleWidget, _super);
    function SettingsGroupTitleWidget(editor, settingsGroup) {
        var _this = _super.call(this) || this;
        _this.editor = editor;
        _this.settingsGroup = settingsGroup;
        _this._onToggled = _this._register(new event_1.Emitter());
        _this.onToggled = _this._onToggled.event;
        _this.create();
        _this._register(_this.editor.onDidChangeConfiguration(function () { return _this.layout(); }));
        _this._register(_this.editor.onDidLayoutChange(function () { return _this.layout(); }));
        _this._register(_this.editor.onDidChangeCursorPosition(function (e) { return _this.onCursorChange(e); }));
        return _this;
    }
    Object.defineProperty(SettingsGroupTitleWidget.prototype, "domNode", {
        get: function () {
            return this._domNode;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SettingsGroupTitleWidget.prototype, "heightInLines", {
        get: function () {
            return 1.5;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SettingsGroupTitleWidget.prototype, "afterLineNumber", {
        get: function () {
            return this._afterLineNumber;
        },
        enumerable: true,
        configurable: true
    });
    SettingsGroupTitleWidget.prototype.create = function () {
        var _this = this;
        this._domNode = DOM.$('.settings-group-title-widget');
        this.titleContainer = DOM.append(this._domNode, DOM.$('.title-container'));
        this.titleContainer.tabIndex = 0;
        this.onclick(this.titleContainer, function () { return _this.toggle(); });
        this.onkeydown(this.titleContainer, function (e) { return _this.onKeyDown(e); });
        var focusTracker = this._register(DOM.trackFocus(this.titleContainer));
        focusTracker.addFocusListener(function () { return _this.toggleFocus(true); });
        focusTracker.addBlurListener(function () { return _this.toggleFocus(false); });
        this.icon = DOM.append(this.titleContainer, DOM.$('.expand-collapse-icon'));
        this.title = DOM.append(this.titleContainer, DOM.$('.title'));
        this.title.textContent = this.settingsGroup.title + (" (" + this.settingsGroup.sections.reduce(function (count, section) { return count + section.settings.length; }, 0) + ")");
        this.layout();
    };
    SettingsGroupTitleWidget.prototype.render = function () {
        var _this = this;
        this._afterLineNumber = this.settingsGroup.range.startLineNumber - 2;
        this.editor.changeViewZones(function (accessor) {
            _this.id = accessor.addZone(_this);
            _this.layout();
        });
    };
    SettingsGroupTitleWidget.prototype.toggleCollapse = function (collapse) {
        DOM.toggleClass(this.titleContainer, 'collapsed', collapse);
    };
    SettingsGroupTitleWidget.prototype.toggleFocus = function (focus) {
        DOM.toggleClass(this.titleContainer, 'focused', focus);
    };
    SettingsGroupTitleWidget.prototype.isCollapsed = function () {
        return DOM.hasClass(this.titleContainer, 'collapsed');
    };
    SettingsGroupTitleWidget.prototype.layout = function () {
        var configuration = this.editor.getConfiguration();
        var layoutInfo = this.editor.getLayoutInfo();
        this._domNode.style.width = layoutInfo.contentWidth - layoutInfo.verticalScrollbarWidth + 'px';
        this.titleContainer.style.lineHeight = configuration.lineHeight + 3 + 'px';
        this.titleContainer.style.fontSize = configuration.fontInfo.fontSize + 'px';
        var iconSize = this.getIconSize();
        this.icon.style.height = iconSize + "px";
        this.icon.style.width = iconSize + "px";
    };
    SettingsGroupTitleWidget.prototype.getIconSize = function () {
        var fontSize = this.editor.getConfiguration().fontInfo.fontSize;
        return fontSize > 8 ? Math.max(fontSize, 16) : 12;
    };
    SettingsGroupTitleWidget.prototype.onKeyDown = function (keyboardEvent) {
        switch (keyboardEvent.keyCode) {
            case 3 /* Enter */:
            case 10 /* Space */:
                this.toggle();
                break;
            case 15 /* LeftArrow */:
                this.collapse(true);
                break;
            case 17 /* RightArrow */:
                this.collapse(false);
                break;
            case 16 /* UpArrow */:
                if (this.settingsGroup.range.startLineNumber - 3 !== 1) {
                    this.editor.focus();
                    var lineNumber_1 = this.settingsGroup.range.startLineNumber - 2;
                    this.editor.setPosition({ lineNumber: lineNumber_1, column: this.editor.getModel().getLineMinColumn(lineNumber_1) });
                }
                break;
            case 18 /* DownArrow */:
                var lineNumber = this.isCollapsed() ? this.settingsGroup.range.startLineNumber : this.settingsGroup.range.startLineNumber - 1;
                this.editor.focus();
                this.editor.setPosition({ lineNumber: lineNumber, column: this.editor.getModel().getLineMinColumn(lineNumber) });
                break;
        }
    };
    SettingsGroupTitleWidget.prototype.toggle = function () {
        this.collapse(!this.isCollapsed());
    };
    SettingsGroupTitleWidget.prototype.collapse = function (collapse) {
        if (collapse !== this.isCollapsed()) {
            DOM.toggleClass(this.titleContainer, 'collapsed', collapse);
            this._onToggled.fire(collapse);
        }
    };
    SettingsGroupTitleWidget.prototype.onCursorChange = function (e) {
        if (e.source !== 'mouse' && this.focusTitle(e.position)) {
            this.titleContainer.focus();
        }
    };
    SettingsGroupTitleWidget.prototype.focusTitle = function (currentPosition) {
        var previousPosition = this.previousPosition;
        this.previousPosition = currentPosition;
        if (!previousPosition) {
            return false;
        }
        if (previousPosition.lineNumber === currentPosition.lineNumber) {
            return false;
        }
        if (currentPosition.lineNumber === this.settingsGroup.range.startLineNumber - 1 || currentPosition.lineNumber === this.settingsGroup.range.startLineNumber - 2) {
            return true;
        }
        if (this.isCollapsed() && currentPosition.lineNumber === this.settingsGroup.range.endLineNumber) {
            return true;
        }
        return false;
    };
    SettingsGroupTitleWidget.prototype.dispose = function () {
        var _this = this;
        this.editor.changeViewZones(function (accessor) {
            accessor.removeZone(_this.id);
        });
        _super.prototype.dispose.call(this);
    };
    return SettingsGroupTitleWidget;
}(widget_1.Widget));
exports.SettingsGroupTitleWidget = SettingsGroupTitleWidget;
var SettingsTabsWidget = (function (_super) {
    __extends(SettingsTabsWidget, _super);
    function SettingsTabsWidget(parent, contextService) {
        var _this = _super.call(this) || this;
        _this.contextService = contextService;
        _this._onSwitch = new event_1.Emitter();
        _this.onSwitch = _this._onSwitch.event;
        _this.create(parent);
        return _this;
    }
    SettingsTabsWidget.prototype.create = function (parent) {
        var _this = this;
        var settingsTabsWidget = DOM.append(parent, DOM.$('.settings-tabs-widget'));
        this.settingsSwitcherBar = this._register(new actionbar_1.ActionBar(settingsTabsWidget, {
            orientation: actionbar_1.ActionsOrientation.HORIZONTAL,
            ariaLabel: nls_1.localize('settingsSwitcherBarAriaLabel', "Settings Switcher"),
            animated: false
        }));
        this.userSettings = new actions_1.Action('userSettings', nls_1.localize('userSettings', "User Settings"), '.settings-tab', true, function () { return _this.onClick(_this.userSettings); });
        this.userSettings.tooltip = this.userSettings.label;
        this.workspaceSettings = new actions_1.Action('workspaceSettings', nls_1.localize('workspaceSettings', "Workspace Settings"), '.settings-tab', this.contextService.hasWorkspace(), function () { return _this.onClick(_this.workspaceSettings); });
        this.workspaceSettings.tooltip = this.workspaceSettings.label;
        this.settingsSwitcherBar.push([this.userSettings, this.workspaceSettings]);
    };
    SettingsTabsWidget.prototype.show = function (configurationTarget) {
        this.userSettings.checked = configurationEditing_1.ConfigurationTarget.USER === configurationTarget;
        this.workspaceSettings.checked = configurationEditing_1.ConfigurationTarget.WORKSPACE === configurationTarget;
    };
    SettingsTabsWidget.prototype.onClick = function (action) {
        if (!action.checked) {
            this._onSwitch.fire();
        }
        return winjs_base_1.TPromise.as(null);
    };
    return SettingsTabsWidget;
}(widget_1.Widget));
SettingsTabsWidget = __decorate([
    __param(1, workspace_1.IWorkspaceContextService)
], SettingsTabsWidget);
exports.SettingsTabsWidget = SettingsTabsWidget;
var SearchWidget = (function (_super) {
    __extends(SearchWidget, _super);
    function SearchWidget(parent, contextViewService, contextMenuService, instantiationService) {
        var _this = _super.call(this) || this;
        _this.contextViewService = contextViewService;
        _this.contextMenuService = contextMenuService;
        _this.instantiationService = instantiationService;
        _this._onDidChange = _this._register(new event_1.Emitter());
        _this.onDidChange = _this._onDidChange.event;
        _this._onEnter = _this._register(new event_1.Emitter());
        _this.onEnter = _this._onEnter.event;
        _this.create(parent);
        return _this;
    }
    SearchWidget.prototype.create = function (parent) {
        this.domNode = DOM.append(parent, DOM.$('div.settings-header-widget'));
        this.createSearchContainer(DOM.append(this.domNode, DOM.$('div.settings-search-container')));
        this.countElement = DOM.append(this.domNode, DOM.$('.settings-count-widget'));
        this.inputBox.inputElement.setAttribute('aria-live', 'assertive');
    };
    SearchWidget.prototype.createSearchContainer = function (searchContainer) {
        var _this = this;
        this.searchContainer = searchContainer;
        var searchInput = DOM.append(this.searchContainer, DOM.$('div.settings-search-input'));
        this.inputBox = this._register(new inputBox_1.InputBox(searchInput, this.contextViewService, {
            ariaLabel: nls_1.localize('SearchSettingsWidget.AriaLabel', "Search settings"),
            placeholder: nls_1.localize('SearchSettingsWidget.Placeholder', "Search Settings")
        }));
        this.inputBox.onDidChange(function (value) { return _this._onDidChange.fire(value); });
        this.onkeyup(this.inputBox.inputElement, function (e) { return _this._onKeyUp(e); });
    };
    SearchWidget.prototype.showMessage = function (message, count) {
        this.countElement.textContent = message;
        this.inputBox.inputElement.setAttribute('aria-label', message);
        DOM.toggleClass(this.countElement, 'no-results', count === 0);
        this.inputBox.inputElement.style.paddingRight = DOM.getTotalWidth(this.countElement) + 20 + 'px';
    };
    SearchWidget.prototype.layout = function (dimension) {
        if (dimension.width < 400) {
            DOM.addClass(this.countElement, 'hide');
            this.inputBox.inputElement.style.paddingRight = '0px';
        }
        else {
            DOM.removeClass(this.countElement, 'hide');
            this.inputBox.inputElement.style.paddingRight = DOM.getTotalWidth(this.countElement) + 20 + 'px';
        }
    };
    SearchWidget.prototype.focus = function () {
        this.inputBox.focus();
    };
    SearchWidget.prototype.clear = function () {
        this.inputBox.value = '';
    };
    SearchWidget.prototype.value = function () {
        return this.inputBox.value;
    };
    SearchWidget.prototype._onKeyUp = function (keyboardEvent) {
        var handled = false;
        switch (keyboardEvent.keyCode) {
            case 3 /* Enter */:
                this._onEnter.fire();
                handled = true;
                break;
            case 9 /* Escape */:
                this.clear();
                handled = true;
                break;
        }
        if (handled) {
            keyboardEvent.preventDefault();
            keyboardEvent.stopPropagation();
        }
    };
    return SearchWidget;
}(widget_1.Widget));
SearchWidget = __decorate([
    __param(1, contextView_1.IContextViewService),
    __param(2, contextView_1.IContextMenuService),
    __param(3, instantiation_1.IInstantiationService)
], SearchWidget);
exports.SearchWidget = SearchWidget;
var FloatingClickWidget = (function (_super) {
    __extends(FloatingClickWidget, _super);
    function FloatingClickWidget(editor, label, keyBindingAction, keybindingService) {
        var _this = _super.call(this) || this;
        _this.editor = editor;
        _this.label = label;
        _this.keyBindingAction = keyBindingAction;
        _this._onClick = _this._register(new event_1.Emitter());
        _this.onClick = _this._onClick.event;
        if (keyBindingAction) {
            var keybinding = keybindingService.lookupKeybinding(keyBindingAction);
            if (keybinding) {
                _this.label += ' (' + keybinding.getLabel() + ')';
            }
        }
        return _this;
    }
    FloatingClickWidget.prototype.render = function () {
        var _this = this;
        this._domNode = DOM.$('.floating-click-widget');
        DOM.append(this._domNode, DOM.$('')).textContent = this.label;
        this.onclick(this._domNode, function (e) { return _this._onClick.fire(); });
        this.editor.addOverlayWidget(this);
    };
    FloatingClickWidget.prototype.dispose = function () {
        this.editor.removeOverlayWidget(this);
        _super.prototype.dispose.call(this);
    };
    FloatingClickWidget.prototype.getId = function () {
        return 'editor.overlayWidget.floatingClickWidget';
    };
    FloatingClickWidget.prototype.getDomNode = function () {
        return this._domNode;
    };
    FloatingClickWidget.prototype.getPosition = function () {
        return {
            preference: editorBrowser_1.OverlayWidgetPositionPreference.BOTTOM_RIGHT_CORNER
        };
    };
    return FloatingClickWidget;
}(widget_1.Widget));
FloatingClickWidget = __decorate([
    __param(3, keybinding_1.IKeybindingService)
], FloatingClickWidget);
exports.FloatingClickWidget = FloatingClickWidget;
var EditPreferenceWidget = (function (_super) {
    __extends(EditPreferenceWidget, _super);
    function EditPreferenceWidget(editor, contextMenuService) {
        var _this = _super.call(this) || this;
        _this.editor = editor;
        _this._onClick = new event_1.Emitter();
        _this._onMouseOver = new event_1.Emitter();
        _this._id = 'preferences.editPreferenceWidget' + EditPreferenceWidget.counter++;
        _this.editor.addOverlayWidget(_this);
        _this._register(_this.editor.onDidScrollChange(function () {
            if (_this._visible) {
                _this._layout();
            }
        }));
        return _this;
    }
    Object.defineProperty(EditPreferenceWidget.prototype, "onClick", {
        get: function () { return this._onClick.event; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(EditPreferenceWidget.prototype, "onMouseOver", {
        get: function () { return this._onMouseOver.event; },
        enumerable: true,
        configurable: true
    });
    EditPreferenceWidget.prototype.dispose = function () {
        this.editor.removeOverlayWidget(this);
        _super.prototype.dispose.call(this);
    };
    EditPreferenceWidget.prototype.getId = function () {
        return this._id;
    };
    EditPreferenceWidget.prototype.getDomNode = function () {
        var _this = this;
        if (!this._domNode) {
            this._domNode = document.createElement('div');
            this._domNode.style.width = '20px';
            this._domNode.style.height = '20px';
            this._domNode.className = 'edit-preferences-widget hidden';
            this.onclick(this._domNode, function (e) { return _this._onClick.fire(); });
            this.onmouseover(this._domNode, function (e) { return _this._onMouseOver.fire(); });
        }
        return this._domNode;
    };
    EditPreferenceWidget.prototype.getPosition = function () {
        return null;
    };
    EditPreferenceWidget.prototype.getLine = function () {
        return this._line;
    };
    EditPreferenceWidget.prototype.show = function (line, preferences) {
        this._preferences = preferences;
        if (!this._visible || this._line !== line) {
            this._line = line;
            this._visible = true;
            this._layout();
        }
    };
    Object.defineProperty(EditPreferenceWidget.prototype, "preferences", {
        get: function () {
            return this._preferences;
        },
        enumerable: true,
        configurable: true
    });
    EditPreferenceWidget.prototype.hide = function () {
        if (this._visible) {
            this._visible = false;
            this._domNode.classList.add('hidden');
        }
    };
    EditPreferenceWidget.prototype._layout = function () {
        var topForLineNumber = this.editor.getTopForLineNumber(this._line);
        var editorScrollTop = this.editor.getScrollTop();
        this._domNode.style.top = topForLineNumber - editorScrollTop - 2 + "px";
        this._domNode.style.left = '0px';
        this._domNode.classList.remove('hidden');
    };
    return EditPreferenceWidget;
}(widget_1.Widget));
EditPreferenceWidget.counter = 1;
EditPreferenceWidget = __decorate([
    __param(1, contextView_1.IContextMenuService)
], EditPreferenceWidget);
exports.EditPreferenceWidget = EditPreferenceWidget;
