// vscode/src/vs/platform/storage/common/storage.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
var instantiation_1 = require('vs/platform/instantiation/common/instantiation');
exports.ID = 'storageService';
exports.IStorageService = instantiation_1.createDecorator(exports.ID);
(function (StorageScope) {
    /**
     * The stored data will be scoped to all workspaces of this domain.
     */
    StorageScope[StorageScope["GLOBAL"] = 0] = "GLOBAL";
    /**
     * The stored data will be scoped to the current workspace.
     */
    StorageScope[StorageScope["WORKSPACE"] = 1] = "WORKSPACE";
})(exports.StorageScope || (exports.StorageScope = {}));
var StorageScope = exports.StorageScope;
exports.NullStorageService = {
    _serviceBrand: undefined,
    store: function () { return undefined; },
    swap: function () { return undefined; },
    remove: function () { return undefined; },
    get: function (a, b, defaultValue) { return defaultValue; },
    getInteger: function (a, b, defaultValue) { return defaultValue; },
    getBoolean: function (a, b, defaultValue) { return defaultValue; },
};
